{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9807e3a4-8edb-46a2-ad4c-88777af9fadc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, glob\n",
    "import pandas as pd\n",
    "from io import StringIO\n",
    "import numpy as np\n",
    "import pysam\n",
    "import pybedtools \n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "pybedtools.helpers.set_tempdir(\"/data/projects/temp\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2505ee28-f7a7-474a-a4b8-755989f3f70e",
   "metadata": {},
   "outputs": [],
   "source": [
    "non_coding_region = \"donor\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6c3ba47b-0edf-43de-bab5-bc0edd7f91b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '/data/projects/DNABERT_snv/Manuscript_11_2023/DBSNP_output/Intersected_data/{}/all_{}_intesected_data.tsv'.format(non_coding_region, non_coding_region)\n",
    "output_path = \"/data/projects/DNABERT_snv/Manuscript_11_2023/DBSNP_output/DNABERT_data/{}\".format(non_coding_region) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "aafe0877-b7be-4e5e-a73f-00dab9627b51",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(output_path):\n",
    "    os.makedirs(output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cc89bac5-1528-4192-95d4-77bda6a6e1dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def seq2kmer(seq):\n",
    "    \"\"\"\n",
    "    Convert original sequence to kmers\n",
    "    \n",
    "    Arguments:\n",
    "    seq -- str, original sequence.\n",
    "    k -- int, kmer of length k specified.\n",
    "    \n",
    "    Returns:\n",
    "    kmers -- str, kmers separated by space\n",
    "    \"\"\"\n",
    "    k=6\n",
    "    kmer = [seq[x:x+k] for x in range(len(seq)+1-k)]\n",
    "    kmers = \" \".join(kmer)\n",
    "    return kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "19aaa7a0-7309-4e97-a93e-dded3164e07f",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = SeqIO.to_dict(SeqIO.parse(\"/data/projects/Resources/Gencode_genome_annotation/GRCh38.primary_assembly.genome.fa\", \"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3bcc2b33-f043-4016-9aba-79948c239d76",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(37045469, 20)\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(data_path, sep= '\\t')\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0ded228f-2999-482f-9e5a-197b732e0cfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(43537490, 11)\n",
      "(43504528, 11)\n"
     ]
    }
   ],
   "source": [
    "# df = pd.read_csv(data_path, sep= '\\t')\n",
    "# print(df.shape)\n",
    "\n",
    "# df = df[df['No_of_variations']<3].reset_index(drop=True)\n",
    "# print(df.shape)\n",
    "# df = df.loc[df['Alternative_Allele']== '.'].reset_index(drop=True)\n",
    "# df\n",
    "#df= df[df['strand']==\"-\"]\n",
    "\n",
    "\n",
    "data= []\n",
    "#df = df.head(10)\n",
    "# Iterate over DataFrame rows\n",
    "for idx, row in df.iterrows():\n",
    "    try:\n",
    "        #print(row)Reference_Allele\tAlternative_Allele\n",
    "        alts = row['Alternative_Nucleotides'].rstrip(',').split(',')\n",
    "        ref_seq = str(genome[row['chr_name']].seq[ row['{}_start'.format(non_coding_region)]:row['{}_end'.format(non_coding_region)]])\n",
    "\n",
    "        # Calculate the variant position relative to the fetched sequence\n",
    "        variant_pos_start = row['Variant_start'] - row['{}_start'.format(non_coding_region)]\n",
    "        variant_pos_end = row['Variant_end']  - row['{}_start'.format(non_coding_region)]\n",
    "        #print(variant_pos_start, variant_pos_end, alts,  row['Reference_Nucleotide'])\n",
    "        #print(ref_seq , \"###\")\n",
    "        # if row['strand'] == '-':\n",
    "        #     ref_seq = str(Seq(ref_seq).reverse_complement())\n",
    "        #print(ref_seq)\n",
    "\n",
    "\n",
    "        #Replace the reference nucleotide with the alternate nucleotide to get the alternate sequence\n",
    "        for alt in alts:\n",
    "            if alt == \"\":  # Deletion\n",
    "                # Fetch additional bases from genome to maintain sequence length\n",
    "                extra_bases = str(genome[row['chr_name']].seq[row['{}_end'.format(non_coding_region)]:row['{}_end'.format(non_coding_region)] + len(row['Reference_Nucleotide'])])\n",
    "                #print(len(row['Reference_Nucleotide']))\n",
    "                alt_seq = ref_seq[:variant_pos_start] + ref_seq[variant_pos_end:] + extra_bases\n",
    "            else:  # SNPs, Insertions, and partial Deletions\n",
    "                extra_bases = str(genome[row['chr_name']].seq[row['{}_end'.format(non_coding_region)]:row['{}_end'.format(non_coding_region)] + len(row['Reference_Nucleotide']) - len(alt)])\n",
    "                #print(extra_bases)\n",
    "                if(variant_pos_start>=0):\n",
    "                    alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:] + extra_bases\n",
    "                else:\n",
    "                    continue\n",
    "            #print(alt_seq)\n",
    "            #Append to the list as a dictionary\n",
    "            data.append({\n",
    "                'chr': row['chr_name'],\n",
    "                'strand': row['strand'],\n",
    "                '{}_coordinates'.format(non_coding_region): str(row['{}_start'.format(non_coding_region)])+'-'+str(row['{}_end'.format(non_coding_region)]),\n",
    "                'Ensemble_Transcript_ID': row['ENSEMBL_Transcript_ID'],\n",
    "                'RS_ID': row['RS_ID'],\n",
    "                'variant_start':  row['Variant_start'],\n",
    "                'variant_end':  row['Variant_end'],\n",
    "                'ref_neucleotide': row['Reference_Nucleotide'],\n",
    "                'alternative_neucleotide': alt,\n",
    "                'reference_seq': ref_seq,\n",
    "                'alt_seq': alt_seq\n",
    "            })\n",
    "            \n",
    "    except KeyError:\n",
    "        print(\"NA\")\n",
    "        # # If sequence is not present, append a default value\n",
    "        # ref_sequences.append(\"NA\")\n",
    "        # alt_sequences.append(\"NA\")\n",
    "        \n",
    "new_df = pd.DataFrame(data)\n",
    "print(new_df.shape)\n",
    "new_df = new_df.drop_duplicates().reset_index(drop=True)\n",
    "print(new_df.shape)\n",
    "merged_list = list(zip(new_df['reference_seq'], new_df['alt_seq']))\n",
    "merged_list = [item.upper() for tup in merged_list for item in tup]\n",
    "#print(merged_list)\n",
    "kmer_lst = list(map(seq2kmer, merged_list))\n",
    "df_kmer = pd.DataFrame(kmer_lst, columns=['Sequence'])\n",
    "df_kmer['Label'] = np.random.choice([0, 1], size=len(df_kmer))\n",
    "\n",
    "new_df.to_csv(output_path + \"/all_data_new.tsv\", sep=\"\\t\", index= False)\n",
    "df_kmer.to_csv(output_path + \"/dev.tsv\", sep=\"\\t\", index= False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e14368ba-5b1a-4f81-8e07-10b8068121fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>donor_coordinates</th>\n",
       "      <th>Ensemble_Transcript_ID</th>\n",
       "      <th>RS_ID</th>\n",
       "      <th>variant_start</th>\n",
       "      <th>variant_end</th>\n",
       "      <th>ref_neucleotide</th>\n",
       "      <th>alternative_neucleotide</th>\n",
       "      <th>reference_seq</th>\n",
       "      <th>alt_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>64315231-64315310</td>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>rs1484286606</td>\n",
       "      <td>64315233</td>\n",
       "      <td>64315245</td>\n",
       "      <td>TATGTTCTACTA</td>\n",
       "      <td>TA</td>\n",
       "      <td>AGTATGTTCTACTAAAGGCCTTGGCCCTTGCCAATTCAGGTGAGTC...</td>\n",
       "      <td>AGTAAAGGCCTTGGCCCTTGCCAATTCAGGTGAGTCTGGGGCAGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>64315231-64315310</td>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>rs750270682</td>\n",
       "      <td>64315250</td>\n",
       "      <td>64315251</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>AGTATGTTCTACTAAAGGCCTTGGCCCTTGCCAATTCAGGTGAGTC...</td>\n",
       "      <td>AGTATGTTCTACTAAAGGCTTTGGCCCTTGCCAATTCAGGTGAGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>64315231-64315310</td>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>rs2035223711</td>\n",
       "      <td>64315259</td>\n",
       "      <td>64315260</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>AGTATGTTCTACTAAAGGCCTTGGCCCTTGCCAATTCAGGTGAGTC...</td>\n",
       "      <td>AGTATGTTCTACTAAAGGCCTTGGCCCTCGCCAATTCAGGTGAGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>64315231-64315310</td>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>rs2035223790</td>\n",
       "      <td>64315260</td>\n",
       "      <td>64315261</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>AGTATGTTCTACTAAAGGCCTTGGCCCTTGCCAATTCAGGTGAGTC...</td>\n",
       "      <td>AGTATGTTCTACTAAAGGCCTTGGCCCTTACCAATTCAGGTGAGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>64315231-64315310</td>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>rs1361791105</td>\n",
       "      <td>64315268</td>\n",
       "      <td>64315269</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>AGTATGTTCTACTAAAGGCCTTGGCCCTTGCCAATTCAGGTGAGTC...</td>\n",
       "      <td>AGTATGTTCTACTAAAGGCCTTGGCCCTTGCCAATTCGGGTGAGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43504523</th>\n",
       "      <td>chr2</td>\n",
       "      <td>-</td>\n",
       "      <td>85839156-85839235</td>\n",
       "      <td>ENST00000640992</td>\n",
       "      <td>rs1487482349</td>\n",
       "      <td>85839224</td>\n",
       "      <td>85839225</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...</td>\n",
       "      <td>AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43504524</th>\n",
       "      <td>chr2</td>\n",
       "      <td>-</td>\n",
       "      <td>85839156-85839235</td>\n",
       "      <td>ENST00000640992</td>\n",
       "      <td>rs3202465</td>\n",
       "      <td>85839225</td>\n",
       "      <td>85839226</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...</td>\n",
       "      <td>AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43504525</th>\n",
       "      <td>chr2</td>\n",
       "      <td>-</td>\n",
       "      <td>85839156-85839235</td>\n",
       "      <td>ENST00000640992</td>\n",
       "      <td>rs115591304</td>\n",
       "      <td>85839226</td>\n",
       "      <td>85839227</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...</td>\n",
       "      <td>AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43504526</th>\n",
       "      <td>chr2</td>\n",
       "      <td>-</td>\n",
       "      <td>85839156-85839235</td>\n",
       "      <td>ENST00000640992</td>\n",
       "      <td>rs1267804946</td>\n",
       "      <td>85839227</td>\n",
       "      <td>85839228</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...</td>\n",
       "      <td>AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43504527</th>\n",
       "      <td>chr2</td>\n",
       "      <td>-</td>\n",
       "      <td>85839156-85839235</td>\n",
       "      <td>ENST00000640992</td>\n",
       "      <td>rs927428411</td>\n",
       "      <td>85839232</td>\n",
       "      <td>85839233</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...</td>\n",
       "      <td>AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43504528 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            chr strand  donor_coordinates Ensemble_Transcript_ID  \\\n",
       "0         chr11      +  64315231-64315310        ENST00000000442   \n",
       "1         chr11      +  64315231-64315310        ENST00000000442   \n",
       "2         chr11      +  64315231-64315310        ENST00000000442   \n",
       "3         chr11      +  64315231-64315310        ENST00000000442   \n",
       "4         chr11      +  64315231-64315310        ENST00000000442   \n",
       "...         ...    ...                ...                    ...   \n",
       "43504523   chr2      -  85839156-85839235        ENST00000640992   \n",
       "43504524   chr2      -  85839156-85839235        ENST00000640992   \n",
       "43504525   chr2      -  85839156-85839235        ENST00000640992   \n",
       "43504526   chr2      -  85839156-85839235        ENST00000640992   \n",
       "43504527   chr2      -  85839156-85839235        ENST00000640992   \n",
       "\n",
       "                 RS_ID  variant_start  variant_end ref_neucleotide  \\\n",
       "0         rs1484286606       64315233     64315245    TATGTTCTACTA   \n",
       "1          rs750270682       64315250     64315251               C   \n",
       "2         rs2035223711       64315259     64315260               T   \n",
       "3         rs2035223790       64315260     64315261               G   \n",
       "4         rs1361791105       64315268     64315269               A   \n",
       "...                ...            ...          ...             ...   \n",
       "43504523  rs1487482349       85839224     85839225               A   \n",
       "43504524     rs3202465       85839225     85839226               G   \n",
       "43504525   rs115591304       85839226     85839227               A   \n",
       "43504526  rs1267804946       85839227     85839228               A   \n",
       "43504527   rs927428411       85839232     85839233               T   \n",
       "\n",
       "         alternative_neucleotide  \\\n",
       "0                             TA   \n",
       "1                              T   \n",
       "2                              C   \n",
       "3                              A   \n",
       "4                              G   \n",
       "...                          ...   \n",
       "43504523                       G   \n",
       "43504524                       A   \n",
       "43504525                       G   \n",
       "43504526                       G   \n",
       "43504527                       A   \n",
       "\n",
       "                                              reference_seq  \\\n",
       "0         AGTATGTTCTACTAAAGGCCTTGGCCCTTGCCAATTCAGGTGAGTC...   \n",
       "1         AGTATGTTCTACTAAAGGCCTTGGCCCTTGCCAATTCAGGTGAGTC...   \n",
       "2         AGTATGTTCTACTAAAGGCCTTGGCCCTTGCCAATTCAGGTGAGTC...   \n",
       "3         AGTATGTTCTACTAAAGGCCTTGGCCCTTGCCAATTCAGGTGAGTC...   \n",
       "4         AGTATGTTCTACTAAAGGCCTTGGCCCTTGCCAATTCAGGTGAGTC...   \n",
       "...                                                     ...   \n",
       "43504523  AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...   \n",
       "43504524  AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...   \n",
       "43504525  AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...   \n",
       "43504526  AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...   \n",
       "43504527  AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...   \n",
       "\n",
       "                                                    alt_seq  \n",
       "0         AGTAAAGGCCTTGGCCCTTGCCAATTCAGGTGAGTCTGGGGCAGGC...  \n",
       "1         AGTATGTTCTACTAAAGGCTTTGGCCCTTGCCAATTCAGGTGAGTC...  \n",
       "2         AGTATGTTCTACTAAAGGCCTTGGCCCTCGCCAATTCAGGTGAGTC...  \n",
       "3         AGTATGTTCTACTAAAGGCCTTGGCCCTTACCAATTCAGGTGAGTC...  \n",
       "4         AGTATGTTCTACTAAAGGCCTTGGCCCTTGCCAATTCGGGTGAGTC...  \n",
       "...                                                     ...  \n",
       "43504523  AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...  \n",
       "43504524  AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...  \n",
       "43504525  AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...  \n",
       "43504526  AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...  \n",
       "43504527  AAGTTTTTATTCTTTTTCATCTTTTTAAACTGGCACACTGCCTGGT...  \n",
       "\n",
       "[43504528 rows x 11 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7f6ff90c-19f5-4d83-b2ab-7dbcf50a30f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGTATG GTATGT TATGTT ATGTTC TGTTCT GTTCTA TTCT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGTAAA GTAAAG TAAAGG AAAGGC AAGGCC AGGCCT GGCC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AGTATG GTATGT TATGTT ATGTTC TGTTCT GTTCTA TTCT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AGTATG GTATGT TATGTT ATGTTC TGTTCT GTTCTA TTCT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGTATG GTATGT TATGTT ATGTTC TGTTCT GTTCTA TTCT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87009051</th>\n",
       "      <td>AAGTTT AGTTTT GTTTTT TTTTTA TTTTAT TTTATT TTAT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87009052</th>\n",
       "      <td>AAGTTT AGTTTT GTTTTT TTTTTA TTTTAT TTTATT TTAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87009053</th>\n",
       "      <td>AAGTTT AGTTTT GTTTTT TTTTTA TTTTAT TTTATT TTAT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87009054</th>\n",
       "      <td>AAGTTT AGTTTT GTTTTT TTTTTA TTTTAT TTTATT TTAT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87009055</th>\n",
       "      <td>AAGTTT AGTTTT GTTTTT TTTTTA TTTTAT TTTATT TTAT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>87009056 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   Sequence  Label\n",
       "0         AGTATG GTATGT TATGTT ATGTTC TGTTCT GTTCTA TTCT...      1\n",
       "1         AGTAAA GTAAAG TAAAGG AAAGGC AAGGCC AGGCCT GGCC...      1\n",
       "2         AGTATG GTATGT TATGTT ATGTTC TGTTCT GTTCTA TTCT...      1\n",
       "3         AGTATG GTATGT TATGTT ATGTTC TGTTCT GTTCTA TTCT...      1\n",
       "4         AGTATG GTATGT TATGTT ATGTTC TGTTCT GTTCTA TTCT...      0\n",
       "...                                                     ...    ...\n",
       "87009051  AAGTTT AGTTTT GTTTTT TTTTTA TTTTAT TTTATT TTAT...      1\n",
       "87009052  AAGTTT AGTTTT GTTTTT TTTTTA TTTTAT TTTATT TTAT...      0\n",
       "87009053  AAGTTT AGTTTT GTTTTT TTTTTA TTTTAT TTTATT TTAT...      1\n",
       "87009054  AAGTTT AGTTTT GTTTTT TTTTTA TTTTAT TTTATT TTAT...      1\n",
       "87009055  AAGTTT AGTTTT GTTTTT TTTTTA TTTTAT TTTATT TTAT...      1\n",
       "\n",
       "[87009056 rows x 2 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_kmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71fc5a7f-b9b2-4f68-b345-c02f59c74d7f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
