{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9807e3a4-8edb-46a2-ad4c-88777af9fadc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, glob\n",
    "import pandas as pd\n",
    "from io import StringIO\n",
    "import numpy as np\n",
    "import pysam\n",
    "import pybedtools \n",
    "from Bio import SeqIO\n",
    "pybedtools.helpers.set_tempdir('/home/pdutta/temp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6c3ba47b-0edf-43de-bab5-bc0edd7f91b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '/home/pdutta/Data/EnhancerPoster/ZBTB33/Intersected_files/all_Enhancer_data_Consensus.tsv'\n",
    "output_path = \"/home/pdutta/Data/EnhancerPoster/ZBTB33/DNABERT_data/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aafe0877-b7be-4e5e-a73f-00dab9627b51",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(output_path):\n",
    "    os.makedirs(output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cc89bac5-1528-4192-95d4-77bda6a6e1dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def seq2kmer(seq):\n",
    "    \"\"\"\n",
    "    Convert original sequence to kmers\n",
    "    \n",
    "    Arguments:\n",
    "    seq -- str, original sequence.\n",
    "    k -- int, kmer of length k specified.\n",
    "    \n",
    "    Returns:\n",
    "    kmers -- str, kmers separated by space\n",
    "    \"\"\"\n",
    "    k=6\n",
    "    kmer = [seq[x:x+k] for x in range(len(seq)+1-k)]\n",
    "    kmers = \" \".join(kmer)\n",
    "    return kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "19aaa7a0-7309-4e97-a93e-dded3164e07f",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = SeqIO.to_dict(SeqIO.parse(\"/home/pdutta/Data/Human_Genome_Data/GRCh38.primary_assembly.genome.fa\", \"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0ded228f-2999-482f-9e5a-197b732e0cfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(439792, 9)\n",
      "(430596, 9)\n",
      "(430596, 9)\n",
      "(497648, 9)\n",
      "(497648, 10)\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(data_path, sep= '\\t')\n",
    "print(df.shape)\n",
    "df = df[df['No_of_variations']<3].reset_index(drop=True)\n",
    "print(df.shape)\n",
    "df = df.loc[df['Alternative_Nucleotides'] != '.'].reset_index(drop=True)\n",
    "print(df.shape)\n",
    "\n",
    "\n",
    "data= []\n",
    "#df = df.head(10)\n",
    "# Iterate over DataFrame rows\n",
    "for idx, row in df.iterrows():\n",
    "    try:\n",
    "        #print(row)\n",
    "        alts = row['Alternative_Nucleotides'].rstrip(',').split(',')\n",
    "        ref_seq = str(genome[row['chr_name']].seq[ row['Enhancer_start']:row['Enhancer_end']])\n",
    "        #ref_sequences.append(ref_sequence)\n",
    "        #print(alts)\n",
    "        #print(ref_sequence)\n",
    "\n",
    "        # Calculate the variant position relative to the fetched sequence\n",
    "        variant_pos_start = row['Variant_start'] - row['Enhancer_start']\n",
    "        variant_pos_end = row['Variant_end']  - row['Enhancer_start']\n",
    "        #print(variant_pos_start, variant_pos_end)\n",
    "\n",
    "\n",
    "        # Replace the reference nucleotide with the alternate nucleotide to get the alternate sequence\n",
    "        for alt in alts:\n",
    "            if alt == \"\":  # Deletion\n",
    "                # Fetch additional bases from genome to maintain sequence length\n",
    "                extra_bases = str(genome[row['chr_name']].seq[row['Enhancer_end']:row['Enhancer_end'] + len(row['Reference_Nucleotide'])])\n",
    "                #print(len(row['Reference_Nucleotide']))\n",
    "                alt_seq = ref_seq[:variant_pos_start] + ref_seq[variant_pos_end:] + extra_bases\n",
    "            else:  # SNPs, Insertions, and partial Deletions\n",
    "                extra_bases = str(genome[row['chr_name']].seq[row['Enhancer_end']:row['Enhancer_end'] + len(row['Reference_Nucleotide']) - len(alt)])\n",
    "                #print(extra_bases)\n",
    "                if(variant_pos_start>=0):\n",
    "                    alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:] + extra_bases\n",
    "                else:\n",
    "                    continue\n",
    "            # Append to the list as a dictionary\n",
    "            data.append({\n",
    "                'chr': row['chr_name'],\n",
    "                'Enhancer_coordinates': str(row['Enhancer_start'])+'-'+str(row['Enhancer_end']),\n",
    "                'dbsnp_id': row['RS_ID'],\n",
    "                'varinat_start':  row['Variant_start'],\n",
    "                'variant_end':  row['Variant_end'],\n",
    "                'ref_neucleotide': row['Reference_Nucleotide'],\n",
    "                'alternative_neucleotide': alt,\n",
    "                'reference_seq': ref_seq,\n",
    "                'alt_seq': alt_seq\n",
    "            })\n",
    "    except KeyError:\n",
    "        # If sequence is not present, append a default value\n",
    "        ref_sequences.append(\"NA\")\n",
    "        alt_sequences.append(\"NA\")\n",
    "\n",
    "        \n",
    "        \n",
    "new_df = pd.DataFrame(data)\n",
    "print(new_df.shape)\n",
    "new_df = new_df.drop_duplicates().reset_index()\n",
    "print(new_df.shape)\n",
    "merged_list = list(zip(new_df['reference_seq'], new_df['alt_seq']))\n",
    "merged_list = [item.upper() for tup in merged_list for item in tup]\n",
    "#print(merged_list)\n",
    "kmer_lst = list(map(seq2kmer, merged_list))\n",
    "df_kmer = pd.DataFrame(kmer_lst, columns=['Sequence'])\n",
    "df_kmer['Label'] = np.random.choice([0, 1], size=len(df_kmer))\n",
    "\n",
    "new_df.to_csv(output_path + \"/all_data_new.tsv\", sep=\"\\t\", index= False)\n",
    "df_kmer.to_csv(output_path + \"/dev.tsv\", sep=\"\\t\", index= False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7f6ff90c-19f5-4d83-b2ab-7dbcf50a30f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_name</th>\n",
       "      <th>Enhancer_start</th>\n",
       "      <th>Enhancer_end</th>\n",
       "      <th>Variant_start</th>\n",
       "      <th>Variant_end</th>\n",
       "      <th>RS_ID</th>\n",
       "      <th>Reference_Nucleotide</th>\n",
       "      <th>No_of_variations</th>\n",
       "      <th>Alternative_Nucleotides</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr19</td>\n",
       "      <td>561375</td>\n",
       "      <td>561574</td>\n",
       "      <td>561375</td>\n",
       "      <td>561376</td>\n",
       "      <td>rs1345678776</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr19</td>\n",
       "      <td>561375</td>\n",
       "      <td>561574</td>\n",
       "      <td>561377</td>\n",
       "      <td>561378</td>\n",
       "      <td>rs570139519</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>G,T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr19</td>\n",
       "      <td>561375</td>\n",
       "      <td>561574</td>\n",
       "      <td>561378</td>\n",
       "      <td>561379</td>\n",
       "      <td>rs1022722361</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>A,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr19</td>\n",
       "      <td>561375</td>\n",
       "      <td>561574</td>\n",
       "      <td>561380</td>\n",
       "      <td>561381</td>\n",
       "      <td>rs1980794075</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr19</td>\n",
       "      <td>561375</td>\n",
       "      <td>561574</td>\n",
       "      <td>561382</td>\n",
       "      <td>561383</td>\n",
       "      <td>rs1980794126</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>A,C,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430591</th>\n",
       "      <td>chr18</td>\n",
       "      <td>2944504</td>\n",
       "      <td>2944703</td>\n",
       "      <td>2944683</td>\n",
       "      <td>2944687</td>\n",
       "      <td>rs2077422449</td>\n",
       "      <td>TTTT</td>\n",
       "      <td>1</td>\n",
       "      <td>TTTTT,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430592</th>\n",
       "      <td>chr18</td>\n",
       "      <td>2944504</td>\n",
       "      <td>2944703</td>\n",
       "      <td>2944688</td>\n",
       "      <td>2944689</td>\n",
       "      <td>rs2077422509</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430593</th>\n",
       "      <td>chr18</td>\n",
       "      <td>2944504</td>\n",
       "      <td>2944703</td>\n",
       "      <td>2944689</td>\n",
       "      <td>2944690</td>\n",
       "      <td>rs1271925579</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430594</th>\n",
       "      <td>chr18</td>\n",
       "      <td>2944504</td>\n",
       "      <td>2944703</td>\n",
       "      <td>2944695</td>\n",
       "      <td>2944696</td>\n",
       "      <td>rs1334793346</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>A,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430595</th>\n",
       "      <td>chr18</td>\n",
       "      <td>2944504</td>\n",
       "      <td>2944703</td>\n",
       "      <td>2944702</td>\n",
       "      <td>2944703</td>\n",
       "      <td>rs1214821368</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>T,</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>430596 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr_name  Enhancer_start  Enhancer_end  Variant_start  Variant_end  \\\n",
       "0         chr19          561375        561574         561375       561376   \n",
       "1         chr19          561375        561574         561377       561378   \n",
       "2         chr19          561375        561574         561378       561379   \n",
       "3         chr19          561375        561574         561380       561381   \n",
       "4         chr19          561375        561574         561382       561383   \n",
       "...         ...             ...           ...            ...          ...   \n",
       "430591    chr18         2944504       2944703        2944683      2944687   \n",
       "430592    chr18         2944504       2944703        2944688      2944689   \n",
       "430593    chr18         2944504       2944703        2944689      2944690   \n",
       "430594    chr18         2944504       2944703        2944695      2944696   \n",
       "430595    chr18         2944504       2944703        2944702      2944703   \n",
       "\n",
       "               RS_ID Reference_Nucleotide  No_of_variations  \\\n",
       "0       rs1345678776                    C                 1   \n",
       "1        rs570139519                    C                 2   \n",
       "2       rs1022722361                    G                 1   \n",
       "3       rs1980794075                    C                 1   \n",
       "4       rs1980794126                    G                 2   \n",
       "...              ...                  ...               ...   \n",
       "430591  rs2077422449                 TTTT                 1   \n",
       "430592  rs2077422509                    C                 1   \n",
       "430593  rs1271925579                    A                 1   \n",
       "430594  rs1334793346                    G                 1   \n",
       "430595  rs1214821368                    G                 1   \n",
       "\n",
       "       Alternative_Nucleotides  \n",
       "0                           T,  \n",
       "1                         G,T,  \n",
       "2                           A,  \n",
       "3                           T,  \n",
       "4                         A,C,  \n",
       "...                        ...  \n",
       "430591                  TTTTT,  \n",
       "430592                      T,  \n",
       "430593                      T,  \n",
       "430594                      A,  \n",
       "430595                      T,  \n",
       "\n",
       "[430596 rows x 9 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2449d935-86ff-41a2-b8bd-c9a68b195d24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>chr</th>\n",
       "      <th>Enhancer_coordinates</th>\n",
       "      <th>dbsnp_id</th>\n",
       "      <th>varinat_start</th>\n",
       "      <th>variant_end</th>\n",
       "      <th>ref_neucleotide</th>\n",
       "      <th>alternative_neucleotide</th>\n",
       "      <th>reference_seq</th>\n",
       "      <th>alt_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>chr19</td>\n",
       "      <td>561375-561574</td>\n",
       "      <td>rs1345678776</td>\n",
       "      <td>561375</td>\n",
       "      <td>561376</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CACGCCTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...</td>\n",
       "      <td>TACGCCTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>chr19</td>\n",
       "      <td>561375-561574</td>\n",
       "      <td>rs570139519</td>\n",
       "      <td>561377</td>\n",
       "      <td>561378</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>CACGCCTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...</td>\n",
       "      <td>CAGGCCTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>chr19</td>\n",
       "      <td>561375-561574</td>\n",
       "      <td>rs570139519</td>\n",
       "      <td>561377</td>\n",
       "      <td>561378</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CACGCCTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...</td>\n",
       "      <td>CATGCCTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>chr19</td>\n",
       "      <td>561375-561574</td>\n",
       "      <td>rs1022722361</td>\n",
       "      <td>561378</td>\n",
       "      <td>561379</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>CACGCCTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...</td>\n",
       "      <td>CACACCTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>chr19</td>\n",
       "      <td>561375-561574</td>\n",
       "      <td>rs1980794075</td>\n",
       "      <td>561380</td>\n",
       "      <td>561381</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CACGCCTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...</td>\n",
       "      <td>CACGCTTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497643</th>\n",
       "      <td>497643</td>\n",
       "      <td>chr18</td>\n",
       "      <td>2944504-2944703</td>\n",
       "      <td>rs2077422449</td>\n",
       "      <td>2944683</td>\n",
       "      <td>2944687</td>\n",
       "      <td>TTTT</td>\n",
       "      <td>TTTTT</td>\n",
       "      <td>CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...</td>\n",
       "      <td>CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497644</th>\n",
       "      <td>497644</td>\n",
       "      <td>chr18</td>\n",
       "      <td>2944504-2944703</td>\n",
       "      <td>rs2077422509</td>\n",
       "      <td>2944688</td>\n",
       "      <td>2944689</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...</td>\n",
       "      <td>CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497645</th>\n",
       "      <td>497645</td>\n",
       "      <td>chr18</td>\n",
       "      <td>2944504-2944703</td>\n",
       "      <td>rs1271925579</td>\n",
       "      <td>2944689</td>\n",
       "      <td>2944690</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...</td>\n",
       "      <td>CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497646</th>\n",
       "      <td>497646</td>\n",
       "      <td>chr18</td>\n",
       "      <td>2944504-2944703</td>\n",
       "      <td>rs1334793346</td>\n",
       "      <td>2944695</td>\n",
       "      <td>2944696</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...</td>\n",
       "      <td>CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497647</th>\n",
       "      <td>497647</td>\n",
       "      <td>chr18</td>\n",
       "      <td>2944504-2944703</td>\n",
       "      <td>rs1214821368</td>\n",
       "      <td>2944702</td>\n",
       "      <td>2944703</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...</td>\n",
       "      <td>CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>497648 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         index    chr Enhancer_coordinates      dbsnp_id  varinat_start  \\\n",
       "0            0  chr19        561375-561574  rs1345678776         561375   \n",
       "1            1  chr19        561375-561574   rs570139519         561377   \n",
       "2            2  chr19        561375-561574   rs570139519         561377   \n",
       "3            3  chr19        561375-561574  rs1022722361         561378   \n",
       "4            4  chr19        561375-561574  rs1980794075         561380   \n",
       "...        ...    ...                  ...           ...            ...   \n",
       "497643  497643  chr18      2944504-2944703  rs2077422449        2944683   \n",
       "497644  497644  chr18      2944504-2944703  rs2077422509        2944688   \n",
       "497645  497645  chr18      2944504-2944703  rs1271925579        2944689   \n",
       "497646  497646  chr18      2944504-2944703  rs1334793346        2944695   \n",
       "497647  497647  chr18      2944504-2944703  rs1214821368        2944702   \n",
       "\n",
       "        variant_end ref_neucleotide alternative_neucleotide  \\\n",
       "0            561376               C                       T   \n",
       "1            561378               C                       G   \n",
       "2            561378               C                       T   \n",
       "3            561379               G                       A   \n",
       "4            561381               C                       T   \n",
       "...             ...             ...                     ...   \n",
       "497643      2944687            TTTT                   TTTTT   \n",
       "497644      2944689               C                       T   \n",
       "497645      2944690               A                       T   \n",
       "497646      2944696               G                       A   \n",
       "497647      2944703               G                       T   \n",
       "\n",
       "                                            reference_seq  \\\n",
       "0       CACGCCTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...   \n",
       "1       CACGCCTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...   \n",
       "2       CACGCCTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...   \n",
       "3       CACGCCTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...   \n",
       "4       CACGCCTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...   \n",
       "...                                                   ...   \n",
       "497643  CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...   \n",
       "497644  CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...   \n",
       "497645  CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...   \n",
       "497646  CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...   \n",
       "497647  CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...   \n",
       "\n",
       "                                                  alt_seq  \n",
       "0       TACGCCTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...  \n",
       "1       CAGGCCTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...  \n",
       "2       CATGCCTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...  \n",
       "3       CACACCTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...  \n",
       "4       CACGCTTGTCATCCCAGCACTTCGGGAGGCCAAGGCGGGTGGATCA...  \n",
       "...                                                   ...  \n",
       "497643  CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...  \n",
       "497644  CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...  \n",
       "497645  CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...  \n",
       "497646  CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...  \n",
       "497647  CTACGCCCAGCTAATTTTTGTATTTCTAGTAGAGACGGGGTTTCAC...  \n",
       "\n",
       "[497648 rows x 10 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "60b4b0de-8a35-4845-8d34-5d1f75c95385",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CACGCC ACGCCT CGCCTG GCCTGT CCTGTC CTGTCA TGTC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TACGCC ACGCCT CGCCTG GCCTGT CCTGTC CTGTCA TGTC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CACGCC ACGCCT CGCCTG GCCTGT CCTGTC CTGTCA TGTC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAGGCC AGGCCT GGCCTG GCCTGT CCTGTC CTGTCA TGTC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CACGCC ACGCCT CGCCTG GCCTGT CCTGTC CTGTCA TGTC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995291</th>\n",
       "      <td>CTACGC TACGCC ACGCCC CGCCCA GCCCAG CCCAGC CCAG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995292</th>\n",
       "      <td>CTACGC TACGCC ACGCCC CGCCCA GCCCAG CCCAGC CCAG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995293</th>\n",
       "      <td>CTACGC TACGCC ACGCCC CGCCCA GCCCAG CCCAGC CCAG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995294</th>\n",
       "      <td>CTACGC TACGCC ACGCCC CGCCCA GCCCAG CCCAGC CCAG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995295</th>\n",
       "      <td>CTACGC TACGCC ACGCCC CGCCCA GCCCAG CCCAGC CCAG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>995296 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Sequence  Label\n",
       "0       CACGCC ACGCCT CGCCTG GCCTGT CCTGTC CTGTCA TGTC...      1\n",
       "1       TACGCC ACGCCT CGCCTG GCCTGT CCTGTC CTGTCA TGTC...      0\n",
       "2       CACGCC ACGCCT CGCCTG GCCTGT CCTGTC CTGTCA TGTC...      0\n",
       "3       CAGGCC AGGCCT GGCCTG GCCTGT CCTGTC CTGTCA TGTC...      1\n",
       "4       CACGCC ACGCCT CGCCTG GCCTGT CCTGTC CTGTCA TGTC...      1\n",
       "...                                                   ...    ...\n",
       "995291  CTACGC TACGCC ACGCCC CGCCCA GCCCAG CCCAGC CCAG...      0\n",
       "995292  CTACGC TACGCC ACGCCC CGCCCA GCCCAG CCCAGC CCAG...      0\n",
       "995293  CTACGC TACGCC ACGCCC CGCCCA GCCCAG CCCAGC CCAG...      1\n",
       "995294  CTACGC TACGCC ACGCCC CGCCCA GCCCAG CCCAGC CCAG...      0\n",
       "995295  CTACGC TACGCC ACGCCC CGCCCA GCCCAG CCCAGC CCAG...      0\n",
       "\n",
       "[995296 rows x 2 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_kmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9141801-94fd-4b9e-a142-798f8fd453e4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
