{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9807e3a4-8edb-46a2-ad4c-88777af9fadc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, glob\n",
    "import pandas as pd\n",
    "from io import StringIO\n",
    "import pysam\n",
    "import pybedtools \n",
    "pybedtools.helpers.set_tempdir('/data/private/pdutta/PSB_Data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6c3ba47b-0edf-43de-bab5-bc0edd7f91b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_wise_files = glob.glob('/data/private/pdutta/PSB_Data/Acceptor/Chrwise/*.tsv')\n",
    "reference_path  = \"/data/projects/Resources/HumanReferenceGenome/\"\n",
    "output_path = \"/data/private/pdutta/PSB_Data/Acceptor/DNABERT_data/\" "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cc89bac5-1528-4192-95d4-77bda6a6e1dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def seq2kmer(seq):\n",
    "    \"\"\"\n",
    "    Convert original sequence to kmers\n",
    "    \n",
    "    Arguments:\n",
    "    seq -- str, original sequence.\n",
    "    k -- int, kmer of length k specified.\n",
    "    \n",
    "    Returns:\n",
    "    kmers -- str, kmers separated by space\n",
    "    \"\"\"\n",
    "    k=6\n",
    "    kmer = [seq[x:x+k] for x in range(len(seq)+1-k)]\n",
    "    kmers = \" \".join(kmer)\n",
    "    return kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0ded228f-2999-482f-9e5a-197b732e0cfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n",
      "(41016818, 17)\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "Interrupted by user",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 8\u001b[0m\n\u001b[1;32m      6\u001b[0m df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(intersect_file, header\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, sep\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28mprint\u001b[39m(df\u001b[38;5;241m.\u001b[39mshape)\n\u001b[0;32m----> 8\u001b[0m \u001b[38;5;28;43minput\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      9\u001b[0m df \u001b[38;5;241m=\u001b[39m df[df[\u001b[38;5;241m15\u001b[39m]\u001b[38;5;241m<\u001b[39m\u001b[38;5;241m3\u001b[39m]\u001b[38;5;241m.\u001b[39mreset_index(drop\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m     10\u001b[0m \u001b[38;5;28mprint\u001b[39m(df\u001b[38;5;241m.\u001b[39mshape)\n",
      "File \u001b[0;32m/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/ipykernel/kernelbase.py:1175\u001b[0m, in \u001b[0;36mKernel.raw_input\u001b[0;34m(self, prompt)\u001b[0m\n\u001b[1;32m   1171\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_allow_stdin:\n\u001b[1;32m   1172\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m StdinNotImplementedError(\n\u001b[1;32m   1173\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mraw_input was called, but this frontend does not support input requests.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1174\u001b[0m     )\n\u001b[0;32m-> 1175\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_input_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1176\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mstr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mprompt\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1177\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_parent_ident\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mshell\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1178\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_parent\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mshell\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1179\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpassword\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m   1180\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/ipykernel/kernelbase.py:1217\u001b[0m, in \u001b[0;36mKernel._input_request\u001b[0;34m(self, prompt, ident, parent, password)\u001b[0m\n\u001b[1;32m   1214\u001b[0m             \u001b[38;5;28;01mbreak\u001b[39;00m\n\u001b[1;32m   1215\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyboardInterrupt\u001b[39;00m:\n\u001b[1;32m   1216\u001b[0m     \u001b[38;5;66;03m# re-raise KeyboardInterrupt, to truncate traceback\u001b[39;00m\n\u001b[0;32m-> 1217\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyboardInterrupt\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mInterrupted by user\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28mNone\u001b[39m\n\u001b[1;32m   1218\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[1;32m   1219\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlog\u001b[38;5;241m.\u001b[39mwarning(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mInvalid Message:\u001b[39m\u001b[38;5;124m\"\u001b[39m, exc_info\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: Interrupted by user"
     ]
    }
   ],
   "source": [
    "for intersect_file in chr_wise_files:\n",
    "    # Read the csv file into a DataFrame\n",
    "    chromosome_name = intersect_file.split('_')[-2]\n",
    "    if (chromosome_name != 'chrM'):\n",
    "        print(chromosome_name)\n",
    "        df = pd.read_csv(intersect_file, header=None, sep= '\\t')\n",
    "        print(df.shape)\n",
    "        df = df[df[15]<3].reset_index(drop=True)\n",
    "        print(df.shape)\n",
    "        df = df.loc[df[14] != '.'].reset_index(drop=True)\n",
    "        print(df.shape)\n",
    "        genome = pysam.FastaFile(reference_path+chromosome_name+\".fa\")\n",
    "        \n",
    "        # Lists to hold the reference and alternative sequences\n",
    "        ref_sequences = []\n",
    "        alt_sequences = []\n",
    "        df = df[df[15]>1]\n",
    "        df = df.head(10)\n",
    "        # Iterate over DataFrame rows\n",
    "        for idx, row in df.iterrows():\n",
    "            try:\n",
    "                ref_sequence = genome.fetch(row[0], row[1], row[2])\n",
    "                ref_sequences.append(ref_sequence)\n",
    "                print(ref_sequence)\n",
    "\n",
    "                # Calculate the variant position relative to the fetched sequence\n",
    "                variant_pos = row['VCF_position'] - row['start']\n",
    "\n",
    "\n",
    "                # Replace the reference nucleotide with the alternate nucleotide to get the alternate sequence\n",
    "                alt_sequence = ref_sequence[:variant_pos] + row['alt_nucleotide'] + ref_sequence[variant_pos + len(row['ref_nucleotide']):]\n",
    "\n",
    "                # Fetch the reference sequence\n",
    "                \n",
    "\n",
    "                # Generate the alternative sequence\n",
    "                alt_sequence = ref_sequence[:row[11] - row[1] - 1] + row[16].replace(\",\", \"\") + ref_sequence[row[11] - row[1]:]\n",
    "                alt_sequences.append(alt_sequence)\n",
    "                print(ref_sequence)\n",
    "                print(alt_sequence)\n",
    "            except KeyError:\n",
    "                # If sequence is not present, append a default value\n",
    "                ref_sequences.append(\"NA\")\n",
    "                alt_sequences.append(\"NA\")\n",
    "\n",
    "        # Add new columns to the DataFrame\n",
    "        df['ref_sequence'] = ref_sequences\n",
    "        df['alt_sequence'] = alt_sequences\n",
    "        merged_list = [i for pair in zip(ref_sequences, alt_sequences) for i in pair]\n",
    "        # Apply function to all items in the list\n",
    "        kmer_lst = list(map(seq2kmer, merged_list))\n",
    "\n",
    "        # Convert to DataFrame\n",
    "        df_kmer = pd.DataFrame(kmer_lst, columns=['Sequence'])\n",
    "        values = [0] * (len(df_kmer) // 2) + [1] * (len(df_kmer) // 2)\n",
    "\n",
    "        # If the DataFrame has an odd number of rows, add one more 0 or 1 to make the length match\n",
    "        if len(df_kmer) % 2:\n",
    "            values += [np.random.choice([0, 1])]\n",
    "        df_kmer['Label'] = values\n",
    "        print(df)\n",
    "        print(df_kmer)\n",
    "        out_folder_path = output_path+chromosome_name\n",
    "        if not os.path.exists(out_folder_path):\n",
    "            os.makedirs(out_folder_path)\n",
    "        df.to_csv(out_folder_path + \"/all_data.tsv\", sep=\"\\t\", index= False)\n",
    "        df_kmer.to_csv(out_folder_path + \"/dev.tsv\", sep=\"\\t\", index= False)\n",
    "        print(\"All the files of \", chromosome_name, \"are saved !!!\\n \")\n",
    "        input()\n",
    "        # Close the reference genome\n",
    "        genome.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7f6ff90c-19f5-4d83-b2ab-7dbcf50a30f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24401417</td>\n",
       "      <td>24401496</td>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24401079</td>\n",
       "      <td>24410779</td>\n",
       "      <td>ENST00000475760</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24401420</td>\n",
       "      <td>24401421</td>\n",
       "      <td>rs767109914</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>C,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24401417</td>\n",
       "      <td>24401496</td>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24401079</td>\n",
       "      <td>24410779</td>\n",
       "      <td>ENST00000475760</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24401421</td>\n",
       "      <td>24401421</td>\n",
       "      <td>rs1557461252</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "      <td>CCACTCCTCAGGTATAGACCCATGAAATACATAAATATGTGTCCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24401417</td>\n",
       "      <td>24401496</td>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24401079</td>\n",
       "      <td>24410779</td>\n",
       "      <td>ENST00000475760</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24401424</td>\n",
       "      <td>24401425</td>\n",
       "      <td>rs1643224572</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>C,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24401417</td>\n",
       "      <td>24401496</td>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24401079</td>\n",
       "      <td>24410779</td>\n",
       "      <td>ENST00000475760</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24401427</td>\n",
       "      <td>24401428</td>\n",
       "      <td>rs1402144573</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>G,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24401417</td>\n",
       "      <td>24401496</td>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24401079</td>\n",
       "      <td>24410779</td>\n",
       "      <td>ENST00000475760</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24401433</td>\n",
       "      <td>24401434</td>\n",
       "      <td>rs1643224756</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>G,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41016813</th>\n",
       "      <td>chr1</td>\n",
       "      <td>236762420</td>\n",
       "      <td>236762499</td>\n",
       "      <td>ENST00000684763.1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236686682</td>\n",
       "      <td>236764595</td>\n",
       "      <td>ENST00000683111</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236762488</td>\n",
       "      <td>236762489</td>\n",
       "      <td>rs727502888</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>A,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41016814</th>\n",
       "      <td>chr1</td>\n",
       "      <td>236762420</td>\n",
       "      <td>236762499</td>\n",
       "      <td>ENST00000684763.1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236686682</td>\n",
       "      <td>236764595</td>\n",
       "      <td>ENST00000683111</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236762489</td>\n",
       "      <td>236762490</td>\n",
       "      <td>rs1362873015</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>A,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41016815</th>\n",
       "      <td>chr1</td>\n",
       "      <td>236762420</td>\n",
       "      <td>236762499</td>\n",
       "      <td>ENST00000684763.1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236686682</td>\n",
       "      <td>236764595</td>\n",
       "      <td>ENST00000683111</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236762493</td>\n",
       "      <td>236762494</td>\n",
       "      <td>rs370992948</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41016816</th>\n",
       "      <td>chr1</td>\n",
       "      <td>236762420</td>\n",
       "      <td>236762499</td>\n",
       "      <td>ENST00000684763.1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236686682</td>\n",
       "      <td>236764595</td>\n",
       "      <td>ENST00000683111</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236762496</td>\n",
       "      <td>236762497</td>\n",
       "      <td>rs746887157</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41016817</th>\n",
       "      <td>chr1</td>\n",
       "      <td>236762420</td>\n",
       "      <td>236762499</td>\n",
       "      <td>ENST00000684763.1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236686682</td>\n",
       "      <td>236764595</td>\n",
       "      <td>ENST00000683111</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236762498</td>\n",
       "      <td>236762499</td>\n",
       "      <td>rs1572153340</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>T,</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41016818 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0          1          2                   3  4     5          6   \\\n",
       "0         chr1   24401417   24401496  ENST00000003583.12  -  chr1   24401079   \n",
       "1         chr1   24401417   24401496  ENST00000003583.12  -  chr1   24401079   \n",
       "2         chr1   24401417   24401496  ENST00000003583.12  -  chr1   24401079   \n",
       "3         chr1   24401417   24401496  ENST00000003583.12  -  chr1   24401079   \n",
       "4         chr1   24401417   24401496  ENST00000003583.12  -  chr1   24401079   \n",
       "...        ...        ...        ...                 ... ..   ...        ...   \n",
       "41016813  chr1  236762420  236762499   ENST00000684763.1  +  chr1  236686682   \n",
       "41016814  chr1  236762420  236762499   ENST00000684763.1  +  chr1  236686682   \n",
       "41016815  chr1  236762420  236762499   ENST00000684763.1  +  chr1  236686682   \n",
       "41016816  chr1  236762420  236762499   ENST00000684763.1  +  chr1  236686682   \n",
       "41016817  chr1  236762420  236762499   ENST00000684763.1  +  chr1  236686682   \n",
       "\n",
       "                 7                8   9     10         11         12  \\\n",
       "0          24410779  ENST00000475760  -1  chr1   24401420   24401421   \n",
       "1          24410779  ENST00000475760  -1  chr1   24401421   24401421   \n",
       "2          24410779  ENST00000475760  -1  chr1   24401424   24401425   \n",
       "3          24410779  ENST00000475760  -1  chr1   24401427   24401428   \n",
       "4          24410779  ENST00000475760  -1  chr1   24401433   24401434   \n",
       "...             ...              ...  ..   ...        ...        ...   \n",
       "41016813  236764595  ENST00000683111   1  chr1  236762488  236762489   \n",
       "41016814  236764595  ENST00000683111   1  chr1  236762489  236762490   \n",
       "41016815  236764595  ENST00000683111   1  chr1  236762493  236762494   \n",
       "41016816  236764595  ENST00000683111   1  chr1  236762496  236762497   \n",
       "41016817  236764595  ENST00000683111   1  chr1  236762498  236762499   \n",
       "\n",
       "                    13 14  15  \\\n",
       "0          rs767109914  T   1   \n",
       "1         rs1557461252  .   1   \n",
       "2         rs1643224572  T   1   \n",
       "3         rs1402144573  A   1   \n",
       "4         rs1643224756  C   1   \n",
       "...                ... ..  ..   \n",
       "41016813   rs727502888  G   1   \n",
       "41016814  rs1362873015  G   1   \n",
       "41016815   rs370992948  C   1   \n",
       "41016816   rs746887157  C   1   \n",
       "41016817  rs1572153340  C   1   \n",
       "\n",
       "                                                         16  \n",
       "0                                                        C,  \n",
       "1         CCACTCCTCAGGTATAGACCCATGAAATACATAAATATGTGTCCCA...  \n",
       "2                                                        C,  \n",
       "3                                                        G,  \n",
       "4                                                        G,  \n",
       "...                                                     ...  \n",
       "41016813                                                 A,  \n",
       "41016814                                                 A,  \n",
       "41016815                                                 T,  \n",
       "41016816                                                 T,  \n",
       "41016817                                                 T,  \n",
       "\n",
       "[41016818 rows x 17 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "5e7fad5b-4410-4203-b57e-bd419412bcad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24383964</td>\n",
       "      <td>24384043</td>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24357005</td>\n",
       "      <td>24413725</td>\n",
       "      <td>ENST00000003583</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24383964</td>\n",
       "      <td>24383965</td>\n",
       "      <td>rs1642394632</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>C,T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24383964</td>\n",
       "      <td>24384043</td>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24357005</td>\n",
       "      <td>24413725</td>\n",
       "      <td>ENST00000003583</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24384014</td>\n",
       "      <td>24384015</td>\n",
       "      <td>rs771170905</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>C,T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24383964</td>\n",
       "      <td>24384043</td>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24356999</td>\n",
       "      <td>24413782</td>\n",
       "      <td>ENST00000337248</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24383964</td>\n",
       "      <td>24383965</td>\n",
       "      <td>rs1642394632</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>C,T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24383964</td>\n",
       "      <td>24384043</td>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24356999</td>\n",
       "      <td>24413782</td>\n",
       "      <td>ENST00000337248</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24384014</td>\n",
       "      <td>24384015</td>\n",
       "      <td>rs771170905</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>C,T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24383964</td>\n",
       "      <td>24384043</td>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24356999</td>\n",
       "      <td>24415097</td>\n",
       "      <td>ENST00000374409</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24383964</td>\n",
       "      <td>24383965</td>\n",
       "      <td>rs1642394632</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>C,T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41016626</th>\n",
       "      <td>chr1</td>\n",
       "      <td>236762420</td>\n",
       "      <td>236762499</td>\n",
       "      <td>ENST00000684763.1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236686639</td>\n",
       "      <td>236764595</td>\n",
       "      <td>ENST00000684050</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236762446</td>\n",
       "      <td>236762447</td>\n",
       "      <td>rs753022801</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>C,T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41016667</th>\n",
       "      <td>chr1</td>\n",
       "      <td>236762420</td>\n",
       "      <td>236762499</td>\n",
       "      <td>ENST00000684763.1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236686639</td>\n",
       "      <td>236764595</td>\n",
       "      <td>ENST00000684502</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236762446</td>\n",
       "      <td>236762447</td>\n",
       "      <td>rs753022801</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>C,T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41016708</th>\n",
       "      <td>chr1</td>\n",
       "      <td>236762420</td>\n",
       "      <td>236762499</td>\n",
       "      <td>ENST00000684763.1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236686674</td>\n",
       "      <td>236764595</td>\n",
       "      <td>ENST00000682692</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236762446</td>\n",
       "      <td>236762447</td>\n",
       "      <td>rs753022801</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>C,T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41016749</th>\n",
       "      <td>chr1</td>\n",
       "      <td>236762420</td>\n",
       "      <td>236762499</td>\n",
       "      <td>ENST00000684763.1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236686675</td>\n",
       "      <td>236764595</td>\n",
       "      <td>ENST00000682966</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236762446</td>\n",
       "      <td>236762447</td>\n",
       "      <td>rs753022801</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>C,T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41016790</th>\n",
       "      <td>chr1</td>\n",
       "      <td>236762420</td>\n",
       "      <td>236762499</td>\n",
       "      <td>ENST00000684763.1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236686682</td>\n",
       "      <td>236764595</td>\n",
       "      <td>ENST00000683111</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236762446</td>\n",
       "      <td>236762447</td>\n",
       "      <td>rs753022801</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>C,T,</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>326426 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0          1          2                   3  4     5          6   \\\n",
       "182       chr1   24383964   24384043  ENST00000003583.12  -  chr1   24357005   \n",
       "200       chr1   24383964   24384043  ENST00000003583.12  -  chr1   24357005   \n",
       "213       chr1   24383964   24384043  ENST00000003583.12  -  chr1   24356999   \n",
       "231       chr1   24383964   24384043  ENST00000003583.12  -  chr1   24356999   \n",
       "244       chr1   24383964   24384043  ENST00000003583.12  -  chr1   24356999   \n",
       "...        ...        ...        ...                 ... ..   ...        ...   \n",
       "41016626  chr1  236762420  236762499   ENST00000684763.1  +  chr1  236686639   \n",
       "41016667  chr1  236762420  236762499   ENST00000684763.1  +  chr1  236686639   \n",
       "41016708  chr1  236762420  236762499   ENST00000684763.1  +  chr1  236686674   \n",
       "41016749  chr1  236762420  236762499   ENST00000684763.1  +  chr1  236686675   \n",
       "41016790  chr1  236762420  236762499   ENST00000684763.1  +  chr1  236686682   \n",
       "\n",
       "                 7                8   9     10         11         12  \\\n",
       "182        24413725  ENST00000003583  -1  chr1   24383964   24383965   \n",
       "200        24413725  ENST00000003583  -1  chr1   24384014   24384015   \n",
       "213        24413782  ENST00000337248  -1  chr1   24383964   24383965   \n",
       "231        24413782  ENST00000337248  -1  chr1   24384014   24384015   \n",
       "244        24415097  ENST00000374409  -1  chr1   24383964   24383965   \n",
       "...             ...              ...  ..   ...        ...        ...   \n",
       "41016626  236764595  ENST00000684050   1  chr1  236762446  236762447   \n",
       "41016667  236764595  ENST00000684502   1  chr1  236762446  236762447   \n",
       "41016708  236764595  ENST00000682692   1  chr1  236762446  236762447   \n",
       "41016749  236764595  ENST00000682966   1  chr1  236762446  236762447   \n",
       "41016790  236764595  ENST00000683111   1  chr1  236762446  236762447   \n",
       "\n",
       "                    13 14  15    16  \n",
       "182       rs1642394632  G   2  C,T,  \n",
       "200        rs771170905  G   2  C,T,  \n",
       "213       rs1642394632  G   2  C,T,  \n",
       "231        rs771170905  G   2  C,T,  \n",
       "244       rs1642394632  G   2  C,T,  \n",
       "...                ... ..  ..   ...  \n",
       "41016626   rs753022801  G   2  C,T,  \n",
       "41016667   rs753022801  G   2  C,T,  \n",
       "41016708   rs753022801  G   2  C,T,  \n",
       "41016749   rs753022801  G   2  C,T,  \n",
       "41016790   rs753022801  G   2  C,T,  \n",
       "\n",
       "[326426 rows x 17 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df[16]=='C,T,']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5b9a614c-b86b-41e3-b553-0fc6a6300dc0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "175344"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_instances = (df[12] >df[2]).sum()\n",
    "num_instances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "c35c1d30-9e2b-410b-8325-b89cc9c6957f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1061</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24361002</td>\n",
       "      <td>24361081</td>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24319425</td>\n",
       "      <td>24364482</td>\n",
       "      <td>ENST00000350501</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24361078</td>\n",
       "      <td>24361086</td>\n",
       "      <td>rs1641082067</td>\n",
       "      <td>AGTCTAGT</td>\n",
       "      <td>1</td>\n",
       "      <td>AGT,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1088</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24361002</td>\n",
       "      <td>24361081</td>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24357005</td>\n",
       "      <td>24413725</td>\n",
       "      <td>ENST00000003583</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24361078</td>\n",
       "      <td>24361086</td>\n",
       "      <td>rs1641082067</td>\n",
       "      <td>AGTCTAGT</td>\n",
       "      <td>1</td>\n",
       "      <td>AGT,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1115</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24361002</td>\n",
       "      <td>24361081</td>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24356999</td>\n",
       "      <td>24413782</td>\n",
       "      <td>ENST00000337248</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24361078</td>\n",
       "      <td>24361086</td>\n",
       "      <td>rs1641082067</td>\n",
       "      <td>AGTCTAGT</td>\n",
       "      <td>1</td>\n",
       "      <td>AGT,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1142</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24361002</td>\n",
       "      <td>24361081</td>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24356999</td>\n",
       "      <td>24415097</td>\n",
       "      <td>ENST00000374409</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24361078</td>\n",
       "      <td>24361086</td>\n",
       "      <td>rs1641082067</td>\n",
       "      <td>AGTCTAGT</td>\n",
       "      <td>1</td>\n",
       "      <td>AGT,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1169</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24361002</td>\n",
       "      <td>24361081</td>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24357005</td>\n",
       "      <td>24416934</td>\n",
       "      <td>ENST00000468303</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24361078</td>\n",
       "      <td>24361086</td>\n",
       "      <td>rs1641082067</td>\n",
       "      <td>AGTCTAGT</td>\n",
       "      <td>1</td>\n",
       "      <td>AGT,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41014654</th>\n",
       "      <td>chr1</td>\n",
       "      <td>236757445</td>\n",
       "      <td>236757524</td>\n",
       "      <td>ENST00000684763.1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236686639</td>\n",
       "      <td>236764595</td>\n",
       "      <td>ENST00000684050</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236757521</td>\n",
       "      <td>236757557</td>\n",
       "      <td>rs1572148914</td>\n",
       "      <td>ATCGCCAGAACCATCAATGAGGTGGAGACTCAGATC</td>\n",
       "      <td>1</td>\n",
       "      <td>ATC,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41014680</th>\n",
       "      <td>chr1</td>\n",
       "      <td>236757445</td>\n",
       "      <td>236757524</td>\n",
       "      <td>ENST00000684763.1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236686639</td>\n",
       "      <td>236764595</td>\n",
       "      <td>ENST00000684502</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236757521</td>\n",
       "      <td>236757557</td>\n",
       "      <td>rs1572148914</td>\n",
       "      <td>ATCGCCAGAACCATCAATGAGGTGGAGACTCAGATC</td>\n",
       "      <td>1</td>\n",
       "      <td>ATC,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41014706</th>\n",
       "      <td>chr1</td>\n",
       "      <td>236757445</td>\n",
       "      <td>236757524</td>\n",
       "      <td>ENST00000684763.1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236686674</td>\n",
       "      <td>236764595</td>\n",
       "      <td>ENST00000682692</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236757521</td>\n",
       "      <td>236757557</td>\n",
       "      <td>rs1572148914</td>\n",
       "      <td>ATCGCCAGAACCATCAATGAGGTGGAGACTCAGATC</td>\n",
       "      <td>1</td>\n",
       "      <td>ATC,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41014732</th>\n",
       "      <td>chr1</td>\n",
       "      <td>236757445</td>\n",
       "      <td>236757524</td>\n",
       "      <td>ENST00000684763.1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236686675</td>\n",
       "      <td>236764595</td>\n",
       "      <td>ENST00000682966</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236757521</td>\n",
       "      <td>236757557</td>\n",
       "      <td>rs1572148914</td>\n",
       "      <td>ATCGCCAGAACCATCAATGAGGTGGAGACTCAGATC</td>\n",
       "      <td>1</td>\n",
       "      <td>ATC,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41014758</th>\n",
       "      <td>chr1</td>\n",
       "      <td>236757445</td>\n",
       "      <td>236757524</td>\n",
       "      <td>ENST00000684763.1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236686682</td>\n",
       "      <td>236764595</td>\n",
       "      <td>ENST00000683111</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>236757521</td>\n",
       "      <td>236757557</td>\n",
       "      <td>rs1572148914</td>\n",
       "      <td>ATCGCCAGAACCATCAATGAGGTGGAGACTCAGATC</td>\n",
       "      <td>1</td>\n",
       "      <td>ATC,</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>175344 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0          1          2                   3  4     5          6   \\\n",
       "1061      chr1   24361002   24361081  ENST00000003583.12  -  chr1   24319425   \n",
       "1088      chr1   24361002   24361081  ENST00000003583.12  -  chr1   24357005   \n",
       "1115      chr1   24361002   24361081  ENST00000003583.12  -  chr1   24356999   \n",
       "1142      chr1   24361002   24361081  ENST00000003583.12  -  chr1   24356999   \n",
       "1169      chr1   24361002   24361081  ENST00000003583.12  -  chr1   24357005   \n",
       "...        ...        ...        ...                 ... ..   ...        ...   \n",
       "41014654  chr1  236757445  236757524   ENST00000684763.1  +  chr1  236686639   \n",
       "41014680  chr1  236757445  236757524   ENST00000684763.1  +  chr1  236686639   \n",
       "41014706  chr1  236757445  236757524   ENST00000684763.1  +  chr1  236686674   \n",
       "41014732  chr1  236757445  236757524   ENST00000684763.1  +  chr1  236686675   \n",
       "41014758  chr1  236757445  236757524   ENST00000684763.1  +  chr1  236686682   \n",
       "\n",
       "                 7                8   9     10         11         12  \\\n",
       "1061       24364482  ENST00000350501   1  chr1   24361078   24361086   \n",
       "1088       24413725  ENST00000003583  -1  chr1   24361078   24361086   \n",
       "1115       24413782  ENST00000337248  -1  chr1   24361078   24361086   \n",
       "1142       24415097  ENST00000374409  -1  chr1   24361078   24361086   \n",
       "1169       24416934  ENST00000468303  -1  chr1   24361078   24361086   \n",
       "...             ...              ...  ..   ...        ...        ...   \n",
       "41014654  236764595  ENST00000684050   1  chr1  236757521  236757557   \n",
       "41014680  236764595  ENST00000684502   1  chr1  236757521  236757557   \n",
       "41014706  236764595  ENST00000682692   1  chr1  236757521  236757557   \n",
       "41014732  236764595  ENST00000682966   1  chr1  236757521  236757557   \n",
       "41014758  236764595  ENST00000683111   1  chr1  236757521  236757557   \n",
       "\n",
       "                    13                                    14  15    16  \n",
       "1061      rs1641082067                              AGTCTAGT   1  AGT,  \n",
       "1088      rs1641082067                              AGTCTAGT   1  AGT,  \n",
       "1115      rs1641082067                              AGTCTAGT   1  AGT,  \n",
       "1142      rs1641082067                              AGTCTAGT   1  AGT,  \n",
       "1169      rs1641082067                              AGTCTAGT   1  AGT,  \n",
       "...                ...                                   ...  ..   ...  \n",
       "41014654  rs1572148914  ATCGCCAGAACCATCAATGAGGTGGAGACTCAGATC   1  ATC,  \n",
       "41014680  rs1572148914  ATCGCCAGAACCATCAATGAGGTGGAGACTCAGATC   1  ATC,  \n",
       "41014706  rs1572148914  ATCGCCAGAACCATCAATGAGGTGGAGACTCAGATC   1  ATC,  \n",
       "41014732  rs1572148914  ATCGCCAGAACCATCAATGAGGTGGAGACTCAGATC   1  ATC,  \n",
       "41014758  rs1572148914  ATCGCCAGAACCATCAATGAGGTGGAGACTCAGATC   1  ATC,  \n",
       "\n",
       "[175344 rows x 17 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df[12] >df[2]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b715fd56-b0f5-4c21-a6c8-1f584b3b73a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2953868</td>\n",
       "      <td>2953947</td>\n",
       "      <td>ENST00000155093.8</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2935381</td>\n",
       "      <td>2982506</td>\n",
       "      <td>ENST00000155093</td>\n",
       "      <td>1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2953879</td>\n",
       "      <td>2953880</td>\n",
       "      <td>rs2051285710</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>C,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2953868</td>\n",
       "      <td>2953947</td>\n",
       "      <td>ENST00000155093.8</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2935281</td>\n",
       "      <td>2982506</td>\n",
       "      <td>ENST00000383052</td>\n",
       "      <td>1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2953879</td>\n",
       "      <td>2953880</td>\n",
       "      <td>rs2051285710</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>C,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2953868</td>\n",
       "      <td>2953947</td>\n",
       "      <td>ENST00000155093.8</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2935505</td>\n",
       "      <td>2961286</td>\n",
       "      <td>ENST00000443793</td>\n",
       "      <td>1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2953879</td>\n",
       "      <td>2953880</td>\n",
       "      <td>rs2051285710</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>C,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2953868</td>\n",
       "      <td>2953947</td>\n",
       "      <td>ENST00000155093.8</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2935651</td>\n",
       "      <td>2981146</td>\n",
       "      <td>ENST00000449237</td>\n",
       "      <td>1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2953879</td>\n",
       "      <td>2953880</td>\n",
       "      <td>rs2051285710</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>C,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2953868</td>\n",
       "      <td>2953947</td>\n",
       "      <td>ENST00000155093.8</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2935500</td>\n",
       "      <td>2978053</td>\n",
       "      <td>ENST00000469869</td>\n",
       "      <td>1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2953879</td>\n",
       "      <td>2953880</td>\n",
       "      <td>rs2051285710</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>C,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107283</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13234787</td>\n",
       "      <td>13234866</td>\n",
       "      <td>ENST00000684326.1</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>13234599</td>\n",
       "      <td>13336129</td>\n",
       "      <td>ENST00000682761</td>\n",
       "      <td>-1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>13234794</td>\n",
       "      <td>13234795</td>\n",
       "      <td>rs1422913673</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>A,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107284</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13234787</td>\n",
       "      <td>13234866</td>\n",
       "      <td>ENST00000684326.1</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>13234599</td>\n",
       "      <td>13336129</td>\n",
       "      <td>ENST00000682761</td>\n",
       "      <td>-1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>13234797</td>\n",
       "      <td>13234798</td>\n",
       "      <td>rs1162204554</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>A,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107285</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13234787</td>\n",
       "      <td>13234866</td>\n",
       "      <td>ENST00000684326.1</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>13234599</td>\n",
       "      <td>13336129</td>\n",
       "      <td>ENST00000682761</td>\n",
       "      <td>-1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>13234801</td>\n",
       "      <td>13234802</td>\n",
       "      <td>rs1355901236</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>C,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107286</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13234787</td>\n",
       "      <td>13234866</td>\n",
       "      <td>ENST00000684326.1</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>13234599</td>\n",
       "      <td>13336129</td>\n",
       "      <td>ENST00000682761</td>\n",
       "      <td>-1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>13234814</td>\n",
       "      <td>13234815</td>\n",
       "      <td>rs1462345242</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>A,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107287</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13234787</td>\n",
       "      <td>13234866</td>\n",
       "      <td>ENST00000684326.1</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>13234599</td>\n",
       "      <td>13336129</td>\n",
       "      <td>ENST00000682761</td>\n",
       "      <td>-1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>13234851</td>\n",
       "      <td>13234852</td>\n",
       "      <td>rs1603217354</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>A,</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>107288 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0         1         2                  3  4     5         6   \\\n",
       "0       chrY   2953868   2953947  ENST00000155093.8  +  chrY   2935381   \n",
       "1       chrY   2953868   2953947  ENST00000155093.8  +  chrY   2935281   \n",
       "2       chrY   2953868   2953947  ENST00000155093.8  +  chrY   2935505   \n",
       "3       chrY   2953868   2953947  ENST00000155093.8  +  chrY   2935651   \n",
       "4       chrY   2953868   2953947  ENST00000155093.8  +  chrY   2935500   \n",
       "...      ...       ...       ...                ... ..   ...       ...   \n",
       "107283  chrY  13234787  13234866  ENST00000684326.1  -  chrY  13234599   \n",
       "107284  chrY  13234787  13234866  ENST00000684326.1  -  chrY  13234599   \n",
       "107285  chrY  13234787  13234866  ENST00000684326.1  -  chrY  13234599   \n",
       "107286  chrY  13234787  13234866  ENST00000684326.1  -  chrY  13234599   \n",
       "107287  chrY  13234787  13234866  ENST00000684326.1  -  chrY  13234599   \n",
       "\n",
       "              7                8   9     10        11        12            13  \\\n",
       "0        2982506  ENST00000155093   1  chrY   2953879   2953880  rs2051285710   \n",
       "1        2982506  ENST00000383052   1  chrY   2953879   2953880  rs2051285710   \n",
       "2        2961286  ENST00000443793   1  chrY   2953879   2953880  rs2051285710   \n",
       "3        2981146  ENST00000449237   1  chrY   2953879   2953880  rs2051285710   \n",
       "4        2978053  ENST00000469869   1  chrY   2953879   2953880  rs2051285710   \n",
       "...          ...              ...  ..   ...       ...       ...           ...   \n",
       "107283  13336129  ENST00000682761  -1  chrY  13234794  13234795  rs1422913673   \n",
       "107284  13336129  ENST00000682761  -1  chrY  13234797  13234798  rs1162204554   \n",
       "107285  13336129  ENST00000682761  -1  chrY  13234801  13234802  rs1355901236   \n",
       "107286  13336129  ENST00000682761  -1  chrY  13234814  13234815  rs1462345242   \n",
       "107287  13336129  ENST00000682761  -1  chrY  13234851  13234852  rs1603217354   \n",
       "\n",
       "       14  15  16  \n",
       "0       T   1  C,  \n",
       "1       T   1  C,  \n",
       "2       T   1  C,  \n",
       "3       T   1  C,  \n",
       "4       T   1  C,  \n",
       "...    ..  ..  ..  \n",
       "107283  G   1  A,  \n",
       "107284  G   1  A,  \n",
       "107285  G   1  C,  \n",
       "107286  C   1  A,  \n",
       "107287  T   1  A,  \n",
       "\n",
       "[107288 rows x 17 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df[df[15]<3].reset_index(drop=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "05335d16-3279-4a6d-80e6-19614322445e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>ref_sequence</th>\n",
       "      <th>alt_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7230758</td>\n",
       "      <td>7230837</td>\n",
       "      <td>ENST00000005340.10</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7228007</td>\n",
       "      <td>7234061</td>\n",
       "      <td>ENST00000574143</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7230760</td>\n",
       "      <td>7230761</td>\n",
       "      <td>rs1296032425</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>A,</td>\n",
       "      <td>AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "      <td>AAGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7230758</td>\n",
       "      <td>7230837</td>\n",
       "      <td>ENST00000005340.10</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7228007</td>\n",
       "      <td>7234061</td>\n",
       "      <td>ENST00000574143</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7230761</td>\n",
       "      <td>7230762</td>\n",
       "      <td>rs143043931</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>T,</td>\n",
       "      <td>AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "      <td>AGTCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7230758</td>\n",
       "      <td>7230837</td>\n",
       "      <td>ENST00000005340.10</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7228007</td>\n",
       "      <td>7234061</td>\n",
       "      <td>ENST00000574143</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7230762</td>\n",
       "      <td>7230763</td>\n",
       "      <td>rs758501117</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>A,</td>\n",
       "      <td>AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "      <td>AGGAGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7230758</td>\n",
       "      <td>7230837</td>\n",
       "      <td>ENST00000005340.10</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7228007</td>\n",
       "      <td>7234061</td>\n",
       "      <td>ENST00000574143</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7230764</td>\n",
       "      <td>7230765</td>\n",
       "      <td>rs1264988936</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>T,</td>\n",
       "      <td>AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "      <td>AGGCGTGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7230758</td>\n",
       "      <td>7230837</td>\n",
       "      <td>ENST00000005340.10</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7228007</td>\n",
       "      <td>7234061</td>\n",
       "      <td>ENST00000574143</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7230765</td>\n",
       "      <td>7230766</td>\n",
       "      <td>rs1282730801</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>T,</td>\n",
       "      <td>AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "      <td>AGGCGGTCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7230758</td>\n",
       "      <td>7230837</td>\n",
       "      <td>ENST00000005340.10</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7228007</td>\n",
       "      <td>7234061</td>\n",
       "      <td>ENST00000574143</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7230766</td>\n",
       "      <td>7230767</td>\n",
       "      <td>rs766636917</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>A,</td>\n",
       "      <td>AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "      <td>AGGCGGGAGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7230758</td>\n",
       "      <td>7230837</td>\n",
       "      <td>ENST00000005340.10</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7228007</td>\n",
       "      <td>7234061</td>\n",
       "      <td>ENST00000574143</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7230767</td>\n",
       "      <td>7230768</td>\n",
       "      <td>rs2071530777</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>C,</td>\n",
       "      <td>AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "      <td>AGGCGGGCCTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7230758</td>\n",
       "      <td>7230837</td>\n",
       "      <td>ENST00000005340.10</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7228007</td>\n",
       "      <td>7234061</td>\n",
       "      <td>ENST00000574143</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7230773</td>\n",
       "      <td>7230774</td>\n",
       "      <td>rs1215823324</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>C,</td>\n",
       "      <td>AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "      <td>AGGCGGGCGTTGTCCTCTGAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7230758</td>\n",
       "      <td>7230837</td>\n",
       "      <td>ENST00000005340.10</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7228007</td>\n",
       "      <td>7234061</td>\n",
       "      <td>ENST00000574143</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7230775</td>\n",
       "      <td>7230776</td>\n",
       "      <td>rs1253332796</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>C,</td>\n",
       "      <td>AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "      <td>AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7230758</td>\n",
       "      <td>7230837</td>\n",
       "      <td>ENST00000005340.10</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7228007</td>\n",
       "      <td>7234061</td>\n",
       "      <td>ENST00000574143</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7230777</td>\n",
       "      <td>7230778</td>\n",
       "      <td>rs989449469</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>T,</td>\n",
       "      <td>AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "      <td>AGGCGGGCGTTGTCATCTTAAATTTCTTCCTTCACCACCCTGCCAA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       0        1        2                   3  4      5        6        7  \\\n",
       "0  chr17  7230758  7230837  ENST00000005340.10  -  chr17  7228007  7234061   \n",
       "1  chr17  7230758  7230837  ENST00000005340.10  -  chr17  7228007  7234061   \n",
       "2  chr17  7230758  7230837  ENST00000005340.10  -  chr17  7228007  7234061   \n",
       "3  chr17  7230758  7230837  ENST00000005340.10  -  chr17  7228007  7234061   \n",
       "4  chr17  7230758  7230837  ENST00000005340.10  -  chr17  7228007  7234061   \n",
       "5  chr17  7230758  7230837  ENST00000005340.10  -  chr17  7228007  7234061   \n",
       "6  chr17  7230758  7230837  ENST00000005340.10  -  chr17  7228007  7234061   \n",
       "7  chr17  7230758  7230837  ENST00000005340.10  -  chr17  7228007  7234061   \n",
       "8  chr17  7230758  7230837  ENST00000005340.10  -  chr17  7228007  7234061   \n",
       "9  chr17  7230758  7230837  ENST00000005340.10  -  chr17  7228007  7234061   \n",
       "\n",
       "                 8  9     10       11       12            13 14  15  16  \\\n",
       "0  ENST00000574143 -1  chr17  7230760  7230761  rs1296032425  G   1  A,   \n",
       "1  ENST00000574143 -1  chr17  7230761  7230762   rs143043931  C   1  T,   \n",
       "2  ENST00000574143 -1  chr17  7230762  7230763   rs758501117  G   1  A,   \n",
       "3  ENST00000574143 -1  chr17  7230764  7230765  rs1264988936  G   1  T,   \n",
       "4  ENST00000574143 -1  chr17  7230765  7230766  rs1282730801  C   1  T,   \n",
       "5  ENST00000574143 -1  chr17  7230766  7230767   rs766636917  G   1  A,   \n",
       "6  ENST00000574143 -1  chr17  7230767  7230768  rs2071530777  T   1  C,   \n",
       "7  ENST00000574143 -1  chr17  7230773  7230774  rs1215823324  T   1  C,   \n",
       "8  ENST00000574143 -1  chr17  7230775  7230776  rs1253332796  T   1  C,   \n",
       "9  ENST00000574143 -1  chr17  7230777  7230778   rs989449469  A   1  T,   \n",
       "\n",
       "                                        ref_sequence  \\\n",
       "0  AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...   \n",
       "1  AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...   \n",
       "2  AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...   \n",
       "3  AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...   \n",
       "4  AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...   \n",
       "5  AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...   \n",
       "6  AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...   \n",
       "7  AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...   \n",
       "8  AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...   \n",
       "9  AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...   \n",
       "\n",
       "                                        alt_sequence  \n",
       "0  AAGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...  \n",
       "1  AGTCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...  \n",
       "2  AGGAGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...  \n",
       "3  AGGCGTGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...  \n",
       "4  AGGCGGTCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...  \n",
       "5  AGGCGGGAGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...  \n",
       "6  AGGCGGGCCTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...  \n",
       "7  AGGCGGGCGTTGTCCTCTGAAATTTCTTCCTTCACCACCCTGCCAA...  \n",
       "8  AGGCGGGCGTTGTCATCTGAAATTTCTTCCTTCACCACCCTGCCAA...  \n",
       "9  AGGCGGGCGTTGTCATCTTAAATTTCTTCCTTCACCACCCTGCCAA...  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "23bc69f0-45f3-4840-8025-f272363c47d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/projects/Resources/HumanReferenceGenome/chr5.fa\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "single positional indexer is out-of-bounds",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[60], line 5\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28mprint\u001b[39m(reference_path\u001b[38;5;241m+\u001b[39mintersect_file\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_\u001b[39m\u001b[38;5;124m'\u001b[39m)[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m2\u001b[39m]\u001b[38;5;241m+\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m.fa\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# This is just for the first row (or the first variant) as an example\u001b[39;00m\n\u001b[0;32m----> 5\u001b[0m row \u001b[38;5;241m=\u001b[39m \u001b[43mdf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43miloc\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m344\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28mprint\u001b[39m(row[\u001b[38;5;241m0\u001b[39m], row[\u001b[38;5;241m1\u001b[39m], row[\u001b[38;5;241m2\u001b[39m], row[\u001b[38;5;241m16\u001b[39m])\n\u001b[1;32m      7\u001b[0m \u001b[38;5;66;03m# Fetch the reference sequence\u001b[39;00m\n",
      "File \u001b[0;32m/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pandas/core/indexing.py:1073\u001b[0m, in \u001b[0;36m_LocationIndexer.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1070\u001b[0m axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxis \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m0\u001b[39m\n\u001b[1;32m   1072\u001b[0m maybe_callable \u001b[38;5;241m=\u001b[39m com\u001b[38;5;241m.\u001b[39mapply_if_callable(key, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj)\n\u001b[0;32m-> 1073\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_axis\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmaybe_callable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pandas/core/indexing.py:1625\u001b[0m, in \u001b[0;36m_iLocIndexer._getitem_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1622\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot index by location index with a non-integer key\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   1624\u001b[0m \u001b[38;5;66;03m# validate the location\u001b[39;00m\n\u001b[0;32m-> 1625\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_integer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1627\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39m_ixs(key, axis\u001b[38;5;241m=\u001b[39maxis)\n",
      "File \u001b[0;32m/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pandas/core/indexing.py:1557\u001b[0m, in \u001b[0;36m_iLocIndexer._validate_integer\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1555\u001b[0m len_axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39m_get_axis(axis))\n\u001b[1;32m   1556\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m key \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m len_axis \u001b[38;5;129;01mor\u001b[39;00m key \u001b[38;5;241m<\u001b[39m \u001b[38;5;241m-\u001b[39mlen_axis:\n\u001b[0;32m-> 1557\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mIndexError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msingle positional indexer is out-of-bounds\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mIndexError\u001b[0m: single positional indexer is out-of-bounds"
     ]
    }
   ],
   "source": [
    "# Load the reference genome\n",
    "genome = pysam.FastaFile(reference_path+chromosome_name+\".fa\")\n",
    "print(reference_path+intersect_file.split('_')[-2]+\".fa\")\n",
    "# This is just for the first row (or the first variant) as an example\n",
    "row = df.iloc[344]\n",
    "print(row[0], row[1], row[2], row[16])\n",
    "# Fetch the reference sequence\n",
    "ref_sequence = genome.fetch(row[0], row[1], row[2])\n",
    "\n",
    "# Now replace the reference nucleotide at the VCF position with the alternative nucleotide to get the alternative sequence\n",
    "alt_sequence = ref_sequence[:row[11] - row[1] - 1] + row[16].replace(\",\", \"\") + ref_sequence[row[11] - row[1]:]\n",
    "\n",
    "print('Ref:', ref_sequence)\n",
    "print('Alt:', alt_sequence)\n",
    "\n",
    "genome.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "19d03b7f-2bac-40c5-831a-7eff70b58805",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1, 'a', 2, 'b', 3, 'c', 4, 'd', 5, 'e']\n"
     ]
    }
   ],
   "source": [
    "list1 = [1, 2, 3, 4, 5]\n",
    "list2 = ['a', 'b', 'c', 'd', 'e']\n",
    "\n",
    "# Merge lists so the result contains alternating elements from both lists\n",
    "merged_list = [i for pair in zip(list1, list2) for i in pair]\n",
    "\n",
    "print(merged_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "edbe9391-aeaf-4fc9-81a5-0ca53dae8c73",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACCATGTCCACCTTTATGCTTTTTAAAGTGAAAAACCATACTAAGAATGAGGCAGCTCAACTTAATAATAAAAACATTTCAAATGTAAAGAAATTTACAA\n",
      "ACGATGTCCACCTTTATGCTTTTTAAAGTGAAAAACCATACTAAGAATGAGGCAGCTCAACTTAATAATAAAAACATTTCAAATGTAAAGAAATTTACAA\n",
      "2\n",
      "GGCAGGCATGGGAAATACAGTGCGTTAAAGGTAGGTTCAGATTGAGTGGAGCAGCTTTAAAGGACTAAGTGTCTGGCCAGTCTTCAATAAACATGAAACT\n",
      "GTAGGCATGGGAAATACAGTGCGTTAAAGGTAGGTTCAGATTGAGTGGAGCAGCTTTAAAGGACTAAGTGTCTGGCCAGTCTTCAATAAACATGAAACT\n",
      "1\n",
      "AAGAGTAGGAAAACATGACAGATTGGGAGGTGTTGAGTTCTGAATTTGGGGACCTAATTGAAATGCAAAGGTATAATCAAGAGAATGAACTTGTCTCTGT\n",
      "ATGGAGTAGGAAAACATGACAGATTGGGAGGTGTTGAGTTCTGAATTTGGGGACCTAATTGAAATGCAAAGGTATAATCAAGAGAATGAACTTGTCTCTGT\n",
      "1\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import pysam\n",
    "\n",
    "# Sample DataFrame\n",
    "df = pd.DataFrame({\n",
    "    'chromosome': ['chr1', 'chr2', 'chr3'],\n",
    "    'start': [110000, 200200, 200300],\n",
    "    'end': [110100, 200300, 200400],\n",
    "    'VCF_position': [110002, 200201, 200301],\n",
    "    'ref_nucleotide': ['C', 'GC', 'A'],\n",
    "    'alt_nucleotide': ['G', 'T', 'TG']\n",
    "})\n",
    "\n",
    "# Load the reference genome\n",
    "genome = pysam.FastaFile(\"/data/projects/Resources/Gencode_genome_annotation/GRCh38.primary_assembly.genome.fa\")\n",
    "\n",
    "# Store the sequences in lists\n",
    "ref_sequences = []\n",
    "alt_sequences = []\n",
    "\n",
    "for _, row in df.iterrows():\n",
    "    # Fetch the reference sequence from start to end\n",
    "    ref_sequence = genome.fetch(row['chromosome'], row['start'], row['end'])\n",
    "    print(ref_sequence)\n",
    "    \n",
    "    # Calculate the variant position relative to the fetched sequence\n",
    "    variant_pos = row['VCF_position'] - row['start']\n",
    "    \n",
    "\n",
    "    # Replace the reference nucleotide with the alternate nucleotide to get the alternate sequence\n",
    "    alt_sequence = ref_sequence[:variant_pos] + row['alt_nucleotide'] + ref_sequence[variant_pos + len(row['ref_nucleotide']):]\n",
    "    print(alt_sequence)\n",
    "    print(variant_pos)\n",
    "    \n",
    "    # Store the sequences\n",
    "    ref_sequences.append(ref_sequence)\n",
    "    alt_sequences.append(alt_sequence)\n",
    "\n",
    "# Add the sequences to the dataframe\n",
    "df['ref_sequence'] = ref_sequences\n",
    "df['alt_sequence'] = alt_sequences\n",
    "\n",
    "# Print the updated dataframe\n",
    "\n",
    "genome.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "63736a71-9f75-43df-a998-f53109a1dcf4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>VCF_position</th>\n",
       "      <th>ref_nucleotide</th>\n",
       "      <th>alt_nucleotide</th>\n",
       "      <th>ref_sequence</th>\n",
       "      <th>alt_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100000</td>\n",
       "      <td>100200</td>\n",
       "      <td>100102</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>ACTAAGCACACAGAGAATAATGTCTAGAATCTGAGTGCCATGTTAT...</td>\n",
       "      <td>ACTAAGCACACAGAGAATAATGTCTAGAATCTGAGTGCCATGTTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>200200</td>\n",
       "      <td>200300</td>\n",
       "      <td>200201</td>\n",
       "      <td>GC</td>\n",
       "      <td>T</td>\n",
       "      <td>GGCAGGCATGGGAAATACAGTGCGTTAAAGGTAGGTTCAGATTGAG...</td>\n",
       "      <td>GTAGGCATGGGAAATACAGTGCGTTAAAGGTAGGTTCAGATTGAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr3</td>\n",
       "      <td>200300</td>\n",
       "      <td>200400</td>\n",
       "      <td>200301</td>\n",
       "      <td>A</td>\n",
       "      <td>TG</td>\n",
       "      <td>AAGAGTAGGAAAACATGACAGATTGGGAGGTGTTGAGTTCTGAATT...</td>\n",
       "      <td>ATGGAGTAGGAAAACATGACAGATTGGGAGGTGTTGAGTTCTGAAT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome   start     end  VCF_position ref_nucleotide alt_nucleotide  \\\n",
       "0       chr1  100000  100200        100102              T              G   \n",
       "1       chr2  200200  200300        200201             GC              T   \n",
       "2       chr3  200300  200400        200301              A             TG   \n",
       "\n",
       "                                        ref_sequence  \\\n",
       "0  ACTAAGCACACAGAGAATAATGTCTAGAATCTGAGTGCCATGTTAT...   \n",
       "1  GGCAGGCATGGGAAATACAGTGCGTTAAAGGTAGGTTCAGATTGAG...   \n",
       "2  AAGAGTAGGAAAACATGACAGATTGGGAGGTGTTGAGTTCTGAATT...   \n",
       "\n",
       "                                        alt_sequence  \n",
       "0  ACTAAGCACACAGAGAATAATGTCTAGAATCTGAGTGCCATGTTAT...  \n",
       "1  GTAGGCATGGGAAATACAGTGCGTTAAAGGTAGGTTCAGATTGAGT...  \n",
       "2  ATGGAGTAGGAAAACATGACAGATTGGGAGGTGTTGAGTTCTGAAT...  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "629db8ab-bf00-45bf-8c54-59e29e507193",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "646941b5-affe-449d-8e9b-be2a13c629a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the genome\n",
    "genome = SeqIO.to_dict(SeqIO.parse(\"/data/projects/Resources/Gencode_genome_annotation/GRCh38.primary_assembly.genome.fa\", \"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "7b6a9e95-f2c4-4ba0-adb0-567d6ea617e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sequences(df):\n",
    "    sequences = []\n",
    "    for _, row in df.iterrows():\n",
    "        chrom = row['chromosome-number']\n",
    "        ref_start = row['varinat_start']\n",
    "        ref_end = row['variant_end']\n",
    "        ref_nucleotide = row['ref_neucleotide']\n",
    "        alts = row['alternative_neucleotide']\n",
    "        number_of_alts = row['number_of_alts']\n",
    "        \n",
    "        # Get reference sequence\n",
    "        ref_seq = str(genome[chrom].seq[ref_start:ref_end])\n",
    "\n",
    "        # Ensure the reference sequence matches what we expect\n",
    "        if ref_seq != ref_nucleotide:\n",
    "            print(f\"Warning: expected reference sequence {ref_nucleotide}, but got {ref_seq}\")\n",
    "\n",
    "        # Replace reference nucleotide with each alternative\n",
    "        for i in range(number_of_alts):\n",
    "            alt_seq = ref_seq.replace(ref_nucleotide, alts.split(',')[i])\n",
    "            sequences.append((ref_seq, alt_seq))\n",
    "\n",
    "    return sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "bfe2aac1-411f-401f-a90d-563414437dda",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Your dataframe (sample)\n",
    "df1 = pd.DataFrame({\n",
    "    'chromosome-number': ['chr1', 'chr1', 'chr1', 'chr1', 'chr1'],\n",
    "    'chromosome_start': [24401417, 24383964, 234429638, 67049811, 85183161],\n",
    "    'chromosome_end': [24401496, 24384043, 234429717, 67049890, 85183240],\n",
    "    'varinat_start': [24401424, 24383964, 234429682, 67049880, 85183235],\n",
    "    'variant_end': [24401425, 24383965, 234429684, 67049882, 85183239],\n",
    "    'ref_neucleotide': ['T', 'G', 'AT', 'AT', 'ATAT'],\n",
    "    'number_of_alts': [1, 2, 1,1,2],\n",
    "    'alternative_neucleotide': ['C,', 'C,T,', 'ATAT,', ',', 'AT,CGT,']\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "c932f0e5-0236-4b2e-b585-6f978ffb9cf8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome-number</th>\n",
       "      <th>chromosome_start</th>\n",
       "      <th>chromosome_end</th>\n",
       "      <th>varinat_start</th>\n",
       "      <th>variant_end</th>\n",
       "      <th>ref_neucleotide</th>\n",
       "      <th>number_of_alts</th>\n",
       "      <th>alternative_neucleotide</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24401417</td>\n",
       "      <td>24401496</td>\n",
       "      <td>24401424</td>\n",
       "      <td>24401425</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>C,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24383964</td>\n",
       "      <td>24384043</td>\n",
       "      <td>24383964</td>\n",
       "      <td>24383965</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>C,T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>234429638</td>\n",
       "      <td>234429717</td>\n",
       "      <td>234429682</td>\n",
       "      <td>234429684</td>\n",
       "      <td>AT</td>\n",
       "      <td>1</td>\n",
       "      <td>ATAT,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67049811</td>\n",
       "      <td>67049890</td>\n",
       "      <td>67049880</td>\n",
       "      <td>67049882</td>\n",
       "      <td>AT</td>\n",
       "      <td>1</td>\n",
       "      <td>,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>85183161</td>\n",
       "      <td>85183240</td>\n",
       "      <td>85183235</td>\n",
       "      <td>85183239</td>\n",
       "      <td>ATAT</td>\n",
       "      <td>2</td>\n",
       "      <td>AT,CGT,</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome-number  chromosome_start  chromosome_end  varinat_start  \\\n",
       "0              chr1          24401417        24401496       24401424   \n",
       "1              chr1          24383964        24384043       24383964   \n",
       "2              chr1         234429638       234429717      234429682   \n",
       "3              chr1          67049811        67049890       67049880   \n",
       "4              chr1          85183161        85183240       85183235   \n",
       "\n",
       "   variant_end ref_neucleotide  number_of_alts alternative_neucleotide  \n",
       "0     24401425               T               1                      C,  \n",
       "1     24383965               G               2                    C,T,  \n",
       "2    234429684              AT               1                   ATAT,  \n",
       "3     67049882              AT               1                       ,  \n",
       "4     85183239            ATAT               2                 AT,CGT,  "
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "26fbacd5-6606-4663-a9df-deabe35ccdd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sequences(df):\n",
    "    sequences = []\n",
    "    for _, row in df.iterrows():\n",
    "        chrom = row['chromosome-number']\n",
    "        ref_start = row['chromosome_start']\n",
    "        ref_end = row['chromosome_end']\n",
    "        variant_start = row['varinat_start']\n",
    "        variant_end = row['variant_end']\n",
    "        ref_nucleotide = row['ref_neucleotide']\n",
    "        alts = row['alternative_neucleotide'].rstrip(',').split(',')\n",
    "        \n",
    "        # Get reference sequence\n",
    "        ref_seq = str(genome[chrom].seq[ref_start:ref_end])\n",
    "\n",
    "        # Adjust for 0-based indexing in python\n",
    "        variant_pos_start = variant_start - ref_start\n",
    "        variant_pos_end = variant_end - ref_start\n",
    "\n",
    "        # Replace reference nucleotide with each alternative\n",
    "        for alt in alts:\n",
    "            if alt == \"\":  # Deletion\n",
    "                # Fetch additional bases from genome to maintain sequence length\n",
    "                extra_bases = str(genome[chrom].seq[ref_end:ref_end + len(ref_nucleotide)])\n",
    "                alt_seq = ref_seq[:variant_pos_start] + ref_seq[variant_pos_end:] + extra_bases\n",
    "            else:  # SNPs, Insertions, and partial Deletions\n",
    "                extra_bases = str(genome[chrom].seq[ref_end:ref_end + len(ref_nucleotide) - len(alt)])\n",
    "                alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:] + extra_bases\n",
    "            sequences.append((ref_seq, alt_seq))\n",
    "\n",
    "    return sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "f243c0a7-06c6-4e73-a00f-15cf08f36cec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TGTTTGATGAAAAGTTCTACTGCATGTTCTCCTAAGCACCTGAAACAGCAAAACACAGCATTTGTAGAGATCATTTCAC\n",
      "TGTTTGACGAAAAGTTCTACTGCATGTTCTCCTAAGCACCTGAAACAGCAAAACACAGCATTTGTAGAGATCATTTCAC \n",
      "\n",
      "GTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAGGGAAGAGAAGGTGGTGTCATCGAGATACTTCAA\n",
      "CTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAGGGAAGAGAAGGTGGTGTCATCGAGATACTTCAA \n",
      "\n",
      "GTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAGGGAAGAGAAGGTGGTGTCATCGAGATACTTCAA\n",
      "TTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAGGGAAGAGAAGGTGGTGTCATCGAGATACTTCAA \n",
      "\n",
      "TTTCCCCATCCTTGGGAAGATGACGATTCTTCCAAAACACTGAAATAAAAAATAAAGTTACTAGGCCATACTTCCCCAA\n",
      "TTTCCCCATCCTTGGGAAGATGACGATTCTTCCAAAACACTGAAATATAAAAAATAAAGTTACTAGGCCATACTTCCCCAA \n",
      "\n",
      "GCAAATACAGTCATTTTAATACCCCAAGAAAAAGTCTTCCTACAAAACAAAAAATTTTAAAATACACACATGACTTTAT\n",
      "GCAAATACAGTCATTTTAATACCCCAAGAAAAAGTCTTCCTACAAAACAAAAAATTTTAAAATACACACGACTTTATTC \n",
      "\n",
      "CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAAGAAAGAAAGAAAAATACGTTATAAAGCAATATG\n",
      "CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAAGAAAGAAAGAAAAATACGTTATAAAGCAATGTT \n",
      "\n",
      "CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAAGAAAGAAAGAAAAATACGTTATAAAGCAATATG\n",
      "CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAAGAAAGAAAGAAAAATACGTTATAAAGCACGTGT \n",
      "\n"
     ]
    }
   ],
   "source": [
    "sequences = get_sequences(df1)\n",
    "for seq in sequences:\n",
    "    print(seq[0])\n",
    "    print(seq[1],\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "fd0be8c5-35de-49f1-ba3e-590629b6cab8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sequences_new(df):\n",
    "    new_df = pd.DataFrame(columns=['reference_id', 'varinat_start', 'variant_end', \n",
    "                                   'ref_neucleotide', 'alternative_neucleotide', \n",
    "                                   'reference_seq', 'alt_seq'])\n",
    "\n",
    "    for idx, row in df.iterrows():\n",
    "        chrom = row['chromosome-number']\n",
    "        ref_start = row['chromosome_start']\n",
    "        ref_end = row['chromosome_end']\n",
    "        variant_start = row['varinat_start']\n",
    "        variant_end = row['variant_end']\n",
    "        ref_nucleotide = row['ref_neucleotide']\n",
    "        alts = row['alternative_neucleotide'].rstrip(',').split(',')\n",
    "        \n",
    "        # Get reference sequence\n",
    "        ref_seq = str(genome[chrom].seq[ref_start:ref_end])\n",
    "\n",
    "        # Adjust for 0-based indexing in python\n",
    "        variant_pos_start = variant_start - ref_start\n",
    "        variant_pos_end = variant_end - ref_start\n",
    "\n",
    "        # Replace reference nucleotide with each alternative\n",
    "        for alt in alts:\n",
    "            if alt == \"\":  # Deletion\n",
    "                # Fetch additional bases from genome to maintain sequence length\n",
    "                extra_bases = str(genome[chrom].seq[ref_end:ref_end + len(ref_nucleotide)])\n",
    "                alt_seq = ref_seq[:variant_pos_start] + ref_seq[variant_pos_end:] + extra_bases\n",
    "            else:  # SNPs, Insertions, and partial Deletions\n",
    "                extra_bases = str(genome[chrom].seq[ref_end:ref_end + len(ref_nucleotide) - len(alt)])\n",
    "                alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:] + extra_bases\n",
    "                \n",
    "            new_df = new_df.append({'reference_id': chrom,\n",
    "                                    'varinat_start': variant_start,\n",
    "                                    'variant_end': variant_end,\n",
    "                                    'ref_neucleotide': ref_nucleotide,\n",
    "                                    'alternative_neucleotide': alt,\n",
    "                                    'reference_seq': ref_seq,\n",
    "                                    'alt_seq': alt_seq}, ignore_index=True)\n",
    "            \n",
    "    return new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "153d29fb-9bd6-40fe-8c7d-8d4a5dd0c867",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/private/pdutta/PSB_Data/ipykernel_1702645/1713894418.py:32: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  new_df = new_df.append({'reference_id': chrom,\n",
      "/data/private/pdutta/PSB_Data/ipykernel_1702645/1713894418.py:32: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  new_df = new_df.append({'reference_id': chrom,\n",
      "/data/private/pdutta/PSB_Data/ipykernel_1702645/1713894418.py:32: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  new_df = new_df.append({'reference_id': chrom,\n",
      "/data/private/pdutta/PSB_Data/ipykernel_1702645/1713894418.py:32: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  new_df = new_df.append({'reference_id': chrom,\n",
      "/data/private/pdutta/PSB_Data/ipykernel_1702645/1713894418.py:32: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  new_df = new_df.append({'reference_id': chrom,\n",
      "/data/private/pdutta/PSB_Data/ipykernel_1702645/1713894418.py:32: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  new_df = new_df.append({'reference_id': chrom,\n",
      "/data/private/pdutta/PSB_Data/ipykernel_1702645/1713894418.py:32: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  new_df = new_df.append({'reference_id': chrom,\n"
     ]
    }
   ],
   "source": [
    "new_df = get_sequences_new(df1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "97fdb342-1d66-4e03-803b-bab14e4de0d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference_id</th>\n",
       "      <th>varinat_start</th>\n",
       "      <th>variant_end</th>\n",
       "      <th>ref_neucleotide</th>\n",
       "      <th>alternative_neucleotide</th>\n",
       "      <th>reference_seq</th>\n",
       "      <th>alt_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24401424</td>\n",
       "      <td>24401425</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>TGTTTGATGAAAAGTTCTACTGCATGTTCTCCTAAGCACCTGAAAC...</td>\n",
       "      <td>TGTTTGACGAAAAGTTCTACTGCATGTTCTCCTAAGCACCTGAAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24383964</td>\n",
       "      <td>24383965</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>GTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAG...</td>\n",
       "      <td>CTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24383964</td>\n",
       "      <td>24383965</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>GTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAG...</td>\n",
       "      <td>TTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>234429682</td>\n",
       "      <td>234429684</td>\n",
       "      <td>AT</td>\n",
       "      <td>ATAT</td>\n",
       "      <td>TTTCCCCATCCTTGGGAAGATGACGATTCTTCCAAAACACTGAAAT...</td>\n",
       "      <td>TTTCCCCATCCTTGGGAAGATGACGATTCTTCCAAAACACTGAAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67049880</td>\n",
       "      <td>67049882</td>\n",
       "      <td>AT</td>\n",
       "      <td></td>\n",
       "      <td>GCAAATACAGTCATTTTAATACCCCAAGAAAAAGTCTTCCTACAAA...</td>\n",
       "      <td>GCAAATACAGTCATTTTAATACCCCAAGAAAAAGTCTTCCTACAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>85183235</td>\n",
       "      <td>85183239</td>\n",
       "      <td>ATAT</td>\n",
       "      <td>AT</td>\n",
       "      <td>CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAA...</td>\n",
       "      <td>CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>85183235</td>\n",
       "      <td>85183239</td>\n",
       "      <td>ATAT</td>\n",
       "      <td>CGT</td>\n",
       "      <td>CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAA...</td>\n",
       "      <td>CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  reference_id varinat_start variant_end ref_neucleotide  \\\n",
       "0         chr1      24401424    24401425               T   \n",
       "1         chr1      24383964    24383965               G   \n",
       "2         chr1      24383964    24383965               G   \n",
       "3         chr1     234429682   234429684              AT   \n",
       "4         chr1      67049880    67049882              AT   \n",
       "5         chr1      85183235    85183239            ATAT   \n",
       "6         chr1      85183235    85183239            ATAT   \n",
       "\n",
       "  alternative_neucleotide                                      reference_seq  \\\n",
       "0                       C  TGTTTGATGAAAAGTTCTACTGCATGTTCTCCTAAGCACCTGAAAC...   \n",
       "1                       C  GTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAG...   \n",
       "2                       T  GTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAG...   \n",
       "3                    ATAT  TTTCCCCATCCTTGGGAAGATGACGATTCTTCCAAAACACTGAAAT...   \n",
       "4                          GCAAATACAGTCATTTTAATACCCCAAGAAAAAGTCTTCCTACAAA...   \n",
       "5                      AT  CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAA...   \n",
       "6                     CGT  CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAA...   \n",
       "\n",
       "                                             alt_seq  \n",
       "0  TGTTTGACGAAAAGTTCTACTGCATGTTCTCCTAAGCACCTGAAAC...  \n",
       "1  CTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAG...  \n",
       "2  TTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAG...  \n",
       "3  TTTCCCCATCCTTGGGAAGATGACGATTCTTCCAAAACACTGAAAT...  \n",
       "4  GCAAATACAGTCATTTTAATACCCCAAGAAAAAGTCTTCCTACAAA...  \n",
       "5  CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAA...  \n",
       "6  CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAA...  "
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "ebadca50-a10f-4c82-b19a-e1c5fb0f0086",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Load the genome\n",
    "genome = SeqIO.to_dict(SeqIO.parse(\"/data/projects/Resources/Gencode_genome_annotation/GRCh38.primary_assembly.genome.fa\", \"fasta\"))\n",
    "\n",
    "\n",
    "def get_sequences(df):\n",
    "    data = []\n",
    "    for idx, row in df.iterrows():\n",
    "        chrom = row['chromosome-number']\n",
    "        ref_start = row['chromosome_start']\n",
    "        ref_end = row['chromosome_end']\n",
    "        variant_start = row['varinat_start']\n",
    "        variant_end = row['variant_end']\n",
    "        ref_nucleotide = row['ref_neucleotide']\n",
    "        alts = row['alternative_neucleotide'].rstrip(',').split(',')\n",
    "        \n",
    "        # Get reference sequence\n",
    "        ref_seq = str(genome[chrom].seq[ref_start:ref_end])\n",
    "\n",
    "        # Adjust for 0-based indexing in python\n",
    "        variant_pos_start = variant_start - ref_start\n",
    "        variant_pos_end = variant_end - ref_start\n",
    "\n",
    "        # Replace reference nucleotide with each alternative\n",
    "        for alt in alts:\n",
    "            if alt == \"\":  # Deletion\n",
    "                # Fetch additional bases from genome to maintain sequence length\n",
    "                extra_bases = str(genome[chrom].seq[ref_end:ref_end + len(ref_nucleotide)])\n",
    "                alt_seq = ref_seq[:variant_pos_start] + ref_seq[variant_pos_end:] + extra_bases\n",
    "            else:  # SNPs, Insertions, and partial Deletions\n",
    "                extra_bases = str(genome[chrom].seq[ref_end:ref_end + len(ref_nucleotide) - len(alt)])\n",
    "                alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:] + extra_bases\n",
    "                \n",
    "            # Append to the list as a dictionary\n",
    "            data.append({\n",
    "                'reference_id': chrom,\n",
    "                'varinat_start': variant_start,\n",
    "                'variant_end': variant_end,\n",
    "                'ref_neucleotide': ref_nucleotide,\n",
    "                'alternative_neucleotide': alt,\n",
    "                'reference_seq': ref_seq,\n",
    "                'alt_seq': alt_seq\n",
    "            })\n",
    "    \n",
    "    # Convert the list of dictionaries to a DataFrame\n",
    "    new_df = pd.DataFrame(data)\n",
    "    return new_df\n",
    "\n",
    "# Your dataframe (sample)\n",
    "df = pd.DataFrame({\n",
    "    'chromosome-number': ['chr1', 'chr1', 'chr1', 'chr1', 'chr1'],\n",
    "    'chromosome_start': [24401417, 24383964, 234429638, 67049811, 85183161],\n",
    "    'chromosome_end': [24401496, 24384043, 234429717, 67049890, 85183240],\n",
    "    'varinat_start': [24401424, 24383964, 234429682, 67049880, 85183235],\n",
    "    'variant_end': [24401425, 24383965, 234429684, 67049882, 85183239],\n",
    "    'ref_neucleotide': ['T', 'G', 'AT', 'AT', 'ATAT'],\n",
    "    'number_of_alts': [1, 2, 1,1,2],\n",
    "    'alternative_neucleotide': ['C,', 'C,T,', 'ATAT,', ',', 'AT,CGT,']\n",
    "})\n",
    "\n",
    "\n",
    "new_df = get_sequences(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "d520a267-3f3b-4adb-953a-900e9fcc988a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference_id</th>\n",
       "      <th>varinat_start</th>\n",
       "      <th>variant_end</th>\n",
       "      <th>ref_neucleotide</th>\n",
       "      <th>alternative_neucleotide</th>\n",
       "      <th>reference_seq</th>\n",
       "      <th>alt_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24401424</td>\n",
       "      <td>24401425</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>TGTTTGATGAAAAGTTCTACTGCATGTTCTCCTAAGCACCTGAAAC...</td>\n",
       "      <td>TGTTTGACGAAAAGTTCTACTGCATGTTCTCCTAAGCACCTGAAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24383964</td>\n",
       "      <td>24383965</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>GTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAG...</td>\n",
       "      <td>CTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24383964</td>\n",
       "      <td>24383965</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>GTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAG...</td>\n",
       "      <td>TTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>234429682</td>\n",
       "      <td>234429684</td>\n",
       "      <td>AT</td>\n",
       "      <td>ATAT</td>\n",
       "      <td>TTTCCCCATCCTTGGGAAGATGACGATTCTTCCAAAACACTGAAAT...</td>\n",
       "      <td>TTTCCCCATCCTTGGGAAGATGACGATTCTTCCAAAACACTGAAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67049880</td>\n",
       "      <td>67049882</td>\n",
       "      <td>AT</td>\n",
       "      <td></td>\n",
       "      <td>GCAAATACAGTCATTTTAATACCCCAAGAAAAAGTCTTCCTACAAA...</td>\n",
       "      <td>GCAAATACAGTCATTTTAATACCCCAAGAAAAAGTCTTCCTACAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>85183235</td>\n",
       "      <td>85183239</td>\n",
       "      <td>ATAT</td>\n",
       "      <td>AT</td>\n",
       "      <td>CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAA...</td>\n",
       "      <td>CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>85183235</td>\n",
       "      <td>85183239</td>\n",
       "      <td>ATAT</td>\n",
       "      <td>CGT</td>\n",
       "      <td>CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAA...</td>\n",
       "      <td>CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  reference_id  varinat_start  variant_end ref_neucleotide  \\\n",
       "0         chr1       24401424     24401425               T   \n",
       "1         chr1       24383964     24383965               G   \n",
       "2         chr1       24383964     24383965               G   \n",
       "3         chr1      234429682    234429684              AT   \n",
       "4         chr1       67049880     67049882              AT   \n",
       "5         chr1       85183235     85183239            ATAT   \n",
       "6         chr1       85183235     85183239            ATAT   \n",
       "\n",
       "  alternative_neucleotide                                      reference_seq  \\\n",
       "0                       C  TGTTTGATGAAAAGTTCTACTGCATGTTCTCCTAAGCACCTGAAAC...   \n",
       "1                       C  GTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAG...   \n",
       "2                       T  GTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAG...   \n",
       "3                    ATAT  TTTCCCCATCCTTGGGAAGATGACGATTCTTCCAAAACACTGAAAT...   \n",
       "4                          GCAAATACAGTCATTTTAATACCCCAAGAAAAAGTCTTCCTACAAA...   \n",
       "5                      AT  CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAA...   \n",
       "6                     CGT  CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAA...   \n",
       "\n",
       "                                             alt_seq  \n",
       "0  TGTTTGACGAAAAGTTCTACTGCATGTTCTCCTAAGCACCTGAAAC...  \n",
       "1  CTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAG...  \n",
       "2  TTGAATAACATTGTAGAACCCAGGTCCTGGGATATCATTCTGAAAG...  \n",
       "3  TTTCCCCATCCTTGGGAAGATGACGATTCTTCCAAAACACTGAAAT...  \n",
       "4  GCAAATACAGTCATTTTAATACCCCAAGAAAAAGTCTTCCTACAAA...  \n",
       "5  CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAA...  \n",
       "6  CCAATTCAGTACTATTTGTAGCAGCCAGGATCCCAGAACCTTAAAA...  "
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d7e03c2-7ab4-45a8-ba22-3bf8c03dc7f4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
