{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "48bb6eb5-6de0-4d27-848d-5c2e2eb29faf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, tarfile\n",
    "import re\n",
    "import multiprocessing\n",
    "import pandas as pd\n",
    "import pyBigWig\n",
    "import pybedtools\n",
    "import zipfile, gzip\n",
    "from gtfparse import read_gtf\n",
    "import pandas as pd\n",
    "#pd.set_option('display.max_columns', None)\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f4f24891-9cb6-4b3d-b75a-13c312c4bcf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_genome_path = \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/\"\n",
    "data_path= \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/gencode_annotation/\"\n",
    "donor_path = \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/Donor_splice_sites/\"\n",
    "non_donor_path = \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/Non_donnor_splice_sites/\"\n",
    "file_name=\"gencodeV38.bb\"\n",
    "gencode_path= \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/gencode_annotation/\"\n",
    "gtf_annotation_filename= \"gencode.v38.annotation.gtf\"\n",
    "gff_annotation_filename= \"gencode.v38.annotation.gff3\"\n",
    "file_path= data_path+file_name\n",
    "gencode_gtf_file_path= gencode_path + gtf_annotation_filename\n",
    "gencode_gff_file_path= gencode_path + gff_annotation_filename\n",
    "#reference_genome_file_path = reference_genome_path + reference_genome_name\n",
    "data_save_path= \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "71349309-5c66-42b7-94e1-7846185b70c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "bb= pyBigWig.open(file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8e286976-54a1-4993-99d4-30f7b4c4880e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'hgnc_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'ont', 'protein_id', 'ccdsid']\n"
     ]
    }
   ],
   "source": [
    "df = read_gtf(gencode_gtf_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bf9498db-5711-4b53-99c0-d8b8de788bcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand',\n",
      "       'frame', 'gene_id', 'gene_type', 'gene_name', 'level', 'hgnc_id',\n",
      "       'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name',\n",
      "       'transcript_support_level', 'tag', 'havana_transcript', 'exon_number',\n",
      "       'exon_id', 'ont', 'protein_id', 'ccdsid'],\n",
      "      dtype='object')\n",
      "(3150424, 25)\n"
     ]
    }
   ],
   "source": [
    "print(df.columns)\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f3c7bf4f-d8cc-475d-84fa-77ca8da25178",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  source feature  start    end  score strand frame  \\\n",
       "2    chr1  HAVANA    exon  11869  12227    NaN      +     0   \n",
       "3    chr1  HAVANA    exon  12613  12721    NaN      +     0   \n",
       "4    chr1  HAVANA    exon  13221  14409    NaN      +     0   \n",
       "\n",
       "             gene_id                           gene_type gene_name level  \\\n",
       "2  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "3  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "4  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "\n",
       "      hgnc_id           havana_gene      transcript_id       transcript_type  \\\n",
       "2  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2  processed_transcript   \n",
       "3  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2  processed_transcript   \n",
       "4  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2  processed_transcript   \n",
       "\n",
       "  transcript_name transcript_support_level    tag     havana_transcript  \\\n",
       "2     DDX11L1-202                        1  basic  OTTHUMT00000362751.1   \n",
       "3     DDX11L1-202                        1  basic  OTTHUMT00000362751.1   \n",
       "4     DDX11L1-202                        1  basic  OTTHUMT00000362751.1   \n",
       "\n",
       "  exon_number            exon_id ont protein_id ccdsid  \n",
       "2           1  ENSE00002234944.1                        \n",
       "3           2  ENSE00003582793.1                        \n",
       "4           3  ENSE00002312635.1                        "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[2:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d082499d-929b-4dfd-b132-2e8b2ddc5cf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "promoter_coordinates = [-249, 50]\n",
    "core_promoter_coordinates= [-34, 35]\n",
    "donor_coordinates= [-199, 200]\n",
    "acceptor_coordinates= [-199, 200]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "58399999-d0ae-4742-929d-9d8e6ca49c7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_data(elements, distance):\n",
    "    #elements = sorted(elements) # sorting the user input\n",
    "    new_elements = [elements[0]] # make a new list for output\n",
    "    for element in elements[1:]: # Iterate over the remaining elements...\n",
    "        if element - new_elements[-1] >= distance: \n",
    "            # this is the condition you described above\n",
    "            new_elements.append(element)\n",
    "    return new_elements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a3fbfc78-90c8-4931-8393-4a248c461354",
   "metadata": {},
   "outputs": [],
   "source": [
    "def string_subs(row, temp_seq):\n",
    "    start = row['coordinates'] -35\n",
    "    end = row['coordinates'] + 34\n",
    "    #print (start, end)\n",
    "    sub_string = temp_seq[start:end]\n",
    "    #print(sub_string)\n",
    "    temp_seq = temp_seq.replace(sub_string, 70*'$')\n",
    "    return temp_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1b6be116-480b-412b-af09-991983f72b4f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1cd5bea5-1a81-4f50-925e-48deae690c73",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr4\n",
      "(62342, 25)\n",
      "190214555\n",
      "length: 190214555\n",
      "(53730, 6)\n",
      "20500\n",
      "20505\n",
      "20510\n",
      "20515\n",
      "20520\n",
      "20525\n",
      "20530\n",
      "20535\n",
      "20540\n",
      "20545\n",
      "20550\n",
      "20555\n",
      "20560\n",
      "20565\n",
      "20570\n",
      "20575\n",
      "20580\n",
      "20585\n",
      "20590\n",
      "20595\n",
      "20600\n",
      "20605\n",
      "20610\n",
      "20615\n",
      "20620\n",
      "20625\n",
      "20630\n",
      "20635\n",
      "20640\n",
      "20645\n",
      "20650\n",
      "20655\n",
      "20660\n",
      "20665\n",
      "20670\n",
      "20675\n",
      "20680\n",
      "20685\n",
      "20690\n",
      "20695\n",
      "20700\n",
      "20705\n",
      "20710\n",
      "20715\n",
      "20720\n",
      "20725\n",
      "20730\n",
      "20735\n",
      "20740\n",
      "20745\n",
      "20750\n",
      "20755\n",
      "20760\n",
      "20765\n",
      "20770\n",
      "20775\n",
      "20780\n",
      "20785\n",
      "20790\n",
      "20795\n",
      "20800\n",
      "20805\n",
      "20810\n",
      "20815\n",
      "20820\n",
      "20825\n",
      "20830\n",
      "20835\n",
      "20840\n",
      "20845\n",
      "20850\n",
      "20855\n",
      "20860\n",
      "20865\n",
      "20870\n",
      "20875\n",
      "20880\n",
      "20885\n",
      "20890\n",
      "20895\n",
      "20900\n",
      "20905\n",
      "20910\n",
      "20915\n",
      "20920\n",
      "20925\n",
      "20930\n",
      "20935\n",
      "20940\n",
      "20945\n",
      "20950\n",
      "20955\n",
      "20960\n",
      "20965\n",
      "20970\n",
      "20975\n",
      "20980\n",
      "20985\n",
      "20990\n",
      "20995\n",
      "21000\n",
      "21005\n",
      "21010\n",
      "21015\n",
      "21020\n",
      "21025\n",
      "21030\n",
      "21035\n",
      "21040\n",
      "21045\n",
      "21050\n",
      "21055\n",
      "21060\n",
      "21065\n",
      "21070\n",
      "21075\n",
      "21080\n",
      "21085\n",
      "21090\n",
      "21095\n",
      "21100\n",
      "21105\n",
      "21110\n",
      "21115\n",
      "21120\n",
      "21125\n",
      "21130\n",
      "21135\n",
      "21140\n",
      "21145\n",
      "21150\n",
      "21155\n",
      "21160\n",
      "21165\n",
      "21170\n",
      "21175\n",
      "21180\n",
      "21185\n",
      "21190\n",
      "21195\n",
      "21200\n",
      "21205\n",
      "21210\n",
      "21215\n",
      "21220\n",
      "21225\n",
      "21230\n",
      "21235\n",
      "21240\n",
      "21245\n",
      "21250\n",
      "21255\n",
      "21260\n",
      "21265\n",
      "21270\n",
      "21275\n",
      "21280\n",
      "21285\n",
      "21290\n",
      "21295\n",
      "21300\n",
      "21305\n",
      "21310\n",
      "21315\n",
      "21320\n",
      "21325\n",
      "21330\n",
      "21335\n",
      "21340\n",
      "21345\n",
      "21350\n",
      "21355\n",
      "21360\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_3451437/1209320208.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     30\u001b[0m     \u001b[0msub_string\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtemp_seq\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mend\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     31\u001b[0m     \u001b[0;31m#print(sub_string)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 32\u001b[0;31m     \u001b[0mtemp_seq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtemp_seq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msub_string\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m70\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0;34m'$'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     33\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     34\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "chr_file= \"chr4.fa\"\n",
    "reference_genome_file_path = reference_genome_path + chr_file\n",
    "chr_name = chr_file.split('.')[0]\n",
    "print (chr_name)\n",
    "df_chr= df[df['seqname']==chr_name]\n",
    "df_new = df_chr[df_chr['feature']=='exon'].reset_index(drop=True)\n",
    "print (df_new.shape)\n",
    "#print (reference_genome_file_path)\n",
    "length= bb.chroms(chr_name)\n",
    "print (length)\n",
    "fasta = pybedtools.example_filename(reference_genome_file_path)\n",
    "temp_seq= pybedtools.BedTool.seq(chr_name+':1-'+str(length), fasta)\n",
    "print (\"length:\", len(temp_seq))\n",
    "\n",
    "pattern_donor= ['GT', 'gt']\n",
    "\n",
    "df_donor = pd.read_csv(donor_path+chr_name+\"_donor_splice_set.csv\")\n",
    "\n",
    "\n",
    "print(df_donor.shape)\n",
    "\n",
    "\n",
    "\n",
    "for index, row in df_donor.loc[20500:21500].iterrows():\n",
    "    if (index%5==0):\n",
    "        print (index)\n",
    "    start = row['coordinates'] -35\n",
    "    end = row['coordinates'] + 34\n",
    "    #print (start, end)\n",
    "    sub_string = temp_seq[start:end]\n",
    "    #print(sub_string)\n",
    "    temp_seq = temp_seq.replace(sub_string, 70*'$')\n",
    "\n",
    "\n",
    "Non_donor_sequence = re.sub('[$]', '', temp_seq)\n",
    "print (\"@@##$$\", len(Non_donor_sequence))\n",
    "non_donor_indices = []\n",
    "for j in pattern_acceptor:\n",
    "    non_donor_indices.append([match.start() for match in re.finditer(j, Non_donor_sequence)])\n",
    "\n",
    "print (len(non_donor_indices[0]))\n",
    "flat_list = [item for sublist in non_donor_indices for item in sublist]\n",
    "print (\"test\")\n",
    "flat_list.sort()\n",
    "print (\"test1\")\n",
    "print (\"@#@\",len(flat_list))\n",
    "removed_flat_list = remove_data(flat_list, 70)\n",
    "print (\"$$$\", len(removed_flat_list))\n",
    "print (removed_flat_list[0:15])\n",
    "\n",
    "list_non_donor = []\n",
    "for i in removed_flat_list[0:10]:\n",
    "    temps =  Non_donor_sequence[i-35:i+34]\n",
    "    list_non_donor.append(temps)\n",
    "pd_df = pd.DataFrame(list_non_donor)\n",
    "print (pd_df)\n",
    "#pd_df.to_csv(non_acceptor_path+\"/\"+chr_name+\"_non_acceptor.csv\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
