{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "48bb6eb5-6de0-4d27-848d-5c2e2eb29faf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, tarfile\n",
    "import pyBigWig\n",
    "import pybedtools\n",
    "import zipfile, gzip\n",
    "from gtfparse import read_gtf\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None)\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f4f24891-9cb6-4b3d-b75a-13c312c4bcf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_genome_path = \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/\"\n",
    "data_path= \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/gencode_annotation/\"\n",
    "file_name=\"gencodeV38.bb\"\n",
    "gencode_path= \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/gencode_annotation/\"\n",
    "gtf_annotation_filename= \"gencode.v38.annotation.gtf\"\n",
    "gff_annotation_filename= \"gencode.v38.annotation.gff3\"\n",
    "file_path= data_path+file_name\n",
    "gencode_gtf_file_path= gencode_path+ gtf_annotation_filename\n",
    "gencode_gff_file_path= gencode_path+ gff_annotation_filename\n",
    "#reference_genome_file_path = reference_genome_path + reference_genome_name\n",
    "data_save_path= \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "271adad9-08d8-4917-bcaf-e22363625388",
   "metadata": {},
   "outputs": [],
   "source": [
    "#reference_genome_file_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "89017f10-1f85-49c8-944d-2fabe5273a4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['chr21.fa.fai',\n",
       " 'chr3.fa.fai',\n",
       " 'chr5.fa.fai',\n",
       " 'chr6.fa.fai',\n",
       " 'chr8.fa.fai',\n",
       " 'chr9.fa.fai',\n",
       " 'chrY.fa.fai',\n",
       " '.DS_Store',\n",
       " 'chr1.fa',\n",
       " 'chr10.fa',\n",
       " 'chr11.fa',\n",
       " 'chr12.fa',\n",
       " 'chr13.fa',\n",
       " 'chr14.fa',\n",
       " 'chr15.fa',\n",
       " 'chr16.fa',\n",
       " 'chr17.fa',\n",
       " 'chr18.fa',\n",
       " 'chr19.fa',\n",
       " 'chr2.fa',\n",
       " 'chr20.fa',\n",
       " 'chr21.fa',\n",
       " 'chr22.fa',\n",
       " 'chr3.fa',\n",
       " 'chr4.fa',\n",
       " 'chr5.fa',\n",
       " 'chr6.fa',\n",
       " 'chr7.fa',\n",
       " 'chr8.fa',\n",
       " 'chr9.fa',\n",
       " 'chrX.fa',\n",
       " 'chrY.fa',\n",
       " 'keep',\n",
       " 'chr10.fa.fai',\n",
       " 'chr11.fa.fai',\n",
       " 'chr12.fa.fai',\n",
       " 'chr13.fa.fai',\n",
       " 'chr14.fa.fai',\n",
       " 'chr15.fa.fai',\n",
       " 'chr16.fa.fai',\n",
       " 'chr17.fa.fai',\n",
       " 'chr18.fa.fai',\n",
       " 'chr19.fa.fai',\n",
       " 'chr2.fa.fai',\n",
       " 'chr20.fa.fai',\n",
       " 'chr22.fa.fai',\n",
       " 'chr4.fa.fai',\n",
       " 'chr7.fa.fai',\n",
       " 'chrX.fa.fai',\n",
       " '._.DS_Store',\n",
       " 'chr1.fa.fai']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr_files= os.listdir(reference_genome_path )\n",
    "chr_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8e286976-54a1-4993-99d4-30f7b4c4880e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'hgnc_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'ont', 'protein_id', 'ccdsid']\n"
     ]
    }
   ],
   "source": [
    "df = read_gtf(gencode_gtf_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bf9498db-5711-4b53-99c0-d8b8de788bcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand',\n",
      "       'frame', 'gene_id', 'gene_type', 'gene_name', 'level', 'hgnc_id',\n",
      "       'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name',\n",
      "       'transcript_support_level', 'tag', 'havana_transcript', 'exon_number',\n",
      "       'exon_id', 'ont', 'protein_id', 'ccdsid'],\n",
      "      dtype='object')\n",
      "(3150424, 25)\n"
     ]
    }
   ],
   "source": [
    "print(df.columns)\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "30f808e1-6f52-447e-ba2b-525ce83222b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010</td>\n",
       "      <td>12057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12697</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001758273.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12975</td>\n",
       "      <td>13052</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00001799933.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>13374</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00001746346.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13453</td>\n",
       "      <td>13670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00001863096.1</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>29534</td>\n",
       "      <td>29570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001890219.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>24738</td>\n",
       "      <td>24891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003507205.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>18268</td>\n",
       "      <td>18366</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003477500.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>17915</td>\n",
       "      <td>18061</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00003565697.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>17606</td>\n",
       "      <td>17742</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00003475637.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>17233</td>\n",
       "      <td>17368</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00003502542.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqname  source     feature  start    end  score strand frame  \\\n",
       "0     chr1  HAVANA        gene  11869  14409    NaN      +     0   \n",
       "1     chr1  HAVANA  transcript  11869  14409    NaN      +     0   \n",
       "2     chr1  HAVANA        exon  11869  12227    NaN      +     0   \n",
       "3     chr1  HAVANA        exon  12613  12721    NaN      +     0   \n",
       "4     chr1  HAVANA        exon  13221  14409    NaN      +     0   \n",
       "5     chr1  HAVANA  transcript  12010  13670    NaN      +     0   \n",
       "6     chr1  HAVANA        exon  12010  12057    NaN      +     0   \n",
       "7     chr1  HAVANA        exon  12179  12227    NaN      +     0   \n",
       "8     chr1  HAVANA        exon  12613  12697    NaN      +     0   \n",
       "9     chr1  HAVANA        exon  12975  13052    NaN      +     0   \n",
       "10    chr1  HAVANA        exon  13221  13374    NaN      +     0   \n",
       "11    chr1  HAVANA        exon  13453  13670    NaN      +     0   \n",
       "12    chr1  HAVANA        gene  14404  29570    NaN      -     0   \n",
       "13    chr1  HAVANA  transcript  14404  29570    NaN      -     0   \n",
       "14    chr1  HAVANA        exon  29534  29570    NaN      -     0   \n",
       "15    chr1  HAVANA        exon  24738  24891    NaN      -     0   \n",
       "16    chr1  HAVANA        exon  18268  18366    NaN      -     0   \n",
       "17    chr1  HAVANA        exon  17915  18061    NaN      -     0   \n",
       "18    chr1  HAVANA        exon  17606  17742    NaN      -     0   \n",
       "19    chr1  HAVANA        exon  17233  17368    NaN      -     0   \n",
       "\n",
       "              gene_id                           gene_type gene_name level  \\\n",
       "0   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "1   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "2   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "3   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "4   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "5   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "6   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "7   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "8   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "9   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "10  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "11  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "12  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "13  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "14  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "15  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "16  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "17  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "18  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "19  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "\n",
       "       hgnc_id           havana_gene      transcript_id  \\\n",
       "0   HGNC:37102  OTTHUMG00000000961.2                      \n",
       "1   HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "2   HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "3   HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "4   HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "5   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "6   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "7   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "8   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "9   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "10  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "11  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "12  HGNC:38034  OTTHUMG00000000958.1                      \n",
       "13  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "14  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "15  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "16  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "17  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "18  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "19  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "\n",
       "                       transcript_type transcript_name  \\\n",
       "0                                                        \n",
       "1                 processed_transcript     DDX11L1-202   \n",
       "2                 processed_transcript     DDX11L1-202   \n",
       "3                 processed_transcript     DDX11L1-202   \n",
       "4                 processed_transcript     DDX11L1-202   \n",
       "5   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "6   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "7   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "8   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "9   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "10  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "11  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "12                                                       \n",
       "13              unprocessed_pseudogene      WASH7P-201   \n",
       "14              unprocessed_pseudogene      WASH7P-201   \n",
       "15              unprocessed_pseudogene      WASH7P-201   \n",
       "16              unprocessed_pseudogene      WASH7P-201   \n",
       "17              unprocessed_pseudogene      WASH7P-201   \n",
       "18              unprocessed_pseudogene      WASH7P-201   \n",
       "19              unprocessed_pseudogene      WASH7P-201   \n",
       "\n",
       "   transcript_support_level                      tag     havana_transcript  \\\n",
       "0                                                                            \n",
       "1                         1                    basic  OTTHUMT00000362751.1   \n",
       "2                         1                    basic  OTTHUMT00000362751.1   \n",
       "3                         1                    basic  OTTHUMT00000362751.1   \n",
       "4                         1                    basic  OTTHUMT00000362751.1   \n",
       "5                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "6                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "7                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "8                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "9                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "10                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "11                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "12                                                                           \n",
       "13                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "14                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "15                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "16                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "17                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "18                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "19                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "\n",
       "   exon_number            exon_id                      ont protein_id ccdsid  \n",
       "0                                                                             \n",
       "1                                                                             \n",
       "2            1  ENSE00002234944.1                                             \n",
       "3            2  ENSE00003582793.1                                             \n",
       "4            3  ENSE00002312635.1                                             \n",
       "5                                  PGO:0000005,PGO:0000019                    \n",
       "6            1  ENSE00001948541.1  PGO:0000005,PGO:0000019                    \n",
       "7            2  ENSE00001671638.2  PGO:0000005,PGO:0000019                    \n",
       "8            3  ENSE00001758273.2  PGO:0000005,PGO:0000019                    \n",
       "9            4  ENSE00001799933.2  PGO:0000005,PGO:0000019                    \n",
       "10           5  ENSE00001746346.2  PGO:0000005,PGO:0000019                    \n",
       "11           6  ENSE00001863096.1  PGO:0000005,PGO:0000019                    \n",
       "12                                                                            \n",
       "13                                             PGO:0000005                    \n",
       "14           1  ENSE00001890219.1              PGO:0000005                    \n",
       "15           2  ENSE00003507205.1              PGO:0000005                    \n",
       "16           3  ENSE00003477500.1              PGO:0000005                    \n",
       "17           4  ENSE00003565697.1              PGO:0000005                    \n",
       "18           5  ENSE00003475637.1              PGO:0000005                    \n",
       "19           6  ENSE00003502542.1              PGO:0000005                    "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "671220d3-1073-49cf-a7d4-7f117e0b5ce2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.head(25).to_csv(\"gencode_annotation.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5c1381e3-3325-4b15-b1b8-80b27b8ac306",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "feature\n",
       "CDS                825028\n",
       "Selenocysteine        119\n",
       "UTR                349402\n",
       "exon              1499012\n",
       "gene                60649\n",
       "start_codon         92997\n",
       "stop_codon          86205\n",
       "transcript         237012\n",
       "dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby(['feature']).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "06945eba-fcdb-4f16-a5ec-fa94e936c8ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(286816, 25)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_chr1 = df[df['seqname']=='chr1']\n",
    "df_chr1.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "4ca223a6-8264-451d-9e50-9252755d1200",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010</td>\n",
       "      <td>12057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12697</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001758273.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12975</td>\n",
       "      <td>13052</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00001799933.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>13374</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00001746346.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13453</td>\n",
       "      <td>13670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00001863096.1</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>29534</td>\n",
       "      <td>29570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001890219.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>24738</td>\n",
       "      <td>24891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003507205.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>18268</td>\n",
       "      <td>18366</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003477500.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>17915</td>\n",
       "      <td>18061</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00003565697.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>17606</td>\n",
       "      <td>17742</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00003475637.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>17233</td>\n",
       "      <td>17368</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00003502542.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>16858</td>\n",
       "      <td>17055</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>7</td>\n",
       "      <td>ENSE00003553898.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>16607</td>\n",
       "      <td>16765</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSE00003621279.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>15796</td>\n",
       "      <td>15947</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>9</td>\n",
       "      <td>ENSE00002030414.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>15005</td>\n",
       "      <td>15038</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>10</td>\n",
       "      <td>ENSE00001935574.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>14404</td>\n",
       "      <td>14501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>11</td>\n",
       "      <td>ENSE00001843071.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqname  source feature  start    end  score strand frame  \\\n",
       "0     chr1  HAVANA    exon  11869  12227    NaN      +     0   \n",
       "1     chr1  HAVANA    exon  12613  12721    NaN      +     0   \n",
       "2     chr1  HAVANA    exon  13221  14409    NaN      +     0   \n",
       "3     chr1  HAVANA    exon  12010  12057    NaN      +     0   \n",
       "4     chr1  HAVANA    exon  12179  12227    NaN      +     0   \n",
       "5     chr1  HAVANA    exon  12613  12697    NaN      +     0   \n",
       "6     chr1  HAVANA    exon  12975  13052    NaN      +     0   \n",
       "7     chr1  HAVANA    exon  13221  13374    NaN      +     0   \n",
       "8     chr1  HAVANA    exon  13453  13670    NaN      +     0   \n",
       "9     chr1  HAVANA    exon  29534  29570    NaN      -     0   \n",
       "10    chr1  HAVANA    exon  24738  24891    NaN      -     0   \n",
       "11    chr1  HAVANA    exon  18268  18366    NaN      -     0   \n",
       "12    chr1  HAVANA    exon  17915  18061    NaN      -     0   \n",
       "13    chr1  HAVANA    exon  17606  17742    NaN      -     0   \n",
       "14    chr1  HAVANA    exon  17233  17368    NaN      -     0   \n",
       "15    chr1  HAVANA    exon  16858  17055    NaN      -     0   \n",
       "16    chr1  HAVANA    exon  16607  16765    NaN      -     0   \n",
       "17    chr1  HAVANA    exon  15796  15947    NaN      -     0   \n",
       "18    chr1  HAVANA    exon  15005  15038    NaN      -     0   \n",
       "19    chr1  HAVANA    exon  14404  14501    NaN      -     0   \n",
       "\n",
       "              gene_id                           gene_type gene_name level  \\\n",
       "0   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "1   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "2   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "3   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "4   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "5   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "6   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "7   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "8   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "9   ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "10  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "11  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "12  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "13  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "14  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "15  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "16  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "17  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "18  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "19  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "\n",
       "       hgnc_id           havana_gene      transcript_id  \\\n",
       "0   HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "1   HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "2   HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "3   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "4   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "5   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "6   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "7   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "8   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "9   HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "10  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "11  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "12  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "13  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "14  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "15  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "16  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "17  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "18  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "19  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "\n",
       "                       transcript_type transcript_name  \\\n",
       "0                 processed_transcript     DDX11L1-202   \n",
       "1                 processed_transcript     DDX11L1-202   \n",
       "2                 processed_transcript     DDX11L1-202   \n",
       "3   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "4   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "5   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "6   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "7   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "8   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "9               unprocessed_pseudogene      WASH7P-201   \n",
       "10              unprocessed_pseudogene      WASH7P-201   \n",
       "11              unprocessed_pseudogene      WASH7P-201   \n",
       "12              unprocessed_pseudogene      WASH7P-201   \n",
       "13              unprocessed_pseudogene      WASH7P-201   \n",
       "14              unprocessed_pseudogene      WASH7P-201   \n",
       "15              unprocessed_pseudogene      WASH7P-201   \n",
       "16              unprocessed_pseudogene      WASH7P-201   \n",
       "17              unprocessed_pseudogene      WASH7P-201   \n",
       "18              unprocessed_pseudogene      WASH7P-201   \n",
       "19              unprocessed_pseudogene      WASH7P-201   \n",
       "\n",
       "   transcript_support_level                      tag     havana_transcript  \\\n",
       "0                         1                    basic  OTTHUMT00000362751.1   \n",
       "1                         1                    basic  OTTHUMT00000362751.1   \n",
       "2                         1                    basic  OTTHUMT00000362751.1   \n",
       "3                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "4                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "5                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "6                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "7                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "8                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "9                        NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "10                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "11                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "12                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "13                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "14                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "15                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "16                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "17                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "18                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "19                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "\n",
       "   exon_number            exon_id                      ont protein_id ccdsid  \n",
       "0            1  ENSE00002234944.1                                             \n",
       "1            2  ENSE00003582793.1                                             \n",
       "2            3  ENSE00002312635.1                                             \n",
       "3            1  ENSE00001948541.1  PGO:0000005,PGO:0000019                    \n",
       "4            2  ENSE00001671638.2  PGO:0000005,PGO:0000019                    \n",
       "5            3  ENSE00001758273.2  PGO:0000005,PGO:0000019                    \n",
       "6            4  ENSE00001799933.2  PGO:0000005,PGO:0000019                    \n",
       "7            5  ENSE00001746346.2  PGO:0000005,PGO:0000019                    \n",
       "8            6  ENSE00001863096.1  PGO:0000005,PGO:0000019                    \n",
       "9            1  ENSE00001890219.1              PGO:0000005                    \n",
       "10           2  ENSE00003507205.1              PGO:0000005                    \n",
       "11           3  ENSE00003477500.1              PGO:0000005                    \n",
       "12           4  ENSE00003565697.1              PGO:0000005                    \n",
       "13           5  ENSE00003475637.1              PGO:0000005                    \n",
       "14           6  ENSE00003502542.1              PGO:0000005                    \n",
       "15           7  ENSE00003553898.1              PGO:0000005                    \n",
       "16           8  ENSE00003621279.1              PGO:0000005                    \n",
       "17           9  ENSE00002030414.1              PGO:0000005                    \n",
       "18          10  ENSE00001935574.1              PGO:0000005                    \n",
       "19          11  ENSE00001843071.1              PGO:0000005                    "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_chr1_transcript= df_chr1[df_chr1['feature']=='exon'].reset_index(drop=True)\n",
    "df_chr1_transcript.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d082499d-929b-4dfd-b132-2e8b2ddc5cf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "promoter_coordinates = [-249, 50]\n",
    "core_promoter_coordinates= [-34, 35]\n",
    "donor_coordinates= [-199, 200]\n",
    "acceptor_coordinates= [-199, 200]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0658de47-c801-426e-9644-3806712abade",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'reference_genome_file_path' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_265975/1130735144.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mfasta\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpybedtools\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexample_filename\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreference_genome_file_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'reference_genome_file_path' is not defined"
     ]
    }
   ],
   "source": [
    "fasta = pybedtools.example_filename(reference_genome_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1e0369d0-4941-4247-92e3-7553cac04e5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def acceptor_sequence(entry, df_acceptor_splice, acceptor_index):\n",
    "    print (\"@@@@\", acceptor_index)\n",
    "    df_acceptor_splice.loc[acceptor_index, ['transcript_id']]= entry['transcript_id']\n",
    "    df_acceptor_splice.loc[acceptor_index, ['exon_id']] = entry['exon_id']\n",
    "    df_acceptor_splice.loc[acceptor_index, ['transcript_type']] = entry['transcript_type']\n",
    "    df_acceptor_splice.loc[acceptor_index, ['strand']]= entry['strand']\n",
    "    if (entry['strand']=='+'):\n",
    "        df_acceptor_splice.loc[acceptor_index, ['coordinates']]= int(entry['start'])-1\n",
    "        acceptor_query= entry['seqname']+':'+str(int(entry['start'])+acceptor_coordinates[0])+'-'+str(int(entry['start'])+acceptor_coordinates[1])\n",
    "        print (\"##$$$\", acceptor_query)\n",
    "        df_acceptor_splice.loc[acceptor_index, ['sequence']] = pybedtools.BedTool.seq(acceptor_query, fasta)\n",
    "        print (\"###\", acceptor_index)\n",
    "    elif (entry['strand']=='-'):\n",
    "        df_acceptor_splice.loc[acceptor_index, ['coordinates']]= int(entry['end'])+1\n",
    "        acceptor_query= entry['seqname']+':'+str(int(entry['end'])+(acceptor_coordinates[1]*(-1)))+'-'+str(int(entry['end'])+(acceptor_coordinates[0]*(-1)))\n",
    "        temp_seq = pybedtools.BedTool.seq(acceptor_query, fasta)\n",
    "        rev_seq = str(Seq(temp_seq).reverse_complement())\n",
    "        df_acceptor_splice.loc[acceptor_index, ['sequence']] = rev_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ccc5d858-46ea-421c-b7c0-f856cedb7869",
   "metadata": {},
   "outputs": [],
   "source": [
    "def donor_sequence(entry, df_donor_splice, donor_index):\n",
    "    df_donor_splice.loc[donor_index, ['transcript_id']]= entry['transcript_id']\n",
    "    df_donor_splice.loc[donor_index, ['exon_id']] = entry['exon_id']\n",
    "    df_donor_splice.loc[donor_index, ['transcript_type']] = entry['transcript_type']\n",
    "    df_donor_splice.loc[donor_index, ['strand']]= entry['strand']\n",
    "    if (entry['strand']=='+'):\n",
    "        df_donor_splice.loc[donor_index, ['coordinates']] = int(entry['end'])+1\n",
    "        donor_query= entry['seqname']+':'+str(int(entry['end'])+donor_coordinates[0])+'-'+str(int(entry['end'])+donor_coordinates[1])\n",
    "        print (\"##$$$\", donor_query, donor_index)\n",
    "        df_donor_splice.loc[donor_index, ['sequence']] = pybedtools.BedTool.seq(donor_query, fasta)\n",
    "    elif (entry['strand']=='-'):\n",
    "        df_donor_splice.loc[donor_index, ['coordinates']] = int(entry['start'])-1\n",
    "        donor_query= entry['seqname']+':'+str(int(entry['start'])+(donor_coordinates[1]*(-1)))+'-'+str(int(entry['start'])+(donor_coordinates[0]*(-1)))\n",
    "        print (\"##$$$\", donor_query, donor_index)\n",
    "        temp_seq = pybedtools.BedTool.seq(donor_query, fasta)\n",
    "        rev_seq = str(Seq(temp_seq).reverse_complement())\n",
    "        df_donor_splice.loc[donor_index, ['sequence']] = rev_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2cbd9c22-f5b8-465f-932f-9aeb64f57a97",
   "metadata": {},
   "outputs": [],
   "source": [
    "def promoter_sequence(entry, df_promoter, promoter_index):\n",
    "    df_promoter.loc[promoter_index, ['gene_id']] = entry['gene_id']\n",
    "    df_promoter.loc[promoter_index, ['transcript_id']]= entry['transcript_id']\n",
    "    df_promoter.loc[promoter_index, ['transcript_type']]= entry['transcript_type']\n",
    "    df_promoter.loc[promoter_index, ['strand']] = entry['strand']\n",
    "    df_promoter.loc[promoter_index, ['exon_id']]=  entry['exon_id']\n",
    "    if (entry['strand']=='+'):\n",
    "        df_promoter.loc[promoter_index,['TSS']]= entry['start']\n",
    "        promoter_query= entry['seqname']+':'+str(int(entry['start'])+promoter_coordinates[0])+'-'+str(int(entry['start'])+promoter_coordinates[1])\n",
    "        df_promoter.loc[promoter_index, ['sequence']] = pybedtools.BedTool.seq(promoter_query, fasta)\n",
    "    elif(entry['strand']=='-'):\n",
    "        df_promoter.loc[promoter_index,['TSS']]= entry['end']\n",
    "        promoter_query= entry['seqname']+':'+str(int(entry['end'])+(promoter_coordinates[1]*(-1)))+'-'+str(int(entry['end'])+(promoter_coordinates[0]*(-1)))\n",
    "        temp_seq = pybedtools.BedTool.seq(promoter_query, fasta)\n",
    "        #print (\"Original\", temp_seq)\n",
    "        rev_seq = str(Seq(temp_seq).reverse_complement())\n",
    "        #print (rev_seq)\n",
    "        df_promoter.loc[promoter_index, ['sequence']] = rev_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "60c21aa4-2a16-4211-b305-bbfea865ebd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def core_promoter_sequence(entry, df_core_promoter, core_promoter_index):\n",
    "    df_core_promoter.loc[core_promoter_index, ['gene_id']] = entry['gene_id']\n",
    "    df_core_promoter.loc[core_promoter_index, ['transcript_id']] = entry['transcript_id']\n",
    "    df_core_promoter.loc[core_promoter_index, ['transcript_type']] = entry['transcript_type']\n",
    "    df_core_promoter.loc[core_promoter_index, ['strand']] = entry['strand']\n",
    "    df_core_promoter.loc[core_promoter_index, ['exon_id']]  =  entry['exon_id']\n",
    "    if (entry['strand']=='+'):\n",
    "        df_core_promoter.loc[core_promoter_index,['TSS']]= entry['start']\n",
    "        core_promoter_query= entry['seqname']+':'+str(int(entry['start'])+core_promoter_coordinates[0])+'-'+str(int(entry['start'])+core_promoter_coordinates[1])\n",
    "        df_core_promoter.loc[core_promoter_index, ['sequence']] = pybedtools.BedTool.seq(core_promoter_query, fasta)\n",
    "    elif(entry['strand']=='-'):\n",
    "        df_core_promoter.loc[core_promoter_index,['TSS']]= entry['end']\n",
    "        core_promoter_query= entry['seqname']+':'+str(int(entry['end'])+(core_promoter_coordinates[1]*(-1)))+'-'+str(int(entry['end'])+(core_promoter_coordinates[0]*(-1)))\n",
    "        temp_seq = pybedtools.BedTool.seq(core_promoter_query, fasta)\n",
    "        rev_seq = str(Seq(temp_seq).reverse_complement())\n",
    "        df_core_promoter.loc[core_promoter_index, ['sequence']] = rev_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "80df8252-7040-441d-93b5-2b03fd7ed87e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n",
      "(136680, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr1\n",
      "\n",
      " ENST00000450305.2 6\n",
      "+\n",
      "TSS 12010\n",
      "##$$$ chr1:11858-12257 0\n",
      "first donor_splice_site 12057\n",
      "Acceptor splice site 12179\n",
      "@@@@ 0\n",
      "##$$$ chr1:11980-12379\n",
      "### 0\n",
      "Donor splice site 12227\n",
      "##$$$ chr1:12028-12427 1\n",
      "Acceptor splice site 12613\n",
      "@@@@ 1\n",
      "##$$$ chr1:12414-12813\n",
      "### 1\n",
      "Donor splice site 12697\n",
      "##$$$ chr1:12498-12897 2\n",
      "Acceptor splice site 12975\n",
      "@@@@ 2\n",
      "##$$$ chr1:12776-13175\n",
      "### 2\n",
      "Donor splice site 13052\n",
      "##$$$ chr1:12853-13252 3\n",
      "Acceptor splice site 13221\n",
      "@@@@ 3\n",
      "##$$$ chr1:13022-13421\n",
      "### 3\n",
      "Donor splice site 13374\n",
      "##$$$ chr1:13175-13574 4\n",
      "Last acceptor_splice_site 13453\n",
      "@@@@ 4\n",
      "##$$$ chr1:13254-13653\n",
      "### 4\n",
      "\n",
      " ENST00000456328.2 3\n",
      "+\n",
      "TSS 11869\n",
      "##$$$ chr1:12028-12427 5\n",
      "first donor_splice_site 12227\n",
      "Acceptor splice site 12613\n",
      "@@@@ 5\n",
      "##$$$ chr1:12414-12813\n",
      "### 5\n",
      "Donor splice site 12721\n",
      "##$$$ chr1:12522-12921 6\n",
      "Last acceptor_splice_site 13221\n",
      "@@@@ 6\n",
      "##$$$ chr1:13022-13421\n",
      "### 6\n",
      "\n",
      " ENST00000488147.1 11\n",
      "-\n",
      "TSS 29570\n",
      "first donor_splice_site 29534\n",
      "##$$$ chr1:29334-29733 7\n",
      "Acceptor splice site 24891\n",
      "@@@@ 7\n",
      "Donor splice site 24738\n",
      "##$$$ chr1:24538-24937 8\n",
      "Acceptor splice site 18366\n",
      "@@@@ 8\n",
      "Donor splice site 18268\n",
      "##$$$ chr1:18068-18467 9\n",
      "Acceptor splice site 18061\n",
      "@@@@ 9\n",
      "Donor splice site 17915\n",
      "##$$$ chr1:17715-18114 10\n",
      "Acceptor splice site 17742\n",
      "@@@@ 10\n",
      "Donor splice site 17606\n",
      "##$$$ chr1:17406-17805 11\n",
      "Acceptor splice site 17368\n",
      "@@@@ 11\n",
      "Donor splice site 17233\n",
      "##$$$ chr1:17033-17432 12\n",
      "Acceptor splice site 17055\n",
      "@@@@ 12\n",
      "Donor splice site 16858\n",
      "##$$$ chr1:16658-17057 13\n",
      "Acceptor splice site 16765\n",
      "@@@@ 13\n",
      "Donor splice site 16607\n",
      "##$$$ chr1:16407-16806 14\n",
      "Acceptor splice site 15947\n",
      "@@@@ 14\n",
      "Donor splice site 15796\n",
      "##$$$ chr1:15596-15995 15\n",
      "Acceptor splice site 15038\n",
      "@@@@ 15\n",
      "Donor splice site 15005\n",
      "##$$$ chr1:14805-15204 16\n",
      "Last acceptor_splice_site 14501\n",
      "@@@@ 16\n"
     ]
    }
   ],
   "source": [
    "global df_promoter, df_core_promoter, df_donor_splice, df_acceptor_splice\n",
    "\n",
    "\n",
    "for file in chr_files:\n",
    "    if file.endswith('.fa'):\n",
    "        reference_genome_file_path = reference_genome_path + file\n",
    "        chr_name = file.split('.')[0]\n",
    "        print (chr_name)\n",
    "        df_chr= df[df['seqname']==chr_name]\n",
    "        df_new = df_chr[df_chr['feature']=='exon'].reset_index(drop=True)\n",
    "        print (df_new.shape)\n",
    "        #print (reference_genome_file_path)\n",
    "        fasta = pybedtools.example_filename(reference_genome_file_path)\n",
    "        \n",
    "        \n",
    "        folder_name = data_save_path + chr_name   \n",
    "        print (folder_name)\n",
    "        if not os.path.exists(folder_name):\n",
    "            os.makedirs(folder_name)\n",
    "        \n",
    "        \n",
    "        df_promoter = pd.DataFrame(columns= ['gene_id','transcript_id','strand','exon_id','transcript_type','TSS', 'sequence'])\n",
    "        df_core_promoter = pd.DataFrame(columns= ['gene_id','transcript_id','strand','exon_id','transcript_type','TSS','sequence'])\n",
    "        df_donor_splice = pd.DataFrame(columns= ['transcript_id','exon_id','transcript_type','strand','coordinates','sequence'])\n",
    "        df_acceptor_splice= pd.DataFrame(columns= ['transcript_id','exon_id', 'transcript_type','strand','coordinates', 'sequence'])\n",
    "        promoter_index = core_promoter_index = donor_index = acceptor_index = 0\n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "        df_by_transcript = df_new.head(20).groupby(\"transcript_id\")\n",
    "        for transcript, entries in df_by_transcript:\n",
    "            print (\"\\n\", transcript,len( entries) )\n",
    "            print (entries.iloc[0]['strand'])\n",
    "            if (entries.iloc[0]['strand']=='+'):\n",
    "                for index, entry in entries.iterrows():\n",
    "                    if (int(entry['exon_number'])==1):\n",
    "                        print(\"TSS\", entry['start'])\n",
    "                        promoter_sequence(entry, df_promoter, promoter_index)\n",
    "                        promoter_index= promoter_index+1\n",
    "                        core_promoter_sequence(entry, df_core_promoter, core_promoter_index)\n",
    "                        core_promoter_index = core_promoter_index +1\n",
    "                        donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                        donor_index =donor_index+ 1\n",
    "                        print(\"first donor_splice_site\", entry['end'])\n",
    "                    elif (int(entry['exon_number'])==len(entries)):\n",
    "                        print (\"Last acceptor_splice_site\",  entry['start'])\n",
    "                        acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                        acceptor_index = acceptor_index +1\n",
    "                    else:\n",
    "                        print (\"Acceptor splice site\", entry['start'])\n",
    "                        acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                        acceptor_index = acceptor_index +1\n",
    "                        print (\"Donor splice site\", entry['end'])\n",
    "                        donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                        donor_index =donor_index+ 1\n",
    "            elif (entries.iloc[0]['strand']=='-'):\n",
    "                for index, entry in entries.iterrows():\n",
    "                    if (int(entry['exon_number'])==1):\n",
    "                        print(\"TSS\", entry['end'])\n",
    "                        print(\"first donor_splice_site\", entry['start'])\n",
    "                        promoter_sequence(entry, df_promoter, promoter_index)\n",
    "                        promoter_index= promoter_index+1\n",
    "                        core_promoter_sequence(entry, df_core_promoter, core_promoter_index)\n",
    "                        core_promoter_index = core_promoter_index +1\n",
    "                        donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                        donor_index =donor_index+ 1\n",
    "                    elif (int(entry['exon_number'])==len(entries)):\n",
    "                        print (\"Last acceptor_splice_site\",  entry['end'])\n",
    "                        acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                        acceptor_index = acceptor_index +1\n",
    "                    else:\n",
    "                        print (\"Acceptor splice site\", entry['end'])\n",
    "                        acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                        acceptor_index = acceptor_index +1\n",
    "                        print (\"Donor splice site\", entry['start'])\n",
    "                        donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                        donor_index =donor_index+ 1\n",
    "        # df_promoter.to_csv(folder_name +\"/\"+ entry['seqname']+\"_promoter.csv\", index=False)\n",
    "        # df_core_promoter.to_csv(folder_name +\"/\"+ entry['seqname']+\"_core_promoter_sequence.csv\", index=False)\n",
    "        # df_acceptor_splice.to_csv(folder_name +\"/\"+ entry['seqname']+\"_acceptor_splice_set_sequence.csv\", index=False)\n",
    "        # df_donor_splice.to_csv(folder_name +\"/\"+ entry['seqname']+\"_donor_splice_set.csv\", index=False)\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cd5bea5-1a81-4f50-925e-48deae690c73",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bd36b20-1acb-41cd-bbff-67f9dd997003",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "6045744b-f030-4aa6-aa35-98910e8fdb12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>TSS</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>12010</td>\n",
       "      <td>gccggtgttgagaatgactgCGCAAATTTGCCGGATTTCCTTTGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>atgcctgtttctccacaaagtgtttacttttggatttttgccagtc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001890219.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>29570</td>\n",
       "      <td>GCTGAGAGACGCAGGAAAGCGGGTCAAggcgtagggctggagggca...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_id      transcript_id strand            exon_id  \\\n",
       "0  ENSG00000223972.5  ENST00000450305.2      +  ENSE00001948541.1   \n",
       "1  ENSG00000223972.5  ENST00000456328.2      +  ENSE00002234944.1   \n",
       "2  ENSG00000227232.5  ENST00000488147.1      -  ENSE00001890219.1   \n",
       "\n",
       "                      transcript_type    TSS  \\\n",
       "0  transcribed_unprocessed_pseudogene  12010   \n",
       "1                processed_transcript  11869   \n",
       "2              unprocessed_pseudogene  29570   \n",
       "\n",
       "                                            sequence  \n",
       "0  gccggtgttgagaatgactgCGCAAATTTGCCGGATTTCCTTTGCT...  \n",
       "1  atgcctgtttctccacaaagtgtttacttttggatttttgccagtc...  \n",
       "2  GCTGAGAGACGCAGGAAAGCGGGTCAAggcgtagggctggagggca...  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "99de0a0b-20a5-4fd1-919f-89703dc8c677",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>TSS</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>ENST00000417324.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001656588.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>36081</td>\n",
       "      <td>CAAGGATTAGTCCTGCTGCCCTGTGCCCAGCTCCCACACAACGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>12010</td>\n",
       "      <td>CACAGGGTCTTGATGCTGTGGTCTTCATCTGCAGGTGTCTGACTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>CCAGCACCGGGTATCATTCACCATTTTTCTTTTCGTTAACTTGCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>ENST00000461467.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001618781.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>36073</td>\n",
       "      <td>AGTCCTGCTGCCCTGTGCCCAGCTCCCACACAACGGGGTTTCGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001841699.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>30267</td>\n",
       "      <td>GCTTTTTTAAAAAGTTAAATATATAAAAAGAGCATCATCAGTCCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001947070.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>29554</td>\n",
       "      <td>GGCACGGAACTCACGCTCACTCCGAGCTCCCGACGTGCACACGGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001890219.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>29570</td>\n",
       "      <td>GGGTAGGGGCGGCCTGACGCTCGGAAGACAACGCATGGGAGCCGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000240361.2</td>\n",
       "      <td>ENST00000492842.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001830178.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>62949</td>\n",
       "      <td>GGTTGCTTCAGAGTCTTCCCTTCTATCTGATTCAGTGGACCAAGTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000268020.3</td>\n",
       "      <td>ENST00000606857.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003698237.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>52473</td>\n",
       "      <td>GAGTCAATGGCCAATGGATAGAGTAAATAATTCTGCGGTATCTAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>ENST00000607096.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003695741.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>30366</td>\n",
       "      <td>CCCTTGCCTTTtagggttgccagattaaaagacaggatgcccagct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00003746039.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>17436</td>\n",
       "      <td>TGGGACCCCACAGTGCACGCTGTGCCCCTGATGATGTGGGAGAGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003812156.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>65419</td>\n",
       "      <td>TTAGTCTCAGTACACAAAGCTCAGACTGGCTATTCCCAGATCTCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000240361.2</td>\n",
       "      <td>ENST00000642116.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003812686.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>57598</td>\n",
       "      <td>CTTTGTCTTTGGGGTATTACATAAATGAAGATTAAGCTATCTGAAT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id      transcript_id strand            exon_id  \\\n",
       "0   ENSG00000237613.2  ENST00000417324.1      -  ENSE00001656588.1   \n",
       "1   ENSG00000223972.5  ENST00000450305.2      +  ENSE00001948541.1   \n",
       "2   ENSG00000223972.5  ENST00000456328.2      +  ENSE00002234944.1   \n",
       "3   ENSG00000237613.2  ENST00000461467.1      -  ENSE00001618781.2   \n",
       "4   ENSG00000243485.5  ENST00000469289.1      +  ENSE00001841699.1   \n",
       "5   ENSG00000243485.5  ENST00000473358.1      +  ENSE00001947070.1   \n",
       "6   ENSG00000227232.5  ENST00000488147.1      -  ENSE00001890219.1   \n",
       "7   ENSG00000240361.2  ENST00000492842.2      +  ENSE00001830178.2   \n",
       "8   ENSG00000268020.3  ENST00000606857.1      +  ENSE00003698237.1   \n",
       "9   ENSG00000284332.1  ENST00000607096.1      +  ENSE00003695741.1   \n",
       "10  ENSG00000278267.1  ENST00000619216.1      -  ENSE00003746039.1   \n",
       "11  ENSG00000186092.7  ENST00000641515.2      +  ENSE00003812156.1   \n",
       "12  ENSG00000240361.2  ENST00000642116.1      +  ENSE00003812686.1   \n",
       "\n",
       "                       transcript_type    TSS  \\\n",
       "0                               lncRNA  36081   \n",
       "1   transcribed_unprocessed_pseudogene  12010   \n",
       "2                 processed_transcript  11869   \n",
       "3                               lncRNA  36073   \n",
       "4                               lncRNA  30267   \n",
       "5                               lncRNA  29554   \n",
       "6               unprocessed_pseudogene  29570   \n",
       "7   transcribed_unprocessed_pseudogene  62949   \n",
       "8               unprocessed_pseudogene  52473   \n",
       "9                                miRNA  30366   \n",
       "10                               miRNA  17436   \n",
       "11                      protein_coding  65419   \n",
       "12                processed_transcript  57598   \n",
       "\n",
       "                                             sequence  \n",
       "0   CAAGGATTAGTCCTGCTGCCCTGTGCCCAGCTCCCACACAACGGGG...  \n",
       "1   CACAGGGTCTTGATGCTGTGGTCTTCATCTGCAGGTGTCTGACTTC...  \n",
       "2   CCAGCACCGGGTATCATTCACCATTTTTCTTTTCGTTAACTTGCCG...  \n",
       "3   AGTCCTGCTGCCCTGTGCCCAGCTCCCACACAACGGGGTTTCGGGG...  \n",
       "4   GCTTTTTTAAAAAGTTAAATATATAAAAAGAGCATCATCAGTCCAA...  \n",
       "5   GGCACGGAACTCACGCTCACTCCGAGCTCCCGACGTGCACACGGCT...  \n",
       "6   GGGTAGGGGCGGCCTGACGCTCGGAAGACAACGCATGGGAGCCGTG...  \n",
       "7   GGTTGCTTCAGAGTCTTCCCTTCTATCTGATTCAGTGGACCAAGTA...  \n",
       "8   GAGTCAATGGCCAATGGATAGAGTAAATAATTCTGCGGTATCTAAA...  \n",
       "9   CCCTTGCCTTTtagggttgccagattaaaagacaggatgcccagct...  \n",
       "10  TGGGACCCCACAGTGCACGCTGTGCCCCTGATGATGTGGGAGAGGA...  \n",
       "11  TTAGTCTCAGTACACAAAGCTCAGACTGGCTATTCCCAGATCTCTT...  \n",
       "12  CTTTGTCTTTGGGGTATTACATAAATGAAGATTAAGCTATCTGAAT...  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_core_promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c32274e1-5f2e-4a43-9db4-cf2a78b3c044",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>strand</th>\n",
       "      <th>coordinates</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000417324.1</td>\n",
       "      <td>ENSE00001656588.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>35720</td>\n",
       "      <td>GGGGGCAGGAATTGGGGGTGACAGGAGAGCTGTAAGGTCTCCAGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000417324.1</td>\n",
       "      <td>ENSE00001669267.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>35276</td>\n",
       "      <td>aagtctcactatgttgctcaggctggtcttgaactcctggcctcaa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>12058</td>\n",
       "      <td>TTTTTCTTTTCGTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>12228</td>\n",
       "      <td>CTGCTGGCCTGTGCCAGGGTGCAAGCTGAGCACTGGAGTGGAGTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001758273.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>12698</td>\n",
       "      <td>TGCAGAGACGGGAGGGGCAGAGCCGCAGGCACAGCCAAGAGGGCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001799933.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>13053</td>\n",
       "      <td>AGCCCTGCTGCTTGCGGTGGCCTATAAAGCCTCCTAGTCTGGCTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001746346.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>13375</td>\n",
       "      <td>GGGGAGGCAGCTGTAACTCAAAGCCTTAGCCTCTGTTCCCACGAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>12228</td>\n",
       "      <td>CTGCTGGCCTGTGCCAGGGTGCAAGCTGAGCACTGGAGTGGAGTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>12722</td>\n",
       "      <td>GCAGGCACAGCCAAGAGGGCTGAAGAAATGGTAGAACGGAGCAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000461467.1</td>\n",
       "      <td>ENSE00001618781.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>35720</td>\n",
       "      <td>GGGGGCAGGAATTGGGGGTGACAGGAGAGCTGTAAGGTCTCCAGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>ENSE00001841699.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>30668</td>\n",
       "      <td>tggttgtttatctgagattcagaattaagcattttatattttattt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSE00001947070.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>30040</td>\n",
       "      <td>CTCTGTGGGCAGCTCCCTTGTGATCTGCTTAGTTCCCACCCCCCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSE00001922571.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>30668</td>\n",
       "      <td>tggttgtttatctgagattcagaattaagcattttatattttattt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00001890219.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>29533</td>\n",
       "      <td>GCGCACCCCGGGCGAGGGCTGCCCGGAGGAGCCGAGGTTGGCGGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003507205.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>24737</td>\n",
       "      <td>TGCCGGTGGTCGTGGGTGGCTCTGACGCTCCGGCTCCGACCCACAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003477500.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>18267</td>\n",
       "      <td>GGGCCTGGCCTGGCAGCTCTGCTGTTAGAAGCAGGAGGTGTGCAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003565697.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>17914</td>\n",
       "      <td>CCTAGCCCTCTGAGCCTCTTGGGGCCCCCTCAGACTTGTCCCTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003475637.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>17605</td>\n",
       "      <td>ATGGCAGCTGATTAGGCCCAGCACCAAATATTCACCATCCCTTGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003502542.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>17232</td>\n",
       "      <td>GGAGAGGAACATGGGCTCAGGACAGCGGGTGTCAGCTTGCCTGACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003553898.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>16857</td>\n",
       "      <td>AGGTCCCAGAGAACTACTTCTATGTGCCAGACCTGGGCCAGGTGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003621279.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>16606</td>\n",
       "      <td>ACACACCTTCTTCCCTAGACACCCCACACTTTGTGTTTCAGACCTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00002030414.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>15795</td>\n",
       "      <td>CCTGAGCAGCCTTTCCTCCTTGCCCTGTTCCCCACCTCCCGGCTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00001935574.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>15004</td>\n",
       "      <td>GTCTCATCAGTGCCCGGTGCTGGGTCAGGGATCGACTGAGGCTCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENST00000492842.2</td>\n",
       "      <td>ENSE00001830178.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>63888</td>\n",
       "      <td>TGGTTTTGTTTTTTGCTCCATGCATGTTTCTCTACGTGTGGCCTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENST00000606857.1</td>\n",
       "      <td>ENSE00003698237.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>53313</td>\n",
       "      <td>AGAATCCTTCTTCAGGAGACTTATCTAAAGCATTTGTGTCATGTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENST00000607096.1</td>\n",
       "      <td>ENSE00003695741.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>30504</td>\n",
       "      <td>TCCGTTGGCTTGCCTCCGGCATTTTTGGCCCTTGCCTTTtagggtt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>ENSE00003746039.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>17368</td>\n",
       "      <td>AGGCCCGGCCTGGCTTAGTGGGGACCCAGGGCATCAGACACAGGTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSE00003812156.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "      <td>65434</td>\n",
       "      <td>TTTGCTGTGATTCTTTAAAAAGCACCTTTAGACTCAGTGAGATAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSE00003813641.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "      <td>65574</td>\n",
       "      <td>GACTATAAAAGTTAGTCTCAGTACACAAAGCTCAGACTGGCTATTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ENST00000642116.1</td>\n",
       "      <td>ENSE00003812686.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>57654</td>\n",
       "      <td>CAAGAGAAACTAATAATGGGCTGGGTATTGTCTGTCTCAGCAGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ENST00000642116.1</td>\n",
       "      <td>ENSE00003812505.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>58857</td>\n",
       "      <td>GTCCCAGACCCTAAATGACTGACAAATTTTAAATCTCCAATAGGAA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id            exon_id                     transcript_type  \\\n",
       "0   ENST00000417324.1  ENSE00001656588.1                              lncRNA   \n",
       "1   ENST00000417324.1  ENSE00001669267.1                              lncRNA   \n",
       "2   ENST00000450305.2  ENSE00001948541.1  transcribed_unprocessed_pseudogene   \n",
       "3   ENST00000450305.2  ENSE00001671638.2  transcribed_unprocessed_pseudogene   \n",
       "4   ENST00000450305.2  ENSE00001758273.2  transcribed_unprocessed_pseudogene   \n",
       "5   ENST00000450305.2  ENSE00001799933.2  transcribed_unprocessed_pseudogene   \n",
       "6   ENST00000450305.2  ENSE00001746346.2  transcribed_unprocessed_pseudogene   \n",
       "7   ENST00000456328.2  ENSE00002234944.1                processed_transcript   \n",
       "8   ENST00000456328.2  ENSE00003582793.1                processed_transcript   \n",
       "9   ENST00000461467.1  ENSE00001618781.2                              lncRNA   \n",
       "10  ENST00000469289.1  ENSE00001841699.1                              lncRNA   \n",
       "11  ENST00000473358.1  ENSE00001947070.1                              lncRNA   \n",
       "12  ENST00000473358.1  ENSE00001922571.1                              lncRNA   \n",
       "13  ENST00000488147.1  ENSE00001890219.1              unprocessed_pseudogene   \n",
       "14  ENST00000488147.1  ENSE00003507205.1              unprocessed_pseudogene   \n",
       "15  ENST00000488147.1  ENSE00003477500.1              unprocessed_pseudogene   \n",
       "16  ENST00000488147.1  ENSE00003565697.1              unprocessed_pseudogene   \n",
       "17  ENST00000488147.1  ENSE00003475637.1              unprocessed_pseudogene   \n",
       "18  ENST00000488147.1  ENSE00003502542.1              unprocessed_pseudogene   \n",
       "19  ENST00000488147.1  ENSE00003553898.1              unprocessed_pseudogene   \n",
       "20  ENST00000488147.1  ENSE00003621279.1              unprocessed_pseudogene   \n",
       "21  ENST00000488147.1  ENSE00002030414.1              unprocessed_pseudogene   \n",
       "22  ENST00000488147.1  ENSE00001935574.1              unprocessed_pseudogene   \n",
       "23  ENST00000492842.2  ENSE00001830178.2  transcribed_unprocessed_pseudogene   \n",
       "24  ENST00000606857.1  ENSE00003698237.1              unprocessed_pseudogene   \n",
       "25  ENST00000607096.1  ENSE00003695741.1                               miRNA   \n",
       "26  ENST00000619216.1  ENSE00003746039.1                               miRNA   \n",
       "27  ENST00000641515.2  ENSE00003812156.1                      protein_coding   \n",
       "28  ENST00000641515.2  ENSE00003813641.1                      protein_coding   \n",
       "29  ENST00000642116.1  ENSE00003812686.1                processed_transcript   \n",
       "30  ENST00000642116.1  ENSE00003812505.1                processed_transcript   \n",
       "\n",
       "   strand coordinates                                           sequence  \n",
       "0       -       35720  GGGGGCAGGAATTGGGGGTGACAGGAGAGCTGTAAGGTCTCCAGTG...  \n",
       "1       -       35276  aagtctcactatgttgctcaggctggtcttgaactcctggcctcaa...  \n",
       "2       +       12058  TTTTTCTTTTCGTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTC...  \n",
       "3       +       12228  CTGCTGGCCTGTGCCAGGGTGCAAGCTGAGCACTGGAGTGGAGTTT...  \n",
       "4       +       12698  TGCAGAGACGGGAGGGGCAGAGCCGCAGGCACAGCCAAGAGGGCTG...  \n",
       "5       +       13053  AGCCCTGCTGCTTGCGGTGGCCTATAAAGCCTCCTAGTCTGGCTCC...  \n",
       "6       +       13375  GGGGAGGCAGCTGTAACTCAAAGCCTTAGCCTCTGTTCCCACGAAG...  \n",
       "7       +       12228  CTGCTGGCCTGTGCCAGGGTGCAAGCTGAGCACTGGAGTGGAGTTT...  \n",
       "8       +       12722  GCAGGCACAGCCAAGAGGGCTGAAGAAATGGTAGAACGGAGCAGCT...  \n",
       "9       -       35720  GGGGGCAGGAATTGGGGGTGACAGGAGAGCTGTAAGGTCTCCAGTG...  \n",
       "10      +       30668  tggttgtttatctgagattcagaattaagcattttatattttattt...  \n",
       "11      +       30040  CTCTGTGGGCAGCTCCCTTGTGATCTGCTTAGTTCCCACCCCCCTT...  \n",
       "12      +       30668  tggttgtttatctgagattcagaattaagcattttatattttattt...  \n",
       "13      -       29533  GCGCACCCCGGGCGAGGGCTGCCCGGAGGAGCCGAGGTTGGCGGAC...  \n",
       "14      -       24737  TGCCGGTGGTCGTGGGTGGCTCTGACGCTCCGGCTCCGACCCACAG...  \n",
       "15      -       18267  GGGCCTGGCCTGGCAGCTCTGCTGTTAGAAGCAGGAGGTGTGCAGA...  \n",
       "16      -       17914  CCTAGCCCTCTGAGCCTCTTGGGGCCCCCTCAGACTTGTCCCTGTG...  \n",
       "17      -       17605  ATGGCAGCTGATTAGGCCCAGCACCAAATATTCACCATCCCTTGGC...  \n",
       "18      -       17232  GGAGAGGAACATGGGCTCAGGACAGCGGGTGTCAGCTTGCCTGACC...  \n",
       "19      -       16857  AGGTCCCAGAGAACTACTTCTATGTGCCAGACCTGGGCCAGGTGCC...  \n",
       "20      -       16606  ACACACCTTCTTCCCTAGACACCCCACACTTTGTGTTTCAGACCTA...  \n",
       "21      -       15795  CCTGAGCAGCCTTTCCTCCTTGCCCTGTTCCCCACCTCCCGGCTCC...  \n",
       "22      -       15004  GTCTCATCAGTGCCCGGTGCTGGGTCAGGGATCGACTGAGGCTCTG...  \n",
       "23      +       63888  TGGTTTTGTTTTTTGCTCCATGCATGTTTCTCTACGTGTGGCCTTT...  \n",
       "24      +       53313  AGAATCCTTCTTCAGGAGACTTATCTAAAGCATTTGTGTCATGTTA...  \n",
       "25      +       30504  TCCGTTGGCTTGCCTCCGGCATTTTTGGCCCTTGCCTTTtagggtt...  \n",
       "26      -       17368  AGGCCCGGCCTGGCTTAGTGGGGACCCAGGGCATCAGACACAGGTA...  \n",
       "27      +       65434  TTTGCTGTGATTCTTTAAAAAGCACCTTTAGACTCAGTGAGATAGC...  \n",
       "28      +       65574  GACTATAAAAGTTAGTCTCAGTACACAAAGCTCAGACTGGCTATTC...  \n",
       "29      +       57654  CAAGAGAAACTAATAATGGGCTGGGTATTGTCTGTCTCAGCAGCAT...  \n",
       "30      +       58857  GTCCCAGACCCTAAATGACTGACAAATTTTAAATCTCCAATAGGAA...  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_donor_splice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "4899bfe3-b2ca-4823-8815-70627ba6dc19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>strand</th>\n",
       "      <th>coordinates</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>12178</td>\n",
       "      <td>GGGTCTTGATGCTGTGGTCTTCATCTGCAGGTGTCTGACTTCCAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001758273.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>12612</td>\n",
       "      <td>CAGGCAGGCCATCGCTGCCACAGAACCCAGTGGATTGGCCTAGGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>12612</td>\n",
       "      <td>CAGGCAGGCCATCGCTGCCACAGAACCCAGTGGATTGGCCTAGGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001799933.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>12974</td>\n",
       "      <td>GGCCGGCGTCTCCTGTCTCCTGGAGAGGCTTCGATGCCCCTCCACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001746346.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>13220</td>\n",
       "      <td>CTGTGCATGAAGGCTGTCAACCAGTCCATAGGCAAGCCTGGCTGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>13220</td>\n",
       "      <td>CTGTGCATGAAGGCTGTCAACCAGTCCATAGGCAAGCCTGGCTGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001863096.1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>13452</td>\n",
       "      <td>GCATAGTGCTCCTGGACCAGTGATACACCCGGCACCCTGTCCTGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00001843071.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>14502</td>\n",
       "      <td>CGTAGGGGGCTCCATGACACCTTCCCCCCCAGACCCAGACTTGGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00001935574.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>15039</td>\n",
       "      <td>GGTGGCTGGGAGGTGGCTCCTCAAACAGCCGCTGTCTCATCAGTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00002030414.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>15948</td>\n",
       "      <td>TGCCCGCGCTCAGGTGGTTGCAGAAGTCAGTGCCCAGGCCCCCACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003621279.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>16766</td>\n",
       "      <td>ACGACCTCATGTACATTGCCGACCTGGGCCCCGGCATTGCCCCCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003553898.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>17056</td>\n",
       "      <td>GAGAGAGCAGCTGGAACAGCAGGTGGGAGGGGTGGGACAGAGGTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003502542.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>17369</td>\n",
       "      <td>GGCCCGGCCTGGCTTAGTGGGGACCCAGGGCATCAGACACAGGTAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003475637.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>17743</td>\n",
       "      <td>CGCCCCCTGGACGAGCGGGCCCTGCAGGTCTGCTGGCCGCGCATAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003565697.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>18062</td>\n",
       "      <td>CCATACCCCTGTGTCCTGAGGCTACTGGGCAGTCCCTCCATTTCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003477500.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>18367</td>\n",
       "      <td>CGACCCTGCAGGTCAGATTATCCTTGTAGAGGCCCCCTGGATGGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003507205.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>24892</td>\n",
       "      <td>TCAGCAGGCATCAGCAGAGTCCCATCTATGCGCTCCTGGCCTCTCA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id            exon_id                     transcript_type  \\\n",
       "0   ENST00000450305.2  ENSE00001671638.2  transcribed_unprocessed_pseudogene   \n",
       "1   ENST00000450305.2  ENSE00001758273.2  transcribed_unprocessed_pseudogene   \n",
       "5   ENST00000456328.2  ENSE00003582793.1                processed_transcript   \n",
       "2   ENST00000450305.2  ENSE00001799933.2  transcribed_unprocessed_pseudogene   \n",
       "3   ENST00000450305.2  ENSE00001746346.2  transcribed_unprocessed_pseudogene   \n",
       "6   ENST00000456328.2  ENSE00002312635.1                processed_transcript   \n",
       "4   ENST00000450305.2  ENSE00001863096.1  transcribed_unprocessed_pseudogene   \n",
       "16  ENST00000488147.1  ENSE00001843071.1              unprocessed_pseudogene   \n",
       "15  ENST00000488147.1  ENSE00001935574.1              unprocessed_pseudogene   \n",
       "14  ENST00000488147.1  ENSE00002030414.1              unprocessed_pseudogene   \n",
       "13  ENST00000488147.1  ENSE00003621279.1              unprocessed_pseudogene   \n",
       "12  ENST00000488147.1  ENSE00003553898.1              unprocessed_pseudogene   \n",
       "11  ENST00000488147.1  ENSE00003502542.1              unprocessed_pseudogene   \n",
       "10  ENST00000488147.1  ENSE00003475637.1              unprocessed_pseudogene   \n",
       "9   ENST00000488147.1  ENSE00003565697.1              unprocessed_pseudogene   \n",
       "8   ENST00000488147.1  ENSE00003477500.1              unprocessed_pseudogene   \n",
       "7   ENST00000488147.1  ENSE00003507205.1              unprocessed_pseudogene   \n",
       "\n",
       "   strand coordinates                                           sequence  \n",
       "0       +       12178  GGGTCTTGATGCTGTGGTCTTCATCTGCAGGTGTCTGACTTCCAGC...  \n",
       "1       +       12612  CAGGCAGGCCATCGCTGCCACAGAACCCAGTGGATTGGCCTAGGTG...  \n",
       "5       +       12612  CAGGCAGGCCATCGCTGCCACAGAACCCAGTGGATTGGCCTAGGTG...  \n",
       "2       +       12974  GGCCGGCGTCTCCTGTCTCCTGGAGAGGCTTCGATGCCCCTCCACA...  \n",
       "3       +       13220  CTGTGCATGAAGGCTGTCAACCAGTCCATAGGCAAGCCTGGCTGCC...  \n",
       "6       +       13220  CTGTGCATGAAGGCTGTCAACCAGTCCATAGGCAAGCCTGGCTGCC...  \n",
       "4       +       13452  GCATAGTGCTCCTGGACCAGTGATACACCCGGCACCCTGTCCTGGA...  \n",
       "16      -       14502  CGTAGGGGGCTCCATGACACCTTCCCCCCCAGACCCAGACTTGGGC...  \n",
       "15      -       15039  GGTGGCTGGGAGGTGGCTCCTCAAACAGCCGCTGTCTCATCAGTGC...  \n",
       "14      -       15948  TGCCCGCGCTCAGGTGGTTGCAGAAGTCAGTGCCCAGGCCCCCACA...  \n",
       "13      -       16766  ACGACCTCATGTACATTGCCGACCTGGGCCCCGGCATTGCCCCCTC...  \n",
       "12      -       17056  GAGAGAGCAGCTGGAACAGCAGGTGGGAGGGGTGGGACAGAGGTGG...  \n",
       "11      -       17369  GGCCCGGCCTGGCTTAGTGGGGACCCAGGGCATCAGACACAGGTAC...  \n",
       "10      -       17743  CGCCCCCTGGACGAGCGGGCCCTGCAGGTCTGCTGGCCGCGCATAT...  \n",
       "9       -       18062  CCATACCCCTGTGTCCTGAGGCTACTGGGCAGTCCCTCCATTTCCC...  \n",
       "8       -       18367  CGACCCTGCAGGTCAGATTATCCTTGTAGAGGCCCCCTGGATGGCA...  \n",
       "7       -       24892  TCAGCAGGCATCAGCAGAGTCCCATCTATGCGCTCCTGGCCTCTCA...  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_acceptor_splice.sort_values(by=['coordinates'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "297c6d07-7e21-44d7-89fb-5e31c54e4c05",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_promoter.to_csv(data_save_path +\"/\"+ entry['seqname']+\"_promoter.csv\", index=False)\n",
    "# df_core_promoter.to_csv(data_save_path +\"/\"+ entry['seqname']+\"_core_promoter_sequence.csv\", index=False)\n",
    "# df_acceptor_splice.to_csv(data_save_path +\"/\"+ entry['seqname']+\"_acceptor_splice_set_sequence.csv\", index=False)\n",
    "# df_donor_splice.to_csv(data_save_path +\"/\"+ entry['seqname']+\"_donor_splice_set.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "8e5827cd-0942-4812-a7de-5a880a371c34",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ttctccacaaagtgtttacttttggatttttgccagtctaacaggtgaAGccctggagattcttattagtgatttgggctggggcctggccatgtgtatttttttaaatttccactgatgattttgctgcatggccggtgttgagaatgactgCGCAAATTTGCCGGATTTCCTTTGCTGTTCCTGCATGTAGTTTAAACGAGATTGCCAGCACCGGGTATCATTCACCATTTTTCTTTTCGTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTATTTGCTGT'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_seq = pybedtools.BedTool.seq('chr1:11628-11927', fasta)\n",
    "test_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "a72c986b-6260-435e-92d2-495a2c5bbd4b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACAGCAAATACACATGAACAGAAAGAAGAGGTCAAAGAAAAGGCTGACGGCAAGTTAACGAAAAGAAAAATGGTGAATGATACCCGGTGCTGGCAATCTCGTTTAAACTACATGCAGGAACAGCAAAGGAAATCCGGCAAATTTGCGcagtcattctcaacaccggccatgcagcaaaatcatcagtggaaatttaaaaaaatacacatggccaggccccagcccaaatcactaataagaatctccagggCTtcacctgttagactggcaaaaatccaaaagtaaacactttgtggagaa\n"
     ]
    }
   ],
   "source": [
    "seq = Seq(test_seq)\n",
    "\n",
    "print (seq.reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8dcec4ae-0d9a-49af-b046-04c107e2a0a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">chr1:1-10\n",
      "GATGAGTCT\n",
      ">chr1:50-55\n",
      "CCATC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "a = pybedtools.BedTool('''\n",
    "chr1 1 10\n",
    "chr1 50 55''', from_string=True)\n",
    "fasta = pybedtools.example_filename('test.fa')\n",
    "a = a.sequence(fi=fasta)\n",
    "print(open(a.seqfn).read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6567b85-bbf4-4245-906b-be1af53a30ab",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28b13586-5eae-4aec-b002-d54f0c97ba5b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "558c0161-5b2a-40f7-ab3e-f59f19032130",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "df_promoter = pd.DataFrame(columns= ['gene_id','transcript_id','strand','exon_id','transcript_type','TSS', 'sequence'])\n",
    "df_core_promoter = pd.DataFrame(columns= ['gene_id','transcript_id','strand','exon_id','transcript_type','TSS','sequence'])\n",
    "df_donor_splice = pd.DataFrame(columns= ['transcript_id','exon_id','transcript_type','strand','coordinates','sequence'])\n",
    "df_acceptor_splice= pd.DataFraame(columns= ['transcript_id','exon_id', 'transcript_type','strand','coordinates', 'sequence'])\n",
    "#global promoter_index, core_promoter_index, donor_index, acceptor_index\n",
    "promoter_index = core_promoter_index = donor_index = acceptor_index = 0\n",
    "\n",
    "df_by_transcript = df_chr1_transcript.head(15).groupby(\"transcript_id\")\n",
    "for transcript, entries in df_by_transcript:\n",
    "    print (\"\\n\", transcript,len( entries) )\n",
    "    print (entries.iloc[0]['strand'])\n",
    "    if (entries.iloc[0]['strand']=='+'):\n",
    "        for index, entry in entries.iterrows():\n",
    "            if (int(entry['exon_number'])==1):\n",
    "                print(\"TSS\", entry['start'])\n",
    "                promoter_sequence(entry, df_promoter, promoter_index)\n",
    "                promoter_index= promoter_index+1\n",
    "                \n",
    "                core_promoter_sequence(entry, df_core_promoter, core_promoter_index)\n",
    "                core_promoter_index = core_promoter_index +1\n",
    "                \n",
    "                donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                donor_index =donor_index+ 1\n",
    "                print(\"first donor_splice_site\", entry['end'])\n",
    "            elif (int(entry['exon_number'])==len(entries)):\n",
    "                print (\"Last acceptor_splice_site\",  entry['start'])\n",
    "                acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                acceptor_index = acceptor_index +1\n",
    "            else:\n",
    "                print (\"Acceptor splice site\", entry['start'])\n",
    "                acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                acceptor_index = acceptor_index +1\n",
    "                print (\"Donor splice site\", entry['end'])\n",
    "                donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                donor_index =donor_index+ 1\n",
    "    elif (entries.iloc[0]['strand']=='-'):\n",
    "        for index, entry in entries.iterrows():\n",
    "            if (int(entry['exon_number'])==1):\n",
    "                print(\"TSS\", entry['end'])\n",
    "                print(\"first donor_splice_site\", entry['start'])\n",
    "                \n",
    "                promoter_sequence(entry, df_promoter, promoter_index)\n",
    "                promoter_index= promoter_index+1\n",
    "                \n",
    "                core_promoter_sequence(entry, df_core_promoter, core_promoter_index)\n",
    "                core_promoter_index = core_promoter_index +1\n",
    "                \n",
    "                donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                donor_index =donor_index+ 1\n",
    "                \n",
    "                \n",
    "            elif (int(entry['exon_number'])==len(entries)):\n",
    "                print (\"Last acceptor_splice_site\",  entry['end'])\n",
    "                acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                acceptor_index = acceptor_index +1\n",
    "            else:\n",
    "                print (\"Acceptor splice site\", entry['end'])\n",
    "                acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                acceptor_index = acceptor_index +1\n",
    "                print (\"Donor splice site\", entry['start'])\n",
    "                donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                donor_index =donor_index+ 1"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
