{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "48bb6eb5-6de0-4d27-848d-5c2e2eb29faf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, tarfile\n",
    "import pyBigWig\n",
    "import pybedtools\n",
    "import zipfile, gzip\n",
    "from gtfparse import read_gtf\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None)\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "f4f24891-9cb6-4b3d-b75a-13c312c4bcf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_genome_path = \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/\"\n",
    "data_path= \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/gencode_annotation/\"\n",
    "file_name=\"gencodeV38.bb\"\n",
    "gencode_path= \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/gencode_annotation/\"\n",
    "gtf_annotation_filename= \"gencode.v38.annotation.gtf\"\n",
    "gff_annotation_filename= \"gencode.v38.annotation.gff3\"\n",
    "file_path= data_path+file_name\n",
    "gencode_gtf_file_path= gencode_path+ gtf_annotation_filename\n",
    "gencode_gff_file_path= gencode_path+ gff_annotation_filename\n",
    "#reference_genome_file_path = reference_genome_path + reference_genome_name\n",
    "data_save_path= \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "271adad9-08d8-4917-bcaf-e22363625388",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_genome_file_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "89017f10-1f85-49c8-944d-2fabe5273a4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['chr10.fa',\n",
       " 'chr11.fa',\n",
       " 'chr12.fa',\n",
       " 'chr13.fa',\n",
       " 'chr14.fa',\n",
       " 'chr15.fa',\n",
       " 'chr16.fa',\n",
       " 'chr17.fa',\n",
       " 'chr18.fa',\n",
       " 'chr19.fa',\n",
       " 'chr2.fa',\n",
       " 'chr20.fa',\n",
       " 'chr21.fa',\n",
       " 'chr22.fa',\n",
       " 'chr3.fa',\n",
       " 'chr4.fa',\n",
       " 'chr5.fa',\n",
       " 'chr6.fa',\n",
       " 'chr7.fa',\n",
       " 'chr8.fa',\n",
       " 'chr9.fa',\n",
       " 'chrX.fa',\n",
       " 'chrY.fa',\n",
       " 'keep']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr_files= os.listdir(reference_genome_path )\n",
    "chr_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8e286976-54a1-4993-99d4-30f7b4c4880e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'hgnc_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'ont', 'protein_id', 'ccdsid']\n"
     ]
    }
   ],
   "source": [
    "df = read_gtf(gencode_gtf_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bf9498db-5711-4b53-99c0-d8b8de788bcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand',\n",
      "       'frame', 'gene_id', 'gene_type', 'gene_name', 'level', 'hgnc_id',\n",
      "       'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name',\n",
      "       'transcript_support_level', 'tag', 'havana_transcript', 'exon_number',\n",
      "       'exon_id', 'ont', 'protein_id', 'ccdsid'],\n",
      "      dtype='object')\n",
      "(3150424, 25)\n"
     ]
    }
   ],
   "source": [
    "print(df.columns)\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "30f808e1-6f52-447e-ba2b-525ce83222b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010</td>\n",
       "      <td>12057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12697</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001758273.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12975</td>\n",
       "      <td>13052</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00001799933.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  source     feature  start    end  score strand frame  \\\n",
       "0    chr1  HAVANA        gene  11869  14409    NaN      +     0   \n",
       "1    chr1  HAVANA  transcript  11869  14409    NaN      +     0   \n",
       "2    chr1  HAVANA        exon  11869  12227    NaN      +     0   \n",
       "3    chr1  HAVANA        exon  12613  12721    NaN      +     0   \n",
       "4    chr1  HAVANA        exon  13221  14409    NaN      +     0   \n",
       "5    chr1  HAVANA  transcript  12010  13670    NaN      +     0   \n",
       "6    chr1  HAVANA        exon  12010  12057    NaN      +     0   \n",
       "7    chr1  HAVANA        exon  12179  12227    NaN      +     0   \n",
       "8    chr1  HAVANA        exon  12613  12697    NaN      +     0   \n",
       "9    chr1  HAVANA        exon  12975  13052    NaN      +     0   \n",
       "\n",
       "             gene_id                           gene_type gene_name level  \\\n",
       "0  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "1  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "2  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "3  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "4  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "5  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "6  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "7  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "8  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "9  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "\n",
       "      hgnc_id           havana_gene      transcript_id  \\\n",
       "0  HGNC:37102  OTTHUMG00000000961.2                      \n",
       "1  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "2  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "3  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "4  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "5  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "6  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "7  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "8  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "9  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "\n",
       "                      transcript_type transcript_name  \\\n",
       "0                                                       \n",
       "1                processed_transcript     DDX11L1-202   \n",
       "2                processed_transcript     DDX11L1-202   \n",
       "3                processed_transcript     DDX11L1-202   \n",
       "4                processed_transcript     DDX11L1-202   \n",
       "5  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "6  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "7  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "8  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "9  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "\n",
       "  transcript_support_level                      tag     havana_transcript  \\\n",
       "0                                                                           \n",
       "1                        1                    basic  OTTHUMT00000362751.1   \n",
       "2                        1                    basic  OTTHUMT00000362751.1   \n",
       "3                        1                    basic  OTTHUMT00000362751.1   \n",
       "4                        1                    basic  OTTHUMT00000362751.1   \n",
       "5                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "6                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "7                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "8                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "9                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "\n",
       "  exon_number            exon_id                      ont protein_id ccdsid  \n",
       "0                                                                            \n",
       "1                                                                            \n",
       "2           1  ENSE00002234944.1                                             \n",
       "3           2  ENSE00003582793.1                                             \n",
       "4           3  ENSE00002312635.1                                             \n",
       "5                                 PGO:0000005,PGO:0000019                    \n",
       "6           1  ENSE00001948541.1  PGO:0000005,PGO:0000019                    \n",
       "7           2  ENSE00001671638.2  PGO:0000005,PGO:0000019                    \n",
       "8           3  ENSE00001758273.2  PGO:0000005,PGO:0000019                    \n",
       "9           4  ENSE00001799933.2  PGO:0000005,PGO:0000019                    "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5c1381e3-3325-4b15-b1b8-80b27b8ac306",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "feature\n",
       "CDS                825028\n",
       "Selenocysteine        119\n",
       "UTR                349402\n",
       "exon              1499012\n",
       "gene                60649\n",
       "start_codon         92997\n",
       "stop_codon          86205\n",
       "transcript         237012\n",
       "dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby(['feature']).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "06945eba-fcdb-4f16-a5ec-fa94e936c8ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(103034, 25)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_chr1 = df[df['seqname']=='chrX']\n",
    "df_chr1.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4ca223a6-8264-451d-9e50-9252755d1200",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>253743</td>\n",
       "      <td>253846</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000228572.7_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>LL0YNC03-29C1.1</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td>OTTHUMG00000046372.2</td>\n",
       "      <td>ENST00000431238.7_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>LL0YNC03-29C1.1-001</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical,PAR</td>\n",
       "      <td>OTTHUMT00000107057.2</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001702291.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>254937</td>\n",
       "      <td>255091</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000228572.7_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>LL0YNC03-29C1.1</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td>OTTHUMG00000046372.2</td>\n",
       "      <td>ENST00000431238.7_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>LL0YNC03-29C1.1-001</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical,PAR</td>\n",
       "      <td>OTTHUMT00000107057.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001655436.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>276322</td>\n",
       "      <td>276394</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000182378.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:23148</td>\n",
       "      <td>OTTHUMG00000022693.6</td>\n",
       "      <td>ENST00000399012.6_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1-202</td>\n",
       "      <td>5</td>\n",
       "      <td>basic,appris_principal_1,CCDS,PAR</td>\n",
       "      <td>OTTHUMT00000058878.2</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001489430.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000381976.1</td>\n",
       "      <td>CCDS14103.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>281482</td>\n",
       "      <td>281684</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000182378.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:23148</td>\n",
       "      <td>OTTHUMG00000022693.6</td>\n",
       "      <td>ENST00000399012.6_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1-202</td>\n",
       "      <td>5</td>\n",
       "      <td>basic,appris_principal_1,CCDS,PAR</td>\n",
       "      <td>OTTHUMT00000058878.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001306908.2</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000381976.1</td>\n",
       "      <td>CCDS14103.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>284167</td>\n",
       "      <td>284314</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000182378.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:23148</td>\n",
       "      <td>OTTHUMG00000022693.6</td>\n",
       "      <td>ENST00000399012.6_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1-202</td>\n",
       "      <td>5</td>\n",
       "      <td>basic,appris_principal_1,CCDS,PAR</td>\n",
       "      <td>OTTHUMT00000058878.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003812308.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000381976.1</td>\n",
       "      <td>CCDS14103.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>288733</td>\n",
       "      <td>288869</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000182378.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:23148</td>\n",
       "      <td>OTTHUMG00000022693.6</td>\n",
       "      <td>ENST00000399012.6_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1-202</td>\n",
       "      <td>5</td>\n",
       "      <td>basic,appris_principal_1,CCDS,PAR</td>\n",
       "      <td>OTTHUMT00000058878.2</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00003812426.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000381976.1</td>\n",
       "      <td>CCDS14103.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>290648</td>\n",
       "      <td>290776</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000182378.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:23148</td>\n",
       "      <td>OTTHUMG00000022693.6</td>\n",
       "      <td>ENST00000399012.6_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1-202</td>\n",
       "      <td>5</td>\n",
       "      <td>basic,appris_principal_1,CCDS,PAR</td>\n",
       "      <td>OTTHUMT00000058878.2</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00003812724.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000381976.1</td>\n",
       "      <td>CCDS14103.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>291499</td>\n",
       "      <td>291654</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000182378.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:23148</td>\n",
       "      <td>OTTHUMG00000022693.6</td>\n",
       "      <td>ENST00000399012.6_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1-202</td>\n",
       "      <td>5</td>\n",
       "      <td>basic,appris_principal_1,CCDS,PAR</td>\n",
       "      <td>OTTHUMT00000058878.2</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00003811639.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000381976.1</td>\n",
       "      <td>CCDS14103.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>293035</td>\n",
       "      <td>293218</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000182378.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:23148</td>\n",
       "      <td>OTTHUMG00000022693.6</td>\n",
       "      <td>ENST00000399012.6_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1-202</td>\n",
       "      <td>5</td>\n",
       "      <td>basic,appris_principal_1,CCDS,PAR</td>\n",
       "      <td>OTTHUMT00000058878.2</td>\n",
       "      <td>7</td>\n",
       "      <td>ENSE00003812899.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000381976.1</td>\n",
       "      <td>CCDS14103.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>299097</td>\n",
       "      <td>303353</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000182378.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:23148</td>\n",
       "      <td>OTTHUMG00000022693.6</td>\n",
       "      <td>ENST00000399012.6_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1-202</td>\n",
       "      <td>5</td>\n",
       "      <td>basic,appris_principal_1,CCDS,PAR</td>\n",
       "      <td>OTTHUMT00000058878.2</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSE00001489377.2</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000381976.1</td>\n",
       "      <td>CCDS14103.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>276324</td>\n",
       "      <td>276394</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000182378.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:23148</td>\n",
       "      <td>OTTHUMG00000022693.6</td>\n",
       "      <td>ENST00000484611.7_PAR_Y</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>PLCXD1-210</td>\n",
       "      <td>4</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000393211.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001536003.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>284167</td>\n",
       "      <td>284314</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000182378.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:23148</td>\n",
       "      <td>OTTHUMG00000022693.6</td>\n",
       "      <td>ENST00000484611.7_PAR_Y</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>PLCXD1-210</td>\n",
       "      <td>4</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000393211.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003657744.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>288733</td>\n",
       "      <td>288869</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000182378.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:23148</td>\n",
       "      <td>OTTHUMG00000022693.6</td>\n",
       "      <td>ENST00000484611.7_PAR_Y</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>PLCXD1-210</td>\n",
       "      <td>4</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000393211.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003581159.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>290648</td>\n",
       "      <td>290792</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000182378.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:23148</td>\n",
       "      <td>OTTHUMG00000022693.6</td>\n",
       "      <td>ENST00000484611.7_PAR_Y</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>PLCXD1-210</td>\n",
       "      <td>4</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000393211.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00001939272.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>291499</td>\n",
       "      <td>291537</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000182378.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:23148</td>\n",
       "      <td>OTTHUMG00000022693.6</td>\n",
       "      <td>ENST00000484611.7_PAR_Y</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>PLCXD1-210</td>\n",
       "      <td>4</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000393211.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00002190604.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqname  source feature   start     end  score strand frame  \\\n",
       "0     chrY  HAVANA    exon  253743  253846    NaN      +     0   \n",
       "1     chrY  HAVANA    exon  254937  255091    NaN      +     0   \n",
       "2     chrY  HAVANA    exon  276322  276394    NaN      +     0   \n",
       "3     chrY  HAVANA    exon  281482  281684    NaN      +     0   \n",
       "4     chrY  HAVANA    exon  284167  284314    NaN      +     0   \n",
       "5     chrY  HAVANA    exon  288733  288869    NaN      +     0   \n",
       "6     chrY  HAVANA    exon  290648  290776    NaN      +     0   \n",
       "7     chrY  HAVANA    exon  291499  291654    NaN      +     0   \n",
       "8     chrY  HAVANA    exon  293035  293218    NaN      +     0   \n",
       "9     chrY  HAVANA    exon  299097  303353    NaN      +     0   \n",
       "10    chrY  HAVANA    exon  276324  276394    NaN      +     0   \n",
       "11    chrY  HAVANA    exon  284167  284314    NaN      +     0   \n",
       "12    chrY  HAVANA    exon  288733  288869    NaN      +     0   \n",
       "13    chrY  HAVANA    exon  290648  290792    NaN      +     0   \n",
       "14    chrY  HAVANA    exon  291499  291537    NaN      +     0   \n",
       "\n",
       "                     gene_id               gene_type        gene_name level  \\\n",
       "0    ENSG00000228572.7_PAR_Y  unprocessed_pseudogene  LL0YNC03-29C1.1     2   \n",
       "1    ENSG00000228572.7_PAR_Y  unprocessed_pseudogene  LL0YNC03-29C1.1     2   \n",
       "2   ENSG00000182378.15_PAR_Y          protein_coding           PLCXD1     2   \n",
       "3   ENSG00000182378.15_PAR_Y          protein_coding           PLCXD1     2   \n",
       "4   ENSG00000182378.15_PAR_Y          protein_coding           PLCXD1     2   \n",
       "5   ENSG00000182378.15_PAR_Y          protein_coding           PLCXD1     2   \n",
       "6   ENSG00000182378.15_PAR_Y          protein_coding           PLCXD1     2   \n",
       "7   ENSG00000182378.15_PAR_Y          protein_coding           PLCXD1     2   \n",
       "8   ENSG00000182378.15_PAR_Y          protein_coding           PLCXD1     2   \n",
       "9   ENSG00000182378.15_PAR_Y          protein_coding           PLCXD1     2   \n",
       "10  ENSG00000182378.15_PAR_Y          protein_coding           PLCXD1     2   \n",
       "11  ENSG00000182378.15_PAR_Y          protein_coding           PLCXD1     2   \n",
       "12  ENSG00000182378.15_PAR_Y          protein_coding           PLCXD1     2   \n",
       "13  ENSG00000182378.15_PAR_Y          protein_coding           PLCXD1     2   \n",
       "14  ENSG00000182378.15_PAR_Y          protein_coding           PLCXD1     2   \n",
       "\n",
       "       hgnc_id           havana_gene            transcript_id  \\\n",
       "0               OTTHUMG00000046372.2  ENST00000431238.7_PAR_Y   \n",
       "1               OTTHUMG00000046372.2  ENST00000431238.7_PAR_Y   \n",
       "2   HGNC:23148  OTTHUMG00000022693.6  ENST00000399012.6_PAR_Y   \n",
       "3   HGNC:23148  OTTHUMG00000022693.6  ENST00000399012.6_PAR_Y   \n",
       "4   HGNC:23148  OTTHUMG00000022693.6  ENST00000399012.6_PAR_Y   \n",
       "5   HGNC:23148  OTTHUMG00000022693.6  ENST00000399012.6_PAR_Y   \n",
       "6   HGNC:23148  OTTHUMG00000022693.6  ENST00000399012.6_PAR_Y   \n",
       "7   HGNC:23148  OTTHUMG00000022693.6  ENST00000399012.6_PAR_Y   \n",
       "8   HGNC:23148  OTTHUMG00000022693.6  ENST00000399012.6_PAR_Y   \n",
       "9   HGNC:23148  OTTHUMG00000022693.6  ENST00000399012.6_PAR_Y   \n",
       "10  HGNC:23148  OTTHUMG00000022693.6  ENST00000484611.7_PAR_Y   \n",
       "11  HGNC:23148  OTTHUMG00000022693.6  ENST00000484611.7_PAR_Y   \n",
       "12  HGNC:23148  OTTHUMG00000022693.6  ENST00000484611.7_PAR_Y   \n",
       "13  HGNC:23148  OTTHUMG00000022693.6  ENST00000484611.7_PAR_Y   \n",
       "14  HGNC:23148  OTTHUMG00000022693.6  ENST00000484611.7_PAR_Y   \n",
       "\n",
       "           transcript_type      transcript_name transcript_support_level  \\\n",
       "0   unprocessed_pseudogene  LL0YNC03-29C1.1-001                       NA   \n",
       "1   unprocessed_pseudogene  LL0YNC03-29C1.1-001                       NA   \n",
       "2           protein_coding           PLCXD1-202                        5   \n",
       "3           protein_coding           PLCXD1-202                        5   \n",
       "4           protein_coding           PLCXD1-202                        5   \n",
       "5           protein_coding           PLCXD1-202                        5   \n",
       "6           protein_coding           PLCXD1-202                        5   \n",
       "7           protein_coding           PLCXD1-202                        5   \n",
       "8           protein_coding           PLCXD1-202                        5   \n",
       "9           protein_coding           PLCXD1-202                        5   \n",
       "10    processed_transcript           PLCXD1-210                        4   \n",
       "11    processed_transcript           PLCXD1-210                        4   \n",
       "12    processed_transcript           PLCXD1-210                        4   \n",
       "13    processed_transcript           PLCXD1-210                        4   \n",
       "14    processed_transcript           PLCXD1-210                        4   \n",
       "\n",
       "                                  tag     havana_transcript exon_number  \\\n",
       "0         basic,Ensembl_canonical,PAR  OTTHUMT00000107057.2           1   \n",
       "1         basic,Ensembl_canonical,PAR  OTTHUMT00000107057.2           2   \n",
       "2   basic,appris_principal_1,CCDS,PAR  OTTHUMT00000058878.2           1   \n",
       "3   basic,appris_principal_1,CCDS,PAR  OTTHUMT00000058878.2           2   \n",
       "4   basic,appris_principal_1,CCDS,PAR  OTTHUMT00000058878.2           3   \n",
       "5   basic,appris_principal_1,CCDS,PAR  OTTHUMT00000058878.2           4   \n",
       "6   basic,appris_principal_1,CCDS,PAR  OTTHUMT00000058878.2           5   \n",
       "7   basic,appris_principal_1,CCDS,PAR  OTTHUMT00000058878.2           6   \n",
       "8   basic,appris_principal_1,CCDS,PAR  OTTHUMT00000058878.2           7   \n",
       "9   basic,appris_principal_1,CCDS,PAR  OTTHUMT00000058878.2           8   \n",
       "10                                PAR  OTTHUMT00000393211.1           1   \n",
       "11                                PAR  OTTHUMT00000393211.1           2   \n",
       "12                                PAR  OTTHUMT00000393211.1           3   \n",
       "13                                PAR  OTTHUMT00000393211.1           4   \n",
       "14                                PAR  OTTHUMT00000393211.1           5   \n",
       "\n",
       "              exon_id          ont         protein_id       ccdsid  \n",
       "0   ENSE00001702291.1  PGO:0000005                                  \n",
       "1   ENSE00001655436.1  PGO:0000005                                  \n",
       "2   ENSE00001489430.1               ENSP00000381976.1  CCDS14103.1  \n",
       "3   ENSE00001306908.2               ENSP00000381976.1  CCDS14103.1  \n",
       "4   ENSE00003812308.1               ENSP00000381976.1  CCDS14103.1  \n",
       "5   ENSE00003812426.1               ENSP00000381976.1  CCDS14103.1  \n",
       "6   ENSE00003812724.1               ENSP00000381976.1  CCDS14103.1  \n",
       "7   ENSE00003811639.1               ENSP00000381976.1  CCDS14103.1  \n",
       "8   ENSE00003812899.1               ENSP00000381976.1  CCDS14103.1  \n",
       "9   ENSE00001489377.2               ENSP00000381976.1  CCDS14103.1  \n",
       "10  ENSE00001536003.1                                               \n",
       "11  ENSE00003657744.1                                               \n",
       "12  ENSE00003581159.1                                               \n",
       "13  ENSE00001939272.2                                               \n",
       "14  ENSE00002190604.1                                               "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_chr1_transcript= df_chr1[df_chr1['feature']=='exon'].reset_index(drop=True)\n",
    "df_chr1_transcript.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "d082499d-929b-4dfd-b132-2e8b2ddc5cf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "promoter_coordinates = [-249, 50]\n",
    "core_promoter_coordinates= [-34, 35]\n",
    "donor_coordinates= [-199, 200]\n",
    "acceptor_coordinates= [-199, 200]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "0658de47-c801-426e-9644-3806712abade",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta = pybedtools.example_filename(reference_genome_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "1e0369d0-4941-4247-92e3-7553cac04e5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def acceptor_sequence(entry, df_acceptor_splice, acceptor_index):\n",
    "    print (\"@@@@\", acceptor_index)\n",
    "    df_acceptor_splice.loc[acceptor_index, ['transcript_id']]= entry['transcript_id']\n",
    "    df_acceptor_splice.loc[acceptor_index, ['exon_id']] = entry['exon_id']\n",
    "    df_acceptor_splice.loc[acceptor_index, ['transcript_type']] = entry['transcript_type']\n",
    "    df_acceptor_splice.loc[acceptor_index, ['strand']]= entry['strand']\n",
    "    if (entry['strand']=='+'):\n",
    "        df_acceptor_splice.loc[acceptor_index, ['coordinates']]= int(entry['start'])-1\n",
    "        acceptor_query= entry['seqname']+':'+str(int(entry['start'])+acceptor_coordinates[0])+'-'+str(int(entry['start'])+acceptor_coordinates[1])\n",
    "        print (\"##$$$\", acceptor_query)\n",
    "        df_acceptor_splice.loc[acceptor_index, ['sequence']] = pybedtools.BedTool.seq(acceptor_query, fasta)\n",
    "        print (\"###\", acceptor_index)\n",
    "    elif (entry['strand']=='-'):\n",
    "        df_acceptor_splice.loc[acceptor_index, ['coordinates']]= int(entry['end'])+1\n",
    "        acceptor_query= entry['seqname']+':'+str(int(entry['end'])+(acceptor_coordinates[1]*(-1)))+'-'+str(int(entry['end'])+(acceptor_coordinates[0]*(-1)))\n",
    "        temp_seq = pybedtools.BedTool.seq(acceptor_query, fasta)\n",
    "        rev_seq = str(Seq(temp_seq).reverse_complement())\n",
    "        df_acceptor_splice.loc[acceptor_index, ['sequence']] = rev_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "ccc5d858-46ea-421c-b7c0-f856cedb7869",
   "metadata": {},
   "outputs": [],
   "source": [
    "def donor_sequence(entry, df_donor_splice, donor_index):\n",
    "    df_donor_splice.loc[donor_index, ['transcript_id']]= entry['transcript_id']\n",
    "    df_donor_splice.loc[donor_index, ['exon_id']] = entry['exon_id']\n",
    "    df_donor_splice.loc[donor_index, ['transcript_type']] = entry['transcript_type']\n",
    "    df_donor_splice.loc[donor_index, ['strand']]= entry['strand']\n",
    "    if (entry['strand']=='+'):\n",
    "        df_donor_splice.loc[donor_index, ['coordinates']] = int(entry['end'])+1\n",
    "        donor_query= entry['seqname']+':'+str(int(entry['end'])+donor_coordinates[0])+'-'+str(int(entry['end'])+donor_coordinates[1])\n",
    "        print (\"##$$$\", donor_query, donor_index)\n",
    "        df_donor_splice.loc[donor_index, ['sequence']] = pybedtools.BedTool.seq(donor_query, fasta)\n",
    "    elif (entry['strand']=='-'):\n",
    "        df_donor_splice.loc[donor_index, ['coordinates']] = int(entry['start'])-1\n",
    "        donor_query= entry['seqname']+':'+str(int(entry['start'])+(donor_coordinates[1]*(-1)))+'-'+str(int(entry['start'])+(donor_coordinates[0]*(-1)))\n",
    "        print (\"##$$$\", donor_query, donor_index)\n",
    "        temp_seq = pybedtools.BedTool.seq(donor_query, fasta)\n",
    "        rev_seq = str(Seq(temp_seq).reverse_complement())\n",
    "        df_donor_splice.loc[donor_index, ['sequence']] = rev_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "2cbd9c22-f5b8-465f-932f-9aeb64f57a97",
   "metadata": {},
   "outputs": [],
   "source": [
    "def promoter_sequence(entry, df_promoter, promoter_index):\n",
    "    df_promoter.loc[promoter_index, ['gene_id']] = entry['gene_id']\n",
    "    df_promoter.loc[promoter_index, ['transcript_id']]= entry['transcript_id']\n",
    "    df_promoter.loc[promoter_index, ['transcript_type']]= entry['transcript_type']\n",
    "    df_promoter.loc[promoter_index, ['strand']] = entry['strand']\n",
    "    df_promoter.loc[promoter_index, ['exon_id']]=  entry['exon_id']\n",
    "    if (entry['strand']=='+'):\n",
    "        df_promoter.loc[promoter_index,['TSS']]= entry['start']\n",
    "        promoter_query= entry['seqname']+':'+str(int(entry['start'])+promoter_coordinates[0])+'-'+str(int(entry['start'])+promoter_coordinates[1])\n",
    "        df_promoter.loc[promoter_index, ['sequence']] = pybedtools.BedTool.seq(promoter_query, fasta)\n",
    "    elif(entry['strand']=='-'):\n",
    "        df_promoter.loc[promoter_index,['TSS']]= entry['end']\n",
    "        promoter_query= entry['seqname']+':'+str(int(entry['end'])+(promoter_coordinates[1]*(-1)))+'-'+str(int(entry['end'])+(promoter_coordinates[0]*(-1)))\n",
    "        temp_seq = pybedtools.BedTool.seq(promoter_query, fasta)\n",
    "        #print (\"Original\", temp_seq)\n",
    "        rev_seq = str(Seq(temp_seq).reverse_complement())\n",
    "        #print (rev_seq)\n",
    "        df_promoter.loc[promoter_index, ['sequence']] = rev_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "60c21aa4-2a16-4211-b305-bbfea865ebd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def core_promoter_sequence(entry, df_core_promoter, core_promoter_index):\n",
    "    df_core_promoter.loc[core_promoter_index, ['gene_id']] = entry['gene_id']\n",
    "    df_core_promoter.loc[core_promoter_index, ['transcript_id']] = entry['transcript_id']\n",
    "    df_core_promoter.loc[core_promoter_index, ['transcript_type']] = entry['transcript_type']\n",
    "    df_core_promoter.loc[core_promoter_index, ['strand']] = entry['strand']\n",
    "    df_core_promoter.loc[core_promoter_index, ['exon_id']]  =  entry['exon_id']\n",
    "    if (entry['strand']=='+'):\n",
    "        df_core_promoter.loc[core_promoter_index,['TSS']]= entry['start']\n",
    "        core_promoter_query= entry['seqname']+':'+str(int(entry['start'])+core_promoter_coordinates[0])+'-'+str(int(entry['start'])+core_promoter_coordinates[1])\n",
    "        df_core_promoter.loc[core_promoter_index, ['sequence']] = pybedtools.BedTool.seq(core_promoter_query, fasta)\n",
    "    elif(entry['strand']=='-'):\n",
    "        df_core_promoter.loc[core_promoter_index,['TSS']]= entry['end']\n",
    "        core_promoter_query= entry['seqname']+':'+str(int(entry['end'])+(core_promoter_coordinates[1]*(-1)))+'-'+str(int(entry['end'])+(core_promoter_coordinates[0]*(-1)))\n",
    "        temp_seq = pybedtools.BedTool.seq(core_promoter_query, fasta)\n",
    "        rev_seq = str(Seq(temp_seq).reverse_complement())\n",
    "        df_core_promoter.loc[core_promoter_index, ['sequence']] = rev_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "80df8252-7040-441d-93b5-2b03fd7ed87e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr10\n",
      "(57804, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr10\n",
      "\n",
      " ENST00000416477.1 6\n",
      "+\n",
      "TSS 44712\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr10.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr10:44702-45101 0\n",
      "first donor_splice_site 44901\n",
      "Acceptor splice site 44952\n",
      "@@@@ 0\n",
      "##$$$ chr10:44753-45152\n",
      "### 0\n",
      "Donor splice site 45204\n",
      "##$$$ chr10:45005-45404 1\n",
      "Acceptor splice site 45309\n",
      "@@@@ 1\n",
      "##$$$ chr10:45110-45509\n",
      "### 1\n",
      "Donor splice site 45405\n",
      "##$$$ chr10:45206-45605 2\n",
      "Acceptor splice site 45834\n",
      "@@@@ 2\n",
      "##$$$ chr10:45635-46034\n",
      "### 2\n",
      "Donor splice site 45882\n",
      "##$$$ chr10:45683-46082 3\n",
      "Acceptor splice site 46245\n",
      "@@@@ 3\n",
      "##$$$ chr10:46046-46445\n",
      "### 3\n",
      "Donor splice site 46359\n",
      "##$$$ chr10:46160-46559 4\n",
      "Last acceptor_splice_site 46842\n",
      "@@@@ 4\n",
      "##$$$ chr10:46643-47042\n",
      "### 4\n",
      "\n",
      " ENST00000562162.2 3\n",
      "-\n",
      "TSS 16549\n",
      "first donor_splice_site 16502\n",
      "##$$$ chr10:16302-16701 5\n",
      "Acceptor splice site 14659\n",
      "@@@@ 5\n",
      "Donor splice site 14497\n",
      "##$$$ chr10:14297-14696 6\n",
      "Last acceptor_splice_site 14299\n",
      "@@@@ 6\n",
      "\n",
      " ENST00000566940.2 3\n",
      "-\n",
      "TSS 20897\n",
      "first donor_splice_site 20786\n",
      "##$$$ chr10:20586-20985 7\n",
      "Acceptor splice site 16544\n",
      "@@@@ 7\n",
      "Donor splice site 16502\n",
      "##$$$ chr10:16302-16701 8\n",
      "Last acceptor_splice_site 14299\n",
      "@@@@ 8\n",
      "\n",
      " ENST00000568584.6 1\n",
      "-\n",
      "TSS 49296\n",
      "first donor_splice_site 49182\n",
      "##$$$ chr10:48982-49381 9\n",
      "\n",
      " ENST00000682078.1 4\n",
      "-\n",
      "TSS 20930\n",
      "first donor_splice_site 20786\n",
      "##$$$ chr10:20586-20985 10\n",
      "Acceptor splice site 16544\n",
      "@@@@ 9\n",
      "Donor splice site 16502\n",
      "##$$$ chr10:16302-16701 11\n",
      "Acceptor splice site 14588\n",
      "@@@@ 10\n",
      "Donor splice site 14497\n",
      "##$$$ chr10:14297-14696 12\n",
      "Last acceptor_splice_site 14299\n",
      "@@@@ 11\n",
      "\n",
      " ENST00000683798.1 3\n",
      "-\n",
      "TSS 20897\n",
      "first donor_splice_site 20690\n",
      "##$$$ chr10:20490-20889 13\n",
      "Acceptor splice site 16544\n",
      "@@@@ 12\n",
      "Donor splice site 16502\n",
      "##$$$ chr10:16302-16701 14\n",
      "Last acceptor_splice_site 14299\n",
      "@@@@ 13\n",
      "chr11\n",
      "(87515, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr11\n",
      "\n",
      " ENST00000424047.1 1\n",
      "-\n",
      "TSS 87586\n",
      "first donor_splice_site 86649\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr11.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr11:86449-86848 0\n",
      "\n",
      " ENST00000519787.1 1\n",
      "+\n",
      "TSS 75780\n",
      "##$$$ chr11:75944-76343 1\n",
      "first donor_splice_site 76143\n",
      "\n",
      " ENST00000526704.6 7\n",
      "-\n",
      "TSS 138998\n",
      "first donor_splice_site 138956\n",
      "##$$$ chr11:138756-139155 2\n",
      "Acceptor splice site 138111\n",
      "@@@@ 0\n",
      "Donor splice site 138047\n",
      "##$$$ chr11:137847-138246 3\n",
      "Acceptor splice site 133778\n",
      "@@@@ 1\n",
      "Donor splice site 133588\n",
      "##$$$ chr11:133388-133787 4\n",
      "Acceptor splice site 131524\n",
      "@@@@ 2\n",
      "Donor splice site 131467\n",
      "##$$$ chr11:131267-131666 5\n",
      "Acceptor splice site 131373\n",
      "@@@@ 3\n",
      "Donor splice site 129572\n",
      "##$$$ chr11:129372-129771 6\n",
      "Acceptor splice site 127648\n",
      "@@@@ 4\n",
      "Donor splice site 127514\n",
      "##$$$ chr11:127314-127713 7\n",
      "Last acceptor_splice_site 127462\n",
      "@@@@ 5\n",
      "\n",
      " ENST00000529266.1 2\n",
      "-\n",
      "TSS 125784\n",
      "first donor_splice_site 125578\n",
      "##$$$ chr11:125378-125777 8\n",
      "Last acceptor_splice_site 121426\n",
      "@@@@ 6\n",
      "\n",
      " ENST00000605013.1 2\n",
      "+\n",
      "TSS 123312\n",
      "##$$$ chr11:124901-125300 9\n",
      "first donor_splice_site 125100\n",
      "Last acceptor_splice_site 125172\n",
      "@@@@ 7\n",
      "##$$$ chr11:124973-125372\n",
      "### 7\n",
      "\n",
      " ENST00000607448.5 3\n",
      "-\n",
      "TSS 139612\n",
      "first donor_splice_site 139489\n",
      "##$$$ chr11:139289-139688 10\n",
      "Acceptor splice site 139152\n",
      "@@@@ 8\n",
      "Donor splice site 138956\n",
      "##$$$ chr11:138756-139155 11\n",
      "Last acceptor_splice_site 138111\n",
      "@@@@ 9\n",
      "\n",
      " ENST00000622626.1 4\n",
      "-\n",
      "TSS 125927\n",
      "first donor_splice_site 125578\n",
      "##$$$ chr11:125378-125777 12\n",
      "Acceptor splice site 121426\n",
      "@@@@ 10\n",
      "Donor splice site 121258\n",
      "##$$$ chr11:121058-121457 13\n",
      "Acceptor splice site 113174\n",
      "@@@@ 11\n",
      "Donor splice site 113116\n",
      "##$$$ chr11:112916-113315 14\n",
      "Last acceptor_splice_site 113111\n",
      "@@@@ 12\n",
      "chr12\n",
      "(84301, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr12\n",
      "\n",
      " ENST00000400706.3 11\n",
      "-\n",
      "TSS 32015\n",
      "first donor_splice_site 31878\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr12.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr12:31678-32077 0\n",
      "Acceptor splice site 26954\n",
      "@@@@ 0\n",
      "Donor splice site 26801\n",
      "##$$$ chr12:26601-27000 1\n",
      "Acceptor splice site 18471\n",
      "@@@@ 1\n",
      "Donor splice site 18373\n",
      "##$$$ chr12:18173-18572 2\n",
      "Acceptor splice site 18183\n",
      "@@@@ 2\n",
      "Donor splice site 18037\n",
      "##$$$ chr12:17837-18236 3\n",
      "Acceptor splice site 17859\n",
      "@@@@ 3\n",
      "Donor splice site 17723\n",
      "##$$$ chr12:17523-17922 4\n",
      "Acceptor splice site 17483\n",
      "@@@@ 4\n",
      "Donor splice site 17348\n",
      "##$$$ chr12:17148-17547 5\n",
      "Acceptor splice site 17170\n",
      "@@@@ 5\n",
      "Donor splice site 16969\n",
      "##$$$ chr12:16769-17168 6\n",
      "Acceptor splice site 16880\n",
      "@@@@ 6\n",
      "Donor splice site 16722\n",
      "##$$$ chr12:16522-16921 7\n",
      "Acceptor splice site 16065\n",
      "@@@@ 7\n",
      "Donor splice site 15913\n",
      "##$$$ chr12:15713-16112 8\n",
      "Acceptor splice site 15153\n",
      "@@@@ 8\n",
      "Donor splice site 15085\n",
      "##$$$ chr12:14885-15284 9\n",
      "Last acceptor_splice_site 14944\n",
      "@@@@ 9\n",
      "\n",
      " ENST00000504074.1 3\n",
      "-\n",
      "TSS 38133\n",
      "first donor_splice_site 37773\n",
      "##$$$ chr12:37573-37972 10\n",
      "Acceptor splice site 37529\n",
      "@@@@ 10\n",
      "Donor splice site 37325\n",
      "##$$$ chr12:37125-37524 11\n",
      "Last acceptor_splice_site 37222\n",
      "@@@@ 11\n",
      "\n",
      " ENST00000535849.1 4\n",
      "+\n",
      "TSS 12310\n",
      "##$$$ chr12:12159-12558 12\n",
      "first donor_splice_site 12358\n",
      "Acceptor splice site 12740\n",
      "@@@@ 12\n",
      "##$$$ chr12:12541-12940\n",
      "### 12\n",
      "Donor splice site 12824\n",
      "##$$$ chr12:12625-13024 13\n",
      "Acceptor splice site 13102\n",
      "@@@@ 13\n",
      "##$$$ chr12:12903-13302\n",
      "### 13\n",
      "Donor splice site 13201\n",
      "##$$$ chr12:13002-13401 14\n",
      "Last acceptor_splice_site 13370\n",
      "@@@@ 14\n",
      "##$$$ chr12:13171-13570\n",
      "### 14\n",
      "\n",
      " ENST00000546223.1 2\n",
      "-\n",
      "TSS 38102\n",
      "first donor_splice_site 37773\n",
      "##$$$ chr12:37573-37972 15\n",
      "Last acceptor_splice_site 37529\n",
      "@@@@ 15\n",
      "chr13\n",
      "(25521, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr13\n",
      "\n",
      " ENST00000623078.1 4\n",
      "-\n",
      "TSS 18178465\n",
      "first donor_splice_site 18177555\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr13.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr13:18177355-18177754 0\n",
      "Acceptor splice site 18176170\n",
      "@@@@ 0\n",
      "Donor splice site 18176018\n",
      "##$$$ chr13:18175818-18176217 1\n",
      "Acceptor splice site 18174512\n",
      "@@@@ 1\n",
      "Donor splice site 18174442\n",
      "##$$$ chr13:18174242-18174641 2\n",
      "Last acceptor_splice_site 18174103\n",
      "@@@@ 2\n",
      "\n",
      " ENST00000623511.1 8\n",
      "+\n",
      "TSS 18195297\n",
      "##$$$ chr13:18195212-18195611 3\n",
      "first donor_splice_site 18195411\n",
      "Acceptor splice site 18218999\n",
      "@@@@ 3\n",
      "##$$$ chr13:18218800-18219199\n",
      "### 3\n",
      "Donor splice site 18219057\n",
      "##$$$ chr13:18218858-18219257 4\n",
      "Acceptor splice site 18219526\n",
      "@@@@ 4\n",
      "##$$$ chr13:18219327-18219726\n",
      "### 4\n",
      "Donor splice site 18219550\n",
      "##$$$ chr13:18219351-18219750 5\n",
      "Acceptor splice site 18222495\n",
      "@@@@ 5\n",
      "##$$$ chr13:18222296-18222695\n",
      "### 5\n",
      "Donor splice site 18222581\n",
      "##$$$ chr13:18222382-18222781 6\n",
      "Acceptor splice site 18227479\n",
      "@@@@ 6\n",
      "##$$$ chr13:18227280-18227679\n",
      "### 6\n",
      "Donor splice site 18227544\n",
      "##$$$ chr13:18227345-18227744 7\n",
      "Acceptor splice site 18229547\n",
      "@@@@ 7\n",
      "##$$$ chr13:18229348-18229747\n",
      "### 7\n",
      "Donor splice site 18229608\n",
      "##$$$ chr13:18229409-18229808 8\n",
      "Acceptor splice site 18230074\n",
      "@@@@ 8\n",
      "##$$$ chr13:18229875-18230274\n",
      "### 8\n",
      "Donor splice site 18230107\n",
      "##$$$ chr13:18229908-18230307 9\n",
      "Last acceptor_splice_site 18230503\n",
      "@@@@ 9\n",
      "##$$$ chr13:18230304-18230703\n",
      "### 9\n",
      "\n",
      " ENST00000651051.1 1\n",
      "+\n",
      "TSS 18211959\n",
      "##$$$ chr13:18211863-18212262 10\n",
      "first donor_splice_site 18212062\n",
      "\n",
      " ENST00000651377.1 7\n",
      "+\n",
      "TSS 18195317\n",
      "##$$$ chr13:18195212-18195611 11\n",
      "first donor_splice_site 18195411\n",
      "Acceptor splice site 18218999\n",
      "@@@@ 10\n",
      "##$$$ chr13:18218800-18219199\n",
      "### 10\n",
      "Donor splice site 18219084\n",
      "##$$$ chr13:18218885-18219284 12\n",
      "Acceptor splice site 18222497\n",
      "@@@@ 11\n",
      "##$$$ chr13:18222298-18222697\n",
      "### 11\n",
      "Donor splice site 18222581\n",
      "##$$$ chr13:18222382-18222781 13\n",
      "Acceptor splice site 18227479\n",
      "@@@@ 12\n",
      "##$$$ chr13:18227280-18227679\n",
      "### 12\n",
      "Donor splice site 18227544\n",
      "##$$$ chr13:18227345-18227744 14\n",
      "Acceptor splice site 18229547\n",
      "@@@@ 13\n",
      "##$$$ chr13:18229348-18229747\n",
      "### 13\n",
      "Donor splice site 18229608\n",
      "##$$$ chr13:18229409-18229808 15\n",
      "Acceptor splice site 18230074\n",
      "@@@@ 14\n",
      "##$$$ chr13:18229875-18230274\n",
      "### 14\n",
      "Donor splice site 18230107\n",
      "##$$$ chr13:18229908-18230307 16\n",
      "Last acceptor_splice_site 18230503\n",
      "@@@@ 15\n",
      "##$$$ chr13:18230304-18230703\n",
      "### 15\n",
      "chr14\n",
      "(53913, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr14\n",
      "\n",
      " ENST00000384179.1 1\n",
      "+\n",
      "TSS 18343038\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr14.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr14:18342945-18343344 0\n",
      "first donor_splice_site 18343144\n",
      "\n",
      " ENST00000514609.6 5\n",
      "+\n",
      "TSS 18333726\n",
      "##$$$ chr14:18333701-18334100 1\n",
      "first donor_splice_site 18333900\n",
      "Acceptor splice site 18337973\n",
      "@@@@ 0\n",
      "##$$$ chr14:18337774-18338173\n",
      "### 0\n",
      "Donor splice site 18338078\n",
      "##$$$ chr14:18337879-18338278 2\n",
      "Acceptor splice site 18338243\n",
      "@@@@ 1\n",
      "##$$$ chr14:18338044-18338443\n",
      "### 1\n",
      "Donor splice site 18338416\n",
      "##$$$ chr14:18338217-18338616 3\n",
      "Acceptor splice site 18340614\n",
      "@@@@ 2\n",
      "##$$$ chr14:18340415-18340814\n",
      "### 2\n",
      "Donor splice site 18340726\n",
      "##$$$ chr14:18340527-18340926 4\n",
      "Last acceptor_splice_site 18341694\n",
      "@@@@ 3\n",
      "##$$$ chr14:18341495-18341894\n",
      "### 3\n",
      "\n",
      " ENST00000547576.2 3\n",
      "-\n",
      "TSS 18412264\n",
      "first donor_splice_site 18412135\n",
      "##$$$ chr14:18411935-18412334 5\n",
      "Acceptor splice site 18411618\n",
      "@@@@ 4\n",
      "Donor splice site 18411540\n",
      "##$$$ chr14:18411340-18411739 6\n",
      "Last acceptor_splice_site 18396941\n",
      "@@@@ 5\n",
      "\n",
      " ENST00000549046.1 1\n",
      "+\n",
      "TSS 18560047\n",
      "##$$$ chr14:18559992-18560391 7\n",
      "first donor_splice_site 18560191\n",
      "\n",
      " ENST00000549284.2 1\n",
      "+\n",
      "TSS 18418775\n",
      "##$$$ chr14:18419074-18419473 8\n",
      "first donor_splice_site 18419273\n",
      "\n",
      " ENST00000551951.1 3\n",
      "+\n",
      "TSS 18562928\n",
      "##$$$ chr14:18562876-18563275 9\n",
      "first donor_splice_site 18563075\n",
      "Acceptor splice site 18563822\n",
      "@@@@ 6\n",
      "##$$$ chr14:18563623-18564022\n",
      "### 6\n",
      "Donor splice site 18563911\n",
      "##$$$ chr14:18563712-18564111 10\n",
      "Last acceptor_splice_site 18565444\n",
      "@@@@ 7\n",
      "##$$$ chr14:18565245-18565644\n",
      "### 7\n",
      "\n",
      " ENST00000552261.1 2\n",
      "-\n",
      "TSS 18596310\n",
      "first donor_splice_site 18596259\n",
      "##$$$ chr14:18596059-18596458 11\n",
      "Last acceptor_splice_site 18556122\n",
      "@@@@ 8\n",
      "\n",
      " ENST00000553170.1 3\n",
      "+\n",
      "TSS 18333826\n",
      "##$$$ chr14:18333697-18334096 12\n",
      "first donor_splice_site 18333896\n",
      "Acceptor splice site 18338243\n",
      "@@@@ 9\n",
      "##$$$ chr14:18338044-18338443\n",
      "### 9\n",
      "Donor splice site 18338416\n",
      "##$$$ chr14:18338217-18338616 13\n",
      "Last acceptor_splice_site 18340614\n",
      "@@@@ 10\n",
      "##$$$ chr14:18340415-18340814\n",
      "### 10\n",
      "\n",
      " ENST00000636054.1 1\n",
      "+\n",
      "TSS 16057472\n",
      "##$$$ chr14:16057423-16057822 14\n",
      "first donor_splice_site 16057622\n",
      "chr15\n",
      "(56560, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr15\n",
      "\n",
      " ENST00000338912.5 2\n",
      "-\n",
      "TSS 19965101\n",
      "first donor_splice_site 19965056\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr15.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr15:19964856-19965255 0\n",
      "Last acceptor_splice_site 19964972\n",
      "@@@@ 0\n",
      "\n",
      " ENST00000364371.1 1\n",
      "+\n",
      "TSS 19883614\n",
      "##$$$ chr15:19883517-19883916 1\n",
      "first donor_splice_site 19883716\n",
      "\n",
      " ENST00000524329.1 6\n",
      "+\n",
      "TSS 19878555\n",
      "##$$$ chr15:19878469-19878868 2\n",
      "first donor_splice_site 19878668\n",
      "Acceptor splice site 19878831\n",
      "@@@@ 1\n",
      "##$$$ chr15:19878632-19879031\n",
      "### 1\n",
      "Donor splice site 19879004\n",
      "##$$$ chr15:19878805-19879204 3\n",
      "Acceptor splice site 19881201\n",
      "@@@@ 2\n",
      "##$$$ chr15:19881002-19881401\n",
      "### 2\n",
      "Donor splice site 19881307\n",
      "##$$$ chr15:19881108-19881507 4\n",
      "Acceptor splice site 19882277\n",
      "@@@@ 3\n",
      "##$$$ chr15:19882078-19882477\n",
      "### 3\n",
      "Donor splice site 19882439\n",
      "##$$$ chr15:19882240-19882639 5\n",
      "Acceptor splice site 19884979\n",
      "@@@@ 4\n",
      "##$$$ chr15:19884780-19885179\n",
      "### 4\n",
      "Donor splice site 19885043\n",
      "##$$$ chr15:19884844-19885243 6\n",
      "Last acceptor_splice_site 19887724\n",
      "@@@@ 5\n",
      "##$$$ chr15:19887525-19887924\n",
      "### 5\n",
      "\n",
      " ENST00000553634.1 1\n",
      "+\n",
      "TSS 19899334\n",
      "##$$$ chr15:19899360-19899759 7\n",
      "first donor_splice_site 19899559\n",
      "\n",
      " ENST00000555979.1 1\n",
      "-\n",
      "TSS 19968352\n",
      "first donor_splice_site 19967414\n",
      "##$$$ chr15:19967214-19967613 8\n",
      "\n",
      " ENST00000558565.2 2\n",
      "-\n",
      "TSS 19988117\n",
      "first donor_splice_site 19988072\n",
      "##$$$ chr15:19987872-19988271 9\n",
      "Last acceptor_splice_site 19987968\n",
      "@@@@ 6\n",
      "\n",
      " ENST00000558824.4 2\n",
      "-\n",
      "TSS 19973218\n",
      "first donor_splice_site 19973173\n",
      "##$$$ chr15:19972973-19973372 10\n",
      "Last acceptor_splice_site 19973088\n",
      "@@@@ 7\n",
      "\n",
      " ENST00000603077.1 1\n",
      "-\n",
      "TSS 20008432\n",
      "first donor_splice_site 20008402\n",
      "##$$$ chr15:20008202-20008601 11\n",
      "\n",
      " ENST00000603326.1 1\n",
      "-\n",
      "TSS 20004815\n",
      "first donor_splice_site 20004797\n",
      "##$$$ chr15:20004597-20004996 12\n",
      "\n",
      " ENST00000604642.1 1\n",
      "-\n",
      "TSS 20003862\n",
      "first donor_splice_site 20003840\n",
      "##$$$ chr15:20003640-20004039 13\n",
      "\n",
      " ENST00000604950.1 1\n",
      "-\n",
      "TSS 20005935\n",
      "first donor_splice_site 20005905\n",
      "##$$$ chr15:20005705-20006104 14\n",
      "\n",
      " ENST00000622901.1 1\n",
      "-\n",
      "TSS 19955602\n",
      "first donor_splice_site 19955300\n",
      "##$$$ chr15:19955100-19955499 15\n",
      "chr16\n",
      "(67511, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr16\n",
      "\n",
      " ENST00000430178.2 5\n",
      "+\n",
      "TSS 11861\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr16.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr16:11709-12108 0\n",
      "first donor_splice_site 11908\n",
      "Acceptor splice site 12294\n",
      "@@@@ 0\n",
      "##$$$ chr16:12095-12494\n",
      "### 0\n",
      "Donor splice site 12378\n",
      "##$$$ chr16:12179-12578 1\n",
      "Acceptor splice site 12663\n",
      "@@@@ 1\n",
      "##$$$ chr16:12464-12863\n",
      "### 1\n",
      "Donor splice site 12733\n",
      "##$$$ chr16:12534-12933 2\n",
      "Acceptor splice site 12906\n",
      "@@@@ 2\n",
      "##$$$ chr16:12707-13106\n",
      "### 2\n",
      "Donor splice site 13076\n",
      "##$$$ chr16:12877-13276 3\n",
      "Last acceptor_splice_site 13156\n",
      "@@@@ 3\n",
      "##$$$ chr16:12957-13356\n",
      "### 3\n",
      "\n",
      " ENST00000513886.1 3\n",
      "+\n",
      "TSS 11555\n",
      "##$$$ chr16:11709-12108 4\n",
      "first donor_splice_site 11908\n",
      "Acceptor splice site 12294\n",
      "@@@@ 4\n",
      "##$$$ chr16:12095-12494\n",
      "### 4\n",
      "Donor splice site 12402\n",
      "##$$$ chr16:12203-12602 5\n",
      "Last acceptor_splice_site 12902\n",
      "@@@@ 5\n",
      "##$$$ chr16:12703-13102\n",
      "### 5\n",
      "\n",
      " ENST00000564273.4 9\n",
      "-\n",
      "TSS 18068\n",
      "first donor_splice_site 17957\n",
      "##$$$ chr16:17757-18156 6\n",
      "Acceptor splice site 17750\n",
      "@@@@ 6\n",
      "Donor splice site 17604\n",
      "##$$$ chr16:17404-17803 7\n",
      "Acceptor splice site 17427\n",
      "@@@@ 7\n",
      "Donor splice site 17291\n",
      "##$$$ chr16:17091-17490 8\n",
      "Acceptor splice site 17051\n",
      "@@@@ 8\n",
      "Donor splice site 16916\n",
      "##$$$ chr16:16716-17115 9\n",
      "Acceptor splice site 16738\n",
      "@@@@ 9\n",
      "Donor splice site 16537\n",
      "##$$$ chr16:16337-16736 10\n",
      "Acceptor splice site 16448\n",
      "@@@@ 10\n",
      "Donor splice site 16290\n",
      "##$$$ chr16:16090-16489 11\n",
      "Acceptor splice site 15633\n",
      "@@@@ 11\n",
      "Donor splice site 15481\n",
      "##$$$ chr16:15281-15680 12\n",
      "Acceptor splice site 14720\n",
      "@@@@ 12\n",
      "Donor splice site 14652\n",
      "##$$$ chr16:14452-14851 13\n",
      "Last acceptor_splice_site 14511\n",
      "@@@@ 13\n",
      "\n",
      " ENST00000568710.1 2\n",
      "-\n",
      "TSS 35195\n",
      "first donor_splice_site 34847\n",
      "##$$$ chr16:34647-35046 14\n",
      "Last acceptor_splice_site 31883\n",
      "@@@@ 14\n",
      "\n",
      " ENST00000615957.1 1\n",
      "-\n",
      "TSS 17119\n",
      "first donor_splice_site 17052\n",
      "##$$$ chr16:16852-17251 15\n",
      "chr17\n",
      "(87010, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr17\n",
      "\n",
      " ENST00000595228.4 3\n",
      "+\n",
      "TSS 137569\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr17.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr17:137446-137845 0\n",
      "first donor_splice_site 137645\n",
      "Acceptor splice site 138203\n",
      "@@@@ 0\n",
      "##$$$ chr17:138004-138403\n",
      "### 0\n",
      "Donor splice site 138402\n",
      "##$$$ chr17:138203-138602 1\n",
      "Last acceptor_splice_site 138910\n",
      "@@@@ 1\n",
      "##$$$ chr17:138711-139110\n",
      "### 1\n",
      "\n",
      " ENST00000613549.3 3\n",
      "-\n",
      "TSS 181650\n",
      "first donor_splice_site 181107\n",
      "##$$$ chr17:180907-181306 2\n",
      "Acceptor splice site 172616\n",
      "@@@@ 2\n",
      "Donor splice site 172537\n",
      "##$$$ chr17:172337-172736 3\n",
      "Last acceptor_splice_site 164204\n",
      "@@@@ 3\n",
      "\n",
      " ENST00000620287.1 1\n",
      "-\n",
      "TSS 90899\n",
      "first donor_splice_site 90796\n",
      "##$$$ chr17:90596-90995 4\n",
      "\n",
      " ENST00000623180.1 5\n",
      "-\n",
      "TSS 76866\n",
      "first donor_splice_site 76723\n",
      "##$$$ chr17:76523-76922 5\n",
      "Acceptor splice site 75878\n",
      "@@@@ 4\n",
      "Donor splice site 75814\n",
      "##$$$ chr17:75614-76013 6\n",
      "Acceptor splice site 71556\n",
      "@@@@ 5\n",
      "Donor splice site 71366\n",
      "##$$$ chr17:71166-71565 7\n",
      "Acceptor splice site 65887\n",
      "@@@@ 6\n",
      "Donor splice site 65830\n",
      "##$$$ chr17:65630-66029 8\n",
      "Last acceptor_splice_site 65736\n",
      "@@@@ 7\n",
      "\n",
      " ENST00000624936.2 8\n",
      "-\n",
      "TSS 118578\n",
      "first donor_splice_site 118383\n",
      "##$$$ chr17:118183-118582 9\n",
      "Acceptor splice site 106864\n",
      "@@@@ 8\n",
      "Donor splice site 106755\n",
      "##$$$ chr17:106555-106954 10\n",
      "Acceptor splice site 105691\n",
      "@@@@ 9\n",
      "Donor splice site 105560\n",
      "##$$$ chr17:105360-105759 11\n",
      "Acceptor splice site 102273\n",
      "@@@@ 10\n",
      "Donor splice site 102058\n",
      "##$$$ chr17:101858-102257 12\n",
      "Acceptor splice site 101174\n",
      "@@@@ 11\n",
      "Donor splice site 101109\n",
      "##$$$ chr17:100909-101308 13\n",
      "Acceptor splice site 98951\n",
      "@@@@ 12\n",
      "Donor splice site 98893\n",
      "##$$$ chr17:98693-99092 14\n",
      "Acceptor splice site 97797\n",
      "@@@@ 13\n",
      "Donor splice site 97269\n",
      "##$$$ chr17:97069-97468 15\n",
      "Last acceptor_splice_site 88705\n",
      "@@@@ 14\n",
      "chr18\n",
      "(26537, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr18\n",
      "\n",
      " ENST00000308911.8 4\n",
      "-\n",
      "TSS 49557\n",
      "first donor_splice_site 49501\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr18.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr18:49301-49700 0\n",
      "Acceptor splice site 49237\n",
      "@@@@ 0\n",
      "Donor splice site 49129\n",
      "##$$$ chr18:48929-49328 1\n",
      "Acceptor splice site 49050\n",
      "@@@@ 1\n",
      "Donor splice site 48940\n",
      "##$$$ chr18:48740-49139 2\n",
      "Last acceptor_splice_site 48447\n",
      "@@@@ 2\n",
      "\n",
      " ENST00000572062.1 2\n",
      "-\n",
      "TSS 16898\n",
      "first donor_splice_site 16856\n",
      "##$$$ chr18:16656-17055 3\n",
      "Last acceptor_splice_site 14653\n",
      "@@@@ 3\n",
      "\n",
      " ENST00000572530.1 1\n",
      "-\n",
      "TSS 73517\n",
      "first donor_splice_site 73312\n",
      "##$$$ chr18:73112-73511 4\n",
      "\n",
      " ENST00000572573.2 3\n",
      "+\n",
      "TSS 11125\n",
      "##$$$ chr18:11396-11795 5\n",
      "first donor_splice_site 11595\n",
      "Acceptor splice site 13152\n",
      "@@@@ 4\n",
      "##$$$ chr18:12953-13352\n",
      "### 4\n",
      "Donor splice site 13354\n",
      "##$$$ chr18:13155-13554 6\n",
      "Last acceptor_splice_site 15617\n",
      "@@@@ 5\n",
      "##$$$ chr18:15418-15817\n",
      "### 5\n",
      "\n",
      " ENST00000572608.1 2\n",
      "-\n",
      "TSS 14958\n",
      "first donor_splice_site 14851\n",
      "##$$$ chr18:14651-15050 7\n",
      "Last acceptor_splice_site 14653\n",
      "@@@@ 6\n",
      "\n",
      " ENST00000575066.2 4\n",
      "+\n",
      "TSS 45034\n",
      "##$$$ chr18:45036-45435 8\n",
      "first donor_splice_site 45235\n",
      "Acceptor splice site 45283\n",
      "@@@@ 7\n",
      "##$$$ chr18:45084-45483\n",
      "### 7\n",
      "Donor splice site 45556\n",
      "##$$$ chr18:45357-45756 9\n",
      "Acceptor splice site 45641\n",
      "@@@@ 8\n",
      "##$$$ chr18:45442-45841\n",
      "### 8\n",
      "Donor splice site 45787\n",
      "##$$$ chr18:45588-45987 10\n",
      "Last acceptor_splice_site 45908\n",
      "@@@@ 9\n",
      "##$$$ chr18:45709-46108\n",
      "### 9\n",
      "\n",
      " ENST00000575820.1 2\n",
      "+\n",
      "TSS 11103\n",
      "##$$$ chr18:11396-11795 11\n",
      "first donor_splice_site 11595\n",
      "Last acceptor_splice_site 15617\n",
      "@@@@ 10\n",
      "##$$$ chr18:15418-15817\n",
      "### 10\n",
      "\n",
      " ENST00000594555.1 2\n",
      "-\n",
      "TSS 48882\n",
      "first donor_splice_site 48753\n",
      "##$$$ chr18:48553-48952 12\n",
      "Last acceptor_splice_site 48615\n",
      "@@@@ 11\n",
      "chr19\n",
      "(80480, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr19\n",
      "\n",
      " ENST00000631796.1 2\n",
      "-\n",
      "TSS 70963\n",
      "first donor_splice_site 70928\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr19.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr19:70728-71127 0\n",
      "Last acceptor_splice_site 66499\n",
      "@@@@ 0\n",
      "\n",
      " ENST00000631994.1 2\n",
      "-\n",
      "TSS 71566\n",
      "first donor_splice_site 71141\n",
      "##$$$ chr19:70941-71340 1\n",
      "Last acceptor_splice_site 66499\n",
      "@@@@ 1\n",
      "\n",
      " ENST00000632089.1 3\n",
      "-\n",
      "TSS 70945\n",
      "first donor_splice_site 70928\n",
      "##$$$ chr19:70728-71127 2\n",
      "Acceptor splice site 66499\n",
      "@@@@ 2\n",
      "Donor splice site 66346\n",
      "##$$$ chr19:66146-66545 3\n",
      "Last acceptor_splice_site 66047\n",
      "@@@@ 3\n",
      "\n",
      " ENST00000632292.1 1\n",
      "-\n",
      "TSS 71626\n",
      "first donor_splice_site 70652\n",
      "##$$$ chr19:70452-70851 4\n",
      "\n",
      " ENST00000632496.1 2\n",
      "-\n",
      "TSS 66420\n",
      "first donor_splice_site 66346\n",
      "##$$$ chr19:66146-66545 5\n",
      "Last acceptor_splice_site 66133\n",
      "@@@@ 4\n",
      "\n",
      " ENST00000632506.1 3\n",
      "-\n",
      "TSS 70976\n",
      "first donor_splice_site 70928\n",
      "##$$$ chr19:70728-71127 6\n",
      "Acceptor splice site 66499\n",
      "@@@@ 5\n",
      "Donor splice site 66346\n",
      "##$$$ chr19:66146-66545 7\n",
      "Last acceptor_splice_site 61894\n",
      "@@@@ 6\n",
      "\n",
      " ENST00000633703.1 3\n",
      "-\n",
      "TSS 70951\n",
      "first donor_splice_site 70928\n",
      "##$$$ chr19:70728-71127 8\n",
      "Acceptor splice site 66499\n",
      "@@@@ 7\n",
      "Donor splice site 66346\n",
      "##$$$ chr19:66146-66545 9\n",
      "Last acceptor_splice_site 64213\n",
      "@@@@ 8\n",
      "\n",
      " ENST00000633719.1 1\n",
      "-\n",
      "TSS 66524\n",
      "first donor_splice_site 62113\n",
      "##$$$ chr19:61913-62312 10\n",
      "\n",
      " ENST00000633742.1 1\n",
      "-\n",
      "TSS 66492\n",
      "first donor_splice_site 66320\n",
      "##$$$ chr19:66120-66519 11\n",
      "\n",
      " ENST00000634023.1 2\n",
      "-\n",
      "TSS 66382\n",
      "first donor_splice_site 66346\n",
      "##$$$ chr19:66146-66545 12\n",
      "Last acceptor_splice_site 65226\n",
      "@@@@ 9\n",
      "chr2\n",
      "(117232, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr2\n",
      "\n",
      " ENST00000327669.5 2\n",
      "-\n",
      "TSS 46505\n",
      "first donor_splice_site 45440\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr2.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr2:45240-45639 0\n",
      "Last acceptor_splice_site 41627\n",
      "@@@@ 0\n",
      "\n",
      " ENST00000403712.6 2\n",
      "-\n",
      "TSS 264032\n",
      "first donor_splice_site 263984\n",
      "##$$$ chr2:263784-264183 1\n",
      "Last acceptor_splice_site 253115\n",
      "@@@@ 1\n",
      "\n",
      " ENST00000437798.1 2\n",
      "+\n",
      "TSS 197569\n",
      "##$$$ chr2:199702-200101 2\n",
      "first donor_splice_site 199901\n",
      "Last acceptor_splice_site 200164\n",
      "@@@@ 2\n",
      "##$$$ chr2:199965-200364\n",
      "### 2\n",
      "\n",
      " ENST00000460464.1 2\n",
      "-\n",
      "TSS 42952\n",
      "first donor_splice_site 42809\n",
      "##$$$ chr2:42609-43008 3\n",
      "Last acceptor_splice_site 41627\n",
      "@@@@ 3\n",
      "\n",
      " ENST00000460974.1 2\n",
      "-\n",
      "TSS 218866\n",
      "first donor_splice_site 218746\n",
      "##$$$ chr2:218546-218945 4\n",
      "Last acceptor_splice_site 218108\n",
      "@@@@ 4\n",
      "\n",
      " ENST00000461026.1 2\n",
      "-\n",
      "TSS 46870\n",
      "first donor_splice_site 46807\n",
      "##$$$ chr2:46607-47006 5\n",
      "Last acceptor_splice_site 41627\n",
      "@@@@ 5\n",
      "\n",
      " ENST00000653001.1 3\n",
      "+\n",
      "TSS 198489\n",
      "##$$$ chr2:198416-198815 6\n",
      "first donor_splice_site 198615\n",
      "Acceptor splice site 200164\n",
      "@@@@ 6\n",
      "##$$$ chr2:199965-200364\n",
      "### 6\n",
      "Donor splice site 200230\n",
      "##$$$ chr2:200031-200430 7\n",
      "Last acceptor_splice_site 209485\n",
      "@@@@ 7\n",
      "##$$$ chr2:209286-209685\n",
      "### 7\n",
      "\n",
      " ENST00000665534.1 5\n",
      "+\n",
      "TSS 198506\n",
      "##$$$ chr2:198416-198815 8\n",
      "first donor_splice_site 198615\n",
      "Acceptor splice site 199744\n",
      "@@@@ 8\n",
      "##$$$ chr2:199545-199944\n",
      "### 8\n",
      "Donor splice site 199901\n",
      "##$$$ chr2:199702-200101 9\n",
      "Acceptor splice site 200164\n",
      "@@@@ 9\n",
      "##$$$ chr2:199965-200364\n",
      "### 9\n",
      "Donor splice site 200230\n",
      "##$$$ chr2:200031-200430 10\n",
      "Acceptor splice site 202916\n",
      "@@@@ 10\n",
      "##$$$ chr2:202717-203116\n",
      "### 10\n",
      "Donor splice site 202988\n",
      "##$$$ chr2:202789-203188 11\n",
      "Last acceptor_splice_site 209485\n",
      "@@@@ 11\n",
      "##$$$ chr2:209286-209685\n",
      "### 11\n",
      "chr20\n",
      "(34275, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr20\n",
      "\n",
      " ENST00000246105.4 2\n",
      "+\n",
      "TSS 227258\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr20.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr20:227147-227546 0\n",
      "first donor_splice_site 227346\n",
      "Last acceptor_splice_site 229278\n",
      "@@@@ 0\n",
      "##$$$ chr20:229079-229478\n",
      "### 0\n",
      "\n",
      " ENST00000334391.5 2\n",
      "-\n",
      "TSS 189711\n",
      "first donor_splice_site 189575\n",
      "##$$$ chr20:189375-189774 1\n",
      "Last acceptor_splice_site 188118\n",
      "@@@@ 1\n",
      "\n",
      " ENST00000382369.9 2\n",
      "-\n",
      "TSS 290429\n",
      "first donor_splice_site 290259\n",
      "##$$$ chr20:290059-290458 2\n",
      "Last acceptor_splice_site 284081\n",
      "@@@@ 2\n",
      "\n",
      " ENST00000382376.4 2\n",
      "+\n",
      "TSS 257724\n",
      "##$$$ chr20:257637-258036 3\n",
      "first donor_splice_site 257836\n",
      "Last acceptor_splice_site 259077\n",
      "@@@@ 3\n",
      "##$$$ chr20:258878-259277\n",
      "### 3\n",
      "\n",
      " ENST00000382388.4 2\n",
      "+\n",
      "TSS 157454\n",
      "##$$$ chr20:157394-157793 4\n",
      "first donor_splice_site 157593\n",
      "Last acceptor_splice_site 158774\n",
      "@@@@ 4\n",
      "##$$$ chr20:158575-158974\n",
      "### 4\n",
      "\n",
      " ENST00000382398.4 2\n",
      "+\n",
      "TSS 142590\n",
      "##$$$ chr20:142487-142886 5\n",
      "first donor_splice_site 142686\n",
      "Last acceptor_splice_site 145415\n",
      "@@@@ 5\n",
      "##$$$ chr20:145216-145615\n",
      "### 5\n",
      "\n",
      " ENST00000382410.3 2\n",
      "+\n",
      "TSS 87672\n",
      "##$$$ chr20:87568-87967 6\n",
      "first donor_splice_site 87767\n",
      "Last acceptor_splice_site 96005\n",
      "@@@@ 6\n",
      "##$$$ chr20:95806-96205\n",
      "### 6\n",
      "\n",
      " ENST00000542572.1 3\n",
      "+\n",
      "TSS 142634\n",
      "##$$$ chr20:142487-142886 7\n",
      "first donor_splice_site 142686\n",
      "Acceptor splice site 145415\n",
      "@@@@ 7\n",
      "##$$$ chr20:145216-145615\n",
      "### 7\n",
      "Donor splice site 145488\n",
      "##$$$ chr20:145289-145688 8\n",
      "Last acceptor_splice_site 145579\n",
      "@@@@ 8\n",
      "##$$$ chr20:145380-145779\n",
      "### 8\n",
      "\n",
      " ENST00000608495.1 1\n",
      "+\n",
      "TSS 267186\n",
      "##$$$ chr20:268658-269057 9\n",
      "first donor_splice_site 268857\n",
      "\n",
      " ENST00000608838.1 2\n",
      "+\n",
      "TSS 87250\n",
      "##$$$ chr20:87160-87559 10\n",
      "first donor_splice_site 87359\n",
      "Last acceptor_splice_site 96005\n",
      "@@@@ 9\n",
      "##$$$ chr20:95806-96205\n",
      "### 9\n",
      "chr21\n",
      "(17971, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr21\n",
      "\n",
      " ENST00000612610.4 7\n",
      "+\n",
      "TSS 5022493\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr21.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr21:5022494-5022893 0\n",
      "first donor_splice_site 5022693\n",
      "Acceptor splice site 5025009\n",
      "@@@@ 0\n",
      "##$$$ chr21:5024810-5025209\n",
      "### 0\n",
      "Donor splice site 5025049\n",
      "##$$$ chr21:5024850-5025249 1\n",
      "Acceptor splice site 5026280\n",
      "@@@@ 1\n",
      "##$$$ chr21:5026081-5026480\n",
      "### 1\n",
      "Donor splice site 5026630\n",
      "##$$$ chr21:5026431-5026830 2\n",
      "Acceptor splice site 5027935\n",
      "@@@@ 2\n",
      "##$$$ chr21:5027736-5028135\n",
      "### 2\n",
      "Donor splice site 5028225\n",
      "##$$$ chr21:5028026-5028425 3\n",
      "Acceptor splice site 5032053\n",
      "@@@@ 3\n",
      "##$$$ chr21:5031854-5032253\n",
      "### 3\n",
      "Donor splice site 5032217\n",
      "##$$$ chr21:5032018-5032417 4\n",
      "Acceptor splice site 5033408\n",
      "@@@@ 4\n",
      "##$$$ chr21:5033209-5033608\n",
      "### 4\n",
      "Donor splice site 5033443\n",
      "##$$$ chr21:5033244-5033643 5\n",
      "Last acceptor_splice_site 5040585\n",
      "@@@@ 5\n",
      "##$$$ chr21:5040386-5040785\n",
      "### 5\n",
      "\n",
      " ENST00000620481.4 6\n",
      "+\n",
      "TSS 5022493\n",
      "##$$$ chr21:5022494-5022893 6\n",
      "first donor_splice_site 5022693\n",
      "Acceptor splice site 5025009\n",
      "@@@@ 6\n",
      "##$$$ chr21:5024810-5025209\n",
      "### 6\n",
      "Donor splice site 5025049\n",
      "##$$$ chr21:5024850-5025249 7\n",
      "Acceptor splice site 5027935\n",
      "@@@@ 7\n",
      "##$$$ chr21:5027736-5028135\n",
      "### 7\n",
      "Donor splice site 5028225\n",
      "##$$$ chr21:5028026-5028425 8\n",
      "Acceptor splice site 5032053\n",
      "@@@@ 8\n",
      "##$$$ chr21:5031854-5032253\n",
      "### 8\n",
      "Donor splice site 5032217\n",
      "##$$$ chr21:5032018-5032417 9\n",
      "Acceptor splice site 5033408\n",
      "@@@@ 9\n",
      "##$$$ chr21:5033209-5033608\n",
      "### 9\n",
      "Donor splice site 5033443\n",
      "##$$$ chr21:5033244-5033643 10\n",
      "Last acceptor_splice_site 5040585\n",
      "@@@@ 10\n",
      "##$$$ chr21:5040386-5040785\n",
      "### 10\n",
      "\n",
      " ENST00000623903.3 3\n",
      "+\n",
      "TSS 5022531\n",
      "##$$$ chr21:5022494-5022893 11\n",
      "first donor_splice_site 5022693\n",
      "Acceptor splice site 5025009\n",
      "@@@@ 11\n",
      "##$$$ chr21:5024810-5025209\n",
      "### 11\n",
      "Donor splice site 5025049\n",
      "##$$$ chr21:5024850-5025249 12\n",
      "Last acceptor_splice_site 5026432\n",
      "@@@@ 12\n",
      "##$$$ chr21:5026233-5026632\n",
      "### 12\n",
      "\n",
      " ENST00000624081.1 4\n",
      "+\n",
      "TSS 5011799\n",
      "##$$$ chr21:5011675-5012074 13\n",
      "first donor_splice_site 5011874\n",
      "Acceptor splice site 5012548\n",
      "@@@@ 13\n",
      "##$$$ chr21:5012349-5012748\n",
      "### 13\n",
      "Donor splice site 5012687\n",
      "##$$$ chr21:5012488-5012887 14\n",
      "Acceptor splice site 5014386\n",
      "@@@@ 14\n",
      "##$$$ chr21:5014187-5014586\n",
      "### 14\n",
      "Donor splice site 5014471\n",
      "##$$$ chr21:5014272-5014671 15\n",
      "Last acceptor_splice_site 5016935\n",
      "@@@@ 15\n",
      "##$$$ chr21:5016736-5017135\n",
      "### 15\n",
      "chr22\n",
      "(31502, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr22\n",
      "\n",
      " ENST00000422332.1 2\n",
      "+\n",
      "TSS 11124337\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr22.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr22:11124180-11124579 0\n",
      "first donor_splice_site 11124379\n",
      "Last acceptor_splice_site 11124508\n",
      "@@@@ 0\n",
      "##$$$ chr22:11124309-11124708\n",
      "### 0\n",
      "\n",
      " ENST00000612732.1 1\n",
      "-\n",
      "TSS 11249959\n",
      "first donor_splice_site 11249809\n",
      "##$$$ chr22:11249609-11250008 1\n",
      "\n",
      " ENST00000615943.1 1\n",
      "-\n",
      "TSS 10736283\n",
      "first donor_splice_site 10736171\n",
      "##$$$ chr22:10735971-10736370 2\n",
      "\n",
      " ENST00000624155.2 2\n",
      "+\n",
      "TSS 11066418\n",
      "##$$$ chr22:11066316-11066715 3\n",
      "first donor_splice_site 11066515\n",
      "Last acceptor_splice_site 11067985\n",
      "@@@@ 1\n",
      "##$$$ chr22:11067786-11068185\n",
      "### 1\n",
      "\n",
      " ENST00000635667.1 9\n",
      "-\n",
      "TSS 10961338\n",
      "first donor_splice_site 10961283\n",
      "##$$$ chr22:10961083-10961482 4\n",
      "Acceptor splice site 10959136\n",
      "@@@@ 2\n",
      "Donor splice site 10959067\n",
      "##$$$ chr22:10958867-10959266 5\n",
      "Acceptor splice site 10950174\n",
      "@@@@ 3\n",
      "Donor splice site 10950049\n",
      "##$$$ chr22:10949849-10950248 6\n",
      "Acceptor splice site 10949269\n",
      "@@@@ 4\n",
      "Donor splice site 10949212\n",
      "##$$$ chr22:10949012-10949411 7\n",
      "Acceptor splice site 10947418\n",
      "@@@@ 5\n",
      "Donor splice site 10947304\n",
      "##$$$ chr22:10947104-10947503 8\n",
      "Acceptor splice site 10945053\n",
      "@@@@ 6\n",
      "Donor splice site 10944967\n",
      "##$$$ chr22:10944767-10945166 9\n",
      "Acceptor splice site 10941780\n",
      "@@@@ 7\n",
      "Donor splice site 10941691\n",
      "##$$$ chr22:10941491-10941890 10\n",
      "Acceptor splice site 10940707\n",
      "@@@@ 8\n",
      "Donor splice site 10940597\n",
      "##$$$ chr22:10940397-10940796 11\n",
      "Last acceptor_splice_site 10939423\n",
      "@@@@ 9\n",
      "\n",
      " ENST00000657566.1 2\n",
      "+\n",
      "TSS 11827523\n",
      "##$$$ chr22:11827459-11827858 12\n",
      "first donor_splice_site 11827658\n",
      "Last acceptor_splice_site 11832761\n",
      "@@@@ 10\n",
      "##$$$ chr22:11832562-11832961\n",
      "### 10\n",
      "\n",
      " ENST00000672261.1 3\n",
      "+\n",
      "TSS 11474744\n",
      "##$$$ chr22:11474705-11475104 13\n",
      "first donor_splice_site 11474904\n",
      "Acceptor splice site 11475146\n",
      "@@@@ 11\n",
      "##$$$ chr22:11474947-11475346\n",
      "### 11\n",
      "Donor splice site 11475238\n",
      "##$$$ chr22:11475039-11475438 14\n",
      "Last acceptor_splice_site 11477178\n",
      "@@@@ 12\n",
      "##$$$ chr22:11476979-11477378\n",
      "### 12\n",
      "chr3\n",
      "(96941, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr3\n",
      "\n",
      " ENST00000440867.1 2\n",
      "+\n",
      "TSS 23757\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr3.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr3:23613-24012 0\n",
      "first donor_splice_site 23812\n",
      "Last acceptor_splice_site 23968\n",
      "@@@@ 0\n",
      "##$$$ chr3:23769-24168\n",
      "### 0\n",
      "\n",
      " ENST00000656877.1 4\n",
      "+\n",
      "TSS 11753\n",
      "##$$$ chr3:11600-11999 1\n",
      "first donor_splice_site 11799\n",
      "Acceptor splice site 20557\n",
      "@@@@ 1\n",
      "##$$$ chr3:20358-20757\n",
      "### 1\n",
      "Donor splice site 20597\n",
      "##$$$ chr3:20398-20797 2\n",
      "Acceptor splice site 23761\n",
      "@@@@ 2\n",
      "##$$$ chr3:23562-23961\n",
      "### 2\n",
      "Donor splice site 23812\n",
      "##$$$ chr3:23613-24012 3\n",
      "Last acceptor_splice_site 23968\n",
      "@@@@ 3\n",
      "##$$$ chr3:23769-24168\n",
      "### 3\n",
      "\n",
      " ENST00000659825.1 3\n",
      "+\n",
      "TSS 11769\n",
      "##$$$ chr3:11600-11999 4\n",
      "first donor_splice_site 11799\n",
      "Acceptor splice site 23761\n",
      "@@@@ 4\n",
      "##$$$ chr3:23562-23961\n",
      "### 4\n",
      "Donor splice site 23812\n",
      "##$$$ chr3:23613-24012 5\n",
      "Last acceptor_splice_site 23968\n",
      "@@@@ 5\n",
      "##$$$ chr3:23769-24168\n",
      "### 5\n",
      "\n",
      " ENST00000660204.1 4\n",
      "+\n",
      "TSS 11768\n",
      "##$$$ chr3:12460-12859 6\n",
      "first donor_splice_site 12659\n",
      "Acceptor splice site 14771\n",
      "@@@@ 6\n",
      "##$$$ chr3:14572-14971\n",
      "### 6\n",
      "Donor splice site 14831\n",
      "##$$$ chr3:14632-15031 7\n",
      "Acceptor splice site 23761\n",
      "@@@@ 7\n",
      "##$$$ chr3:23562-23961\n",
      "### 7\n",
      "Donor splice site 23812\n",
      "##$$$ chr3:23613-24012 8\n",
      "Last acceptor_splice_site 23968\n",
      "@@@@ 8\n",
      "##$$$ chr3:23769-24168\n",
      "### 8\n",
      "\n",
      " ENST00000660491.1 3\n",
      "+\n",
      "TSS 11768\n",
      "##$$$ chr3:12460-12859 9\n",
      "first donor_splice_site 12659\n",
      "Acceptor splice site 20891\n",
      "@@@@ 9\n",
      "##$$$ chr3:20692-21091\n",
      "### 9\n",
      "Donor splice site 20953\n",
      "##$$$ chr3:20754-21153 10\n",
      "Last acceptor_splice_site 23968\n",
      "@@@@ 10\n",
      "##$$$ chr3:23769-24168\n",
      "### 10\n",
      "\n",
      " ENST00000661675.1 4\n",
      "+\n",
      "TSS 11745\n",
      "##$$$ chr3:11600-11999 11\n",
      "first donor_splice_site 11799\n",
      "Acceptor splice site 14771\n",
      "@@@@ 11\n",
      "##$$$ chr3:14572-14971\n",
      "### 11\n",
      "Donor splice site 14831\n",
      "##$$$ chr3:14632-15031 12\n",
      "Acceptor splice site 23761\n",
      "@@@@ 12\n",
      "##$$$ chr3:23562-23961\n",
      "### 12\n",
      "Donor splice site 23812\n",
      "##$$$ chr3:23613-24012 13\n",
      "Last acceptor_splice_site 23968\n",
      "@@@@ 13\n",
      "##$$$ chr3:23769-24168\n",
      "### 13\n",
      "chr4\n",
      "(62342, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr4\n",
      "\n",
      " ENST00000502981.2 3\n",
      "+\n",
      "TSS 53329\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr4.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr4:53292-53691 0\n",
      "first donor_splice_site 53491\n",
      "Acceptor splice site 59430\n",
      "@@@@ 0\n",
      "##$$$ chr4:59231-59630\n",
      "### 0\n",
      "Donor splice site 59556\n",
      "##$$$ chr4:59357-59756 1\n",
      "Last acceptor_splice_site 60058\n",
      "@@@@ 1\n",
      "##$$$ chr4:59859-60258\n",
      "### 1\n",
      "\n",
      " ENST00000503774.1 1\n",
      "-\n",
      "TSS 50124\n",
      "first donor_splice_site 49554\n",
      "##$$$ chr4:49354-49753 2\n",
      "\n",
      " ENST00000509152.3 4\n",
      "+\n",
      "TSS 53304\n",
      "##$$$ chr4:53292-53691 3\n",
      "first donor_splice_site 53491\n",
      "Acceptor splice site 59430\n",
      "@@@@ 2\n",
      "##$$$ chr4:59231-59630\n",
      "### 2\n",
      "Donor splice site 59556\n",
      "##$$$ chr4:59357-59756 4\n",
      "Acceptor splice site 60058\n",
      "@@@@ 3\n",
      "##$$$ chr4:59859-60258\n",
      "### 3\n",
      "Donor splice site 60153\n",
      "##$$$ chr4:59954-60353 5\n",
      "Last acceptor_splice_site 67840\n",
      "@@@@ 4\n",
      "##$$$ chr4:67641-68040\n",
      "### 4\n",
      "\n",
      " ENST00000510175.6 1\n",
      "+\n",
      "TSS 124501\n",
      "##$$$ chr4:124474-124873 6\n",
      "first donor_splice_site 124673\n",
      "\n",
      " ENST00000511645.1 1\n",
      "-\n",
      "TSS 107114\n",
      "first donor_splice_site 105959\n",
      "##$$$ chr4:105759-106158 7\n",
      "\n",
      " ENST00000608255.2 2\n",
      "+\n",
      "TSS 53286\n",
      "##$$$ chr4:53292-53691 8\n",
      "first donor_splice_site 53491\n",
      "Last acceptor_splice_site 85731\n",
      "@@@@ 5\n",
      "##$$$ chr4:85532-85931\n",
      "### 5\n",
      "\n",
      " ENST00000609518.5 3\n",
      "+\n",
      "TSS 53295\n",
      "##$$$ chr4:53292-53691 9\n",
      "first donor_splice_site 53491\n",
      "Acceptor splice site 59430\n",
      "@@@@ 6\n",
      "##$$$ chr4:59231-59630\n",
      "### 6\n",
      "Donor splice site 59556\n",
      "##$$$ chr4:59357-59756 10\n",
      "Last acceptor_splice_site 85731\n",
      "@@@@ 7\n",
      "##$$$ chr4:85532-85931\n",
      "### 7\n",
      "\n",
      " ENST00000610261.6 4\n",
      "+\n",
      "TSS 53321\n",
      "##$$$ chr4:53292-53691 11\n",
      "first donor_splice_site 53491\n",
      "Acceptor splice site 59430\n",
      "@@@@ 8\n",
      "##$$$ chr4:59231-59630\n",
      "### 8\n",
      "Donor splice site 59556\n",
      "##$$$ chr4:59357-59756 12\n",
      "Acceptor splice site 60058\n",
      "@@@@ 9\n",
      "##$$$ chr4:59859-60258\n",
      "### 9\n",
      "Donor splice site 60153\n",
      "##$$$ chr4:59954-60353 13\n",
      "Last acceptor_splice_site 85731\n",
      "@@@@ 10\n",
      "##$$$ chr4:85532-85931\n",
      "### 10\n",
      "\n",
      " ENST00000616656.1 1\n",
      "-\n",
      "TSS 49956\n",
      "first donor_splice_site 49096\n",
      "##$$$ chr4:48896-49295 14\n",
      "chr5\n",
      "(68092, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr5\n",
      "\n",
      " ENST00000506872.1 1\n",
      "+\n",
      "TSS 58198\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr5.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr5:58716-59115 0\n",
      "first donor_splice_site 58915\n",
      "\n",
      " ENST00000637938.1 19\n",
      "+\n",
      "TSS 92151\n",
      "##$$$ chr5:92077-92476 1\n",
      "first donor_splice_site 92276\n",
      "Acceptor splice site 113251\n",
      "@@@@ 0\n",
      "##$$$ chr5:113052-113451\n",
      "### 0\n",
      "Donor splice site 113448\n",
      "##$$$ chr5:113249-113648 2\n",
      "Acceptor splice site 139483\n",
      "@@@@ 1\n",
      "##$$$ chr5:139284-139683\n",
      "### 1\n",
      "Donor splice site 140716\n",
      "##$$$ chr5:140517-140916 3\n",
      "Acceptor splice site 143047\n",
      "@@@@ 2\n",
      "##$$$ chr5:142848-143247\n",
      "### 2\n",
      "Donor splice site 143256\n",
      "##$$$ chr5:143057-143456 4\n",
      "Acceptor splice site 143380\n",
      "@@@@ 3\n",
      "##$$$ chr5:143181-143580\n",
      "### 3\n",
      "Donor splice site 143503\n",
      "##$$$ chr5:143304-143703 5\n",
      "Acceptor splice site 144827\n",
      "@@@@ 4\n",
      "##$$$ chr5:144628-145027\n",
      "### 4\n",
      "Donor splice site 144920\n",
      "##$$$ chr5:144721-145120 6\n",
      "Acceptor splice site 151513\n",
      "@@@@ 5\n",
      "##$$$ chr5:151314-151713\n",
      "### 5\n",
      "Donor splice site 151599\n",
      "##$$$ chr5:151400-151799 7\n",
      "Acceptor splice site 154875\n",
      "@@@@ 6\n",
      "##$$$ chr5:154676-155075\n",
      "### 6\n",
      "Donor splice site 154991\n",
      "##$$$ chr5:154792-155191 8\n",
      "Acceptor splice site 155345\n",
      "@@@@ 7\n",
      "##$$$ chr5:155146-155545\n",
      "### 7\n",
      "Donor splice site 155443\n",
      "##$$$ chr5:155244-155643 9\n",
      "Acceptor splice site 156071\n",
      "@@@@ 8\n",
      "##$$$ chr5:155872-156271\n",
      "### 8\n",
      "Donor splice site 156210\n",
      "##$$$ chr5:156011-156410 10\n",
      "Acceptor splice site 156773\n",
      "@@@@ 9\n",
      "##$$$ chr5:156574-156973\n",
      "### 9\n",
      "Donor splice site 156911\n",
      "##$$$ chr5:156712-157111 11\n",
      "Acceptor splice site 161783\n",
      "@@@@ 10\n",
      "##$$$ chr5:161584-161983\n",
      "### 10\n",
      "Donor splice site 161944\n",
      "##$$$ chr5:161745-162144 12\n",
      "Acceptor splice site 162722\n",
      "@@@@ 11\n",
      "##$$$ chr5:162523-162922\n",
      "### 11\n",
      "Donor splice site 163548\n",
      "##$$$ chr5:163349-163748 13\n",
      "Acceptor splice site 169340\n",
      "@@@@ 12\n",
      "##$$$ chr5:169141-169540\n",
      "### 12\n",
      "Donor splice site 169592\n",
      "##$$$ chr5:169393-169792 14\n",
      "Acceptor splice site 171043\n",
      "@@@@ 13\n",
      "##$$$ chr5:170844-171243\n",
      "### 13\n",
      "Donor splice site 171132\n",
      "##$$$ chr5:170933-171332 15\n",
      "Acceptor splice site 171214\n",
      "@@@@ 14\n",
      "##$$$ chr5:171015-171414\n",
      "### 14\n",
      "Donor splice site 171444\n",
      "##$$$ chr5:171245-171644 16\n",
      "Acceptor splice site 172897\n",
      "@@@@ 15\n",
      "##$$$ chr5:172698-173097\n",
      "### 15\n",
      "Donor splice site 173067\n",
      "##$$$ chr5:172868-173267 17\n",
      "Acceptor splice site 173918\n",
      "@@@@ 16\n",
      "##$$$ chr5:173719-174118\n",
      "### 16\n",
      "Donor splice site 174098\n",
      "##$$$ chr5:173899-174298 18\n",
      "Last acceptor_splice_site 181514\n",
      "@@@@ 17\n",
      "##$$$ chr5:181315-181714\n",
      "### 17\n",
      "chr6\n",
      "(71093, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr6\n",
      "\n",
      " ENST00000405102.1 1\n",
      "-\n",
      "TSS 106856\n",
      "first donor_splice_site 105919\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr6.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr6:105719-106118 0\n",
      "\n",
      " ENST00000406017.3 1\n",
      "-\n",
      "TSS 145083\n",
      "first donor_splice_site 142272\n",
      "##$$$ chr6:142072-142471 1\n",
      "\n",
      " ENST00000407941.1 3\n",
      "-\n",
      "TSS 170631\n",
      "first donor_splice_site 170352\n",
      "##$$$ chr6:170152-170551 2\n",
      "Acceptor splice site 169753\n",
      "@@@@ 0\n",
      "Donor splice site 169545\n",
      "##$$$ chr6:169345-169744 3\n",
      "Last acceptor_splice_site 167826\n",
      "@@@@ 1\n",
      "\n",
      " ENST00000436899.5 3\n",
      "-\n",
      "TSS 144885\n",
      "first donor_splice_site 144536\n",
      "##$$$ chr6:144336-144735 4\n",
      "Acceptor splice site 140379\n",
      "@@@@ 2\n",
      "Donor splice site 140211\n",
      "##$$$ chr6:140011-140410 5\n",
      "Last acceptor_splice_site 132117\n",
      "@@@@ 3\n",
      "\n",
      " ENST00000580119.2 8\n",
      "-\n",
      "TSS 205519\n",
      "first donor_splice_site 204887\n",
      "##$$$ chr6:204687-205086 6\n",
      "Acceptor splice site 198509\n",
      "@@@@ 4\n",
      "Donor splice site 198165\n",
      "##$$$ chr6:197965-198364 7\n",
      "Acceptor splice site 189649\n",
      "@@@@ 5\n",
      "Donor splice site 189476\n",
      "##$$$ chr6:189276-189675 8\n",
      "Acceptor splice site 187906\n",
      "@@@@ 6\n",
      "Donor splice site 187736\n",
      "##$$$ chr6:187536-187935 9\n",
      "Acceptor splice site 186561\n",
      "@@@@ 7\n",
      "Donor splice site 186356\n",
      "##$$$ chr6:186156-186555 10\n",
      "Acceptor splice site 184809\n",
      "@@@@ 8\n",
      "Donor splice site 184593\n",
      "##$$$ chr6:184393-184792 11\n",
      "Acceptor splice site 182474\n",
      "@@@@ 9\n",
      "Donor splice site 182366\n",
      "##$$$ chr6:182166-182565 12\n",
      "Last acceptor_splice_site 181695\n",
      "@@@@ 10\n",
      "\n",
      " ENST00000604449.1 1\n",
      "+\n",
      "TSS 95124\n",
      "##$$$ chr6:95255-95654 13\n",
      "first donor_splice_site 95454\n",
      "\n",
      " ENST00000651295.1 3\n",
      "-\n",
      "TSS 204112\n",
      "first donor_splice_site 203998\n",
      "##$$$ chr6:203798-204197 14\n",
      "Acceptor splice site 198509\n",
      "@@@@ 11\n",
      "Donor splice site 198165\n",
      "##$$$ chr6:197965-198364 15\n",
      "Last acceptor_splice_site 189649\n",
      "@@@@ 12\n",
      "chr7\n",
      "(71618, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr7\n",
      "\n",
      " ENST00000465755.2 3\n",
      "-\n",
      "TSS 95683\n",
      "first donor_splice_site 95564\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr7.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr7:95364-95763 0\n",
      "Acceptor splice site 71833\n",
      "@@@@ 0\n",
      "Donor splice site 71677\n",
      "##$$$ chr7:71477-71876 1\n",
      "Last acceptor_splice_site 71571\n",
      "@@@@ 1\n",
      "\n",
      " ENST00000478759.2 1\n",
      "+\n",
      "TSS 77038\n",
      "##$$$ chr7:77154-77553 2\n",
      "first donor_splice_site 77353\n",
      "\n",
      " ENST00000615488.1 1\n",
      "+\n",
      "TSS 37032\n",
      "##$$$ chr7:37278-37677 3\n",
      "first donor_splice_site 37477\n",
      "\n",
      " ENST00000632586.1 3\n",
      "-\n",
      "TSS 35489\n",
      "first donor_splice_site 35335\n",
      "##$$$ chr7:35135-35534 4\n",
      "Acceptor splice site 33792\n",
      "@@@@ 2\n",
      "Donor splice site 33681\n",
      "##$$$ chr7:33481-33880 5\n",
      "Last acceptor_splice_site 31606\n",
      "@@@@ 3\n",
      "\n",
      " ENST00000633163.1 4\n",
      "-\n",
      "TSS 35457\n",
      "first donor_splice_site 35335\n",
      "##$$$ chr7:35135-35534 6\n",
      "Acceptor splice site 31606\n",
      "@@@@ 4\n",
      "Donor splice site 31060\n",
      "##$$$ chr7:30860-31259 7\n",
      "Acceptor splice site 21029\n",
      "@@@@ 5\n",
      "Donor splice site 20834\n",
      "##$$$ chr7:20634-21033 8\n",
      "Last acceptor_splice_site 19895\n",
      "@@@@ 6\n",
      "\n",
      " ENST00000635373.1 4\n",
      "-\n",
      "TSS 31194\n",
      "first donor_splice_site 31060\n",
      "##$$$ chr7:30860-31259 9\n",
      "Acceptor splice site 26402\n",
      "@@@@ 7\n",
      "Donor splice site 26289\n",
      "##$$$ chr7:26089-26488 10\n",
      "Acceptor splice site 21029\n",
      "@@@@ 8\n",
      "Donor splice site 20834\n",
      "##$$$ chr7:20634-21033 11\n",
      "Last acceptor_splice_site 19172\n",
      "@@@@ 9\n",
      "\n",
      " ENST00000635506.1 2\n",
      "+\n",
      "TSS 24314\n",
      "##$$$ chr7:24166-24565 12\n",
      "first donor_splice_site 24365\n",
      "Last acceptor_splice_site 26965\n",
      "@@@@ 10\n",
      "##$$$ chr7:26766-27165\n",
      "### 10\n",
      "\n",
      " ENST00000635602.1 2\n",
      "+\n",
      "TSS 12704\n",
      "##$$$ chr7:12623-13022 13\n",
      "first donor_splice_site 12822\n",
      "Last acceptor_splice_site 26965\n",
      "@@@@ 11\n",
      "##$$$ chr7:26766-27165\n",
      "### 11\n",
      "chr8\n",
      "(56255, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr8\n",
      "\n",
      " ENST00000320901.4 1\n",
      "-\n",
      "TSS 167024\n",
      "first donor_splice_site 166086\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr8.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr8:165886-166285 0\n",
      "\n",
      " ENST00000508026.3 1\n",
      "-\n",
      "TSS 200562\n",
      "first donor_splice_site 200385\n",
      "##$$$ chr8:200185-200584 1\n",
      "\n",
      " ENST00000518652.1 2\n",
      "+\n",
      "TSS 72617\n",
      "##$$$ chr8:72502-72901 2\n",
      "first donor_splice_site 72701\n",
      "Last acceptor_splice_site 78905\n",
      "@@@@ 0\n",
      "##$$$ chr8:78706-79105\n",
      "### 0\n",
      "\n",
      " ENST00000520139.2 2\n",
      "-\n",
      "TSS 64320\n",
      "first donor_splice_site 64269\n",
      "##$$$ chr8:64069-64468 3\n",
      "Last acceptor_splice_site 64175\n",
      "@@@@ 1\n",
      "\n",
      " ENST00000520478.1 1\n",
      "-\n",
      "TSS 150728\n",
      "first donor_splice_site 150584\n",
      "##$$$ chr8:150384-150783 4\n",
      "\n",
      " ENST00000521145.5 1\n",
      "+\n",
      "TSS 232137\n",
      "##$$$ chr8:232068-232467 5\n",
      "first donor_splice_site 232267\n",
      "\n",
      " ENST00000521854.1 1\n",
      "-\n",
      "TSS 213707\n",
      "first donor_splice_site 213251\n",
      "##$$$ chr8:213051-213450 6\n",
      "\n",
      " ENST00000522481.3 2\n",
      "+\n",
      "TSS 72601\n",
      "##$$$ chr8:72474-72873 7\n",
      "first donor_splice_site 72673\n",
      "Last acceptor_splice_site 78905\n",
      "@@@@ 2\n",
      "##$$$ chr8:78706-79105\n",
      "### 2\n",
      "\n",
      " ENST00000522984.2 2\n",
      "-\n",
      "TSS 202897\n",
      "first donor_splice_site 202822\n",
      "##$$$ chr8:202622-203021 8\n",
      "Last acceptor_splice_site 202747\n",
      "@@@@ 3\n",
      "\n",
      " ENST00000523033.1 1\n",
      "+\n",
      "TSS 155274\n",
      "##$$$ chr8:155282-155681 9\n",
      "first donor_splice_site 155481\n",
      "\n",
      " ENST00000606507.1 3\n",
      "-\n",
      "TSS 232231\n",
      "first donor_splice_site 232129\n",
      "##$$$ chr8:231929-232328 10\n",
      "Acceptor splice site 230907\n",
      "@@@@ 4\n",
      "Donor splice site 230804\n",
      "##$$$ chr8:230604-231003 11\n",
      "Last acceptor_splice_site 230170\n",
      "@@@@ 5\n",
      "\n",
      " ENST00000606975.1 3\n",
      "-\n",
      "TSS 232211\n",
      "first donor_splice_site 232129\n",
      "##$$$ chr8:231929-232328 12\n",
      "Acceptor splice site 219417\n",
      "@@@@ 6\n",
      "Donor splice site 219290\n",
      "##$$$ chr8:219090-219489 13\n",
      "Last acceptor_splice_site 213710\n",
      "@@@@ 7\n",
      "chr9\n",
      "(55272, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chr9\n",
      "\n",
      " ENST00000421620.2 6\n",
      "+\n",
      "TSS 12134\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr9.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chr9:11991-12390 0\n",
      "first donor_splice_site 12190\n",
      "Acceptor splice site 12291\n",
      "@@@@ 0\n",
      "##$$$ chr9:12092-12491\n",
      "### 0\n",
      "Donor splice site 12340\n",
      "##$$$ chr9:12141-12540 1\n",
      "Acceptor splice site 12726\n",
      "@@@@ 1\n",
      "##$$$ chr9:12527-12926\n",
      "### 1\n",
      "Donor splice site 12834\n",
      "##$$$ chr9:12635-13034 2\n",
      "Acceptor splice site 13088\n",
      "@@@@ 2\n",
      "##$$$ chr9:12889-13288\n",
      "### 2\n",
      "Donor splice site 13157\n",
      "##$$$ chr9:12958-13357 3\n",
      "Acceptor splice site 13338\n",
      "@@@@ 3\n",
      "##$$$ chr9:13139-13538\n",
      "### 3\n",
      "Donor splice site 13487\n",
      "##$$$ chr9:13288-13687 4\n",
      "Last acceptor_splice_site 13566\n",
      "@@@@ 4\n",
      "##$$$ chr9:13367-13766\n",
      "### 4\n",
      "\n",
      " ENST00000442898.5 11\n",
      "-\n",
      "TSS 29739\n",
      "first donor_splice_site 29602\n",
      "##$$$ chr9:29402-29801 5\n",
      "Acceptor splice site 25004\n",
      "@@@@ 5\n",
      "Donor splice site 24851\n",
      "##$$$ chr9:24651-25050 6\n",
      "Acceptor splice site 18492\n",
      "@@@@ 6\n",
      "Donor splice site 18381\n",
      "##$$$ chr9:18181-18580 7\n",
      "Acceptor splice site 18174\n",
      "@@@@ 7\n",
      "Donor splice site 18028\n",
      "##$$$ chr9:17828-18227 8\n",
      "Acceptor splice site 17855\n",
      "@@@@ 8\n",
      "Donor splice site 17719\n",
      "##$$$ chr9:17519-17918 9\n",
      "Acceptor splice site 17479\n",
      "@@@@ 9\n",
      "Donor splice site 17344\n",
      "##$$$ chr9:17144-17543 10\n",
      "Acceptor splice site 17166\n",
      "@@@@ 10\n",
      "Donor splice site 16965\n",
      "##$$$ chr9:16765-17164 11\n",
      "Acceptor splice site 16876\n",
      "@@@@ 11\n",
      "Donor splice site 16718\n",
      "##$$$ chr9:16518-16917 12\n",
      "Acceptor splice site 16061\n",
      "@@@@ 12\n",
      "Donor splice site 15909\n",
      "##$$$ chr9:15709-16108 13\n",
      "Acceptor splice site 15149\n",
      "@@@@ 13\n",
      "Donor splice site 15081\n",
      "##$$$ chr9:14881-15280 14\n",
      "Last acceptor_splice_site 14940\n",
      "@@@@ 14\n",
      "\n",
      " ENST00000642633.1 3\n",
      "-\n",
      "TSS 73865\n",
      "first donor_splice_site 73683\n",
      "##$$$ chr9:73483-73882 15\n",
      "Acceptor splice site 52614\n",
      "@@@@ 15\n",
      "Donor splice site 52452\n",
      "##$$$ chr9:52252-52651 16\n",
      "Last acceptor_splice_site 45504\n",
      "@@@@ 16\n",
      "chrX\n",
      "(47267, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chrX\n",
      "\n",
      " ENST00000399012.6 8\n",
      "+\n",
      "TSS 276322\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chrX.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chrX:276195-276594 0\n",
      "first donor_splice_site 276394\n",
      "Acceptor splice site 281482\n",
      "@@@@ 0\n",
      "##$$$ chrX:281283-281682\n",
      "### 0\n",
      "Donor splice site 281684\n",
      "##$$$ chrX:281485-281884 1\n",
      "Acceptor splice site 284167\n",
      "@@@@ 1\n",
      "##$$$ chrX:283968-284367\n",
      "### 1\n",
      "Donor splice site 284314\n",
      "##$$$ chrX:284115-284514 2\n",
      "Acceptor splice site 288733\n",
      "@@@@ 2\n",
      "##$$$ chrX:288534-288933\n",
      "### 2\n",
      "Donor splice site 288869\n",
      "##$$$ chrX:288670-289069 3\n",
      "Acceptor splice site 290648\n",
      "@@@@ 3\n",
      "##$$$ chrX:290449-290848\n",
      "### 3\n",
      "Donor splice site 290776\n",
      "##$$$ chrX:290577-290976 4\n",
      "Acceptor splice site 291499\n",
      "@@@@ 4\n",
      "##$$$ chrX:291300-291699\n",
      "### 4\n",
      "Donor splice site 291654\n",
      "##$$$ chrX:291455-291854 5\n",
      "Acceptor splice site 293035\n",
      "@@@@ 5\n",
      "##$$$ chrX:292836-293235\n",
      "### 5\n",
      "Donor splice site 293218\n",
      "##$$$ chrX:293019-293418 6\n",
      "Last acceptor_splice_site 299097\n",
      "@@@@ 6\n",
      "##$$$ chrX:298898-299297\n",
      "### 6\n",
      "\n",
      " ENST00000430923.7 5\n",
      "+\n",
      "TSS 276353\n",
      "##$$$ chrX:276195-276594 7\n",
      "first donor_splice_site 276394\n",
      "Acceptor splice site 284167\n",
      "@@@@ 7\n",
      "##$$$ chrX:283968-284367\n",
      "### 7\n",
      "Donor splice site 284314\n",
      "##$$$ chrX:284115-284514 8\n",
      "Acceptor splice site 288733\n",
      "@@@@ 8\n",
      "##$$$ chrX:288534-288933\n",
      "### 8\n",
      "Donor splice site 288869\n",
      "##$$$ chrX:288670-289069 9\n",
      "Acceptor splice site 290648\n",
      "@@@@ 9\n",
      "##$$$ chrX:290449-290848\n",
      "### 9\n",
      "Donor splice site 290776\n",
      "##$$$ chrX:290577-290976 10\n",
      "Last acceptor_splice_site 291499\n",
      "@@@@ 10\n",
      "##$$$ chrX:291300-291699\n",
      "### 10\n",
      "\n",
      " ENST00000431238.7 2\n",
      "+\n",
      "TSS 253743\n",
      "##$$$ chrX:253647-254046 11\n",
      "first donor_splice_site 253846\n",
      "Last acceptor_splice_site 254937\n",
      "@@@@ 11\n",
      "##$$$ chrX:254738-255137\n",
      "### 11\n",
      "\n",
      " ENST00000484611.7 5\n",
      "+\n",
      "TSS 276324\n",
      "##$$$ chrX:276195-276594 12\n",
      "first donor_splice_site 276394\n",
      "Acceptor splice site 284167\n",
      "@@@@ 12\n",
      "##$$$ chrX:283968-284367\n",
      "### 12\n",
      "Donor splice site 284314\n",
      "##$$$ chrX:284115-284514 13\n",
      "Acceptor splice site 288733\n",
      "@@@@ 13\n",
      "##$$$ chrX:288534-288933\n",
      "### 13\n",
      "Donor splice site 288869\n",
      "##$$$ chrX:288670-289069 14\n",
      "Acceptor splice site 290648\n",
      "@@@@ 14\n",
      "##$$$ chrX:290449-290848\n",
      "### 14\n",
      "Donor splice site 290792\n",
      "##$$$ chrX:290593-290992 15\n",
      "Last acceptor_splice_site 291499\n",
      "@@@@ 15\n",
      "##$$$ chrX:291300-291699\n",
      "### 15\n",
      "chrY\n",
      "(5283, 25)\n",
      "/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/chrY\n",
      "\n",
      " ENST00000399012.6_PAR_Y 8\n",
      "+\n",
      "TSS 276322\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chrY.fa.fai not found, generating...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##$$$ chrY:276195-276594 0\n",
      "first donor_splice_site 276394\n",
      "Acceptor splice site 281482\n",
      "@@@@ 0\n",
      "##$$$ chrY:281283-281682\n",
      "### 0\n",
      "Donor splice site 281684\n",
      "##$$$ chrY:281485-281884 1\n",
      "Acceptor splice site 284167\n",
      "@@@@ 1\n",
      "##$$$ chrY:283968-284367\n",
      "### 1\n",
      "Donor splice site 284314\n",
      "##$$$ chrY:284115-284514 2\n",
      "Acceptor splice site 288733\n",
      "@@@@ 2\n",
      "##$$$ chrY:288534-288933\n",
      "### 2\n",
      "Donor splice site 288869\n",
      "##$$$ chrY:288670-289069 3\n",
      "Acceptor splice site 290648\n",
      "@@@@ 3\n",
      "##$$$ chrY:290449-290848\n",
      "### 3\n",
      "Donor splice site 290776\n",
      "##$$$ chrY:290577-290976 4\n",
      "Acceptor splice site 291499\n",
      "@@@@ 4\n",
      "##$$$ chrY:291300-291699\n",
      "### 4\n",
      "Donor splice site 291654\n",
      "##$$$ chrY:291455-291854 5\n",
      "Acceptor splice site 293035\n",
      "@@@@ 5\n",
      "##$$$ chrY:292836-293235\n",
      "### 5\n",
      "Donor splice site 293218\n",
      "##$$$ chrY:293019-293418 6\n",
      "Last acceptor_splice_site 299097\n",
      "@@@@ 6\n",
      "##$$$ chrY:298898-299297\n",
      "### 6\n",
      "\n",
      " ENST00000430923.7_PAR_Y 5\n",
      "+\n",
      "TSS 276353\n",
      "##$$$ chrY:276195-276594 7\n",
      "first donor_splice_site 276394\n",
      "Acceptor splice site 284167\n",
      "@@@@ 7\n",
      "##$$$ chrY:283968-284367\n",
      "### 7\n",
      "Donor splice site 284314\n",
      "##$$$ chrY:284115-284514 8\n",
      "Acceptor splice site 288733\n",
      "@@@@ 8\n",
      "##$$$ chrY:288534-288933\n",
      "### 8\n",
      "Donor splice site 288869\n",
      "##$$$ chrY:288670-289069 9\n",
      "Acceptor splice site 290648\n",
      "@@@@ 9\n",
      "##$$$ chrY:290449-290848\n",
      "### 9\n",
      "Donor splice site 290776\n",
      "##$$$ chrY:290577-290976 10\n",
      "Last acceptor_splice_site 291499\n",
      "@@@@ 10\n",
      "##$$$ chrY:291300-291699\n",
      "### 10\n",
      "\n",
      " ENST00000431238.7_PAR_Y 2\n",
      "+\n",
      "TSS 253743\n",
      "##$$$ chrY:253647-254046 11\n",
      "first donor_splice_site 253846\n",
      "Last acceptor_splice_site 254937\n",
      "@@@@ 11\n",
      "##$$$ chrY:254738-255137\n",
      "### 11\n",
      "\n",
      " ENST00000484611.7_PAR_Y 5\n",
      "+\n",
      "TSS 276324\n",
      "##$$$ chrY:276195-276594 12\n",
      "first donor_splice_site 276394\n",
      "Acceptor splice site 284167\n",
      "@@@@ 12\n",
      "##$$$ chrY:283968-284367\n",
      "### 12\n",
      "Donor splice site 284314\n",
      "##$$$ chrY:284115-284514 13\n",
      "Acceptor splice site 288733\n",
      "@@@@ 13\n",
      "##$$$ chrY:288534-288933\n",
      "### 13\n",
      "Donor splice site 288869\n",
      "##$$$ chrY:288670-289069 14\n",
      "Acceptor splice site 290648\n",
      "@@@@ 14\n",
      "##$$$ chrY:290449-290848\n",
      "### 14\n",
      "Donor splice site 290792\n",
      "##$$$ chrY:290593-290992 15\n",
      "Last acceptor_splice_site 291499\n",
      "@@@@ 15\n",
      "##$$$ chrY:291300-291699\n",
      "### 15\n"
     ]
    }
   ],
   "source": [
    "global df_promoter, df_core_promoter, df_donor_splice, df_acceptor_splice\n",
    "\n",
    "\n",
    "for file in chr_files:\n",
    "    if file.endswith('.fa'):\n",
    "        reference_genome_file_path = reference_genome_path + file\n",
    "        chr_name = file.split('.')[0]\n",
    "        print (chr_name)\n",
    "        df_chr= df[df['seqname']==chr_name]\n",
    "        df_new = df_chr[df_chr['feature']=='exon'].reset_index(drop=True)\n",
    "        print (df_new.shape)\n",
    "        #print (reference_genome_file_path)\n",
    "        fasta = pybedtools.example_filename(reference_genome_file_path)\n",
    "        \n",
    "        \n",
    "        folder_name = data_save_path + chr_name   \n",
    "        print (folder_name)\n",
    "        if not os.path.exists(folder_name):\n",
    "            os.makedirs(folder_name)\n",
    "        \n",
    "        \n",
    "        df_promoter = pd.DataFrame(columns= ['gene_id','transcript_id','strand','exon_id','transcript_type','TSS', 'sequence'])\n",
    "        df_core_promoter = pd.DataFrame(columns= ['gene_id','transcript_id','strand','exon_id','transcript_type','TSS','sequence'])\n",
    "        df_donor_splice = pd.DataFrame(columns= ['transcript_id','exon_id','transcript_type','strand','coordinates','sequence'])\n",
    "        df_acceptor_splice= pd.DataFrame(columns= ['transcript_id','exon_id', 'transcript_type','strand','coordinates', 'sequence'])\n",
    "        promoter_index = core_promoter_index = donor_index = acceptor_index = 0\n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "        df_by_transcript = df_new.head(20).groupby(\"transcript_id\")\n",
    "        for transcript, entries in df_by_transcript:\n",
    "            print (\"\\n\", transcript,len( entries) )\n",
    "            print (entries.iloc[0]['strand'])\n",
    "            if (entries.iloc[0]['strand']=='+'):\n",
    "                for index, entry in entries.iterrows():\n",
    "                    if (int(entry['exon_number'])==1):\n",
    "                        print(\"TSS\", entry['start'])\n",
    "                        promoter_sequence(entry, df_promoter, promoter_index)\n",
    "                        promoter_index= promoter_index+1\n",
    "                        core_promoter_sequence(entry, df_core_promoter, core_promoter_index)\n",
    "                        core_promoter_index = core_promoter_index +1\n",
    "                        donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                        donor_index =donor_index+ 1\n",
    "                        print(\"first donor_splice_site\", entry['end'])\n",
    "                    elif (int(entry['exon_number'])==len(entries)):\n",
    "                        print (\"Last acceptor_splice_site\",  entry['start'])\n",
    "                        acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                        acceptor_index = acceptor_index +1\n",
    "                    else:\n",
    "                        print (\"Acceptor splice site\", entry['start'])\n",
    "                        acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                        acceptor_index = acceptor_index +1\n",
    "                        print (\"Donor splice site\", entry['end'])\n",
    "                        donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                        donor_index =donor_index+ 1\n",
    "            elif (entries.iloc[0]['strand']=='-'):\n",
    "                for index, entry in entries.iterrows():\n",
    "                    if (int(entry['exon_number'])==1):\n",
    "                        print(\"TSS\", entry['end'])\n",
    "                        print(\"first donor_splice_site\", entry['start'])\n",
    "                        promoter_sequence(entry, df_promoter, promoter_index)\n",
    "                        promoter_index= promoter_index+1\n",
    "                        core_promoter_sequence(entry, df_core_promoter, core_promoter_index)\n",
    "                        core_promoter_index = core_promoter_index +1\n",
    "                        donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                        donor_index =donor_index+ 1\n",
    "                    elif (int(entry['exon_number'])==len(entries)):\n",
    "                        print (\"Last acceptor_splice_site\",  entry['end'])\n",
    "                        acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                        acceptor_index = acceptor_index +1\n",
    "                    else:\n",
    "                        print (\"Acceptor splice site\", entry['end'])\n",
    "                        acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                        acceptor_index = acceptor_index +1\n",
    "                        print (\"Donor splice site\", entry['start'])\n",
    "                        donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                        donor_index =donor_index+ 1\n",
    "        df_promoter.to_csv(folder_name +\"/\"+ entry['seqname']+\"_promoter.csv\", index=False)\n",
    "        df_core_promoter.to_csv(folder_name +\"/\"+ entry['seqname']+\"_core_promoter_sequence.csv\", index=False)\n",
    "        df_acceptor_splice.to_csv(folder_name +\"/\"+ entry['seqname']+\"_acceptor_splice_set_sequence.csv\", index=False)\n",
    "        df_donor_splice.to_csv(folder_name +\"/\"+ entry['seqname']+\"_donor_splice_set.csv\", index=False)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cd5bea5-1a81-4f50-925e-48deae690c73",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "df_promoter = pd.DataFrame(columns= ['gene_id','transcript_id','strand','exon_id','transcript_type','TSS', 'sequence'])\n",
    "df_core_promoter = pd.DataFrame(columns= ['gene_id','transcript_id','strand','exon_id','transcript_type','TSS','sequence'])\n",
    "df_donor_splice = pd.DataFrame(columns= ['transcript_id','exon_id','transcript_type','strand','coordinates','sequence'])\n",
    "df_acceptor_splice= pd.DataFrame(columns= ['transcript_id','exon_id', 'transcript_type','strand','coordinates', 'sequence'])\n",
    "#global promoter_index, core_promoter_index, donor_index, acceptor_index\n",
    "promoter_index = core_promoter_index = donor_index = acceptor_index = 0\n",
    "\n",
    "df_by_transcript = df_chr1_transcript.head(20).groupby(\"transcript_id\")\n",
    "for transcript, entries in df_by_transcript:\n",
    "    print (\"\\n\", transcript,len( entries) )\n",
    "    print (entries.iloc[0]['strand'])\n",
    "    if (entries.iloc[0]['strand']=='+'):\n",
    "        for index, entry in entries.iterrows():\n",
    "            if (int(entry['exon_number'])==1):\n",
    "                print(\"TSS\", entry['start'])\n",
    "                promoter_sequence(entry, df_promoter, promoter_index)\n",
    "                promoter_index= promoter_index+1\n",
    "                \n",
    "                core_promoter_sequence(entry, df_core_promoter, core_promoter_index)\n",
    "                core_promoter_index = core_promoter_index +1\n",
    "                \n",
    "                donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                donor_index =donor_index+ 1\n",
    "                print(\"first donor_splice_site\", entry['end'])\n",
    "            elif (int(entry['exon_number'])==len(entries)):\n",
    "                print (\"Last acceptor_splice_site\",  entry['start'])\n",
    "                acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                acceptor_index = acceptor_index +1\n",
    "            else:\n",
    "                print (\"Acceptor splice site\", entry['start'])\n",
    "                acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                acceptor_index = acceptor_index +1\n",
    "                print (\"Donor splice site\", entry['end'])\n",
    "                donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                donor_index =donor_index+ 1\n",
    "    elif (entries.iloc[0]['strand']=='-'):\n",
    "        for index, entry in entries.iterrows():\n",
    "            if (int(entry['exon_number'])==1):\n",
    "                print(\"TSS\", entry['end'])\n",
    "                print(\"first donor_splice_site\", entry['start'])\n",
    "                \n",
    "                promoter_sequence(entry, df_promoter, promoter_index)\n",
    "                promoter_index= promoter_index+1\n",
    "                \n",
    "                core_promoter_sequence(entry, df_core_promoter, core_promoter_index)\n",
    "                core_promoter_index = core_promoter_index +1\n",
    "                \n",
    "                donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                donor_index =donor_index+ 1\n",
    "                \n",
    "                \n",
    "            elif (int(entry['exon_number'])==len(entries)):\n",
    "                print (\"Last acceptor_splice_site\",  entry['end'])\n",
    "                acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                acceptor_index = acceptor_index +1\n",
    "            else:\n",
    "                print (\"Acceptor splice site\", entry['end'])\n",
    "                acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                acceptor_index = acceptor_index +1\n",
    "                print (\"Donor splice site\", entry['start'])\n",
    "                donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                donor_index =donor_index+ 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bd36b20-1acb-41cd-bbff-67f9dd997003",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "6045744b-f030-4aa6-aa35-98910e8fdb12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>TSS</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>ENST00000417324.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001656588.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>36081</td>\n",
       "      <td>AGAATTGGTTTGATTTACATAATCCTGCCCCTGAGATTTAGCCCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>12010</td>\n",
       "      <td>gccggtgttgagaatgactgCGCAAATTTGCCGGATTTCCTTTGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>atgcctgtttctccacaaagtgtttacttttggatttttgccagtc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>ENST00000461467.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001618781.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>36073</td>\n",
       "      <td>TTTGATTTACATAATCCTGCCCCTGAGATTTAGCCCCACCCCTGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENST00000466430.5</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001606755.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>120932</td>\n",
       "      <td>TCTTTACTTGCACTGACAAATATGAACGCTGGGGAATTTGGAAatg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001841699.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>30267</td>\n",
       "      <td>ttgggcagctcgtgttcaatgggtagagtttcaggctggggtgatg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENST00000471248.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001934975.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>129173</td>\n",
       "      <td>gactccatcttataaaaggaaaaaagaaagaaaagaaaaaTTCCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001947070.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>29554</td>\n",
       "      <td>CCGGACGCCTCCTCACCTGCGAGCCGCCCTCCCGGAAGCTCCCGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENST00000477740.5</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001919246.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>129217</td>\n",
       "      <td>agctgagatcgcaccagtgcactccagcctgggcaacagagtgaga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001890219.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>29570</td>\n",
       "      <td>GCTGAGAGACGCAGGAAAGCGGGTCAAggcgtagggctggagggca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSG00000240361.2</td>\n",
       "      <td>ENST00000492842.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001830178.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>62949</td>\n",
       "      <td>AGTTCATGTATTTGAGCAATTGAGAACATTTAAAGCAATAGACTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000268020.3</td>\n",
       "      <td>ENST00000606857.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003698237.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>52473</td>\n",
       "      <td>AATAAATAACTTATTATCACCGCATTTCCCCAAttcatttatcttt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>ENST00000607096.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003695741.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>30366</td>\n",
       "      <td>ttaatcccactgaactgtatgctgaaaaatggtttagacggtgaat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00003746039.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>17436</td>\n",
       "      <td>GCAGAAGAGGGACTTGGGGGTCTTCCCAGCAACATCAGCTCTGTCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003812156.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>65419</td>\n",
       "      <td>GCATTATTACTTACTCTCTTGTTAACCTATCTGGATTTTAATTTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSG00000240361.2</td>\n",
       "      <td>ENST00000642116.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003812686.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>57598</td>\n",
       "      <td>AATGTAGTACCTCAAATCCTTCCTTCCCTCCCCCTATTTCATCAGC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id      transcript_id strand            exon_id  \\\n",
       "0   ENSG00000237613.2  ENST00000417324.1      -  ENSE00001656588.1   \n",
       "1   ENSG00000223972.5  ENST00000450305.2      +  ENSE00001948541.1   \n",
       "2   ENSG00000223972.5  ENST00000456328.2      +  ENSE00002234944.1   \n",
       "3   ENSG00000237613.2  ENST00000461467.1      -  ENSE00001618781.2   \n",
       "4   ENSG00000238009.6  ENST00000466430.5      -  ENSE00001606755.2   \n",
       "5   ENSG00000243485.5  ENST00000469289.1      +  ENSE00001841699.1   \n",
       "6   ENSG00000238009.6  ENST00000471248.1      -  ENSE00001934975.1   \n",
       "7   ENSG00000243485.5  ENST00000473358.1      +  ENSE00001947070.1   \n",
       "8   ENSG00000238009.6  ENST00000477740.5      -  ENSE00001919246.1   \n",
       "9   ENSG00000227232.5  ENST00000488147.1      -  ENSE00001890219.1   \n",
       "10  ENSG00000240361.2  ENST00000492842.2      +  ENSE00001830178.2   \n",
       "11  ENSG00000268020.3  ENST00000606857.1      +  ENSE00003698237.1   \n",
       "12  ENSG00000284332.1  ENST00000607096.1      +  ENSE00003695741.1   \n",
       "13  ENSG00000278267.1  ENST00000619216.1      -  ENSE00003746039.1   \n",
       "14  ENSG00000186092.7  ENST00000641515.2      +  ENSE00003812156.1   \n",
       "15  ENSG00000240361.2  ENST00000642116.1      +  ENSE00003812686.1   \n",
       "\n",
       "                       transcript_type     TSS  \\\n",
       "0                               lncRNA   36081   \n",
       "1   transcribed_unprocessed_pseudogene   12010   \n",
       "2                 processed_transcript   11869   \n",
       "3                               lncRNA   36073   \n",
       "4                               lncRNA  120932   \n",
       "5                               lncRNA   30267   \n",
       "6                               lncRNA  129173   \n",
       "7                               lncRNA   29554   \n",
       "8                               lncRNA  129217   \n",
       "9               unprocessed_pseudogene   29570   \n",
       "10  transcribed_unprocessed_pseudogene   62949   \n",
       "11              unprocessed_pseudogene   52473   \n",
       "12                               miRNA   30366   \n",
       "13                               miRNA   17436   \n",
       "14                      protein_coding   65419   \n",
       "15                processed_transcript   57598   \n",
       "\n",
       "                                             sequence  \n",
       "0   AGAATTGGTTTGATTTACATAATCCTGCCCCTGAGATTTAGCCCCA...  \n",
       "1   gccggtgttgagaatgactgCGCAAATTTGCCGGATTTCCTTTGCT...  \n",
       "2   atgcctgtttctccacaaagtgtttacttttggatttttgccagtc...  \n",
       "3   TTTGATTTACATAATCCTGCCCCTGAGATTTAGCCCCACCCCTGAG...  \n",
       "4   TCTTTACTTGCACTGACAAATATGAACGCTGGGGAATTTGGAAatg...  \n",
       "5   ttgggcagctcgtgttcaatgggtagagtttcaggctggggtgatg...  \n",
       "6   gactccatcttataaaaggaaaaaagaaagaaaagaaaaaTTCCAT...  \n",
       "7   CCGGACGCCTCCTCACCTGCGAGCCGCCCTCCCGGAAGCTCCCGCC...  \n",
       "8   agctgagatcgcaccagtgcactccagcctgggcaacagagtgaga...  \n",
       "9   GCTGAGAGACGCAGGAAAGCGGGTCAAggcgtagggctggagggca...  \n",
       "10  AGTTCATGTATTTGAGCAATTGAGAACATTTAAAGCAATAGACTGC...  \n",
       "11  AATAAATAACTTATTATCACCGCATTTCCCCAAttcatttatcttt...  \n",
       "12  ttaatcccactgaactgtatgctgaaaaatggtttagacggtgaat...  \n",
       "13  GCAGAAGAGGGACTTGGGGGTCTTCCCAGCAACATCAGCTCTGTCA...  \n",
       "14  GCATTATTACTTACTCTCTTGTTAACCTATCTGGATTTTAATTTTG...  \n",
       "15  AATGTAGTACCTCAAATCCTTCCTTCCCTCCCCCTATTTCATCAGC...  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "99de0a0b-20a5-4fd1-919f-89703dc8c677",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>TSS</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>ENST00000417324.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001656588.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>36081</td>\n",
       "      <td>CAAGGATTAGTCCTGCTGCCCTGTGCCCAGCTCCCACACAACGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>12010</td>\n",
       "      <td>CACAGGGTCTTGATGCTGTGGTCTTCATCTGCAGGTGTCTGACTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>CCAGCACCGGGTATCATTCACCATTTTTCTTTTCGTTAACTTGCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>ENST00000461467.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001618781.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>36073</td>\n",
       "      <td>AGTCCTGCTGCCCTGTGCCCAGCTCCCACACAACGGGGTTTCGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENST00000466430.5</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001606755.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>120932</td>\n",
       "      <td>TTTCCATGTGATTTTCCTCTTAATTTTTTTCTAGCTGATCCATATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001841699.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>30267</td>\n",
       "      <td>GCTTTTTTAAAAAGTTAAATATATAAAAAGAGCATCATCAGTCCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENST00000471248.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001934975.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>129173</td>\n",
       "      <td>AGCATCTTAAAATGATTCAACAGGAGGAGATAAGGAAGCTCGAGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001947070.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>29554</td>\n",
       "      <td>GGCACGGAACTCACGCTCACTCCGAGCTCCCGACGTGCACACGGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENST00000477740.5</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001919246.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>129217</td>\n",
       "      <td>GTGTCACAACTGCACATTGGTTTCACAGCTGCAGGACAAGTTCGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001890219.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>29570</td>\n",
       "      <td>GGGTAGGGGCGGCCTGACGCTCGGAAGACAACGCATGGGAGCCGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSG00000240361.2</td>\n",
       "      <td>ENST00000492842.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001830178.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>62949</td>\n",
       "      <td>GGTTGCTTCAGAGTCTTCCCTTCTATCTGATTCAGTGGACCAAGTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000268020.3</td>\n",
       "      <td>ENST00000606857.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003698237.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>52473</td>\n",
       "      <td>GAGTCAATGGCCAATGGATAGAGTAAATAATTCTGCGGTATCTAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>ENST00000607096.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003695741.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>30366</td>\n",
       "      <td>CCCTTGCCTTTtagggttgccagattaaaagacaggatgcccagct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00003746039.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>17436</td>\n",
       "      <td>TGGGACCCCACAGTGCACGCTGTGCCCCTGATGATGTGGGAGAGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003812156.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>65419</td>\n",
       "      <td>TTAGTCTCAGTACACAAAGCTCAGACTGGCTATTCCCAGATCTCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSG00000240361.2</td>\n",
       "      <td>ENST00000642116.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003812686.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>57598</td>\n",
       "      <td>CTTTGTCTTTGGGGTATTACATAAATGAAGATTAAGCTATCTGAAT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id      transcript_id strand            exon_id  \\\n",
       "0   ENSG00000237613.2  ENST00000417324.1      -  ENSE00001656588.1   \n",
       "1   ENSG00000223972.5  ENST00000450305.2      +  ENSE00001948541.1   \n",
       "2   ENSG00000223972.5  ENST00000456328.2      +  ENSE00002234944.1   \n",
       "3   ENSG00000237613.2  ENST00000461467.1      -  ENSE00001618781.2   \n",
       "4   ENSG00000238009.6  ENST00000466430.5      -  ENSE00001606755.2   \n",
       "5   ENSG00000243485.5  ENST00000469289.1      +  ENSE00001841699.1   \n",
       "6   ENSG00000238009.6  ENST00000471248.1      -  ENSE00001934975.1   \n",
       "7   ENSG00000243485.5  ENST00000473358.1      +  ENSE00001947070.1   \n",
       "8   ENSG00000238009.6  ENST00000477740.5      -  ENSE00001919246.1   \n",
       "9   ENSG00000227232.5  ENST00000488147.1      -  ENSE00001890219.1   \n",
       "10  ENSG00000240361.2  ENST00000492842.2      +  ENSE00001830178.2   \n",
       "11  ENSG00000268020.3  ENST00000606857.1      +  ENSE00003698237.1   \n",
       "12  ENSG00000284332.1  ENST00000607096.1      +  ENSE00003695741.1   \n",
       "13  ENSG00000278267.1  ENST00000619216.1      -  ENSE00003746039.1   \n",
       "14  ENSG00000186092.7  ENST00000641515.2      +  ENSE00003812156.1   \n",
       "15  ENSG00000240361.2  ENST00000642116.1      +  ENSE00003812686.1   \n",
       "\n",
       "                       transcript_type     TSS  \\\n",
       "0                               lncRNA   36081   \n",
       "1   transcribed_unprocessed_pseudogene   12010   \n",
       "2                 processed_transcript   11869   \n",
       "3                               lncRNA   36073   \n",
       "4                               lncRNA  120932   \n",
       "5                               lncRNA   30267   \n",
       "6                               lncRNA  129173   \n",
       "7                               lncRNA   29554   \n",
       "8                               lncRNA  129217   \n",
       "9               unprocessed_pseudogene   29570   \n",
       "10  transcribed_unprocessed_pseudogene   62949   \n",
       "11              unprocessed_pseudogene   52473   \n",
       "12                               miRNA   30366   \n",
       "13                               miRNA   17436   \n",
       "14                      protein_coding   65419   \n",
       "15                processed_transcript   57598   \n",
       "\n",
       "                                             sequence  \n",
       "0   CAAGGATTAGTCCTGCTGCCCTGTGCCCAGCTCCCACACAACGGGG...  \n",
       "1   CACAGGGTCTTGATGCTGTGGTCTTCATCTGCAGGTGTCTGACTTC...  \n",
       "2   CCAGCACCGGGTATCATTCACCATTTTTCTTTTCGTTAACTTGCCG...  \n",
       "3   AGTCCTGCTGCCCTGTGCCCAGCTCCCACACAACGGGGTTTCGGGG...  \n",
       "4   TTTCCATGTGATTTTCCTCTTAATTTTTTTCTAGCTGATCCATATG...  \n",
       "5   GCTTTTTTAAAAAGTTAAATATATAAAAAGAGCATCATCAGTCCAA...  \n",
       "6   AGCATCTTAAAATGATTCAACAGGAGGAGATAAGGAAGCTCGAGGA...  \n",
       "7   GGCACGGAACTCACGCTCACTCCGAGCTCCCGACGTGCACACGGCT...  \n",
       "8   GTGTCACAACTGCACATTGGTTTCACAGCTGCAGGACAAGTTCGAG...  \n",
       "9   GGGTAGGGGCGGCCTGACGCTCGGAAGACAACGCATGGGAGCCGTG...  \n",
       "10  GGTTGCTTCAGAGTCTTCCCTTCTATCTGATTCAGTGGACCAAGTA...  \n",
       "11  GAGTCAATGGCCAATGGATAGAGTAAATAATTCTGCGGTATCTAAA...  \n",
       "12  CCCTTGCCTTTtagggttgccagattaaaagacaggatgcccagct...  \n",
       "13  TGGGACCCCACAGTGCACGCTGTGCCCCTGATGATGTGGGAGAGGA...  \n",
       "14  TTAGTCTCAGTACACAAAGCTCAGACTGGCTATTCCCAGATCTCTT...  \n",
       "15  CTTTGTCTTTGGGGTATTACATAAATGAAGATTAAGCTATCTGAAT...  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_core_promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "c32274e1-5f2e-4a43-9db4-cf2a78b3c044",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>strand</th>\n",
       "      <th>coordinates</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000417324.1</td>\n",
       "      <td>ENSE00001656588.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>35720</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000417324.1</td>\n",
       "      <td>ENSE00001669267.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>35276</td>\n",
       "      <td>tgt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>12058</td>\n",
       "      <td>GCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>12228</td>\n",
       "      <td>AGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001758273.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>12698</td>\n",
       "      <td>TGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001799933.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>13053</td>\n",
       "      <td>GGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001746346.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>13375</td>\n",
       "      <td>AGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>12228</td>\n",
       "      <td>AGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>12722</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000461467.1</td>\n",
       "      <td>ENSE00001618781.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>35720</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000466430.5</td>\n",
       "      <td>ENSE00001606755.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>120774</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000466430.5</td>\n",
       "      <td>ENSE00001957285.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>112699</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000466430.5</td>\n",
       "      <td>ENSE00001944529.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>92090</td>\n",
       "      <td>ggt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>ENSE00001841699.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>30668</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000471248.1</td>\n",
       "      <td>ENSE00001934975.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>129054</td>\n",
       "      <td>CGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSE00001947070.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>30040</td>\n",
       "      <td>ggt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSE00001922571.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>30668</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENST00000477740.5</td>\n",
       "      <td>ENSE00001919246.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>129054</td>\n",
       "      <td>CGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENST00000477740.5</td>\n",
       "      <td>ENSE00001171005.3</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>120720</td>\n",
       "      <td>TGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENST00000477740.5</td>\n",
       "      <td>ENSE00001957285.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>112699</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00001890219.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>29533</td>\n",
       "      <td>CGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003507205.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>24737</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003477500.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>18267</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003565697.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>17914</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003475637.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>17605</td>\n",
       "      <td>TGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003502542.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>17232</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003553898.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>16857</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003621279.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>16606</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00002030414.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>15795</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00001935574.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>15004</td>\n",
       "      <td>CGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ENST00000492842.2</td>\n",
       "      <td>ENSE00001830178.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>63888</td>\n",
       "      <td>ACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ENST00000606857.1</td>\n",
       "      <td>ENSE00003698237.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>53313</td>\n",
       "      <td>CGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ENST00000607096.1</td>\n",
       "      <td>ENSE00003695741.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>30504</td>\n",
       "      <td>ata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>ENSE00003746039.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>17368</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSE00003812156.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "      <td>65434</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSE00003813641.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "      <td>65574</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>ENST00000642116.1</td>\n",
       "      <td>ENSE00003812686.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>57654</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ENST00000642116.1</td>\n",
       "      <td>ENSE00003812505.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>58857</td>\n",
       "      <td>TGT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id            exon_id                     transcript_type  \\\n",
       "0   ENST00000417324.1  ENSE00001656588.1                              lncRNA   \n",
       "1   ENST00000417324.1  ENSE00001669267.1                              lncRNA   \n",
       "2   ENST00000450305.2  ENSE00001948541.1  transcribed_unprocessed_pseudogene   \n",
       "3   ENST00000450305.2  ENSE00001671638.2  transcribed_unprocessed_pseudogene   \n",
       "4   ENST00000450305.2  ENSE00001758273.2  transcribed_unprocessed_pseudogene   \n",
       "5   ENST00000450305.2  ENSE00001799933.2  transcribed_unprocessed_pseudogene   \n",
       "6   ENST00000450305.2  ENSE00001746346.2  transcribed_unprocessed_pseudogene   \n",
       "7   ENST00000456328.2  ENSE00002234944.1                processed_transcript   \n",
       "8   ENST00000456328.2  ENSE00003582793.1                processed_transcript   \n",
       "9   ENST00000461467.1  ENSE00001618781.2                              lncRNA   \n",
       "10  ENST00000466430.5  ENSE00001606755.2                              lncRNA   \n",
       "11  ENST00000466430.5  ENSE00001957285.1                              lncRNA   \n",
       "12  ENST00000466430.5  ENSE00001944529.1                              lncRNA   \n",
       "13  ENST00000469289.1  ENSE00001841699.1                              lncRNA   \n",
       "14  ENST00000471248.1  ENSE00001934975.1                              lncRNA   \n",
       "15  ENST00000473358.1  ENSE00001947070.1                              lncRNA   \n",
       "16  ENST00000473358.1  ENSE00001922571.1                              lncRNA   \n",
       "17  ENST00000477740.5  ENSE00001919246.1                              lncRNA   \n",
       "18  ENST00000477740.5  ENSE00001171005.3                              lncRNA   \n",
       "19  ENST00000477740.5  ENSE00001957285.1                              lncRNA   \n",
       "20  ENST00000488147.1  ENSE00001890219.1              unprocessed_pseudogene   \n",
       "21  ENST00000488147.1  ENSE00003507205.1              unprocessed_pseudogene   \n",
       "22  ENST00000488147.1  ENSE00003477500.1              unprocessed_pseudogene   \n",
       "23  ENST00000488147.1  ENSE00003565697.1              unprocessed_pseudogene   \n",
       "24  ENST00000488147.1  ENSE00003475637.1              unprocessed_pseudogene   \n",
       "25  ENST00000488147.1  ENSE00003502542.1              unprocessed_pseudogene   \n",
       "26  ENST00000488147.1  ENSE00003553898.1              unprocessed_pseudogene   \n",
       "27  ENST00000488147.1  ENSE00003621279.1              unprocessed_pseudogene   \n",
       "28  ENST00000488147.1  ENSE00002030414.1              unprocessed_pseudogene   \n",
       "29  ENST00000488147.1  ENSE00001935574.1              unprocessed_pseudogene   \n",
       "30  ENST00000492842.2  ENSE00001830178.2  transcribed_unprocessed_pseudogene   \n",
       "31  ENST00000606857.1  ENSE00003698237.1              unprocessed_pseudogene   \n",
       "32  ENST00000607096.1  ENSE00003695741.1                               miRNA   \n",
       "33  ENST00000619216.1  ENSE00003746039.1                               miRNA   \n",
       "34  ENST00000641515.2  ENSE00003812156.1                      protein_coding   \n",
       "35  ENST00000641515.2  ENSE00003813641.1                      protein_coding   \n",
       "36  ENST00000642116.1  ENSE00003812686.1                processed_transcript   \n",
       "37  ENST00000642116.1  ENSE00003812505.1                processed_transcript   \n",
       "\n",
       "   strand coordinates sequence  \n",
       "0       -       35720      GGT  \n",
       "1       -       35276      tgt  \n",
       "2       +       12058      GCA  \n",
       "3       +       12228      AGT  \n",
       "4       +       12698      TGT  \n",
       "5       +       13053      GGC  \n",
       "6       +       13375      AGT  \n",
       "7       +       12228      AGT  \n",
       "8       +       12722      GGT  \n",
       "9       -       35720      GGT  \n",
       "10      -      120774      GGT  \n",
       "11      -      112699      GGT  \n",
       "12      -       92090      ggt  \n",
       "13      +       30668      GGT  \n",
       "14      -      129054      CGT  \n",
       "15      +       30040      ggt  \n",
       "16      +       30668      GGT  \n",
       "17      -      129054      CGT  \n",
       "18      -      120720      TGT  \n",
       "19      -      112699      GGT  \n",
       "20      -       29533      CGT  \n",
       "21      -       24737      GGT  \n",
       "22      -       18267      GGT  \n",
       "23      -       17914      GGT  \n",
       "24      -       17605      TGT  \n",
       "25      -       17232      GGT  \n",
       "26      -       16857      GGT  \n",
       "27      -       16606      GGT  \n",
       "28      -       15795      GGT  \n",
       "29      -       15004      CGG  \n",
       "30      +       63888      ACA  \n",
       "31      +       53313      CGT  \n",
       "32      +       30504      ata  \n",
       "33      -       17368      GGT  \n",
       "34      +       65434      GGT  \n",
       "35      +       65574      GGT  \n",
       "36      +       57654      GGT  \n",
       "37      +       58857      TGT  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_donor_splice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "4899bfe3-b2ca-4823-8815-70627ba6dc19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>strand</th>\n",
       "      <th>coordinates</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>12178</td>\n",
       "      <td>GAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001758273.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>12612</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001799933.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>12974</td>\n",
       "      <td>AGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001746346.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>13220</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001863096.1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>13452</td>\n",
       "      <td>AGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>12612</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>13220</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSE00001922571.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>30563</td>\n",
       "      <td>AGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSE00001827679.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>30975</td>\n",
       "      <td>agt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003507205.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>24892</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003477500.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>18367</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003565697.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>18062</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003475637.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>17743</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003502542.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>17369</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003553898.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>17056</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003621279.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>16766</td>\n",
       "      <td>AGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00002030414.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>15948</td>\n",
       "      <td>AGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00001935574.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>15039</td>\n",
       "      <td>AGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00001843071.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>14502</td>\n",
       "      <td>GTG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id            exon_id                     transcript_type  \\\n",
       "0   ENST00000450305.2  ENSE00001671638.2  transcribed_unprocessed_pseudogene   \n",
       "1   ENST00000450305.2  ENSE00001758273.2  transcribed_unprocessed_pseudogene   \n",
       "2   ENST00000450305.2  ENSE00001799933.2  transcribed_unprocessed_pseudogene   \n",
       "3   ENST00000450305.2  ENSE00001746346.2  transcribed_unprocessed_pseudogene   \n",
       "4   ENST00000450305.2  ENSE00001863096.1  transcribed_unprocessed_pseudogene   \n",
       "5   ENST00000456328.2  ENSE00003582793.1                processed_transcript   \n",
       "6   ENST00000456328.2  ENSE00002312635.1                processed_transcript   \n",
       "7   ENST00000473358.1  ENSE00001922571.1                              lncRNA   \n",
       "8   ENST00000473358.1  ENSE00001827679.1                              lncRNA   \n",
       "9   ENST00000488147.1  ENSE00003507205.1              unprocessed_pseudogene   \n",
       "10  ENST00000488147.1  ENSE00003477500.1              unprocessed_pseudogene   \n",
       "11  ENST00000488147.1  ENSE00003565697.1              unprocessed_pseudogene   \n",
       "12  ENST00000488147.1  ENSE00003475637.1              unprocessed_pseudogene   \n",
       "13  ENST00000488147.1  ENSE00003502542.1              unprocessed_pseudogene   \n",
       "14  ENST00000488147.1  ENSE00003553898.1              unprocessed_pseudogene   \n",
       "15  ENST00000488147.1  ENSE00003621279.1              unprocessed_pseudogene   \n",
       "16  ENST00000488147.1  ENSE00002030414.1              unprocessed_pseudogene   \n",
       "17  ENST00000488147.1  ENSE00001935574.1              unprocessed_pseudogene   \n",
       "18  ENST00000488147.1  ENSE00001843071.1              unprocessed_pseudogene   \n",
       "\n",
       "   strand coordinates sequence  \n",
       "0       +       12178      GAT  \n",
       "1       +       12612      AGG  \n",
       "2       +       12974      AGC  \n",
       "3       +       13220      AGG  \n",
       "4       +       13452      AGT  \n",
       "5       +       12612      AGG  \n",
       "6       +       13220      AGG  \n",
       "7       +       30563      AGT  \n",
       "8       +       30975      agt  \n",
       "9       -       24892      AGG  \n",
       "10      -       18367      AGG  \n",
       "11      -       18062      AGG  \n",
       "12      -       17743      AGG  \n",
       "13      -       17369      AGG  \n",
       "14      -       17056      AGG  \n",
       "15      -       16766      AGA  \n",
       "16      -       15948      AGT  \n",
       "17      -       15039      AGT  \n",
       "18      -       14502      GTG  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_acceptor_splice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "297c6d07-7e21-44d7-89fb-5e31c54e4c05",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_promoter.to_csv(data_save_path +\"/\"+ entry['seqname']+\"_promoter.csv\", index=False)\n",
    "# df_core_promoter.to_csv(data_save_path +\"/\"+ entry['seqname']+\"_core_promoter_sequence.csv\", index=False)\n",
    "# df_acceptor_splice.to_csv(data_save_path +\"/\"+ entry['seqname']+\"_acceptor_splice_set_sequence.csv\", index=False)\n",
    "# df_donor_splice.to_csv(data_save_path +\"/\"+ entry['seqname']+\"_donor_splice_set.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "8e5827cd-0942-4812-a7de-5a880a371c34",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ttctccacaaagtgtttacttttggatttttgccagtctaacaggtgaAGccctggagattcttattagtgatttgggctggggcctggccatgtgtatttttttaaatttccactgatgattttgctgcatggccggtgttgagaatgactgCGCAAATTTGCCGGATTTCCTTTGCTGTTCCTGCATGTAGTTTAAACGAGATTGCCAGCACCGGGTATCATTCACCATTTTTCTTTTCGTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTATTTGCTGT'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_seq = pybedtools.BedTool.seq('chr1:11628-11927', fasta)\n",
    "test_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "a72c986b-6260-435e-92d2-495a2c5bbd4b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACAGCAAATACACATGAACAGAAAGAAGAGGTCAAAGAAAAGGCTGACGGCAAGTTAACGAAAAGAAAAATGGTGAATGATACCCGGTGCTGGCAATCTCGTTTAAACTACATGCAGGAACAGCAAAGGAAATCCGGCAAATTTGCGcagtcattctcaacaccggccatgcagcaaaatcatcagtggaaatttaaaaaaatacacatggccaggccccagcccaaatcactaataagaatctccagggCTtcacctgttagactggcaaaaatccaaaagtaaacactttgtggagaa\n"
     ]
    }
   ],
   "source": [
    "seq = Seq(test_seq)\n",
    "\n",
    "print (seq.reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8dcec4ae-0d9a-49af-b046-04c107e2a0a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">chr1:1-10\n",
      "GATGAGTCT\n",
      ">chr1:50-55\n",
      "CCATC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "a = pybedtools.BedTool('''\n",
    "chr1 1 10\n",
    "chr1 50 55''', from_string=True)\n",
    "fasta = pybedtools.example_filename('test.fa')\n",
    "a = a.sequence(fi=fasta)\n",
    "print(open(a.seqfn).read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6567b85-bbf4-4245-906b-be1af53a30ab",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28b13586-5eae-4aec-b002-d54f0c97ba5b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "558c0161-5b2a-40f7-ab3e-f59f19032130",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
