{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "48bb6eb5-6de0-4d27-848d-5c2e2eb29faf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, tarfile\n",
    "import pyBigWig\n",
    "import pybedtools\n",
    "import zipfile, gzip\n",
    "from gtfparse import read_gtf\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f4f24891-9cb6-4b3d-b75a-13c312c4bcf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_genome_path = \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/\"\n",
    "reference_genome_name = \"chr1.fa\"\n",
    "data_path= \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/gencode_annotation/\"\n",
    "file_name=\"gencodeV38.bb\"\n",
    "gencode_path= \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/gencode_annotation/\"\n",
    "gtf_annotation_filename= \"gencode.v38.annotation.gtf\"\n",
    "gff_annotation_filename= \"gencode.v38.annotation.gff3\"\n",
    "file_path= data_path+file_name\n",
    "gencode_gtf_file_path= gencode_path+ gtf_annotation_filename\n",
    "gencode_gff_file_path= gencode_path+ gff_annotation_filename\n",
    "reference_genome_file_path = reference_genome_path + reference_genome_name\n",
    "data_save_path= \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "096b8442-f1dd-47dc-abe6-2f836f7f7d31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/gencode_annotation/gencodeV38.bb'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8e286976-54a1-4993-99d4-30f7b4c4880e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'hgnc_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'ont', 'protein_id', 'ccdsid']\n"
     ]
    }
   ],
   "source": [
    "df = read_gtf(gencode_gtf_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bf9498db-5711-4b53-99c0-d8b8de788bcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand',\n",
      "       'frame', 'gene_id', 'gene_type', 'gene_name', 'level', 'hgnc_id',\n",
      "       'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name',\n",
      "       'transcript_support_level', 'tag', 'havana_transcript', 'exon_number',\n",
      "       'exon_id', 'ont', 'protein_id', 'ccdsid'],\n",
      "      dtype='object')\n",
      "(3150424, 25)\n"
     ]
    }
   ],
   "source": [
    "print(df.columns)\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "30f808e1-6f52-447e-ba2b-525ce83222b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010</td>\n",
       "      <td>12057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12697</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001758273.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12975</td>\n",
       "      <td>13052</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00001799933.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  source     feature  start    end  score strand frame  \\\n",
       "0    chr1  HAVANA        gene  11869  14409    NaN      +     0   \n",
       "1    chr1  HAVANA  transcript  11869  14409    NaN      +     0   \n",
       "2    chr1  HAVANA        exon  11869  12227    NaN      +     0   \n",
       "3    chr1  HAVANA        exon  12613  12721    NaN      +     0   \n",
       "4    chr1  HAVANA        exon  13221  14409    NaN      +     0   \n",
       "5    chr1  HAVANA  transcript  12010  13670    NaN      +     0   \n",
       "6    chr1  HAVANA        exon  12010  12057    NaN      +     0   \n",
       "7    chr1  HAVANA        exon  12179  12227    NaN      +     0   \n",
       "8    chr1  HAVANA        exon  12613  12697    NaN      +     0   \n",
       "9    chr1  HAVANA        exon  12975  13052    NaN      +     0   \n",
       "\n",
       "             gene_id                           gene_type gene_name level  \\\n",
       "0  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "1  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "2  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "3  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "4  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "5  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "6  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "7  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "8  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "9  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "\n",
       "      hgnc_id           havana_gene      transcript_id  \\\n",
       "0  HGNC:37102  OTTHUMG00000000961.2                      \n",
       "1  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "2  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "3  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "4  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "5  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "6  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "7  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "8  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "9  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "\n",
       "                      transcript_type transcript_name  \\\n",
       "0                                                       \n",
       "1                processed_transcript     DDX11L1-202   \n",
       "2                processed_transcript     DDX11L1-202   \n",
       "3                processed_transcript     DDX11L1-202   \n",
       "4                processed_transcript     DDX11L1-202   \n",
       "5  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "6  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "7  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "8  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "9  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "\n",
       "  transcript_support_level                      tag     havana_transcript  \\\n",
       "0                                                                           \n",
       "1                        1                    basic  OTTHUMT00000362751.1   \n",
       "2                        1                    basic  OTTHUMT00000362751.1   \n",
       "3                        1                    basic  OTTHUMT00000362751.1   \n",
       "4                        1                    basic  OTTHUMT00000362751.1   \n",
       "5                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "6                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "7                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "8                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "9                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "\n",
       "  exon_number            exon_id                      ont protein_id ccdsid  \n",
       "0                                                                            \n",
       "1                                                                            \n",
       "2           1  ENSE00002234944.1                                             \n",
       "3           2  ENSE00003582793.1                                             \n",
       "4           3  ENSE00002312635.1                                             \n",
       "5                                 PGO:0000005,PGO:0000019                    \n",
       "6           1  ENSE00001948541.1  PGO:0000005,PGO:0000019                    \n",
       "7           2  ENSE00001671638.2  PGO:0000005,PGO:0000019                    \n",
       "8           3  ENSE00001758273.2  PGO:0000005,PGO:0000019                    \n",
       "9           4  ENSE00001799933.2  PGO:0000005,PGO:0000019                    "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5c1381e3-3325-4b15-b1b8-80b27b8ac306",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "feature\n",
       "CDS                825028\n",
       "Selenocysteine        119\n",
       "UTR                349402\n",
       "exon              1499012\n",
       "gene                60649\n",
       "start_codon         92997\n",
       "stop_codon          86205\n",
       "transcript         237012\n",
       "dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby(['feature']).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "06945eba-fcdb-4f16-a5ec-fa94e936c8ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010</td>\n",
       "      <td>12057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12697</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001758273.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12975</td>\n",
       "      <td>13052</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00001799933.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>13374</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00001746346.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13453</td>\n",
       "      <td>13670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00001863096.1</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>29534</td>\n",
       "      <td>29570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001890219.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqname  source     feature  start    end  score strand frame  \\\n",
       "0     chr1  HAVANA        gene  11869  14409    NaN      +     0   \n",
       "1     chr1  HAVANA  transcript  11869  14409    NaN      +     0   \n",
       "2     chr1  HAVANA        exon  11869  12227    NaN      +     0   \n",
       "3     chr1  HAVANA        exon  12613  12721    NaN      +     0   \n",
       "4     chr1  HAVANA        exon  13221  14409    NaN      +     0   \n",
       "5     chr1  HAVANA  transcript  12010  13670    NaN      +     0   \n",
       "6     chr1  HAVANA        exon  12010  12057    NaN      +     0   \n",
       "7     chr1  HAVANA        exon  12179  12227    NaN      +     0   \n",
       "8     chr1  HAVANA        exon  12613  12697    NaN      +     0   \n",
       "9     chr1  HAVANA        exon  12975  13052    NaN      +     0   \n",
       "10    chr1  HAVANA        exon  13221  13374    NaN      +     0   \n",
       "11    chr1  HAVANA        exon  13453  13670    NaN      +     0   \n",
       "12    chr1  HAVANA        gene  14404  29570    NaN      -     0   \n",
       "13    chr1  HAVANA  transcript  14404  29570    NaN      -     0   \n",
       "14    chr1  HAVANA        exon  29534  29570    NaN      -     0   \n",
       "\n",
       "              gene_id                           gene_type gene_name level  \\\n",
       "0   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "1   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "2   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "3   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "4   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "5   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "6   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "7   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "8   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "9   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "10  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "11  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "12  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "13  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "14  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "\n",
       "       hgnc_id           havana_gene      transcript_id  \\\n",
       "0   HGNC:37102  OTTHUMG00000000961.2                      \n",
       "1   HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "2   HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "3   HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "4   HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "5   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "6   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "7   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "8   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "9   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "10  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "11  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "12  HGNC:38034  OTTHUMG00000000958.1                      \n",
       "13  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "14  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "\n",
       "                       transcript_type transcript_name  \\\n",
       "0                                                        \n",
       "1                 processed_transcript     DDX11L1-202   \n",
       "2                 processed_transcript     DDX11L1-202   \n",
       "3                 processed_transcript     DDX11L1-202   \n",
       "4                 processed_transcript     DDX11L1-202   \n",
       "5   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "6   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "7   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "8   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "9   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "10  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "11  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "12                                                       \n",
       "13              unprocessed_pseudogene      WASH7P-201   \n",
       "14              unprocessed_pseudogene      WASH7P-201   \n",
       "\n",
       "   transcript_support_level                      tag     havana_transcript  \\\n",
       "0                                                                            \n",
       "1                         1                    basic  OTTHUMT00000362751.1   \n",
       "2                         1                    basic  OTTHUMT00000362751.1   \n",
       "3                         1                    basic  OTTHUMT00000362751.1   \n",
       "4                         1                    basic  OTTHUMT00000362751.1   \n",
       "5                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "6                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "7                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "8                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "9                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "10                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "11                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "12                                                                           \n",
       "13                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "14                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "\n",
       "   exon_number            exon_id                      ont protein_id ccdsid  \n",
       "0                                                                             \n",
       "1                                                                             \n",
       "2            1  ENSE00002234944.1                                             \n",
       "3            2  ENSE00003582793.1                                             \n",
       "4            3  ENSE00002312635.1                                             \n",
       "5                                  PGO:0000005,PGO:0000019                    \n",
       "6            1  ENSE00001948541.1  PGO:0000005,PGO:0000019                    \n",
       "7            2  ENSE00001671638.2  PGO:0000005,PGO:0000019                    \n",
       "8            3  ENSE00001758273.2  PGO:0000005,PGO:0000019                    \n",
       "9            4  ENSE00001799933.2  PGO:0000005,PGO:0000019                    \n",
       "10           5  ENSE00001746346.2  PGO:0000005,PGO:0000019                    \n",
       "11           6  ENSE00001863096.1  PGO:0000005,PGO:0000019                    \n",
       "12                                                                            \n",
       "13                                             PGO:0000005                    \n",
       "14           1  ENSE00001890219.1              PGO:0000005                    "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_chr1 = df[df['seqname']=='chr1']\n",
    "df_chr1.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4ca223a6-8264-451d-9e50-9252755d1200",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010</td>\n",
       "      <td>12057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>8</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12697</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001758273.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>9</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12975</td>\n",
       "      <td>13052</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00001799933.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>10</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>13374</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00001746346.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>11</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13453</td>\n",
       "      <td>13670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00001863096.1</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>14</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>29534</td>\n",
       "      <td>29570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001890219.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index seqname  source feature  start    end  score strand frame  \\\n",
       "0      2    chr1  HAVANA    exon  11869  12227    NaN      +     0   \n",
       "1      3    chr1  HAVANA    exon  12613  12721    NaN      +     0   \n",
       "2      4    chr1  HAVANA    exon  13221  14409    NaN      +     0   \n",
       "3      6    chr1  HAVANA    exon  12010  12057    NaN      +     0   \n",
       "4      7    chr1  HAVANA    exon  12179  12227    NaN      +     0   \n",
       "5      8    chr1  HAVANA    exon  12613  12697    NaN      +     0   \n",
       "6      9    chr1  HAVANA    exon  12975  13052    NaN      +     0   \n",
       "7     10    chr1  HAVANA    exon  13221  13374    NaN      +     0   \n",
       "8     11    chr1  HAVANA    exon  13453  13670    NaN      +     0   \n",
       "9     14    chr1  HAVANA    exon  29534  29570    NaN      -     0   \n",
       "\n",
       "             gene_id                           gene_type gene_name level  \\\n",
       "0  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "1  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "2  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "3  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "4  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "5  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "6  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "7  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "8  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "9  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "\n",
       "      hgnc_id           havana_gene      transcript_id  \\\n",
       "0  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "1  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "2  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "3  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "4  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "5  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "6  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "7  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "8  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "9  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "\n",
       "                      transcript_type transcript_name  \\\n",
       "0                processed_transcript     DDX11L1-202   \n",
       "1                processed_transcript     DDX11L1-202   \n",
       "2                processed_transcript     DDX11L1-202   \n",
       "3  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "4  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "5  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "6  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "7  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "8  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "9              unprocessed_pseudogene      WASH7P-201   \n",
       "\n",
       "  transcript_support_level                      tag     havana_transcript  \\\n",
       "0                        1                    basic  OTTHUMT00000362751.1   \n",
       "1                        1                    basic  OTTHUMT00000362751.1   \n",
       "2                        1                    basic  OTTHUMT00000362751.1   \n",
       "3                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "4                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "5                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "6                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "7                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "8                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "9                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "\n",
       "  exon_number            exon_id                      ont protein_id ccdsid  \n",
       "0           1  ENSE00002234944.1                                             \n",
       "1           2  ENSE00003582793.1                                             \n",
       "2           3  ENSE00002312635.1                                             \n",
       "3           1  ENSE00001948541.1  PGO:0000005,PGO:0000019                    \n",
       "4           2  ENSE00001671638.2  PGO:0000005,PGO:0000019                    \n",
       "5           3  ENSE00001758273.2  PGO:0000005,PGO:0000019                    \n",
       "6           4  ENSE00001799933.2  PGO:0000005,PGO:0000019                    \n",
       "7           5  ENSE00001746346.2  PGO:0000005,PGO:0000019                    \n",
       "8           6  ENSE00001863096.1  PGO:0000005,PGO:0000019                    \n",
       "9           1  ENSE00001890219.1              PGO:0000005                    "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_chr1_transcript= df_chr1[df_chr1['feature']=='exon'].reset_index()\n",
    "df_chr1_transcript.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cd5bea5-1a81-4f50-925e-48deae690c73",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_by_transcript = df.groupby(\"state\")[\"last_name\"].count()\n",
    "n_by_state.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6045744b-f030-4aa6-aa35-98910e8fdb12",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "99de0a0b-20a5-4fd1-919f-89703dc8c677",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta = pybedtools.example_filename(reference_genome_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c32274e1-5f2e-4a43-9db4-cf2a78b3c044",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_TSS_set = pd.DataFrame(columns= ['exon_id','exon_number','transcript_id','transcript_type','sequence'])\n",
    "df_core_promoter_set = pd.DataFrame(columns= ['exon_id','exon_number','transcript_id','transcript_type','sequence'])\n",
    "df_donor_splice_set = pd.DataFrame(columns= ['exon_id','exon_number','transcript_id','transcript_type','sequence'])\n",
    "df_acceptor_splice_set = pd.DataFrame(columns= ['exon_id','exon_number','transcript_id','transcript_type','sequence'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7cb5d429-63f0-4ab9-b3e6-64a62441cc51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000450305.2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000456328.2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  seqname  source     feature  start    end  score strand  \\\n",
       "transcript_id                                                               \n",
       "                     chr1  HAVANA        gene  11869  14409    NaN      +   \n",
       "ENST00000450305.2    chr1  HAVANA  transcript  12010  13670    NaN      +   \n",
       "ENST00000456328.2    chr1  HAVANA  transcript  11869  14409    NaN      +   \n",
       "\n",
       "                  frame            gene_id  \\\n",
       "transcript_id                                \n",
       "                      0  ENSG00000223972.5   \n",
       "ENST00000450305.2     0  ENSG00000223972.5   \n",
       "ENST00000456328.2     0  ENSG00000223972.5   \n",
       "\n",
       "                                            gene_type gene_name level  \\\n",
       "transcript_id                                                           \n",
       "                   transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "ENST00000450305.2  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "ENST00000456328.2  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "\n",
       "                      hgnc_id           havana_gene  \\\n",
       "transcript_id                                         \n",
       "                   HGNC:37102  OTTHUMG00000000961.2   \n",
       "ENST00000450305.2  HGNC:37102  OTTHUMG00000000961.2   \n",
       "ENST00000456328.2  HGNC:37102  OTTHUMG00000000961.2   \n",
       "\n",
       "                                      transcript_type transcript_name  \\\n",
       "transcript_id                                                           \n",
       "                                                                        \n",
       "ENST00000450305.2  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "ENST00000456328.2                processed_transcript     DDX11L1-202   \n",
       "\n",
       "                  transcript_support_level                      tag  \\\n",
       "transcript_id                                                         \n",
       "                                                                      \n",
       "ENST00000450305.2                       NA  basic,Ensembl_canonical   \n",
       "ENST00000456328.2                        1                    basic   \n",
       "\n",
       "                      havana_transcript exon_number exon_id  \\\n",
       "transcript_id                                                 \n",
       "                                                              \n",
       "ENST00000450305.2  OTTHUMT00000002844.2                       \n",
       "ENST00000456328.2  OTTHUMT00000362751.1                       \n",
       "\n",
       "                                       ont protein_id ccdsid  \n",
       "transcript_id                                                 \n",
       "                                                              \n",
       "ENST00000450305.2  PGO:0000005,PGO:0000019                    \n",
       "ENST00000456328.2                                             "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7aff974-1410-4f95-bdb9-db269f03939c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35cb1e37-3b25-437e-9234-3a12b7991a78",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8d031d0-606b-4067-a822-053fed748b5b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "40fe2ce5-5b5e-41cc-a178-09a77c9c8358",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9c5f9184-a7db-4110-8fe7-b24acb8988c0",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 chr1:11869-14409\n",
      "1 chr1:11869-14409\n",
      "2 chr1:11869-12227\n",
      "3 chr1:12613-12721\n",
      "4 chr1:13221-14409\n",
      "5 chr1:12010-13670\n",
      "6 chr1:12010-12057\n",
      "7 chr1:12179-12227\n",
      "8 chr1:12613-12697\n",
      "9 chr1:12975-13052\n",
      "10 chr1:13221-13374\n",
      "11 chr1:13453-13670\n",
      "12 chr1:14404-29570\n",
      "13 chr1:14404-29570\n",
      "14 chr1:29534-29570\n",
      "15 chr1:24738-24891\n",
      "16 chr1:18268-18366\n",
      "17 chr1:17915-18061\n",
      "18 chr1:17606-17742\n",
      "19 chr1:17233-17368\n",
      "20 chr1:16858-17055\n",
      "21 chr1:16607-16765\n",
      "22 chr1:15796-15947\n",
      "23 chr1:15005-15038\n",
      "24 chr1:14404-14501\n",
      "25 chr1:17369-17436\n",
      "26 chr1:17369-17436\n",
      "27 chr1:17369-17436\n",
      "28 chr1:29554-31109\n",
      "29 chr1:29554-31097\n",
      "30 chr1:29554-30039\n",
      "31 chr1:30564-30667\n",
      "32 chr1:30976-31097\n",
      "33 chr1:30267-31109\n",
      "34 chr1:30267-30667\n",
      "35 chr1:30976-31109\n",
      "36 chr1:30366-30503\n",
      "37 chr1:30366-30503\n",
      "38 chr1:30366-30503\n",
      "39 chr1:34554-36081\n"
     ]
    }
   ],
   "source": [
    "df_transcript = pd.DataFrame(columns= ['transcript_id','gene_id','transcript_type','transcript_name','sequence'])\n",
    "df_exon = pd.DataFrame(columns= ['exon_id','exon_number','transcript_id','gene_id','transcript_type','gene_type','sequence'])\n",
    "df_gene = pd.DataFrame(columns= ['gene_id','gene_type','gene_name','sequence'])\n",
    "transcript_index = exon_index = gene_index = 0\n",
    "for index, row in df_chr1.head(40).iterrows():\n",
    "    #print (\"###\", index)\n",
    "    if (row['feature'] ==\"gene\"):\n",
    "        df_gene.loc[gene_index, ['gene_id']]= row['gene_id']\n",
    "        df_gene.loc[gene_index, ['gene_type']]= row['gene_type']\n",
    "        df_gene.loc[gene_index, ['gene_name']]= row['gene_name']\n",
    "        query= row['seqname']+':'+str(row['start'])+'-'+str(row['end'])\n",
    "        print (index, query)\n",
    "        df_gene.loc[gene_index, ['sequence']] = BedTool.seq(query, fasta)\n",
    "        gene_index= gene_index+1\n",
    "    elif (row['feature'] ==\"transcript\"):\n",
    "        df_transcript.loc[transcript_index, ['transcript_id']]= row['transcript_id']\n",
    "        df_transcript.loc[transcript_index, ['gene_id']]= row['gene_id']\n",
    "        df_transcript.loc[transcript_index, ['transcript_type']]= row['transcript_type']\n",
    "        df_transcript.loc[transcript_index, ['transcript_name']]= row['transcript_name']\n",
    "        query= row['seqname']+':'+str(row['start'])+'-'+str(row['end'])\n",
    "        print (index, query)\n",
    "        df_transcript.loc[transcript_index, ['sequence']] = BedTool.seq(query, fasta)\n",
    "        transcript_index= transcript_index+1\n",
    "    elif (row['feature'] ==\"exon\"):\n",
    "        df_exon.loc[exon_index, ['exon_id']]= row['exon_id']\n",
    "        df_exon.loc[exon_index, ['exon_number']]= row['exon_number']\n",
    "        df_exon.loc[exon_index, ['transcript_id']]= row['transcript_id']\n",
    "        df_exon.loc[exon_index, ['gene_id']]= row['gene_id']\n",
    "        df_exon.loc[exon_index, ['transcript_type']]= row['transcript_type']\n",
    "        df_exon.loc[exon_index, ['gene_type']]= row['gene_type']\n",
    "        query= row['seqname']+':'+str(row['start'])+'-'+str(row['end'])\n",
    "        print (index, query)\n",
    "        df_exon.loc[exon_index, ['sequence']] = BedTool.seq(query, fasta)\n",
    "        exon_index= exon_index+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "77a2f7e5-bffb-4796-84bd-c94e6d7a5d6e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>GTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>TTTCTGCTCAGTTCTTTATTGATTGGTGTGCCGTTTTCTCTGGAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>CTACAGAGGCGACATGGGGGTCAGGCAAGCTGACACCCGCTGTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>GTGCACACGGCTCCCATGCGTTGTCTTCCGAGCGTCAGGCCGCCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>ggatgcccagctagtttgaattttagataaacaacgaataatttcg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>agcagcaGGAGtgttttaattaaaagaaggcagttgctgtaaccaa...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_id                           gene_type    gene_name  \\\n",
       "0  ENSG00000223972.5  transcribed_unprocessed_pseudogene      DDX11L1   \n",
       "1  ENSG00000227232.5              unprocessed_pseudogene       WASH7P   \n",
       "2  ENSG00000278267.1                               miRNA    MIR6859-1   \n",
       "3  ENSG00000243485.5                              lncRNA  MIR1302-2HG   \n",
       "4  ENSG00000284332.1                               miRNA    MIR1302-2   \n",
       "5  ENSG00000237613.2                              lncRNA      FAM138A   \n",
       "\n",
       "                                            sequence  \n",
       "0  GTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCAT...  \n",
       "1  TTTCTGCTCAGTTCTTTATTGATTGGTGTGCCGTTTTCTCTGGAAG...  \n",
       "2  CTACAGAGGCGACATGGGGGTCAGGCAAGCTGACACCCGCTGTCCT...  \n",
       "3  GTGCACACGGCTCCCATGCGTTGTCTTCCGAGCGTCAGGCCGCCCC...  \n",
       "4  ggatgcccagctagtttgaattttagataaacaacgaataatttcg...  \n",
       "5  agcagcaGGAGtgttttaattaaaagaaggcagttgctgtaaccaa...  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4ef37fa4-28e6-42ca-be75-4e1c8bdccc32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>GTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>GTGTCTGACTTCCAGCAACTGCTGGCCTGTGCCAGGGTGCAAGCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>TTTCTGCTCAGTTCTTTATTGATTGGTGTGCCGTTTTCTCTGGAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1-201</td>\n",
       "      <td>CTACAGAGGCGACATGGGGGTCAGGCAAGCTGACACCCGCTGTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-202</td>\n",
       "      <td>GTGCACACGGCTCCCATGCGTTGTCTTCCGAGCGTCAGGCCGCCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-201</td>\n",
       "      <td>TCATCAGTCCAAAGTCCAGCAGTTGTCCCTCCTGGAATCCGTTGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000607096.1</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR1302-2-201</td>\n",
       "      <td>ggatgcccagctagtttgaattttagataaacaacgaataatttcg...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id            gene_id                     transcript_type  \\\n",
       "0  ENST00000456328.2  ENSG00000223972.5                processed_transcript   \n",
       "1  ENST00000450305.2  ENSG00000223972.5  transcribed_unprocessed_pseudogene   \n",
       "2  ENST00000488147.1  ENSG00000227232.5              unprocessed_pseudogene   \n",
       "3  ENST00000619216.1  ENSG00000278267.1                               miRNA   \n",
       "4  ENST00000473358.1  ENSG00000243485.5                              lncRNA   \n",
       "5  ENST00000469289.1  ENSG00000243485.5                              lncRNA   \n",
       "6  ENST00000607096.1  ENSG00000284332.1                               miRNA   \n",
       "\n",
       "   transcript_name                                           sequence  \n",
       "0      DDX11L1-202  GTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCAT...  \n",
       "1      DDX11L1-201  GTGTCTGACTTCCAGCAACTGCTGGCCTGTGCCAGGGTGCAAGCTG...  \n",
       "2       WASH7P-201  TTTCTGCTCAGTTCTTTATTGATTGGTGTGCCGTTTTCTCTGGAAG...  \n",
       "3    MIR6859-1-201  CTACAGAGGCGACATGGGGGTCAGGCAAGCTGACACCCGCTGTCCT...  \n",
       "4  MIR1302-2HG-202  GTGCACACGGCTCCCATGCGTTGTCTTCCGAGCGTCAGGCCGCCCC...  \n",
       "5  MIR1302-2HG-201  TCATCAGTCCAAAGTCCAGCAGTTGTCCCTCCTGGAATCCGTTGGC...  \n",
       "6    MIR1302-2-201  ggatgcccagctagtttgaattttagataaacaacgaataatttcg...  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ba3815fb-d278-43df-a1cf-2a5c259682af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>GTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>GTGTGTGGTGATGCCAGGCATGCCCTTCCCCAGCATCAGGTCTCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>GCAGGGCCATCAGGCACCAAAGGGATTCTGCCAGCATAGTGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>GTGTCTGACTTCCAGCAACTGCTGGCCTGTGCCAGGGTGCAAGCTGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>TTGGAGGAAAGATGAGTGAGAGCATCAACTTCTCTCACAACCTAGGCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSE00001758273.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>GTGTGTGGTGATGCCAGGCATGCCCTTCCCCAGCATCAGGTCTCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSE00001799933.2</td>\n",
       "      <td>4</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>CTCAGAGCCCAGGCCAGGGGCCCCCAAGAAAGGCTCTGGTGGAGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSE00001746346.2</td>\n",
       "      <td>5</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>GCAGGGCCATCAGGCACCAAAGGGATTCTGCCAGCATAGTGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSE00001863096.1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>TTTCACCAGAAGTAGGCCTCTTCCTGACAGGCAGCTGCACCACTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSE00001890219.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>GCTCACTCCGAGCTCCCGACGTGCACACGGCTCCCAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSE00003507205.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>CTGCTGAAGATGTCTCCAGAGACCTTCTGCAGGTACTGCAGGGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSE00003477500.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>CTTGATGGCCTTCTTGCTGCCCTTGATCTTCTCAATCTTGGCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSE00003565697.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>CTGCAGGGCCCGCTCGTCCAGGGGGCGGTGCTTGCTCTGGATCCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSE00003475637.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>AGGTTCTCGGTGGTGTTGAAGAGCAGCAAGGAGCTGACAGAGCTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSE00003502542.1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>CTGCTGTTCCAGCTGCTCTCTCTTGCTGATGGACAAGGGGGCATCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSE00003553898.1</td>\n",
       "      <td>7</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>CTTGAGAGGCTCGGCTACCTCAGTGTGGAAGGTGGGCAGTTCTGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENSE00003621279.1</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>CTGAAGGAGACGCGCTGCTGCTGCTGTCGTCCTGCCTGGCGCCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENSE00002030414.1</td>\n",
       "      <td>9</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>CTtgctcctgctccttctgctgctgcttctccagctttcgctcctt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSE00001935574.1</td>\n",
       "      <td>10</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>GACATCAAGTGCCCACCTTGGCTCGTGGCTCTCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSE00001843071.1</td>\n",
       "      <td>11</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>TTTCTGCTCAGTTCTTTATTGATTGGTGTGCCGTTTTCTCTGGAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSE00003746039.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>CTACAGAGGCGACATGGGGGTCAGGCAAGCTGACACCCGCTGTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENSE00001947070.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>GTGCACACGGCTCCCATGCGTTGTCTTCCGAGCGTCAGGCCGCCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENSE00001922571.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TTTTGTCCGCCTTCCCTGCCTCCTCTTCTGGGGGAGTTAGATCGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENSE00001827679.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>tgaggatgcgaagagaaggtggctgtctgcaaaccaggaagagagc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENSE00001841699.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TCATCAGTCCAAAGTCCAGCAGTTGTCCCTCCTGGAATCCGTTGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENSE00001890064.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>tgaggatgcgaagagaaggtggctgtctgcaaaccaggaagagagc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENSE00003695741.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000607096.1</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>ggatgcccagctagtttgaattttagataaacaacgaataatttcg...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              exon_id exon_number      transcript_id            gene_id  \\\n",
       "0   ENSE00002234944.1           1  ENST00000456328.2  ENSG00000223972.5   \n",
       "1   ENSE00003582793.1           2  ENST00000456328.2  ENSG00000223972.5   \n",
       "2   ENSE00002312635.1           3  ENST00000456328.2  ENSG00000223972.5   \n",
       "3   ENSE00001948541.1           1  ENST00000450305.2  ENSG00000223972.5   \n",
       "4   ENSE00001671638.2           2  ENST00000450305.2  ENSG00000223972.5   \n",
       "5   ENSE00001758273.2           3  ENST00000450305.2  ENSG00000223972.5   \n",
       "6   ENSE00001799933.2           4  ENST00000450305.2  ENSG00000223972.5   \n",
       "7   ENSE00001746346.2           5  ENST00000450305.2  ENSG00000223972.5   \n",
       "8   ENSE00001863096.1           6  ENST00000450305.2  ENSG00000223972.5   \n",
       "9   ENSE00001890219.1           1  ENST00000488147.1  ENSG00000227232.5   \n",
       "10  ENSE00003507205.1           2  ENST00000488147.1  ENSG00000227232.5   \n",
       "11  ENSE00003477500.1           3  ENST00000488147.1  ENSG00000227232.5   \n",
       "12  ENSE00003565697.1           4  ENST00000488147.1  ENSG00000227232.5   \n",
       "13  ENSE00003475637.1           5  ENST00000488147.1  ENSG00000227232.5   \n",
       "14  ENSE00003502542.1           6  ENST00000488147.1  ENSG00000227232.5   \n",
       "15  ENSE00003553898.1           7  ENST00000488147.1  ENSG00000227232.5   \n",
       "16  ENSE00003621279.1           8  ENST00000488147.1  ENSG00000227232.5   \n",
       "17  ENSE00002030414.1           9  ENST00000488147.1  ENSG00000227232.5   \n",
       "18  ENSE00001935574.1          10  ENST00000488147.1  ENSG00000227232.5   \n",
       "19  ENSE00001843071.1          11  ENST00000488147.1  ENSG00000227232.5   \n",
       "20  ENSE00003746039.1           1  ENST00000619216.1  ENSG00000278267.1   \n",
       "21  ENSE00001947070.1           1  ENST00000473358.1  ENSG00000243485.5   \n",
       "22  ENSE00001922571.1           2  ENST00000473358.1  ENSG00000243485.5   \n",
       "23  ENSE00001827679.1           3  ENST00000473358.1  ENSG00000243485.5   \n",
       "24  ENSE00001841699.1           1  ENST00000469289.1  ENSG00000243485.5   \n",
       "25  ENSE00001890064.1           2  ENST00000469289.1  ENSG00000243485.5   \n",
       "26  ENSE00003695741.1           1  ENST00000607096.1  ENSG00000284332.1   \n",
       "\n",
       "                       transcript_type                           gene_type  \\\n",
       "0                 processed_transcript  transcribed_unprocessed_pseudogene   \n",
       "1                 processed_transcript  transcribed_unprocessed_pseudogene   \n",
       "2                 processed_transcript  transcribed_unprocessed_pseudogene   \n",
       "3   transcribed_unprocessed_pseudogene  transcribed_unprocessed_pseudogene   \n",
       "4   transcribed_unprocessed_pseudogene  transcribed_unprocessed_pseudogene   \n",
       "5   transcribed_unprocessed_pseudogene  transcribed_unprocessed_pseudogene   \n",
       "6   transcribed_unprocessed_pseudogene  transcribed_unprocessed_pseudogene   \n",
       "7   transcribed_unprocessed_pseudogene  transcribed_unprocessed_pseudogene   \n",
       "8   transcribed_unprocessed_pseudogene  transcribed_unprocessed_pseudogene   \n",
       "9               unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "10              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "11              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "12              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "13              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "14              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "15              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "16              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "17              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "18              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "19              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "20                               miRNA                               miRNA   \n",
       "21                              lncRNA                              lncRNA   \n",
       "22                              lncRNA                              lncRNA   \n",
       "23                              lncRNA                              lncRNA   \n",
       "24                              lncRNA                              lncRNA   \n",
       "25                              lncRNA                              lncRNA   \n",
       "26                               miRNA                               miRNA   \n",
       "\n",
       "                                             sequence  \n",
       "0   GTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCAT...  \n",
       "1   GTGTGTGGTGATGCCAGGCATGCCCTTCCCCAGCATCAGGTCTCCA...  \n",
       "2   GCAGGGCCATCAGGCACCAAAGGGATTCTGCCAGCATAGTGCTCCT...  \n",
       "3    GTGTCTGACTTCCAGCAACTGCTGGCCTGTGCCAGGGTGCAAGCTGAG  \n",
       "4   TTGGAGGAAAGATGAGTGAGAGCATCAACTTCTCTCACAACCTAGGCCA  \n",
       "5   GTGTGTGGTGATGCCAGGCATGCCCTTCCCCAGCATCAGGTCTCCA...  \n",
       "6   CTCAGAGCCCAGGCCAGGGGCCCCCAAGAAAGGCTCTGGTGGAGAA...  \n",
       "7   GCAGGGCCATCAGGCACCAAAGGGATTCTGCCAGCATAGTGCTCCT...  \n",
       "8   TTTCACCAGAAGTAGGCCTCTTCCTGACAGGCAGCTGCACCACTGC...  \n",
       "9               GCTCACTCCGAGCTCCCGACGTGCACACGGCTCCCAT  \n",
       "10  CTGCTGAAGATGTCTCCAGAGACCTTCTGCAGGTACTGCAGGGCAT...  \n",
       "11  CTTGATGGCCTTCTTGCTGCCCTTGATCTTCTCAATCTTGGCCTGG...  \n",
       "12  CTGCAGGGCCCGCTCGTCCAGGGGGCGGTGCTTGCTCTGGATCCTG...  \n",
       "13  AGGTTCTCGGTGGTGTTGAAGAGCAGCAAGGAGCTGACAGAGCTGA...  \n",
       "14  CTGCTGTTCCAGCTGCTCTCTCTTGCTGATGGACAAGGGGGCATCA...  \n",
       "15  CTTGAGAGGCTCGGCTACCTCAGTGTGGAAGGTGGGCAGTTCTGGA...  \n",
       "16  CTGAAGGAGACGCGCTGCTGCTGCTGTCGTCCTGCCTGGCGCCTTG...  \n",
       "17  CTtgctcctgctccttctgctgctgcttctccagctttcgctcctt...  \n",
       "18                 GACATCAAGTGCCCACCTTGGCTCGTGGCTCTCA  \n",
       "19  TTTCTGCTCAGTTCTTTATTGATTGGTGTGCCGTTTTCTCTGGAAG...  \n",
       "20  CTACAGAGGCGACATGGGGGTCAGGCAAGCTGACACCCGCTGTCCT...  \n",
       "21  GTGCACACGGCTCCCATGCGTTGTCTTCCGAGCGTCAGGCCGCCCC...  \n",
       "22  TTTTGTCCGCCTTCCCTGCCTCCTCTTCTGGGGGAGTTAGATCGAG...  \n",
       "23  tgaggatgcgaagagaaggtggctgtctgcaaaccaggaagagagc...  \n",
       "24  TCATCAGTCCAAAGTCCAGCAGTTGTCCCTCCTGGAATCCGTTGGC...  \n",
       "25  tgaggatgcgaagagaaggtggctgtctgcaaaccaggaagagagc...  \n",
       "26  ggatgcccagctagtttgaattttagataaacaacgaataatttcg...  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "aacbcd0d-bb32-4b8d-81e9-bba99ac9e342",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = pybedtools.BedTool(gencode_gff_file_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "5ad8d2ef-d97d-4972-8c3c-98b604e90224",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\tHAVANA\tgene\t11869\t14409\t.\t+\t.\tID=ENSG00000223972.5;gene_id=ENSG00000223972.5;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;level=2;hgnc_id=HGNC:37102;havana_gene=OTTHUMG00000000961.2\n",
      "\n",
      "chr1\tHAVANA\ttranscript\t11869\t14409\t.\t+\t.\tID=ENST00000456328.2;Parent=ENSG00000223972.5;gene_id=ENSG00000223972.5;transcript_id=ENST00000456328.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=processed_transcript;transcript_name=DDX11L1-202;level=2;transcript_support_level=1;hgnc_id=HGNC:37102;tag=basic;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000362751.1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for i in a[0:2]:\n",
    "    print (i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "85f2fd7b-6872-45c2-916b-b9f2e42e4809",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTATTTGCTGTCTCTTAGCCCAGACTTCCCGTGTCCTTTCCACCGGGCCTTTGAGAGGTCACAGGGTCTTGATGCTGTGGTCTTCATCTGCAGGTGTCTGACTTCCAGCAACTGCTGGCCTGTGCCAGGGTGCAAGCTGAGCACTGGAGTGGAGTTTTCCTGTGGAGAGGAGCCATGCCTAGAGTGGGATGGGCCATTGTTCATCTTCTGGCCCCTGTTGTCTGCATGTAACTTAATACCACAACCAGGCATAGGGGAAAGATTGGAGGAAAGATGAGTGAGAGCATCAACTTCTCTCACAACCTAGGCCAGTAAGTAGTGCTTGTGCTCATCTCCTTGGCTGTGATACGTGGCCGGCCCTCGCTCCAGCAGCTGGACCCCTACCTGCCGTCTGCTGCCATCGGAGCCCAAAGCCGGGCTGTGACTGCTCAGACCAGCCGGCTGGAGGGAGGGGCTCAGCAGGTCTGGCTTTGGCCCTGGGAGAGCAGGTGGAAGATCAGGCAGGCCATCGCTGCCACAGAACCCAGTGGATTGGCCTAGGTGGGATCTCTGAGCTCAACAAGCCCTCTCTGGGTGGTAGGTGCAGAGACGGGAGGGGCAGAGCCGCAGGCACAGCCAAGAGGGCTGAAGAAATGGTAGAACGGAGCAGCTGGTGATGTGTGGGCCCACCGGCCCCAGGCTCCTGTCTCCCCCCAGGTGTGTGGTGATGCCAGGCATGCCCTTCCCCAGCATCAGGTCTCCAGAGCTGCAGAAGACGACGGCCGACTTGGATCACACTCTTGTGAGTGTCCCCAGTGTTGCAGAGGTGAGAGGAGAGTAGACAGTGAGTGGGAGTGGCGTCGCCCCTAGGGCTCTACGGGGCCGGCGTCTCCTGTCTCCTGGAGAGGCTTCGATGCCCCTCCACACCCTCTTGATCTTCCCTGTGATGTCATCTGGAGCCCTGCTGCTTGCGGTGGCCTATAAAGCCTCCTAGTCTGGCTCCAAGGCCTGGCAGAGTCTTTCCCAGGGAAAGCTACAAGCAGCAAACAGTCTGCATGGGTCATCCCCTTCACTCCCAGCTCAGAGCCCAGGCCAGGGGCCCCCAAGAAAGGCTCTGGTGGAGAACCTGTGCATGAAGGCTGTCAACCAGTCCATAGGCAAGCCTGGCTGCCTCCAGCTGGGTCGACAGACAGGGGCTGGAGAAGGGGAGAAGAGGAAAGTGAGGTTGCCTGCCCTGTCTCCTACCTGAGGCTGAGGAAGGAGAAGGGGATGCACTGTTGGGGAGGCAGCTGTAACTCAAAGCCTTAGCCTCTGTTCCCACGAAGGCAGGGCCATCAGGCACCAAAGGGATTCTGCCAGCATAGTGCTCCTGGACCAGTGATACACCCGGCACCCTGTCCTGGACACGCTGTTGGCCTGGATCTGAGCCCTGGTGGAGGTCAAAGCCACCTTTGGTTCTGCCATTGCTGCTGTGTGGAAGTTCACTCCTGCCTTTTCCTTTCCCTAGAGCCTCCACCACCCCGAGATCACATTTCTCACTGCCTTTTGTCTGCCCAGTTTCACCAGAAGTAGGCCTCTTCCTGACAGGCAGCTGCACCACTGCCTGGCGCTGTGCCCTTCCTTTGCTCTGCCCGCTGGAGACGGTGTTTGTCATGGGCCTGGTCTGCAGGGATCCTGCTACAAAGGTGAAACCCAGGAGAGTGTGGAGTCCAGAGTGTTGCCAGGACCCAGGCACAGGCATTAGTGCCCGTTGGAGAAAACAGGGGAATCCCGAAGAAATGGTGGGTCCTGGCCATCCGTGAGATCTTCCCAGGGCAGCTCCCCTCTGTGGAATCCAATCTGTCTTCCATCCTGCGTGGCCGAGGGCCAGGCTTCTCACTGGGCCTCTGCAGGAGGCTGCCATTTGTCCTGCCCACCTTCTTAGAAGCGAGACGGAGCAGACCCATCTGCTACTGCCCTTTCTATAATAACTAAAGTTAGCTGCCCTGGACTATTCACCCCCTAGTCTCAATTTAAGAAGATCCCCATGGCCACAGGGCCCCTGCCTGGGGGCTTGTCACCTCCCCCACCTTCTTCCTGAGTCATTCCTGCAGCCTTGCTCCCTAACCTGCCCCACAGCCTTGCCTGGATTTCTATCTCCCTGGCTTGGTGCCAGTTCCTCCAAGTCGATGGCACCTCCCTCCCTCTCAACCACTTGAGCAAACTCCAAGACATCTTCTACCCCAACACCAGCAATTGTGCCAAGGGCCATTAGGCTCTCAGCATGACTATTTTTAGAGACCCCGTGTCTGTCACTGAAACCTTTTTTGTGGGAGACTATTCCTCCCATCTGCAACAGCTGCCCCTGCTGACTGCCCTTCTCTCCTCCCTCTCATCCCAGAGAAACAGGTCAGCTGGGAGCTTCTGCCCCCACTGCCTAGGGACCAACAGGGGCAGGAGGCAGTCACTGACCCCGAGACGTTTGCATCCTGCACAGCTAGAGATCCTTTATTAAAAGCACACTGTTGGTTTCTG'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BedTool.seq('chr1:11869-14409', fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "297c6d07-7e21-44d7-89fb-5e31c54e4c05",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_exon.to_csv(data_save_path +\"/\"+ row['seqname']+\"_exon_sequence.csv\", index=False)\n",
    "df_transcript.to_csv(data_save_path +\"/\"+ row['seqname']+\"_transcript_sequence.csv\", index=False)\n",
    "df_gene.to_csv(data_save_path +\"/\"+ row['seqname']+\"_gene_sequence.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e5827cd-0942-4812-a7de-5a880a371c34",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8dcec4ae-0d9a-49af-b046-04c107e2a0a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">chr1:1-10\n",
      "GATGAGTCT\n",
      ">chr1:50-55\n",
      "CCATC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "a = pybedtools.BedTool('''\n",
    "chr1 1 10\n",
    "chr1 50 55''', from_string=True)\n",
    "fasta = pybedtools.example_filename('test.fa')\n",
    "a = a.sequence(fi=fasta)\n",
    "print(open(a.seqfn).read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d6567b85-bbf4-4245-906b-be1af53a30ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "b = a.save_seqs('example.fa')\n",
    "assert open(b.fn).read() == open(a.fn).read()\n",
    "if os.path.exists('example.fa'):\n",
    "    os.unlink('example.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "28b13586-5eae-4aec-b002-d54f0c97ba5b",
   "metadata": {},
   "outputs": [
    {
     "ename": "SystemError",
     "evalue": "<method 'chroms' of 'pyBigWig.bigWigFile' objects> returned a result with an error set",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;31mTypeError\u001b[0m: argument 1 must be str, not int",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mSystemError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_1463639/564859714.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mbb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchroms\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mSystemError\u001b[0m: <method 'chroms' of 'pyBigWig.bigWigFile' objects> returned a result with an error set"
     ]
    }
   ],
   "source": [
    "bb.chroms(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "558c0161-5b2a-40f7-ab3e-f59f19032130",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
