{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "48bb6eb5-6de0-4d27-848d-5c2e2eb29faf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, tarfile\n",
    "import pyBigWig\n",
    "import pybedtools\n",
    "import zipfile, gzip\n",
    "from gtfparse import read_gtf\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None)\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f4f24891-9cb6-4b3d-b75a-13c312c4bcf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_genome_path = \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/\"\n",
    "data_path= \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/gencode_annotation/\"\n",
    "file_name=\"gencodeV38.bb\"\n",
    "gencode_path= \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/gencode_annotation/\"\n",
    "gtf_annotation_filename= \"gencode.v38.annotation.gtf\"\n",
    "gff_annotation_filename= \"gencode.v38.annotation.gff3\"\n",
    "file_path= data_path+file_name\n",
    "gencode_gtf_file_path= gencode_path+ gtf_annotation_filename\n",
    "gencode_gff_file_path= gencode_path+ gff_annotation_filename\n",
    "#reference_genome_file_path = reference_genome_path + reference_genome_name\n",
    "data_save_path= \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "89017f10-1f85-49c8-944d-2fabe5273a4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['chr21.fa.fai',\n",
       " 'chr3.fa.fai',\n",
       " 'chr5.fa.fai',\n",
       " 'chr6.fa.fai',\n",
       " 'chr8.fa.fai',\n",
       " 'chr9.fa.fai',\n",
       " 'chrY.fa.fai',\n",
       " '.DS_Store',\n",
       " 'chr1.fa',\n",
       " 'chr10.fa',\n",
       " 'chr11.fa',\n",
       " 'chr12.fa',\n",
       " 'chr13.fa',\n",
       " 'chr14.fa',\n",
       " 'chr15.fa',\n",
       " 'chr16.fa',\n",
       " 'chr17.fa',\n",
       " 'chr18.fa',\n",
       " 'chr19.fa',\n",
       " 'chr2.fa',\n",
       " 'chr20.fa',\n",
       " 'chr21.fa',\n",
       " 'chr22.fa',\n",
       " 'chr3.fa',\n",
       " 'chr4.fa',\n",
       " 'chr5.fa',\n",
       " 'chr6.fa',\n",
       " 'chr7.fa',\n",
       " 'chr8.fa',\n",
       " 'chr9.fa',\n",
       " 'chrX.fa',\n",
       " 'chrY.fa',\n",
       " 'keep',\n",
       " 'chr10.fa.fai',\n",
       " 'chr11.fa.fai',\n",
       " 'chr12.fa.fai',\n",
       " 'chr13.fa.fai',\n",
       " 'chr14.fa.fai',\n",
       " 'chr15.fa.fai',\n",
       " 'chr16.fa.fai',\n",
       " 'chr17.fa.fai',\n",
       " 'chr18.fa.fai',\n",
       " 'chr19.fa.fai',\n",
       " 'chr2.fa.fai',\n",
       " 'chr20.fa.fai',\n",
       " 'chr22.fa.fai',\n",
       " 'chr4.fa.fai',\n",
       " 'chr7.fa.fai',\n",
       " 'chrX.fa.fai',\n",
       " '._.DS_Store',\n",
       " 'chr1.fa.fai']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr_files= os.listdir(reference_genome_path )\n",
    "chr_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8e286976-54a1-4993-99d4-30f7b4c4880e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'hgnc_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'ont', 'protein_id', 'ccdsid']\n"
     ]
    }
   ],
   "source": [
    "df = read_gtf(gencode_gtf_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bf9498db-5711-4b53-99c0-d8b8de788bcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand',\n",
      "       'frame', 'gene_id', 'gene_type', 'gene_name', 'level', 'hgnc_id',\n",
      "       'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name',\n",
      "       'transcript_support_level', 'tag', 'havana_transcript', 'exon_number',\n",
      "       'exon_id', 'ont', 'protein_id', 'ccdsid'],\n",
      "      dtype='object')\n",
      "(3150424, 25)\n"
     ]
    }
   ],
   "source": [
    "print(df.columns)\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5c1381e3-3325-4b15-b1b8-80b27b8ac306",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "feature\n",
       "CDS                825028\n",
       "Selenocysteine        119\n",
       "UTR                349402\n",
       "exon              1499012\n",
       "gene                60649\n",
       "start_codon         92997\n",
       "stop_codon          86205\n",
       "transcript         237012\n",
       "dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby(['feature']).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d082499d-929b-4dfd-b132-2e8b2ddc5cf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "promoter_coordinates = [-249, 50]\n",
    "core_promoter_coordinates= [-34, 35]\n",
    "donor_coordinates= [-199, 200]\n",
    "acceptor_coordinates= [-199, 200]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1e0369d0-4941-4247-92e3-7553cac04e5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def non_acceptor_sequence(entry, df_acceptor_splice, acceptor_index):\n",
    "    print (\"@@@@\", acceptor_index)\n",
    "    df_acceptor_splice.loc[acceptor_index, ['transcript_id']]= entry['transcript_id']\n",
    "    df_acceptor_splice.loc[acceptor_index, ['exon_id']] = entry['exon_id']\n",
    "    df_acceptor_splice.loc[acceptor_index, ['transcript_type']] = entry['transcript_type']\n",
    "    df_acceptor_splice.loc[acceptor_index, ['strand']]= entry['strand']\n",
    "    if (entry['strand']=='+'):\n",
    "        df_acceptor_splice.loc[acceptor_index, ['coordinates']]= int(entry['start'])-1\n",
    "        acceptor_query= entry['seqname']+':'+str(int(entry['start'])+acceptor_coordinates[0])+'-'+str(int(entry['start'])+acceptor_coordinates[1])\n",
    "        print (\"##$$$\", acceptor_query)\n",
    "        df_acceptor_splice.loc[acceptor_index, ['sequence']] = pybedtools.BedTool.seq(acceptor_query, fasta)\n",
    "        print (\"###\", acceptor_index)\n",
    "    elif (entry['strand']=='-'):\n",
    "        df_acceptor_splice.loc[acceptor_index, ['coordinates']]= int(entry['end'])+1\n",
    "        acceptor_query= entry['seqname']+':'+str(int(entry['end'])+(acceptor_coordinates[1]*(-1)))+'-'+str(int(entry['end'])+(acceptor_coordinates[0]*(-1)))\n",
    "        temp_seq = pybedtools.BedTool.seq(acceptor_query, fasta)\n",
    "        rev_seq = str(Seq(temp_seq).reverse_complement())\n",
    "        df_acceptor_splice.loc[acceptor_index, ['sequence']] = rev_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ccc5d858-46ea-421c-b7c0-f856cedb7869",
   "metadata": {},
   "outputs": [],
   "source": [
    "def donor_sequence(entry, df_donor_splice, donor_index):\n",
    "    df_donor_splice.loc[donor_index, ['transcript_id']]= entry['transcript_id']\n",
    "    df_donor_splice.loc[donor_index, ['exon_id']] = entry['exon_id']\n",
    "    df_donor_splice.loc[donor_index, ['transcript_type']] = entry['transcript_type']\n",
    "    df_donor_splice.loc[donor_index, ['strand']]= entry['strand']\n",
    "    if (entry['strand']=='+'):\n",
    "        df_donor_splice.loc[donor_index, ['coordinates']] = int(entry['end'])+1\n",
    "        donor_query= entry['seqname']+':'+str(int(entry['end'])+donor_coordinates[0])+'-'+str(int(entry['end'])+donor_coordinates[1])\n",
    "        print (\"##$$$\", donor_query, donor_index)\n",
    "        df_donor_splice.loc[donor_index, ['sequence']] = pybedtools.BedTool.seq(donor_query, fasta)\n",
    "    elif (entry['strand']=='-'):\n",
    "        df_donor_splice.loc[donor_index, ['coordinates']] = int(entry['start'])-1\n",
    "        donor_query= entry['seqname']+':'+str(int(entry['start'])+(donor_coordinates[1]*(-1)))+'-'+str(int(entry['start'])+(donor_coordinates[0]*(-1)))\n",
    "        print (\"##$$$\", donor_query, donor_index)\n",
    "        temp_seq = pybedtools.BedTool.seq(donor_query, fasta)\n",
    "        rev_seq = str(Seq(temp_seq).reverse_complement())\n",
    "        df_donor_splice.loc[donor_index, ['sequence']] = rev_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2cbd9c22-f5b8-465f-932f-9aeb64f57a97",
   "metadata": {},
   "outputs": [],
   "source": [
    "def promoter_sequence(entry, df_promoter, promoter_index):\n",
    "    df_promoter.loc[promoter_index, ['gene_id']] = entry['gene_id']\n",
    "    df_promoter.loc[promoter_index, ['transcript_id']]= entry['transcript_id']\n",
    "    df_promoter.loc[promoter_index, ['transcript_type']]= entry['transcript_type']\n",
    "    df_promoter.loc[promoter_index, ['strand']] = entry['strand']\n",
    "    df_promoter.loc[promoter_index, ['exon_id']]=  entry['exon_id']\n",
    "    if (entry['strand']=='+'):\n",
    "        df_promoter.loc[promoter_index,['TSS']]= entry['start']\n",
    "        promoter_query= entry['seqname']+':'+str(int(entry['start'])+promoter_coordinates[0])+'-'+str(int(entry['start'])+promoter_coordinates[1])\n",
    "        df_promoter.loc[promoter_index, ['sequence']] = pybedtools.BedTool.seq(promoter_query, fasta)\n",
    "    elif(entry['strand']=='-'):\n",
    "        df_promoter.loc[promoter_index,['TSS']]= entry['end']\n",
    "        promoter_query= entry['seqname']+':'+str(int(entry['end'])+(promoter_coordinates[1]*(-1)))+'-'+str(int(entry['end'])+(promoter_coordinates[0]*(-1)))\n",
    "        temp_seq = pybedtools.BedTool.seq(promoter_query, fasta)\n",
    "        #print (\"Original\", temp_seq)\n",
    "        rev_seq = str(Seq(temp_seq).reverse_complement())\n",
    "        #print (rev_seq)\n",
    "        df_promoter.loc[promoter_index, ['sequence']] = rev_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "60c21aa4-2a16-4211-b305-bbfea865ebd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def core_promoter_sequence(entry, df_core_promoter, core_promoter_index):\n",
    "    df_core_promoter.loc[core_promoter_index, ['gene_id']] = entry['gene_id']\n",
    "    df_core_promoter.loc[core_promoter_index, ['transcript_id']] = entry['transcript_id']\n",
    "    df_core_promoter.loc[core_promoter_index, ['transcript_type']] = entry['transcript_type']\n",
    "    df_core_promoter.loc[core_promoter_index, ['strand']] = entry['strand']\n",
    "    df_core_promoter.loc[core_promoter_index, ['exon_id']]  =  entry['exon_id']\n",
    "    if (entry['strand']=='+'):\n",
    "        df_core_promoter.loc[core_promoter_index,['TSS']]= entry['start']\n",
    "        core_promoter_query= entry['seqname']+':'+str(int(entry['start'])+core_promoter_coordinates[0])+'-'+str(int(entry['start'])+core_promoter_coordinates[1])\n",
    "        df_core_promoter.loc[core_promoter_index, ['sequence']] = pybedtools.BedTool.seq(core_promoter_query, fasta)\n",
    "    elif(entry['strand']=='-'):\n",
    "        df_core_promoter.loc[core_promoter_index,['TSS']]= entry['end']\n",
    "        core_promoter_query= entry['seqname']+':'+str(int(entry['end'])+(core_promoter_coordinates[1]*(-1)))+'-'+str(int(entry['end'])+(core_promoter_coordinates[0]*(-1)))\n",
    "        temp_seq = pybedtools.BedTool.seq(core_promoter_query, fasta)\n",
    "        rev_seq = str(Seq(temp_seq).reverse_complement())\n",
    "        df_core_promoter.loc[core_promoter_index, ['sequence']] = rev_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cd5bea5-1a81-4f50-925e-48deae690c73",
   "metadata": {},
   "outputs": [],
   "source": [
    "global df_promoter, df_core_promoter, df_donor_splice, df_acceptor_splice\n",
    "\n",
    "\n",
    "for file in chr_files:\n",
    "    if file.endswith('.fa'):\n",
    "        reference_genome_file_path = reference_genome_path + file\n",
    "        chr_name = file.split('.')[0]\n",
    "        print (chr_name)\n",
    "        df_chr= df[df['seqname']==chr_name]\n",
    "        df_new = df_chr[df_chr['feature']=='exon'].reset_index(drop=True)\n",
    "        print (df_new.shape)\n",
    "        #print (reference_genome_file_path)\n",
    "        fasta = pybedtools.example_filename(reference_genome_file_path)\n",
    "        \n",
    "        \n",
    "        folder_name = data_save_path + chr_name   \n",
    "        print (folder_name)\n",
    "        if not os.path.exists(folder_name):\n",
    "            os.makedirs(folder_name)\n",
    "        \n",
    "        \n",
    "        #df_promoter = pd.DataFrame(columns= ['gene_id','transcript_id','strand','exon_id','transcript_type','TSS', 'sequence'])\n",
    "        #df_core_promoter = pd.DataFrame(columns= ['gene_id','transcript_id','strand','exon_id','transcript_type','TSS','sequence'])\n",
    "        df_donor_splice = pd.DataFrame(columns= ['transcript_id','exon_id','transcript_type','strand','coordinates','sequence'])\n",
    "        df_acceptor_splice= pd.DataFrame(columns= ['transcript_id','exon_id', 'transcript_type','strand','coordinates', 'sequence'])\n",
    "        promoter_index = core_promoter_index = donor_index = acceptor_index = 0\n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "        df_by_transcript = df_new.head(20).groupby(\"transcript_id\")\n",
    "        for transcript, entries in df_by_transcript:\n",
    "            print (\"\\n\", transcript,len( entries) )\n",
    "            print (entries.iloc[0]['strand'])\n",
    "            if (entries.iloc[0]['strand']=='+'):\n",
    "                for index, entry in entries.iterrows():\n",
    "                    if (int(entry['exon_number'])==1):\n",
    "                        print(\"TSS\", entry['start'])\n",
    "                        #promoter_sequence(entry, df_promoter, promoter_index)\n",
    "                        #promoter_index= promoter_index+1\n",
    "                        #core_promoter_sequence(entry, df_core_promoter, core_promoter_index)\n",
    "                        #core_promoter_index = core_promoter_index +1\n",
    "                        donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                        donor_index =donor_index+ 1\n",
    "                    elif (int(entry['exon_number'])==len(entries)):\n",
    "                        print (\"Last acceptor_splice_site\",  entry['start'])\n",
    "                        acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                        acceptor_index = acceptor_index +1\n",
    "                    else:\n",
    "                        print (\"Acceptor splice site\", entry['start'])\n",
    "                        acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                        acceptor_index = acceptor_index +1\n",
    "                        print (\"Donor splice site\", entry['end'])\n",
    "                        donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                        donor_index =donor_index+ 1\n",
    "            elif (entries.iloc[0]['strand']=='-'):\n",
    "                for index, entry in entries.iterrows():\n",
    "                    if (int(entry['exon_number'])==1):\n",
    "                        print(\"TSS\", entry['end'])\n",
    "                        print(\"first donor_splice_site\", entry['start'])\n",
    "                        #promoter_sequence(entry, df_promoter, promoter_index)\n",
    "                        #promoter_index= promoter_index+1\n",
    "                        #core_promoter_sequence(entry, df_core_promoter, core_promoter_index)\n",
    "                        #core_promoter_index = core_promoter_index +1\n",
    "                        donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                        donor_index =donor_index+ 1\n",
    "                    elif (int(entry['exon_number'])==len(entries)):\n",
    "                        print (\"Last acceptor_splice_site\",  entry['end'])\n",
    "                        acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                        acceptor_index = acceptor_index +1\n",
    "                    else:\n",
    "                        print (\"Acceptor splice site\", entry['end'])\n",
    "                        acceptor_sequence(entry, df_acceptor_splice, acceptor_index)\n",
    "                        acceptor_index = acceptor_index +1\n",
    "                        print (\"Donor splice site\", entry['start'])\n",
    "                        donor_sequence(entry, df_donor_splice, donor_index)\n",
    "                        donor_index =donor_index+ 1\n",
    "        df_promoter.to_csv(folder_name +\"/\"+ entry['seqname']+\"_promoter.csv\", index=False)\n",
    "        df_core_promoter.to_csv(folder_name +\"/\"+ entry['seqname']+\"_core_promoter_sequence.csv\", index=False)\n",
    "        df_acceptor_splice.to_csv(folder_name +\"/\"+ entry['seqname']+\"_acceptor_splice_set_sequence.csv\", index=False)\n",
    "        df_donor_splice.to_csv(folder_name +\"/\"+ entry['seqname']+\"_donor_splice_set.csv\", index=False)\n",
    "        break()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bd36b20-1acb-41cd-bbff-67f9dd997003",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "6045744b-f030-4aa6-aa35-98910e8fdb12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>TSS</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>ENST00000417324.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001656588.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>36081</td>\n",
       "      <td>AGAATTGGTTTGATTTACATAATCCTGCCCCTGAGATTTAGCCCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>12010</td>\n",
       "      <td>gccggtgttgagaatgactgCGCAAATTTGCCGGATTTCCTTTGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>atgcctgtttctccacaaagtgtttacttttggatttttgccagtc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>ENST00000461467.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001618781.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>36073</td>\n",
       "      <td>TTTGATTTACATAATCCTGCCCCTGAGATTTAGCCCCACCCCTGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENST00000466430.5</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001606755.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>120932</td>\n",
       "      <td>TCTTTACTTGCACTGACAAATATGAACGCTGGGGAATTTGGAAatg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001841699.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>30267</td>\n",
       "      <td>ttgggcagctcgtgttcaatgggtagagtttcaggctggggtgatg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENST00000471248.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001934975.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>129173</td>\n",
       "      <td>gactccatcttataaaaggaaaaaagaaagaaaagaaaaaTTCCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001947070.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>29554</td>\n",
       "      <td>CCGGACGCCTCCTCACCTGCGAGCCGCCCTCCCGGAAGCTCCCGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENST00000477740.5</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001919246.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>129217</td>\n",
       "      <td>agctgagatcgcaccagtgcactccagcctgggcaacagagtgaga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001890219.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>29570</td>\n",
       "      <td>GCTGAGAGACGCAGGAAAGCGGGTCAAggcgtagggctggagggca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSG00000240361.2</td>\n",
       "      <td>ENST00000492842.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001830178.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>62949</td>\n",
       "      <td>AGTTCATGTATTTGAGCAATTGAGAACATTTAAAGCAATAGACTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000268020.3</td>\n",
       "      <td>ENST00000606857.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003698237.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>52473</td>\n",
       "      <td>AATAAATAACTTATTATCACCGCATTTCCCCAAttcatttatcttt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>ENST00000607096.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003695741.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>30366</td>\n",
       "      <td>ttaatcccactgaactgtatgctgaaaaatggtttagacggtgaat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00003746039.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>17436</td>\n",
       "      <td>GCAGAAGAGGGACTTGGGGGTCTTCCCAGCAACATCAGCTCTGTCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003812156.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>65419</td>\n",
       "      <td>GCATTATTACTTACTCTCTTGTTAACCTATCTGGATTTTAATTTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSG00000240361.2</td>\n",
       "      <td>ENST00000642116.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003812686.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>57598</td>\n",
       "      <td>AATGTAGTACCTCAAATCCTTCCTTCCCTCCCCCTATTTCATCAGC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id      transcript_id strand            exon_id  \\\n",
       "0   ENSG00000237613.2  ENST00000417324.1      -  ENSE00001656588.1   \n",
       "1   ENSG00000223972.5  ENST00000450305.2      +  ENSE00001948541.1   \n",
       "2   ENSG00000223972.5  ENST00000456328.2      +  ENSE00002234944.1   \n",
       "3   ENSG00000237613.2  ENST00000461467.1      -  ENSE00001618781.2   \n",
       "4   ENSG00000238009.6  ENST00000466430.5      -  ENSE00001606755.2   \n",
       "5   ENSG00000243485.5  ENST00000469289.1      +  ENSE00001841699.1   \n",
       "6   ENSG00000238009.6  ENST00000471248.1      -  ENSE00001934975.1   \n",
       "7   ENSG00000243485.5  ENST00000473358.1      +  ENSE00001947070.1   \n",
       "8   ENSG00000238009.6  ENST00000477740.5      -  ENSE00001919246.1   \n",
       "9   ENSG00000227232.5  ENST00000488147.1      -  ENSE00001890219.1   \n",
       "10  ENSG00000240361.2  ENST00000492842.2      +  ENSE00001830178.2   \n",
       "11  ENSG00000268020.3  ENST00000606857.1      +  ENSE00003698237.1   \n",
       "12  ENSG00000284332.1  ENST00000607096.1      +  ENSE00003695741.1   \n",
       "13  ENSG00000278267.1  ENST00000619216.1      -  ENSE00003746039.1   \n",
       "14  ENSG00000186092.7  ENST00000641515.2      +  ENSE00003812156.1   \n",
       "15  ENSG00000240361.2  ENST00000642116.1      +  ENSE00003812686.1   \n",
       "\n",
       "                       transcript_type     TSS  \\\n",
       "0                               lncRNA   36081   \n",
       "1   transcribed_unprocessed_pseudogene   12010   \n",
       "2                 processed_transcript   11869   \n",
       "3                               lncRNA   36073   \n",
       "4                               lncRNA  120932   \n",
       "5                               lncRNA   30267   \n",
       "6                               lncRNA  129173   \n",
       "7                               lncRNA   29554   \n",
       "8                               lncRNA  129217   \n",
       "9               unprocessed_pseudogene   29570   \n",
       "10  transcribed_unprocessed_pseudogene   62949   \n",
       "11              unprocessed_pseudogene   52473   \n",
       "12                               miRNA   30366   \n",
       "13                               miRNA   17436   \n",
       "14                      protein_coding   65419   \n",
       "15                processed_transcript   57598   \n",
       "\n",
       "                                             sequence  \n",
       "0   AGAATTGGTTTGATTTACATAATCCTGCCCCTGAGATTTAGCCCCA...  \n",
       "1   gccggtgttgagaatgactgCGCAAATTTGCCGGATTTCCTTTGCT...  \n",
       "2   atgcctgtttctccacaaagtgtttacttttggatttttgccagtc...  \n",
       "3   TTTGATTTACATAATCCTGCCCCTGAGATTTAGCCCCACCCCTGAG...  \n",
       "4   TCTTTACTTGCACTGACAAATATGAACGCTGGGGAATTTGGAAatg...  \n",
       "5   ttgggcagctcgtgttcaatgggtagagtttcaggctggggtgatg...  \n",
       "6   gactccatcttataaaaggaaaaaagaaagaaaagaaaaaTTCCAT...  \n",
       "7   CCGGACGCCTCCTCACCTGCGAGCCGCCCTCCCGGAAGCTCCCGCC...  \n",
       "8   agctgagatcgcaccagtgcactccagcctgggcaacagagtgaga...  \n",
       "9   GCTGAGAGACGCAGGAAAGCGGGTCAAggcgtagggctggagggca...  \n",
       "10  AGTTCATGTATTTGAGCAATTGAGAACATTTAAAGCAATAGACTGC...  \n",
       "11  AATAAATAACTTATTATCACCGCATTTCCCCAAttcatttatcttt...  \n",
       "12  ttaatcccactgaactgtatgctgaaaaatggtttagacggtgaat...  \n",
       "13  GCAGAAGAGGGACTTGGGGGTCTTCCCAGCAACATCAGCTCTGTCA...  \n",
       "14  GCATTATTACTTACTCTCTTGTTAACCTATCTGGATTTTAATTTTG...  \n",
       "15  AATGTAGTACCTCAAATCCTTCCTTCCCTCCCCCTATTTCATCAGC...  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "99de0a0b-20a5-4fd1-919f-89703dc8c677",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>TSS</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>ENST00000417324.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001656588.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>36081</td>\n",
       "      <td>CAAGGATTAGTCCTGCTGCCCTGTGCCCAGCTCCCACACAACGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>12010</td>\n",
       "      <td>CACAGGGTCTTGATGCTGTGGTCTTCATCTGCAGGTGTCTGACTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>CCAGCACCGGGTATCATTCACCATTTTTCTTTTCGTTAACTTGCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>ENST00000461467.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001618781.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>36073</td>\n",
       "      <td>AGTCCTGCTGCCCTGTGCCCAGCTCCCACACAACGGGGTTTCGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENST00000466430.5</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001606755.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>120932</td>\n",
       "      <td>TTTCCATGTGATTTTCCTCTTAATTTTTTTCTAGCTGATCCATATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001841699.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>30267</td>\n",
       "      <td>GCTTTTTTAAAAAGTTAAATATATAAAAAGAGCATCATCAGTCCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENST00000471248.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001934975.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>129173</td>\n",
       "      <td>AGCATCTTAAAATGATTCAACAGGAGGAGATAAGGAAGCTCGAGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001947070.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>29554</td>\n",
       "      <td>GGCACGGAACTCACGCTCACTCCGAGCTCCCGACGTGCACACGGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENST00000477740.5</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001919246.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>129217</td>\n",
       "      <td>GTGTCACAACTGCACATTGGTTTCACAGCTGCAGGACAAGTTCGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00001890219.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>29570</td>\n",
       "      <td>GGGTAGGGGCGGCCTGACGCTCGGAAGACAACGCATGGGAGCCGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSG00000240361.2</td>\n",
       "      <td>ENST00000492842.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00001830178.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>62949</td>\n",
       "      <td>GGTTGCTTCAGAGTCTTCCCTTCTATCTGATTCAGTGGACCAAGTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000268020.3</td>\n",
       "      <td>ENST00000606857.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003698237.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>52473</td>\n",
       "      <td>GAGTCAATGGCCAATGGATAGAGTAAATAATTCTGCGGTATCTAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>ENST00000607096.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003695741.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>30366</td>\n",
       "      <td>CCCTTGCCTTTtagggttgccagattaaaagacaggatgcccagct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSE00003746039.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>17436</td>\n",
       "      <td>TGGGACCCCACAGTGCACGCTGTGCCCCTGATGATGTGGGAGAGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003812156.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>65419</td>\n",
       "      <td>TTAGTCTCAGTACACAAAGCTCAGACTGGCTATTCCCAGATCTCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSG00000240361.2</td>\n",
       "      <td>ENST00000642116.1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSE00003812686.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>57598</td>\n",
       "      <td>CTTTGTCTTTGGGGTATTACATAAATGAAGATTAAGCTATCTGAAT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id      transcript_id strand            exon_id  \\\n",
       "0   ENSG00000237613.2  ENST00000417324.1      -  ENSE00001656588.1   \n",
       "1   ENSG00000223972.5  ENST00000450305.2      +  ENSE00001948541.1   \n",
       "2   ENSG00000223972.5  ENST00000456328.2      +  ENSE00002234944.1   \n",
       "3   ENSG00000237613.2  ENST00000461467.1      -  ENSE00001618781.2   \n",
       "4   ENSG00000238009.6  ENST00000466430.5      -  ENSE00001606755.2   \n",
       "5   ENSG00000243485.5  ENST00000469289.1      +  ENSE00001841699.1   \n",
       "6   ENSG00000238009.6  ENST00000471248.1      -  ENSE00001934975.1   \n",
       "7   ENSG00000243485.5  ENST00000473358.1      +  ENSE00001947070.1   \n",
       "8   ENSG00000238009.6  ENST00000477740.5      -  ENSE00001919246.1   \n",
       "9   ENSG00000227232.5  ENST00000488147.1      -  ENSE00001890219.1   \n",
       "10  ENSG00000240361.2  ENST00000492842.2      +  ENSE00001830178.2   \n",
       "11  ENSG00000268020.3  ENST00000606857.1      +  ENSE00003698237.1   \n",
       "12  ENSG00000284332.1  ENST00000607096.1      +  ENSE00003695741.1   \n",
       "13  ENSG00000278267.1  ENST00000619216.1      -  ENSE00003746039.1   \n",
       "14  ENSG00000186092.7  ENST00000641515.2      +  ENSE00003812156.1   \n",
       "15  ENSG00000240361.2  ENST00000642116.1      +  ENSE00003812686.1   \n",
       "\n",
       "                       transcript_type     TSS  \\\n",
       "0                               lncRNA   36081   \n",
       "1   transcribed_unprocessed_pseudogene   12010   \n",
       "2                 processed_transcript   11869   \n",
       "3                               lncRNA   36073   \n",
       "4                               lncRNA  120932   \n",
       "5                               lncRNA   30267   \n",
       "6                               lncRNA  129173   \n",
       "7                               lncRNA   29554   \n",
       "8                               lncRNA  129217   \n",
       "9               unprocessed_pseudogene   29570   \n",
       "10  transcribed_unprocessed_pseudogene   62949   \n",
       "11              unprocessed_pseudogene   52473   \n",
       "12                               miRNA   30366   \n",
       "13                               miRNA   17436   \n",
       "14                      protein_coding   65419   \n",
       "15                processed_transcript   57598   \n",
       "\n",
       "                                             sequence  \n",
       "0   CAAGGATTAGTCCTGCTGCCCTGTGCCCAGCTCCCACACAACGGGG...  \n",
       "1   CACAGGGTCTTGATGCTGTGGTCTTCATCTGCAGGTGTCTGACTTC...  \n",
       "2   CCAGCACCGGGTATCATTCACCATTTTTCTTTTCGTTAACTTGCCG...  \n",
       "3   AGTCCTGCTGCCCTGTGCCCAGCTCCCACACAACGGGGTTTCGGGG...  \n",
       "4   TTTCCATGTGATTTTCCTCTTAATTTTTTTCTAGCTGATCCATATG...  \n",
       "5   GCTTTTTTAAAAAGTTAAATATATAAAAAGAGCATCATCAGTCCAA...  \n",
       "6   AGCATCTTAAAATGATTCAACAGGAGGAGATAAGGAAGCTCGAGGA...  \n",
       "7   GGCACGGAACTCACGCTCACTCCGAGCTCCCGACGTGCACACGGCT...  \n",
       "8   GTGTCACAACTGCACATTGGTTTCACAGCTGCAGGACAAGTTCGAG...  \n",
       "9   GGGTAGGGGCGGCCTGACGCTCGGAAGACAACGCATGGGAGCCGTG...  \n",
       "10  GGTTGCTTCAGAGTCTTCCCTTCTATCTGATTCAGTGGACCAAGTA...  \n",
       "11  GAGTCAATGGCCAATGGATAGAGTAAATAATTCTGCGGTATCTAAA...  \n",
       "12  CCCTTGCCTTTtagggttgccagattaaaagacaggatgcccagct...  \n",
       "13  TGGGACCCCACAGTGCACGCTGTGCCCCTGATGATGTGGGAGAGGA...  \n",
       "14  TTAGTCTCAGTACACAAAGCTCAGACTGGCTATTCCCAGATCTCTT...  \n",
       "15  CTTTGTCTTTGGGGTATTACATAAATGAAGATTAAGCTATCTGAAT...  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_core_promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "c32274e1-5f2e-4a43-9db4-cf2a78b3c044",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>strand</th>\n",
       "      <th>coordinates</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000417324.1</td>\n",
       "      <td>ENSE00001656588.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>35720</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000417324.1</td>\n",
       "      <td>ENSE00001669267.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>35276</td>\n",
       "      <td>tgt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>12058</td>\n",
       "      <td>GCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>12228</td>\n",
       "      <td>AGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001758273.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>12698</td>\n",
       "      <td>TGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001799933.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>13053</td>\n",
       "      <td>GGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001746346.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>13375</td>\n",
       "      <td>AGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>12228</td>\n",
       "      <td>AGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>12722</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000461467.1</td>\n",
       "      <td>ENSE00001618781.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>35720</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000466430.5</td>\n",
       "      <td>ENSE00001606755.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>120774</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000466430.5</td>\n",
       "      <td>ENSE00001957285.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>112699</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000466430.5</td>\n",
       "      <td>ENSE00001944529.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>92090</td>\n",
       "      <td>ggt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>ENSE00001841699.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>30668</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000471248.1</td>\n",
       "      <td>ENSE00001934975.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>129054</td>\n",
       "      <td>CGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSE00001947070.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>30040</td>\n",
       "      <td>ggt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSE00001922571.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>30668</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENST00000477740.5</td>\n",
       "      <td>ENSE00001919246.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>129054</td>\n",
       "      <td>CGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENST00000477740.5</td>\n",
       "      <td>ENSE00001171005.3</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>120720</td>\n",
       "      <td>TGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENST00000477740.5</td>\n",
       "      <td>ENSE00001957285.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>112699</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00001890219.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>29533</td>\n",
       "      <td>CGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003507205.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>24737</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003477500.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>18267</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003565697.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>17914</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003475637.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>17605</td>\n",
       "      <td>TGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003502542.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>17232</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003553898.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>16857</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003621279.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>16606</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00002030414.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>15795</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00001935574.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>15004</td>\n",
       "      <td>CGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ENST00000492842.2</td>\n",
       "      <td>ENSE00001830178.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>63888</td>\n",
       "      <td>ACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ENST00000606857.1</td>\n",
       "      <td>ENSE00003698237.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>53313</td>\n",
       "      <td>CGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ENST00000607096.1</td>\n",
       "      <td>ENSE00003695741.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>30504</td>\n",
       "      <td>ata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>ENSE00003746039.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>17368</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSE00003812156.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "      <td>65434</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSE00003813641.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "      <td>65574</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>ENST00000642116.1</td>\n",
       "      <td>ENSE00003812686.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>57654</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ENST00000642116.1</td>\n",
       "      <td>ENSE00003812505.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>58857</td>\n",
       "      <td>TGT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id            exon_id                     transcript_type  \\\n",
       "0   ENST00000417324.1  ENSE00001656588.1                              lncRNA   \n",
       "1   ENST00000417324.1  ENSE00001669267.1                              lncRNA   \n",
       "2   ENST00000450305.2  ENSE00001948541.1  transcribed_unprocessed_pseudogene   \n",
       "3   ENST00000450305.2  ENSE00001671638.2  transcribed_unprocessed_pseudogene   \n",
       "4   ENST00000450305.2  ENSE00001758273.2  transcribed_unprocessed_pseudogene   \n",
       "5   ENST00000450305.2  ENSE00001799933.2  transcribed_unprocessed_pseudogene   \n",
       "6   ENST00000450305.2  ENSE00001746346.2  transcribed_unprocessed_pseudogene   \n",
       "7   ENST00000456328.2  ENSE00002234944.1                processed_transcript   \n",
       "8   ENST00000456328.2  ENSE00003582793.1                processed_transcript   \n",
       "9   ENST00000461467.1  ENSE00001618781.2                              lncRNA   \n",
       "10  ENST00000466430.5  ENSE00001606755.2                              lncRNA   \n",
       "11  ENST00000466430.5  ENSE00001957285.1                              lncRNA   \n",
       "12  ENST00000466430.5  ENSE00001944529.1                              lncRNA   \n",
       "13  ENST00000469289.1  ENSE00001841699.1                              lncRNA   \n",
       "14  ENST00000471248.1  ENSE00001934975.1                              lncRNA   \n",
       "15  ENST00000473358.1  ENSE00001947070.1                              lncRNA   \n",
       "16  ENST00000473358.1  ENSE00001922571.1                              lncRNA   \n",
       "17  ENST00000477740.5  ENSE00001919246.1                              lncRNA   \n",
       "18  ENST00000477740.5  ENSE00001171005.3                              lncRNA   \n",
       "19  ENST00000477740.5  ENSE00001957285.1                              lncRNA   \n",
       "20  ENST00000488147.1  ENSE00001890219.1              unprocessed_pseudogene   \n",
       "21  ENST00000488147.1  ENSE00003507205.1              unprocessed_pseudogene   \n",
       "22  ENST00000488147.1  ENSE00003477500.1              unprocessed_pseudogene   \n",
       "23  ENST00000488147.1  ENSE00003565697.1              unprocessed_pseudogene   \n",
       "24  ENST00000488147.1  ENSE00003475637.1              unprocessed_pseudogene   \n",
       "25  ENST00000488147.1  ENSE00003502542.1              unprocessed_pseudogene   \n",
       "26  ENST00000488147.1  ENSE00003553898.1              unprocessed_pseudogene   \n",
       "27  ENST00000488147.1  ENSE00003621279.1              unprocessed_pseudogene   \n",
       "28  ENST00000488147.1  ENSE00002030414.1              unprocessed_pseudogene   \n",
       "29  ENST00000488147.1  ENSE00001935574.1              unprocessed_pseudogene   \n",
       "30  ENST00000492842.2  ENSE00001830178.2  transcribed_unprocessed_pseudogene   \n",
       "31  ENST00000606857.1  ENSE00003698237.1              unprocessed_pseudogene   \n",
       "32  ENST00000607096.1  ENSE00003695741.1                               miRNA   \n",
       "33  ENST00000619216.1  ENSE00003746039.1                               miRNA   \n",
       "34  ENST00000641515.2  ENSE00003812156.1                      protein_coding   \n",
       "35  ENST00000641515.2  ENSE00003813641.1                      protein_coding   \n",
       "36  ENST00000642116.1  ENSE00003812686.1                processed_transcript   \n",
       "37  ENST00000642116.1  ENSE00003812505.1                processed_transcript   \n",
       "\n",
       "   strand coordinates sequence  \n",
       "0       -       35720      GGT  \n",
       "1       -       35276      tgt  \n",
       "2       +       12058      GCA  \n",
       "3       +       12228      AGT  \n",
       "4       +       12698      TGT  \n",
       "5       +       13053      GGC  \n",
       "6       +       13375      AGT  \n",
       "7       +       12228      AGT  \n",
       "8       +       12722      GGT  \n",
       "9       -       35720      GGT  \n",
       "10      -      120774      GGT  \n",
       "11      -      112699      GGT  \n",
       "12      -       92090      ggt  \n",
       "13      +       30668      GGT  \n",
       "14      -      129054      CGT  \n",
       "15      +       30040      ggt  \n",
       "16      +       30668      GGT  \n",
       "17      -      129054      CGT  \n",
       "18      -      120720      TGT  \n",
       "19      -      112699      GGT  \n",
       "20      -       29533      CGT  \n",
       "21      -       24737      GGT  \n",
       "22      -       18267      GGT  \n",
       "23      -       17914      GGT  \n",
       "24      -       17605      TGT  \n",
       "25      -       17232      GGT  \n",
       "26      -       16857      GGT  \n",
       "27      -       16606      GGT  \n",
       "28      -       15795      GGT  \n",
       "29      -       15004      CGG  \n",
       "30      +       63888      ACA  \n",
       "31      +       53313      CGT  \n",
       "32      +       30504      ata  \n",
       "33      -       17368      GGT  \n",
       "34      +       65434      GGT  \n",
       "35      +       65574      GGT  \n",
       "36      +       57654      GGT  \n",
       "37      +       58857      TGT  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_donor_splice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "4899bfe3-b2ca-4823-8815-70627ba6dc19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>strand</th>\n",
       "      <th>coordinates</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>12178</td>\n",
       "      <td>GAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001758273.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>12612</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001799933.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>12974</td>\n",
       "      <td>AGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001746346.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>13220</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001863096.1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>+</td>\n",
       "      <td>13452</td>\n",
       "      <td>AGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>12612</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>13220</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSE00001922571.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>30563</td>\n",
       "      <td>AGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSE00001827679.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>30975</td>\n",
       "      <td>agt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003507205.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>24892</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003477500.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>18367</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003565697.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>18062</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003475637.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>17743</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003502542.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>17369</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003553898.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>17056</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00003621279.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>16766</td>\n",
       "      <td>AGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00002030414.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>15948</td>\n",
       "      <td>AGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00001935574.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>15039</td>\n",
       "      <td>AGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSE00001843071.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>-</td>\n",
       "      <td>14502</td>\n",
       "      <td>GTG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id            exon_id                     transcript_type  \\\n",
       "0   ENST00000450305.2  ENSE00001671638.2  transcribed_unprocessed_pseudogene   \n",
       "1   ENST00000450305.2  ENSE00001758273.2  transcribed_unprocessed_pseudogene   \n",
       "2   ENST00000450305.2  ENSE00001799933.2  transcribed_unprocessed_pseudogene   \n",
       "3   ENST00000450305.2  ENSE00001746346.2  transcribed_unprocessed_pseudogene   \n",
       "4   ENST00000450305.2  ENSE00001863096.1  transcribed_unprocessed_pseudogene   \n",
       "5   ENST00000456328.2  ENSE00003582793.1                processed_transcript   \n",
       "6   ENST00000456328.2  ENSE00002312635.1                processed_transcript   \n",
       "7   ENST00000473358.1  ENSE00001922571.1                              lncRNA   \n",
       "8   ENST00000473358.1  ENSE00001827679.1                              lncRNA   \n",
       "9   ENST00000488147.1  ENSE00003507205.1              unprocessed_pseudogene   \n",
       "10  ENST00000488147.1  ENSE00003477500.1              unprocessed_pseudogene   \n",
       "11  ENST00000488147.1  ENSE00003565697.1              unprocessed_pseudogene   \n",
       "12  ENST00000488147.1  ENSE00003475637.1              unprocessed_pseudogene   \n",
       "13  ENST00000488147.1  ENSE00003502542.1              unprocessed_pseudogene   \n",
       "14  ENST00000488147.1  ENSE00003553898.1              unprocessed_pseudogene   \n",
       "15  ENST00000488147.1  ENSE00003621279.1              unprocessed_pseudogene   \n",
       "16  ENST00000488147.1  ENSE00002030414.1              unprocessed_pseudogene   \n",
       "17  ENST00000488147.1  ENSE00001935574.1              unprocessed_pseudogene   \n",
       "18  ENST00000488147.1  ENSE00001843071.1              unprocessed_pseudogene   \n",
       "\n",
       "   strand coordinates sequence  \n",
       "0       +       12178      GAT  \n",
       "1       +       12612      AGG  \n",
       "2       +       12974      AGC  \n",
       "3       +       13220      AGG  \n",
       "4       +       13452      AGT  \n",
       "5       +       12612      AGG  \n",
       "6       +       13220      AGG  \n",
       "7       +       30563      AGT  \n",
       "8       +       30975      agt  \n",
       "9       -       24892      AGG  \n",
       "10      -       18367      AGG  \n",
       "11      -       18062      AGG  \n",
       "12      -       17743      AGG  \n",
       "13      -       17369      AGG  \n",
       "14      -       17056      AGG  \n",
       "15      -       16766      AGA  \n",
       "16      -       15948      AGT  \n",
       "17      -       15039      AGT  \n",
       "18      -       14502      GTG  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_acceptor_splice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "297c6d07-7e21-44d7-89fb-5e31c54e4c05",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_promoter.to_csv(data_save_path +\"/\"+ entry['seqname']+\"_promoter.csv\", index=False)\n",
    "# df_core_promoter.to_csv(data_save_path +\"/\"+ entry['seqname']+\"_core_promoter_sequence.csv\", index=False)\n",
    "# df_acceptor_splice.to_csv(data_save_path +\"/\"+ entry['seqname']+\"_acceptor_splice_set_sequence.csv\", index=False)\n",
    "# df_donor_splice.to_csv(data_save_path +\"/\"+ entry['seqname']+\"_donor_splice_set.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "8e5827cd-0942-4812-a7de-5a880a371c34",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ttctccacaaagtgtttacttttggatttttgccagtctaacaggtgaAGccctggagattcttattagtgatttgggctggggcctggccatgtgtatttttttaaatttccactgatgattttgctgcatggccggtgttgagaatgactgCGCAAATTTGCCGGATTTCCTTTGCTGTTCCTGCATGTAGTTTAAACGAGATTGCCAGCACCGGGTATCATTCACCATTTTTCTTTTCGTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTATTTGCTGT'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_seq = pybedtools.BedTool.seq('chr1:11628-11927', fasta)\n",
    "test_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "a72c986b-6260-435e-92d2-495a2c5bbd4b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACAGCAAATACACATGAACAGAAAGAAGAGGTCAAAGAAAAGGCTGACGGCAAGTTAACGAAAAGAAAAATGGTGAATGATACCCGGTGCTGGCAATCTCGTTTAAACTACATGCAGGAACAGCAAAGGAAATCCGGCAAATTTGCGcagtcattctcaacaccggccatgcagcaaaatcatcagtggaaatttaaaaaaatacacatggccaggccccagcccaaatcactaataagaatctccagggCTtcacctgttagactggcaaaaatccaaaagtaaacactttgtggagaa\n"
     ]
    }
   ],
   "source": [
    "seq = Seq(test_seq)\n",
    "\n",
    "print (seq.reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8dcec4ae-0d9a-49af-b046-04c107e2a0a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_genome_file_path = '/mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/chr1.fa'\n",
    "fasta = pybedtools.example_filename(reference_genome_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "5bebd1ed-1188-4d7f-b7b8-6de3e80dbec9",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp_seq = pybedtools.BedTool.seq('chr1:1-248956422', fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "306e063e-f1fa-466d-8421-fd0862e86a76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "248956422"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(temp_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "0584843b-d173-43ae-a790-da4b29558b72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "248956471"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "sys.getsizeof(temp_seq,'U6') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "1bda15e6-d171-44b1-91d1-66dc5632ff14",
   "metadata": {},
   "outputs": [],
   "source": [
    "np_test_seq  = np.array(list(temp_seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "6e633014-9719-488e-a35c-9a8d2f14e758",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "995825792"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sys.getsizeof(np_test_seq,'U6') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "d6567b85-bbf4-4245-906b-be1af53a30ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_str= \"HGGVHGJBKJGNBJHBNBHJBNHGJHBHB\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "e78746dd-cce0-4c48-980e-02b47147971b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "test_str_new = np.array(test_str)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "3acc7ead-6d8d-4bbb-bd90-9126542f34c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array('HGGVHGJBKJGNBJHBNBHJBNHGJHBHB', dtype='<U29')"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_str_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "28b13586-5eae-4aec-b002-d54f0c97ba5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "position = [[10,12], [16,20]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc2e4e0d-39f5-488c-a648-094c4e4753bd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "a90ca177-7bf9-4f00-bf3f-3729831d50ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[10, 12]\n",
      "$$$\n",
      "$$$\n",
      "HGGVHGJBKJ$$$JHB$$$$$NHGJHBHB\n",
      "[16, 20]\n",
      "$$$$$\n",
      "$$$$$\n",
      "HGGVHGJBKJ$$$JHB$$$$$NHGJHBHB\n",
      "HGGVHGJBKJJHBNHGJHBHB\n"
     ]
    }
   ],
   "source": [
    "for i in position:\n",
    "    print (i)\n",
    "    sub_string = test_str[i[0]:i[1]+1]\n",
    "    print(sub_string)\n",
    "    substitute_str = \"$\"* int(i[1]+1-i[0])\n",
    "    print (substitute_str)\n",
    "    test_str = test_str.replace(sub_string, substitute_str)\n",
    "    print (test_str)\n",
    "line = re.sub('[$]', '', test_str)\n",
    "print (line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "9aee7d57-f924-4f95-a9dd-c5ffded3d5b2",
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "too many indices for array: array is 0-dimensional, but 1 were indexed",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_355239/2495947432.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# test_str_new[position[0][0]:position[0][1]] = 'NN'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mtest_str_new\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m: too many indices for array: array is 0-dimensional, but 1 were indexed"
     ]
    }
   ],
   "source": [
    "# test_str_new[position[0][0]:position[0][1]] = 'NN'\n",
    "test_str_new[0][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "558c0161-5b2a-40f7-ab3e-f59f19032130",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_str = test_str[0:position[0]]+ test_str[position[1]:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "050601b6-fa71-4bd3-9143-fffb5e5f2a02",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'$GGV$GJBKJGNBJ$BNB$JBN$GJ$B$B'"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_str = test_str.replace(test_str[0], '$')\n",
    "test_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b806e74a-5aad-49d2-8b84-8fb7a22e004c",
   "metadata": {},
   "outputs": [],
   "source": [
    "for "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "8ee74b1e-3398-4f74-8270-960f068cf624",
   "metadata": {},
   "outputs": [],
   "source": [
    "check_sr= \"\"\n",
    "for i in range(10):\n",
    "    check_sr= check_sr+ '0'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "929ff8f0-8eb2-4e07-bd0b-6b8f52a4f85c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0000000000'"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_sr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "0c9f1896-5273-4810-93cc-8331ec182ba6",
   "metadata": {},
   "outputs": [],
   "source": [
    "check_sr1= \"\"\n",
    "check_sr1 = [\"0\" for i in range(10)  ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "56d58a23-92cb-431f-897e-fec30c98a14a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['0', '0', '0', '0', '0', '0', '0', '0', '0', '0']"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_sr1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "33b01025-6519-41d5-93e1-3209677b2336",
   "metadata": {},
   "outputs": [],
   "source": [
    "checksr2= ''.join([\"0\" for i in range(10)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "816c3882-a88e-495e-9acb-89e700ddae3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0000000000'"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "checksr2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "8aecce6c-df92-45b2-a58b-af178829484a",
   "metadata": {},
   "outputs": [],
   "source": [
    "A = np.array([0,1,2,3,4,5,6])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "924c3c2d-f2e8-43c3-ba60-aa57eae053e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 1, 2, 3, 4, 5, 6])"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "de823636-ec0a-4078-b643-b34983997ccd",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: 'N'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_355239/2028919401.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mA\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"N\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m: invalid literal for int() with base 10: 'N'"
     ]
    }
   ],
   "source": [
    "A[0:2] = \"N\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "f7f274a6-d2c9-449b-b47d-f43638f23556",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (2966232614.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"/tmp/ipykernel_355239/2966232614.py\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    for  i in range([10:12]):\u001b[0m\n\u001b[0m                       ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "for  i in range([10:12]):\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "42b1f67b-b360-4843-8b29-3a1de7320837",
   "metadata": {},
   "outputs": [],
   "source": [
    "str1 = \"0\"*5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "2adea436-c0cc-44af-8ccf-b31b23ee5e3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'00000'"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "66020428-81da-44c9-9a83-d4d928c50168",
   "metadata": {},
   "outputs": [],
   "source": [
    "p = range(1,4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "d5aae4aa-a22a-439c-8900-23e97b633c3c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "range(1, 4)"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "ec2e5f24-dd94-4de3-aa14-4a30a32da357",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "000f4669-b885-4407-9ea0-3693eac5defb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-1\n"
     ]
    }
   ],
   "source": [
    "a_string = \"Welme to Kite! CO\"\n",
    "\n",
    "index = a_string.find(\"co\")\n",
    "\n",
    "print(index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "6a8569ad-62c3-477e-b822-86dee6eee201",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[8, 13, 18, 23, 55]\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "string = 'i think that that that that student wrote there is not that right'\n",
    "pattern = 'that'\n",
    "\n",
    "print([match.start() for match in re.finditer(pattern, string)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "261c6de7-4d1a-439c-b61d-4493884aaa13",
   "metadata": {},
   "outputs": [],
   "source": [
    "pattern1= 'GT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "33e030f8-1697-4887-867f-dea07e9daafc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5479664"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Non_GT_donor = [match.start() for match in re.finditer(pattern1, temp_seq)]\n",
    "len(Non_GT_donor)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "6f10325a-0f7c-46f0-9d95-9d6126aefbea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[11453, 11492, 11497, 11806, 11816, 11819, 11844, 11868, 11879, 11908]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Non_GT_donor[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "664c470b-f818-4716-80f4-8aa1eae5b3a2",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'NoneType' object is not subscriptable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_355239/1920682704.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mx\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m: 'NoneType' object is not subscriptable"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3385de7-1d46-466f-a226-0575d87d9001",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
