{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4cba45fd-0341-4d37-87aa-70c7a32af5e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/pdutta/anaconda3/envs/dgcca/lib/python3.7/site-packages/xenaPython/__init__.py:110: FutureWarning: Possible nested set at position 7\n",
      "  re.sub(r\"^[^[]+[[]([^]]*)[]].*$\", r\"\\1\", query, flags=re.DOTALL))\n"
     ]
    }
   ],
   "source": [
    "import xenaPython as xena"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ef54b3e6-f0fd-4d1c-ae9e-52266eae2810",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'publicHub': 'https://ucscpublic.xenahubs.net',\n",
       " 'tcgaHub': 'https://tcga.xenahubs.net',\n",
       " 'icgcHub': 'https://icgc.xenahubs.net',\n",
       " 'toilHub': 'https://toil.xenahubs.net',\n",
       " 'pcawgHub': 'https://pcawg.xenahubs.net',\n",
       " 'singlecellHub': 'https://singlecell.xenahubs.net',\n",
       " 'pancanAtlasHub': 'https://pancanatlas.xenahubs.net',\n",
       " 'treehouseHub': 'https://xena.treehouse.gi.ucsc.edu',\n",
       " 'gdcHub': 'https://gdc.xenahubs.net',\n",
       " 'reference': 'https://reference.xenahubs.net'}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xena.PUBLIC_HUBS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0dcd5360-113f-46bf-af1e-fbf9759f5dac",
   "metadata": {},
   "outputs": [],
   "source": [
    "host = xena.PUBLIC_HUBS['tcgaHub']\n",
    "dataset = \"survival/BRCA_survival.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "c567b7cd-e999-472e-acd3-e076f40485fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = xena.dataset_samples (host, dataset, None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "825fdf0e-e390-458c-9107-6aa50c5553b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "features = xena.dataset_field(host, dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "cf6649da-17b8-4760-8bbf-2b8a4d47af13",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['DFI',\n",
       " 'DFI.time',\n",
       " 'DSS',\n",
       " 'DSS.time',\n",
       " 'OS',\n",
       " 'OS.time',\n",
       " 'PFI',\n",
       " 'PFI.time',\n",
       " 'Redaction',\n",
       " '_PATIENT',\n",
       " 'sampleID']"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "419a2a54-b36c-4d6e-9ce4-0c79804b89af",
   "metadata": {},
   "outputs": [],
   "source": [
    "[position, features] = xena.dataset_probe_values(host, dataset, samples, features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "82a288c4-c0da-4c9d-b4ac-a0a94dbe2a3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1236"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(features[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a61dbe9b-0146-4ae1-b82c-89fff81fc7a0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b4f27a41-0397-49dc-957b-18136abee20a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1236]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xena.dataset_samples_n_dense_matrix(host, dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "8d49fc11-5976-4aa3-99e8-c34726aab0c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'rows': {'sampleID': ['TCGA-3C-AAAU-01',\n",
       "   'TCGA-3C-AALI-01',\n",
       "   'TCGA-3C-AALJ-01',\n",
       "   'TCGA-3C-AALK-01',\n",
       "   'TCGA-4H-AAAK-01'],\n",
       "  'rna-vaf': [None, None, None, None, None],\n",
       "  'genes': [None, None, None, None, None],\n",
       "  'dna-vaf': [None, None, None, None, None],\n",
       "  'position': [None, None, None, None, None],\n",
       "  'alt': [None, None, None, None, None],\n",
       "  'effect': [None, None, None, None, None],\n",
       "  'ref': [None, None, None, None, None],\n",
       "  'amino-acid': [None, None, None, None, None],\n",
       "  'altGene': [None, None, None, None, None]}}"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xena.sparse_data_examples(host, dataset, 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "51d977db-dd66-4598-b459-8d638e4ae5ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'name': 'TCGA.OV.sampleMap/HumanMethylation27', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.OV.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.OV.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.OV.sampleMap/OV_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.OV.sampleMap/RPPA_RBN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.OV.sampleMap/HT_HG-U133A', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.OV.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.OV.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.OV.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.OV.sampleMap/HiSeq', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.OV.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.OV.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.OV.sampleMap/AgilentG4502A_07_2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.OV.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.OV.sampleMap/HiSeq_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.OV.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.OV.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.OV.sampleMap/TCGA_OV_gene_expression_subtype',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.OV.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.OV.sampleMap/AgilentG4502A_07_3', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.DLBC.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.DLBC.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.DLBC.sampleMap/DLBC_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.DLBC.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.DLBC.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.DLBC.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.DLBC.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.DLBC.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.DLBC.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.DLBC.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.DLBC.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.DLBC.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRC.sampleMap/HumanMethylation27', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRC.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRC.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRC.sampleMap/RPPA_RBN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRC.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRC.sampleMap/KIRC_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.KIRC.sampleMap/miRNA_GA_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRC.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRC.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRC.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.KIRC.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRC.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRC.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRC.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.KIRC.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRC.sampleMap/AgilentG4502A_07_3', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.SARC.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.SARC.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.SARC.sampleMap/SARC_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.SARC.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.SARC.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.SARC.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.SARC.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.SARC.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.SARC.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.SARC.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.SARC.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.SARC.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PANCAN.sampleMap/HumanMethylation27', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PANCAN.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PANCAN.sampleMap/miRNA_GA_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PANCAN.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PANCAN.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PANCAN.sampleMap/SNP6_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.PANCAN.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COAD.sampleMap/HumanMethylation27', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COAD.sampleMap/GAV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COAD.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COAD.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COAD.sampleMap/RPPA_RBN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COAD.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COAD.sampleMap/miRNA_GA_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COAD.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COAD.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COAD.sampleMap/COAD_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.COAD.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.COAD.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COAD.sampleMap/GAV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COAD.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COAD.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COAD.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.COAD.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COAD.sampleMap/AgilentG4502A_07_3', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PRAD.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PRAD.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PRAD.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PRAD.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PRAD.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PRAD.sampleMap/PRAD_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.PRAD.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.PRAD.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PRAD.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PRAD.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PRAD.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.PRAD.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUSC.sampleMap/HumanMethylation27', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUSC.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUSC.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUSC.sampleMap/RPPA_RBN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUSC.sampleMap/HT_HG-U133A', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUSC.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUSC.sampleMap/miRNA_GA_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUSC.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUSC.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUSC.sampleMap/LUSC_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.LUSC.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.LUSC.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUSC.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUSC.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUSC.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.LUSC.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUSC.sampleMap/AgilentG4502A_07_3', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ACC.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ACC.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ACC.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ACC.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ACC.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ACC.sampleMap/ACC_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.ACC.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.ACC.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ACC.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ACC.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ACC.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.ACC.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_LUSC.HiSeq',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_OV.Agilent',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_KIRC.HiSeq.V2',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_GBM.U133A',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_KIRC.HiSeq',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_COAD.Agilent',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_PRAD.HiSeq.V2',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_COAD.HiSeq.V2',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_LUSC.HiSeq.V2',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_KICH.HiSeq.V2',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_BRCA.HiSeq',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_OV.U133A', 'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_BLCA.HiSeq.V2',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_GBM.Agilent',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_LIHC.HiSeq',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_GBM.HiSeq.V2',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_KIRP.HiSeq.V2',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_STAD.HiSeq',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_LUAD.HiSeq.V2',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_HNSC.HiSeq',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_UCEC.GA.V2',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_CESC.HiSeq.V2',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_READ.HiSeq.V2',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_HNSC.HiSeq.V2',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_BRCA.Agilent',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_BLCA.HiSeq',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_UCEC.GA', 'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_BRCA.HiSeq.V2',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_LUAD.HiSeq',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_THCA.HiSeq.V2',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_UCEC.HiSeq.V2',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_LIHC.HiSeq.V2',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/separate_processed/RABIT_READ.Agilent',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/pancan/RABIT_pancan.HiSeq.V2', 'type': 'clinicalMatrix'},\n",
       " {'name': 'RABIT/pancan/RABIT_pancan.HiSeq', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.KICH.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KICH.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KICH.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KICH.sampleMap/KICH_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.KICH.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KICH.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KICH.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.KICH.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KICH.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KICH.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KICH.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.KICH.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'probeMap/illuminaMethyl450_hg18_GPL13534', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/ucscGenesHg18Probe', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/illuminaMethyl27K_hg18_gpl8490', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/gencode.v23.annotation.gene.probemap', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/gencode.v19.annotation.exon.probemap', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/refGene_hg19.probeMap', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/probesAffySNP6', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/gencode.v23.basic.annotation.gene.probemap',\n",
       "  'type': 'probeMap'},\n",
       " {'name': 'probeMap/Illumina1M.hg19.probeMap', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/gencode.v27.annotation.gene.probemap', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/agilentHumanCGH2X400K_GPL9777.probemap',\n",
       "  'type': 'probeMap'},\n",
       " {'name': 'probeMap/illuminaMethyl450_hg19_GPL16304_TCGAlegacy',\n",
       "  'type': 'probeMap'},\n",
       " {'name': 'probeMap/probesAgilent1AWholeGenome_hg18', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/miRBase_primary_transcript_hg18', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/GPL6244.hg19.probeMap', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/probeAffy500K', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/AffymetrixHumanExon1ST_hg18', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/gencode.v19.comprehensive.transcript', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/probesAffy500K', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/ucscGenesHg19Probe', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/HumanHT-12_V3_0_R3_11283641_A', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/probeAffy250KstyI', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/unc_v2_exon_hg19_probe', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/gencode.v19.annotation.gene.probemap', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/illuminaMethyl450_hg38_GDC', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/illuminaMethyl27_hg38_GDC', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/gencode.v23.basic.annotation.transcript.probemap',\n",
       "  'type': 'probeMap'},\n",
       " {'name': 'probeMap/gencode.v23.annotation.transcript.probemap',\n",
       "  'type': 'probeMap'},\n",
       " {'name': 'probeMap/Illumina1M.hg18.probeMap', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/illuminaMethyl27K_hg18_gpl8490_TCGAlegacy',\n",
       "  'type': 'probeMap'},\n",
       " {'name': 'probeMap/agilentHuman4X44K_hg18.probeMap', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/miRBase_primary_transcript_hg19', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/gencode.vM13.basic.gene', 'type': None},\n",
       " {'name': 'probeMap/gencode.v22.annotation.gene.probeMap', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/illuminaMethyl450_hg19_GPL16304', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/gencode.annotation.gene.probemap', 'type': 'probeMap'},\n",
       " {'name': 'TCGA.UCS.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCS.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCS.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCS.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCS.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCS.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.UCS.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCS.sampleMap/UCS_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.UCS.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCS.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCS.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.UCS.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COADREAD.sampleMap/HumanMethylation27',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COADREAD.sampleMap/HumanMethylation450',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COADREAD.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COADREAD.sampleMap/RPPA_RBN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COADREAD.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COADREAD.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COADREAD.sampleMap/COADREAD_clinicalMatrix',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.COADREAD.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.COADREAD.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COADREAD.sampleMap/SNP6_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.COADREAD.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.COADREAD.sampleMap/AgilentG4502A_07_3',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUNG.sampleMap/HumanMethylation27', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUNG.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUNG.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUNG.sampleMap/RPPA_RBN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUNG.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUNG.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUNG.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.LUNG.sampleMap/LUNG_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.LUNG.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUNG.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.LUNG.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUNG.sampleMap/AgilentG4502A_07_3', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUAD.sampleMap/HumanMethylation27', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUAD.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUAD.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUAD.sampleMap/RPPA_RBN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUAD.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUAD.sampleMap/miRNA_GA_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUAD.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUAD.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUAD.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.LUAD.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUAD.sampleMap/LUAD_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.LUAD.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUAD.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LUAD.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.LUAD.sampleMap/AgilentG4502A_07_3', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.FPPP.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.FPPP.sampleMap/FPPP_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.LAML.sampleMap/HumanMethylation27', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LAML.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LAML.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LAML.sampleMap/GA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LAML.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LAML.sampleMap/miRNA_GA_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LAML.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LAML.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LAML.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.LAML.sampleMap/GA_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LAML.sampleMap/LAML_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.LAML.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LAML.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.LAML.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBM.sampleMap/HumanMethylation27', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBM.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBM.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBM.sampleMap/RPPA_RBN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBM.sampleMap/HT_HG-U133A', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBM.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBM.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBM.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBM.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.GBM.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBM.sampleMap/AgilentG4502A_07_2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBM.sampleMap/GBM_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.GBM.sampleMap/AgilentG4502A_07_1', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBM.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBM.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBM.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.GBM.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRP.sampleMap/HumanMethylation27', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRP.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRP.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRP.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRP.sampleMap/miRNA_GA_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRP.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRP.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRP.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.KIRP.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRP.sampleMap/KIRP_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.KIRP.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRP.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRP.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.KIRP.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.KIRP.sampleMap/AgilentG4502A_07_3', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PAAD.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PAAD.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PAAD.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PAAD.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PAAD.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PAAD.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.PAAD.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PAAD.sampleMap/PAAD_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.PAAD.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PAAD.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PAAD.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.PAAD.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.CHOL.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.CHOL.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.CHOL.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.CHOL.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.CHOL.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.CHOL.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.CHOL.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.CHOL.sampleMap/CHOL_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.CHOL.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.CHOL.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.CHOL.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.CHOL.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.CESC.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.CESC.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.CESC.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.CESC.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.CESC.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.CESC.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.CESC.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.CESC.sampleMap/CESC_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.CESC.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.CESC.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.CESC.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.CESC.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.SKCM.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.SKCM.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.SKCM.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.SKCM.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.SKCM.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.SKCM.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.SKCM.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.SKCM.sampleMap/SKCM_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.SKCM.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.SKCM.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.SKCM.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.SKCM.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LGG.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LGG.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LGG.sampleMap/oneoff_TCGA_LGG_MethylMix',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LGG.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LGG.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LGG.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LGG.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.LGG.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LGG.sampleMap/LGG_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.LGG.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LGG.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LGG.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.LGG.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LGG.sampleMap/AgilentG4502A_07_3', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PCPG.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PCPG.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PCPG.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PCPG.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PCPG.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PCPG.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.PCPG.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PCPG.sampleMap/PCPG_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.PCPG.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PCPG.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.PCPG.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.PCPG.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.TGCT.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.TGCT.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.TGCT.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.TGCT.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.TGCT.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.TGCT.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.TGCT.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.TGCT.sampleMap/TGCT_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.TGCT.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.TGCT.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.TGCT.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.TGCT.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BLCA.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BLCA.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BLCA.sampleMap/RPPA_RBN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BLCA.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BLCA.sampleMap/miRNA_GA_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BLCA.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BLCA.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BLCA.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.BLCA.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BLCA.sampleMap/BLCA_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.BLCA.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BLCA.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BLCA.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.BLCA.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.THYM.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.THYM.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.THYM.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.THYM.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.THYM.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.THYM.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.THYM.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.THYM.sampleMap/THYM_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.THYM.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.THYM.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.THYM.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.THYM.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BRCA.sampleMap/HumanMethylation27', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BRCA.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BRCA.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BRCA.sampleMap/RPPA_RBN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BRCA.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BRCA.sampleMap/miRNA_GA_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BRCA.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BRCA.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BRCA.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.BRCA.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BRCA.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BRCA.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BRCA.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.BRCA.sampleMap/BRCA_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.BRCA.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.BRCA.sampleMap/AgilentG4502A_07_3', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UVM.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UVM.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UVM.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UVM.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UVM.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UVM.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.UVM.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UVM.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UVM.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UVM.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.UVM.sampleMap/UVM_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.UVM.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCEC.sampleMap/HumanMethylation27', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCEC.sampleMap/GAV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCEC.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCEC.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCEC.sampleMap/RPPA_RBN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCEC.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCEC.sampleMap/miRNA_GA_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCEC.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCEC.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCEC.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.UCEC.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCEC.sampleMap/GAV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCEC.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCEC.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCEC.sampleMap/UCEC_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.UCEC.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.UCEC.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.UCEC.sampleMap/AgilentG4502A_07_3', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LIHC.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LIHC.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LIHC.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LIHC.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LIHC.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LIHC.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.LIHC.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LIHC.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LIHC.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LIHC.sampleMap/LIHC_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.LIHC.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.LIHC.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBMLGG.sampleMap/HumanMethylation450',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBMLGG.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBMLGG.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBMLGG.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBMLGG.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.GBMLGG.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.GBMLGG.sampleMap/GBMLGG_clinicalMatrix',\n",
       "  'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.GBMLGG.sampleMap/SNP6_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.GBMLGG.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.THCA.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.THCA.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.THCA.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.THCA.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.THCA.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.THCA.sampleMap/mutation_broad_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.THCA.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.THCA.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.THCA.sampleMap/THCA_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.THCA.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.THCA.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.THCA.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.THCA.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.HNSC.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.HNSC.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.HNSC.sampleMap/RPPA_RBN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.HNSC.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.HNSC.sampleMap/miRNA_GA_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.HNSC.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.HNSC.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.HNSC.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.HNSC.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.HNSC.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.HNSC.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.HNSC.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.HNSC.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.HNSC.sampleMap/HNSC_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.ESCA.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ESCA.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ESCA.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ESCA.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ESCA.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ESCA.sampleMap/HiSeq', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ESCA.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.ESCA.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ESCA.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ESCA.sampleMap/HiSeq_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ESCA.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ESCA.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.ESCA.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.ESCA.sampleMap/ESCA_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.STAD.sampleMap/HumanMethylation27', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.STAD.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.STAD.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.STAD.sampleMap/GA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.STAD.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.STAD.sampleMap/miRNA_GA_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.STAD.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.STAD.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.STAD.sampleMap/HiSeq', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.STAD.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.STAD.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.STAD.sampleMap/GA_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.STAD.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.STAD.sampleMap/HiSeq_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.STAD.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.STAD.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.STAD.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.STAD.sampleMap/STAD_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.READ.sampleMap/HumanMethylation27', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.READ.sampleMap/GAV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.READ.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.READ.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.READ.sampleMap/RPPA_RBN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.READ.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.READ.sampleMap/miRNA_GA_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.READ.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.READ.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.READ.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.READ.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.READ.sampleMap/GAV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.READ.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.READ.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.READ.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.READ.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.READ.sampleMap/READ_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'TCGA.READ.sampleMap/AgilentG4502A_07_3', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.MESO.sampleMap/HumanMethylation450', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.MESO.sampleMap/Gistic2_CopyNumber_Gistic2_all_data_by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.MESO.sampleMap/HiSeqV2_percentile', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.MESO.sampleMap/HiSeqV2_PANCAN', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.MESO.sampleMap/HiSeqV2', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.MESO.sampleMap/SNP6_nocnv_genomicSegment',\n",
       "  'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.MESO.sampleMap/miRNA_HiSeq_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.MESO.sampleMap/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.MESO.sampleMap/RPPA', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.MESO.sampleMap/SNP6_genomicSegment', 'type': 'genomicSegment'},\n",
       " {'name': 'TCGA.MESO.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.MESO.sampleMap/MESO_clinicalMatrix', 'type': 'clinicalMatrix'},\n",
       " {'name': 'probeMap/hugo_gencode_good_hg19_V24lift37_probemap',\n",
       "  'type': 'probeMap'},\n",
       " {'name': 'probeMap/hugo_refgene_good_hg18_probemap', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/unc_exon_hg19_probe', 'type': 'probeMap'},\n",
       " {'name': 'exon_TCGA.hg18.Feb2011.gaf.probemap', 'type': 'probeMap'},\n",
       " {'name': 'TCGA.LUAD.sampleMap/HiSeqV2_exon', 'type': 'genomicMatrix'},\n",
       " {'name': 'paradigmPathwayV1.probeMap', 'type': 'probeMap'},\n",
       " {'name': 'unc_v2_exon_hg19_probe_TCGA', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/affyU133_hg19', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/hugo_gencode_good_hg38_probemap', 'type': 'probeMap'},\n",
       " {'name': 'probeMap/hugo_gencode_good_hg38_v23comp_probemap',\n",
       "  'type': 'probeMap'},\n",
       " {'name': 'merge_merged_reals/ACC_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/BLCA_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/BRCA_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/CESC_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/CHOL_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/COAD_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/DLBC_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/ESCA_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/GBM_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/HNSC_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/KICH_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/KIRC_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/KIRP_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/LAML_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/LGG_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/LIHC_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/LUAD_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/LUSC_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/MESO_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/OV_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/PAAD_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/PCPG_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/PRAD_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/READ_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/SARC_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/SKCM_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/STAD_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/TGCT_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/THCA_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/THYM_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/UCEC_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/UCS_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'merge_merged_reals/UVM_merge_merged_reals.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'mc3/ACC_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/BLCA_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/BRCA_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/CESC_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/CHOL_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/COAD_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/DLBC_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/ESCA_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/GBM_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/HNSC_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/KICH_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/KIRC_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/KIRP_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/LGG_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/LIHC_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/LUAD_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/LUSC_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/MESO_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/OV_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/PAAD_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/PCPG_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/PRAD_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/READ_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/SARC_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/SKCM_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/STAD_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/TGCT_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/THCA_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/THYM_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/UCEC_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/UCS_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/UVM_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3_gene_level/ACC_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/BLCA_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/BRCA_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/CESC_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/CHOL_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/COAD_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/DLBC_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/ESCA_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/GBM_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/HNSC_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/KICH_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/KIRC_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/KIRP_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/LGG_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/LIHC_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/LUAD_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/LUSC_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/MESO_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/OV_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/PAAD_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/PCPG_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/PRAD_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/READ_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/SARC_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/SKCM_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/STAD_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/TGCT_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/THCA_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/THYM_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/UCEC_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/UCS_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/UVM_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/ACC_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/BLCA_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/BRCA_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/CESC_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/CHOL_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/COAD_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/DLBC_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/ESCA_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/GBM_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/HNSC_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/KICH_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/KIRC_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/KIRP_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/LAML_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/LGG_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/LIHC_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/LUAD_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/LUSC_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/MESO_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/OV_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/PAAD_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/PCPG_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/PRAD_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/READ_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/SARC_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/SKCM_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/STAD_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/TGCT_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/THCA_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/THYM_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/UCEC_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/UCS_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z/UVM_PanCan33_ssGSEA_1387GeneSets_NonZero_sample_level_Z.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'survival/ACC_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/BLCA_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/BRCA_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/CESC_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/CHOL_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/COADREAD_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/COAD_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/DLBC_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/ESCA_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/GBMLGG_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/GBM_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/HNSC_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/KICH_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/KIRC_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/KIRP_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/LAML_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/LGG_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/LIHC_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/LUAD_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/LUNG_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/MESO_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/OV_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/PAAD_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/PCPG_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/PRAD_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/READ_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/SARC_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/SKCM_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/STAD_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/TGCT_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/THCA_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/THYM_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/UCEC_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/UCS_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'survival/UVM_survival.txt', 'type': 'clinicalMatrix'},\n",
       " {'name': 'mc3/LUNG_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/GBMLGG_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3/COADREAD_mc3.txt', 'type': 'mutationVector'},\n",
       " {'name': 'mc3_gene_level/LUNG_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/GBMLGG_mc3_gene_level.txt', 'type': 'genomicMatrix'},\n",
       " {'name': 'mc3_gene_level/COADREAD_mc3_gene_level.txt',\n",
       "  'type': 'genomicMatrix'},\n",
       " {'name': 'TCGA.LAML.sampleMap/mutation_wustl', 'type': 'mutationVector'},\n",
       " {'name': 'TCGA.LAML.sampleMap/mutation_wustl_gene', 'type': 'genomicMatrix'},\n",
       " {'name': 'survival/LUSC_survival.txt', 'type': 'clinicalMatrix'}]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xena.all_datasets(host)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "d338aa25-822a-499c-9197-909abfb149c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[564]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "42271e51-71da-4d88-96fb-4e94c996f3cb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package xenaPython:\n",
      "\n",
      "NAME\n",
      "    xenaPython - Methods for querying data from UCSC Xena hubs\n",
      "\n",
      "DESCRIPTION\n",
      "    Data rows are associated with \"sample\" IDs.\n",
      "    Sample IDs are unique within a \"cohort\".\n",
      "    A \"dataset\" is a particular assay of a cohort, e.g. gene expression.\n",
      "    Datasets have associated metadata, specifying their data type and cohort.\n",
      "    \n",
      "    There are three primary data types: dense matrix (samples by probes),\n",
      "    sparse (sample, position, variant), and segmented (sample, position, value).\n",
      "    \n",
      "    \n",
      "    Dense matrices can be genotypic or phenotypic. Phenotypic matrices have\n",
      "    associated field metadata (descriptive names, codes, etc.).\n",
      "    \n",
      "    Genotypic matricies may have an associated probeMap, which maps probes to\n",
      "    genomic locations. If a matrix has hugo probeMap, the probes themselves\n",
      "    are gene names. Otherwise, a probeMap is used to map a gene location to a\n",
      "    set of probes.\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    convert\n",
      "    example\n",
      "    jupyter\n",
      "    xenaAPI\n",
      "    xenaQuery\n",
      "\n",
      "FUNCTIONS\n",
      "    all_cohorts(host, exclude)\n",
      "    \n",
      "    all_datasets(host)\n",
      "    \n",
      "    all_datasets_n(host)\n",
      "        Count the number datasets with non-null cohort\n",
      "    \n",
      "    all_field_metadata(host, dataset)\n",
      "        Metadata for all dataset fields (phenotypic datasets)\n",
      "    \n",
      "    cohort_samples(host, cohort, limit)\n",
      "        All samples in cohort\n",
      "    \n",
      "    cohort_summary(host, exclude)\n",
      "        Count datasets per-cohort, excluding the given dataset types\n",
      "        \n",
      "        xena.cohort_summary(xena.PUBLIC_HUBS[\"pancanAtlasHub\"], [\"probeMap\"])\n",
      "    \n",
      "    dataset_fetch(host, dataset, samples, probes)\n",
      "        Probe values for give samples\n",
      "    \n",
      "    dataset_field(host, dataset)\n",
      "        All field (probe) names in dataset\n",
      "    \n",
      "    dataset_field_examples(host, dataset, count)\n",
      "        Field names in dataset, up to <count>\n",
      "    \n",
      "    dataset_field_n(host, dataset)\n",
      "        Number of fields in dataset\n",
      "    \n",
      "    dataset_gene_probe_avg(host, dataset, samples, genes)\n",
      "        Probe average, per-gene, for given samples\n",
      "    \n",
      "    dataset_gene_probes_values(host, dataset, samples, genes)\n",
      "        Probe values in gene, and probe genomic positions, for given samples\n",
      "    \n",
      "    dataset_list(host, cohorts)\n",
      "        Dataset metadata for datasets in the given cohorts\n",
      "    \n",
      "    dataset_metadata(host, dataset)\n",
      "        Dataset metadata\n",
      "    \n",
      "    dataset_probe_signature(host, dataset, samples, probes, weights)\n",
      "        Computed probe signature for given samples and weight array\n",
      "    \n",
      "    dataset_probe_values(host, dataset, samples, probes)\n",
      "        Probe values for given samples, and probe genomic positions\n",
      "        \n",
      "        host = xena.PUBLIC_HUBS[\"pancanAtlasHub\"]\n",
      "        dataset = \"EB++AdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.xena\"\n",
      "        samples = xena.dataset_samples(host, dataset, None)\n",
      "        [position, [foxm1, tp53]] = xena.dataset_probe_values(host, dataset, samples, [\"FOXM1\", \"TP53\"])\n",
      "    \n",
      "    dataset_samples(host, dataset, limit)\n",
      "        All samples in dataset (optional limit)\n",
      "        \n",
      "        samples = xena.dataset_samples(xena.PUBLIC_HUBS[\"pancanAtlasHub\"], \"EB++AdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.xena\", None)\n",
      "    \n",
      "    dataset_samples_n_dense_matrix(host, dataset)\n",
      "        All samples in dataset (faster, for dense matrix dataset only)\n",
      "    \n",
      "    dataset_sources(host, dataset)\n",
      "    \n",
      "    dataset_status(host, dataset)\n",
      "    \n",
      "    datasets_null_rows(host)\n",
      "    \n",
      "    feature_list(host, dataset)\n",
      "        Dataset field names and long titles (phenotypic datasets)\n",
      "    \n",
      "    field_codes(host, dataset, fields)\n",
      "        Codes for categorical fields\n",
      "    \n",
      "    field_metadata(host, dataset, fields)\n",
      "        Metadata for given fields (phenotypic datasets)\n",
      "    \n",
      "    gene_transcripts(host, dataset, gene)\n",
      "        Gene transcripts\n",
      "    \n",
      "    load_file(name, host='http://127.0.0.1:7222')\n",
      "        # XXX need timeout and error handling\n",
      "        # should this do the file copy, too?\n",
      "    \n",
      "    load_ipython_extension(ipython)\n",
      "        jupyter support method\n",
      "    \n",
      "    match_fields(host, dataset, names)\n",
      "        Find fields matching names (must be lower-case)\n",
      "    \n",
      "    open_browser(url='heatmap/', columns=[])\n",
      "    \n",
      "    probe_count(host, dataset)\n",
      "    \n",
      "    probemap_list(host)\n",
      "        Find probemaps\n",
      "    \n",
      "    ref_gene_exons(host, dataset, genes)\n",
      "        Gene model\n",
      "    \n",
      "    ref_gene_position(host, dataset, gene)\n",
      "        Gene position from gene model\n",
      "    \n",
      "    ref_gene_range(host, dataset, chr, start, end)\n",
      "        Gene models overlapping range\n",
      "    \n",
      "    segment_data_examples(host, dataset, count)\n",
      "        Initial segmented data rows, with limit\n",
      "    \n",
      "    segmented_data_range(host, dataset, samples, chr, start, end)\n",
      "        Segmented (copy number) data overlapping range\n",
      "    \n",
      "    sparse_data(host, dataset, samples, genes)\n",
      "        Sparse (mutation) data rows for genes\n",
      "    \n",
      "    sparse_data_examples(host, dataset, count)\n",
      "        Initial sparse data rows, with limit\n",
      "    \n",
      "    sparse_data_match_field(host, field, dataset, names)\n",
      "        Genes in sparse (mutation) dataset matching given names\n",
      "    \n",
      "    sparse_data_match_field_slow(host, field, dataset, names)\n",
      "        Genes in sparse (mutation) dataset matching given names, case-insensitive (names must be lower-case)\n",
      "    \n",
      "    sparse_data_match_partial_field(host, field, dataset, names, limit)\n",
      "        Partial match genes in sparse (mutation) dataset\n",
      "    \n",
      "    sparse_data_range(host, dataset, samples, chr, start, end)\n",
      "        Sparse (mutation) data rows overlapping the given range, for the given samples\n",
      "    \n",
      "    transcript_expression(host, transcripts, studyA, subtypeA, studyB, subtypeB, dataset)\n",
      "    \n",
      "    unload_ipython_extension(ipython)\n",
      "        jupyter support method\n",
      "\n",
      "DATA\n",
      "    LOCAL_HUB = 'https://local.xena.ucsc.edu:7223'\n",
      "    PUBLIC_HUBS = {'gdcHub': 'https://gdc.xenahubs.net', 'icgcHub': 'https...\n",
      "    QUERIES = {'all_cohorts': ';allCohorts\\n(fn [exclude]\\n\\t(map :cohort\\...\n",
      "    __warningregistry__ = {'version': 10, ('Possible nested set at positio...\n",
      "    excludeType = ['probeMap', 'probemap', 'genePredExt']\n",
      "    ipython_instance = None\n",
      "\n",
      "FILE\n",
      "    /home/pdutta/anaconda3/envs/dgcca/lib/python3.7/site-packages/xenaPython/__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import xenaPython \n",
    "help(xenaPython)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88ee1051-229d-4117-94bc-26d97e9b5d8f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
