{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "13e86a2e-92b3-4644-a96d-547bf513e828",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, glob\n",
    "import pandas as pd\n",
    "import pysam\n",
    "import numpy as np\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c2380204-9847-4c6f-b152-114f02e0139d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_type = \"Brain\"\n",
    "intersected_base_path = \"/data/projects/GDC_Cancer_Wise/New_data/{}/Generated_files/Intersected_Data/300bp_TFBS\".format(cancer_type)\n",
    "reference_genome_path=\"/data/projects/Resources/Gencode_genome_annotation/GRCh38.primary_assembly.genome.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5a4aa801-312d-4380-adb6-0c0ebcd9ba78",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_path = \"/data/projects/GDC_Cancer_Wise/New_data/{}/Generated_files/DNABERT_Data/Somatic/TFBS_300bp\".format(cancer_type)\n",
    "if not os.path.exists(output_path):\n",
    "    os.makedirs(output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1e9c0e52-cf4d-48b1-9173-cd32e57d8d95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>H2AFZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>H2AK5ac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>H2BK12ac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>H2BK5ac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>H3F3A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>H3K14ac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>H3K27ac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>H3K27me3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>H3K36me3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>H3K4me1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>H3K4me2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>H3K79me1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>H3K79me2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>H3K9ac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>H3K9me1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>H3K9me3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>H4K20me1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>H4K5ac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>H4K8ac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>IKZF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>RAD21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>RBM22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>RUNX3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>TRIM22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ZBTB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ZNF143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NFE2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>PRDM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SCRT2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ZFHX2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ZNF121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ZNF366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>ZNF770</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           0\n",
       "0        FOS\n",
       "1      H2AFZ\n",
       "2    H2AK5ac\n",
       "3   H2BK12ac\n",
       "4    H2BK5ac\n",
       "5      H3F3A\n",
       "6    H3K14ac\n",
       "7    H3K27ac\n",
       "8   H3K27me3\n",
       "9   H3K36me3\n",
       "10   H3K4me1\n",
       "11   H3K4me2\n",
       "12  H3K79me1\n",
       "13  H3K79me2\n",
       "14    H3K9ac\n",
       "15   H3K9me1\n",
       "16   H3K9me3\n",
       "17  H4K20me1\n",
       "18    H4K5ac\n",
       "19    H4K8ac\n",
       "20     IKZF1\n",
       "21     RAD21\n",
       "22     RBM22\n",
       "23     RUNX3\n",
       "24    TRIM22\n",
       "25    ZBTB33\n",
       "26    ZNF143\n",
       "27      NFE2\n",
       "28     PRDM1\n",
       "29     SCRT2\n",
       "30     ZFHX2\n",
       "31    ZNF121\n",
       "32    ZNF366\n",
       "33    ZNF770"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_models_df= pd.read_csv(\"/home/campus.stonybrook.edu/pdutta/Github/Postdoc/DNABERT_data_processing/TFBS/important_tfbs.txt\", sep=\"\\t\", header=None)\n",
    "top_models_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c05f597d-17d5-4ea2-9dae-2903b1c0fe98",
   "metadata": {},
   "outputs": [],
   "source": [
    "def seq2kmer(seq):\n",
    "    \"\"\"\n",
    "    Convert original sequence to kmers\n",
    "    \n",
    "    Arguments:\n",
    "    seq -- str, original sequence.\n",
    "    k -- int, kmer of length k specified.\n",
    "    \n",
    "    Returns:\n",
    "    kmers -- str, kmers separated by space\n",
    "    \"\"\"\n",
    "    k=6\n",
    "    kmer = [seq[x:x+k] for x in range(len(seq)+1-k)]\n",
    "    kmers = \" \".join(kmer)\n",
    "    return kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a30b3260-fbcf-4037-9916-961c31c8c4f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sequences(df, reference_fasta):\n",
    "    data = []\n",
    "    for idx, row in df.iterrows():\n",
    "        #print(row)\n",
    "        chrom = row[0]\n",
    "        ref_start = row[1]\n",
    "        ref_end = row[2]\n",
    "        variant_start = row['START_POS']\n",
    "        variant_end = row['END_POS']\n",
    "        ref_nucleotide = row['REF']\n",
    "        alt = row[\"ALT\"]\n",
    "        \n",
    "        # Adjust for 0-based indexing in python\n",
    "        variant_pos_start = variant_start - ref_start\n",
    "        variant_pos_end = variant_end - ref_start\n",
    "        #print(ref_nucleotide , alt, variant_pos_start, variant_pos_end)\n",
    "        \n",
    "        \n",
    "        # Get reference sequence\n",
    "        #print(chrom)\n",
    "        ref_seq = reference_fasta.fetch(chrom, ref_start, ref_end)\n",
    "        #print(ref_seq)\n",
    "        \n",
    "        # Handle insertion and deletion to get the correct alt sequence\n",
    "        # Identify if the variant is an insertion or deletion\n",
    "        if len(ref_nucleotide) < len(alt):  # Insertion\n",
    "            delete_size =  len(alt) - len(ref_nucleotide)\n",
    "            #print(variant_pos_start, variant_pos_end ,delete_size)\n",
    "            alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:len(ref_seq) - delete_size]\n",
    "\n",
    "        elif len(ref_nucleotide) > len(alt):  # Deletion\n",
    "            insert_size = len(ref_nucleotide) - len(alt)\n",
    "            #print(insert_size)\n",
    "            extra_bases = reference_fasta.fetch(chrom, ref_end, ref_end + insert_size)\n",
    "            #print(extra_bases)\n",
    "            alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:] + extra_bases\n",
    "\n",
    "        else:  # SNV\n",
    "            alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:]\n",
    "        # print(ref_seq)\n",
    "        # print(alt_seq)\n",
    "        # input()\n",
    "\n",
    "\n",
    "        data.append({\n",
    "            'chr': chrom,\n",
    "            #'strand': row['strand'],\n",
    "            #'Transcript_ID': row['transcript_id'],\n",
    "            'Chip_Seq_start': row[1],\n",
    "            'Chip_Seq_end': row[2],\n",
    "            'varinat_start': variant_start,\n",
    "            'variant_end': variant_end,\n",
    "            'ref_neucleotide': ref_nucleotide,\n",
    "            'alternative_neucleotide': alt,\n",
    "            'reference_seq': ref_seq,\n",
    "            'alt_seq': alt_seq\n",
    "        })\n",
    "            \n",
    "    \n",
    "    # Convert the list of dictionaries to a DataFrame\n",
    "    new_df = pd.DataFrame(data)\n",
    "    # print(new_df.shape)\n",
    "    # print(new_df)\n",
    "    new_df = new_df.drop_duplicates().reset_index()\n",
    "    #print(new_df.shape)\n",
    "    #print(\"&*\")\n",
    "    data = []\n",
    "    merged_list = list(zip(new_df['reference_seq'], new_df['alt_seq']))\n",
    "    merged_list = [item.upper() for tup in merged_list for item in tup]\n",
    "    #print(merged_list)\n",
    "    kmer_lst = list(map(seq2kmer, merged_list))\n",
    "    df_kmer = pd.DataFrame(kmer_lst, columns=['Sequence'])\n",
    "    df_kmer['Label'] = np.random.choice([0, 1], size=len(df_kmer))\n",
    "    #print(df_kmer.shape)\n",
    "    return new_df, df_kmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ef4c248a-8b67-4dd9-8992-23b3d81450f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#####The reference file is loaded######\n"
     ]
    }
   ],
   "source": [
    "## Load the reference file\n",
    "reference_fasta = pysam.FastaFile(reference_genome_path)\n",
    "print(\"#####The reference file is loaded######\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "933aa443-d84e-48a9-b824-e5df3470e39b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/FOS/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for FOS***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/H2AFZ/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H2AFZ***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/H2AK5ac/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H2AK5ac***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/H2BK12ac/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H2BK12ac***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/H2BK5ac/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H2BK5ac***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/H3F3A/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H3F3A***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/H3K14ac/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H3K14ac***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/H3K27ac/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H3K27ac***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/H3K27me3/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H3K27me3***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/H3K36me3/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H3K36me3***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/H3K4me1/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H3K4me1***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/H3K4me2/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H3K4me2***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/H3K79me1/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H3K79me1***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/H3K79me2/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H3K79me2***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/H3K9ac/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H3K9ac***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/H3K9me1/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H3K9me1***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/H3K9me3/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H3K9me3***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/H4K20me1/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H4K20me1***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/H4K5ac/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H4K5ac***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/H4K8ac/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H4K8ac***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/IKZF1/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for IKZF1***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/RAD21/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for RAD21***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/RBM22/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for RBM22***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/RUNX3/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for RUNX3***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/TRIM22/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for TRIM22***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/ZBTB33/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZBTB33***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/ZNF143/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZNF143***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/NFE2/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for NFE2***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/PRDM1/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for PRDM1***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/SCRT2/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for SCRT2***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/ZFHX2/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZFHX2***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/ZNF121/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZNF121***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/ZNF366/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZNF366***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/300bp_TFBS/ZNF770/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZNF770***\n",
      "Missing files recorded at /data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/DNABERT_Data/Somatic/TFBS_300bp/missing_files.txt\n"
     ]
    }
   ],
   "source": [
    "missing_files = []\n",
    "\n",
    "\n",
    "\n",
    "for index, row in top_models_df.iterrows():\n",
    "    intersected_data = f\"{intersected_base_path}/{row[0]}/intersected_vcf_data.pkl\"\n",
    "    print(intersected_data)\n",
    "    \n",
    "    try:\n",
    "        with open(intersected_data, \"rb\") as file:\n",
    "            loaded_dictionary = pickle.load(file)\n",
    "        print(f\"****All the VCF files are loaded for {row[0]}***\") \n",
    "\n",
    "        dnabert_raw_data = {}\n",
    "        for key, value in loaded_dictionary.items():\n",
    "            new_df, df_kmer = get_sequences(value, reference_fasta)\n",
    "            dnabert_raw_data[key] = new_df\n",
    "            # Apply mutations\n",
    "            # Assuming grouped_df is already created and sorted properly\n",
    "            # Modify your grouped DataFrame construction and apply the new function\n",
    "            out_folder_path = f\"{output_path}/{row[0]}/Patient_wise/{key}\"\n",
    "            if not os.path.exists(out_folder_path):\n",
    "                os.makedirs(out_folder_path)\n",
    "            df_kmer.to_csv(out_folder_path + \"/dev.tsv\", sep=\"\\t\", index=False)\n",
    "        \n",
    "        with open(f\"{output_path}/{row[0]}/raw_vcf_data.pkl\", \"wb\") as file:\n",
    "            pickle.dump(dnabert_raw_data, file)\n",
    "\n",
    "    except FileNotFoundError:\n",
    "        print(f\"File not found for {row[0]}, skipping...\")\n",
    "        missing_files.append(row[0])\n",
    "\n",
    "# Optionally, save the list of missing files/tags to a text file\n",
    "missing_files_path = f\"{output_path}/missing_files.txt\"\n",
    "with open(missing_files_path, \"w\") as f:\n",
    "    for tag in missing_files:\n",
    "        f.write(f\"{tag}\\n\")\n",
    "\n",
    "print(f\"Missing files recorded at {missing_files_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d00e115-459f-4e81-ae6a-e0b38ce19e74",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
