{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "98323797-7a8b-4c7a-b2b1-ed9a72bad141",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import plotly.graph_objects as go\n",
    "import seaborn as sns\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "a4508903-30fe-4403-b7a4-eb0d9f40cc69",
   "metadata": {},
   "outputs": [],
   "source": [
    "from lifelines import CoxPHFitter \n",
    "from lifelines import KaplanMeierFitter\n",
    "from lifelines.statistics import logrank_test, multivariate_logrank_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "7010ecb2-dfd9-4b50-bf2e-ca1c50531e57",
   "metadata": {},
   "outputs": [],
   "source": [
    "attention_score_path = \"/data/projects/GDC_Cancer_Wise/Brain/Data/Visualization_data/Acceptor/Prediction_result/atten_score80.npy\"\n",
    "prediction_path = \"/data/projects/GDC_Cancer_Wise/Brain/Data/Visualization_data/Acceptor/Prediction_result/pred_results.npy\"\n",
    "data_path = \"/data/projects/GDC_Cancer_Wise/Brain/Data/Visualization_data/Acceptor/dev.tsv\"\n",
    "raw_data_path = \"/data/projects/GDC_Cancer_Wise/Brain/Data/Visualization_data/Acceptor/raw_data.tsv\"\n",
    "data_statistics = \"/data/projects/GDC_Cancer_Wise/Brain/Data/Postprocessing_result/Acceptor/important_snp.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "ebb2e00d-343a-480c-abb2-0a6a1afd523b",
   "metadata": {},
   "outputs": [],
   "source": [
    "attention_data = np.load(attention_score_path)\n",
    "pred_result = np.load(prediction_path)\n",
    "data = pd.read_csv(data_path, sep=\"\\t\") \n",
    "raw_data = pd.read_csv(raw_data_path, sep=\"\\t\")\n",
    "df_snp = pd.read_csv(data_statistics, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "2209be19-7182-4084-a0e5-44b62886942d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.01490009, 0.00858043, 0.00673186, ..., 0.00683163, 0.00867788,\n",
       "        0.01443886],\n",
       "       [0.0142613 , 0.00871497, 0.00696233, ..., 0.02421607, 0.03467445,\n",
       "        0.0659325 ],\n",
       "       [0.02704413, 0.01780735, 0.01480795, ..., 0.08836011, 0.13167963,\n",
       "        0.26301744],\n",
       "       ...,\n",
       "       [0.01499418, 0.01007978, 0.00943549, ..., 0.00348013, 0.00170559,\n",
       "        0.00089852],\n",
       "       [0.01763524, 0.02300015, 0.01869645, ..., 0.03688714, 0.04093168,\n",
       "        0.06625   ],\n",
       "       [0.01098231, 0.01113181, 0.00911861, ..., 0.01283992, 0.0114428 ,\n",
       "        0.01403958]])"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attention_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b2cf59b0-95a6-42ac-94b0-7a21f6fc202c",
   "metadata": {},
   "outputs": [],
   "source": [
    "attention_data = attention_data[:, 1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "1a5cb102-908e-4e79-8440-0c6f654643eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "79"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(attention_data[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "007b45fd-9790-402d-834d-b969e3abc976",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([7.0022678e-01, 3.2903254e-02, 9.8029608e-01, ..., 9.6866232e-04,\n",
       "       7.9034662e-01, 3.8581815e-02], dtype=float32)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pred_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "857d38df-9a8f-4cd6-98f9-cd08aa92ba21",
   "metadata": {},
   "outputs": [],
   "source": [
    "#seq_wise_vis_dir = \"/data/projects/Enhancer/DNABERT_Output/Visualization_Enhancer_Encode/Predictions/V\"\n",
    "seqwise_attention_path = os.path.dirname(attention_score_path)+\"/Nucleotide_wise_attention_map\"\n",
    "if not os.path.exists(seqwise_attention_path):\n",
    "    os.makedirs(seqwise_attention_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "c107a815-d1f0-43d0-8b5f-548ed6fbd60f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1652"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(attention_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "d51731d2-2398-4edf-ab75-ca53248c7787",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Acceptor_start</th>\n",
       "      <th>Acceptor_end</th>\n",
       "      <th>varinat_start</th>\n",
       "      <th>variant_end</th>\n",
       "      <th>ref_neucleotide</th>\n",
       "      <th>alternative_neucleotide</th>\n",
       "      <th>reference_seq</th>\n",
       "      <th>alt_seq</th>\n",
       "      <th>Ref_Score</th>\n",
       "      <th>Alt_Score</th>\n",
       "      <th>patient_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000246151.9, ENST00000374524.1, ENST0000...</td>\n",
       "      <td>23786274</td>\n",
       "      <td>23786353</td>\n",
       "      <td>23786322</td>\n",
       "      <td>23786323</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ATAATTCATACCTTCTTTCCCTATTCCTGTCATCCTCTAGAATTTC...</td>\n",
       "      <td>ATAATTCATACCTTCTTTCCCTATTCCTGTCATCCTCTAGAATTTC...</td>\n",
       "      <td>0.700227</td>\n",
       "      <td>0.032903</td>\n",
       "      <td>['6fee80f8-11cb-4de7-aded-e586708d2368', 'f98d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000261464.10, ENST00000336184.6, ENST000...</td>\n",
       "      <td>211371261</td>\n",
       "      <td>211371340</td>\n",
       "      <td>211371265</td>\n",
       "      <td>211371266</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ACTAATTTTTTAAACATGATTATCCATTTTGTAATGAAAGGTTTTT...</td>\n",
       "      <td>ACTAGTTTTTTAAACATGATTATCCATTTTGTAATGAAAGGTTTTT...</td>\n",
       "      <td>0.980296</td>\n",
       "      <td>0.004916</td>\n",
       "      <td>['e849abef-b55e-4124-b760-09101a7c4390', 'b020...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000271588.9</td>\n",
       "      <td>186078066</td>\n",
       "      <td>186078145</td>\n",
       "      <td>186078129</td>\n",
       "      <td>186078130</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>TAAGATTAGAGGAGTAAACATAGTGGGATATTATTTTCAGGAGGGC...</td>\n",
       "      <td>TAAGATTAGAGGAGTAAACATAGTGGGATATTATTTTCAGGAGGGC...</td>\n",
       "      <td>0.992662</td>\n",
       "      <td>0.020286</td>\n",
       "      <td>['0332b017-17d5-4083-8fc4-9d6f8fdbbbde', '8faa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000292357.8, ENST00000338302.7, ENST0000...</td>\n",
       "      <td>156912729</td>\n",
       "      <td>156912808</td>\n",
       "      <td>156912797</td>\n",
       "      <td>156912798</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GCCAAGATGCCATTCTGAGTGAGCACCCCATTCCACACAGGAATCC...</td>\n",
       "      <td>GCCAAGATGCCATTCTGAGTGAGCACCCCATTCCACACAGGAATCC...</td>\n",
       "      <td>0.974504</td>\n",
       "      <td>0.078847</td>\n",
       "      <td>['5fd77ba9-5015-4d8b-86a0-582e5c76bdd6', '54c6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000294428.7</td>\n",
       "      <td>64804693</td>\n",
       "      <td>64804772</td>\n",
       "      <td>64804703</td>\n",
       "      <td>64804707</td>\n",
       "      <td>TAAA</td>\n",
       "      <td>TAA</td>\n",
       "      <td>TTTTCAAAATTAAACTGACAACGTTTTGTCATTTTCACAGAGCTCA...</td>\n",
       "      <td>TTTTCAAAATTAACTGACAACGTTTTGTCATTTTCACAGAGCTCAG...</td>\n",
       "      <td>0.999974</td>\n",
       "      <td>0.064063</td>\n",
       "      <td>['d85192d7-683f-4214-b8de-ae2e4ff2405c', '25be...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>821</th>\n",
       "      <td>821</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000438499.1</td>\n",
       "      <td>6668090</td>\n",
       "      <td>6668169</td>\n",
       "      <td>6668146</td>\n",
       "      <td>6668147</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>GATTCTCCTCCAATGCATCACATTTTTATTCCCCTGTCCCTGTGGG...</td>\n",
       "      <td>GATTCTCCTCCAATGCATCACATTTTTATTCCCCTGTCCCTGTGGG...</td>\n",
       "      <td>0.741388</td>\n",
       "      <td>0.010471</td>\n",
       "      <td>['e7b36f2d-9ab1-4a46-b669-65367b841752', 'c087...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>822</th>\n",
       "      <td>822</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000480178.1</td>\n",
       "      <td>9748034</td>\n",
       "      <td>9748113</td>\n",
       "      <td>9748044</td>\n",
       "      <td>9748045</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>AGGAAATGCTTGAGACAGTTCCAGCGGCCACATCTCCCTCTGTGTG...</td>\n",
       "      <td>AGGAAATGCTGGAGACAGTTCCAGCGGCCACATCTCCCTCTGTGTG...</td>\n",
       "      <td>0.636375</td>\n",
       "      <td>0.043710</td>\n",
       "      <td>['cf245938-dd0c-42fe-9a39-09f29ff12f23', '0bce...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>823</th>\n",
       "      <td>823</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000494525.1</td>\n",
       "      <td>21733718</td>\n",
       "      <td>21733797</td>\n",
       "      <td>21733771</td>\n",
       "      <td>21733772</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>TCCTCTTCAATTCTTTGATATCACTGAGTTCTTTTACCACTGGAAA...</td>\n",
       "      <td>TCCTCTTCAATTCTTTGATATCACTGAGTTCTTTTACCACTGGAAA...</td>\n",
       "      <td>0.702985</td>\n",
       "      <td>0.007422</td>\n",
       "      <td>['41d50a3c-49fc-4d8a-9c25-bbbcd3a5da6f']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>824</th>\n",
       "      <td>824</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000624824.1</td>\n",
       "      <td>15702688</td>\n",
       "      <td>15702767</td>\n",
       "      <td>15702716</td>\n",
       "      <td>15702717</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>TGGCACACGGCGCCTTGTGCACTTTCCAACATTTTGATGCTAACCG...</td>\n",
       "      <td>TGGCACACGGCGCCTTGTGCACTTTCCAGCATTTTGATGCTAACCG...</td>\n",
       "      <td>0.939159</td>\n",
       "      <td>0.000969</td>\n",
       "      <td>['0332b017-17d5-4083-8fc4-9d6f8fdbbbde', '8faa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>825</th>\n",
       "      <td>825</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000649154.1</td>\n",
       "      <td>120015469</td>\n",
       "      <td>120015548</td>\n",
       "      <td>120015503</td>\n",
       "      <td>120015504</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CTGGTCCGAGTTCCGGCGCGTGGACTTCGGGTCGGACGCCTGCAGG...</td>\n",
       "      <td>CTGGTCCGAGTTCCGGCGCGTGGACTTCGGGTCGTACGCCTGCAGG...</td>\n",
       "      <td>0.790346</td>\n",
       "      <td>0.038582</td>\n",
       "      <td>['55952783-8c2c-4528-a544-b4c962ab634a']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>826 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0   chr strand  \\\n",
       "0             0  chr1      +   \n",
       "1             1  chr1      +   \n",
       "2             2  chr1      +   \n",
       "3             3  chr1      +   \n",
       "4             4  chr1      +   \n",
       "..          ...   ...    ...   \n",
       "821         821  chrX      -   \n",
       "822         822  chrX      -   \n",
       "823         823  chrX      -   \n",
       "824         824  chrX      -   \n",
       "825         825  chrX      -   \n",
       "\n",
       "                                         Transcript_ID  Acceptor_start  \\\n",
       "0    ENST00000246151.9, ENST00000374524.1, ENST0000...        23786274   \n",
       "1    ENST00000261464.10, ENST00000336184.6, ENST000...       211371261   \n",
       "2                                    ENST00000271588.9       186078066   \n",
       "3    ENST00000292357.8, ENST00000338302.7, ENST0000...       156912729   \n",
       "4                                    ENST00000294428.7        64804693   \n",
       "..                                                 ...             ...   \n",
       "821                                  ENST00000438499.1         6668090   \n",
       "822                                  ENST00000480178.1         9748034   \n",
       "823                                  ENST00000494525.1        21733718   \n",
       "824                                  ENST00000624824.1        15702688   \n",
       "825                                  ENST00000649154.1       120015469   \n",
       "\n",
       "     Acceptor_end  varinat_start  variant_end ref_neucleotide  \\\n",
       "0        23786353       23786322     23786323               G   \n",
       "1       211371340      211371265    211371266               A   \n",
       "2       186078145      186078129    186078130               C   \n",
       "3       156912808      156912797    156912798               G   \n",
       "4        64804772       64804703     64804707            TAAA   \n",
       "..            ...            ...          ...             ...   \n",
       "821       6668169        6668146      6668147               G   \n",
       "822       9748113        9748044      9748045               T   \n",
       "823      21733797       21733771     21733772               G   \n",
       "824      15702767       15702716     15702717               A   \n",
       "825     120015548      120015503    120015504               G   \n",
       "\n",
       "    alternative_neucleotide  \\\n",
       "0                         A   \n",
       "1                         G   \n",
       "2                         T   \n",
       "3                         A   \n",
       "4                       TAA   \n",
       "..                      ...   \n",
       "821                       T   \n",
       "822                       G   \n",
       "823                       A   \n",
       "824                       G   \n",
       "825                       T   \n",
       "\n",
       "                                         reference_seq  \\\n",
       "0    ATAATTCATACCTTCTTTCCCTATTCCTGTCATCCTCTAGAATTTC...   \n",
       "1    ACTAATTTTTTAAACATGATTATCCATTTTGTAATGAAAGGTTTTT...   \n",
       "2    TAAGATTAGAGGAGTAAACATAGTGGGATATTATTTTCAGGAGGGC...   \n",
       "3    GCCAAGATGCCATTCTGAGTGAGCACCCCATTCCACACAGGAATCC...   \n",
       "4    TTTTCAAAATTAAACTGACAACGTTTTGTCATTTTCACAGAGCTCA...   \n",
       "..                                                 ...   \n",
       "821  GATTCTCCTCCAATGCATCACATTTTTATTCCCCTGTCCCTGTGGG...   \n",
       "822  AGGAAATGCTTGAGACAGTTCCAGCGGCCACATCTCCCTCTGTGTG...   \n",
       "823  TCCTCTTCAATTCTTTGATATCACTGAGTTCTTTTACCACTGGAAA...   \n",
       "824  TGGCACACGGCGCCTTGTGCACTTTCCAACATTTTGATGCTAACCG...   \n",
       "825  CTGGTCCGAGTTCCGGCGCGTGGACTTCGGGTCGGACGCCTGCAGG...   \n",
       "\n",
       "                                               alt_seq  Ref_Score  Alt_Score  \\\n",
       "0    ATAATTCATACCTTCTTTCCCTATTCCTGTCATCCTCTAGAATTTC...   0.700227   0.032903   \n",
       "1    ACTAGTTTTTTAAACATGATTATCCATTTTGTAATGAAAGGTTTTT...   0.980296   0.004916   \n",
       "2    TAAGATTAGAGGAGTAAACATAGTGGGATATTATTTTCAGGAGGGC...   0.992662   0.020286   \n",
       "3    GCCAAGATGCCATTCTGAGTGAGCACCCCATTCCACACAGGAATCC...   0.974504   0.078847   \n",
       "4    TTTTCAAAATTAACTGACAACGTTTTGTCATTTTCACAGAGCTCAG...   0.999974   0.064063   \n",
       "..                                                 ...        ...        ...   \n",
       "821  GATTCTCCTCCAATGCATCACATTTTTATTCCCCTGTCCCTGTGGG...   0.741388   0.010471   \n",
       "822  AGGAAATGCTGGAGACAGTTCCAGCGGCCACATCTCCCTCTGTGTG...   0.636375   0.043710   \n",
       "823  TCCTCTTCAATTCTTTGATATCACTGAGTTCTTTTACCACTGGAAA...   0.702985   0.007422   \n",
       "824  TGGCACACGGCGCCTTGTGCACTTTCCAGCATTTTGATGCTAACCG...   0.939159   0.000969   \n",
       "825  CTGGTCCGAGTTCCGGCGCGTGGACTTCGGGTCGTACGCCTGCAGG...   0.790346   0.038582   \n",
       "\n",
       "                                            patient_id  \n",
       "0    ['6fee80f8-11cb-4de7-aded-e586708d2368', 'f98d...  \n",
       "1    ['e849abef-b55e-4124-b760-09101a7c4390', 'b020...  \n",
       "2    ['0332b017-17d5-4083-8fc4-9d6f8fdbbbde', '8faa...  \n",
       "3    ['5fd77ba9-5015-4d8b-86a0-582e5c76bdd6', '54c6...  \n",
       "4    ['d85192d7-683f-4214-b8de-ae2e4ff2405c', '25be...  \n",
       "..                                                 ...  \n",
       "821  ['e7b36f2d-9ab1-4a46-b669-65367b841752', 'c087...  \n",
       "822  ['cf245938-dd0c-42fe-9a39-09f29ff12f23', '0bce...  \n",
       "823           ['41d50a3c-49fc-4d8a-9c25-bbbcd3a5da6f']  \n",
       "824  ['0332b017-17d5-4083-8fc4-9d6f8fdbbbde', '8faa...  \n",
       "825           ['55952783-8c2c-4528-a544-b4c962ab634a']  \n",
       "\n",
       "[826 rows x 15 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "8a8f561f-b5bb-4cb3-9b5b-f4f140ec7428",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "79\n"
     ]
    }
   ],
   "source": [
    "merged_list = list(zip(raw_data['reference_seq'], raw_data['alt_seq']))\n",
    "all_seq = [item.upper() for tup in merged_list for item in tup ]\n",
    "kmers = [list(seq) for seq in all_seq]\n",
    "\n",
    "num_columns = attention_data.shape[1]\n",
    "print(num_columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caa72526-8ddb-4e9e-961b-8765f0a7ab7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the length of the region to be visualized around the variant\n",
    "buffer_length = 10\n",
    "\n",
    "# Iterate over sequences and attention scores\n",
    "for i, sequence in enumerate(kmers):\n",
    "    # Calculate the relative positions of variant start and end within the sequence\n",
    "    relative_variant_start = raw_data['varinat_start'].iloc[i // 2] - raw_data['Acceptor_start'].iloc[i // 2]\n",
    "    relative_variant_end = raw_data['variant_end'].iloc[i // 2] - raw_data['Acceptor_start'].iloc[i // 2]\n",
    "    #print(relative_variant_start, relative_variant_end)\n",
    "\n",
    "    # Adjust indices for buffer length, ensuring they are within sequence boundaries\n",
    "    start_index = max(0, relative_variant_start - buffer_length)\n",
    "    end_index = min(len(sequence), relative_variant_end + buffer_length)\n",
    "\n",
    "    # Check if slicing range is valid\n",
    "    if start_index >= end_index:\n",
    "        print(\"Invalid slice range, skipping...\")\n",
    "        continue\n",
    "\n",
    "    # Slice the sequence and attention data\n",
    "    sliced_sequence = sequence[start_index:end_index]\n",
    "    sliced_attention = attention_data[i, start_index:end_index]\n",
    "    #print(sliced_sequence)\n",
    "    #print(sliced_attention)\n",
    "\n",
    "    # Plot heatmap\n",
    "    plt.figure(figsize=(18, 3))\n",
    "    sns.heatmap([sliced_attention], cmap='YlGnBu', cbar=True, annot=[sliced_sequence], fmt='', annot_kws={\"size\": 8})\n",
    "    original_start_index=  max(raw_data['Acceptor_start'].iloc[i//2], raw_data['varinat_start'].iloc[i//2] - buffer_length)\n",
    "    original_end_index=  min(raw_data['Acceptor_end'].iloc[i//2], raw_data['variant_end'].iloc[i//2] + buffer_length)\n",
    "    seq_type = \"Reference Sequence\" if i % 2 == 0 else \"Alternative Sequence\"\n",
    "    title = f\"{seq_type} around Variant {raw_data['chr'].iloc[i // 2]}:{original_start_index}-{original_end_index}\"\n",
    "    #print(title)\n",
    "    plt.title(title)\n",
    "    plt.savefig(seqwise_attention_path + \"/\" + f\"{title}.png\", bbox_inches='tight')\n",
    "    plt.close()\n",
    "    #input()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "5c03eccc-1ec8-40dd-a0f7-f849fa4cde64",
   "metadata": {},
   "outputs": [],
   "source": [
    "#seq_wise_vis_dir = \"/data/projects/Enhancer/DNABERT_Output/Visualization_Enhancer_Encode/Predictions/V\"\n",
    "seqwise_attention_path = os.path.dirname(attention_score_path)+\"/Nucleotide_wise_attention_map_paried_sequence\"\n",
    "if not os.path.exists(seqwise_attention_path):\n",
    "    os.makedirs(seqwise_attention_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "fc3f4664-46f3-4950-a0ed-e0b87fe18934",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Column</th>\n",
       "      <th>Max Value</th>\n",
       "      <th>Non-NaN Count</th>\n",
       "      <th>Non-NaN Indices</th>\n",
       "      <th>Percentage</th>\n",
       "      <th>Variant Region Difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr16_58267438_58267517_58267455_58267458</td>\n",
       "      <td>36.161531</td>\n",
       "      <td>32</td>\n",
       "      <td>[6, 7, 25, 32, 49, 63, 65, 80, 97, 115, 128, 1...</td>\n",
       "      <td>10.958904</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10_73797336_73797415_73797341_73797342</td>\n",
       "      <td>35.699415</td>\n",
       "      <td>3</td>\n",
       "      <td>[17, 219, 266]</td>\n",
       "      <td>1.027397</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr12_23190232_23190311_23190241_23190245</td>\n",
       "      <td>34.806988</td>\n",
       "      <td>4</td>\n",
       "      <td>[26, 163, 196, 290]</td>\n",
       "      <td>1.369863</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr4_52750790_52750869_52750817_52750818</td>\n",
       "      <td>34.397403</td>\n",
       "      <td>2</td>\n",
       "      <td>[76, 240]</td>\n",
       "      <td>0.684932</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3_175460004_175460083_175460009_175460012</td>\n",
       "      <td>34.386584</td>\n",
       "      <td>2</td>\n",
       "      <td>[122, 257]</td>\n",
       "      <td>0.684932</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3780</th>\n",
       "      <td>chr1_209623686_209623765_209623748_209623749</td>\n",
       "      <td>5.003059</td>\n",
       "      <td>6</td>\n",
       "      <td>[73, 108, 140, 156, 167, 206]</td>\n",
       "      <td>2.054795</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3781</th>\n",
       "      <td>chr6_29828502_29828581_29828528_29828529</td>\n",
       "      <td>5.002682</td>\n",
       "      <td>193</td>\n",
       "      <td>[0, 1, 2, 3, 6, 7, 10, 11, 13, 14, 15, 16, 17,...</td>\n",
       "      <td>66.095890</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3782</th>\n",
       "      <td>chr19_35259199_35259278_35259204_35259205</td>\n",
       "      <td>5.002419</td>\n",
       "      <td>1</td>\n",
       "      <td>[44]</td>\n",
       "      <td>0.342466</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3783</th>\n",
       "      <td>chr13_110492029_110492108_110492064_110492065</td>\n",
       "      <td>5.000697</td>\n",
       "      <td>22</td>\n",
       "      <td>[43, 52, 53, 64, 66, 72, 98, 106, 115, 118, 13...</td>\n",
       "      <td>7.534247</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3784</th>\n",
       "      <td>chr7_899451_899530_899526_899527</td>\n",
       "      <td>5.000186</td>\n",
       "      <td>1</td>\n",
       "      <td>[10]</td>\n",
       "      <td>0.342466</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3785 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Column  Max Value  Non-NaN Count  \\\n",
       "0         chr16_58267438_58267517_58267455_58267458  36.161531             32   \n",
       "1         chr10_73797336_73797415_73797341_73797342  35.699415              3   \n",
       "2         chr12_23190232_23190311_23190241_23190245  34.806988              4   \n",
       "3          chr4_52750790_52750869_52750817_52750818  34.397403              2   \n",
       "4      chr3_175460004_175460083_175460009_175460012  34.386584              2   \n",
       "...                                             ...        ...            ...   \n",
       "3780   chr1_209623686_209623765_209623748_209623749   5.003059              6   \n",
       "3781       chr6_29828502_29828581_29828528_29828529   5.002682            193   \n",
       "3782      chr19_35259199_35259278_35259204_35259205   5.002419              1   \n",
       "3783  chr13_110492029_110492108_110492064_110492065   5.000697             22   \n",
       "3784               chr7_899451_899530_899526_899527   5.000186              1   \n",
       "\n",
       "                                        Non-NaN Indices  Percentage  \\\n",
       "0     [6, 7, 25, 32, 49, 63, 65, 80, 97, 115, 128, 1...   10.958904   \n",
       "1                                        [17, 219, 266]    1.027397   \n",
       "2                                   [26, 163, 196, 290]    1.369863   \n",
       "3                                             [76, 240]    0.684932   \n",
       "4                                            [122, 257]    0.684932   \n",
       "...                                                 ...         ...   \n",
       "3780                      [73, 108, 140, 156, 167, 206]    2.054795   \n",
       "3781  [0, 1, 2, 3, 6, 7, 10, 11, 13, 14, 15, 16, 17,...   66.095890   \n",
       "3782                                               [44]    0.342466   \n",
       "3783  [43, 52, 53, 64, 66, 72, 98, 106, 115, 118, 13...    7.534247   \n",
       "3784                                               [10]    0.342466   \n",
       "\n",
       "      Variant Region Difference  \n",
       "0                             3  \n",
       "1                             1  \n",
       "2                             4  \n",
       "3                             1  \n",
       "4                             3  \n",
       "...                         ...  \n",
       "3780                          1  \n",
       "3781                          1  \n",
       "3782                          1  \n",
       "3783                          1  \n",
       "3784                          1  \n",
       "\n",
       "[3785 rows x 6 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_snp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "c24ee367-b74b-4919-9e52-58284bd58f7e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>_PATIENT</th>\n",
       "      <th>OS</th>\n",
       "      <th>OS.time</th>\n",
       "      <th>DSS</th>\n",
       "      <th>DSS.time</th>\n",
       "      <th>DFI</th>\n",
       "      <th>DFI.time</th>\n",
       "      <th>PFI</th>\n",
       "      <th>PFI.time</th>\n",
       "      <th>Redaction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-CS-4938-01</td>\n",
       "      <td>TCGA-CS-4938</td>\n",
       "      <td>0</td>\n",
       "      <td>3574.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3574.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>3574.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-CS-4941-01</td>\n",
       "      <td>TCGA-CS-4941</td>\n",
       "      <td>1</td>\n",
       "      <td>234.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>234.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>9.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-CS-4942-01</td>\n",
       "      <td>TCGA-CS-4942</td>\n",
       "      <td>1</td>\n",
       "      <td>1335.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1335.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1184.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-CS-4943-01</td>\n",
       "      <td>TCGA-CS-4943</td>\n",
       "      <td>1</td>\n",
       "      <td>1106.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1106.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1106.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-CS-4944-01</td>\n",
       "      <td>TCGA-CS-4944</td>\n",
       "      <td>0</td>\n",
       "      <td>1828.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1828.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1828.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524</th>\n",
       "      <td>TCGA-WY-A85A-01</td>\n",
       "      <td>TCGA-WY-A85A</td>\n",
       "      <td>0</td>\n",
       "      <td>1320.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1320.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1320.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>525</th>\n",
       "      <td>TCGA-WY-A85B-01</td>\n",
       "      <td>TCGA-WY-A85B</td>\n",
       "      <td>0</td>\n",
       "      <td>1393.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1393.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1393.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1393.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>526</th>\n",
       "      <td>TCGA-WY-A85C-01</td>\n",
       "      <td>TCGA-WY-A85C</td>\n",
       "      <td>0</td>\n",
       "      <td>1426.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1426.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>809.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>527</th>\n",
       "      <td>TCGA-WY-A85D-01</td>\n",
       "      <td>TCGA-WY-A85D</td>\n",
       "      <td>0</td>\n",
       "      <td>1147.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1147.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>887.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>528</th>\n",
       "      <td>TCGA-WY-A85E-01</td>\n",
       "      <td>TCGA-WY-A85E</td>\n",
       "      <td>0</td>\n",
       "      <td>633.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>633.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>532.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>529 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              sample      _PATIENT  OS  OS.time  DSS  DSS.time  DFI  DFI.time  \\\n",
       "0    TCGA-CS-4938-01  TCGA-CS-4938   0   3574.0  0.0    3574.0  NaN       NaN   \n",
       "1    TCGA-CS-4941-01  TCGA-CS-4941   1    234.0  1.0     234.0  NaN       NaN   \n",
       "2    TCGA-CS-4942-01  TCGA-CS-4942   1   1335.0  1.0    1335.0  NaN       NaN   \n",
       "3    TCGA-CS-4943-01  TCGA-CS-4943   1   1106.0  1.0    1106.0  NaN       NaN   \n",
       "4    TCGA-CS-4944-01  TCGA-CS-4944   0   1828.0  0.0    1828.0  NaN       NaN   \n",
       "..               ...           ...  ..      ...  ...       ...  ...       ...   \n",
       "524  TCGA-WY-A85A-01  TCGA-WY-A85A   0   1320.0  0.0    1320.0  NaN       NaN   \n",
       "525  TCGA-WY-A85B-01  TCGA-WY-A85B   0   1393.0  0.0    1393.0  0.0    1393.0   \n",
       "526  TCGA-WY-A85C-01  TCGA-WY-A85C   0   1426.0  0.0    1426.0  NaN       NaN   \n",
       "527  TCGA-WY-A85D-01  TCGA-WY-A85D   0   1147.0  0.0    1147.0  NaN       NaN   \n",
       "528  TCGA-WY-A85E-01  TCGA-WY-A85E   0    633.0  0.0     633.0  NaN       NaN   \n",
       "\n",
       "     PFI  PFI.time  Redaction  \n",
       "0      0    3574.0        NaN  \n",
       "1      1       9.0        NaN  \n",
       "2      1    1184.0        NaN  \n",
       "3      1    1106.0        NaN  \n",
       "4      0    1828.0        NaN  \n",
       "..   ...       ...        ...  \n",
       "524    0    1320.0        NaN  \n",
       "525    0    1393.0        NaN  \n",
       "526    1     809.0        NaN  \n",
       "527    1     887.0        NaN  \n",
       "528    1     532.0        NaN  \n",
       "\n",
       "[529 rows x 11 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.set_option('display.max_columns', None)\n",
    "df_clinical = pd.read_csv(\"/data/projects/GDC_Cancer_Wise/Brain/Clinical_files/TCGA_XENA_LGG_survival.txt\", sep=\"\\t\")\n",
    "#df_clinical = df_clinical[~(df_clinical.iloc[:, 3:] == \"'--\").all(axis=1)]\n",
    "df_clinical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "b5dc975e-cbb3-4b0e-9b4c-bd8d1f02453b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['case_id',\n",
       " 'case_submitter_id',\n",
       " 'project_id',\n",
       " 'age_at_index',\n",
       " 'age_is_obfuscated',\n",
       " 'cause_of_death',\n",
       " 'cause_of_death_source',\n",
       " 'country_of_residence_at_enrollment',\n",
       " 'days_to_birth',\n",
       " 'days_to_death',\n",
       " 'ethnicity',\n",
       " 'gender',\n",
       " 'occupation_duration_years',\n",
       " 'premature_at_birth',\n",
       " 'race',\n",
       " 'vital_status',\n",
       " 'weeks_gestation_at_birth',\n",
       " 'year_of_birth',\n",
       " 'year_of_death',\n",
       " 'adrenal_hormone',\n",
       " 'age_at_diagnosis',\n",
       " 'ajcc_clinical_m',\n",
       " 'ajcc_clinical_n',\n",
       " 'ajcc_clinical_stage',\n",
       " 'ajcc_clinical_t',\n",
       " 'ajcc_pathologic_m',\n",
       " 'ajcc_pathologic_n',\n",
       " 'ajcc_pathologic_stage',\n",
       " 'ajcc_pathologic_t',\n",
       " 'ajcc_staging_system_edition',\n",
       " 'anaplasia_present',\n",
       " 'anaplasia_present_type',\n",
       " 'ann_arbor_b_symptoms',\n",
       " 'ann_arbor_b_symptoms_described',\n",
       " 'ann_arbor_clinical_stage',\n",
       " 'ann_arbor_extranodal_involvement',\n",
       " 'ann_arbor_pathologic_stage',\n",
       " 'best_overall_response',\n",
       " 'breslow_thickness',\n",
       " 'burkitt_lymphoma_clinical_variant',\n",
       " 'child_pugh_classification',\n",
       " 'circumferential_resection_margin',\n",
       " 'classification_of_tumor',\n",
       " 'cog_liver_stage',\n",
       " 'cog_neuroblastoma_risk_group',\n",
       " 'cog_renal_stage',\n",
       " 'cog_rhabdomyosarcoma_risk_group',\n",
       " 'days_to_best_overall_response',\n",
       " 'days_to_diagnosis',\n",
       " 'days_to_last_follow_up',\n",
       " 'days_to_last_known_disease_status',\n",
       " 'days_to_recurrence',\n",
       " 'eln_risk_classification',\n",
       " 'enneking_msts_grade',\n",
       " 'enneking_msts_metastasis',\n",
       " 'enneking_msts_stage',\n",
       " 'enneking_msts_tumor_site',\n",
       " 'esophageal_columnar_dysplasia_degree',\n",
       " 'esophageal_columnar_metaplasia_present',\n",
       " 'figo_stage',\n",
       " 'figo_staging_edition_year',\n",
       " 'first_symptom_prior_to_diagnosis',\n",
       " 'gastric_esophageal_junction_involvement',\n",
       " 'gleason_grade_group',\n",
       " 'gleason_grade_tertiary',\n",
       " 'gleason_patterns_percent',\n",
       " 'goblet_cells_columnar_mucosa_present',\n",
       " 'greatest_tumor_dimension',\n",
       " 'gross_tumor_weight',\n",
       " 'icd_10_code',\n",
       " 'igcccg_stage',\n",
       " 'inpc_grade',\n",
       " 'inpc_histologic_group',\n",
       " 'inrg_stage',\n",
       " 'inss_stage',\n",
       " 'international_prognostic_index',\n",
       " 'irs_group',\n",
       " 'irs_stage',\n",
       " 'ishak_fibrosis_score',\n",
       " 'iss_stage',\n",
       " 'largest_extrapelvic_peritoneal_focus',\n",
       " 'last_known_disease_status',\n",
       " 'laterality',\n",
       " 'lymph_node_involved_site',\n",
       " 'lymph_nodes_positive',\n",
       " 'lymph_nodes_tested',\n",
       " 'lymphatic_invasion_present',\n",
       " 'margin_distance',\n",
       " 'margins_involved_site',\n",
       " 'masaoka_stage',\n",
       " 'medulloblastoma_molecular_classification',\n",
       " 'metastasis_at_diagnosis',\n",
       " 'metastasis_at_diagnosis_site',\n",
       " 'method_of_diagnosis',\n",
       " 'micropapillary_features',\n",
       " 'mitosis_karyorrhexis_index',\n",
       " 'mitotic_count',\n",
       " 'morphology',\n",
       " 'non_nodal_regional_disease',\n",
       " 'non_nodal_tumor_deposits',\n",
       " 'ovarian_specimen_status',\n",
       " 'ovarian_surface_involvement',\n",
       " 'papillary_renal_cell_type',\n",
       " 'percent_tumor_invasion',\n",
       " 'perineural_invasion_present',\n",
       " 'peripancreatic_lymph_nodes_positive',\n",
       " 'peripancreatic_lymph_nodes_tested',\n",
       " 'peritoneal_fluid_cytological_status',\n",
       " 'pregnant_at_diagnosis',\n",
       " 'primary_diagnosis',\n",
       " 'primary_disease',\n",
       " 'primary_gleason_grade',\n",
       " 'prior_malignancy',\n",
       " 'prior_treatment',\n",
       " 'progression_or_recurrence',\n",
       " 'residual_disease',\n",
       " 'satellite_nodule_present',\n",
       " 'secondary_gleason_grade',\n",
       " 'site_of_resection_or_biopsy',\n",
       " 'sites_of_involvement',\n",
       " 'supratentorial_localization',\n",
       " 'synchronous_malignancy',\n",
       " 'tissue_or_organ_of_origin',\n",
       " 'transglottic_extension',\n",
       " 'tumor_confined_to_organ_of_origin',\n",
       " 'tumor_depth',\n",
       " 'tumor_focality',\n",
       " 'tumor_grade',\n",
       " 'tumor_largest_dimension_diameter',\n",
       " 'tumor_regression_grade',\n",
       " 'tumor_stage',\n",
       " 'vascular_invasion_present',\n",
       " 'vascular_invasion_type',\n",
       " 'weiss_assessment_score',\n",
       " 'who_cns_grade',\n",
       " 'who_nte_grade',\n",
       " 'wilms_tumor_histologic_subtype',\n",
       " 'year_of_diagnosis',\n",
       " 'chemo_concurrent_to_radiation',\n",
       " 'days_to_treatment_end',\n",
       " 'days_to_treatment_start',\n",
       " 'initial_disease_status',\n",
       " 'number_of_cycles',\n",
       " 'reason_treatment_ended',\n",
       " 'regimen_or_line_of_therapy',\n",
       " 'route_of_administration',\n",
       " 'therapeutic_agents',\n",
       " 'treatment_anatomic_site',\n",
       " 'treatment_arm',\n",
       " 'treatment_dose',\n",
       " 'treatment_dose_units',\n",
       " 'treatment_effect',\n",
       " 'treatment_effect_indicator',\n",
       " 'treatment_frequency',\n",
       " 'treatment_intent_type',\n",
       " 'treatment_or_therapy',\n",
       " 'treatment_outcome',\n",
       " 'treatment_type']"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df_clinical.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "a670b6b8-ae64-446a-9568-0b60a4c47c88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Acceptor_start</th>\n",
       "      <th>Acceptor_end</th>\n",
       "      <th>varinat_start</th>\n",
       "      <th>variant_end</th>\n",
       "      <th>ref_neucleotide</th>\n",
       "      <th>alternative_neucleotide</th>\n",
       "      <th>reference_seq</th>\n",
       "      <th>alt_seq</th>\n",
       "      <th>Ref_Score</th>\n",
       "      <th>Alt_Score</th>\n",
       "      <th>patient_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000246151.9, ENST00000374524.1, ENST0000...</td>\n",
       "      <td>23786274</td>\n",
       "      <td>23786353</td>\n",
       "      <td>23786322</td>\n",
       "      <td>23786323</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ATAATTCATACCTTCTTTCCCTATTCCTGTCATCCTCTAGAATTTC...</td>\n",
       "      <td>ATAATTCATACCTTCTTTCCCTATTCCTGTCATCCTCTAGAATTTC...</td>\n",
       "      <td>0.700227</td>\n",
       "      <td>0.032903</td>\n",
       "      <td>['6fee80f8-11cb-4de7-aded-e586708d2368', 'f98d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000261464.10, ENST00000336184.6, ENST000...</td>\n",
       "      <td>211371261</td>\n",
       "      <td>211371340</td>\n",
       "      <td>211371265</td>\n",
       "      <td>211371266</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ACTAATTTTTTAAACATGATTATCCATTTTGTAATGAAAGGTTTTT...</td>\n",
       "      <td>ACTAGTTTTTTAAACATGATTATCCATTTTGTAATGAAAGGTTTTT...</td>\n",
       "      <td>0.980296</td>\n",
       "      <td>0.004916</td>\n",
       "      <td>['e849abef-b55e-4124-b760-09101a7c4390', 'b020...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000271588.9</td>\n",
       "      <td>186078066</td>\n",
       "      <td>186078145</td>\n",
       "      <td>186078129</td>\n",
       "      <td>186078130</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>TAAGATTAGAGGAGTAAACATAGTGGGATATTATTTTCAGGAGGGC...</td>\n",
       "      <td>TAAGATTAGAGGAGTAAACATAGTGGGATATTATTTTCAGGAGGGC...</td>\n",
       "      <td>0.992662</td>\n",
       "      <td>0.020286</td>\n",
       "      <td>['0332b017-17d5-4083-8fc4-9d6f8fdbbbde', '8faa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000292357.8, ENST00000338302.7, ENST0000...</td>\n",
       "      <td>156912729</td>\n",
       "      <td>156912808</td>\n",
       "      <td>156912797</td>\n",
       "      <td>156912798</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GCCAAGATGCCATTCTGAGTGAGCACCCCATTCCACACAGGAATCC...</td>\n",
       "      <td>GCCAAGATGCCATTCTGAGTGAGCACCCCATTCCACACAGGAATCC...</td>\n",
       "      <td>0.974504</td>\n",
       "      <td>0.078847</td>\n",
       "      <td>['5fd77ba9-5015-4d8b-86a0-582e5c76bdd6', '54c6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000294428.7</td>\n",
       "      <td>64804693</td>\n",
       "      <td>64804772</td>\n",
       "      <td>64804703</td>\n",
       "      <td>64804707</td>\n",
       "      <td>TAAA</td>\n",
       "      <td>TAA</td>\n",
       "      <td>TTTTCAAAATTAAACTGACAACGTTTTGTCATTTTCACAGAGCTCA...</td>\n",
       "      <td>TTTTCAAAATTAACTGACAACGTTTTGTCATTTTCACAGAGCTCAG...</td>\n",
       "      <td>0.999974</td>\n",
       "      <td>0.064063</td>\n",
       "      <td>['d85192d7-683f-4214-b8de-ae2e4ff2405c', '25be...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>821</th>\n",
       "      <td>821</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000438499.1</td>\n",
       "      <td>6668090</td>\n",
       "      <td>6668169</td>\n",
       "      <td>6668146</td>\n",
       "      <td>6668147</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>GATTCTCCTCCAATGCATCACATTTTTATTCCCCTGTCCCTGTGGG...</td>\n",
       "      <td>GATTCTCCTCCAATGCATCACATTTTTATTCCCCTGTCCCTGTGGG...</td>\n",
       "      <td>0.741388</td>\n",
       "      <td>0.010471</td>\n",
       "      <td>['e7b36f2d-9ab1-4a46-b669-65367b841752', 'c087...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>822</th>\n",
       "      <td>822</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000480178.1</td>\n",
       "      <td>9748034</td>\n",
       "      <td>9748113</td>\n",
       "      <td>9748044</td>\n",
       "      <td>9748045</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>AGGAAATGCTTGAGACAGTTCCAGCGGCCACATCTCCCTCTGTGTG...</td>\n",
       "      <td>AGGAAATGCTGGAGACAGTTCCAGCGGCCACATCTCCCTCTGTGTG...</td>\n",
       "      <td>0.636375</td>\n",
       "      <td>0.043710</td>\n",
       "      <td>['cf245938-dd0c-42fe-9a39-09f29ff12f23', '0bce...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>823</th>\n",
       "      <td>823</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000494525.1</td>\n",
       "      <td>21733718</td>\n",
       "      <td>21733797</td>\n",
       "      <td>21733771</td>\n",
       "      <td>21733772</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>TCCTCTTCAATTCTTTGATATCACTGAGTTCTTTTACCACTGGAAA...</td>\n",
       "      <td>TCCTCTTCAATTCTTTGATATCACTGAGTTCTTTTACCACTGGAAA...</td>\n",
       "      <td>0.702985</td>\n",
       "      <td>0.007422</td>\n",
       "      <td>['41d50a3c-49fc-4d8a-9c25-bbbcd3a5da6f']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>824</th>\n",
       "      <td>824</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000624824.1</td>\n",
       "      <td>15702688</td>\n",
       "      <td>15702767</td>\n",
       "      <td>15702716</td>\n",
       "      <td>15702717</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>TGGCACACGGCGCCTTGTGCACTTTCCAACATTTTGATGCTAACCG...</td>\n",
       "      <td>TGGCACACGGCGCCTTGTGCACTTTCCAGCATTTTGATGCTAACCG...</td>\n",
       "      <td>0.939159</td>\n",
       "      <td>0.000969</td>\n",
       "      <td>['0332b017-17d5-4083-8fc4-9d6f8fdbbbde', '8faa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>825</th>\n",
       "      <td>825</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000649154.1</td>\n",
       "      <td>120015469</td>\n",
       "      <td>120015548</td>\n",
       "      <td>120015503</td>\n",
       "      <td>120015504</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CTGGTCCGAGTTCCGGCGCGTGGACTTCGGGTCGGACGCCTGCAGG...</td>\n",
       "      <td>CTGGTCCGAGTTCCGGCGCGTGGACTTCGGGTCGTACGCCTGCAGG...</td>\n",
       "      <td>0.790346</td>\n",
       "      <td>0.038582</td>\n",
       "      <td>['55952783-8c2c-4528-a544-b4c962ab634a']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>826 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0   chr strand  \\\n",
       "0             0  chr1      +   \n",
       "1             1  chr1      +   \n",
       "2             2  chr1      +   \n",
       "3             3  chr1      +   \n",
       "4             4  chr1      +   \n",
       "..          ...   ...    ...   \n",
       "821         821  chrX      -   \n",
       "822         822  chrX      -   \n",
       "823         823  chrX      -   \n",
       "824         824  chrX      -   \n",
       "825         825  chrX      -   \n",
       "\n",
       "                                         Transcript_ID  Acceptor_start  \\\n",
       "0    ENST00000246151.9, ENST00000374524.1, ENST0000...        23786274   \n",
       "1    ENST00000261464.10, ENST00000336184.6, ENST000...       211371261   \n",
       "2                                    ENST00000271588.9       186078066   \n",
       "3    ENST00000292357.8, ENST00000338302.7, ENST0000...       156912729   \n",
       "4                                    ENST00000294428.7        64804693   \n",
       "..                                                 ...             ...   \n",
       "821                                  ENST00000438499.1         6668090   \n",
       "822                                  ENST00000480178.1         9748034   \n",
       "823                                  ENST00000494525.1        21733718   \n",
       "824                                  ENST00000624824.1        15702688   \n",
       "825                                  ENST00000649154.1       120015469   \n",
       "\n",
       "     Acceptor_end  varinat_start  variant_end ref_neucleotide  \\\n",
       "0        23786353       23786322     23786323               G   \n",
       "1       211371340      211371265    211371266               A   \n",
       "2       186078145      186078129    186078130               C   \n",
       "3       156912808      156912797    156912798               G   \n",
       "4        64804772       64804703     64804707            TAAA   \n",
       "..            ...            ...          ...             ...   \n",
       "821       6668169        6668146      6668147               G   \n",
       "822       9748113        9748044      9748045               T   \n",
       "823      21733797       21733771     21733772               G   \n",
       "824      15702767       15702716     15702717               A   \n",
       "825     120015548      120015503    120015504               G   \n",
       "\n",
       "    alternative_neucleotide  \\\n",
       "0                         A   \n",
       "1                         G   \n",
       "2                         T   \n",
       "3                         A   \n",
       "4                       TAA   \n",
       "..                      ...   \n",
       "821                       T   \n",
       "822                       G   \n",
       "823                       A   \n",
       "824                       G   \n",
       "825                       T   \n",
       "\n",
       "                                         reference_seq  \\\n",
       "0    ATAATTCATACCTTCTTTCCCTATTCCTGTCATCCTCTAGAATTTC...   \n",
       "1    ACTAATTTTTTAAACATGATTATCCATTTTGTAATGAAAGGTTTTT...   \n",
       "2    TAAGATTAGAGGAGTAAACATAGTGGGATATTATTTTCAGGAGGGC...   \n",
       "3    GCCAAGATGCCATTCTGAGTGAGCACCCCATTCCACACAGGAATCC...   \n",
       "4    TTTTCAAAATTAAACTGACAACGTTTTGTCATTTTCACAGAGCTCA...   \n",
       "..                                                 ...   \n",
       "821  GATTCTCCTCCAATGCATCACATTTTTATTCCCCTGTCCCTGTGGG...   \n",
       "822  AGGAAATGCTTGAGACAGTTCCAGCGGCCACATCTCCCTCTGTGTG...   \n",
       "823  TCCTCTTCAATTCTTTGATATCACTGAGTTCTTTTACCACTGGAAA...   \n",
       "824  TGGCACACGGCGCCTTGTGCACTTTCCAACATTTTGATGCTAACCG...   \n",
       "825  CTGGTCCGAGTTCCGGCGCGTGGACTTCGGGTCGGACGCCTGCAGG...   \n",
       "\n",
       "                                               alt_seq  Ref_Score  Alt_Score  \\\n",
       "0    ATAATTCATACCTTCTTTCCCTATTCCTGTCATCCTCTAGAATTTC...   0.700227   0.032903   \n",
       "1    ACTAGTTTTTTAAACATGATTATCCATTTTGTAATGAAAGGTTTTT...   0.980296   0.004916   \n",
       "2    TAAGATTAGAGGAGTAAACATAGTGGGATATTATTTTCAGGAGGGC...   0.992662   0.020286   \n",
       "3    GCCAAGATGCCATTCTGAGTGAGCACCCCATTCCACACAGGAATCC...   0.974504   0.078847   \n",
       "4    TTTTCAAAATTAACTGACAACGTTTTGTCATTTTCACAGAGCTCAG...   0.999974   0.064063   \n",
       "..                                                 ...        ...        ...   \n",
       "821  GATTCTCCTCCAATGCATCACATTTTTATTCCCCTGTCCCTGTGGG...   0.741388   0.010471   \n",
       "822  AGGAAATGCTGGAGACAGTTCCAGCGGCCACATCTCCCTCTGTGTG...   0.636375   0.043710   \n",
       "823  TCCTCTTCAATTCTTTGATATCACTGAGTTCTTTTACCACTGGAAA...   0.702985   0.007422   \n",
       "824  TGGCACACGGCGCCTTGTGCACTTTCCAGCATTTTGATGCTAACCG...   0.939159   0.000969   \n",
       "825  CTGGTCCGAGTTCCGGCGCGTGGACTTCGGGTCGTACGCCTGCAGG...   0.790346   0.038582   \n",
       "\n",
       "                                            patient_id  \n",
       "0    ['6fee80f8-11cb-4de7-aded-e586708d2368', 'f98d...  \n",
       "1    ['e849abef-b55e-4124-b760-09101a7c4390', 'b020...  \n",
       "2    ['0332b017-17d5-4083-8fc4-9d6f8fdbbbde', '8faa...  \n",
       "3    ['5fd77ba9-5015-4d8b-86a0-582e5c76bdd6', '54c6...  \n",
       "4    ['d85192d7-683f-4214-b8de-ae2e4ff2405c', '25be...  \n",
       "..                                                 ...  \n",
       "821  ['e7b36f2d-9ab1-4a46-b669-65367b841752', 'c087...  \n",
       "822  ['cf245938-dd0c-42fe-9a39-09f29ff12f23', '0bce...  \n",
       "823           ['41d50a3c-49fc-4d8a-9c25-bbbcd3a5da6f']  \n",
       "824  ['0332b017-17d5-4083-8fc4-9d6f8fdbbbde', '8faa...  \n",
       "825           ['55952783-8c2c-4528-a544-b4c962ab634a']  \n",
       "\n",
       "[826 rows x 15 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def KMPlot(df_kmf, saved_path):\n",
    "    kmf = KaplanMeierFitter()\n",
    "    T = df_kmf[\"time\"]\n",
    "    E = df_kmf[\"status\"]\n",
    "    plt.figure(figsize=(20,20))\n",
    "    for i in range(2):\n",
    "        T = df_kmf[df_kmf['label']==i]['time'].to_list()\n",
    "        E = df_kmf[df_kmf['label']==i]['status'].to_list()\n",
    "        if(i==1):\n",
    "            kmf.fit(T, E, label=\"Affected Patients\")\n",
    "        else if (i==0):\n",
    "            kmf.fit(T, E, label=\"Non-affected Patients\")\n",
    "        ax = kmf.plot_survival_function(ax=ax)\n",
    "    result = multivariate_logrank_test(df_kmf['time'], df_kmf['label'], df_kmf['status'])\n",
    "    #print(\"##\",result.test_statistic)\n",
    "    print(k,result.p_value)\n",
    "    #print(\"IO\",result.print_summary())\n",
    "    plt.title(\"KM Plot for cluster number =\"+str(k)+\"\\n[pvalue: {}]\".format(result.p_value));"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "53d52a3b-60df-4157-b151-71c996adba01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "20\n",
      "40\n",
      "60\n",
      "80\n",
      "100\n",
      "120\n",
      "140\n",
      "160\n",
      "180\n",
      "200\n",
      "220\n",
      "240\n",
      "260\n",
      "280\n",
      "300\n",
      "320\n",
      "340\n",
      "360\n",
      "380\n",
      "400\n",
      "420\n",
      "440\n",
      "460\n",
      "480\n",
      "500\n",
      "520\n",
      "540\n",
      "560\n",
      "580\n",
      "600\n",
      "620\n",
      "640\n",
      "660\n",
      "680\n",
      "700\n",
      "720\n",
      "740\n",
      "760\n",
      "780\n",
      "800\n",
      "820\n",
      "840\n",
      "860\n",
      "880\n",
      "900\n",
      "920\n",
      "940\n",
      "960\n",
      "980\n",
      "1000\n",
      "1020\n",
      "1040\n",
      "1060\n",
      "1080\n",
      "1100\n",
      "1120\n",
      "1140\n",
      "1160\n",
      "1180\n",
      "1200\n",
      "1220\n",
      "1240\n",
      "1260\n",
      "1280\n",
      "1300\n",
      "1320\n",
      "1340\n",
      "1360\n",
      "1380\n",
      "1400\n",
      "1420\n",
      "1440\n",
      "1460\n",
      "1480\n",
      "1500\n",
      "1520\n",
      "1540\n",
      "1560\n",
      "1580\n",
      "1600\n",
      "1620\n",
      "1640\n"
     ]
    }
   ],
   "source": [
    "buffer_length = 20  # Define buffer length\n",
    "\n",
    "# Iterate over pairs of sequences (reference and alternative) and attention scores\n",
    "for i in range(0, len(kmers), 2):\n",
    "    if(i%20==0):\n",
    "        print(i)\n",
    "    sequence_ref = kmers[i]\n",
    "    sequence_alt = kmers[i + 1]\n",
    "    index = raw_data['chr'].iloc[i // 2]+\"_\"+str(raw_data['Acceptor_start'].iloc[i // 2])+\"_\"+str(raw_data['Acceptor_end'].iloc[i // 2])+\"_\"+str(raw_data['varinat_start'].iloc[i // 2])+\"_\"+str(raw_data['variant_end'].iloc[i // 2])\n",
    "    #print(index)\n",
    "    percentage_value = df_snp[df_snp['Column']==index]['Percentage'].values[0]\n",
    "    #print(percentage_value)\n",
    "    lower_bound = math.floor(percentage_value)\n",
    "    upper_bound = math.ceil(percentage_value)\n",
    "    #print(lower_bound , upper_bound)\n",
    "    directory_name = f\"{lower_bound}_{upper_bound}\"\n",
    "    #print(\"##\", directory_name)\n",
    "\n",
    "    # Calculate relative positions of variant start and end within the sequence\n",
    "    relative_variant_start = raw_data['varinat_start'].iloc[i // 2] - raw_data['Acceptor_start'].iloc[i // 2]\n",
    "    relative_variant_end = raw_data['variant_end'].iloc[i // 2] - raw_data['Acceptor_start'].iloc[i // 2]\n",
    "\n",
    "    # Adjust indices for buffer length, ensuring they are within sequence boundaries\n",
    "    start_index = max(0, relative_variant_start - buffer_length)\n",
    "    end_index = min(len(sequence_ref), relative_variant_end + buffer_length)\n",
    "\n",
    "    # Check if slicing range is valid\n",
    "    if start_index >= end_index:\n",
    "        print(\"Invalid slice range, skipping...\")\n",
    "        continue\n",
    "\n",
    "    # Slice the sequences and ` data\n",
    "    sliced_sequence_ref = sequence_ref[start_index:end_index]\n",
    "    sliced_sequence_alt = sequence_alt[start_index:end_index]\n",
    "    sliced_attention_ref = attention_data[i, start_index:end_index]\n",
    "    sliced_attention_alt = attention_data[i + 1, start_index:end_index]\n",
    "    \n",
    "    original_start_index=  max(raw_data['Acceptor_start'].iloc[i//2], raw_data['varinat_start'].iloc[i//2] - buffer_length)\n",
    "    original_end_index=  min(raw_data['Acceptor_end'].iloc[i//2], raw_data['variant_end'].iloc[i//2] + buffer_length)\n",
    "    \n",
    "    sub_seqwise_attention_path = seqwise_attention_path + \"/\" +directory_name\n",
    "    if not os.path.exists(sub_seqwise_attention_path):\n",
    "        os.makedirs(sub_seqwise_attention_path)\n",
    "\n",
    "    \n",
    "    # Determine the number of ticks needed, with a tick for every 5 nucleotides\n",
    "    num_ticks = (original_end_index - original_start_index) // 5 + 1  # Ensure we include the end index\n",
    "\n",
    "    # Generate the tick positions and labels\n",
    "    tick_positions = [i * 5 for i in range(num_ticks)]  # Positions in the sequence\n",
    "    tick_labels = [original_start_index + pos for pos in tick_positions]  # Map to actual nucleotide positions\n",
    "\n",
    "    # Ensure tick positions and labels do not exceed the length of the sequence or the end index\n",
    "    tick_positions = [pos for pos, label in zip(tick_positions, tick_labels) if label <= original_end_index]\n",
    "    tick_labels = [label for label in tick_labels if label <= original_end_index]\n",
    "    # print(plot_range)\n",
    "    # print(tick_positions)\n",
    "    # print(tick_labels)\n",
    "    # Create a new figure for each iteration\n",
    "    fig, axs = plt.subplots(2, 1, figsize=(18, 3))\n",
    "    plt.subplots_adjust(hspace=1)  # Adjust spacing between subplots\n",
    "    \n",
    "    #y_position = 1.08\n",
    "    \n",
    "        # Add text for the Acceptor Region\n",
    "    plt.text(\n",
    "        x=0.5, \n",
    "        y=1.08, \n",
    "        s=f\"Acceptor Region: {raw_data['Acceptor_start'].iloc[i // 2]}-{raw_data['Acceptor_end'].iloc[i // 2]}\", \n",
    "        horizontalalignment='center', \n",
    "        verticalalignment='center', \n",
    "        transform=plt.gcf().transFigure\n",
    "    )\n",
    "    #y_position -= 0.04  # Adjust the y position for the next line\n",
    "\n",
    "    # Add text for the Variant Region\n",
    "    # Replace variant_region_info with the appropriate variable or value\n",
    "    plt.text(\n",
    "        x=0.5, \n",
    "        y=1.04, \n",
    "        s=f\"Variant Region: {raw_data['varinat_start'].iloc[i // 2]}-{raw_data['variant_end'].iloc[i // 2]}\", \n",
    "        horizontalalignment='center', \n",
    "        verticalalignment='center', \n",
    "        transform=plt.gcf().transFigure\n",
    "    )\n",
    "    #y_position -= 0.04  # Adjust the y position for the next line\n",
    "\n",
    "    # Add text for the Region Plotted\n",
    "    # Replace plotted_region_info with the appropriate variable or value\n",
    "    plt.text(\n",
    "        x=0.5, \n",
    "        y=1.00, \n",
    "        s=f\"Region Plotted: {original_start_index}-{original_end_index}\", \n",
    "        horizontalalignment='center', \n",
    "        verticalalignment='center', \n",
    "        transform=plt.gcf().transFigure\n",
    "    )   # Create subplots\n",
    "    #y_position -= 0.08  # Adjust the y position for the next line\n",
    "    \n",
    "    \n",
    "    # Plot heatmap for reference sequence\n",
    "    sns.heatmap([sliced_attention_ref], cmap='YlGnBu', cbar=True, annot=[sliced_sequence_ref], fmt='', annot_kws={\"size\": 8}, ax=axs[0])\n",
    "    axs[0].set_xticks(tick_positions)  # Set the tick positions on the x-axis\n",
    "    axs[0].set_xticklabels(tick_labels)  # Set the tick labels\n",
    "\n",
    "    # Plot heatmap for alternative sequence\n",
    "    sns.heatmap([sliced_attention_alt], cmap='YlGnBu', cbar=True, annot=[sliced_sequence_alt], fmt='', annot_kws={\"size\": 8}, ax=axs[1])\n",
    "    axs[1].set_title(f\"Alternative Sequence {raw_data['chr'].iloc[i // 2]}:{original_start_index}-{original_end_index}\")\n",
    "    axs[1].set_xticks(tick_positions)  # Set the tick positions on the x-axis for the second subplot\n",
    "    axs[1].set_xticklabels(tick_labels)  # Set the tick labels\n",
    "\n",
    "\n",
    "    # Save the plot\n",
    "    plt.savefig(sub_seqwise_attention_path + \"/\" + f\"Paired_sequence_{raw_data['chr'].iloc[i // 2]}_{original_start_index}_{original_end_index}.png\", bbox_inches='tight')\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60cc1b09-b9ae-4108-ad47-fd930f8982dd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f64caf96-d7c1-4af1-97d0-ef2ba60355a6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e448456a-d6bd-4aa3-9614-2811136765c5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
