{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fbbadeb3-81ba-4b36-8aea-3a3d64520f94",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d33e9c15-af01-42c2-b68f-496b78099c1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"//data/private/pdutta/VCF_data/MergedTransVCF/merged_chr2.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1bccceca-7247-4a08-a0de-a5594ad049ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>transcription_start_site</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>transcript_start</th>\n",
       "      <th>transcript_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>mutations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>234</td>\n",
       "      <td>21044073</td>\n",
       "      <td>chr2</td>\n",
       "      <td>21001429</td>\n",
       "      <td>21044073</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSG00000084674</td>\n",
       "      <td>ENST00000233242</td>\n",
       "      <td>rs1663876300, rs1663674989, rs886055565, rs166...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>242</td>\n",
       "      <td>19358623</td>\n",
       "      <td>chr2</td>\n",
       "      <td>19351485</td>\n",
       "      <td>19358623</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSG00000143867</td>\n",
       "      <td>ENST00000272223</td>\n",
       "      <td>rs1041878582, rs1664849630, rs1664939091, rs16...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>250</td>\n",
       "      <td>25121673</td>\n",
       "      <td>chr2</td>\n",
       "      <td>25121673</td>\n",
       "      <td>25154361</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000084710</td>\n",
       "      <td>ENST00000264719</td>\n",
       "      <td>rs747999572, rs959120992, rs1346903942, rs1190...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>263</td>\n",
       "      <td>11539843</td>\n",
       "      <td>chr2</td>\n",
       "      <td>11539843</td>\n",
       "      <td>11588243</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000196208</td>\n",
       "      <td>ENST00000263834</td>\n",
       "      <td>rs373909340, rs1572797221, rs1237923365, rs119...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>264</td>\n",
       "      <td>11556353</td>\n",
       "      <td>chr2</td>\n",
       "      <td>11556353</td>\n",
       "      <td>11642786</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000196208</td>\n",
       "      <td>ENST00000234142</td>\n",
       "      <td>rs750767918, rs373909340, rs1572797221, rs1572...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  transcription_start_site chromosome_name  transcript_start  \\\n",
       "0         234                  21044073            chr2          21001429   \n",
       "1         242                  19358623            chr2          19351485   \n",
       "2         250                  25121673            chr2          25121673   \n",
       "3         263                  11539843            chr2          11539843   \n",
       "4         264                  11556353            chr2          11556353   \n",
       "\n",
       "   transcript_end  strand  ensembl_gene_id ensembl_transcript_id  \\\n",
       "0        21044073      -1  ENSG00000084674       ENST00000233242   \n",
       "1        19358623      -1  ENSG00000143867       ENST00000272223   \n",
       "2        25154361       1  ENSG00000084710       ENST00000264719   \n",
       "3        11588243       1  ENSG00000196208       ENST00000263834   \n",
       "4        11642786       1  ENSG00000196208       ENST00000234142   \n",
       "\n",
       "                                           mutations  \n",
       "0  rs1663876300, rs1663674989, rs886055565, rs166...  \n",
       "1  rs1041878582, rs1664849630, rs1664939091, rs16...  \n",
       "2  rs747999572, rs959120992, rs1346903942, rs1190...  \n",
       "3  rs373909340, rs1572797221, rs1237923365, rs119...  \n",
       "4  rs750767918, rs373909340, rs1572797221, rs1572...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c740f0a6-930f-4a1c-b864-65d094436a61",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_202846/1371329063.py:1: DtypeWarning: Columns (23,24,28) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df_DBSNP = pd.read_csv(\"/data/private/pdutta/VCF_data/Merged_DBSNP_Chrwise/chr2_data_merged.csv\")\n"
     ]
    }
   ],
   "source": [
    "df_DBSNP = pd.read_csv(\"/data/private/pdutta/VCF_data/Merged_DBSNP_Chrwise/chr2_data_merged.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "89d1b91e-47be-45d4-98e5-f161994cc962",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>ref</th>\n",
       "      <th>altCount</th>\n",
       "      <th>alts</th>\n",
       "      <th>shiftBases</th>\n",
       "      <th>freqSourceCount</th>\n",
       "      <th>minorAlleleFreq</th>\n",
       "      <th>...</th>\n",
       "      <th>alleleTotals</th>\n",
       "      <th>soTermCount</th>\n",
       "      <th>soTerms</th>\n",
       "      <th>clinVarCount</th>\n",
       "      <th>clinVarAccs</th>\n",
       "      <th>clinVarSigs</th>\n",
       "      <th>submitterCount</th>\n",
       "      <th>submitters</th>\n",
       "      <th>pubMedIdCount</th>\n",
       "      <th>pubMedIds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>31032285</td>\n",
       "      <td>31032286</td>\n",
       "      <td>rs1000000004</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>C,</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>-inf,0.000213523,0.000313574,-inf,-inf,-inf,-i...</td>\n",
       "      <td>...</td>\n",
       "      <td>0,14050,264690,0,0,0,0,0,0,16760,0,0,140186,0,...</td>\n",
       "      <td>1</td>\n",
       "      <td>1627,</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>GNOMAD,HUMAN_LONGEVITY,TOMMO_GENOMICS,TOPMED,</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>95169762</td>\n",
       "      <td>95169763</td>\n",
       "      <td>rs1000000009</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>G,</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>-inf,0,7.55601e-06,-inf,-inf,-inf,-inf,-inf,-i...</td>\n",
       "      <td>...</td>\n",
       "      <td>0,14050,264690,0,0,0,0,0,0,0,0,0,139988,0,0,0,...</td>\n",
       "      <td>1</td>\n",
       "      <td>1627,</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>GNOMAD,TOPMED,</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>26429613</td>\n",
       "      <td>26429614</td>\n",
       "      <td>rs1000000016</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>G,</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>-inf,0,3.778e-06,-inf,-inf,-inf,-inf,-inf,-inf...</td>\n",
       "      <td>...</td>\n",
       "      <td>0,14050,264690,0,0,0,0,0,0,0,0,0,140184,0,0,0,...</td>\n",
       "      <td>1</td>\n",
       "      <td>1627,</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>GNOMAD,HUMAN_LONGEVITY,TOPMED,</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>12732230</td>\n",
       "      <td>12732231</td>\n",
       "      <td>rs1000000040</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>C,</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>-inf,0,3.778e-06,-inf,-inf,-inf,-inf,-inf,-inf...</td>\n",
       "      <td>...</td>\n",
       "      <td>0,14050,264690,0,0,0,0,0,0,0,0,0,140182,0,0,0,...</td>\n",
       "      <td>1</td>\n",
       "      <td>1627,</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>GNOMAD,TOPMED,</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>12572350</td>\n",
       "      <td>12572351</td>\n",
       "      <td>rs1000000045</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>C,</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>-inf,7.11744e-05,3.778e-05,-inf,-inf,-inf,-inf...</td>\n",
       "      <td>...</td>\n",
       "      <td>0,14050,264690,0,0,0,0,0,0,0,0,0,140178,0,0,0,...</td>\n",
       "      <td>2</td>\n",
       "      <td>2152,1627,</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>GNOMAD,TOPMED,</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome     start       end          name ref  altCount alts  shiftBases  \\\n",
       "0       chr2  31032285  31032286  rs1000000004   T         1   C,           0   \n",
       "1       chr2  95169762  95169763  rs1000000009   A         1   G,           0   \n",
       "2       chr2  26429613  26429614  rs1000000016   A         1   G,           0   \n",
       "3       chr2  12732230  12732231  rs1000000040   A         1   C,           0   \n",
       "4       chr2  12572350  12572351  rs1000000045   T         1   C,           0   \n",
       "\n",
       "   freqSourceCount                                    minorAlleleFreq  ...  \\\n",
       "0               31  -inf,0.000213523,0.000313574,-inf,-inf,-inf,-i...  ...   \n",
       "1               31  -inf,0,7.55601e-06,-inf,-inf,-inf,-inf,-inf,-i...  ...   \n",
       "2               31  -inf,0,3.778e-06,-inf,-inf,-inf,-inf,-inf,-inf...  ...   \n",
       "3               31  -inf,0,3.778e-06,-inf,-inf,-inf,-inf,-inf,-inf...  ...   \n",
       "4               31  -inf,7.11744e-05,3.778e-05,-inf,-inf,-inf,-inf...  ...   \n",
       "\n",
       "                                        alleleTotals soTermCount     soTerms  \\\n",
       "0  0,14050,264690,0,0,0,0,0,0,16760,0,0,140186,0,...           1       1627,   \n",
       "1  0,14050,264690,0,0,0,0,0,0,0,0,0,139988,0,0,0,...           1       1627,   \n",
       "2  0,14050,264690,0,0,0,0,0,0,0,0,0,140184,0,0,0,...           1       1627,   \n",
       "3  0,14050,264690,0,0,0,0,0,0,0,0,0,140182,0,0,0,...           1       1627,   \n",
       "4  0,14050,264690,0,0,0,0,0,0,0,0,0,140178,0,0,0,...           2  2152,1627,   \n",
       "\n",
       "  clinVarCount clinVarAccs  clinVarSigs  submitterCount  \\\n",
       "0            0         NaN          NaN               4   \n",
       "1            0         NaN          NaN               2   \n",
       "2            0         NaN          NaN               3   \n",
       "3            0         NaN          NaN               2   \n",
       "4            0         NaN          NaN               2   \n",
       "\n",
       "                                      submitters pubMedIdCount pubMedIds  \n",
       "0  GNOMAD,HUMAN_LONGEVITY,TOMMO_GENOMICS,TOPMED,             0       NaN  \n",
       "1                                 GNOMAD,TOPMED,             0       NaN  \n",
       "2                 GNOMAD,HUMAN_LONGEVITY,TOPMED,             0       NaN  \n",
       "3                                 GNOMAD,TOPMED,             0       NaN  \n",
       "4                                 GNOMAD,TOPMED,             0       NaN  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_DBSNP.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ac54b3bb-e527-4b3f-86e0-6ee034714a24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>ref</th>\n",
       "      <th>altCount</th>\n",
       "      <th>alts</th>\n",
       "      <th>shiftBases</th>\n",
       "      <th>freqSourceCount</th>\n",
       "      <th>minorAlleleFreq</th>\n",
       "      <th>...</th>\n",
       "      <th>alleleTotals</th>\n",
       "      <th>soTermCount</th>\n",
       "      <th>soTerms</th>\n",
       "      <th>clinVarCount</th>\n",
       "      <th>clinVarAccs</th>\n",
       "      <th>clinVarSigs</th>\n",
       "      <th>submitterCount</th>\n",
       "      <th>submitters</th>\n",
       "      <th>pubMedIdCount</th>\n",
       "      <th>pubMedIds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>38774325</th>\n",
       "      <td>chr2</td>\n",
       "      <td>21031412</td>\n",
       "      <td>21031413</td>\n",
       "      <td>rs1663876300</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>T,</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>-inf,0,-inf,-inf,-inf,-inf,-inf,-inf,-inf,-inf...</td>\n",
       "      <td>...</td>\n",
       "      <td>0,10680,0,0,0,0,0,0,0,0,0,0,140150,0,0,0,0,0,0...</td>\n",
       "      <td>1</td>\n",
       "      <td>1627,</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>GNOMAD,</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chromosome     start       end          name ref  altCount alts  \\\n",
       "38774325       chr2  21031412  21031413  rs1663876300   C         1   T,   \n",
       "\n",
       "          shiftBases  freqSourceCount  \\\n",
       "38774325           0               31   \n",
       "\n",
       "                                            minorAlleleFreq  ...  \\\n",
       "38774325  -inf,0,-inf,-inf,-inf,-inf,-inf,-inf,-inf,-inf...  ...   \n",
       "\n",
       "                                               alleleTotals soTermCount  \\\n",
       "38774325  0,10680,0,0,0,0,0,0,0,0,0,0,140150,0,0,0,0,0,0...           1   \n",
       "\n",
       "          soTerms clinVarCount clinVarAccs  clinVarSigs  submitterCount  \\\n",
       "38774325    1627,            0         NaN          NaN               1   \n",
       "\n",
       "          submitters pubMedIdCount pubMedIds  \n",
       "38774325     GNOMAD,             0       NaN  \n",
       "\n",
       "[1 rows x 29 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_DBSNP.loc[df_DBSNP['name'] == 'rs1663876300']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e3f4e3e5-e54b-4b2a-a31e-6f4b59de8564",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['chromosome',\n",
       " 'start',\n",
       " 'end',\n",
       " 'name',\n",
       " 'ref',\n",
       " 'altCount',\n",
       " 'alts',\n",
       " 'shiftBases',\n",
       " 'freqSourceCount',\n",
       " 'minorAlleleFreq',\n",
       " 'majorAllele',\n",
       " 'minorAllele',\n",
       " 'maxFuncImpact',\n",
       " 'class',\n",
       " 'ucscNotes',\n",
       " '_dataOffset',\n",
       " '_dataLen',\n",
       " 'freqSourceCount.1',\n",
       " 'alleleCounts',\n",
       " 'alleleTotals',\n",
       " 'soTermCount',\n",
       " 'soTerms',\n",
       " 'clinVarCount',\n",
       " 'clinVarAccs',\n",
       " 'clinVarSigs',\n",
       " 'submitterCount',\n",
       " 'submitters',\n",
       " 'pubMedIdCount',\n",
       " 'pubMedIds']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df_DBSNP.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3fcddd7-e89b-41be-b4ff-1d5c6b178462",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
