{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9807e3a4-8edb-46a2-ad4c-88777af9fadc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, glob\n",
    "import pandas as pd\n",
    "from io import StringIO\n",
    "import numpy as np\n",
    "import pysam\n",
    "import pybedtools \n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "pybedtools.helpers.set_tempdir(\"/data/projects/temp\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2505ee28-f7a7-474a-a4b8-755989f3f70e",
   "metadata": {},
   "outputs": [],
   "source": [
    "non_coding_region = \"donor\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6c3ba47b-0edf-43de-bab5-bc0edd7f91b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '/data/projects/DNABERT_snv/Manuscript_11_2023/Intersected_data/{}_ICGC_intersected.tsv'.format(non_coding_region)\n",
    "output_path = \"/data/projects/DNABERT_snv/Manuscript_11_2023/DNABERT_data/{}\".format(non_coding_region) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aafe0877-b7be-4e5e-a73f-00dab9627b51",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(output_path):\n",
    "    os.makedirs(output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cc89bac5-1528-4192-95d4-77bda6a6e1dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def seq2kmer(seq):\n",
    "    \"\"\"\n",
    "    Convert original sequence to kmers\n",
    "    \n",
    "    Arguments:\n",
    "    seq -- str, original sequence.\n",
    "    k -- int, kmer of length k specified.\n",
    "    \n",
    "    Returns:\n",
    "    kmers -- str, kmers separated by space\n",
    "    \"\"\"\n",
    "    k=6\n",
    "    kmer = [seq[x:x+k] for x in range(len(seq)+1-k)]\n",
    "    kmers = \" \".join(kmer)\n",
    "    return kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "19aaa7a0-7309-4e97-a93e-dded3164e07f",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = SeqIO.to_dict(SeqIO.parse(\"/data/projects/Resources/Gencode_genome_annotation/GRCh38.primary_assembly.genome.fa\", \"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3bcc2b33-f043-4016-9aba-79948c239d76",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5792692, 22)\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(data_path, sep= '\\t')\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "0ded228f-2999-482f-9e5a-197b732e0cfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5786746, 11)\n",
      "(5783396, 11)\n"
     ]
    }
   ],
   "source": [
    "# df = pd.read_csv(data_path, sep= '\\t')\n",
    "# print(df.shape)\n",
    "\n",
    "# df = df[df['No_of_variations']<3].reset_index(drop=True)\n",
    "# print(df.shape)\n",
    "# df = df.loc[df['Alternative_Allele']== '.'].reset_index(drop=True)\n",
    "# df\n",
    "#df= df[df['strand']==\"-\"]\n",
    "\n",
    "\n",
    "data= []\n",
    "#df = df.head(10)\n",
    "# Iterate over DataFrame rows\n",
    "for idx, row in df.iterrows():\n",
    "    try:\n",
    "        #print(row)Reference_Allele\tAlternative_Allele\n",
    "        alts = row['Alternative_Allele'].rstrip(',').split(',')\n",
    "        ref_seq = str(genome[row['chr_name']].seq[ row['Donor_start']:row['Donor_end']])\n",
    "\n",
    "        # Calculate the variant position relative to the fetched sequence\n",
    "        variant_pos_start = row['Start'] - row['Donor_start']\n",
    "        variant_pos_end = row['End']  - row['Donor_start']\n",
    "        # print(variant_pos_start, variant_pos_end, alts,  row['Reference_Allele'])\n",
    "        # print(ref_seq)\n",
    "        # if row['strand'] == '-':\n",
    "        #     ref_seq = str(Seq(ref_seq).reverse_complement())\n",
    "        #print(ref_seq)\n",
    "\n",
    "\n",
    "        #Replace the reference nucleotide with the alternate nucleotide to get the alternate sequence\n",
    "        for alt in alts:\n",
    "            if alt == \".\":  # Deletion\n",
    "                # Fetch additional bases from genome to maintain sequence length\n",
    "                extra_bases = str(genome[row['chr_name']].seq[row['Donor_end']:row['Donor_end'] + len(row['Reference_Allele'])])\n",
    "                #print(len(row['Reference_Nucleotide']))\n",
    "                alt_seq = ref_seq[:variant_pos_start] + ref_seq[variant_pos_end:] + extra_bases\n",
    "            else:  # SNPs, Insertions, and partial Deletions\n",
    "                extra_bases = str(genome[row['chr_name']].seq[row['Donor_end']:row['Donor_end'] + len(row['Reference_Allele']) - len(alt)])\n",
    "                #print(extra_bases)\n",
    "                if(variant_pos_start>=0):\n",
    "                    alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:] + extra_bases\n",
    "                else:\n",
    "                    continue\n",
    "            #print(alt_seq)\n",
    "            #Append to the list as a dictionary\n",
    "            data.append({\n",
    "                'chr': row['chr_name'],\n",
    "                'strand': row['strand'],\n",
    "                'Donor_coordinates': str(row['Donor_start'])+'-'+str(row['Donor_end']),\n",
    "                'Ensemble_Transcript_ID': row['ENSEMBL_Transcript_ID'],\n",
    "                'Mutation_ID': row['Mutation_ID'],\n",
    "                'varinat_start':  row['Start'],\n",
    "                'variant_end':  row['End'],\n",
    "                'ref_neucleotide': row['Reference_Allele'],\n",
    "                'alternative_neucleotide': alt,\n",
    "                'reference_seq': ref_seq,\n",
    "                'alt_seq': alt_seq\n",
    "            })\n",
    "            \n",
    "    except KeyError:\n",
    "        print(\"NA\")\n",
    "        # # If sequence is not present, append a default value\n",
    "        # ref_sequences.append(\"NA\")\n",
    "        # alt_sequences.append(\"NA\")\n",
    "\n",
    "        \n",
    "        \n",
    "new_df = pd.DataFrame(data)\n",
    "print(new_df.shape)\n",
    "new_df = new_df.drop_duplicates().reset_index(drop=True)\n",
    "print(new_df.shape)\n",
    "merged_list = list(zip(new_df['reference_seq'], new_df['alt_seq']))\n",
    "merged_list = [item.upper() for tup in merged_list for item in tup]\n",
    "#print(merged_list)\n",
    "kmer_lst = list(map(seq2kmer, merged_list))\n",
    "df_kmer = pd.DataFrame(kmer_lst, columns=['Sequence'])\n",
    "df_kmer['Label'] = np.random.choice([0, 1], size=len(df_kmer))\n",
    "\n",
    "new_df.to_csv(output_path + \"/all_data_new.tsv\", sep=\"\\t\", index= False)\n",
    "df_kmer.to_csv(output_path + \"/dev.tsv\", sep=\"\\t\", index= False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e14368ba-5b1a-4f81-8e07-10b8068121fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>Donor_coordinates</th>\n",
       "      <th>Ensemble_Transcript_ID</th>\n",
       "      <th>Mutation_ID</th>\n",
       "      <th>varinat_start</th>\n",
       "      <th>variant_end</th>\n",
       "      <th>ref_neucleotide</th>\n",
       "      <th>alternative_neucleotide</th>\n",
       "      <th>reference_seq</th>\n",
       "      <th>alt_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>127588526-127588605</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>MU80692579</td>\n",
       "      <td>127588529</td>\n",
       "      <td>127588530</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CGCGGATCTTCGGGAAGAAGCAGATGCGGATTCTCATGGGTGAGGC...</td>\n",
       "      <td>CGCTGATCTTCGGGAAGAAGCAGATGCGGATTCTCATGGGTGAGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>127588526-127588605</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>MU90753368</td>\n",
       "      <td>127588553</td>\n",
       "      <td>127588554</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>CGCGGATCTTCGGGAAGAAGCAGATGCGGATTCTCATGGGTGAGGC...</td>\n",
       "      <td>CGCGGATCTTCGGGAAGAAGCAGATGCAGATTCTCATGGGTGAGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>127588526-127588605</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>MU115107606</td>\n",
       "      <td>127588569</td>\n",
       "      <td>127588570</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CGCGGATCTTCGGGAAGAAGCAGATGCGGATTCTCATGGGTGAGGC...</td>\n",
       "      <td>CGCGGATCTTCGGGAAGAAGCAGATGCGGATTCTCATGGGTGATGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>127588526-127588605</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>MU84868698</td>\n",
       "      <td>127588581</td>\n",
       "      <td>127588582</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CGCGGATCTTCGGGAAGAAGCAGATGCGGATTCTCATGGGTGAGGC...</td>\n",
       "      <td>CGCGGATCTTCGGGAAGAAGCAGATGCGGATTCTCATGGGTGAGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>127588526-127588605</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>MU83704948</td>\n",
       "      <td>127588591</td>\n",
       "      <td>127588592</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CGCGGATCTTCGGGAAGAAGCAGATGCGGATTCTCATGGGTGAGGC...</td>\n",
       "      <td>CGCGGATCTTCGGGAAGAAGCAGATGCGGATTCTCATGGGTGAGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5783391</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>63710688-63710767</td>\n",
       "      <td>ENST00000640999</td>\n",
       "      <td>MU28664773</td>\n",
       "      <td>63710752</td>\n",
       "      <td>63710753</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>CCAATAACCCCTTTCTACCAAGAACCCTTTCCCACTCACCGTATGC...</td>\n",
       "      <td>CCAATAACCCCTTTCTACCAAGAACCCTTTCCCACTCACCGTATGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5783392</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>63710688-63710767</td>\n",
       "      <td>ENST00000640999</td>\n",
       "      <td>MU29299841</td>\n",
       "      <td>63710760</td>\n",
       "      <td>63710761</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CCAATAACCCCTTTCTACCAAGAACCCTTTCCCACTCACCGTATGC...</td>\n",
       "      <td>CCAATAACCCCTTTCTACCAAGAACCCTTTCCCACTCACCGTATGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5783393</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>63710688-63710767</td>\n",
       "      <td>ENST00000640999</td>\n",
       "      <td>MU129502912</td>\n",
       "      <td>63710765</td>\n",
       "      <td>63710766</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>CCAATAACCCCTTTCTACCAAGAACCCTTTCCCACTCACCGTATGC...</td>\n",
       "      <td>CCAATAACCCCTTTCTACCAAGAACCCTTTCCCACTCACCGTATGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5783394</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>63713366-63713445</td>\n",
       "      <td>ENST00000640999</td>\n",
       "      <td>MU84546198</td>\n",
       "      <td>63713396</td>\n",
       "      <td>63713397</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CCCACCCTCCCTCCATGTGACACACTCATGGTTTATTACCTGCAAG...</td>\n",
       "      <td>CCCACCCTCCCTCCATGTGACACACTCATGTTTTATTACCTGCAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5783395</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>63713366-63713445</td>\n",
       "      <td>ENST00000640999</td>\n",
       "      <td>MU130439032</td>\n",
       "      <td>63713425</td>\n",
       "      <td>63713426</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>CCCACCCTCCCTCCATGTGACACACTCATGGTTTATTACCTGCAAG...</td>\n",
       "      <td>CCCACCCTCCCTCCATGTGACACACTCATGGTTTATTACCTGCAAG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5783396 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           chr strand    Donor_coordinates Ensemble_Transcript_ID  \\\n",
       "0         chr7      +  127588526-127588605        ENST00000000233   \n",
       "1         chr7      +  127588526-127588605        ENST00000000233   \n",
       "2         chr7      +  127588526-127588605        ENST00000000233   \n",
       "3         chr7      +  127588526-127588605        ENST00000000233   \n",
       "4         chr7      +  127588526-127588605        ENST00000000233   \n",
       "...        ...    ...                  ...                    ...   \n",
       "5783391  chr17      -    63710688-63710767        ENST00000640999   \n",
       "5783392  chr17      -    63710688-63710767        ENST00000640999   \n",
       "5783393  chr17      -    63710688-63710767        ENST00000640999   \n",
       "5783394  chr17      -    63713366-63713445        ENST00000640999   \n",
       "5783395  chr17      -    63713366-63713445        ENST00000640999   \n",
       "\n",
       "         Mutation_ID  varinat_start  variant_end ref_neucleotide  \\\n",
       "0         MU80692579      127588529    127588530               G   \n",
       "1         MU90753368      127588553    127588554               G   \n",
       "2        MU115107606      127588569    127588570               G   \n",
       "3         MU84868698      127588581    127588582               G   \n",
       "4         MU83704948      127588591    127588592               G   \n",
       "...              ...            ...          ...             ...   \n",
       "5783391   MU28664773       63710752     63710753               C   \n",
       "5783392   MU29299841       63710760     63710761               C   \n",
       "5783393  MU129502912       63710765     63710766               T   \n",
       "5783394   MU84546198       63713396     63713397               G   \n",
       "5783395  MU130439032       63713425     63713426               C   \n",
       "\n",
       "        alternative_neucleotide  \\\n",
       "0                             T   \n",
       "1                             A   \n",
       "2                             T   \n",
       "3                             T   \n",
       "4                             T   \n",
       "...                         ...   \n",
       "5783391                       A   \n",
       "5783392                       T   \n",
       "5783393                       C   \n",
       "5783394                       T   \n",
       "5783395                       G   \n",
       "\n",
       "                                             reference_seq  \\\n",
       "0        CGCGGATCTTCGGGAAGAAGCAGATGCGGATTCTCATGGGTGAGGC...   \n",
       "1        CGCGGATCTTCGGGAAGAAGCAGATGCGGATTCTCATGGGTGAGGC...   \n",
       "2        CGCGGATCTTCGGGAAGAAGCAGATGCGGATTCTCATGGGTGAGGC...   \n",
       "3        CGCGGATCTTCGGGAAGAAGCAGATGCGGATTCTCATGGGTGAGGC...   \n",
       "4        CGCGGATCTTCGGGAAGAAGCAGATGCGGATTCTCATGGGTGAGGC...   \n",
       "...                                                    ...   \n",
       "5783391  CCAATAACCCCTTTCTACCAAGAACCCTTTCCCACTCACCGTATGC...   \n",
       "5783392  CCAATAACCCCTTTCTACCAAGAACCCTTTCCCACTCACCGTATGC...   \n",
       "5783393  CCAATAACCCCTTTCTACCAAGAACCCTTTCCCACTCACCGTATGC...   \n",
       "5783394  CCCACCCTCCCTCCATGTGACACACTCATGGTTTATTACCTGCAAG...   \n",
       "5783395  CCCACCCTCCCTCCATGTGACACACTCATGGTTTATTACCTGCAAG...   \n",
       "\n",
       "                                                   alt_seq  \n",
       "0        CGCTGATCTTCGGGAAGAAGCAGATGCGGATTCTCATGGGTGAGGC...  \n",
       "1        CGCGGATCTTCGGGAAGAAGCAGATGCAGATTCTCATGGGTGAGGC...  \n",
       "2        CGCGGATCTTCGGGAAGAAGCAGATGCGGATTCTCATGGGTGATGC...  \n",
       "3        CGCGGATCTTCGGGAAGAAGCAGATGCGGATTCTCATGGGTGAGGC...  \n",
       "4        CGCGGATCTTCGGGAAGAAGCAGATGCGGATTCTCATGGGTGAGGC...  \n",
       "...                                                    ...  \n",
       "5783391  CCAATAACCCCTTTCTACCAAGAACCCTTTCCCACTCACCGTATGC...  \n",
       "5783392  CCAATAACCCCTTTCTACCAAGAACCCTTTCCCACTCACCGTATGC...  \n",
       "5783393  CCAATAACCCCTTTCTACCAAGAACCCTTTCCCACTCACCGTATGC...  \n",
       "5783394  CCCACCCTCCCTCCATGTGACACACTCATGTTTTATTACCTGCAAG...  \n",
       "5783395  CCCACCCTCCCTCCATGTGACACACTCATGGTTTATTACCTGCAAG...  \n",
       "\n",
       "[5783396 rows x 11 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "7f6ff90c-19f5-4d83-b2ab-7dbcf50a30f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CGCGGA GCGGAT CGGATC GGATCT GATCTT ATCTTC TCTT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CGCTGA GCTGAT CTGATC TGATCT GATCTT ATCTTC TCTT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CGCGGA GCGGAT CGGATC GGATCT GATCTT ATCTTC TCTT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CGCGGA GCGGAT CGGATC GGATCT GATCTT ATCTTC TCTT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CGCGGA GCGGAT CGGATC GGATCT GATCTT ATCTTC TCTT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11566787</th>\n",
       "      <td>CCAATA CAATAA AATAAC ATAACC TAACCC AACCCC ACCC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11566788</th>\n",
       "      <td>CCCACC CCACCC CACCCT ACCCTC CCCTCC CCTCCC CTCC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11566789</th>\n",
       "      <td>CCCACC CCACCC CACCCT ACCCTC CCCTCC CCTCCC CTCC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11566790</th>\n",
       "      <td>CCCACC CCACCC CACCCT ACCCTC CCCTCC CCTCCC CTCC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11566791</th>\n",
       "      <td>CCCACC CCACCC CACCCT ACCCTC CCCTCC CCTCCC CTCC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11566792 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   Sequence  Label\n",
       "0         CGCGGA GCGGAT CGGATC GGATCT GATCTT ATCTTC TCTT...      1\n",
       "1         CGCTGA GCTGAT CTGATC TGATCT GATCTT ATCTTC TCTT...      0\n",
       "2         CGCGGA GCGGAT CGGATC GGATCT GATCTT ATCTTC TCTT...      1\n",
       "3         CGCGGA GCGGAT CGGATC GGATCT GATCTT ATCTTC TCTT...      1\n",
       "4         CGCGGA GCGGAT CGGATC GGATCT GATCTT ATCTTC TCTT...      1\n",
       "...                                                     ...    ...\n",
       "11566787  CCAATA CAATAA AATAAC ATAACC TAACCC AACCCC ACCC...      0\n",
       "11566788  CCCACC CCACCC CACCCT ACCCTC CCCTCC CCTCCC CTCC...      1\n",
       "11566789  CCCACC CCACCC CACCCT ACCCTC CCCTCC CCTCCC CTCC...      1\n",
       "11566790  CCCACC CCACCC CACCCT ACCCTC CCCTCC CCTCCC CTCC...      1\n",
       "11566791  CCCACC CCACCC CACCCT ACCCTC CCCTCC CCTCCC CTCC...      1\n",
       "\n",
       "[11566792 rows x 2 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_kmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d66e4a84-e7e2-4afa-9282-b86ab19680ff",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
