{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9807e3a4-8edb-46a2-ad4c-88777af9fadc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, glob\n",
    "import pandas as pd\n",
    "from io import StringIO\n",
    "import numpy as np\n",
    "import pysam\n",
    "import pybedtools \n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "pybedtools.helpers.set_tempdir(\"/data/projects/temp\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2505ee28-f7a7-474a-a4b8-755989f3f70e",
   "metadata": {},
   "outputs": [],
   "source": [
    "non_coding_region = \"core_prom\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6c3ba47b-0edf-43de-bab5-bc0edd7f91b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '/data/projects/DNABERT_snv/Manuscript_11_2023/Intersected_data/{}_ICGC_intersected.tsv'.format(non_coding_region)\n",
    "output_path = \"/data/projects/DNABERT_snv/Manuscript_11_2023/DNABERT_data/{}\".format(non_coding_region) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "aafe0877-b7be-4e5e-a73f-00dab9627b51",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(output_path):\n",
    "    os.makedirs(output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cc89bac5-1528-4192-95d4-77bda6a6e1dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def seq2kmer(seq):\n",
    "    \"\"\"\n",
    "    Convert original sequence to kmers\n",
    "    \n",
    "    Arguments:\n",
    "    seq -- str, original sequence.\n",
    "    k -- int, kmer of length k specified.\n",
    "    \n",
    "    Returns:\n",
    "    kmers -- str, kmers separated by space\n",
    "    \"\"\"\n",
    "    k=6\n",
    "    kmer = [seq[x:x+k] for x in range(len(seq)+1-k)]\n",
    "    kmers = \" \".join(kmer)\n",
    "    return kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "19aaa7a0-7309-4e97-a93e-dded3164e07f",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = SeqIO.to_dict(SeqIO.parse(\"/data/projects/Resources/Gencode_genome_annotation/GRCh38.primary_assembly.genome.fa\", \"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3bcc2b33-f043-4016-9aba-79948c239d76",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(746364, 22)\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(data_path, sep= '\\t')\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0ded228f-2999-482f-9e5a-197b732e0cfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(745265, 11)\n",
      "(745265, 11)\n"
     ]
    }
   ],
   "source": [
    "# df = pd.read_csv(data_path, sep= '\\t')\n",
    "# print(df.shape)\n",
    "\n",
    "# df = df[df['No_of_variations']<3].reset_index(drop=True)\n",
    "# print(df.shape)\n",
    "# df = df.loc[df['Alternative_Allele']== '.'].reset_index(drop=True)\n",
    "# df\n",
    "#df= df[df['strand']==\"-\"]\n",
    "\n",
    "\n",
    "data= []\n",
    "#df = df.head(10)\n",
    "# Iterate over DataFrame rows\n",
    "for idx, row in df.iterrows():\n",
    "    try:\n",
    "        #print(row)Reference_Allele\tAlternative_Allele\n",
    "        alts = row['Alternative_Allele'].rstrip(',').split(',')\n",
    "        ref_seq = str(genome[row['chr_name']].seq[ row['Core_prom_start']:row['Core_prom_end']])\n",
    "\n",
    "        # Calculate the variant position relative to the fetched sequence\n",
    "        variant_pos_start = row['Start'] - row['Core_prom_start']\n",
    "        variant_pos_end = row['End']  - row['Core_prom_start']\n",
    "        # print(variant_pos_start, variant_pos_end, alts,  row['Reference_Allele'])\n",
    "        # print(ref_seq)\n",
    "        # if row['strand'] == '-':\n",
    "        #     ref_seq = str(Seq(ref_seq).reverse_complement())\n",
    "        #print(ref_seq)\n",
    "\n",
    "\n",
    "        #Replace the reference nucleotide with the alternate nucleotide to get the alternate sequence\n",
    "        for alt in alts:\n",
    "            if alt == \".\":  # Deletion\n",
    "                # Fetch additional bases from genome to maintain sequence length\n",
    "                extra_bases = str(genome[row['chr_name']].seq[row['Core_prom_end']:row['Core_prom_end'] + len(row['Reference_Allele'])])\n",
    "                #print(len(row['Reference_Nucleotide']))\n",
    "                alt_seq = ref_seq[:variant_pos_start] + ref_seq[variant_pos_end:] + extra_bases\n",
    "            else:  # SNPs, Insertions, and partial Deletions\n",
    "                extra_bases = str(genome[row['chr_name']].seq[row['Core_prom_end']:row['Core_prom_end'] + len(row['Reference_Allele']) - len(alt)])\n",
    "                #print(extra_bases)\n",
    "                if(variant_pos_start>=0):\n",
    "                    alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:] + extra_bases\n",
    "                else:\n",
    "                    continue\n",
    "            #print(alt_seq)\n",
    "            #Append to the list as a dictionary\n",
    "            data.append({\n",
    "                'chr': row['chr_name'],\n",
    "                'strand': row['strand'],\n",
    "                'Core_prom_coordinates': str(row['Core_prom_start'])+'-'+str(row['Core_prom_end']),\n",
    "                'Ensemble_Transcript_ID': row['ENSEMBL_Transcript_ID'],\n",
    "                'Mutation_ID': row['Mutation_ID'],\n",
    "                'varinat_start':  row['Start'],\n",
    "                'variant_end':  row['End'],\n",
    "                'ref_neucleotide': row['Reference_Allele'],\n",
    "                'alternative_neucleotide': alt,\n",
    "                'reference_seq': ref_seq,\n",
    "                'alt_seq': alt_seq\n",
    "            })\n",
    "            \n",
    "    except KeyError:\n",
    "        print(\"NA\")\n",
    "        # # If sequence is not present, append a default value\n",
    "        # ref_sequences.append(\"NA\")\n",
    "        # alt_sequences.append(\"NA\")\n",
    "\n",
    "        \n",
    "        \n",
    "new_df = pd.DataFrame(data)\n",
    "print(new_df.shape)\n",
    "new_df = new_df.drop_duplicates().reset_index(drop=True)\n",
    "print(new_df.shape)\n",
    "merged_list = list(zip(new_df['reference_seq'], new_df['alt_seq']))\n",
    "merged_list = [item.upper() for tup in merged_list for item in tup]\n",
    "#print(merged_list)\n",
    "kmer_lst = list(map(seq2kmer, merged_list))\n",
    "df_kmer = pd.DataFrame(kmer_lst, columns=['Sequence'])\n",
    "df_kmer['Label'] = np.random.choice([0, 1], size=len(df_kmer))\n",
    "\n",
    "new_df.to_csv(output_path + \"/all_data_new.tsv\", sep=\"\\t\", index= False)\n",
    "df_kmer.to_csv(output_path + \"/dev.tsv\", sep=\"\\t\", index= False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e14368ba-5b1a-4f81-8e07-10b8068121fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>Core_prom_coordinates</th>\n",
       "      <th>Ensemble_Transcript_ID</th>\n",
       "      <th>Mutation_ID</th>\n",
       "      <th>varinat_start</th>\n",
       "      <th>variant_end</th>\n",
       "      <th>ref_neucleotide</th>\n",
       "      <th>alternative_neucleotide</th>\n",
       "      <th>reference_seq</th>\n",
       "      <th>alt_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>127588366-127588455</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>MU133392649</td>\n",
       "      <td>127588387</td>\n",
       "      <td>127588388</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GTCGGGAGGGCAGCGACGCGCGGAGGCGGCGGCGGAGCCTCCTCCT...</td>\n",
       "      <td>GTCGGGAGGGCAGCGACGCGCAGAGGCGGCGGCGGAGCCTCCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>127588366-127588455</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>MU39478069</td>\n",
       "      <td>127588429</td>\n",
       "      <td>127588430</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>GTCGGGAGGGCAGCGACGCGCGGAGGCGGCGGCGGAGCCTCCTCCT...</td>\n",
       "      <td>GTCGGGAGGGCAGCGACGCGCGGAGGCGGCGGCGGAGCCTCCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr12</td>\n",
       "      <td>-</td>\n",
       "      <td>8949601-8949690</td>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>MU49180650</td>\n",
       "      <td>8949659</td>\n",
       "      <td>8949660</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>TCAGAGGCCAGCGTTCCCCCTAGCGCCTCGCTGTGCCCCACTCTGG...</td>\n",
       "      <td>TCAGAGGCCAGCGTTCCCCCTAGCGCCTCGCTGTGCCCCACTCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12</td>\n",
       "      <td>-</td>\n",
       "      <td>8949601-8949690</td>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>MU37117518</td>\n",
       "      <td>8949660</td>\n",
       "      <td>8949661</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>TCAGAGGCCAGCGTTCCCCCTAGCGCCTCGCTGTGCCCCACTCTGG...</td>\n",
       "      <td>TCAGAGGCCAGCGTTCCCCCTAGCGCCTCGCTGTGCCCCACTCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>64305479-64305568</td>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>MU46907284</td>\n",
       "      <td>64305485</td>\n",
       "      <td>64305486</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CCCGCCCCCTGCTTTGCATGCGCACGGCCGGCCCCACCCCCGCTGT...</td>\n",
       "      <td>CCCGCCTCCTGCTTTGCATGCGCACGGCCGGCCCCACCCCCGCTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>745260</th>\n",
       "      <td>chr20</td>\n",
       "      <td>+</td>\n",
       "      <td>45416088-45416177</td>\n",
       "      <td>ENST00000640996</td>\n",
       "      <td>MU66504628</td>\n",
       "      <td>45416128</td>\n",
       "      <td>45416129</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCCGCTGG...</td>\n",
       "      <td>CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCGGCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>745261</th>\n",
       "      <td>chr20</td>\n",
       "      <td>+</td>\n",
       "      <td>45416088-45416177</td>\n",
       "      <td>ENST00000640996</td>\n",
       "      <td>MU128987750</td>\n",
       "      <td>45416129</td>\n",
       "      <td>45416130</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCCGCTGG...</td>\n",
       "      <td>CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCCACTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>745262</th>\n",
       "      <td>chr20</td>\n",
       "      <td>+</td>\n",
       "      <td>45416088-45416177</td>\n",
       "      <td>ENST00000640996</td>\n",
       "      <td>MU120160683</td>\n",
       "      <td>45416141</td>\n",
       "      <td>45416142</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCCGCTGG...</td>\n",
       "      <td>CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCCGCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>745263</th>\n",
       "      <td>chr20</td>\n",
       "      <td>+</td>\n",
       "      <td>45416088-45416177</td>\n",
       "      <td>ENST00000640996</td>\n",
       "      <td>MU93024897</td>\n",
       "      <td>45416144</td>\n",
       "      <td>45416145</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCCGCTGG...</td>\n",
       "      <td>CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCCGCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>745264</th>\n",
       "      <td>chr20</td>\n",
       "      <td>+</td>\n",
       "      <td>45416088-45416177</td>\n",
       "      <td>ENST00000640996</td>\n",
       "      <td>MU633820</td>\n",
       "      <td>45416172</td>\n",
       "      <td>45416173</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCCGCTGG...</td>\n",
       "      <td>CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCCGCTGG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>745265 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr strand Core_prom_coordinates Ensemble_Transcript_ID  \\\n",
       "0        chr7      +   127588366-127588455        ENST00000000233   \n",
       "1        chr7      +   127588366-127588455        ENST00000000233   \n",
       "2       chr12      -       8949601-8949690        ENST00000000412   \n",
       "3       chr12      -       8949601-8949690        ENST00000000412   \n",
       "4       chr11      +     64305479-64305568        ENST00000000442   \n",
       "...       ...    ...                   ...                    ...   \n",
       "745260  chr20      +     45416088-45416177        ENST00000640996   \n",
       "745261  chr20      +     45416088-45416177        ENST00000640996   \n",
       "745262  chr20      +     45416088-45416177        ENST00000640996   \n",
       "745263  chr20      +     45416088-45416177        ENST00000640996   \n",
       "745264  chr20      +     45416088-45416177        ENST00000640996   \n",
       "\n",
       "        Mutation_ID  varinat_start  variant_end ref_neucleotide  \\\n",
       "0       MU133392649      127588387    127588388               G   \n",
       "1        MU39478069      127588429    127588430               T   \n",
       "2        MU49180650        8949659      8949660               G   \n",
       "3        MU37117518        8949660      8949661               G   \n",
       "4        MU46907284       64305485     64305486               C   \n",
       "...             ...            ...          ...             ...   \n",
       "745260   MU66504628       45416128     45416129               C   \n",
       "745261  MU128987750       45416129     45416130               G   \n",
       "745262  MU120160683       45416141     45416142               G   \n",
       "745263   MU93024897       45416144     45416145               G   \n",
       "745264     MU633820       45416172     45416173               C   \n",
       "\n",
       "       alternative_neucleotide  \\\n",
       "0                            A   \n",
       "1                            C   \n",
       "2                            A   \n",
       "3                            A   \n",
       "4                            T   \n",
       "...                        ...   \n",
       "745260                       G   \n",
       "745261                       A   \n",
       "745262                       A   \n",
       "745263                       T   \n",
       "745264                       T   \n",
       "\n",
       "                                            reference_seq  \\\n",
       "0       GTCGGGAGGGCAGCGACGCGCGGAGGCGGCGGCGGAGCCTCCTCCT...   \n",
       "1       GTCGGGAGGGCAGCGACGCGCGGAGGCGGCGGCGGAGCCTCCTCCT...   \n",
       "2       TCAGAGGCCAGCGTTCCCCCTAGCGCCTCGCTGTGCCCCACTCTGG...   \n",
       "3       TCAGAGGCCAGCGTTCCCCCTAGCGCCTCGCTGTGCCCCACTCTGG...   \n",
       "4       CCCGCCCCCTGCTTTGCATGCGCACGGCCGGCCCCACCCCCGCTGT...   \n",
       "...                                                   ...   \n",
       "745260  CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCCGCTGG...   \n",
       "745261  CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCCGCTGG...   \n",
       "745262  CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCCGCTGG...   \n",
       "745263  CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCCGCTGG...   \n",
       "745264  CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCCGCTGG...   \n",
       "\n",
       "                                                  alt_seq  \n",
       "0       GTCGGGAGGGCAGCGACGCGCAGAGGCGGCGGCGGAGCCTCCTCCT...  \n",
       "1       GTCGGGAGGGCAGCGACGCGCGGAGGCGGCGGCGGAGCCTCCTCCT...  \n",
       "2       TCAGAGGCCAGCGTTCCCCCTAGCGCCTCGCTGTGCCCCACTCTGG...  \n",
       "3       TCAGAGGCCAGCGTTCCCCCTAGCGCCTCGCTGTGCCCCACTCTGG...  \n",
       "4       CCCGCCTCCTGCTTTGCATGCGCACGGCCGGCCCCACCCCCGCTGT...  \n",
       "...                                                   ...  \n",
       "745260  CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCGGCTGG...  \n",
       "745261  CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCCACTGG...  \n",
       "745262  CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCCGCTGG...  \n",
       "745263  CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCCGCTGG...  \n",
       "745264  CCTTTCTTATTGGATACCGGCTTTTCATGGGGCCGAGCGCCGCTGG...  \n",
       "\n",
       "[745265 rows x 11 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7f6ff90c-19f5-4d83-b2ab-7dbcf50a30f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GTCGGG TCGGGA CGGGAG GGGAGG GGAGGG GAGGGC AGGG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GTCGGG TCGGGA CGGGAG GGGAGG GGAGGG GAGGGC AGGG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTCGGG TCGGGA CGGGAG GGGAGG GGAGGG GAGGGC AGGG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTCGGG TCGGGA CGGGAG GGGAGG GGAGGG GAGGGC AGGG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCAGAG CAGAGG AGAGGC GAGGCC AGGCCA GGCCAG GCCA...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1490525</th>\n",
       "      <td>CCTTTC CTTTCT TTTCTT TTCTTA TCTTAT CTTATT TTAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1490526</th>\n",
       "      <td>CCTTTC CTTTCT TTTCTT TTCTTA TCTTAT CTTATT TTAT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1490527</th>\n",
       "      <td>CCTTTC CTTTCT TTTCTT TTCTTA TCTTAT CTTATT TTAT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1490528</th>\n",
       "      <td>CCTTTC CTTTCT TTTCTT TTCTTA TCTTAT CTTATT TTAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1490529</th>\n",
       "      <td>CCTTTC CTTTCT TTTCTT TTCTTA TCTTAT CTTATT TTAT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1490530 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  Sequence  Label\n",
       "0        GTCGGG TCGGGA CGGGAG GGGAGG GGAGGG GAGGGC AGGG...      0\n",
       "1        GTCGGG TCGGGA CGGGAG GGGAGG GGAGGG GAGGGC AGGG...      0\n",
       "2        GTCGGG TCGGGA CGGGAG GGGAGG GGAGGG GAGGGC AGGG...      1\n",
       "3        GTCGGG TCGGGA CGGGAG GGGAGG GGAGGG GAGGGC AGGG...      1\n",
       "4        TCAGAG CAGAGG AGAGGC GAGGCC AGGCCA GGCCAG GCCA...      1\n",
       "...                                                    ...    ...\n",
       "1490525  CCTTTC CTTTCT TTTCTT TTCTTA TCTTAT CTTATT TTAT...      0\n",
       "1490526  CCTTTC CTTTCT TTTCTT TTCTTA TCTTAT CTTATT TTAT...      1\n",
       "1490527  CCTTTC CTTTCT TTTCTT TTCTTA TCTTAT CTTATT TTAT...      1\n",
       "1490528  CCTTTC CTTTCT TTTCTT TTCTTA TCTTAT CTTATT TTAT...      0\n",
       "1490529  CCTTTC CTTTCT TTTCTT TTCTTA TCTTAT CTTATT TTAT...      1\n",
       "\n",
       "[1490530 rows x 2 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_kmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d66e4a84-e7e2-4afa-9282-b86ab19680ff",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
