{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "9807e3a4-8edb-46a2-ad4c-88777af9fadc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, glob\n",
    "import pandas as pd\n",
    "from io import StringIO\n",
    "import numpy as np\n",
    "import pysam\n",
    "import pybedtools \n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "pybedtools.helpers.set_tempdir(\"/data/projects/temp\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "2505ee28-f7a7-474a-a4b8-755989f3f70e",
   "metadata": {},
   "outputs": [],
   "source": [
    "non_coding_region = \"acceptor\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6c3ba47b-0edf-43de-bab5-bc0edd7f91b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '/data/projects/DNABERT_snv/Manuscript_11_2023/Intersected_data/{}_ICGC_intersected.tsv'.format(non_coding_region)\n",
    "output_path = \"/data/projects/DNABERT_snv/Manuscript_11_2023/DNABERT_data/{}\".format(non_coding_region) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "aafe0877-b7be-4e5e-a73f-00dab9627b51",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(output_path):\n",
    "    os.makedirs(output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "cc89bac5-1528-4192-95d4-77bda6a6e1dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def seq2kmer(seq):\n",
    "    \"\"\"\n",
    "    Convert original sequence to kmers\n",
    "    \n",
    "    Arguments:\n",
    "    seq -- str, original sequence.\n",
    "    k -- int, kmer of length k specified.\n",
    "    \n",
    "    Returns:\n",
    "    kmers -- str, kmers separated by space\n",
    "    \"\"\"\n",
    "    k=6\n",
    "    kmer = [seq[x:x+k] for x in range(len(seq)+1-k)]\n",
    "    kmers = \" \".join(kmer)\n",
    "    return kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "19aaa7a0-7309-4e97-a93e-dded3164e07f",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = SeqIO.to_dict(SeqIO.parse(\"/data/projects/Resources/Gencode_genome_annotation/GRCh38.primary_assembly.genome.fa\", \"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "3bcc2b33-f043-4016-9aba-79948c239d76",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6001604, 22)\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(data_path, sep= '\\t')\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "0ded228f-2999-482f-9e5a-197b732e0cfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5995702, 11)\n",
      "(5992039, 11)\n"
     ]
    }
   ],
   "source": [
    "# df = pd.read_csv(data_path, sep= '\\t')\n",
    "# print(df.shape)\n",
    "\n",
    "# df = df[df['No_of_variations']<3].reset_index(drop=True)\n",
    "# print(df.shape)\n",
    "# df = df.loc[df['Alternative_Allele']== '.'].reset_index(drop=True)\n",
    "# df\n",
    "#df= df[df['strand']==\"-\"]\n",
    "\n",
    "\n",
    "data= []\n",
    "df = df.head(10)\n",
    "# Iterate over DataFrame rows\n",
    "for idx, row in df.iterrows():\n",
    "    try:\n",
    "        #print(row)Reference_Allele\tAlternative_Allele\n",
    "        alts = row['Alternative_Allele'].rstrip(',').split(',')\n",
    "        ref_seq = str(genome[row['chr_name']].seq[ row['Acceptor_start']:row['Acceptor_end']])\n",
    "\n",
    "        # Calculate the variant position relative to the fetched sequence\n",
    "        variant_pos_start = row['Start'] - row['Acceptor_start']\n",
    "        variant_pos_end = row['End']  - row['Acceptor_start']\n",
    "        #print(variant_pos_start, variant_pos_end, alts,  row['Reference_Allele'])\n",
    "        # print(ref_seq)\n",
    "        # if row['strand'] == '-':\n",
    "        #     ref_seq = str(Seq(ref_seq).reverse_complement())\n",
    "        #print(ref_seq)\n",
    "\n",
    "\n",
    "        #Replace the reference nucleotide with the alternate nucleotide to get the alternate sequence\n",
    "        for alt in alts:\n",
    "            if alt == \".\":  # Deletion\n",
    "                # Fetch additional bases from genome to maintain sequence length\n",
    "                extra_bases = str(genome[row['chr_name']].seq[row['Acceptor_end']:row['Acceptor_end'] + len(row['Reference_Allele'])])\n",
    "                #print(len(row['Reference_Nucleotide']))\n",
    "                alt_seq = ref_seq[:variant_pos_start] + ref_seq[variant_pos_end:] + extra_bases\n",
    "            else:  # SNPs, Insertions, and partial Deletions\n",
    "                extra_bases = str(genome[row['chr_name']].seq[row['Acceptor_end']:row['Acceptor_end'] + len(row['Reference_Allele']) - len(alt)])\n",
    "                #print(extra_bases)\n",
    "                if(variant_pos_start>=0):\n",
    "                    alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:] + extra_bases\n",
    "                else:\n",
    "                    continue\n",
    "            #print(alt_seq)\n",
    "            #Append to the list as a dictionary\n",
    "            data.append({\n",
    "                'chr': row['chr_name'],\n",
    "                'strand': row['strand'],\n",
    "                'acceptor_coordinates': str(row['Acceptor_start'])+'-'+str(row['Acceptor_end']),\n",
    "                'Ensemble_Transcript_ID': row['Transcript_ID'],\n",
    "                'Mutation_ID': row['Mutation_ID'],\n",
    "                'varinat_start':  row['Start'],\n",
    "                'variant_end':  row['End'],\n",
    "                'ref_neucleotide': row['Reference_Allele'],\n",
    "                'alternative_neucleotide': alt,\n",
    "                'reference_seq': ref_seq,\n",
    "                'alt_seq': alt_seq\n",
    "            })\n",
    "            \n",
    "    except KeyError:\n",
    "        print(\"NA\")\n",
    "        # # If sequence is not present, append a default value\n",
    "        # ref_sequences.append(\"NA\")\n",
    "        # alt_sequences.append(\"NA\")\n",
    "        \n",
    "new_df = pd.DataFrame(data)\n",
    "print(new_df.shape)\n",
    "new_df = new_df.drop_duplicates().reset_index(drop=True)\n",
    "print(new_df.shape)\n",
    "merged_list = list(zip(new_df['reference_seq'], new_df['alt_seq']))\n",
    "merged_list = [item.upper() for tup in merged_list for item in tup]\n",
    "#print(merged_list)\n",
    "kmer_lst = list(map(seq2kmer, merged_list))\n",
    "df_kmer = pd.DataFrame(kmer_lst, columns=['Sequence'])\n",
    "df_kmer['Label'] = np.random.choice([0, 1], size=len(df_kmer))\n",
    "\n",
    "new_df.to_csv(output_path + \"/all_data_new.tsv\", sep=\"\\t\", index= False)\n",
    "df_kmer.to_csv(output_path + \"/dev.tsv\", sep=\"\\t\", index= False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "e14368ba-5b1a-4f81-8e07-10b8068121fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>acceptor_coordinates</th>\n",
       "      <th>Ensemble_Transcript_ID</th>\n",
       "      <th>Mutation_ID</th>\n",
       "      <th>varinat_start</th>\n",
       "      <th>variant_end</th>\n",
       "      <th>ref_neucleotide</th>\n",
       "      <th>alternative_neucleotide</th>\n",
       "      <th>reference_seq</th>\n",
       "      <th>alt_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>127589043-127589122</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>MU82778890</td>\n",
       "      <td>127589078</td>\n",
       "      <td>127589079</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>TCGCTCCCATCTCCATCCCTGTGCCCCTTTCCGTTGCAGTTGGCTT...</td>\n",
       "      <td>TCGCTCCCATCTCCATCCCTGTGCCCCTTTCCGTTTCAGTTGGCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>127589043-127589122</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>MU85932120</td>\n",
       "      <td>127589090</td>\n",
       "      <td>127589091</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>TCGCTCCCATCTCCATCCCTGTGCCCCTTTCCGTTGCAGTTGGCTT...</td>\n",
       "      <td>TCGCTCCCATCTCCATCCCTGTGCCCCTTTCCGTTGCAGTTGGCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>127589043-127589122</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>MU84868715</td>\n",
       "      <td>127589096</td>\n",
       "      <td>127589097</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>TCGCTCCCATCTCCATCCCTGTGCCCCTTTCCGTTGCAGTTGGCTT...</td>\n",
       "      <td>TCGCTCCCATCTCCATCCCTGTGCCCCTTTCCGTTGCAGTTGGCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>127589043-127589122</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>MU82319183</td>\n",
       "      <td>127589104</td>\n",
       "      <td>127589105</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>TCGCTCCCATCTCCATCCCTGTGCCCCTTTCCGTTGCAGTTGGCTT...</td>\n",
       "      <td>TCGCTCCCATCTCCATCCCTGTGCCCCTTTCCGTTGCAGTTGGCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>127588371-127588450</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>MU133310861</td>\n",
       "      <td>127588386</td>\n",
       "      <td>127588387</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>GAGGGCAGCGACGCGCGGAGGCGGCGGCGGAGCCTCCTCCTGCTGC...</td>\n",
       "      <td>GAGGGCAGCGACGCGAGGAGGCGGCGGCGGAGCCTCCTCCTGCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5992034</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>63713488-63713567</td>\n",
       "      <td>ENST00000640999</td>\n",
       "      <td>MU84976091</td>\n",
       "      <td>63713494</td>\n",
       "      <td>63713495</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>TTGCTAGATTCACAGTCATCAGGTCCTCAAATCCTTTGCCTAAAAG...</td>\n",
       "      <td>TTGCTATATTCACAGTCATCAGGTCCTCAAATCCTTTGCCTAAAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5992035</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>63713488-63713567</td>\n",
       "      <td>ENST00000640999</td>\n",
       "      <td>MU129479225</td>\n",
       "      <td>63713502</td>\n",
       "      <td>63713503</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>TTGCTAGATTCACAGTCATCAGGTCCTCAAATCCTTTGCCTAAAAG...</td>\n",
       "      <td>TTGCTAGATTCACAATCATCAGGTCCTCAAATCCTTTGCCTAAAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5992036</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>63713488-63713567</td>\n",
       "      <td>ENST00000640999</td>\n",
       "      <td>MU29299852</td>\n",
       "      <td>63713515</td>\n",
       "      <td>63713516</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>TTGCTAGATTCACAGTCATCAGGTCCTCAAATCCTTTGCCTAAAAG...</td>\n",
       "      <td>TTGCTAGATTCACAGTCATCAGGTCCTGAAATCCTTTGCCTAAAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5992037</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>63713488-63713567</td>\n",
       "      <td>ENST00000640999</td>\n",
       "      <td>MU121174896</td>\n",
       "      <td>63713550</td>\n",
       "      <td>63713551</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>TTGCTAGATTCACAGTCATCAGGTCCTCAAATCCTTTGCCTAAAAG...</td>\n",
       "      <td>TTGCTAGATTCACAGTCATCAGGTCCTCAAATCCTTTGCCTAAAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5992038</th>\n",
       "      <td>chr17</td>\n",
       "      <td>-</td>\n",
       "      <td>63713488-63713567</td>\n",
       "      <td>ENST00000640999</td>\n",
       "      <td>MU762371</td>\n",
       "      <td>63713556</td>\n",
       "      <td>63713557</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>TTGCTAGATTCACAGTCATCAGGTCCTCAAATCCTTTGCCTAAAAG...</td>\n",
       "      <td>TTGCTAGATTCACAGTCATCAGGTCCTCAAATCCTTTGCCTAAAAG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5992039 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           chr strand acceptor_coordinates Ensemble_Transcript_ID  \\\n",
       "0         chr7      +  127589043-127589122        ENST00000000233   \n",
       "1         chr7      +  127589043-127589122        ENST00000000233   \n",
       "2         chr7      +  127589043-127589122        ENST00000000233   \n",
       "3         chr7      +  127589043-127589122        ENST00000000233   \n",
       "4         chr7      +  127588371-127588450        ENST00000000233   \n",
       "...        ...    ...                  ...                    ...   \n",
       "5992034  chr17      -    63713488-63713567        ENST00000640999   \n",
       "5992035  chr17      -    63713488-63713567        ENST00000640999   \n",
       "5992036  chr17      -    63713488-63713567        ENST00000640999   \n",
       "5992037  chr17      -    63713488-63713567        ENST00000640999   \n",
       "5992038  chr17      -    63713488-63713567        ENST00000640999   \n",
       "\n",
       "         Mutation_ID  varinat_start  variant_end ref_neucleotide  \\\n",
       "0         MU82778890      127589078    127589079               G   \n",
       "1         MU85932120      127589090    127589091               G   \n",
       "2         MU84868715      127589096    127589097               G   \n",
       "3         MU82319183      127589104    127589105               G   \n",
       "4        MU133310861      127588386    127588387               C   \n",
       "...              ...            ...          ...             ...   \n",
       "5992034   MU84976091       63713494     63713495               G   \n",
       "5992035  MU129479225       63713502     63713503               G   \n",
       "5992036   MU29299852       63713515     63713516               C   \n",
       "5992037  MU121174896       63713550     63713551               A   \n",
       "5992038     MU762371       63713556     63713557               G   \n",
       "\n",
       "        alternative_neucleotide  \\\n",
       "0                             T   \n",
       "1                             T   \n",
       "2                             T   \n",
       "3                             T   \n",
       "4                             A   \n",
       "...                         ...   \n",
       "5992034                       T   \n",
       "5992035                       A   \n",
       "5992036                       G   \n",
       "5992037                       T   \n",
       "5992038                       C   \n",
       "\n",
       "                                             reference_seq  \\\n",
       "0        TCGCTCCCATCTCCATCCCTGTGCCCCTTTCCGTTGCAGTTGGCTT...   \n",
       "1        TCGCTCCCATCTCCATCCCTGTGCCCCTTTCCGTTGCAGTTGGCTT...   \n",
       "2        TCGCTCCCATCTCCATCCCTGTGCCCCTTTCCGTTGCAGTTGGCTT...   \n",
       "3        TCGCTCCCATCTCCATCCCTGTGCCCCTTTCCGTTGCAGTTGGCTT...   \n",
       "4        GAGGGCAGCGACGCGCGGAGGCGGCGGCGGAGCCTCCTCCTGCTGC...   \n",
       "...                                                    ...   \n",
       "5992034  TTGCTAGATTCACAGTCATCAGGTCCTCAAATCCTTTGCCTAAAAG...   \n",
       "5992035  TTGCTAGATTCACAGTCATCAGGTCCTCAAATCCTTTGCCTAAAAG...   \n",
       "5992036  TTGCTAGATTCACAGTCATCAGGTCCTCAAATCCTTTGCCTAAAAG...   \n",
       "5992037  TTGCTAGATTCACAGTCATCAGGTCCTCAAATCCTTTGCCTAAAAG...   \n",
       "5992038  TTGCTAGATTCACAGTCATCAGGTCCTCAAATCCTTTGCCTAAAAG...   \n",
       "\n",
       "                                                   alt_seq  \n",
       "0        TCGCTCCCATCTCCATCCCTGTGCCCCTTTCCGTTTCAGTTGGCTT...  \n",
       "1        TCGCTCCCATCTCCATCCCTGTGCCCCTTTCCGTTGCAGTTGGCTT...  \n",
       "2        TCGCTCCCATCTCCATCCCTGTGCCCCTTTCCGTTGCAGTTGGCTT...  \n",
       "3        TCGCTCCCATCTCCATCCCTGTGCCCCTTTCCGTTGCAGTTGGCTT...  \n",
       "4        GAGGGCAGCGACGCGAGGAGGCGGCGGCGGAGCCTCCTCCTGCTGC...  \n",
       "...                                                    ...  \n",
       "5992034  TTGCTATATTCACAGTCATCAGGTCCTCAAATCCTTTGCCTAAAAG...  \n",
       "5992035  TTGCTAGATTCACAATCATCAGGTCCTCAAATCCTTTGCCTAAAAG...  \n",
       "5992036  TTGCTAGATTCACAGTCATCAGGTCCTGAAATCCTTTGCCTAAAAG...  \n",
       "5992037  TTGCTAGATTCACAGTCATCAGGTCCTCAAATCCTTTGCCTAAAAG...  \n",
       "5992038  TTGCTAGATTCACAGTCATCAGGTCCTCAAATCCTTTGCCTAAAAG...  \n",
       "\n",
       "[5992039 rows x 11 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "7f6ff90c-19f5-4d83-b2ab-7dbcf50a30f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGCTC CGCTCC GCTCCC CTCCCA TCCCAT CCCATC CCAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGCTC CGCTCC GCTCCC CTCCCA TCCCAT CCCATC CCAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGCTC CGCTCC GCTCCC CTCCCA TCCCAT CCCATC CCAT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGCTC CGCTCC GCTCCC CTCCCA TCCCAT CCCATC CCAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGCTC CGCTCC GCTCCC CTCCCA TCCCAT CCCATC CCAT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TCGCTC CGCTCC GCTCCC CTCCCA TCCCAT CCCATC CCAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TCGCTC CGCTCC GCTCCC CTCCCA TCCCAT CCCATC CCAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TCGCTC CGCTCC GCTCCC CTCCCA TCCCAT CCCATC CCAT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GAGGGC AGGGCA GGGCAG GGCAGC GCAGCG CAGCGA AGCG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GAGGGC AGGGCA GGGCAG GGCAGC GCAGCG CAGCGA AGCG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GAGGGC AGGGCA GGGCAG GGCAGC GCAGCG CAGCGA AGCG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GAGGGC AGGGCA GGGCAG GGCAGC GCAGCG CAGCGA AGCG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GAGGGC AGGGCA GGGCAG GGCAGC GCAGCG CAGCGA AGCG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GAGGGC AGGGCA GGGCAG GGCAGC GCAGCG CAGCGA AGCG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TCAGGA CAGGAG AGGAGT GGAGTT GAGTTT AGTTTT GTTT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TCGGGA CGGGAG GGGAGT GGAGTT GAGTTT AGTTTT GTTT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>TCAGGA CAGGAG AGGAGT GGAGTT GAGTTT AGTTTT GTTT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>TCAGGA CAGGAG AGGAGT GGAGTT GAGTTT AGTTTT GTTT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>TCAGGA CAGGAG AGGAGT GGAGTT GAGTTT AGTTTT GTTT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>TCAGGA CAGGAG AGGAGT GGAGTT GAGTTT AGTTTT GTTT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Sequence  Label\n",
       "0   TCGCTC CGCTCC GCTCCC CTCCCA TCCCAT CCCATC CCAT...      0\n",
       "1   TCGCTC CGCTCC GCTCCC CTCCCA TCCCAT CCCATC CCAT...      0\n",
       "2   TCGCTC CGCTCC GCTCCC CTCCCA TCCCAT CCCATC CCAT...      1\n",
       "3   TCGCTC CGCTCC GCTCCC CTCCCA TCCCAT CCCATC CCAT...      0\n",
       "4   TCGCTC CGCTCC GCTCCC CTCCCA TCCCAT CCCATC CCAT...      1\n",
       "5   TCGCTC CGCTCC GCTCCC CTCCCA TCCCAT CCCATC CCAT...      0\n",
       "6   TCGCTC CGCTCC GCTCCC CTCCCA TCCCAT CCCATC CCAT...      0\n",
       "7   TCGCTC CGCTCC GCTCCC CTCCCA TCCCAT CCCATC CCAT...      1\n",
       "8   GAGGGC AGGGCA GGGCAG GGCAGC GCAGCG CAGCGA AGCG...      1\n",
       "9   GAGGGC AGGGCA GGGCAG GGCAGC GCAGCG CAGCGA AGCG...      0\n",
       "10  GAGGGC AGGGCA GGGCAG GGCAGC GCAGCG CAGCGA AGCG...      0\n",
       "11  GAGGGC AGGGCA GGGCAG GGCAGC GCAGCG CAGCGA AGCG...      0\n",
       "12  GAGGGC AGGGCA GGGCAG GGCAGC GCAGCG CAGCGA AGCG...      0\n",
       "13  GAGGGC AGGGCA GGGCAG GGCAGC GCAGCG CAGCGA AGCG...      1\n",
       "14  TCAGGA CAGGAG AGGAGT GGAGTT GAGTTT AGTTTT GTTT...      0\n",
       "15  TCGGGA CGGGAG GGGAGT GGAGTT GAGTTT AGTTTT GTTT...      0\n",
       "16  TCAGGA CAGGAG AGGAGT GGAGTT GAGTTT AGTTTT GTTT...      0\n",
       "17  TCAGGA CAGGAG AGGAGT GGAGTT GAGTTT AGTTTT GTTT...      1\n",
       "18  TCAGGA CAGGAG AGGAGT GGAGTT GAGTTT AGTTTT GTTT...      1\n",
       "19  TCAGGA CAGGAG AGGAGT GGAGTT GAGTTT AGTTTT GTTT...      0"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_kmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71fc5a7f-b9b2-4f68-b345-c02f59c74d7f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
