{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "48bb6eb5-6de0-4d27-848d-5c2e2eb29faf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, tarfile\n",
    "import pyBigWig\n",
    "import pybedtools\n",
    "import zipfile, gzip\n",
    "from gtfparse import read_gtf\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f4f24891-9cb6-4b3d-b75a-13c312c4bcf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_genome_path = \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/genome_by_sequence/\"\n",
    "reference_genome_name = \"chr1.fa\"\n",
    "data_path= \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/gencode_annotation/\"\n",
    "file_name=\"gencodeV38.bb\"\n",
    "gencode_path= \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/gencode_annotation/\"\n",
    "gtf_annotation_filename= \"gencode.v38.annotation.gtf\"\n",
    "gff_annotation_filename= \"gencode.v38.annotation.gff3\"\n",
    "file_path= data_path+file_name\n",
    "gencode_gtf_file_path= gencode_path+ gtf_annotation_filename\n",
    "gencode_gff_file_path= gencode_path+ gff_annotation_filename\n",
    "reference_genome_file_path = reference_genome_path + reference_genome_name\n",
    "data_save_path= \"/mnt/data05/shared/pdutta_data/Human_Genome_Data/region_wise_sequence\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "096b8442-f1dd-47dc-abe6-2f836f7f7d31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/data05/shared/pdutta_data/Human_Genome_Data/UCSC/gencode_annotation/gencodeV38.bb'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8e286976-54a1-4993-99d4-30f7b4c4880e",
   "metadata": {},
   "outputs": [],
   "source": [
    "bb= pyBigWig.open(file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bf9498db-5711-4b53-99c0-d8b8de788bcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20716\n",
      "{'version': 4, 'nLevels': 8, 'nBasesCovered': 1827967802, 'minVal': 1, 'maxVal': 226, 'sumData': 11784798039, 'sumSquared': 217150380791}\n",
      "11868 \n",
      " 14409 \n",
      " ENST00000456328.2\t0\t+\t11868\t11868\t16724991\t3\t359,109,1189,\t0,744,1352,\tuc286dmu.1\tnone\tnone\t-1,-1,-1,\tnone\tDDX11L1\tnone\tnone\tpseudo\thavana_homo_sapiens\tprocessed_transcript\tbasic\t2\tcanonical,basic,all\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "list1 = bb.entries('chr1', 0,248956422)\n",
    "#print (list1[0])\n",
    "\n",
    "print (len(list1))\n",
    "print(bb.header())\n",
    "\n",
    "for i in list1[0:1]:\n",
    "    domain, level, url = i\n",
    "    print (domain,\"\\n\",level,\"\\n\",url)\n",
    "    print (\"\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "30f808e1-6f52-447e-ba2b-525ce83222b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "248956422"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bb.chroms('chr1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5c1381e3-3325-4b15-b1b8-80b27b8ac306",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: []\n"
     ]
    }
   ],
   "source": [
    "df = read_gtf(gencode_gff_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "06945eba-fcdb-4f16-a5ec-fa94e936c8ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand',\n",
      "       'frame'],\n",
      "      dtype='object')\n",
      "(3148136, 8)\n"
     ]
    }
   ],
   "source": [
    "print(df.columns)\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "99de0a0b-20a5-4fd1-919f-89703dc8c677",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010</td>\n",
       "      <td>12057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12697</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12975</td>\n",
       "      <td>13052</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  source     feature  start    end  score strand frame\n",
       "0    chr1  HAVANA        gene  11869  14409    NaN      +     0\n",
       "1    chr1  HAVANA  transcript  11869  14409    NaN      +     0\n",
       "2    chr1  HAVANA        exon  11869  12227    NaN      +     0\n",
       "3    chr1  HAVANA        exon  12613  12721    NaN      +     0\n",
       "4    chr1  HAVANA        exon  13221  14409    NaN      +     0\n",
       "5    chr1  HAVANA  transcript  12010  13670    NaN      +     0\n",
       "6    chr1  HAVANA        exon  12010  12057    NaN      +     0\n",
       "7    chr1  HAVANA        exon  12179  12227    NaN      +     0\n",
       "8    chr1  HAVANA        exon  12613  12697    NaN      +     0\n",
       "9    chr1  HAVANA        exon  12975  13052    NaN      +     0"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c32274e1-5f2e-4a43-9db4-cf2a78b3c044",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "feature\n",
       "CDS                                        825914\n",
       "exon                                      1499012\n",
       "five_prime_UTR                             163360\n",
       "gene                                        60649\n",
       "start_codon                                 92997\n",
       "stop_codon                                  86205\n",
       "stop_codon_redefined_as_selenocysteine        119\n",
       "three_prime_UTR                            182868\n",
       "transcript                                 237012\n",
       "dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby(['feature']).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7cb5d429-63f0-4ab9-b3e6-64a62441cc51",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'hgnc_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'ont', 'protein_id', 'ccdsid']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand',\n",
      "       'frame', 'gene_id', 'gene_type', 'gene_name', 'level', 'hgnc_id',\n",
      "       'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name',\n",
      "       'transcript_support_level', 'tag', 'havana_transcript', 'exon_number',\n",
      "       'exon_id', 'ont', 'protein_id', 'ccdsid'],\n",
      "      dtype='object')\n",
      "(3150424, 25)\n"
     ]
    }
   ],
   "source": [
    "df1 = read_gtf(gencode_gtf_file_path)\n",
    "print(df1.columns)\n",
    "print(df1.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e7aff974-1410-4f95-bdb9-db269f03939c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010</td>\n",
       "      <td>12057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12697</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001758273.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12975</td>\n",
       "      <td>13052</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00001799933.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>13374</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00001746346.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13453</td>\n",
       "      <td>13670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00001863096.1</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>29534</td>\n",
       "      <td>29570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001890219.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>24738</td>\n",
       "      <td>24891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003507205.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>18268</td>\n",
       "      <td>18366</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003477500.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>17915</td>\n",
       "      <td>18061</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00003565697.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>17606</td>\n",
       "      <td>17742</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00003475637.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>17233</td>\n",
       "      <td>17368</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSE00003502542.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>16858</td>\n",
       "      <td>17055</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>7</td>\n",
       "      <td>ENSE00003553898.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>16607</td>\n",
       "      <td>16765</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSE00003621279.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>15796</td>\n",
       "      <td>15947</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>9</td>\n",
       "      <td>ENSE00002030414.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>15005</td>\n",
       "      <td>15038</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>10</td>\n",
       "      <td>ENSE00001935574.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>14404</td>\n",
       "      <td>14501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>11</td>\n",
       "      <td>ENSE00001843071.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqname  source     feature  start    end  score strand frame  \\\n",
       "0     chr1  HAVANA        gene  11869  14409    NaN      +     0   \n",
       "1     chr1  HAVANA  transcript  11869  14409    NaN      +     0   \n",
       "2     chr1  HAVANA        exon  11869  12227    NaN      +     0   \n",
       "3     chr1  HAVANA        exon  12613  12721    NaN      +     0   \n",
       "4     chr1  HAVANA        exon  13221  14409    NaN      +     0   \n",
       "5     chr1  HAVANA  transcript  12010  13670    NaN      +     0   \n",
       "6     chr1  HAVANA        exon  12010  12057    NaN      +     0   \n",
       "7     chr1  HAVANA        exon  12179  12227    NaN      +     0   \n",
       "8     chr1  HAVANA        exon  12613  12697    NaN      +     0   \n",
       "9     chr1  HAVANA        exon  12975  13052    NaN      +     0   \n",
       "10    chr1  HAVANA        exon  13221  13374    NaN      +     0   \n",
       "11    chr1  HAVANA        exon  13453  13670    NaN      +     0   \n",
       "12    chr1  HAVANA        gene  14404  29570    NaN      -     0   \n",
       "13    chr1  HAVANA  transcript  14404  29570    NaN      -     0   \n",
       "14    chr1  HAVANA        exon  29534  29570    NaN      -     0   \n",
       "15    chr1  HAVANA        exon  24738  24891    NaN      -     0   \n",
       "16    chr1  HAVANA        exon  18268  18366    NaN      -     0   \n",
       "17    chr1  HAVANA        exon  17915  18061    NaN      -     0   \n",
       "18    chr1  HAVANA        exon  17606  17742    NaN      -     0   \n",
       "19    chr1  HAVANA        exon  17233  17368    NaN      -     0   \n",
       "20    chr1  HAVANA        exon  16858  17055    NaN      -     0   \n",
       "21    chr1  HAVANA        exon  16607  16765    NaN      -     0   \n",
       "22    chr1  HAVANA        exon  15796  15947    NaN      -     0   \n",
       "23    chr1  HAVANA        exon  15005  15038    NaN      -     0   \n",
       "24    chr1  HAVANA        exon  14404  14501    NaN      -     0   \n",
       "\n",
       "              gene_id                           gene_type gene_name level  \\\n",
       "0   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "1   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "2   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "3   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "4   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "5   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "6   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "7   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "8   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "9   ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "10  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "11  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2   \n",
       "12  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "13  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "14  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "15  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "16  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "17  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "18  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "19  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "20  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "21  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "22  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "23  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "24  ENSG00000227232.5              unprocessed_pseudogene    WASH7P     2   \n",
       "\n",
       "       hgnc_id           havana_gene      transcript_id  \\\n",
       "0   HGNC:37102  OTTHUMG00000000961.2                      \n",
       "1   HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "2   HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "3   HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "4   HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "5   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "6   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "7   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "8   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "9   HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "10  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "11  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "12  HGNC:38034  OTTHUMG00000000958.1                      \n",
       "13  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "14  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "15  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "16  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "17  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "18  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "19  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "20  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "21  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "22  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "23  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "24  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "\n",
       "                       transcript_type transcript_name  \\\n",
       "0                                                        \n",
       "1                 processed_transcript     DDX11L1-202   \n",
       "2                 processed_transcript     DDX11L1-202   \n",
       "3                 processed_transcript     DDX11L1-202   \n",
       "4                 processed_transcript     DDX11L1-202   \n",
       "5   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "6   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "7   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "8   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "9   transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "10  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "11  transcribed_unprocessed_pseudogene     DDX11L1-201   \n",
       "12                                                       \n",
       "13              unprocessed_pseudogene      WASH7P-201   \n",
       "14              unprocessed_pseudogene      WASH7P-201   \n",
       "15              unprocessed_pseudogene      WASH7P-201   \n",
       "16              unprocessed_pseudogene      WASH7P-201   \n",
       "17              unprocessed_pseudogene      WASH7P-201   \n",
       "18              unprocessed_pseudogene      WASH7P-201   \n",
       "19              unprocessed_pseudogene      WASH7P-201   \n",
       "20              unprocessed_pseudogene      WASH7P-201   \n",
       "21              unprocessed_pseudogene      WASH7P-201   \n",
       "22              unprocessed_pseudogene      WASH7P-201   \n",
       "23              unprocessed_pseudogene      WASH7P-201   \n",
       "24              unprocessed_pseudogene      WASH7P-201   \n",
       "\n",
       "   transcript_support_level                      tag     havana_transcript  \\\n",
       "0                                                                            \n",
       "1                         1                    basic  OTTHUMT00000362751.1   \n",
       "2                         1                    basic  OTTHUMT00000362751.1   \n",
       "3                         1                    basic  OTTHUMT00000362751.1   \n",
       "4                         1                    basic  OTTHUMT00000362751.1   \n",
       "5                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "6                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "7                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "8                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "9                        NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "10                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "11                       NA  basic,Ensembl_canonical  OTTHUMT00000002844.2   \n",
       "12                                                                           \n",
       "13                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "14                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "15                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "16                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "17                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "18                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "19                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "20                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "21                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "22                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "23                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "24                       NA  basic,Ensembl_canonical  OTTHUMT00000002839.1   \n",
       "\n",
       "   exon_number            exon_id                      ont protein_id ccdsid  \n",
       "0                                                                             \n",
       "1                                                                             \n",
       "2            1  ENSE00002234944.1                                             \n",
       "3            2  ENSE00003582793.1                                             \n",
       "4            3  ENSE00002312635.1                                             \n",
       "5                                  PGO:0000005,PGO:0000019                    \n",
       "6            1  ENSE00001948541.1  PGO:0000005,PGO:0000019                    \n",
       "7            2  ENSE00001671638.2  PGO:0000005,PGO:0000019                    \n",
       "8            3  ENSE00001758273.2  PGO:0000005,PGO:0000019                    \n",
       "9            4  ENSE00001799933.2  PGO:0000005,PGO:0000019                    \n",
       "10           5  ENSE00001746346.2  PGO:0000005,PGO:0000019                    \n",
       "11           6  ENSE00001863096.1  PGO:0000005,PGO:0000019                    \n",
       "12                                                                            \n",
       "13                                             PGO:0000005                    \n",
       "14           1  ENSE00001890219.1              PGO:0000005                    \n",
       "15           2  ENSE00003507205.1              PGO:0000005                    \n",
       "16           3  ENSE00003477500.1              PGO:0000005                    \n",
       "17           4  ENSE00003565697.1              PGO:0000005                    \n",
       "18           5  ENSE00003475637.1              PGO:0000005                    \n",
       "19           6  ENSE00003502542.1              PGO:0000005                    \n",
       "20           7  ENSE00003553898.1              PGO:0000005                    \n",
       "21           8  ENSE00003621279.1              PGO:0000005                    \n",
       "22           9  ENSE00002030414.1              PGO:0000005                    \n",
       "23          10  ENSE00001935574.1              PGO:0000005                    \n",
       "24          11  ENSE00001843071.1              PGO:0000005                    "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1.head(25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b2f14802-3052-459b-a491-c7be58371913",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_2= df1[df1['seqname']=='chr13']\n",
    "df_3= df_2[df_2['feature']=='gene']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9256b928-759e-4a5b-9317-66dfc606f355",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [seqname, source, feature, start, end, score, strand, frame, gene_id, gene_type, gene_name, level, hgnc_id, havana_gene, transcript_id, transcript_type, transcript_name, transcript_support_level, tag, havana_transcript, exon_number, exon_id, ont, protein_id, ccdsid]\n",
       "Index: []"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "df_3.loc[df_3['start'] == 15999973]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "35cb1e37-3b25-437e-9234-3a12b7991a78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "feature\n",
       "CDS                825028\n",
       "Selenocysteine        119\n",
       "UTR                349402\n",
       "exon              1499012\n",
       "gene                60649\n",
       "start_codon         92997\n",
       "stop_codon          86205\n",
       "transcript         237012\n",
       "dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1.groupby(['feature']).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a8d031d0-606b-4067-a822-053fed748b5b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286811</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>248912690</td>\n",
       "      <td>248912795</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000200495.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>RNU6-1205P</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:48168</td>\n",
       "      <td></td>\n",
       "      <td>ENST00000363625.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>RNU6-1205P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286812</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>248912690</td>\n",
       "      <td>248912795</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000200495.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>RNU6-1205P</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:48168</td>\n",
       "      <td></td>\n",
       "      <td>ENST00000363625.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>RNU6-1205P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001438388.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286813</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>248936581</td>\n",
       "      <td>248937043</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000233084.2</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>RPL23AP25</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:35672</td>\n",
       "      <td>OTTHUMG00000040422.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>pseudo_consens</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286814</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>248936581</td>\n",
       "      <td>248937043</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000233084.2</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>RPL23AP25</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:35672</td>\n",
       "      <td>OTTHUMG00000040422.1</td>\n",
       "      <td>ENST00000430973.1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>RPL23AP25-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>pseudo_consens,basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000097297.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PGO:0000004</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286815</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>248936581</td>\n",
       "      <td>248937043</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000233084.2</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>RPL23AP25</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:35672</td>\n",
       "      <td>OTTHUMG00000040422.1</td>\n",
       "      <td>ENST00000430973.1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>RPL23AP25-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>pseudo_consens,basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000097297.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001666859.1</td>\n",
       "      <td>PGO:0000004</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>286816 rows  25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname   source     feature      start        end  score strand frame  \\\n",
       "0         chr1   HAVANA        gene      11869      14409    NaN      +     0   \n",
       "1         chr1   HAVANA  transcript      11869      14409    NaN      +     0   \n",
       "2         chr1   HAVANA        exon      11869      12227    NaN      +     0   \n",
       "3         chr1   HAVANA        exon      12613      12721    NaN      +     0   \n",
       "4         chr1   HAVANA        exon      13221      14409    NaN      +     0   \n",
       "...        ...      ...         ...        ...        ...    ...    ...   ...   \n",
       "286811    chr1  ENSEMBL  transcript  248912690  248912795    NaN      -     0   \n",
       "286812    chr1  ENSEMBL        exon  248912690  248912795    NaN      -     0   \n",
       "286813    chr1   HAVANA        gene  248936581  248937043    NaN      +     0   \n",
       "286814    chr1   HAVANA  transcript  248936581  248937043    NaN      +     0   \n",
       "286815    chr1   HAVANA        exon  248936581  248937043    NaN      +     0   \n",
       "\n",
       "                  gene_id                           gene_type   gene_name  \\\n",
       "0       ENSG00000223972.5  transcribed_unprocessed_pseudogene     DDX11L1   \n",
       "1       ENSG00000223972.5  transcribed_unprocessed_pseudogene     DDX11L1   \n",
       "2       ENSG00000223972.5  transcribed_unprocessed_pseudogene     DDX11L1   \n",
       "3       ENSG00000223972.5  transcribed_unprocessed_pseudogene     DDX11L1   \n",
       "4       ENSG00000223972.5  transcribed_unprocessed_pseudogene     DDX11L1   \n",
       "...                   ...                                 ...         ...   \n",
       "286811  ENSG00000200495.1                               snRNA  RNU6-1205P   \n",
       "286812  ENSG00000200495.1                               snRNA  RNU6-1205P   \n",
       "286813  ENSG00000233084.2                processed_pseudogene   RPL23AP25   \n",
       "286814  ENSG00000233084.2                processed_pseudogene   RPL23AP25   \n",
       "286815  ENSG00000233084.2                processed_pseudogene   RPL23AP25   \n",
       "\n",
       "       level     hgnc_id           havana_gene      transcript_id  \\\n",
       "0          2  HGNC:37102  OTTHUMG00000000961.2                      \n",
       "1          2  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "2          2  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "3          2  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "4          2  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "...      ...         ...                   ...                ...   \n",
       "286811     3  HGNC:48168                        ENST00000363625.1   \n",
       "286812     3  HGNC:48168                        ENST00000363625.1   \n",
       "286813     1  HGNC:35672  OTTHUMG00000040422.1                      \n",
       "286814     1  HGNC:35672  OTTHUMG00000040422.1  ENST00000430973.1   \n",
       "286815     1  HGNC:35672  OTTHUMG00000040422.1  ENST00000430973.1   \n",
       "\n",
       "             transcript_type transcript_name transcript_support_level  \\\n",
       "0                                                                       \n",
       "1       processed_transcript     DDX11L1-202                        1   \n",
       "2       processed_transcript     DDX11L1-202                        1   \n",
       "3       processed_transcript     DDX11L1-202                        1   \n",
       "4       processed_transcript     DDX11L1-202                        1   \n",
       "...                      ...             ...                      ...   \n",
       "286811                 snRNA  RNU6-1205P-201                       NA   \n",
       "286812                 snRNA  RNU6-1205P-201                       NA   \n",
       "286813                                                                  \n",
       "286814  processed_pseudogene   RPL23AP25-201                       NA   \n",
       "286815  processed_pseudogene   RPL23AP25-201                       NA   \n",
       "\n",
       "                                           tag     havana_transcript  \\\n",
       "0                                                                      \n",
       "1                                        basic  OTTHUMT00000362751.1   \n",
       "2                                        basic  OTTHUMT00000362751.1   \n",
       "3                                        basic  OTTHUMT00000362751.1   \n",
       "4                                        basic  OTTHUMT00000362751.1   \n",
       "...                                        ...                   ...   \n",
       "286811                 basic,Ensembl_canonical                         \n",
       "286812                 basic,Ensembl_canonical                         \n",
       "286813                          pseudo_consens                         \n",
       "286814  pseudo_consens,basic,Ensembl_canonical  OTTHUMT00000097297.1   \n",
       "286815  pseudo_consens,basic,Ensembl_canonical  OTTHUMT00000097297.1   \n",
       "\n",
       "       exon_number            exon_id          ont protein_id ccdsid  \n",
       "0                                                                     \n",
       "1                                                                     \n",
       "2                1  ENSE00002234944.1                                 \n",
       "3                2  ENSE00003582793.1                                 \n",
       "4                3  ENSE00002312635.1                                 \n",
       "...            ...                ...          ...        ...    ...  \n",
       "286811                                                                \n",
       "286812           1  ENSE00001438388.1                                 \n",
       "286813                                                                \n",
       "286814                                 PGO:0000004                    \n",
       "286815           1  ENSE00001666859.1  PGO:0000004                    \n",
       "\n",
       "[286816 rows x 25 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_chr1 = df1[df1['seqname']=='chr1']\n",
    "df_chr1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "40fe2ce5-5b5e-41cc-a178-09a77c9c8358",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta = pybedtools.example_filename(reference_genome_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9c5f9184-a7db-4110-8fe7-b24acb8988c0",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 chr1:11869-14409\n",
      "1 chr1:11869-14409\n",
      "2 chr1:11869-12227\n",
      "3 chr1:12613-12721\n",
      "4 chr1:13221-14409\n",
      "5 chr1:12010-13670\n",
      "6 chr1:12010-12057\n",
      "7 chr1:12179-12227\n",
      "8 chr1:12613-12697\n",
      "9 chr1:12975-13052\n",
      "10 chr1:13221-13374\n",
      "11 chr1:13453-13670\n",
      "12 chr1:14404-29570\n",
      "13 chr1:14404-29570\n",
      "14 chr1:29534-29570\n",
      "15 chr1:24738-24891\n",
      "16 chr1:18268-18366\n",
      "17 chr1:17915-18061\n",
      "18 chr1:17606-17742\n",
      "19 chr1:17233-17368\n",
      "20 chr1:16858-17055\n",
      "21 chr1:16607-16765\n",
      "22 chr1:15796-15947\n",
      "23 chr1:15005-15038\n",
      "24 chr1:14404-14501\n",
      "25 chr1:17369-17436\n",
      "26 chr1:17369-17436\n",
      "27 chr1:17369-17436\n",
      "28 chr1:29554-31109\n",
      "29 chr1:29554-31097\n",
      "30 chr1:29554-30039\n",
      "31 chr1:30564-30667\n",
      "32 chr1:30976-31097\n",
      "33 chr1:30267-31109\n",
      "34 chr1:30267-30667\n",
      "35 chr1:30976-31109\n",
      "36 chr1:30366-30503\n",
      "37 chr1:30366-30503\n",
      "38 chr1:30366-30503\n",
      "39 chr1:34554-36081\n"
     ]
    }
   ],
   "source": [
    "df_transcript = pd.DataFrame(columns= ['transcript_id','gene_id','transcript_type','transcript_name','sequence'])\n",
    "df_exon = pd.DataFrame(columns= ['exon_id','exon_number','transcript_id','gene_id','transcript_type','gene_type','sequence'])\n",
    "df_gene = pd.DataFrame(columns= ['gene_id','gene_type','gene_name','sequence'])\n",
    "transcript_index=exon_index = gene_index= 0\n",
    "for index, row in df_chr1.head(40).iterrows():\n",
    "    #print (\"###\", index)\n",
    "    if (row['feature'] ==\"gene\"):\n",
    "        df_gene.loc[gene_index, ['gene_id']]= row['gene_id']\n",
    "        df_gene.loc[gene_index, ['gene_type']]= row['gene_type']\n",
    "        df_gene.loc[gene_index, ['gene_name']]= row['gene_name']\n",
    "        query= row['seqname']+':'+str(row['start'])+'-'+str(row['end'])\n",
    "        print (index, query)\n",
    "        df_gene.loc[gene_index, ['sequence']] = BedTool.seq(query, fasta)\n",
    "        gene_index= gene_index+1\n",
    "    elif (row['feature'] ==\"transcript\"):\n",
    "        df_transcript.loc[transcript_index, ['transcript_id']]= row['transcript_id']\n",
    "        df_transcript.loc[transcript_index, ['gene_id']]= row['gene_id']\n",
    "        df_transcript.loc[transcript_index, ['transcript_type']]= row['transcript_type']\n",
    "        df_transcript.loc[transcript_index, ['transcript_name']]= row['transcript_name']\n",
    "        query= row['seqname']+':'+str(row['start'])+'-'+str(row['end'])\n",
    "        print (index, query)\n",
    "        df_transcript.loc[transcript_index, ['sequence']] = BedTool.seq(query, fasta)\n",
    "        transcript_index= transcript_index+1\n",
    "    elif (row['feature'] ==\"exon\"):\n",
    "        df_exon.loc[exon_index, ['exon_id']]= row['exon_id']\n",
    "        df_exon.loc[exon_index, ['exon_number']]= row['exon_number']\n",
    "        df_exon.loc[exon_index, ['transcript_id']]= row['transcript_id']\n",
    "        df_exon.loc[exon_index, ['gene_id']]= row['gene_id']\n",
    "        df_exon.loc[exon_index, ['transcript_type']]= row['transcript_type']\n",
    "        df_exon.loc[exon_index, ['gene_type']]= row['gene_type']\n",
    "        query= row['seqname']+':'+str(row['start'])+'-'+str(row['end'])\n",
    "        print (index, query)\n",
    "        df_exon.loc[exon_index, ['sequence']] = BedTool.seq(query, fasta)\n",
    "        exon_index= exon_index+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "77a2f7e5-bffb-4796-84bd-c94e6d7a5d6e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>GTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>TTTCTGCTCAGTTCTTTATTGATTGGTGTGCCGTTTTCTCTGGAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>CTACAGAGGCGACATGGGGGTCAGGCAAGCTGACACCCGCTGTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>GTGCACACGGCTCCCATGCGTTGTCTTCCGAGCGTCAGGCCGCCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>ggatgcccagctagtttgaattttagataaacaacgaataatttcg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>agcagcaGGAGtgttttaattaaaagaaggcagttgctgtaaccaa...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_id                           gene_type    gene_name  \\\n",
       "0  ENSG00000223972.5  transcribed_unprocessed_pseudogene      DDX11L1   \n",
       "1  ENSG00000227232.5              unprocessed_pseudogene       WASH7P   \n",
       "2  ENSG00000278267.1                               miRNA    MIR6859-1   \n",
       "3  ENSG00000243485.5                              lncRNA  MIR1302-2HG   \n",
       "4  ENSG00000284332.1                               miRNA    MIR1302-2   \n",
       "5  ENSG00000237613.2                              lncRNA      FAM138A   \n",
       "\n",
       "                                            sequence  \n",
       "0  GTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCAT...  \n",
       "1  TTTCTGCTCAGTTCTTTATTGATTGGTGTGCCGTTTTCTCTGGAAG...  \n",
       "2  CTACAGAGGCGACATGGGGGTCAGGCAAGCTGACACCCGCTGTCCT...  \n",
       "3  GTGCACACGGCTCCCATGCGTTGTCTTCCGAGCGTCAGGCCGCCCC...  \n",
       "4  ggatgcccagctagtttgaattttagataaacaacgaataatttcg...  \n",
       "5  agcagcaGGAGtgttttaattaaaagaaggcagttgctgtaaccaa...  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4ef37fa4-28e6-42ca-be75-4e1c8bdccc32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>GTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>GTGTCTGACTTCCAGCAACTGCTGGCCTGTGCCAGGGTGCAAGCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>TTTCTGCTCAGTTCTTTATTGATTGGTGTGCCGTTTTCTCTGGAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1-201</td>\n",
       "      <td>CTACAGAGGCGACATGGGGGTCAGGCAAGCTGACACCCGCTGTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-202</td>\n",
       "      <td>GTGCACACGGCTCCCATGCGTTGTCTTCCGAGCGTCAGGCCGCCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-201</td>\n",
       "      <td>TCATCAGTCCAAAGTCCAGCAGTTGTCCCTCCTGGAATCCGTTGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000607096.1</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR1302-2-201</td>\n",
       "      <td>ggatgcccagctagtttgaattttagataaacaacgaataatttcg...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id            gene_id                     transcript_type  \\\n",
       "0  ENST00000456328.2  ENSG00000223972.5                processed_transcript   \n",
       "1  ENST00000450305.2  ENSG00000223972.5  transcribed_unprocessed_pseudogene   \n",
       "2  ENST00000488147.1  ENSG00000227232.5              unprocessed_pseudogene   \n",
       "3  ENST00000619216.1  ENSG00000278267.1                               miRNA   \n",
       "4  ENST00000473358.1  ENSG00000243485.5                              lncRNA   \n",
       "5  ENST00000469289.1  ENSG00000243485.5                              lncRNA   \n",
       "6  ENST00000607096.1  ENSG00000284332.1                               miRNA   \n",
       "\n",
       "   transcript_name                                           sequence  \n",
       "0      DDX11L1-202  GTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCAT...  \n",
       "1      DDX11L1-201  GTGTCTGACTTCCAGCAACTGCTGGCCTGTGCCAGGGTGCAAGCTG...  \n",
       "2       WASH7P-201  TTTCTGCTCAGTTCTTTATTGATTGGTGTGCCGTTTTCTCTGGAAG...  \n",
       "3    MIR6859-1-201  CTACAGAGGCGACATGGGGGTCAGGCAAGCTGACACCCGCTGTCCT...  \n",
       "4  MIR1302-2HG-202  GTGCACACGGCTCCCATGCGTTGTCTTCCGAGCGTCAGGCCGCCCC...  \n",
       "5  MIR1302-2HG-201  TCATCAGTCCAAAGTCCAGCAGTTGTCCCTCCTGGAATCCGTTGGC...  \n",
       "6    MIR1302-2-201  ggatgcccagctagtttgaattttagataaacaacgaataatttcg...  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ba3815fb-d278-43df-a1cf-2a5c259682af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>GTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>GTGTGTGGTGATGCCAGGCATGCCCTTCCCCAGCATCAGGTCTCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>GCAGGGCCATCAGGCACCAAAGGGATTCTGCCAGCATAGTGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>GTGTCTGACTTCCAGCAACTGCTGGCCTGTGCCAGGGTGCAAGCTGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>TTGGAGGAAAGATGAGTGAGAGCATCAACTTCTCTCACAACCTAGGCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSE00001758273.2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>GTGTGTGGTGATGCCAGGCATGCCCTTCCCCAGCATCAGGTCTCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSE00001799933.2</td>\n",
       "      <td>4</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>CTCAGAGCCCAGGCCAGGGGCCCCCAAGAAAGGCTCTGGTGGAGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSE00001746346.2</td>\n",
       "      <td>5</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>GCAGGGCCATCAGGCACCAAAGGGATTCTGCCAGCATAGTGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSE00001863096.1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>TTTCACCAGAAGTAGGCCTCTTCCTGACAGGCAGCTGCACCACTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSE00001890219.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>GCTCACTCCGAGCTCCCGACGTGCACACGGCTCCCAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSE00003507205.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>CTGCTGAAGATGTCTCCAGAGACCTTCTGCAGGTACTGCAGGGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSE00003477500.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>CTTGATGGCCTTCTTGCTGCCCTTGATCTTCTCAATCTTGGCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSE00003565697.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>CTGCAGGGCCCGCTCGTCCAGGGGGCGGTGCTTGCTCTGGATCCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSE00003475637.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>AGGTTCTCGGTGGTGTTGAAGAGCAGCAAGGAGCTGACAGAGCTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSE00003502542.1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>CTGCTGTTCCAGCTGCTCTCTCTTGCTGATGGACAAGGGGGCATCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSE00003553898.1</td>\n",
       "      <td>7</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>CTTGAGAGGCTCGGCTACCTCAGTGTGGAAGGTGGGCAGTTCTGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENSE00003621279.1</td>\n",
       "      <td>8</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>CTGAAGGAGACGCGCTGCTGCTGCTGTCGTCCTGCCTGGCGCCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENSE00002030414.1</td>\n",
       "      <td>9</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>CTtgctcctgctccttctgctgctgcttctccagctttcgctcctt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSE00001935574.1</td>\n",
       "      <td>10</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>GACATCAAGTGCCCACCTTGGCTCGTGGCTCTCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSE00001843071.1</td>\n",
       "      <td>11</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>TTTCTGCTCAGTTCTTTATTGATTGGTGTGCCGTTTTCTCTGGAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSE00003746039.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>CTACAGAGGCGACATGGGGGTCAGGCAAGCTGACACCCGCTGTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENSE00001947070.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>GTGCACACGGCTCCCATGCGTTGTCTTCCGAGCGTCAGGCCGCCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENSE00001922571.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TTTTGTCCGCCTTCCCTGCCTCCTCTTCTGGGGGAGTTAGATCGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENSE00001827679.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>tgaggatgcgaagagaaggtggctgtctgcaaaccaggaagagagc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENSE00001841699.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TCATCAGTCCAAAGTCCAGCAGTTGTCCCTCCTGGAATCCGTTGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENSE00001890064.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>tgaggatgcgaagagaaggtggctgtctgcaaaccaggaagagagc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENSE00003695741.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000607096.1</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>ggatgcccagctagtttgaattttagataaacaacgaataatttcg...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              exon_id exon_number      transcript_id            gene_id  \\\n",
       "0   ENSE00002234944.1           1  ENST00000456328.2  ENSG00000223972.5   \n",
       "1   ENSE00003582793.1           2  ENST00000456328.2  ENSG00000223972.5   \n",
       "2   ENSE00002312635.1           3  ENST00000456328.2  ENSG00000223972.5   \n",
       "3   ENSE00001948541.1           1  ENST00000450305.2  ENSG00000223972.5   \n",
       "4   ENSE00001671638.2           2  ENST00000450305.2  ENSG00000223972.5   \n",
       "5   ENSE00001758273.2           3  ENST00000450305.2  ENSG00000223972.5   \n",
       "6   ENSE00001799933.2           4  ENST00000450305.2  ENSG00000223972.5   \n",
       "7   ENSE00001746346.2           5  ENST00000450305.2  ENSG00000223972.5   \n",
       "8   ENSE00001863096.1           6  ENST00000450305.2  ENSG00000223972.5   \n",
       "9   ENSE00001890219.1           1  ENST00000488147.1  ENSG00000227232.5   \n",
       "10  ENSE00003507205.1           2  ENST00000488147.1  ENSG00000227232.5   \n",
       "11  ENSE00003477500.1           3  ENST00000488147.1  ENSG00000227232.5   \n",
       "12  ENSE00003565697.1           4  ENST00000488147.1  ENSG00000227232.5   \n",
       "13  ENSE00003475637.1           5  ENST00000488147.1  ENSG00000227232.5   \n",
       "14  ENSE00003502542.1           6  ENST00000488147.1  ENSG00000227232.5   \n",
       "15  ENSE00003553898.1           7  ENST00000488147.1  ENSG00000227232.5   \n",
       "16  ENSE00003621279.1           8  ENST00000488147.1  ENSG00000227232.5   \n",
       "17  ENSE00002030414.1           9  ENST00000488147.1  ENSG00000227232.5   \n",
       "18  ENSE00001935574.1          10  ENST00000488147.1  ENSG00000227232.5   \n",
       "19  ENSE00001843071.1          11  ENST00000488147.1  ENSG00000227232.5   \n",
       "20  ENSE00003746039.1           1  ENST00000619216.1  ENSG00000278267.1   \n",
       "21  ENSE00001947070.1           1  ENST00000473358.1  ENSG00000243485.5   \n",
       "22  ENSE00001922571.1           2  ENST00000473358.1  ENSG00000243485.5   \n",
       "23  ENSE00001827679.1           3  ENST00000473358.1  ENSG00000243485.5   \n",
       "24  ENSE00001841699.1           1  ENST00000469289.1  ENSG00000243485.5   \n",
       "25  ENSE00001890064.1           2  ENST00000469289.1  ENSG00000243485.5   \n",
       "26  ENSE00003695741.1           1  ENST00000607096.1  ENSG00000284332.1   \n",
       "\n",
       "                       transcript_type                           gene_type  \\\n",
       "0                 processed_transcript  transcribed_unprocessed_pseudogene   \n",
       "1                 processed_transcript  transcribed_unprocessed_pseudogene   \n",
       "2                 processed_transcript  transcribed_unprocessed_pseudogene   \n",
       "3   transcribed_unprocessed_pseudogene  transcribed_unprocessed_pseudogene   \n",
       "4   transcribed_unprocessed_pseudogene  transcribed_unprocessed_pseudogene   \n",
       "5   transcribed_unprocessed_pseudogene  transcribed_unprocessed_pseudogene   \n",
       "6   transcribed_unprocessed_pseudogene  transcribed_unprocessed_pseudogene   \n",
       "7   transcribed_unprocessed_pseudogene  transcribed_unprocessed_pseudogene   \n",
       "8   transcribed_unprocessed_pseudogene  transcribed_unprocessed_pseudogene   \n",
       "9               unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "10              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "11              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "12              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "13              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "14              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "15              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "16              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "17              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "18              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "19              unprocessed_pseudogene              unprocessed_pseudogene   \n",
       "20                               miRNA                               miRNA   \n",
       "21                              lncRNA                              lncRNA   \n",
       "22                              lncRNA                              lncRNA   \n",
       "23                              lncRNA                              lncRNA   \n",
       "24                              lncRNA                              lncRNA   \n",
       "25                              lncRNA                              lncRNA   \n",
       "26                               miRNA                               miRNA   \n",
       "\n",
       "                                             sequence  \n",
       "0   GTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCAT...  \n",
       "1   GTGTGTGGTGATGCCAGGCATGCCCTTCCCCAGCATCAGGTCTCCA...  \n",
       "2   GCAGGGCCATCAGGCACCAAAGGGATTCTGCCAGCATAGTGCTCCT...  \n",
       "3    GTGTCTGACTTCCAGCAACTGCTGGCCTGTGCCAGGGTGCAAGCTGAG  \n",
       "4   TTGGAGGAAAGATGAGTGAGAGCATCAACTTCTCTCACAACCTAGGCCA  \n",
       "5   GTGTGTGGTGATGCCAGGCATGCCCTTCCCCAGCATCAGGTCTCCA...  \n",
       "6   CTCAGAGCCCAGGCCAGGGGCCCCCAAGAAAGGCTCTGGTGGAGAA...  \n",
       "7   GCAGGGCCATCAGGCACCAAAGGGATTCTGCCAGCATAGTGCTCCT...  \n",
       "8   TTTCACCAGAAGTAGGCCTCTTCCTGACAGGCAGCTGCACCACTGC...  \n",
       "9               GCTCACTCCGAGCTCCCGACGTGCACACGGCTCCCAT  \n",
       "10  CTGCTGAAGATGTCTCCAGAGACCTTCTGCAGGTACTGCAGGGCAT...  \n",
       "11  CTTGATGGCCTTCTTGCTGCCCTTGATCTTCTCAATCTTGGCCTGG...  \n",
       "12  CTGCAGGGCCCGCTCGTCCAGGGGGCGGTGCTTGCTCTGGATCCTG...  \n",
       "13  AGGTTCTCGGTGGTGTTGAAGAGCAGCAAGGAGCTGACAGAGCTGA...  \n",
       "14  CTGCTGTTCCAGCTGCTCTCTCTTGCTGATGGACAAGGGGGCATCA...  \n",
       "15  CTTGAGAGGCTCGGCTACCTCAGTGTGGAAGGTGGGCAGTTCTGGA...  \n",
       "16  CTGAAGGAGACGCGCTGCTGCTGCTGTCGTCCTGCCTGGCGCCTTG...  \n",
       "17  CTtgctcctgctccttctgctgctgcttctccagctttcgctcctt...  \n",
       "18                 GACATCAAGTGCCCACCTTGGCTCGTGGCTCTCA  \n",
       "19  TTTCTGCTCAGTTCTTTATTGATTGGTGTGCCGTTTTCTCTGGAAG...  \n",
       "20  CTACAGAGGCGACATGGGGGTCAGGCAAGCTGACACCCGCTGTCCT...  \n",
       "21  GTGCACACGGCTCCCATGCGTTGTCTTCCGAGCGTCAGGCCGCCCC...  \n",
       "22  TTTTGTCCGCCTTCCCTGCCTCCTCTTCTGGGGGAGTTAGATCGAG...  \n",
       "23  tgaggatgcgaagagaaggtggctgtctgcaaaccaggaagagagc...  \n",
       "24  TCATCAGTCCAAAGTCCAGCAGTTGTCCCTCCTGGAATCCGTTGGC...  \n",
       "25  tgaggatgcgaagagaaggtggctgtctgcaaaccaggaagagagc...  \n",
       "26  ggatgcccagctagtttgaattttagataaacaacgaataatttcg...  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "aacbcd0d-bb32-4b8d-81e9-bba99ac9e342",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = pybedtools.BedTool(gencode_gff_file_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "5ad8d2ef-d97d-4972-8c3c-98b604e90224",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\tHAVANA\tgene\t11869\t14409\t.\t+\t.\tID=ENSG00000223972.5;gene_id=ENSG00000223972.5;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;level=2;hgnc_id=HGNC:37102;havana_gene=OTTHUMG00000000961.2\n",
      "\n",
      "chr1\tHAVANA\ttranscript\t11869\t14409\t.\t+\t.\tID=ENST00000456328.2;Parent=ENSG00000223972.5;gene_id=ENSG00000223972.5;transcript_id=ENST00000456328.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=processed_transcript;transcript_name=DDX11L1-202;level=2;transcript_support_level=1;hgnc_id=HGNC:37102;tag=basic;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000362751.1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for i in a[0:2]:\n",
    "    print (i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "85f2fd7b-6872-45c2-916b-b9f2e42e4809",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTATTTGCTGTCTCTTAGCCCAGACTTCCCGTGTCCTTTCCACCGGGCCTTTGAGAGGTCACAGGGTCTTGATGCTGTGGTCTTCATCTGCAGGTGTCTGACTTCCAGCAACTGCTGGCCTGTGCCAGGGTGCAAGCTGAGCACTGGAGTGGAGTTTTCCTGTGGAGAGGAGCCATGCCTAGAGTGGGATGGGCCATTGTTCATCTTCTGGCCCCTGTTGTCTGCATGTAACTTAATACCACAACCAGGCATAGGGGAAAGATTGGAGGAAAGATGAGTGAGAGCATCAACTTCTCTCACAACCTAGGCCAGTAAGTAGTGCTTGTGCTCATCTCCTTGGCTGTGATACGTGGCCGGCCCTCGCTCCAGCAGCTGGACCCCTACCTGCCGTCTGCTGCCATCGGAGCCCAAAGCCGGGCTGTGACTGCTCAGACCAGCCGGCTGGAGGGAGGGGCTCAGCAGGTCTGGCTTTGGCCCTGGGAGAGCAGGTGGAAGATCAGGCAGGCCATCGCTGCCACAGAACCCAGTGGATTGGCCTAGGTGGGATCTCTGAGCTCAACAAGCCCTCTCTGGGTGGTAGGTGCAGAGACGGGAGGGGCAGAGCCGCAGGCACAGCCAAGAGGGCTGAAGAAATGGTAGAACGGAGCAGCTGGTGATGTGTGGGCCCACCGGCCCCAGGCTCCTGTCTCCCCCCAGGTGTGTGGTGATGCCAGGCATGCCCTTCCCCAGCATCAGGTCTCCAGAGCTGCAGAAGACGACGGCCGACTTGGATCACACTCTTGTGAGTGTCCCCAGTGTTGCAGAGGTGAGAGGAGAGTAGACAGTGAGTGGGAGTGGCGTCGCCCCTAGGGCTCTACGGGGCCGGCGTCTCCTGTCTCCTGGAGAGGCTTCGATGCCCCTCCACACCCTCTTGATCTTCCCTGTGATGTCATCTGGAGCCCTGCTGCTTGCGGTGGCCTATAAAGCCTCCTAGTCTGGCTCCAAGGCCTGGCAGAGTCTTTCCCAGGGAAAGCTACAAGCAGCAAACAGTCTGCATGGGTCATCCCCTTCACTCCCAGCTCAGAGCCCAGGCCAGGGGCCCCCAAGAAAGGCTCTGGTGGAGAACCTGTGCATGAAGGCTGTCAACCAGTCCATAGGCAAGCCTGGCTGCCTCCAGCTGGGTCGACAGACAGGGGCTGGAGAAGGGGAGAAGAGGAAAGTGAGGTTGCCTGCCCTGTCTCCTACCTGAGGCTGAGGAAGGAGAAGGGGATGCACTGTTGGGGAGGCAGCTGTAACTCAAAGCCTTAGCCTCTGTTCCCACGAAGGCAGGGCCATCAGGCACCAAAGGGATTCTGCCAGCATAGTGCTCCTGGACCAGTGATACACCCGGCACCCTGTCCTGGACACGCTGTTGGCCTGGATCTGAGCCCTGGTGGAGGTCAAAGCCACCTTTGGTTCTGCCATTGCTGCTGTGTGGAAGTTCACTCCTGCCTTTTCCTTTCCCTAGAGCCTCCACCACCCCGAGATCACATTTCTCACTGCCTTTTGTCTGCCCAGTTTCACCAGAAGTAGGCCTCTTCCTGACAGGCAGCTGCACCACTGCCTGGCGCTGTGCCCTTCCTTTGCTCTGCCCGCTGGAGACGGTGTTTGTCATGGGCCTGGTCTGCAGGGATCCTGCTACAAAGGTGAAACCCAGGAGAGTGTGGAGTCCAGAGTGTTGCCAGGACCCAGGCACAGGCATTAGTGCCCGTTGGAGAAAACAGGGGAATCCCGAAGAAATGGTGGGTCCTGGCCATCCGTGAGATCTTCCCAGGGCAGCTCCCCTCTGTGGAATCCAATCTGTCTTCCATCCTGCGTGGCCGAGGGCCAGGCTTCTCACTGGGCCTCTGCAGGAGGCTGCCATTTGTCCTGCCCACCTTCTTAGAAGCGAGACGGAGCAGACCCATCTGCTACTGCCCTTTCTATAATAACTAAAGTTAGCTGCCCTGGACTATTCACCCCCTAGTCTCAATTTAAGAAGATCCCCATGGCCACAGGGCCCCTGCCTGGGGGCTTGTCACCTCCCCCACCTTCTTCCTGAGTCATTCCTGCAGCCTTGCTCCCTAACCTGCCCCACAGCCTTGCCTGGATTTCTATCTCCCTGGCTTGGTGCCAGTTCCTCCAAGTCGATGGCACCTCCCTCCCTCTCAACCACTTGAGCAAACTCCAAGACATCTTCTACCCCAACACCAGCAATTGTGCCAAGGGCCATTAGGCTCTCAGCATGACTATTTTTAGAGACCCCGTGTCTGTCACTGAAACCTTTTTTGTGGGAGACTATTCCTCCCATCTGCAACAGCTGCCCCTGCTGACTGCCCTTCTCTCCTCCCTCTCATCCCAGAGAAACAGGTCAGCTGGGAGCTTCTGCCCCCACTGCCTAGGGACCAACAGGGGCAGGAGGCAGTCACTGACCCCGAGACGTTTGCATCCTGCACAGCTAGAGATCCTTTATTAAAAGCACACTGTTGGTTTCTG'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BedTool.seq('chr1:11869-14409', fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "297c6d07-7e21-44d7-89fb-5e31c54e4c05",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_exon.to_csv(data_save_path +\"/\"+ row['seqname']+\"_exon_sequence.csv\", index=False)\n",
    "df_transcript.to_csv(data_save_path +\"/\"+ row['seqname']+\"_transcript_sequence.csv\", index=False)\n",
    "df_gene.to_csv(data_save_path +\"/\"+ row['seqname']+\"_gene_sequence.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e5827cd-0942-4812-a7de-5a880a371c34",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8dcec4ae-0d9a-49af-b046-04c107e2a0a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">chr1:1-10\n",
      "GATGAGTCT\n",
      ">chr1:50-55\n",
      "CCATC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "a = pybedtools.BedTool('''\n",
    "chr1 1 10\n",
    "chr1 50 55''', from_string=True)\n",
    "fasta = pybedtools.example_filename('test.fa')\n",
    "a = a.sequence(fi=fasta)\n",
    "print(open(a.seqfn).read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d6567b85-bbf4-4245-906b-be1af53a30ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "b = a.save_seqs('example.fa')\n",
    "assert open(b.fn).read() == open(a.fn).read()\n",
    "if os.path.exists('example.fa'):\n",
    "    os.unlink('example.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "28b13586-5eae-4aec-b002-d54f0c97ba5b",
   "metadata": {},
   "outputs": [
    {
     "ename": "SystemError",
     "evalue": "<method 'chroms' of 'pyBigWig.bigWigFile' objects> returned a result with an error set",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;31mTypeError\u001b[0m: argument 1 must be str, not int",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mSystemError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_1463639/564859714.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mbb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchroms\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mSystemError\u001b[0m: <method 'chroms' of 'pyBigWig.bigWigFile' objects> returned a result with an error set"
     ]
    }
   ],
   "source": [
    "bb.chroms(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "558c0161-5b2a-40f7-ab3e-f59f19032130",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_new= df1[df1['feature']=='transcript']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "394a42b8-3286-4536-a83b-ac62745448f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:50039</td>\n",
       "      <td></td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29554</td>\n",
       "      <td>31097</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:52482</td>\n",
       "      <td>OTTHUMG00000000959.2</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-202</td>\n",
       "      <td>5</td>\n",
       "      <td>dotter_confirmed,not_best_in_genome_evidence,b...</td>\n",
       "      <td>OTTHUMT00000002840.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3150406</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7462</td>\n",
       "      <td></td>\n",
       "      <td>ENST00000361681.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND6-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical,appris_principal_1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000354665.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3150411</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TE</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7479</td>\n",
       "      <td></td>\n",
       "      <td>ENST00000387459.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TE-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3150414</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7427</td>\n",
       "      <td></td>\n",
       "      <td>ENST00000361789.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-CYB-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical,appris_principal_1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000354554.2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3150419</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7499</td>\n",
       "      <td></td>\n",
       "      <td>ENST00000387460.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3150422</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7494</td>\n",
       "      <td></td>\n",
       "      <td>ENST00000387461.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>237012 rows  25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname   source     feature  start    end  score strand frame  \\\n",
       "1          chr1   HAVANA  transcript  11869  14409    NaN      +     0   \n",
       "5          chr1   HAVANA  transcript  12010  13670    NaN      +     0   \n",
       "13         chr1   HAVANA  transcript  14404  29570    NaN      -     0   \n",
       "26         chr1  ENSEMBL  transcript  17369  17436    NaN      -     0   \n",
       "29         chr1   HAVANA  transcript  29554  31097    NaN      +     0   \n",
       "...         ...      ...         ...    ...    ...    ...    ...   ...   \n",
       "3150406    chrM  ENSEMBL  transcript  14149  14673    NaN      -     0   \n",
       "3150411    chrM  ENSEMBL  transcript  14674  14742    NaN      -     0   \n",
       "3150414    chrM  ENSEMBL  transcript  14747  15887    NaN      +     0   \n",
       "3150419    chrM  ENSEMBL  transcript  15888  15953    NaN      +     0   \n",
       "3150422    chrM  ENSEMBL  transcript  15956  16023    NaN      -     0   \n",
       "\n",
       "                   gene_id                           gene_type    gene_name  \\\n",
       "1        ENSG00000223972.5  transcribed_unprocessed_pseudogene      DDX11L1   \n",
       "5        ENSG00000223972.5  transcribed_unprocessed_pseudogene      DDX11L1   \n",
       "13       ENSG00000227232.5              unprocessed_pseudogene       WASH7P   \n",
       "26       ENSG00000278267.1                               miRNA    MIR6859-1   \n",
       "29       ENSG00000243485.5                              lncRNA  MIR1302-2HG   \n",
       "...                    ...                                 ...          ...   \n",
       "3150406  ENSG00000198695.2                      protein_coding       MT-ND6   \n",
       "3150411  ENSG00000210194.1                             Mt_tRNA        MT-TE   \n",
       "3150414  ENSG00000198727.2                      protein_coding       MT-CYB   \n",
       "3150419  ENSG00000210195.2                             Mt_tRNA        MT-TT   \n",
       "3150422  ENSG00000210196.2                             Mt_tRNA        MT-TP   \n",
       "\n",
       "        level     hgnc_id           havana_gene      transcript_id  \\\n",
       "1           2  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "5           2  HGNC:37102  OTTHUMG00000000961.2  ENST00000450305.2   \n",
       "13          2  HGNC:38034  OTTHUMG00000000958.1  ENST00000488147.1   \n",
       "26          3  HGNC:50039                        ENST00000619216.1   \n",
       "29          2  HGNC:52482  OTTHUMG00000000959.2  ENST00000473358.1   \n",
       "...       ...         ...                   ...                ...   \n",
       "3150406     3   HGNC:7462                        ENST00000361681.2   \n",
       "3150411     3   HGNC:7479                        ENST00000387459.1   \n",
       "3150414     3   HGNC:7427                        ENST00000361789.2   \n",
       "3150419     3   HGNC:7499                        ENST00000387460.2   \n",
       "3150422     3   HGNC:7494                        ENST00000387461.2   \n",
       "\n",
       "                            transcript_type  transcript_name  \\\n",
       "1                      processed_transcript      DDX11L1-202   \n",
       "5        transcribed_unprocessed_pseudogene      DDX11L1-201   \n",
       "13                   unprocessed_pseudogene       WASH7P-201   \n",
       "26                                    miRNA    MIR6859-1-201   \n",
       "29                                   lncRNA  MIR1302-2HG-202   \n",
       "...                                     ...              ...   \n",
       "3150406                      protein_coding       MT-ND6-201   \n",
       "3150411                             Mt_tRNA        MT-TE-201   \n",
       "3150414                      protein_coding       MT-CYB-201   \n",
       "3150419                             Mt_tRNA        MT-TT-201   \n",
       "3150422                             Mt_tRNA        MT-TP-201   \n",
       "\n",
       "        transcript_support_level  \\\n",
       "1                              1   \n",
       "5                             NA   \n",
       "13                            NA   \n",
       "26                            NA   \n",
       "29                             5   \n",
       "...                          ...   \n",
       "3150406                       NA   \n",
       "3150411                       NA   \n",
       "3150414                       NA   \n",
       "3150419                       NA   \n",
       "3150422                       NA   \n",
       "\n",
       "                                                       tag  \\\n",
       "1                                                    basic   \n",
       "5                                  basic,Ensembl_canonical   \n",
       "13                                 basic,Ensembl_canonical   \n",
       "26                                 basic,Ensembl_canonical   \n",
       "29       dotter_confirmed,not_best_in_genome_evidence,b...   \n",
       "...                                                    ...   \n",
       "3150406         basic,Ensembl_canonical,appris_principal_1   \n",
       "3150411                            basic,Ensembl_canonical   \n",
       "3150414         basic,Ensembl_canonical,appris_principal_1   \n",
       "3150419                            basic,Ensembl_canonical   \n",
       "3150422                            basic,Ensembl_canonical   \n",
       "\n",
       "            havana_transcript exon_number exon_id                      ont  \\\n",
       "1        OTTHUMT00000362751.1                                                \n",
       "5        OTTHUMT00000002844.2                      PGO:0000005,PGO:0000019   \n",
       "13       OTTHUMT00000002839.1                                  PGO:0000005   \n",
       "26                                                                           \n",
       "29       OTTHUMT00000002840.1                                                \n",
       "...                       ...         ...     ...                      ...   \n",
       "3150406                                                                      \n",
       "3150411                                                                      \n",
       "3150414                                                                      \n",
       "3150419                                                                      \n",
       "3150422                                                                      \n",
       "\n",
       "                protein_id ccdsid  \n",
       "1                                  \n",
       "5                                  \n",
       "13                                 \n",
       "26                                 \n",
       "29                                 \n",
       "...                    ...    ...  \n",
       "3150406  ENSP00000354665.2         \n",
       "3150411                            \n",
       "3150414  ENSP00000354554.2         \n",
       "3150419                            \n",
       "3150422                            \n",
       "\n",
       "[237012 rows x 25 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1531b2d5-5786-428a-9a5b-692c0466cbed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_type\n",
       "IG_C_gene                                 23\n",
       "IG_C_pseudogene                            9\n",
       "IG_D_gene                                 37\n",
       "IG_J_gene                                 18\n",
       "IG_J_pseudogene                            3\n",
       "IG_V_gene                                145\n",
       "IG_V_pseudogene                          187\n",
       "IG_pseudogene                              1\n",
       "Mt_rRNA                                    2\n",
       "Mt_tRNA                                   22\n",
       "TEC                                     1077\n",
       "TR_C_gene                                  6\n",
       "TR_D_gene                                  4\n",
       "TR_J_gene                                 79\n",
       "TR_J_pseudogene                            4\n",
       "TR_V_gene                                106\n",
       "TR_V_pseudogene                           33\n",
       "lncRNA                                 47675\n",
       "miRNA                                   1879\n",
       "misc_RNA                                2212\n",
       "polymorphic_pseudogene                   124\n",
       "processed_pseudogene                   10165\n",
       "protein_coding                        162022\n",
       "pseudogene                                15\n",
       "rRNA                                      47\n",
       "rRNA_pseudogene                          497\n",
       "ribozyme                                   8\n",
       "sRNA                                       5\n",
       "scRNA                                      1\n",
       "scaRNA                                    49\n",
       "snRNA                                   1901\n",
       "snoRNA                                   943\n",
       "transcribed_processed_pseudogene        1140\n",
       "transcribed_unitary_pseudogene           742\n",
       "transcribed_unprocessed_pseudogene      3103\n",
       "translated_processed_pseudogene            3\n",
       "translated_unprocessed_pseudogene          1\n",
       "unitary_pseudogene                       103\n",
       "unprocessed_pseudogene                  2620\n",
       "vault_RNA                                  1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_new.groupby(\"gene_type\").size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "05aed19e-f7be-4ca1-a68e-f882d087d1ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_new1= df1[df1['havana_gene'].notnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f80d68b3-58a7-4d01-8150-95faeba82693",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3150419</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7499</td>\n",
       "      <td></td>\n",
       "      <td>ENST00000387460.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3150420</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7499</td>\n",
       "      <td></td>\n",
       "      <td>ENST00000387460.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001544475.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3150421</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7494</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3150422</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7494</td>\n",
       "      <td></td>\n",
       "      <td>ENST00000387461.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3150423</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7494</td>\n",
       "      <td></td>\n",
       "      <td>ENST00000387461.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001544473.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3150424 rows  25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname   source     feature  start    end  score strand frame  \\\n",
       "0          chr1   HAVANA        gene  11869  14409    NaN      +     0   \n",
       "1          chr1   HAVANA  transcript  11869  14409    NaN      +     0   \n",
       "2          chr1   HAVANA        exon  11869  12227    NaN      +     0   \n",
       "3          chr1   HAVANA        exon  12613  12721    NaN      +     0   \n",
       "4          chr1   HAVANA        exon  13221  14409    NaN      +     0   \n",
       "...         ...      ...         ...    ...    ...    ...    ...   ...   \n",
       "3150419    chrM  ENSEMBL  transcript  15888  15953    NaN      +     0   \n",
       "3150420    chrM  ENSEMBL        exon  15888  15953    NaN      +     0   \n",
       "3150421    chrM  ENSEMBL        gene  15956  16023    NaN      -     0   \n",
       "3150422    chrM  ENSEMBL  transcript  15956  16023    NaN      -     0   \n",
       "3150423    chrM  ENSEMBL        exon  15956  16023    NaN      -     0   \n",
       "\n",
       "                   gene_id                           gene_type gene_name  \\\n",
       "0        ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1   \n",
       "1        ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1   \n",
       "2        ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1   \n",
       "3        ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1   \n",
       "4        ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1   \n",
       "...                    ...                                 ...       ...   \n",
       "3150419  ENSG00000210195.2                             Mt_tRNA     MT-TT   \n",
       "3150420  ENSG00000210195.2                             Mt_tRNA     MT-TT   \n",
       "3150421  ENSG00000210196.2                             Mt_tRNA     MT-TP   \n",
       "3150422  ENSG00000210196.2                             Mt_tRNA     MT-TP   \n",
       "3150423  ENSG00000210196.2                             Mt_tRNA     MT-TP   \n",
       "\n",
       "        level     hgnc_id           havana_gene      transcript_id  \\\n",
       "0           2  HGNC:37102  OTTHUMG00000000961.2                      \n",
       "1           2  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "2           2  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "3           2  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "4           2  HGNC:37102  OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "...       ...         ...                   ...                ...   \n",
       "3150419     3   HGNC:7499                        ENST00000387460.2   \n",
       "3150420     3   HGNC:7499                        ENST00000387460.2   \n",
       "3150421     3   HGNC:7494                                            \n",
       "3150422     3   HGNC:7494                        ENST00000387461.2   \n",
       "3150423     3   HGNC:7494                        ENST00000387461.2   \n",
       "\n",
       "              transcript_type transcript_name transcript_support_level  \\\n",
       "0                                                                        \n",
       "1        processed_transcript     DDX11L1-202                        1   \n",
       "2        processed_transcript     DDX11L1-202                        1   \n",
       "3        processed_transcript     DDX11L1-202                        1   \n",
       "4        processed_transcript     DDX11L1-202                        1   \n",
       "...                       ...             ...                      ...   \n",
       "3150419               Mt_tRNA       MT-TT-201                       NA   \n",
       "3150420               Mt_tRNA       MT-TT-201                       NA   \n",
       "3150421                                                                  \n",
       "3150422               Mt_tRNA       MT-TP-201                       NA   \n",
       "3150423               Mt_tRNA       MT-TP-201                       NA   \n",
       "\n",
       "                             tag     havana_transcript exon_number  \\\n",
       "0                                                                    \n",
       "1                          basic  OTTHUMT00000362751.1               \n",
       "2                          basic  OTTHUMT00000362751.1           1   \n",
       "3                          basic  OTTHUMT00000362751.1           2   \n",
       "4                          basic  OTTHUMT00000362751.1           3   \n",
       "...                          ...                   ...         ...   \n",
       "3150419  basic,Ensembl_canonical                                     \n",
       "3150420  basic,Ensembl_canonical                                 1   \n",
       "3150421                                                              \n",
       "3150422  basic,Ensembl_canonical                                     \n",
       "3150423  basic,Ensembl_canonical                                 1   \n",
       "\n",
       "                   exon_id ont protein_id ccdsid  \n",
       "0                                                 \n",
       "1                                                 \n",
       "2        ENSE00002234944.1                        \n",
       "3        ENSE00003582793.1                        \n",
       "4        ENSE00002312635.1                        \n",
       "...                    ...  ..        ...    ...  \n",
       "3150419                                           \n",
       "3150420  ENSE00001544475.2                        \n",
       "3150421                                           \n",
       "3150422                                           \n",
       "3150423  ENSE00001544473.2                        \n",
       "\n",
       "[3150424 rows x 25 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_new1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f0edf05-6c32-4043-9720-63bc3ea7ea31",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
