{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3d7a618f-c3c7-450b-95bb-c05198d173ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import pandas as pd\n",
    "import pybedtools\n",
    "from io import StringIO\n",
    "import pybedtools \n",
    "pybedtools.helpers.set_tempdir(\"/data/projects/temp\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2ed55fab-a81a-48dc-ad14-fe7bba09685b",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_file_path= \"/data/projects/Resources/Gencode_genome_annotation/GRCh38.primary_assembly.genome.fa\"\n",
    "reference_fasta = pysam.FastaFile(reference_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c211af81-f021-44bf-81c3-3ccd14cde6d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::idx_find_and_load] Could not retrieve index file for '/data/private/pdutta/Collab_data/Oliver_data/New_data/background_boundaries.SpliceAI.vcf.gz'\n"
     ]
    }
   ],
   "source": [
    "vcf_file_path = \"/data/private/pdutta/Collab_data/Oliver_data/New_data/background_boundaries.SpliceAI.vcf.gz\"\n",
    "vcf_file = pysam.VariantFile(vcf_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9098ac97-5872-43fa-ac57-76923bcdc7cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::idx_find_and_load] Could not retrieve index file for '/data/private/pdutta/Collab_data/Oliver_data/New_data/sigma_boundaries.SpliceAI.vcf.gz'\n"
     ]
    }
   ],
   "source": [
    "Splice_VCF = \"/data/private/pdutta/Collab_data/Oliver_data/New_data/sigma_boundaries.SpliceAI.vcf.gz\"\n",
    "splice_vcf_file = pysam.VariantFile(Splice_VCF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d37c594d-12f1-4cda-ab6a-fc987d2b4744",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>strand</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>acceptor_40BP</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>ENSE00003692979.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>GTGAAATGATCTCTACAAATGCTGTGTTTTGCTGTTTCAGGTGCTT...</td>\n",
       "      <td>24401417</td>\n",
       "      <td>24401496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>ENSE00003638911.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>TTGAAGTATCTCGATGACACCACCTTCTCTTCCCTTTCAGAATGAT...</td>\n",
       "      <td>24383964</td>\n",
       "      <td>24384043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>ENSE00003688858.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>TTATGTGACAATGCTGATTCCTTTGGTTGTCTCCTTACAGTGCGCC...</td>\n",
       "      <td>24379784</td>\n",
       "      <td>24379863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>ENSE00003631141.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>GAAAGGTACTGAGTACATTGGGTGTATTTTCTCCCCCTAGGCCTCT...</td>\n",
       "      <td>24373771</td>\n",
       "      <td>24373850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000003583.12</td>\n",
       "      <td>ENSE00003516207.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>AAGAGTTCCTTATTCTGTCCTAAAATTTCTTTCTCCTTAGGGCATT...</td>\n",
       "      <td>24369800</td>\n",
       "      <td>24369879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255761</th>\n",
       "      <td>ENST00000684226.1</td>\n",
       "      <td>ENSE00003767626.1</td>\n",
       "      <td>non_stop_decay</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>ACTTTGGGTTACTTTATGACTGTTGCTTTTTTCCTTCTAGCCTGCC...</td>\n",
       "      <td>13297809</td>\n",
       "      <td>13297888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255762</th>\n",
       "      <td>ENST00000684226.1</td>\n",
       "      <td>ENSE00003761778.1</td>\n",
       "      <td>non_stop_decay</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>TTCAGTAGTTACTCTCACAAATAGTTTTTTCCCACAATAGGTATTG...</td>\n",
       "      <td>13260365</td>\n",
       "      <td>13260444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255763</th>\n",
       "      <td>ENST00000684226.1</td>\n",
       "      <td>ENSE00001713248.1</td>\n",
       "      <td>non_stop_decay</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>ACTGTCCAATACTTTTTGTTTATTTATTTATTTAATATAGGTGGAG...</td>\n",
       "      <td>13251148</td>\n",
       "      <td>13251227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255764</th>\n",
       "      <td>ENST00000684226.1</td>\n",
       "      <td>ENSE00003917271.1</td>\n",
       "      <td>non_stop_decay</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>TCCTGCAAACCCCAATGTGAACTTGTGGTGCCTTTTCCAGGCCTCC...</td>\n",
       "      <td>13234787</td>\n",
       "      <td>13234866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255765</th>\n",
       "      <td>ENST00000684326.1</td>\n",
       "      <td>ENSE00003920852.1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>TCCTGCAAACCCCAATGTGAACTTGTGGTGCCTTTTCCAGGCCTCC...</td>\n",
       "      <td>13234787</td>\n",
       "      <td>13234866</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1255766 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              transcript_id            exon_id       transcript_type strand  \\\n",
       "0        ENST00000003583.12  ENSE00003692979.1        protein_coding      -   \n",
       "1        ENST00000003583.12  ENSE00003638911.1        protein_coding      -   \n",
       "2        ENST00000003583.12  ENSE00003688858.1        protein_coding      -   \n",
       "3        ENST00000003583.12  ENSE00003631141.1        protein_coding      -   \n",
       "4        ENST00000003583.12  ENSE00003516207.1        protein_coding      -   \n",
       "...                     ...                ...                   ...    ...   \n",
       "1255761   ENST00000684226.1  ENSE00003767626.1        non_stop_decay      -   \n",
       "1255762   ENST00000684226.1  ENSE00003761778.1        non_stop_decay      -   \n",
       "1255763   ENST00000684226.1  ENSE00001713248.1        non_stop_decay      -   \n",
       "1255764   ENST00000684226.1  ENSE00003917271.1        non_stop_decay      -   \n",
       "1255765   ENST00000684326.1  ENSE00003920852.1  processed_transcript      -   \n",
       "\n",
       "        chr_name                                      acceptor_40BP     start  \\\n",
       "0           chr1  GTGAAATGATCTCTACAAATGCTGTGTTTTGCTGTTTCAGGTGCTT...  24401417   \n",
       "1           chr1  TTGAAGTATCTCGATGACACCACCTTCTCTTCCCTTTCAGAATGAT...  24383964   \n",
       "2           chr1  TTATGTGACAATGCTGATTCCTTTGGTTGTCTCCTTACAGTGCGCC...  24379784   \n",
       "3           chr1  GAAAGGTACTGAGTACATTGGGTGTATTTTCTCCCCCTAGGCCTCT...  24373771   \n",
       "4           chr1  AAGAGTTCCTTATTCTGTCCTAAAATTTCTTTCTCCTTAGGGCATT...  24369800   \n",
       "...          ...                                                ...       ...   \n",
       "1255761     chrY  ACTTTGGGTTACTTTATGACTGTTGCTTTTTTCCTTCTAGCCTGCC...  13297809   \n",
       "1255762     chrY  TTCAGTAGTTACTCTCACAAATAGTTTTTTCCCACAATAGGTATTG...  13260365   \n",
       "1255763     chrY  ACTGTCCAATACTTTTTGTTTATTTATTTATTTAATATAGGTGGAG...  13251148   \n",
       "1255764     chrY  TCCTGCAAACCCCAATGTGAACTTGTGGTGCCTTTTCCAGGCCTCC...  13234787   \n",
       "1255765     chrY  TCCTGCAAACCCCAATGTGAACTTGTGGTGCCTTTTCCAGGCCTCC...  13234787   \n",
       "\n",
       "              end  \n",
       "0        24401496  \n",
       "1        24384043  \n",
       "2        24379863  \n",
       "3        24373850  \n",
       "4        24369879  \n",
       "...           ...  \n",
       "1255761  13297888  \n",
       "1255762  13260444  \n",
       "1255763  13251227  \n",
       "1255764  13234866  \n",
       "1255765  13234866  \n",
       "\n",
       "[1255766 rows x 8 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_acceptor = pd.read_csv(\"/data/private/pdutta/DNABERT_data/Splice_sites/Acceptor_raw_files/Acceptor_80BP_bed.tsv\", sep=\"\\t\")\n",
    "df_acceptor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "36214060-ae21-48a0-a0de-346596e3b172",
   "metadata": {},
   "outputs": [],
   "source": [
    "str_io_acceptor = StringIO()\n",
    "df_acceptor[['chr_name','start','end','transcript_id','strand']].to_csv(str_io_acceptor, sep='\\t', index=False, header=False)\n",
    "str_io_acceptor.seek(0)\n",
    "acceptor_bed = pybedtools.BedTool(str_io_acceptor.read(), from_string=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "eb5f0383-1f3b-4449-b671-56fa427edbe4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Conversion to BED format complete. File saved as: /data/private/pdutta/Collab_data/Oliver_data/New_data/background_boundaries.SpliceAI.bed\n"
     ]
    }
   ],
   "source": [
    "# Create a BED file to write to\n",
    "bed_file_path = \"/data/private/pdutta/Collab_data/Oliver_data/New_data/background_boundaries.SpliceAI.bed\"\n",
    "bed_entries = []\n",
    "with open(bed_file_path, \"w\") as bed_file:\n",
    "    # Iterate over each record in the VCF file\n",
    "    for rec in vcf_file.fetch():\n",
    "        # Get the start position (0-based for BED)\n",
    "        start = rec.pos\n",
    "        # For BED, the end position is the start + length of the reference allele\n",
    "        end = start + len(rec.ref) -1\n",
    "        # Create the BED format line (chromosome, start, end, name/ref, score, strand are optional)\n",
    "        bed_line = f\"chr{rec.chrom}\\t{start}\\t{end}\\t{rec.ref}\\t{','.join(rec.alts)}\\n\"\n",
    "        # Write the line to the BED file\n",
    "        bed_file.write(bed_line)\n",
    "\n",
    "print(f\"Conversion to BED format complete. File saved as: {bed_file_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2f10ae4a-2824-4e45-a37e-a41581516d54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(output.bed)>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# List to store BED entries\n",
    "bed_entries = []\n",
    "\n",
    "# Iterate over each record in the VCF file and create BED entries\n",
    "for rec in vcf_file.fetch():\n",
    "    # Get the start position (0-based for BED)\n",
    "    start = rec.pos\n",
    "    # For BED, the end position is the start + length of the reference allele\n",
    "    end = start + len(rec.ref)-1\n",
    "    # Create the BED format entry\n",
    "    bed_entry = pybedtools.create_interval_from_list([\"chr\"+rec.chrom, start, end, rec.ref, ','.join(rec.alts)])\n",
    "    # Append the entry to the list\n",
    "    bed_entries.append(bed_entry)\n",
    "\n",
    "# Create a BedTool object from the list of entries\n",
    "vcf_bed = pybedtools.BedTool(bed_entries)\n",
    "vcf_bed.saveas('output.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d81e45f2-2162-4f43-a451-8fd698d31da5",
   "metadata": {},
   "outputs": [],
   "source": [
    "intersect_bed = acceptor_bed.intersect(vcf_bed, wa=True, wb=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "4cc93f8e-ce3f-42c9-948d-39985e794af4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t114780176\t114780255\tENST00000060969.6\t-\tchr1\t114780192\t114780192\tC\tT\n",
      "\n",
      "chr1\t114779245\t114779324\tENST00000060969.6\t-\tchr1\t114779245\t114779245\tT\tC\n",
      "\n",
      "chr1\t114776420\t114776499\tENST00000060969.6\t-\tchr1\t114776425\t114776425\tA\tG\n",
      "\n",
      "chr1\t114776420\t114776499\tENST00000060969.6\t-\tchr1\t114776441\t114776441\tC\tG\n",
      "\n",
      "chr1\t170719685\t170719764\tENST00000239461.11\t+\tchr1\t170719745\t170719745\tA\tG\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for i, interval in enumerate(intersect_bed):\n",
    "    if i < 5:\n",
    "        print(interval)\n",
    "    else:\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f84f0ec1-2f38-4cd5-b3a2-fb465963f353",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_name</th>\n",
       "      <th>Acceptor_start</th>\n",
       "      <th>Acceptor_end</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_start</th>\n",
       "      <th>Variant_end</th>\n",
       "      <th>Reference_Nucleotide</th>\n",
       "      <th>Alternative_Nucleotides</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>114780176</td>\n",
       "      <td>114780255</td>\n",
       "      <td>ENST00000060969.6</td>\n",
       "      <td>-</td>\n",
       "      <td>114780192</td>\n",
       "      <td>114780192</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>114779245</td>\n",
       "      <td>114779324</td>\n",
       "      <td>ENST00000060969.6</td>\n",
       "      <td>-</td>\n",
       "      <td>114779245</td>\n",
       "      <td>114779245</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>114776420</td>\n",
       "      <td>114776499</td>\n",
       "      <td>ENST00000060969.6</td>\n",
       "      <td>-</td>\n",
       "      <td>114776425</td>\n",
       "      <td>114776425</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>114776420</td>\n",
       "      <td>114776499</td>\n",
       "      <td>ENST00000060969.6</td>\n",
       "      <td>-</td>\n",
       "      <td>114776441</td>\n",
       "      <td>114776441</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>170719685</td>\n",
       "      <td>170719764</td>\n",
       "      <td>ENST00000239461.11</td>\n",
       "      <td>+</td>\n",
       "      <td>170719745</td>\n",
       "      <td>170719745</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17975</th>\n",
       "      <td>chr22</td>\n",
       "      <td>22550295</td>\n",
       "      <td>22550374</td>\n",
       "      <td>ENST00000543184.5</td>\n",
       "      <td>-</td>\n",
       "      <td>22550312</td>\n",
       "      <td>22550312</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17976</th>\n",
       "      <td>chr22</td>\n",
       "      <td>32538071</td>\n",
       "      <td>32538150</td>\n",
       "      <td>ENST00000619146.1</td>\n",
       "      <td>-</td>\n",
       "      <td>32538072</td>\n",
       "      <td>32538072</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17977</th>\n",
       "      <td>chr22</td>\n",
       "      <td>32538071</td>\n",
       "      <td>32538150</td>\n",
       "      <td>ENST00000619146.1</td>\n",
       "      <td>-</td>\n",
       "      <td>32538085</td>\n",
       "      <td>32538085</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17978</th>\n",
       "      <td>chr22</td>\n",
       "      <td>32528969</td>\n",
       "      <td>32529048</td>\n",
       "      <td>ENST00000619146.1</td>\n",
       "      <td>-</td>\n",
       "      <td>32528980</td>\n",
       "      <td>32528980</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17979</th>\n",
       "      <td>chr22</td>\n",
       "      <td>32528969</td>\n",
       "      <td>32529048</td>\n",
       "      <td>ENST00000619146.1</td>\n",
       "      <td>-</td>\n",
       "      <td>32528981</td>\n",
       "      <td>32528981</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17980 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr_name  Acceptor_start  Acceptor_end       Transcript_id Strand  \\\n",
       "0         chr1       114780176     114780255   ENST00000060969.6      -   \n",
       "1         chr1       114779245     114779324   ENST00000060969.6      -   \n",
       "2         chr1       114776420     114776499   ENST00000060969.6      -   \n",
       "3         chr1       114776420     114776499   ENST00000060969.6      -   \n",
       "4         chr1       170719685     170719764  ENST00000239461.11      +   \n",
       "...        ...             ...           ...                 ...    ...   \n",
       "17975    chr22        22550295      22550374   ENST00000543184.5      -   \n",
       "17976    chr22        32538071      32538150   ENST00000619146.1      -   \n",
       "17977    chr22        32538071      32538150   ENST00000619146.1      -   \n",
       "17978    chr22        32528969      32529048   ENST00000619146.1      -   \n",
       "17979    chr22        32528969      32529048   ENST00000619146.1      -   \n",
       "\n",
       "       Variant_start  Variant_end Reference_Nucleotide Alternative_Nucleotides  \n",
       "0          114780192    114780192                    C                       T  \n",
       "1          114779245    114779245                    T                       C  \n",
       "2          114776425    114776425                    A                       G  \n",
       "3          114776441    114776441                    C                       G  \n",
       "4          170719745    170719745                    A                       G  \n",
       "...              ...          ...                  ...                     ...  \n",
       "17975       22550312     22550312                    C                       T  \n",
       "17976       32538072     32538072                    C                       G  \n",
       "17977       32538085     32538085                    T                       C  \n",
       "17978       32528980     32528980                    C                       A  \n",
       "17979       32528981     32528981                    C                       G  \n",
       "\n",
       "[17980 rows x 9 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersect_df = intersect_bed.to_dataframe()\n",
    "intersect_df=intersect_df.drop(columns=['strand'])\n",
    "new_columns = ['chr_name', 'Acceptor_start', 'Acceptor_end', 'Transcript_id', 'Strand', 'Variant_start', 'Variant_end',  'Reference_Nucleotide',  'Alternative_Nucleotides']\n",
    "intersect_df.columns = new_columns\n",
    "intersect_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bcef468-f3dc-4eb8-b731-eb01a57e9443",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
