{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "13e86a2e-92b3-4644-a96d-547bf513e828",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, glob\n",
    "import pandas as pd\n",
    "import pysam\n",
    "import numpy as np\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c2380204-9847-4c6f-b152-114f02e0139d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_type = \"Brain\"\n",
    "intersected_base_path = \"/data/private/pdutta_new/GDC_cancer_wise/{}/Generated_files/Intersected_files/100bp_TFBS/TFBS_test\".format(cancer_type)\n",
    "reference_genome_path=\"/data/projects/Resources/Gencode_genome_annotation/GRCh38.primary_assembly.genome.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5a4aa801-312d-4380-adb6-0c0ebcd9ba78",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_path = \"/data/projects/GDC_Cancer_Wise/New_data/{}/Generated_files/DNABERT_Data/Somatic/TFBS_300bp_test\".format(cancer_type)\n",
    "if not os.path.exists(output_path):\n",
    "    os.makedirs(output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1e9c0e52-cf4d-48b1-9173-cd32e57d8d95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>H2AFZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>H2AK5ac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>H2BK12ac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>H2BK5ac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>H3F3A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>H3K14ac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>H3K27ac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>H3K27me3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>H3K36me3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>H3K4me1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>H3K4me2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>H3K79me1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>H3K79me2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>H3K9ac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>H3K9me1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>H3K9me3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>H4K20me1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>H4K5ac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>H4K8ac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>IKZF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>RAD21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>RBM22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>RUNX3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>TRIM22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ZBTB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ZNF143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NFE2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>PRDM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SCRT2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ZFHX2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ZNF121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ZNF366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>ZNF770</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           0\n",
       "0        FOS\n",
       "1      H2AFZ\n",
       "2    H2AK5ac\n",
       "3   H2BK12ac\n",
       "4    H2BK5ac\n",
       "5      H3F3A\n",
       "6    H3K14ac\n",
       "7    H3K27ac\n",
       "8   H3K27me3\n",
       "9   H3K36me3\n",
       "10   H3K4me1\n",
       "11   H3K4me2\n",
       "12  H3K79me1\n",
       "13  H3K79me2\n",
       "14    H3K9ac\n",
       "15   H3K9me1\n",
       "16   H3K9me3\n",
       "17  H4K20me1\n",
       "18    H4K5ac\n",
       "19    H4K8ac\n",
       "20     IKZF1\n",
       "21     RAD21\n",
       "22     RBM22\n",
       "23     RUNX3\n",
       "24    TRIM22\n",
       "25    ZBTB33\n",
       "26    ZNF143\n",
       "27      NFE2\n",
       "28     PRDM1\n",
       "29     SCRT2\n",
       "30     ZFHX2\n",
       "31    ZNF121\n",
       "32    ZNF366\n",
       "33    ZNF770"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_models_df= pd.read_csv(\"/home/campus.stonybrook.edu/pdutta/Github/Postdoc/DNABERT_data_processing/TFBS/important_tfbs.txt\", sep=\"\\t\", header=None)\n",
    "top_models_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "57c9bf9b-a33c-4650-a467-a83dbc19b7e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CBFA2T3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>POLR2A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         0\n",
       "0  CBFA2T3\n",
       "1   POLR2A"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_models_df = pd.DataFrame({0: ['CBFA2T3', 'POLR2A']})\n",
    "\n",
    "top_models_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c05f597d-17d5-4ea2-9dae-2903b1c0fe98",
   "metadata": {},
   "outputs": [],
   "source": [
    "def seq2kmer(seq):\n",
    "    \"\"\"\n",
    "    Convert original sequence to kmers\n",
    "    \n",
    "    Arguments:\n",
    "    seq -- str, original sequence.\n",
    "    k -- int, kmer of length k specified.\n",
    "    \n",
    "    Returns:\n",
    "    kmers -- str, kmers separated by space\n",
    "    \"\"\"\n",
    "    k=6\n",
    "    kmer = [seq[x:x+k] for x in range(len(seq)+1-k)]\n",
    "    kmers = \" \".join(kmer)\n",
    "    return kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2aba05c4-0239-41fb-a73f-9da05ade230f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def apply_mutations_absolute(ref_seq, chip_start, starts, ends, refs, alts):\n",
    "    mutations = sorted(zip(starts, ends, refs, alts), key=lambda x: x[0])  # Sort by start position\n",
    "    seq_list = list(ref_seq)\n",
    "    offset = chip_start  # This is the absolute starting position of the sequence on the chromosome\n",
    "    #print(mutations, seq_list, offset)\n",
    "    #input()\n",
    "\n",
    "    # Apply each mutation\n",
    "    for start, end, ref, alt in mutations:\n",
    "        rel_pos = start - offset  # Calculate the relative position in the sequence\n",
    "        #print(rel_pos)\n",
    "        if ''.join(seq_list[rel_pos:rel_pos+len(ref)]) == ref:  # Check if the reference matches\n",
    "            seq_list[rel_pos:rel_pos+len(ref)] = list(alt)  # Replace with the alternative\n",
    "        #else:\n",
    "            # If the reference doesn't match, log an error or handle it accordingly\n",
    "            #print(f\"Expected ref '{ref}' at position {rel_pos}, but found '{''.join(seq_list[rel_pos:rel_pos+len(ref)])}'\")\n",
    "    #print(''.join(seq_list))\n",
    "    #input()\n",
    "\n",
    "    return ''.join(seq_list)  # Convert list back to string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a30b3260-fbcf-4037-9916-961c31c8c4f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sequences(df, reference_fasta):\n",
    "    data = []\n",
    "    for idx, row in df.iterrows():\n",
    "        #print(row)\n",
    "        chrom = row[0]\n",
    "        ref_start = row[1]\n",
    "        ref_end = row[2]\n",
    "        variant_start = row['START_POS']\n",
    "        variant_end = row['END_POS']\n",
    "        ref_nucleotide = row['REF']\n",
    "        alt = row[\"ALT\"]\n",
    "        \n",
    "        # Adjust for 0-based indexing in python\n",
    "        variant_pos_start = variant_start - ref_start\n",
    "        variant_pos_end = variant_end - ref_start\n",
    "        #print(ref_nucleotide , alt, variant_pos_start, variant_pos_end)\n",
    "        \n",
    "        \n",
    "        # Get reference sequence\n",
    "        #print(chrom)\n",
    "        ref_seq = reference_fasta.fetch(chrom, ref_start, ref_end)\n",
    "        #print(ref_seq)\n",
    "        \n",
    "        # Handle insertion and deletion to get the correct alt sequence\n",
    "        # Identify if the variant is an insertion or deletion\n",
    "        if len(ref_nucleotide) < len(alt):  # Insertion\n",
    "            delete_size =  len(alt) - len(ref_nucleotide)\n",
    "            #print(variant_pos_start, variant_pos_end ,delete_size)\n",
    "            alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:len(ref_seq) - delete_size]\n",
    "\n",
    "        elif len(ref_nucleotide) > len(alt):  # Deletion\n",
    "            insert_size = len(ref_nucleotide) - len(alt)\n",
    "            #print(insert_size)\n",
    "            extra_bases = reference_fasta.fetch(chrom, ref_end, ref_end + insert_size)\n",
    "            #print(extra_bases)\n",
    "            alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:] + extra_bases\n",
    "\n",
    "        else:  # SNV\n",
    "            alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:]\n",
    "        # print(ref_seq)\n",
    "        # print(alt_seq)\n",
    "        # input()\n",
    "\n",
    "\n",
    "        data.append({\n",
    "            'chr': chrom,\n",
    "            #'strand': row['strand'],\n",
    "            #'Transcript_ID': row['transcript_id'],\n",
    "            'Chip_Seq_start': row[1],\n",
    "            'Chip_Seq_end': row[2],\n",
    "            'varinat_start': variant_start,\n",
    "            'variant_end': variant_end,\n",
    "            'ref_neucleotide': ref_nucleotide,\n",
    "            'alternative_neucleotide': alt,\n",
    "            'reference_seq': ref_seq,\n",
    "            'alt_seq': alt_seq\n",
    "        })\n",
    "            \n",
    "    \n",
    "    # Convert the list of dictionaries to a DataFrame\n",
    "    new_df = pd.DataFrame(data)\n",
    "    # print(new_df.shape)\n",
    "    # print(new_df)\n",
    "    new_df = new_df.drop_duplicates().reset_index()\n",
    "    #print(new_df.shape)\n",
    "    #print(\"&*\")\n",
    "    data = []\n",
    "    merged_list = list(zip(new_df['reference_seq'], new_df['alt_seq']))\n",
    "    merged_list = [item.upper() for tup in merged_list for item in tup]\n",
    "    #print(merged_list)\n",
    "    kmer_lst = list(map(seq2kmer, merged_list))\n",
    "    df_kmer = pd.DataFrame(kmer_lst, columns=['Sequence'])\n",
    "    df_kmer['Label'] = np.random.choice([0, 1], size=len(df_kmer))\n",
    "    #print(df_kmer.shape)\n",
    "    \n",
    "    \n",
    "    grouped_df = new_df.groupby(['chr', 'Chip_Seq_start', 'Chip_Seq_end']).agg({\n",
    "                                'varinat_start': lambda x: list(x),  # Get unique starts\n",
    "                                'variant_end': lambda x: list(x),    # Get unique ends\n",
    "                                'ref_neucleotide': lambda x: list(x),  # Unique reference nucleotides\n",
    "                                'alternative_neucleotide': lambda x: list(x),  # Unique alternative nucleotides\n",
    "                                'reference_seq': 'first',  # Assuming you might want just the first sequence\n",
    "                                'alt_seq':  lambda x: list(set(x))  # Assuming you might want all the alt sequence\n",
    "                            }).reset_index()\n",
    "    grouped_df['mutated_sequence'] = grouped_df.apply(lambda row: apply_mutations_absolute(\n",
    "                row['reference_seq'], row['Chip_Seq_start'], row['varinat_start'],\n",
    "                row['variant_end'], row['ref_neucleotide'], row['alternative_neucleotide']), axis=1)\n",
    "    \n",
    "    \n",
    "    merged_list_regionwise = list(zip(grouped_df['reference_seq'], grouped_df['mutated_sequence']))\n",
    "    merged_list_regionwise = [item.upper() for tup in merged_list_regionwise for item in tup]\n",
    "    regionwise_kmer_list = list(map(seq2kmer, merged_list_regionwise))\n",
    "    df_kmer_region = pd.DataFrame(regionwise_kmer_list, columns=['Sequence'])\n",
    "    df_kmer_region['Label'] = np.random.choice([0, 1], size=len(df_kmer_region))\n",
    "    return new_df, df_kmer, grouped_df, df_kmer_region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ef4c248a-8b67-4dd9-8992-23b3d81450f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#####The reference file is loaded######\n"
     ]
    }
   ],
   "source": [
    "## Load the reference file\n",
    "reference_fasta = pysam.FastaFile(reference_genome_path)\n",
    "print(\"#####The reference file is loaded######\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "933aa443-d84e-48a9-b824-e5df3470e39b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/private/pdutta_new/GDC_cancer_wise/Brain/Generated_files/Intersected_files/100bp_TFBS/TFBS_test/CBFA2T3/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for CBFA2T3***\n",
      "/data/private/pdutta_new/GDC_cancer_wise/Brain/Generated_files/Intersected_files/100bp_TFBS/TFBS_test/POLR2A/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for POLR2A***\n",
      "Missing files recorded at /data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/DNABERT_Data/Somatic/TFBS_300bp_test/missing_files.txt\n"
     ]
    }
   ],
   "source": [
    "missing_files = []\n",
    "\n",
    "\n",
    "\n",
    "for index, row in top_models_df.iterrows():\n",
    "    intersected_data = f\"{intersected_base_path}/{row[0]}/intersected_vcf_data.pkl\"\n",
    "    print(intersected_data)\n",
    "    \n",
    "    try:\n",
    "        with open(intersected_data, \"rb\") as file:\n",
    "            loaded_dictionary = pickle.load(file)\n",
    "        print(f\"****All the VCF files are loaded for {row[0]}***\") \n",
    "\n",
    "        dnabert_raw_data = {}\n",
    "        dnabert_raw_data_regionwise = {}\n",
    "        for key, value in loaded_dictionary.items():\n",
    "            new_df, df_kmer, grouped_df, df_kmer_region = get_sequences(value, reference_fasta)\n",
    "            dnabert_raw_data[key] = new_df\n",
    "            dnabert_raw_data_regionwise[key] = grouped_df\n",
    "            # display(df_kmer)\n",
    "            # display(new_df)\n",
    "            \n",
    "            out_folder_path = f\"{output_path}/{row[0]}/Patient_wise/variant_wise/{key}\"\n",
    "            if not os.path.exists(out_folder_path):\n",
    "                os.makedirs(out_folder_path)\n",
    "            df_kmer.to_csv(out_folder_path + \"/dev.tsv\", sep=\"\\t\", index=False)\n",
    "\n",
    "            region_out_folder_path = f\"{output_path}/{row[0]}/Patient_wise/region_wise/{key}\"\n",
    "            if not os.path.exists(region_out_folder_path):\n",
    "                os.makedirs(region_out_folder_path)\n",
    "            df_kmer_region.to_csv(region_out_folder_path + \"/dev.tsv\", sep=\"\\t\", index=False)\n",
    "        \n",
    "        with open(f\"{output_path}/{row[0]}/variantwise_raw_vcf_data.pkl\", \"wb\") as file:\n",
    "            pickle.dump(dnabert_raw_data, file)\n",
    "        with open(f\"{output_path}/{row[0]}/regionwise_raw_vcf_data.pkl\", \"wb\") as file:\n",
    "            pickle.dump(dnabert_raw_data_regionwise, file)\n",
    "\n",
    "    except FileNotFoundError:\n",
    "        print(f\"File not found for {row[0]}, skipping...\")\n",
    "        missing_files.append(row[0])\n",
    "\n",
    "# Optionally, save the list of missing files/tags to a text file\n",
    "missing_files_path = f\"{output_path}/missing_files.txt\"\n",
    "with open(missing_files_path, \"w\") as f:\n",
    "    for tag in missing_files:\n",
    "        f.write(f\"{tag}\\n\")\n",
    "\n",
    "print(f\"Missing files recorded at {missing_files_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3d3978e4-e2c0-4bc1-864d-e76c01f2e8b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(missing_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4d00e115-459f-4e81-ae6a-e0b38ce19e74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>Chip_Seq_start</th>\n",
       "      <th>Chip_Seq_end</th>\n",
       "      <th>varinat_start</th>\n",
       "      <th>variant_end</th>\n",
       "      <th>ref_neucleotide</th>\n",
       "      <th>alternative_neucleotide</th>\n",
       "      <th>reference_seq</th>\n",
       "      <th>alt_seq</th>\n",
       "      <th>mutated_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19190</td>\n",
       "      <td>19291</td>\n",
       "      <td>[19189]</td>\n",
       "      <td>[19191]</td>\n",
       "      <td>[GC]</td>\n",
       "      <td>[G]</td>\n",
       "      <td>CCCCTGTAGCTCCCCTACCTCCAAGAGCCCAGCCCTTGCCCACAGG...</td>\n",
       "      <td>[CCCCTGTAGCTCCCCTACCTCCAAGAGCCCAGCCCTTGCCCACAG...</td>\n",
       "      <td>CCCCTGTAGCTCCCCTACCTCCAAGAGCCCAGCCCTTGCCCACAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>38986</td>\n",
       "      <td>39087</td>\n",
       "      <td>[39023]</td>\n",
       "      <td>[39025]</td>\n",
       "      <td>[TC]</td>\n",
       "      <td>[T]</td>\n",
       "      <td>AGTGCAAAATGAAAGAAGACTGTCAGAGACCCCAAACTCTGCTGTC...</td>\n",
       "      <td>[AGTGCAAAATGAAAGAAGACTGTCAGAGACCCCAAACTTGCTGTC...</td>\n",
       "      <td>AGTGCAAAATGAAAGAAGACTGTCAGAGACCCCAAACTTGCTGTCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>109484</td>\n",
       "      <td>109585</td>\n",
       "      <td>[109574]</td>\n",
       "      <td>[109579]</td>\n",
       "      <td>[CGTGT]</td>\n",
       "      <td>[C]</td>\n",
       "      <td>TACGTAAGAGTGCTTGAGGGCTAATTTTATGAAAGCTTTGGGAAGT...</td>\n",
       "      <td>[TACGTAAGAGTGCTTGAGGGCTAATTTTATGAAAGCTTTGGGAAG...</td>\n",
       "      <td>TACGTAAGAGTGCTTGAGGGCTAATTTTATGAAAGCTTTGGGAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>122513</td>\n",
       "      <td>122614</td>\n",
       "      <td>[122585]</td>\n",
       "      <td>[122588]</td>\n",
       "      <td>[CAT]</td>\n",
       "      <td>[C]</td>\n",
       "      <td>TCATTCTTTTGTTTATATAGTCAATATCTCTATCTCAATTGGATCT...</td>\n",
       "      <td>[TCATTCTTTTGTTTATATAGTCAATATCTCTATCTCAATTGGATC...</td>\n",
       "      <td>TCATTCTTTTGTTTATATAGTCAATATCTCTATCTCAATTGGATCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>189375</td>\n",
       "      <td>189476</td>\n",
       "      <td>[189391]</td>\n",
       "      <td>[189394]</td>\n",
       "      <td>[ACC]</td>\n",
       "      <td>[A]</td>\n",
       "      <td>ACGGCTCACTGAGCAAACCCCGAGTCCCGACCACCGCCTCAGTGTG...</td>\n",
       "      <td>[ACGGCTCACTGAGCAAACCGAGTCCCGACCACCGCCTCAGTGTGG...</td>\n",
       "      <td>ACGGCTCACTGAGCAAACCGAGTCCCGACCACCGCCTCAGTGTGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59387</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56856031</td>\n",
       "      <td>56856132</td>\n",
       "      <td>[56856102, 56856116]</td>\n",
       "      <td>[56856105, 56856117]</td>\n",
       "      <td>[CAA, A]</td>\n",
       "      <td>[C, AG]</td>\n",
       "      <td>CGTCTCTCACATCTGAACTGCATGCTGAGTGGGCAAGTTGGTTGTA...</td>\n",
       "      <td>[CGTCTCTCACATCTGAACTGCATGCTGAGTGGGCAAGTTGGTTGT...</td>\n",
       "      <td>CGTCTCTCACATCTGAACTGCATGCTGAGTGGGCAAGTTGGTTGTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59388</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56862596</td>\n",
       "      <td>56862697</td>\n",
       "      <td>[56862598, 56862620, 56862620]</td>\n",
       "      <td>[56862599, 56862621, 56862624]</td>\n",
       "      <td>[C, C, CTTT]</td>\n",
       "      <td>[CT, CT, C]</td>\n",
       "      <td>GACTTTAAAGCATAGATGAAAAATCTTTTTTTTTTTTTTAGACAAC...</td>\n",
       "      <td>[GACTTTAAAGCATAGATGAAAAATCTTTTTTTTTTTTTTTAGACA...</td>\n",
       "      <td>GACTTTTAAAGCATAGATGAAAAATCTTTTTTTTTTTTTTAGACAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59389</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56865020</td>\n",
       "      <td>56865121</td>\n",
       "      <td>[56865118]</td>\n",
       "      <td>[56865119]</td>\n",
       "      <td>[C]</td>\n",
       "      <td>[CA]</td>\n",
       "      <td>CCCAGTGGAAATTCCAATCAGTTTTGTGAAGGCTGCTAAGTCAACT...</td>\n",
       "      <td>[CCCAGTGGAAATTCCAATCAGTTTTGTGAAGGCTGCTAAGTCAAC...</td>\n",
       "      <td>CCCAGTGGAAATTCCAATCAGTTTTGTGAAGGCTGCTAAGTCAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59390</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56880069</td>\n",
       "      <td>56880170</td>\n",
       "      <td>[56880149]</td>\n",
       "      <td>[56880152]</td>\n",
       "      <td>[CAA]</td>\n",
       "      <td>[C]</td>\n",
       "      <td>TTTAGCAGACACGGGGTTTTACCGTGTTAGCCAGGATGGTCTCGAT...</td>\n",
       "      <td>[TTTAGCAGACACGGGGTTTTACCGTGTTAGCCAGGATGGTCTCGA...</td>\n",
       "      <td>TTTAGCAGACACGGGGTTTTACCGTGTTAGCCAGGATGGTCTCGAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59391</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56883200</td>\n",
       "      <td>56883301</td>\n",
       "      <td>[56883289]</td>\n",
       "      <td>[56883293]</td>\n",
       "      <td>[TGGA]</td>\n",
       "      <td>[T]</td>\n",
       "      <td>AGTTCTGATTATTACTGTTAAGAGAAAAGCACAGAGATAAAATAAC...</td>\n",
       "      <td>[AGTTCTGATTATTACTGTTAAGAGAAAAGCACAGAGATAAAATAA...</td>\n",
       "      <td>AGTTCTGATTATTACTGTTAAGAGAAAAGCACAGAGATAAAATAAC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>59392 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr  Chip_Seq_start  Chip_Seq_end                   varinat_start  \\\n",
       "0      chr1           19190         19291                         [19189]   \n",
       "1      chr1           38986         39087                         [39023]   \n",
       "2      chr1          109484        109585                        [109574]   \n",
       "3      chr1          122513        122614                        [122585]   \n",
       "4      chr1          189375        189476                        [189391]   \n",
       "...     ...             ...           ...                             ...   \n",
       "59387  chrY        56856031      56856132            [56856102, 56856116]   \n",
       "59388  chrY        56862596      56862697  [56862598, 56862620, 56862620]   \n",
       "59389  chrY        56865020      56865121                      [56865118]   \n",
       "59390  chrY        56880069      56880170                      [56880149]   \n",
       "59391  chrY        56883200      56883301                      [56883289]   \n",
       "\n",
       "                          variant_end ref_neucleotide alternative_neucleotide  \\\n",
       "0                             [19191]            [GC]                     [G]   \n",
       "1                             [39025]            [TC]                     [T]   \n",
       "2                            [109579]         [CGTGT]                     [C]   \n",
       "3                            [122588]           [CAT]                     [C]   \n",
       "4                            [189394]           [ACC]                     [A]   \n",
       "...                               ...             ...                     ...   \n",
       "59387            [56856105, 56856117]        [CAA, A]                 [C, AG]   \n",
       "59388  [56862599, 56862621, 56862624]    [C, C, CTTT]             [CT, CT, C]   \n",
       "59389                      [56865119]             [C]                    [CA]   \n",
       "59390                      [56880152]           [CAA]                     [C]   \n",
       "59391                      [56883293]          [TGGA]                     [T]   \n",
       "\n",
       "                                           reference_seq  \\\n",
       "0      CCCCTGTAGCTCCCCTACCTCCAAGAGCCCAGCCCTTGCCCACAGG...   \n",
       "1      AGTGCAAAATGAAAGAAGACTGTCAGAGACCCCAAACTCTGCTGTC...   \n",
       "2      TACGTAAGAGTGCTTGAGGGCTAATTTTATGAAAGCTTTGGGAAGT...   \n",
       "3      TCATTCTTTTGTTTATATAGTCAATATCTCTATCTCAATTGGATCT...   \n",
       "4      ACGGCTCACTGAGCAAACCCCGAGTCCCGACCACCGCCTCAGTGTG...   \n",
       "...                                                  ...   \n",
       "59387  CGTCTCTCACATCTGAACTGCATGCTGAGTGGGCAAGTTGGTTGTA...   \n",
       "59388  GACTTTAAAGCATAGATGAAAAATCTTTTTTTTTTTTTTAGACAAC...   \n",
       "59389  CCCAGTGGAAATTCCAATCAGTTTTGTGAAGGCTGCTAAGTCAACT...   \n",
       "59390  TTTAGCAGACACGGGGTTTTACCGTGTTAGCCAGGATGGTCTCGAT...   \n",
       "59391  AGTTCTGATTATTACTGTTAAGAGAAAAGCACAGAGATAAAATAAC...   \n",
       "\n",
       "                                                 alt_seq  \\\n",
       "0      [CCCCTGTAGCTCCCCTACCTCCAAGAGCCCAGCCCTTGCCCACAG...   \n",
       "1      [AGTGCAAAATGAAAGAAGACTGTCAGAGACCCCAAACTTGCTGTC...   \n",
       "2      [TACGTAAGAGTGCTTGAGGGCTAATTTTATGAAAGCTTTGGGAAG...   \n",
       "3      [TCATTCTTTTGTTTATATAGTCAATATCTCTATCTCAATTGGATC...   \n",
       "4      [ACGGCTCACTGAGCAAACCGAGTCCCGACCACCGCCTCAGTGTGG...   \n",
       "...                                                  ...   \n",
       "59387  [CGTCTCTCACATCTGAACTGCATGCTGAGTGGGCAAGTTGGTTGT...   \n",
       "59388  [GACTTTAAAGCATAGATGAAAAATCTTTTTTTTTTTTTTTAGACA...   \n",
       "59389  [CCCAGTGGAAATTCCAATCAGTTTTGTGAAGGCTGCTAAGTCAAC...   \n",
       "59390  [TTTAGCAGACACGGGGTTTTACCGTGTTAGCCAGGATGGTCTCGA...   \n",
       "59391  [AGTTCTGATTATTACTGTTAAGAGAAAAGCACAGAGATAAAATAA...   \n",
       "\n",
       "                                        mutated_sequence  \n",
       "0      CCCCTGTAGCTCCCCTACCTCCAAGAGCCCAGCCCTTGCCCACAGG...  \n",
       "1      AGTGCAAAATGAAAGAAGACTGTCAGAGACCCCAAACTTGCTGTCA...  \n",
       "2      TACGTAAGAGTGCTTGAGGGCTAATTTTATGAAAGCTTTGGGAAGT...  \n",
       "3      TCATTCTTTTGTTTATATAGTCAATATCTCTATCTCAATTGGATCT...  \n",
       "4      ACGGCTCACTGAGCAAACCGAGTCCCGACCACCGCCTCAGTGTGGT...  \n",
       "...                                                  ...  \n",
       "59387  CGTCTCTCACATCTGAACTGCATGCTGAGTGGGCAAGTTGGTTGTA...  \n",
       "59388  GACTTTTAAAGCATAGATGAAAAATCTTTTTTTTTTTTTTAGACAA...  \n",
       "59389  CCCAGTGGAAATTCCAATCAGTTTTGTGAAGGCTGCTAAGTCAACT...  \n",
       "59390  TTTAGCAGACACGGGGTTTTACCGTGTTAGCCAGGATGGTCTCGAT...  \n",
       "59391  AGTTCTGATTATTACTGTTAAGAGAAAAGCACAGAGATAAAATAAC...  \n",
       "\n",
       "[59392 rows x 10 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "61c5a0ae-26ff-4c9f-a618-971a477205da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chr                                                                     chr1\n",
       "Chip_Seq_start                                                     123086175\n",
       "Chip_Seq_end                                                       123086276\n",
       "varinat_start                              [123086218, 123086246, 123086260]\n",
       "variant_end                                [123086219, 123086247, 123086261]\n",
       "ref_neucleotide                                                    [T, A, T]\n",
       "alternative_neucleotide                                            [A, G, C]\n",
       "reference_seq              ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGTGG...\n",
       "alt_seq                    [ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGTG...\n",
       "mutated_sequence           ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGAGG...\n",
       "Name: 4, dtype: object"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped_df.iloc[4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "3d99bdfb-a1c7-41d2-98c7-71630d9694ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGTGGAGATTTCAGCCGCTTTGAGGTCAATGGTAGAAAAGGAAATAACTTCGTAGAAAAA',\n",
       " 'ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGTGGAGATTTCAGCCGCTTTGAGGTCAATAGTAGAAAAGGAAACAACTTCGTAGAAAAA',\n",
       " 'ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGAGGAGATTTCAGCCGCTTTGAGGTCAATAGTAGAAAAGGAAATAACTTCGTAGAAAAA']"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped_df.iloc[4]['alt_seq']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "eea5e048-ab92-4d7d-b29a-ccd7f8c35559",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGAGGAGATTTCAGCCGCTTTGAGGTCAATGGTAGAAAAGGAAACAACTTCGTAGAAAAA'"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped_df.iloc[4]['mutated_sequence']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "d8b4e67f-5cb2-469d-af11-cccb85374fe0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGTGGAGATTTCAGCCGCTTTGAGGTCAATAGTAGAAAAGGAAATAACTTCGTAGAAAAA'"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped_df.iloc[4]['reference_seq']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0c799547-3f0a-4f64-b6d3-87b373c340c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>chr</th>\n",
       "      <th>Chip_Seq_start</th>\n",
       "      <th>Chip_Seq_end</th>\n",
       "      <th>varinat_start</th>\n",
       "      <th>variant_end</th>\n",
       "      <th>ref_neucleotide</th>\n",
       "      <th>alternative_neucleotide</th>\n",
       "      <th>reference_seq</th>\n",
       "      <th>alt_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>54707858</td>\n",
       "      <td>54707958</td>\n",
       "      <td>54707863</td>\n",
       "      <td>54707864</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>TGCATCCTCAGATGGATCCTCAGATCCTTGCCCTCCCACTGTCACC...</td>\n",
       "      <td>TGCATTCTCAGATGGATCCTCAGATCCTTGCCCTCCCACTGTCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>chr5</td>\n",
       "      <td>415490</td>\n",
       "      <td>415590</td>\n",
       "      <td>415508</td>\n",
       "      <td>415509</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>GGCCGAGTCTCCCTGGTCGGGTGGGAGGCCTAGGGGCCGAGTCTGC...</td>\n",
       "      <td>GGCCGAGTCTCCCTGGTCTGGTGGGAGGCCTAGGGGCCGAGTCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>chr5</td>\n",
       "      <td>415490</td>\n",
       "      <td>415590</td>\n",
       "      <td>415510</td>\n",
       "      <td>415511</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>GGCCGAGTCTCCCTGGTCGGGTGGGAGGCCTAGGGGCCGAGTCTGC...</td>\n",
       "      <td>GGCCGAGTCTCCCTGGTCGGCTGGGAGGCCTAGGGGCCGAGTCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>chr5</td>\n",
       "      <td>415490</td>\n",
       "      <td>415590</td>\n",
       "      <td>415564</td>\n",
       "      <td>415565</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>GGCCGAGTCTCCCTGGTCGGGTGGGAGGCCTAGGGGCCGAGTCTGC...</td>\n",
       "      <td>GGCCGAGTCTCCCTGGTCGGGTGGGAGGCCTAGGGGCCGAGTCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>chr16</td>\n",
       "      <td>89068861</td>\n",
       "      <td>89068961</td>\n",
       "      <td>89068902</td>\n",
       "      <td>89068903</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>TGGTGATGATGATGGTGGTGATGATGATGGTGGTGATGGTGATGAT...</td>\n",
       "      <td>TGGTGATGATGATGGTGGTGATGATGATGGTGGTGATGGTGGTGAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>105</td>\n",
       "      <td>chr19</td>\n",
       "      <td>41789861</td>\n",
       "      <td>41789962</td>\n",
       "      <td>41789887</td>\n",
       "      <td>41789888</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>AGAAAGAAAGAAAGAAAGAAAGAAAAGAAAGAAAGAAAGAAAGAAA...</td>\n",
       "      <td>AGAAAGAAAGAAAGAAAGAAAGAAAAAAAAGAAAGAAAGAAAGAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>106</td>\n",
       "      <td>chr13</td>\n",
       "      <td>57431125</td>\n",
       "      <td>57431226</td>\n",
       "      <td>57431173</td>\n",
       "      <td>57431174</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CTGAAGGAACATACCGTAAGATAAGAAGAATCATATATGTCAGACC...</td>\n",
       "      <td>CTGAAGGAACATACCGTAAGATAAGAAGAATCATATATGTCAGACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>107</td>\n",
       "      <td>chr20</td>\n",
       "      <td>29508463</td>\n",
       "      <td>29508564</td>\n",
       "      <td>29508498</td>\n",
       "      <td>29508499</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>CTAATCTTGAGCTCCTGGCCTCTAGTGATTCTCCTGCCTTGGGCTC...</td>\n",
       "      <td>CTAATCTTGAGCTCCTGGCCTCTAGTGATTCTCCTACCTTGGGCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>108</td>\n",
       "      <td>chr20</td>\n",
       "      <td>29508463</td>\n",
       "      <td>29508564</td>\n",
       "      <td>29508500</td>\n",
       "      <td>29508501</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>CTAATCTTGAGCTCCTGGCCTCTAGTGATTCTCCTGCCTTGGGCTC...</td>\n",
       "      <td>CTAATCTTGAGCTCCTGGCCTCTAGTGATTCTCCTGCGTTGGGCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>109</td>\n",
       "      <td>chr17</td>\n",
       "      <td>21663288</td>\n",
       "      <td>21663389</td>\n",
       "      <td>21663365</td>\n",
       "      <td>21663366</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>GGCGATGGTAGCAGGAGAAGATGCTACATGTTTCCAATAACATGAA...</td>\n",
       "      <td>GGCGATGGTAGCAGGAGAAGATGCTACATGTTTCCAATAACATGAA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>110 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     index    chr  Chip_Seq_start  Chip_Seq_end  varinat_start  variant_end  \\\n",
       "0        0   chr1        54707858      54707958       54707863     54707864   \n",
       "1        1   chr5          415490        415590         415508       415509   \n",
       "2        2   chr5          415490        415590         415510       415511   \n",
       "3        3   chr5          415490        415590         415564       415565   \n",
       "4        4  chr16        89068861      89068961       89068902     89068903   \n",
       "..     ...    ...             ...           ...            ...          ...   \n",
       "105    105  chr19        41789861      41789962       41789887     41789888   \n",
       "106    106  chr13        57431125      57431226       57431173     57431174   \n",
       "107    107  chr20        29508463      29508564       29508498     29508499   \n",
       "108    108  chr20        29508463      29508564       29508500     29508501   \n",
       "109    109  chr17        21663288      21663389       21663365     21663366   \n",
       "\n",
       "    ref_neucleotide alternative_neucleotide  \\\n",
       "0                 C                       T   \n",
       "1                 G                       T   \n",
       "2                 G                       C   \n",
       "3                 A                       G   \n",
       "4                 A                       G   \n",
       "..              ...                     ...   \n",
       "105               G                       A   \n",
       "106               C                       T   \n",
       "107               G                       A   \n",
       "108               C                       G   \n",
       "109               A                       C   \n",
       "\n",
       "                                         reference_seq  \\\n",
       "0    TGCATCCTCAGATGGATCCTCAGATCCTTGCCCTCCCACTGTCACC...   \n",
       "1    GGCCGAGTCTCCCTGGTCGGGTGGGAGGCCTAGGGGCCGAGTCTGC...   \n",
       "2    GGCCGAGTCTCCCTGGTCGGGTGGGAGGCCTAGGGGCCGAGTCTGC...   \n",
       "3    GGCCGAGTCTCCCTGGTCGGGTGGGAGGCCTAGGGGCCGAGTCTGC...   \n",
       "4    TGGTGATGATGATGGTGGTGATGATGATGGTGGTGATGGTGATGAT...   \n",
       "..                                                 ...   \n",
       "105  AGAAAGAAAGAAAGAAAGAAAGAAAAGAAAGAAAGAAAGAAAGAAA...   \n",
       "106  CTGAAGGAACATACCGTAAGATAAGAAGAATCATATATGTCAGACC...   \n",
       "107  CTAATCTTGAGCTCCTGGCCTCTAGTGATTCTCCTGCCTTGGGCTC...   \n",
       "108  CTAATCTTGAGCTCCTGGCCTCTAGTGATTCTCCTGCCTTGGGCTC...   \n",
       "109  GGCGATGGTAGCAGGAGAAGATGCTACATGTTTCCAATAACATGAA...   \n",
       "\n",
       "                                               alt_seq  \n",
       "0    TGCATTCTCAGATGGATCCTCAGATCCTTGCCCTCCCACTGTCACC...  \n",
       "1    GGCCGAGTCTCCCTGGTCTGGTGGGAGGCCTAGGGGCCGAGTCTGC...  \n",
       "2    GGCCGAGTCTCCCTGGTCGGCTGGGAGGCCTAGGGGCCGAGTCTGC...  \n",
       "3    GGCCGAGTCTCCCTGGTCGGGTGGGAGGCCTAGGGGCCGAGTCTGC...  \n",
       "4    TGGTGATGATGATGGTGGTGATGATGATGGTGGTGATGGTGGTGAT...  \n",
       "..                                                 ...  \n",
       "105  AGAAAGAAAGAAAGAAAGAAAGAAAAAAAAGAAAGAAAGAAAGAAA...  \n",
       "106  CTGAAGGAACATACCGTAAGATAAGAAGAATCATATATGTCAGACC...  \n",
       "107  CTAATCTTGAGCTCCTGGCCTCTAGTGATTCTCCTACCTTGGGCTC...  \n",
       "108  CTAATCTTGAGCTCCTGGCCTCTAGTGATTCTCCTGCGTTGGGCTC...  \n",
       "109  GGCGATGGTAGCAGGAGAAGATGCTACATGTTTCCAATAACATGAA...  \n",
       "\n",
       "[110 rows x 10 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6642ec8c-5d3b-4848-8049-3b8d5578c181",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>chr</th>\n",
       "      <th>Chip_Seq_start</th>\n",
       "      <th>Chip_Seq_end</th>\n",
       "      <th>varinat_start</th>\n",
       "      <th>variant_end</th>\n",
       "      <th>ref_neucleotide</th>\n",
       "      <th>alternative_neucleotide</th>\n",
       "      <th>reference_seq</th>\n",
       "      <th>alt_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [index, chr, Chip_Seq_start, Chip_Seq_end, varinat_start, variant_end, ref_neucleotide, alternative_neucleotide, reference_seq, alt_seq]\n",
       "Index: []"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df[new_df['alternative_neucleotide'].apply(len) == 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d435b3e2-d871-4595-b51f-48083bb389f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>chr</th>\n",
       "      <th>Chip_Seq_start</th>\n",
       "      <th>Chip_Seq_end</th>\n",
       "      <th>varinat_start</th>\n",
       "      <th>variant_end</th>\n",
       "      <th>ref_neucleotide</th>\n",
       "      <th>alternative_neucleotide</th>\n",
       "      <th>reference_seq</th>\n",
       "      <th>alt_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>62</td>\n",
       "      <td>chr1</td>\n",
       "      <td>123086175</td>\n",
       "      <td>123086276</td>\n",
       "      <td>123086218</td>\n",
       "      <td>123086219</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGTGG...</td>\n",
       "      <td>ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>63</td>\n",
       "      <td>chr1</td>\n",
       "      <td>123086175</td>\n",
       "      <td>123086276</td>\n",
       "      <td>123086246</td>\n",
       "      <td>123086247</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGTGG...</td>\n",
       "      <td>ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>64</td>\n",
       "      <td>chr1</td>\n",
       "      <td>123086175</td>\n",
       "      <td>123086276</td>\n",
       "      <td>123086260</td>\n",
       "      <td>123086261</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGTGG...</td>\n",
       "      <td>ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGTGG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index   chr  Chip_Seq_start  Chip_Seq_end  varinat_start  variant_end  \\\n",
       "62     62  chr1       123086175     123086276      123086218    123086219   \n",
       "63     63  chr1       123086175     123086276      123086246    123086247   \n",
       "64     64  chr1       123086175     123086276      123086260    123086261   \n",
       "\n",
       "   ref_neucleotide alternative_neucleotide  \\\n",
       "62               T                       A   \n",
       "63               A                       G   \n",
       "64               T                       C   \n",
       "\n",
       "                                        reference_seq  \\\n",
       "62  ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGTGG...   \n",
       "63  ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGTGG...   \n",
       "64  ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGTGG...   \n",
       "\n",
       "                                              alt_seq  \n",
       "62  ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGAGG...  \n",
       "63  ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGTGG...  \n",
       "64  ACAGAGAGCAGACTTGAAACACTCTTTTTGTGGAATTTGCAAGTGG...  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df[\n",
    "    (new_df['chr'] == 'chr1') & \n",
    "    (new_df['Chip_Seq_start'] == 123086175) & \n",
    "    (new_df['Chip_Seq_end'] == 123086276)\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3453192-2afb-43a2-9438-6b37e13840ca",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
