{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b5386f14-1efe-4907-b26d-2f5a0557ed75",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, glob\n",
    "import pandas as pd\n",
    "import pysam\n",
    "import numpy as np\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c2380204-9847-4c6f-b152-114f02e0139d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_type = \"Brain\"\n",
    "non_coding_region=\"acceptor\"\n",
    "intersected_data = \"/home/pdutta/Data/Cancer_wiseGDC/Data/{}/Generated_files/Intersected_Data/intersected_vcf_{}_data.pkl\".format(cancer_type, non_coding_region)\n",
    "reference_path  = \"/home/pdutta/Data/Human_Genome_Data/GRCh38.primary_assembly.genome.fa\"\n",
    "output_path = \"/home/pdutta/Data/Cancer_wiseGDC/Data/{}/Generated_files/DNABERT_Data/{}/\".format(cancer_type, non_coding_region)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c05f597d-17d5-4ea2-9dae-2903b1c0fe98",
   "metadata": {},
   "outputs": [],
   "source": [
    "def seq2kmer(seq):\n",
    "    \"\"\"\n",
    "    Convert original sequence to kmers\n",
    "    \n",
    "    Arguments:\n",
    "    seq -- str, original sequence.\n",
    "    k -- int, kmer of length k specified.\n",
    "    \n",
    "    Returns:\n",
    "    kmers -- str, kmers separated by space\n",
    "    \"\"\"\n",
    "    k=6\n",
    "    kmer = [seq[x:x+k] for x in range(len(seq)+1-k)]\n",
    "    kmers = \" \".join(kmer)\n",
    "    return kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a30b3260-fbcf-4037-9916-961c31c8c4f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sequences(df, reference_fasta):\n",
    "    data = []\n",
    "    for idx, row in df.iterrows():\n",
    "        #print(row)\n",
    "        chrom = row['chr_name']\n",
    "        ref_start = row['start']\n",
    "        ref_end = row['end']\n",
    "        variant_start = row['START_POS']\n",
    "        variant_end = row['END_POS']\n",
    "        ref_nucleotide = row['REF']\n",
    "        alt = row[\"ALT\"]\n",
    "        \n",
    "        # Adjust for 0-based indexing in python\n",
    "        variant_pos_start = variant_start - ref_start\n",
    "        variant_pos_end = variant_end - ref_start\n",
    "        #print(ref_nucleotide , alt, variant_pos_start, variant_pos_end)\n",
    "        \n",
    "        \n",
    "        # Get reference sequence\n",
    "        #print(chrom)\n",
    "        ref_seq = reference_fasta.fetch(chrom, ref_start, ref_end)\n",
    "        #print(ref_seq)\n",
    "        \n",
    "        # Handle insertion and deletion to get the correct alt sequence\n",
    "        # Identify if the variant is an insertion or deletion\n",
    "        if len(ref_nucleotide) < len(alt):  # Insertion\n",
    "            delete_size =  len(alt) - len(ref_nucleotide)\n",
    "            #print(variant_pos_start, variant_pos_end ,delete_size)\n",
    "            alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:len(ref_seq) - delete_size]\n",
    "\n",
    "        elif len(ref_nucleotide) > len(alt):  # Deletion\n",
    "            insert_size = len(ref_nucleotide) - len(alt)\n",
    "            #print(insert_size)\n",
    "            extra_bases = reference_fasta.fetch(chrom, ref_end, ref_end + insert_size)\n",
    "            #print(extra_bases)\n",
    "            alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:] + extra_bases\n",
    "\n",
    "        else:  # SNV\n",
    "            alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:]\n",
    "        # print(ref_seq)\n",
    "        # print(alt_seq)\n",
    "        # input()\n",
    "\n",
    "\n",
    "        data.append({\n",
    "            'chr': chrom,\n",
    "            'strand': row['strand'],\n",
    "            'Transcript_ID': row['transcript_id'],\n",
    "            'Acceptor_start': row['start'],\n",
    "            'Acceptor_end': row['end'],\n",
    "            'varinat_start': variant_start,\n",
    "            'variant_end': variant_end,\n",
    "            'ref_neucleotide': ref_nucleotide,\n",
    "            'alternative_neucleotide': alt,\n",
    "            'reference_seq': ref_seq,\n",
    "            'alt_seq': alt_seq\n",
    "        })\n",
    "            \n",
    "    \n",
    "    # Convert the list of dictionaries to a DataFrame\n",
    "    new_df = pd.DataFrame(data)\n",
    "    # print(new_df.shape)\n",
    "    # print(new_df)\n",
    "    new_df = new_df.drop_duplicates().reset_index()\n",
    "    print(new_df.shape)\n",
    "    print(\"&*\")\n",
    "    data = []\n",
    "    merged_list = list(zip(new_df['reference_seq'], new_df['alt_seq']))\n",
    "    merged_list = [item.upper() for tup in merged_list for item in tup]\n",
    "    #print(merged_list)\n",
    "    kmer_lst = list(map(seq2kmer, merged_list))\n",
    "    df_kmer = pd.DataFrame(kmer_lst, columns=['Sequence'])\n",
    "    df_kmer['Label'] = np.random.choice([0, 1], size=len(df_kmer))\n",
    "    print(df_kmer.shape)\n",
    "    return new_df, df_kmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f6a7eca6-4572-44fe-8751-05376982e8e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "****All the VCF files are loaded***\n"
     ]
    }
   ],
   "source": [
    "## Load the VCF files from pickle\n",
    "with open(intersected_data, \"rb\") as file:\n",
    "    loaded_dictionary = pickle.load(file)\n",
    "print(\"****All the VCF files are loaded***\")   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "33c0d3e4-b033-4755-9cf3-e9258f9af482",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#####The reference file is loaded######\n"
     ]
    }
   ],
   "source": [
    "## Load the reference file\n",
    "reference_fasta = pysam.FastaFile(reference_path)\n",
    "print(\"#####The reference file is loaded######\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "40649ac8-11a3-4475-b391-1e144696b89e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence names in the FASTA file: ['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chrX', 'chrY', 'chrM', 'GL000008.2', 'GL000009.2', 'GL000194.1', 'GL000195.1', 'GL000205.2', 'GL000208.1', 'GL000213.1', 'GL000214.1', 'GL000216.2', 'GL000218.1', 'GL000219.1', 'GL000220.1', 'GL000221.1', 'GL000224.1', 'GL000225.1', 'GL000226.1', 'KI270302.1', 'KI270303.1', 'KI270304.1', 'KI270305.1', 'KI270310.1', 'KI270311.1', 'KI270312.1', 'KI270315.1', 'KI270316.1', 'KI270317.1', 'KI270320.1', 'KI270322.1', 'KI270329.1', 'KI270330.1', 'KI270333.1', 'KI270334.1', 'KI270335.1', 'KI270336.1', 'KI270337.1', 'KI270338.1', 'KI270340.1', 'KI270362.1', 'KI270363.1', 'KI270364.1', 'KI270366.1', 'KI270371.1', 'KI270372.1', 'KI270373.1', 'KI270374.1', 'KI270375.1', 'KI270376.1', 'KI270378.1', 'KI270379.1', 'KI270381.1', 'KI270382.1', 'KI270383.1', 'KI270384.1', 'KI270385.1', 'KI270386.1', 'KI270387.1', 'KI270388.1', 'KI270389.1', 'KI270390.1', 'KI270391.1', 'KI270392.1', 'KI270393.1', 'KI270394.1', 'KI270395.1', 'KI270396.1', 'KI270411.1', 'KI270412.1', 'KI270414.1', 'KI270417.1', 'KI270418.1', 'KI270419.1', 'KI270420.1', 'KI270422.1', 'KI270423.1', 'KI270424.1', 'KI270425.1', 'KI270429.1', 'KI270435.1', 'KI270438.1', 'KI270442.1', 'KI270448.1', 'KI270465.1', 'KI270466.1', 'KI270467.1', 'KI270468.1', 'KI270507.1', 'KI270508.1', 'KI270509.1', 'KI270510.1', 'KI270511.1', 'KI270512.1', 'KI270515.1', 'KI270516.1', 'KI270517.1', 'KI270518.1', 'KI270519.1', 'KI270521.1', 'KI270522.1', 'KI270528.1', 'KI270529.1', 'KI270530.1', 'KI270538.1', 'KI270539.1', 'KI270544.1', 'KI270548.1', 'KI270579.1', 'KI270580.1', 'KI270581.1', 'KI270582.1', 'KI270583.1', 'KI270584.1', 'KI270587.1', 'KI270588.1', 'KI270589.1', 'KI270590.1', 'KI270591.1', 'KI270593.1', 'KI270706.1', 'KI270707.1', 'KI270708.1', 'KI270709.1', 'KI270710.1', 'KI270711.1', 'KI270712.1', 'KI270713.1', 'KI270714.1', 'KI270715.1', 'KI270716.1', 'KI270717.1', 'KI270718.1', 'KI270719.1', 'KI270720.1', 'KI270721.1', 'KI270722.1', 'KI270723.1', 'KI270724.1', 'KI270725.1', 'KI270726.1', 'KI270727.1', 'KI270728.1', 'KI270729.1', 'KI270730.1', 'KI270731.1', 'KI270732.1', 'KI270733.1', 'KI270734.1', 'KI270735.1', 'KI270736.1', 'KI270737.1', 'KI270738.1', 'KI270739.1', 'KI270740.1', 'KI270741.1', 'KI270742.1', 'KI270743.1', 'KI270744.1', 'KI270745.1', 'KI270746.1', 'KI270747.1', 'KI270748.1', 'KI270749.1', 'KI270750.1', 'KI270751.1', 'KI270752.1', 'KI270753.1', 'KI270754.1', 'KI270755.1', 'KI270756.1', 'KI270757.1']\n"
     ]
    }
   ],
   "source": [
    "# List all sequence names\n",
    "sequence_names = reference_fasta.references\n",
    "print(\"Sequence names in the FASTA file:\", sequence_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "70ebeec1-32a5-4856-93c8-7904efaa4c8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CODE STARTED\n",
      "54d621a5-8e45-4bbc-92c4-1c7f36212361 (5236, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/54d621a5-8e45-4bbc-92c4-1c7f36212361 is created\n",
      "(5235, 12)\n",
      "&*\n",
      "(10470, 2)\n",
      "CODE STARTED\n",
      "ef6c3ecc-0bb6-4035-9c45-57740f5bcaa1 (7679, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/ef6c3ecc-0bb6-4035-9c45-57740f5bcaa1 is created\n",
      "(7677, 12)\n",
      "&*\n",
      "(15354, 2)\n",
      "CODE STARTED\n",
      "0f158d5f-16c7-4fb7-b48c-0ff392f1706b (5290, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/0f158d5f-16c7-4fb7-b48c-0ff392f1706b is created\n",
      "(5290, 12)\n",
      "&*\n",
      "(10580, 2)\n",
      "CODE STARTED\n",
      "da5fbf8f-862d-49a6-8119-86f508063e29 (5240, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/da5fbf8f-862d-49a6-8119-86f508063e29 is created\n",
      "(5239, 12)\n",
      "&*\n",
      "(10478, 2)\n",
      "CODE STARTED\n",
      "24abc004-6712-4df9-83f2-bd98dcc6ea5e (5228, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/24abc004-6712-4df9-83f2-bd98dcc6ea5e is created\n",
      "(5228, 12)\n",
      "&*\n",
      "(10456, 2)\n",
      "CODE STARTED\n",
      "535e0e9e-9757-45b3-aafd-2b80908de134 (5478, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/535e0e9e-9757-45b3-aafd-2b80908de134 is created\n",
      "(5478, 12)\n",
      "&*\n",
      "(10956, 2)\n",
      "CODE STARTED\n",
      "b2eaaa93-e870-4b1c-a835-b1d932c75b88 (6497, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/b2eaaa93-e870-4b1c-a835-b1d932c75b88 is created\n",
      "(6496, 12)\n",
      "&*\n",
      "(12992, 2)\n",
      "CODE STARTED\n",
      "a58b5cd5-4625-4070-b672-4eb92cd7775c (6910, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/a58b5cd5-4625-4070-b672-4eb92cd7775c is created\n",
      "(6905, 12)\n",
      "&*\n",
      "(13810, 2)\n",
      "CODE STARTED\n",
      "cf565879-9f53-4fd1-b0da-065869293342 (5290, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/cf565879-9f53-4fd1-b0da-065869293342 is created\n",
      "(5290, 12)\n",
      "&*\n",
      "(10580, 2)\n",
      "CODE STARTED\n",
      "0fc04044-0b33-4af6-a800-01d8cda12702 (5396, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/0fc04044-0b33-4af6-a800-01d8cda12702 is created\n",
      "(5396, 12)\n",
      "&*\n",
      "(10792, 2)\n",
      "CODE STARTED\n",
      "cc884f4d-e62d-4214-bfaa-81052a02a246 (5408, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/cc884f4d-e62d-4214-bfaa-81052a02a246 is created\n",
      "(5408, 12)\n",
      "&*\n",
      "(10816, 2)\n",
      "CODE STARTED\n",
      "d101eddd-ad00-4a42-9a95-ece5bd2332a2 (4970, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/d101eddd-ad00-4a42-9a95-ece5bd2332a2 is created\n",
      "(4969, 12)\n",
      "&*\n",
      "(9938, 2)\n",
      "CODE STARTED\n",
      "b5fb53b7-7a70-48a2-84ff-ca6ebe62d053 (5434, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/b5fb53b7-7a70-48a2-84ff-ca6ebe62d053 is created\n",
      "(5434, 12)\n",
      "&*\n",
      "(10868, 2)\n",
      "CODE STARTED\n",
      "7bed15a4-3b1e-487e-954d-e6b55e296dc0 (5520, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/7bed15a4-3b1e-487e-954d-e6b55e296dc0 is created\n",
      "(5520, 12)\n",
      "&*\n",
      "(11040, 2)\n",
      "CODE STARTED\n",
      "db856976-cd46-494d-bdc8-b4f3608cf10d (6978, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/db856976-cd46-494d-bdc8-b4f3608cf10d is created\n",
      "(6975, 12)\n",
      "&*\n",
      "(13950, 2)\n",
      "CODE STARTED\n",
      "2c9bbb8f-4b4d-4c0e-a158-1fb0fb95aca7 (5172, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/2c9bbb8f-4b4d-4c0e-a158-1fb0fb95aca7 is created\n",
      "(5167, 12)\n",
      "&*\n",
      "(10334, 2)\n",
      "CODE STARTED\n",
      "b78b66b2-2983-4965-99ff-8a7e45a07d8f (8106, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/b78b66b2-2983-4965-99ff-8a7e45a07d8f is created\n",
      "(8104, 12)\n",
      "&*\n",
      "(16208, 2)\n",
      "CODE STARTED\n",
      "fd27f3fc-6f0b-4794-8555-3111ad520493 (5313, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/fd27f3fc-6f0b-4794-8555-3111ad520493 is created\n",
      "(5308, 12)\n",
      "&*\n",
      "(10616, 2)\n",
      "CODE STARTED\n",
      "f00eed55-0ad9-41e8-bdf2-144b43f9b9b8 (5141, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/f00eed55-0ad9-41e8-bdf2-144b43f9b9b8 is created\n",
      "(5141, 12)\n",
      "&*\n",
      "(10282, 2)\n",
      "CODE STARTED\n",
      "22653df8-0c37-4f24-8832-95dd239f49b6 (5432, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/22653df8-0c37-4f24-8832-95dd239f49b6 is created\n",
      "(5432, 12)\n",
      "&*\n",
      "(10864, 2)\n",
      "CODE STARTED\n",
      "07f93ac0-50e0-4c1d-b4da-898062a5b6b8 (5625, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/07f93ac0-50e0-4c1d-b4da-898062a5b6b8 is created\n",
      "(5625, 12)\n",
      "&*\n",
      "(11250, 2)\n",
      "CODE STARTED\n",
      "6cb08eef-31fc-4ad7-9cc7-6176c4ea09a1 (5327, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/6cb08eef-31fc-4ad7-9cc7-6176c4ea09a1 is created\n",
      "(5327, 12)\n",
      "&*\n",
      "(10654, 2)\n",
      "CODE STARTED\n",
      "90c2456b-af8b-4e31-ae86-7de05a0f4124 (6603, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/90c2456b-af8b-4e31-ae86-7de05a0f4124 is created\n",
      "(6601, 12)\n",
      "&*\n",
      "(13202, 2)\n",
      "CODE STARTED\n",
      "1cc1ee58-e45c-4d85-b163-6d097f9d8f1f (5428, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/1cc1ee58-e45c-4d85-b163-6d097f9d8f1f is created\n",
      "(5428, 12)\n",
      "&*\n",
      "(10856, 2)\n",
      "CODE STARTED\n",
      "a5dba504-f9af-4ad7-8dd2-4d719440b569 (6474, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/a5dba504-f9af-4ad7-8dd2-4d719440b569 is created\n",
      "(6474, 12)\n",
      "&*\n",
      "(12948, 2)\n",
      "CODE STARTED\n",
      "3adc47fa-b27a-4df1-98ab-201ebab3c6bf (7777, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/3adc47fa-b27a-4df1-98ab-201ebab3c6bf is created\n",
      "(7775, 12)\n",
      "&*\n",
      "(15550, 2)\n",
      "CODE STARTED\n",
      "b7416dd2-f9ef-4fdc-8dd3-5e9b9e7169a3 (5348, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/b7416dd2-f9ef-4fdc-8dd3-5e9b9e7169a3 is created\n",
      "(5348, 12)\n",
      "&*\n",
      "(10696, 2)\n",
      "CODE STARTED\n",
      "f3ad237d-b424-4ca4-9659-2ff88a42e584 (5000, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/f3ad237d-b424-4ca4-9659-2ff88a42e584 is created\n",
      "(4999, 12)\n",
      "&*\n",
      "(9998, 2)\n",
      "CODE STARTED\n",
      "7d01d7c7-5138-4883-a4e3-c763151943df (5195, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/7d01d7c7-5138-4883-a4e3-c763151943df is created\n",
      "(5194, 12)\n",
      "&*\n",
      "(10388, 2)\n",
      "CODE STARTED\n",
      "2eb74322-a9f0-4dec-befc-b086386102a5 (5343, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/2eb74322-a9f0-4dec-befc-b086386102a5 is created\n",
      "(5341, 12)\n",
      "&*\n",
      "(10682, 2)\n",
      "CODE STARTED\n",
      "e5bd9afb-3fbd-4688-a376-795a51fd5079 (5468, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/e5bd9afb-3fbd-4688-a376-795a51fd5079 is created\n",
      "(5468, 12)\n",
      "&*\n",
      "(10936, 2)\n",
      "CODE STARTED\n",
      "f8b8e0df-4234-4839-b0ee-34a2c9ef4985 (5158, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/f8b8e0df-4234-4839-b0ee-34a2c9ef4985 is created\n",
      "(5157, 12)\n",
      "&*\n",
      "(10314, 2)\n",
      "CODE STARTED\n",
      "5d65b3c0-5c29-48aa-b6fb-37baf78fe409 (5248, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/5d65b3c0-5c29-48aa-b6fb-37baf78fe409 is created\n",
      "(5248, 12)\n",
      "&*\n",
      "(10496, 2)\n",
      "CODE STARTED\n",
      "09435dfd-67f9-4a28-9201-9e82d5bd04de (5470, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/09435dfd-67f9-4a28-9201-9e82d5bd04de is created\n",
      "(5470, 12)\n",
      "&*\n",
      "(10940, 2)\n",
      "CODE STARTED\n",
      "aae9d68e-f0eb-486b-9520-0b9f547f1ad3 (5323, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/aae9d68e-f0eb-486b-9520-0b9f547f1ad3 is created\n",
      "(5322, 12)\n",
      "&*\n",
      "(10644, 2)\n",
      "CODE STARTED\n",
      "c12f0212-31ee-4aa2-88c4-d29c7c9059cb (5247, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/c12f0212-31ee-4aa2-88c4-d29c7c9059cb is created\n",
      "(5247, 12)\n",
      "&*\n",
      "(10494, 2)\n",
      "CODE STARTED\n",
      "3d9d175b-20dc-4916-afa0-95b9acbbec02 (5445, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/3d9d175b-20dc-4916-afa0-95b9acbbec02 is created\n",
      "(5445, 12)\n",
      "&*\n",
      "(10890, 2)\n",
      "CODE STARTED\n",
      "ca52ea0e-5021-463b-8821-5e7c765812ff (7910, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/ca52ea0e-5021-463b-8821-5e7c765812ff is created\n",
      "(7904, 12)\n",
      "&*\n",
      "(15808, 2)\n",
      "CODE STARTED\n",
      "eef24407-c46d-45c3-9934-86902ac3ceac (5111, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/eef24407-c46d-45c3-9934-86902ac3ceac is created\n",
      "(5111, 12)\n",
      "&*\n",
      "(10222, 2)\n",
      "CODE STARTED\n",
      "3e28bc65-988d-4112-9397-7773ec79a9a9 (6733, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/3e28bc65-988d-4112-9397-7773ec79a9a9 is created\n",
      "(6731, 12)\n",
      "&*\n",
      "(13462, 2)\n",
      "CODE STARTED\n",
      "1d3daf9d-7740-484e-850b-827a6b7d451d (5420, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/1d3daf9d-7740-484e-850b-827a6b7d451d is created\n",
      "(5420, 12)\n",
      "&*\n",
      "(10840, 2)\n",
      "CODE STARTED\n",
      "bce00d1a-9598-4fef-8e21-da2c5290d906 (5218, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/bce00d1a-9598-4fef-8e21-da2c5290d906 is created\n",
      "(5218, 12)\n",
      "&*\n",
      "(10436, 2)\n",
      "CODE STARTED\n",
      "8bac8b71-abb7-4ecc-9b61-7112fed7da15 (5384, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/8bac8b71-abb7-4ecc-9b61-7112fed7da15 is created\n",
      "(5383, 12)\n",
      "&*\n",
      "(10766, 2)\n",
      "CODE STARTED\n",
      "8da9bde9-4b8c-46a2-bcc2-a620a186bab3 (5293, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/8da9bde9-4b8c-46a2-bcc2-a620a186bab3 is created\n",
      "(5293, 12)\n",
      "&*\n",
      "(10586, 2)\n",
      "CODE STARTED\n",
      "2a14ea45-6901-488a-af43-1e6287ec8bb0 (5305, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/2a14ea45-6901-488a-af43-1e6287ec8bb0 is created\n",
      "(5304, 12)\n",
      "&*\n",
      "(10608, 2)\n",
      "CODE STARTED\n",
      "a732538e-135a-43dc-a88b-b122a1a36749 (5437, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/a732538e-135a-43dc-a88b-b122a1a36749 is created\n",
      "(5432, 12)\n",
      "&*\n",
      "(10864, 2)\n",
      "CODE STARTED\n",
      "9a81060d-2b6d-4a97-a131-a5cf6c4c3ee4 (5480, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/9a81060d-2b6d-4a97-a131-a5cf6c4c3ee4 is created\n",
      "(5480, 12)\n",
      "&*\n",
      "(10960, 2)\n",
      "CODE STARTED\n",
      "9f7f9c66-39bb-48b1-ae8e-91acc8028425 (5261, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/9f7f9c66-39bb-48b1-ae8e-91acc8028425 is created\n",
      "(5261, 12)\n",
      "&*\n",
      "(10522, 2)\n",
      "CODE STARTED\n",
      "4193ad47-7cab-48fe-9466-aec494172f72 (5213, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/4193ad47-7cab-48fe-9466-aec494172f72 is created\n",
      "(5212, 12)\n",
      "&*\n",
      "(10424, 2)\n",
      "CODE STARTED\n",
      "dc3f72c3-79b4-411c-8275-a85e9d081abb (5404, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/dc3f72c3-79b4-411c-8275-a85e9d081abb is created\n",
      "(5404, 12)\n",
      "&*\n",
      "(10808, 2)\n",
      "CODE STARTED\n",
      "3adefeae-8c71-47a0-a9d5-3d0005aa0ef4 (5406, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/3adefeae-8c71-47a0-a9d5-3d0005aa0ef4 is created\n",
      "(5406, 12)\n",
      "&*\n",
      "(10812, 2)\n",
      "CODE STARTED\n",
      "b01b9f70-fc7d-43d1-bc8a-6c435ffd3705 (5584, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/b01b9f70-fc7d-43d1-bc8a-6c435ffd3705 is created\n",
      "(5584, 12)\n",
      "&*\n",
      "(11168, 2)\n",
      "CODE STARTED\n",
      "9a3edc2a-0eb3-4d88-965d-2801cd19df51 (4571, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/9a3edc2a-0eb3-4d88-965d-2801cd19df51 is created\n",
      "(4571, 12)\n",
      "&*\n",
      "(9142, 2)\n",
      "CODE STARTED\n",
      "d6cb373e-d9bc-43a1-8edc-2199b38b95f7 (5396, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/d6cb373e-d9bc-43a1-8edc-2199b38b95f7 is created\n",
      "(5395, 12)\n",
      "&*\n",
      "(10790, 2)\n",
      "CODE STARTED\n",
      "7d9bd465-2e99-4b07-902f-752fa348efb8 (6658, 24)\n",
      "/home/pdutta/Data/Cancer_wiseGDC/Data/Brain/Generated_files/DNABERT_Data/acceptor/Patient_wise/7d9bd465-2e99-4b07-902f-752fa348efb8 is created\n",
      "(6656, 12)\n",
      "&*\n",
      "(13312, 2)\n"
     ]
    }
   ],
   "source": [
    "dnabert_raw_data = {}\n",
    "for key, value in loaded_dictionary.items():\n",
    "    #display(value[['chr_name', 'start', 'end', 'strand', 'transcript_type', 'CHROM', 'START_POS', 'END_POS', 'ID', 'REF', 'ALT']])\n",
    "    print(\"CODE STARTED\")\n",
    "    print(key, value.shape)\n",
    "    out_folder_path = output_path+\"Patient_wise/\"+ key.split(\"_\")[0].split(\".\")[0]\n",
    "    print(out_folder_path, \"is created\")\n",
    "    #ins_del_df = value[(value['ALT'].str.len() > value['REF'].str.len()) | (value['REF'].str.len() > value['ALT'].str.len())]\n",
    "    new_df, df_kmer = get_sequences(value, reference_fasta)\n",
    "    dnabert_raw_data[key.split(\"_\")[0].split(\".\")[0]] = new_df\n",
    "    \n",
    "    #input()\n",
    "    if not os.path.exists(out_folder_path):\n",
    "        os.makedirs(out_folder_path)\n",
    "    df_kmer.to_csv(out_folder_path + \"/dev.tsv\", sep=\"\\t\", index= False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "802fdc95-492f-4ff6-a884-cee2cee2e498",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(output_path+\"/raw_{}_vcf_data.pkl\".format(non_coding_region), \"wb\") as file:\n",
    "    pickle.dump(dnabert_raw_data, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b62f3450-e730-4951-9d7d-69f9fc4e16a2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
