{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "cdd0a298-de34-4915-8bae-192d11cb06ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, tarfile\n",
    "import re, glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import zipfile, gzip\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.utils import shuffle\n",
    "import itertools\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "c6f90d5f-d0af-47a7-8d89-fa101ffeecb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_file_path = \"/data/private/pdutta/Collab_data/Oliver_data/APOE_TREM2/TREM2_new.vcf\"\n",
    "reference_file_path = \"/data/projects/Resources/HumanReferenceGenome/chr19.fa\"\n",
    "output_path = \"/data/private/pdutta/Collab_data/Oliver_data/DNABERT_data/TREM2_CrProm\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "56537918-9aed-40e0-bd4c-30b85b4f10c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(output_path):\n",
    "    os.makedirs(output_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfdba9d5-bcd7-4508-9bf3-accb1e527093",
   "metadata": {},
   "source": [
    "## Getting the VCF file and get the coordinates and alternative allele information "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "eb11de29-bd63-4c8b-a640-f2e5c30781b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_file = pysam.VariantFile(vcf_file_path)\n",
    "reference_fasta = pysam.FastaFile(reference_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "f8cf0b5f-e0cb-4ac6-b6d0-aee02c4966a2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alternative Allele ('A',) 6 41158608 C\n",
      "Alternative Allele ('T',) 6 41158620 C\n",
      "Alternative Allele ('A',) 6 41158657 C\n",
      "Alternative Allele ('A',) 6 41158680 G\n",
      "Alternative Allele ('T',) 6 41158685 G\n",
      "Alternative Allele ('T',) 6 41158691 C\n",
      "Alternative Allele ('T',) 6 41158720 C\n",
      "Alternative Allele ('G',) 6 41158721 C\n",
      "Alternative Allele ('T',) 6 41158747 C\n",
      "Alternative Allele ('T',) 6 41158767 C\n",
      "Alternative Allele ('A',) 6 41158768 G\n",
      "Alternative Allele ('A',) 6 41158875 G\n",
      "Alternative Allele ('A',) 6 41158881 G\n",
      "Alternative Allele ('C',) 6 41158904 A\n",
      "Alternative Allele ('G',) 6 41158917 A\n",
      "Alternative Allele ('C',) 6 41158927 G\n",
      "Alternative Allele ('T',) 6 41158963 C\n",
      "Alternative Allele ('A',) 6 41159042 G\n",
      "Alternative Allele ('T',) 6 41159052 C\n",
      "Alternative Allele ('C',) 6 41159063 G\n",
      "Alternative Allele ('G',) 6 41159068 T\n",
      "Alternative Allele ('A',) 6 41159805 G\n",
      "Alternative Allele ('T',) 6 41159823 C\n",
      "Alternative Allele ('A',) 6 41159845 G\n",
      "Alternative Allele ('T',) 6 41159867 C\n",
      "Alternative Allele ('A',) 6 41159868 G\n",
      "Alternative Allele ('A',) 6 41159875 C\n",
      "Alternative Allele ('T',) 6 41159877 G\n",
      "Alternative Allele ('C',) 6 41159882 T\n",
      "Alternative Allele ('A',) 6 41161366 C\n",
      "Alternative Allele ('T',) 6 41161395 C\n",
      "Alternative Allele ('C',) 6 41161450 G\n",
      "Alternative Allele ('A',) 6 41161457 G\n",
      "Alternative Allele ('T',) 6 41161469 C\n",
      "Alternative Allele ('A',) 6 41161470 G\n",
      "Alternative Allele ('T',) 6 41161514 C\n",
      "Alternative Allele ('A',) 6 41161515 G\n",
      "Alternative Allele ('G',) 6 41161519 C\n",
      "Alternative Allele ('C',) 6 41161537 G\n",
      "Alternative Allele ('A',) 6 41161557 G\n",
      "Alternative Allele ('A',) 6 41161562 G\n",
      "Alternative Allele ('G',) 6 41161592 GTGT\n",
      "Alternative Allele ('A',) 6 41161606 G\n",
      "Alternative Allele ('A',) 6 41161607 G\n",
      "Alternative Allele ('G',) 6 41161608 A\n",
      "Alternative Allele ('G',) 6 41163086 A\n",
      "Alternative Allele ('T',) 6 41163134 G\n"
     ]
    }
   ],
   "source": [
    "position_list = []\n",
    "alternative_allele_list = []\n",
    "for rec in vcf_file.fetch():\n",
    "    # You can access all data about the variant like this:\n",
    "    print('Alternative Allele',  rec.alts, rec.chrom, rec.pos, rec.ref)\n",
    "    position_list.append(rec.pos)\n",
    "    alternative_allele_list.append(rec.alts[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "2f124c7e-09ee-4008-a4c0-dc2e6bf6d171",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41158608 41163134\n"
     ]
    }
   ],
   "source": [
    "min_position = min(position_list)\n",
    "max_position = max(position_list)\n",
    "print(min_position, max_position)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "d04acf8a-76f4-434d-8aed-6f74869db4e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "position_list= np.array(position_list)\n",
    "alternative_allele_list = np.array(alternative_allele_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7fe5abe-ddbd-4969-a5e3-2cb25d8540f0",
   "metadata": {},
   "source": [
    "## Get the co-ordinates from Core Promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "d5d06849-9c30-4a94-9a8f-5f7fca5a095e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>TSS</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>ENSG00000004059.11</td>\n",
       "      <td>ENST00000000233.10</td>\n",
       "      <td>ENSE00001872691.2</td>\n",
       "      <td>+</td>\n",
       "      <td>127588411</td>\n",
       "      <td>127588366</td>\n",
       "      <td>127588455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7</td>\n",
       "      <td>ENSG00000001626.16</td>\n",
       "      <td>ENST00000003084.11</td>\n",
       "      <td>ENSE00001343851.2</td>\n",
       "      <td>+</td>\n",
       "      <td>117480025</td>\n",
       "      <td>117479980</td>\n",
       "      <td>117480069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr7</td>\n",
       "      <td>ENSG00000001630.17</td>\n",
       "      <td>ENST00000003100.13</td>\n",
       "      <td>ENSE00001216550.10</td>\n",
       "      <td>-</td>\n",
       "      <td>92134477</td>\n",
       "      <td>92134433</td>\n",
       "      <td>92134522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7</td>\n",
       "      <td>ENSG00000002933.9</td>\n",
       "      <td>ENST00000004103.8</td>\n",
       "      <td>ENSE00001827596.2</td>\n",
       "      <td>+</td>\n",
       "      <td>150800769</td>\n",
       "      <td>150800724</td>\n",
       "      <td>150800813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr7</td>\n",
       "      <td>ENSG00000004799.8</td>\n",
       "      <td>ENST00000005178.6</td>\n",
       "      <td>ENSE00001133344.6</td>\n",
       "      <td>-</td>\n",
       "      <td>95596516</td>\n",
       "      <td>95596472</td>\n",
       "      <td>95596561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236970</th>\n",
       "      <td>chr15</td>\n",
       "      <td>ENSG00000103876.14</td>\n",
       "      <td>ENST00000684569.1</td>\n",
       "      <td>ENSE00003921718.1</td>\n",
       "      <td>+</td>\n",
       "      <td>80153010</td>\n",
       "      <td>80152965</td>\n",
       "      <td>80153054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236971</th>\n",
       "      <td>chr15</td>\n",
       "      <td>ENSG00000213614.11</td>\n",
       "      <td>ENST00000684602.1</td>\n",
       "      <td>ENSE00002620879.1</td>\n",
       "      <td>-</td>\n",
       "      <td>72375989</td>\n",
       "      <td>72375945</td>\n",
       "      <td>72376034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236972</th>\n",
       "      <td>chr15</td>\n",
       "      <td>ENSG00000213614.11</td>\n",
       "      <td>ENST00000684667.1</td>\n",
       "      <td>ENSE00002584031.1</td>\n",
       "      <td>-</td>\n",
       "      <td>72375972</td>\n",
       "      <td>72375928</td>\n",
       "      <td>72376017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236973</th>\n",
       "      <td>chr15</td>\n",
       "      <td>ENSG00000104133.16</td>\n",
       "      <td>ENST00000684676.1</td>\n",
       "      <td>ENSE00003922470.1</td>\n",
       "      <td>-</td>\n",
       "      <td>44663688</td>\n",
       "      <td>44663644</td>\n",
       "      <td>44663733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236974</th>\n",
       "      <td>chr15</td>\n",
       "      <td>ENSG00000183060.16</td>\n",
       "      <td>ENST00000684762.1</td>\n",
       "      <td>ENSE00001781053.1</td>\n",
       "      <td>-</td>\n",
       "      <td>99733437</td>\n",
       "      <td>99733393</td>\n",
       "      <td>99733482</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>236975 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Chr             gene_id       transcript_id             exon_id  \\\n",
       "0        chr7  ENSG00000004059.11  ENST00000000233.10   ENSE00001872691.2   \n",
       "1        chr7  ENSG00000001626.16  ENST00000003084.11   ENSE00001343851.2   \n",
       "2        chr7  ENSG00000001630.17  ENST00000003100.13  ENSE00001216550.10   \n",
       "3        chr7   ENSG00000002933.9   ENST00000004103.8   ENSE00001827596.2   \n",
       "4        chr7   ENSG00000004799.8   ENST00000005178.6   ENSE00001133344.6   \n",
       "...       ...                 ...                 ...                 ...   \n",
       "236970  chr15  ENSG00000103876.14   ENST00000684569.1   ENSE00003921718.1   \n",
       "236971  chr15  ENSG00000213614.11   ENST00000684602.1   ENSE00002620879.1   \n",
       "236972  chr15  ENSG00000213614.11   ENST00000684667.1   ENSE00002584031.1   \n",
       "236973  chr15  ENSG00000104133.16   ENST00000684676.1   ENSE00003922470.1   \n",
       "236974  chr15  ENSG00000183060.16   ENST00000684762.1   ENSE00001781053.1   \n",
       "\n",
       "       strand        TSS      start        end  \n",
       "0           +  127588411  127588366  127588455  \n",
       "1           +  117480025  117479980  117480069  \n",
       "2           -   92134477   92134433   92134522  \n",
       "3           +  150800769  150800724  150800813  \n",
       "4           -   95596516   95596472   95596561  \n",
       "...       ...        ...        ...        ...  \n",
       "236970      +   80153010   80152965   80153054  \n",
       "236971      -   72375989   72375945   72376034  \n",
       "236972      -   72375972   72375928   72376017  \n",
       "236973      -   44663688   44663644   44663733  \n",
       "236974      -   99733437   99733393   99733482  \n",
       "\n",
       "[236975 rows x 8 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "core_prom_bed = pd.read_csv(\"/data/private/pdutta/DNABERT_data/Core_promoters/Core_Prom_bed.tsv\", sep= \"\\t\")\n",
    "core_prom_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "262d0703-b701-4d7b-ba09-00842b1c9a23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>TSS</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000001036.14</td>\n",
       "      <td>ENST00000002165.11</td>\n",
       "      <td>ENSE00003705756.2</td>\n",
       "      <td>-</td>\n",
       "      <td>143511720</td>\n",
       "      <td>143511676</td>\n",
       "      <td>143511765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000008197.5</td>\n",
       "      <td>ENST00000008391.4</td>\n",
       "      <td>ENSE00001843453.2</td>\n",
       "      <td>+</td>\n",
       "      <td>50713526</td>\n",
       "      <td>50713481</td>\n",
       "      <td>50713570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000010017.13</td>\n",
       "      <td>ENST00000011619.6</td>\n",
       "      <td>ENSE00001815244.2</td>\n",
       "      <td>-</td>\n",
       "      <td>13711835</td>\n",
       "      <td>13711791</td>\n",
       "      <td>13711880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000010818.10</td>\n",
       "      <td>ENST00000012134.7</td>\n",
       "      <td>ENSE00003804252.1</td>\n",
       "      <td>-</td>\n",
       "      <td>142945031</td>\n",
       "      <td>142944987</td>\n",
       "      <td>142945076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000026297.17</td>\n",
       "      <td>ENST00000028008.9</td>\n",
       "      <td>ENSE00001855489.2</td>\n",
       "      <td>-</td>\n",
       "      <td>166956584</td>\n",
       "      <td>166956540</td>\n",
       "      <td>166956629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10995</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000155906.20</td>\n",
       "      <td>ENST00000684715.1</td>\n",
       "      <td>ENSE00003917755.1</td>\n",
       "      <td>-</td>\n",
       "      <td>151452146</td>\n",
       "      <td>151452102</td>\n",
       "      <td>151452191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10996</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000135317.14</td>\n",
       "      <td>ENST00000684717.1</td>\n",
       "      <td>ENSE00003916063.1</td>\n",
       "      <td>-</td>\n",
       "      <td>85593897</td>\n",
       "      <td>85593853</td>\n",
       "      <td>85593942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10997</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000172469.17</td>\n",
       "      <td>ENST00000684753.1</td>\n",
       "      <td>ENSE00003920710.1</td>\n",
       "      <td>+</td>\n",
       "      <td>95577865</td>\n",
       "      <td>95577820</td>\n",
       "      <td>95577909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10998</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000155906.20</td>\n",
       "      <td>ENST00000684765.1</td>\n",
       "      <td>ENSE00003917755.1</td>\n",
       "      <td>-</td>\n",
       "      <td>151452146</td>\n",
       "      <td>151452102</td>\n",
       "      <td>151452191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10999</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000112319.20</td>\n",
       "      <td>ENST00000684773.1</td>\n",
       "      <td>ENSE00003918385.1</td>\n",
       "      <td>+</td>\n",
       "      <td>133241632</td>\n",
       "      <td>133241587</td>\n",
       "      <td>133241676</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11000 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr             gene_id       transcript_id            exon_id strand  \\\n",
       "0      chr6  ENSG00000001036.14  ENST00000002165.11  ENSE00003705756.2      -   \n",
       "1      chr6   ENSG00000008197.5   ENST00000008391.4  ENSE00001843453.2      +   \n",
       "2      chr6  ENSG00000010017.13   ENST00000011619.6  ENSE00001815244.2      -   \n",
       "3      chr6  ENSG00000010818.10   ENST00000012134.7  ENSE00003804252.1      -   \n",
       "4      chr6  ENSG00000026297.17   ENST00000028008.9  ENSE00001855489.2      -   \n",
       "...     ...                 ...                 ...                ...    ...   \n",
       "10995  chr6  ENSG00000155906.20   ENST00000684715.1  ENSE00003917755.1      -   \n",
       "10996  chr6  ENSG00000135317.14   ENST00000684717.1  ENSE00003916063.1      -   \n",
       "10997  chr6  ENSG00000172469.17   ENST00000684753.1  ENSE00003920710.1      +   \n",
       "10998  chr6  ENSG00000155906.20   ENST00000684765.1  ENSE00003917755.1      -   \n",
       "10999  chr6  ENSG00000112319.20   ENST00000684773.1  ENSE00003918385.1      +   \n",
       "\n",
       "             TSS      start        end  \n",
       "0      143511720  143511676  143511765  \n",
       "1       50713526   50713481   50713570  \n",
       "2       13711835   13711791   13711880  \n",
       "3      142945031  142944987  142945076  \n",
       "4      166956584  166956540  166956629  \n",
       "...          ...        ...        ...  \n",
       "10995  151452146  151452102  151452191  \n",
       "10996   85593897   85593853   85593942  \n",
       "10997   95577865   95577820   95577909  \n",
       "10998  151452146  151452102  151452191  \n",
       "10999  133241632  133241587  133241676  \n",
       "\n",
       "[11000 rows x 8 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "core_prom_19  = core_prom_bed[core_prom_bed['Chr']=='chr6'].reset_index(drop=True)\n",
    "core_prom_19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "257d9813-481b-46c4-ac73-d2fa85c52fd9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>TSS</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000001036.14</td>\n",
       "      <td>ENST00000002165.11</td>\n",
       "      <td>ENSE00003705756.2</td>\n",
       "      <td>-</td>\n",
       "      <td>143511720</td>\n",
       "      <td>143511676</td>\n",
       "      <td>143511765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000008197.5</td>\n",
       "      <td>ENST00000008391.4</td>\n",
       "      <td>ENSE00001843453.2</td>\n",
       "      <td>+</td>\n",
       "      <td>50713526</td>\n",
       "      <td>50713481</td>\n",
       "      <td>50713570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000010818.10</td>\n",
       "      <td>ENST00000012134.7</td>\n",
       "      <td>ENSE00003804252.1</td>\n",
       "      <td>-</td>\n",
       "      <td>142945031</td>\n",
       "      <td>142944987</td>\n",
       "      <td>142945076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000026297.17</td>\n",
       "      <td>ENST00000028008.9</td>\n",
       "      <td>ENSE00001855489.2</td>\n",
       "      <td>-</td>\n",
       "      <td>166956584</td>\n",
       "      <td>166956540</td>\n",
       "      <td>166956629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000047932.14</td>\n",
       "      <td>ENST00000052569.10</td>\n",
       "      <td>ENSE00000919104.8</td>\n",
       "      <td>-</td>\n",
       "      <td>117602528</td>\n",
       "      <td>117602484</td>\n",
       "      <td>117602573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10995</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000155906.20</td>\n",
       "      <td>ENST00000684715.1</td>\n",
       "      <td>ENSE00003917755.1</td>\n",
       "      <td>-</td>\n",
       "      <td>151452146</td>\n",
       "      <td>151452102</td>\n",
       "      <td>151452191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10996</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000135317.14</td>\n",
       "      <td>ENST00000684717.1</td>\n",
       "      <td>ENSE00003916063.1</td>\n",
       "      <td>-</td>\n",
       "      <td>85593897</td>\n",
       "      <td>85593853</td>\n",
       "      <td>85593942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10997</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000172469.17</td>\n",
       "      <td>ENST00000684753.1</td>\n",
       "      <td>ENSE00003920710.1</td>\n",
       "      <td>+</td>\n",
       "      <td>95577865</td>\n",
       "      <td>95577820</td>\n",
       "      <td>95577909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10998</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000155906.20</td>\n",
       "      <td>ENST00000684765.1</td>\n",
       "      <td>ENSE00003917755.1</td>\n",
       "      <td>-</td>\n",
       "      <td>151452146</td>\n",
       "      <td>151452102</td>\n",
       "      <td>151452191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10999</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000112319.20</td>\n",
       "      <td>ENST00000684773.1</td>\n",
       "      <td>ENSE00003918385.1</td>\n",
       "      <td>+</td>\n",
       "      <td>133241632</td>\n",
       "      <td>133241587</td>\n",
       "      <td>133241676</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6801 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr             gene_id       transcript_id            exon_id strand  \\\n",
       "0      chr6  ENSG00000001036.14  ENST00000002165.11  ENSE00003705756.2      -   \n",
       "1      chr6   ENSG00000008197.5   ENST00000008391.4  ENSE00001843453.2      +   \n",
       "3      chr6  ENSG00000010818.10   ENST00000012134.7  ENSE00003804252.1      -   \n",
       "4      chr6  ENSG00000026297.17   ENST00000028008.9  ENSE00001855489.2      -   \n",
       "5      chr6  ENSG00000047932.14  ENST00000052569.10  ENSE00000919104.8      -   \n",
       "...     ...                 ...                 ...                ...    ...   \n",
       "10995  chr6  ENSG00000155906.20   ENST00000684715.1  ENSE00003917755.1      -   \n",
       "10996  chr6  ENSG00000135317.14   ENST00000684717.1  ENSE00003916063.1      -   \n",
       "10997  chr6  ENSG00000172469.17   ENST00000684753.1  ENSE00003920710.1      +   \n",
       "10998  chr6  ENSG00000155906.20   ENST00000684765.1  ENSE00003917755.1      -   \n",
       "10999  chr6  ENSG00000112319.20   ENST00000684773.1  ENSE00003918385.1      +   \n",
       "\n",
       "             TSS      start        end  \n",
       "0      143511720  143511676  143511765  \n",
       "1       50713526   50713481   50713570  \n",
       "3      142945031  142944987  142945076  \n",
       "4      166956584  166956540  166956629  \n",
       "5      117602528  117602484  117602573  \n",
       "...          ...        ...        ...  \n",
       "10995  151452146  151452102  151452191  \n",
       "10996   85593897   85593853   85593942  \n",
       "10997   95577865   95577820   95577909  \n",
       "10998  151452146  151452102  151452191  \n",
       "10999  133241632  133241587  133241676  \n",
       "\n",
       "[6801 rows x 8 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "core_prom_19[(core_prom_19['start']>=min_position)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1597dc25-9c38-47c6-8b26-ba5dd3d259e0",
   "metadata": {},
   "source": [
    "## Get the Core promoter regions which contains the variant information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "19087b84-84a5-4f12-b2d2-6323edbccf48",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([41158608, 41158620, 41158657, 41158680, 41158685, 41158691,\n",
       "       41158720, 41158721, 41158747, 41158767, 41158768, 41158875,\n",
       "       41158881, 41158904, 41158917, 41158927, 41158963, 41159042,\n",
       "       41159052, 41159063, 41159068, 41159805, 41159823, 41159845,\n",
       "       41159867, 41159868, 41159875, 41159877, 41159882, 41161366,\n",
       "       41161395, 41161450, 41161457, 41161469, 41161470, 41161514,\n",
       "       41161515, 41161519, 41161537, 41161557, 41161562, 41161592,\n",
       "       41161606, 41161607, 41161608, 41163086, 41163134])"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "position_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "eb43055b-72c3-4672-9ead-072873b1119e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "41158608"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min_position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "4ab8325e-8ed9-4d4d-8ee3-5945180e7c58",
   "metadata": {},
   "outputs": [],
   "source": [
    "core_prom_19['vcf_position'] = pd.Series([[] for _ in range(len(core_prom_19))], index=core_prom_19.index)\n",
    "core_prom_19['alts'] = pd.Series([[] for _ in range(len(core_prom_19))], index=core_prom_19.index)\n",
    "\n",
    "for alt, position in zip(alternative_allele_list, position_list):\n",
    "    # Check if the position falls within the range of any row\n",
    "    mask = (core_prom_19['start'] <= position) & (core_prom_19['end'] >= position)\n",
    "    # If it does, store the index of the row\n",
    "    #core_prom_19.loc[mask, 'vcf_position'] = core_prom_19.loc[mask, 'vcf_position'].apply(lambda x: x.append(position) if pd.notnull(x) else [position])\n",
    "    core_prom_19.loc[mask, 'vcf_position'] = core_prom_19.loc[mask, 'vcf_position'].apply(lambda x: x + [position])\n",
    "    core_prom_19.loc[mask, 'alts'] = core_prom_19.loc[mask, 'alts'].apply(lambda x: x + [alt])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "1d0902c3-1230-48fb-88f2-a7ad60eb615a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>TSS</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>vcf_position</th>\n",
       "      <th>alts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1429</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000095970.17</td>\n",
       "      <td>ENST00000373113.8</td>\n",
       "      <td>ENSE00001922142.2</td>\n",
       "      <td>-</td>\n",
       "      <td>41163116</td>\n",
       "      <td>41163072</td>\n",
       "      <td>41163161</td>\n",
       "      <td>[41163086, 41163134]</td>\n",
       "      <td>[G, T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1430</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000095970.17</td>\n",
       "      <td>ENST00000373122.8</td>\n",
       "      <td>ENSE00002318238.2</td>\n",
       "      <td>-</td>\n",
       "      <td>41163103</td>\n",
       "      <td>41163059</td>\n",
       "      <td>41163148</td>\n",
       "      <td>[41163086, 41163134]</td>\n",
       "      <td>[G, T]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chr             gene_id      transcript_id            exon_id strand  \\\n",
       "1429  chr6  ENSG00000095970.17  ENST00000373113.8  ENSE00001922142.2      -   \n",
       "1430  chr6  ENSG00000095970.17  ENST00000373122.8  ENSE00002318238.2      -   \n",
       "\n",
       "           TSS     start       end          vcf_position    alts  \n",
       "1429  41163116  41163072  41163161  [41163086, 41163134]  [G, T]  \n",
       "1430  41163103  41163059  41163148  [41163086, 41163134]  [G, T]  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_rows= core_prom_19[core_prom_19['vcf_position'].apply(lambda x: len(x) > 0)]\n",
    "selected_rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "d5377734-eade-4537-af49-58537f9a27b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41158608 41163134\n"
     ]
    }
   ],
   "source": [
    "print(min_position , max_position)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "7ad1aa86-1cf1-43f2-9713-4d8cf2974d7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = SeqIO.to_dict(SeqIO.parse(\"/data/projects/Resources/Gencode_genome_annotation/GRCh38.primary_assembly.genome.fa\", \"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "6e0776af-7543-46d0-bb00-538e709f2ffe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGCCTTGTGCAAGATCTCGTCTTTCCCTTGAACTGCAGAAAAGAGA\n",
      "41163072 41163086 14 G\n",
      "GAGGCTCCATGCCGCCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGCCTTGTGCAAGATCTCGTCTTTCCCTTGAACTGCAGAAAAGAGA A\n",
      "41163072 41163134 62 T\n",
      "GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGCCTTGTGCAAGATCTCTTCTTTCCCTTGAACTGCAGAAAAGAGA G\n",
      "ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGCCTTGTGCAAGATCTCGTCTTTCCCTTGAAC\n",
      "41163059 41163086 27 G\n",
      "ATGAGCAGCCGGAGAGGCTCCATGCCGCCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGCCTTGTGCAAGATCTCGTCTTTCCCTTGAAC A\n",
      "41163059 41163134 75 T\n",
      "ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGCCTTGTGCAAGATCTCTTCTTTCCCTTGAAC G\n"
     ]
    }
   ],
   "source": [
    "data = []\n",
    "for _, row in selected_rows.iterrows():\n",
    "    chrom = row['Chr']\n",
    "    ref_start = row['start']\n",
    "    ref_end = row['end']\n",
    "    variant_position_list = row['vcf_position']\n",
    "    alts_list = row['alts']\n",
    "\n",
    "    # Get reference sequence\n",
    "    ref_seq = str(genome[chrom].seq[ref_start:ref_end])\n",
    "    print(ref_seq)\n",
    "    \n",
    "    for variant_position, alt in zip(variant_position_list, alts_list):\n",
    "        seq_variant_pos = variant_position - ref_start\n",
    "        print(ref_start, variant_position, seq_variant_pos, alt)\n",
    "        alt_seq = ref_seq[:seq_variant_pos-1] + alt + ref_seq[seq_variant_pos:]\n",
    "        print(alt_seq, ref_seq[seq_variant_pos-1])\n",
    "        data.append({\n",
    "                'chr': chrom,\n",
    "                'strand': row['strand'],\n",
    "                'Core_Prom_start': ref_start,\n",
    "                'Core_Prom_end': ref_end,\n",
    "                'VCF_position': variant_position,\n",
    "                'ref_neucleotide': ref_seq[seq_variant_pos-1],\n",
    "                'alt_neucleotide': alt,\n",
    "                'reference_seq': ref_seq,\n",
    "                'alt_seq': alt_seq\n",
    "            })\n",
    "new_df = pd.DataFrame(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "8c9b61f9-e906-4010-9853-d8ff872bbab5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>Core_Prom_start</th>\n",
       "      <th>Core_Prom_end</th>\n",
       "      <th>VCF_position</th>\n",
       "      <th>ref_neucleotide</th>\n",
       "      <th>alt_neucleotide</th>\n",
       "      <th>reference_seq</th>\n",
       "      <th>alt_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163072</td>\n",
       "      <td>41163161</td>\n",
       "      <td>41163086</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...</td>\n",
       "      <td>GAGGCTCCATGCCGCCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163072</td>\n",
       "      <td>41163161</td>\n",
       "      <td>41163134</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...</td>\n",
       "      <td>GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163059</td>\n",
       "      <td>41163148</td>\n",
       "      <td>41163086</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGC...</td>\n",
       "      <td>ATGAGCAGCCGGAGAGGCTCCATGCCGCCCTTCCCCAGCCAAGGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163059</td>\n",
       "      <td>41163148</td>\n",
       "      <td>41163134</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGC...</td>\n",
       "      <td>ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr strand  Core_Prom_start  Core_Prom_end  VCF_position ref_neucleotide  \\\n",
       "0  chr6      -         41163072       41163161      41163086               A   \n",
       "1  chr6      -         41163072       41163161      41163134               G   \n",
       "2  chr6      -         41163059       41163148      41163086               A   \n",
       "3  chr6      -         41163059       41163148      41163134               G   \n",
       "\n",
       "  alt_neucleotide                                      reference_seq  \\\n",
       "0               G  GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...   \n",
       "1               T  GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...   \n",
       "2               G  ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGC...   \n",
       "3               T  ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGC...   \n",
       "\n",
       "                                             alt_seq  \n",
       "0  GAGGCTCCATGCCGCCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...  \n",
       "1  GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...  \n",
       "2  ATGAGCAGCCGGAGAGGCTCCATGCCGCCCTTCCCCAGCCAAGGGC...  \n",
       "3  ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGC...  "
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "1259d1c9-7528-4d42-b818-849129750b1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1993411/2717605502.py:4: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  df_data = df_data.append({'chr': new_df.iloc[i]['chr'], 'strand': new_df.iloc[i]['strand'], 'Core_Prom_start': new_df.iloc[i]['Core_Prom_start'], 'Core_Prom_end': new_df.iloc[i]['Core_Prom_end'], 'VCF_position': new_df.iloc[i]['VCF_position'], \"Data\": new_df.iloc[i]['reference_seq'], 'Label': 1}, ignore_index=True)\n",
      "/tmp/ipykernel_1993411/2717605502.py:5: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  df_data = df_data.append({'chr': new_df.iloc[i]['chr'], 'strand': new_df.iloc[i]['strand'], 'Core_Prom_start': new_df.iloc[i]['Core_Prom_start'], 'Core_Prom_end': new_df.iloc[i]['Core_Prom_end'], 'VCF_position': new_df.iloc[i]['VCF_position'], \"Data\": new_df.iloc[i]['alt_seq'], 'Label': 0}, ignore_index=True)\n",
      "/tmp/ipykernel_1993411/2717605502.py:4: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  df_data = df_data.append({'chr': new_df.iloc[i]['chr'], 'strand': new_df.iloc[i]['strand'], 'Core_Prom_start': new_df.iloc[i]['Core_Prom_start'], 'Core_Prom_end': new_df.iloc[i]['Core_Prom_end'], 'VCF_position': new_df.iloc[i]['VCF_position'], \"Data\": new_df.iloc[i]['reference_seq'], 'Label': 1}, ignore_index=True)\n",
      "/tmp/ipykernel_1993411/2717605502.py:5: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  df_data = df_data.append({'chr': new_df.iloc[i]['chr'], 'strand': new_df.iloc[i]['strand'], 'Core_Prom_start': new_df.iloc[i]['Core_Prom_start'], 'Core_Prom_end': new_df.iloc[i]['Core_Prom_end'], 'VCF_position': new_df.iloc[i]['VCF_position'], \"Data\": new_df.iloc[i]['alt_seq'], 'Label': 0}, ignore_index=True)\n",
      "/tmp/ipykernel_1993411/2717605502.py:4: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  df_data = df_data.append({'chr': new_df.iloc[i]['chr'], 'strand': new_df.iloc[i]['strand'], 'Core_Prom_start': new_df.iloc[i]['Core_Prom_start'], 'Core_Prom_end': new_df.iloc[i]['Core_Prom_end'], 'VCF_position': new_df.iloc[i]['VCF_position'], \"Data\": new_df.iloc[i]['reference_seq'], 'Label': 1}, ignore_index=True)\n",
      "/tmp/ipykernel_1993411/2717605502.py:5: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  df_data = df_data.append({'chr': new_df.iloc[i]['chr'], 'strand': new_df.iloc[i]['strand'], 'Core_Prom_start': new_df.iloc[i]['Core_Prom_start'], 'Core_Prom_end': new_df.iloc[i]['Core_Prom_end'], 'VCF_position': new_df.iloc[i]['VCF_position'], \"Data\": new_df.iloc[i]['alt_seq'], 'Label': 0}, ignore_index=True)\n",
      "/tmp/ipykernel_1993411/2717605502.py:4: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  df_data = df_data.append({'chr': new_df.iloc[i]['chr'], 'strand': new_df.iloc[i]['strand'], 'Core_Prom_start': new_df.iloc[i]['Core_Prom_start'], 'Core_Prom_end': new_df.iloc[i]['Core_Prom_end'], 'VCF_position': new_df.iloc[i]['VCF_position'], \"Data\": new_df.iloc[i]['reference_seq'], 'Label': 1}, ignore_index=True)\n",
      "/tmp/ipykernel_1993411/2717605502.py:5: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  df_data = df_data.append({'chr': new_df.iloc[i]['chr'], 'strand': new_df.iloc[i]['strand'], 'Core_Prom_start': new_df.iloc[i]['Core_Prom_start'], 'Core_Prom_end': new_df.iloc[i]['Core_Prom_end'], 'VCF_position': new_df.iloc[i]['VCF_position'], \"Data\": new_df.iloc[i]['alt_seq'], 'Label': 0}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# Melt the DataFrame to a long format\n",
    "df_data = pd.DataFrame(columns=['chr', 'strand','Core_Prom_start', 'Core_Prom_end', \"VCF_position\", \"Data\" ,\"Label\"])\n",
    "for i in range(len(new_df)):\n",
    "    df_data = df_data.append({'chr': new_df.iloc[i]['chr'], 'strand': new_df.iloc[i]['strand'], 'Core_Prom_start': new_df.iloc[i]['Core_Prom_start'], 'Core_Prom_end': new_df.iloc[i]['Core_Prom_end'], 'VCF_position': new_df.iloc[i]['VCF_position'], \"Data\": new_df.iloc[i]['reference_seq'], 'Label': 1}, ignore_index=True)\n",
    "    df_data = df_data.append({'chr': new_df.iloc[i]['chr'], 'strand': new_df.iloc[i]['strand'], 'Core_Prom_start': new_df.iloc[i]['Core_Prom_start'], 'Core_Prom_end': new_df.iloc[i]['Core_Prom_end'], 'VCF_position': new_df.iloc[i]['VCF_position'], \"Data\": new_df.iloc[i]['alt_seq'], 'Label': 0}, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "95144528-d32e-4023-9d73-aab414a692c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>Core_Prom_start</th>\n",
       "      <th>Core_Prom_end</th>\n",
       "      <th>VCF_position</th>\n",
       "      <th>Data</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163072</td>\n",
       "      <td>41163161</td>\n",
       "      <td>41163086</td>\n",
       "      <td>GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163072</td>\n",
       "      <td>41163161</td>\n",
       "      <td>41163086</td>\n",
       "      <td>GAGGCTCCATGCCGCCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163072</td>\n",
       "      <td>41163161</td>\n",
       "      <td>41163134</td>\n",
       "      <td>GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163072</td>\n",
       "      <td>41163161</td>\n",
       "      <td>41163134</td>\n",
       "      <td>GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163059</td>\n",
       "      <td>41163148</td>\n",
       "      <td>41163086</td>\n",
       "      <td>ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163059</td>\n",
       "      <td>41163148</td>\n",
       "      <td>41163086</td>\n",
       "      <td>ATGAGCAGCCGGAGAGGCTCCATGCCGCCCTTCCCCAGCCAAGGGC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163059</td>\n",
       "      <td>41163148</td>\n",
       "      <td>41163134</td>\n",
       "      <td>ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163059</td>\n",
       "      <td>41163148</td>\n",
       "      <td>41163134</td>\n",
       "      <td>ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr strand Core_Prom_start Core_Prom_end VCF_position  \\\n",
       "0  chr6      -        41163072      41163161     41163086   \n",
       "1  chr6      -        41163072      41163161     41163086   \n",
       "2  chr6      -        41163072      41163161     41163134   \n",
       "3  chr6      -        41163072      41163161     41163134   \n",
       "4  chr6      -        41163059      41163148     41163086   \n",
       "5  chr6      -        41163059      41163148     41163086   \n",
       "6  chr6      -        41163059      41163148     41163134   \n",
       "7  chr6      -        41163059      41163148     41163134   \n",
       "\n",
       "                                                Data Label  \n",
       "0  GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...     1  \n",
       "1  GAGGCTCCATGCCGCCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...     0  \n",
       "2  GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...     1  \n",
       "3  GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...     0  \n",
       "4  ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGC...     1  \n",
       "5  ATGAGCAGCCGGAGAGGCTCCATGCCGCCCTTCCCCAGCCAAGGGC...     0  \n",
       "6  ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGC...     1  \n",
       "7  ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGC...     0  "
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "457c85b5-32ba-48a8-b802-0898043c82cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def seq2kmer(seq, k):\n",
    "    \"\"\"\n",
    "    Convert original sequence to kmers\n",
    "    \n",
    "    Arguments:\n",
    "    seq -- str, original sequence.\n",
    "    k -- int, kmer of length k specified.\n",
    "    \n",
    "    Returns:\n",
    "    kmers -- str, kmers separated by space\n",
    "    \"\"\"\n",
    "    kmer = [seq[x:x+k] for x in range(len(seq)+1-k)]\n",
    "    kmers = \" \".join(kmer)\n",
    "    return kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "a8d9a0d4-abc3-4eb7-837b-53a6e0964d1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>Core_Prom_start</th>\n",
       "      <th>Core_Prom_end</th>\n",
       "      <th>VCF_position</th>\n",
       "      <th>Data</th>\n",
       "      <th>Label</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163072</td>\n",
       "      <td>41163161</td>\n",
       "      <td>41163086</td>\n",
       "      <td>GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGGCT AGGCTC GGCTCC GCTCCA CTCCAT TCCATG CCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163072</td>\n",
       "      <td>41163161</td>\n",
       "      <td>41163086</td>\n",
       "      <td>GAGGCTCCATGCCGCCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGGCT AGGCTC GGCTCC GCTCCA CTCCAT TCCATG CCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163072</td>\n",
       "      <td>41163161</td>\n",
       "      <td>41163134</td>\n",
       "      <td>GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGGCT AGGCTC GGCTCC GCTCCA CTCCAT TCCATG CCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163072</td>\n",
       "      <td>41163161</td>\n",
       "      <td>41163134</td>\n",
       "      <td>GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGGCT AGGCTC GGCTCC GCTCCA CTCCAT TCCATG CCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163059</td>\n",
       "      <td>41163148</td>\n",
       "      <td>41163086</td>\n",
       "      <td>ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGC...</td>\n",
       "      <td>1</td>\n",
       "      <td>ATGAGC TGAGCA GAGCAG AGCAGC GCAGCC CAGCCG AGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163059</td>\n",
       "      <td>41163148</td>\n",
       "      <td>41163086</td>\n",
       "      <td>ATGAGCAGCCGGAGAGGCTCCATGCCGCCCTTCCCCAGCCAAGGGC...</td>\n",
       "      <td>0</td>\n",
       "      <td>ATGAGC TGAGCA GAGCAG AGCAGC GCAGCC CAGCCG AGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163059</td>\n",
       "      <td>41163148</td>\n",
       "      <td>41163134</td>\n",
       "      <td>ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGC...</td>\n",
       "      <td>1</td>\n",
       "      <td>ATGAGC TGAGCA GAGCAG AGCAGC GCAGCC CAGCCG AGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>41163059</td>\n",
       "      <td>41163148</td>\n",
       "      <td>41163134</td>\n",
       "      <td>ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGC...</td>\n",
       "      <td>0</td>\n",
       "      <td>ATGAGC TGAGCA GAGCAG AGCAGC GCAGCC CAGCCG AGCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr strand Core_Prom_start Core_Prom_end VCF_position  \\\n",
       "0  chr6      -        41163072      41163161     41163086   \n",
       "1  chr6      -        41163072      41163161     41163086   \n",
       "2  chr6      -        41163072      41163161     41163134   \n",
       "3  chr6      -        41163072      41163161     41163134   \n",
       "4  chr6      -        41163059      41163148     41163086   \n",
       "5  chr6      -        41163059      41163148     41163086   \n",
       "6  chr6      -        41163059      41163148     41163134   \n",
       "7  chr6      -        41163059      41163148     41163134   \n",
       "\n",
       "                                                Data Label  \\\n",
       "0  GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...     1   \n",
       "1  GAGGCTCCATGCCGCCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...     0   \n",
       "2  GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...     1   \n",
       "3  GAGGCTCCATGCCACCCTTCCCCAGCCAAGGGCAGAAGCAGAGTGC...     0   \n",
       "4  ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGC...     1   \n",
       "5  ATGAGCAGCCGGAGAGGCTCCATGCCGCCCTTCCCCAGCCAAGGGC...     0   \n",
       "6  ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGC...     1   \n",
       "7  ATGAGCAGCCGGAGAGGCTCCATGCCACCCTTCCCCAGCCAAGGGC...     0   \n",
       "\n",
       "                                            Sequence  \n",
       "0  GAGGCT AGGCTC GGCTCC GCTCCA CTCCAT TCCATG CCAT...  \n",
       "1  GAGGCT AGGCTC GGCTCC GCTCCA CTCCAT TCCATG CCAT...  \n",
       "2  GAGGCT AGGCTC GGCTCC GCTCCA CTCCAT TCCATG CCAT...  \n",
       "3  GAGGCT AGGCTC GGCTCC GCTCCA CTCCAT TCCATG CCAT...  \n",
       "4  ATGAGC TGAGCA GAGCAG AGCAGC GCAGCC CAGCCG AGCC...  \n",
       "5  ATGAGC TGAGCA GAGCAG AGCAGC GCAGCC CAGCCG AGCC...  \n",
       "6  ATGAGC TGAGCA GAGCAG AGCAGC GCAGCC CAGCCG AGCC...  \n",
       "7  ATGAGC TGAGCA GAGCAG AGCAGC GCAGCC CAGCCG AGCC...  "
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_data['Sequence'] = df_data['Data'].apply(seq2kmer, args=(6,))\n",
    "df_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "833bf6ba-17b3-48c7-9cd7-bf8ddfaba086",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_selected.to_csv(output_path+\"/selected_Donor_splice_sites.tsv\", sep=\"\\t\", index=False)\n",
    "new_df.to_csv(output_path+\"/intermediate.tsv\",  sep=\"\\t\", index=False)\n",
    "df_data[['Sequence','Label']].to_csv(output_path+\"/dev.tsv\" , sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69f69d2e-e235-444f-87cb-9848d06dc2f8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
