{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4b34682f-35e8-4b7b-9fda-8a8893280317",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, tarfile\n",
    "import pandas as pd\n",
    "import pybedtools\n",
    "pybedtools.helpers.set_tempdir('/data/projects/temp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0811431e-8d96-4098-b8b3-48a15bab1eb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_path= \"/data/projects/Resources/Gencode_genome_annotation/\"\n",
    "gtf_annotation_filename= \"gencode.v40.annotation.gtf\"\n",
    "gencode_gtf_file_path= gencode_path+ gtf_annotation_filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7fcbf04e-d140-4025-95a2-995e3c353911",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the column names for a GTF file\n",
    "col_names = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "\n",
    "# Read the GTF file, skipping lines that start with '#'\n",
    "df = pd.read_csv(gencode_gtf_file_path, sep='\\t', comment='#', names=col_names, header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cce4006a-1944-4f3e-ae5c-109c00409686",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; gene_type \"transc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283857</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210195.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283858</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210195.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283859</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; gene_type \"Mt_tRN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283860</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283861</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3283862 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname   source     feature  start    end score strand frame  \\\n",
       "0          chr1   HAVANA        gene  11869  14409     .      +     .   \n",
       "1          chr1   HAVANA  transcript  11869  14409     .      +     .   \n",
       "2          chr1   HAVANA        exon  11869  12227     .      +     .   \n",
       "3          chr1   HAVANA        exon  12613  12721     .      +     .   \n",
       "4          chr1   HAVANA        exon  13221  14409     .      +     .   \n",
       "...         ...      ...         ...    ...    ...   ...    ...   ...   \n",
       "3283857    chrM  ENSEMBL  transcript  15888  15953     .      +     .   \n",
       "3283858    chrM  ENSEMBL        exon  15888  15953     .      +     .   \n",
       "3283859    chrM  ENSEMBL        gene  15956  16023     .      -     .   \n",
       "3283860    chrM  ENSEMBL  transcript  15956  16023     .      -     .   \n",
       "3283861    chrM  ENSEMBL        exon  15956  16023     .      -     .   \n",
       "\n",
       "                                                 attribute  \n",
       "0        gene_id \"ENSG00000223972.5\"; gene_type \"transc...  \n",
       "1        gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  \n",
       "2        gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  \n",
       "3        gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  \n",
       "4        gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  \n",
       "...                                                    ...  \n",
       "3283857  gene_id \"ENSG00000210195.2\"; transcript_id \"EN...  \n",
       "3283858  gene_id \"ENSG00000210195.2\"; transcript_id \"EN...  \n",
       "3283859  gene_id \"ENSG00000210196.2\"; gene_type \"Mt_tRN...  \n",
       "3283860  gene_id \"ENSG00000210196.2\"; transcript_id \"EN...  \n",
       "3283861  gene_id \"ENSG00000210196.2\"; transcript_id \"EN...  \n",
       "\n",
       "[3283862 rows x 9 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5f0f39ab-ff7c-4025-b355-8b6eb51d1910",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[['seqname', 'start', 'end', 'source', 'feature','strand', 'frame', 'attribute']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "620fc960-5832-4b34-8420-eed37384e970",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; gene_type \"transc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; gene_type \"unproc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000278267.1\"; gene_type \"miRNA\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000243485.5\"; gene_type \"lncRNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284332.1\"; gene_type \"miRNA\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283843</th>\n",
       "      <td>chrM</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198695.2\"; gene_type \"protei...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283848</th>\n",
       "      <td>chrM</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210194.1\"; gene_type \"Mt_tRN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283851</th>\n",
       "      <td>chrM</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198727.2\"; gene_type \"protei...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283856</th>\n",
       "      <td>chrM</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210195.2\"; gene_type \"Mt_tRN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283859</th>\n",
       "      <td>chrM</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; gene_type \"Mt_tRN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61544 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname  start    end   source feature strand frame  \\\n",
       "0          chr1  11869  14409   HAVANA    gene      +     .   \n",
       "12         chr1  14404  29570   HAVANA    gene      -     .   \n",
       "25         chr1  17369  17436  ENSEMBL    gene      -     .   \n",
       "28         chr1  29554  31109   HAVANA    gene      +     .   \n",
       "36         chr1  30366  30503  ENSEMBL    gene      +     .   \n",
       "...         ...    ...    ...      ...     ...    ...   ...   \n",
       "3283843    chrM  14149  14673  ENSEMBL    gene      -     .   \n",
       "3283848    chrM  14674  14742  ENSEMBL    gene      -     .   \n",
       "3283851    chrM  14747  15887  ENSEMBL    gene      +     .   \n",
       "3283856    chrM  15888  15953  ENSEMBL    gene      +     .   \n",
       "3283859    chrM  15956  16023  ENSEMBL    gene      -     .   \n",
       "\n",
       "                                                 attribute  \n",
       "0        gene_id \"ENSG00000223972.5\"; gene_type \"transc...  \n",
       "12       gene_id \"ENSG00000227232.5\"; gene_type \"unproc...  \n",
       "25       gene_id \"ENSG00000278267.1\"; gene_type \"miRNA\"...  \n",
       "28       gene_id \"ENSG00000243485.5\"; gene_type \"lncRNA...  \n",
       "36       gene_id \"ENSG00000284332.1\"; gene_type \"miRNA\"...  \n",
       "...                                                    ...  \n",
       "3283843  gene_id \"ENSG00000198695.2\"; gene_type \"protei...  \n",
       "3283848  gene_id \"ENSG00000210194.1\"; gene_type \"Mt_tRN...  \n",
       "3283851  gene_id \"ENSG00000198727.2\"; gene_type \"protei...  \n",
       "3283856  gene_id \"ENSG00000210195.2\"; gene_type \"Mt_tRN...  \n",
       "3283859  gene_id \"ENSG00000210196.2\"; gene_type \"Mt_tRN...  \n",
       "\n",
       "[61544 rows x 8 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene = df[df['feature']==\"gene\"]\n",
    "df_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3b798473-b445-4617-a03f-d530f9120c52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Column</th>\n",
       "      <th>Max Value</th>\n",
       "      <th>Non-NaN Count</th>\n",
       "      <th>Non-NaN Indices</th>\n",
       "      <th>Percentage</th>\n",
       "      <th>Variant Region Difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr16_58267438_58267517_58267455_58267458</td>\n",
       "      <td>36.161531</td>\n",
       "      <td>32</td>\n",
       "      <td>[6, 7, 25, 32, 49, 63, 65, 80, 97, 115, 128, 1...</td>\n",
       "      <td>10.958904</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10_73797336_73797415_73797341_73797342</td>\n",
       "      <td>35.699415</td>\n",
       "      <td>3</td>\n",
       "      <td>[17, 219, 266]</td>\n",
       "      <td>1.027397</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr12_23190232_23190311_23190241_23190245</td>\n",
       "      <td>34.806988</td>\n",
       "      <td>4</td>\n",
       "      <td>[26, 163, 196, 290]</td>\n",
       "      <td>1.369863</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr4_52750790_52750869_52750817_52750818</td>\n",
       "      <td>34.397403</td>\n",
       "      <td>2</td>\n",
       "      <td>[76, 240]</td>\n",
       "      <td>0.684932</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3_175460004_175460083_175460009_175460012</td>\n",
       "      <td>34.386584</td>\n",
       "      <td>2</td>\n",
       "      <td>[122, 257]</td>\n",
       "      <td>0.684932</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3780</th>\n",
       "      <td>chr1_209623686_209623765_209623748_209623749</td>\n",
       "      <td>5.003059</td>\n",
       "      <td>6</td>\n",
       "      <td>[73, 108, 140, 156, 167, 206]</td>\n",
       "      <td>2.054795</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3781</th>\n",
       "      <td>chr6_29828502_29828581_29828528_29828529</td>\n",
       "      <td>5.002682</td>\n",
       "      <td>193</td>\n",
       "      <td>[0, 1, 2, 3, 6, 7, 10, 11, 13, 14, 15, 16, 17,...</td>\n",
       "      <td>66.095890</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3782</th>\n",
       "      <td>chr19_35259199_35259278_35259204_35259205</td>\n",
       "      <td>5.002419</td>\n",
       "      <td>1</td>\n",
       "      <td>[44]</td>\n",
       "      <td>0.342466</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3783</th>\n",
       "      <td>chr13_110492029_110492108_110492064_110492065</td>\n",
       "      <td>5.000697</td>\n",
       "      <td>22</td>\n",
       "      <td>[43, 52, 53, 64, 66, 72, 98, 106, 115, 118, 13...</td>\n",
       "      <td>7.534247</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3784</th>\n",
       "      <td>chr7_899451_899530_899526_899527</td>\n",
       "      <td>5.000186</td>\n",
       "      <td>1</td>\n",
       "      <td>[10]</td>\n",
       "      <td>0.342466</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3785 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Column  Max Value  Non-NaN Count  \\\n",
       "0         chr16_58267438_58267517_58267455_58267458  36.161531             32   \n",
       "1         chr10_73797336_73797415_73797341_73797342  35.699415              3   \n",
       "2         chr12_23190232_23190311_23190241_23190245  34.806988              4   \n",
       "3          chr4_52750790_52750869_52750817_52750818  34.397403              2   \n",
       "4      chr3_175460004_175460083_175460009_175460012  34.386584              2   \n",
       "...                                             ...        ...            ...   \n",
       "3780   chr1_209623686_209623765_209623748_209623749   5.003059              6   \n",
       "3781       chr6_29828502_29828581_29828528_29828529   5.002682            193   \n",
       "3782      chr19_35259199_35259278_35259204_35259205   5.002419              1   \n",
       "3783  chr13_110492029_110492108_110492064_110492065   5.000697             22   \n",
       "3784               chr7_899451_899530_899526_899527   5.000186              1   \n",
       "\n",
       "                                        Non-NaN Indices  Percentage  \\\n",
       "0     [6, 7, 25, 32, 49, 63, 65, 80, 97, 115, 128, 1...   10.958904   \n",
       "1                                        [17, 219, 266]    1.027397   \n",
       "2                                   [26, 163, 196, 290]    1.369863   \n",
       "3                                             [76, 240]    0.684932   \n",
       "4                                            [122, 257]    0.684932   \n",
       "...                                                 ...         ...   \n",
       "3780                      [73, 108, 140, 156, 167, 206]    2.054795   \n",
       "3781  [0, 1, 2, 3, 6, 7, 10, 11, 13, 14, 15, 16, 17,...   66.095890   \n",
       "3782                                               [44]    0.342466   \n",
       "3783  [43, 52, 53, 64, 66, 72, 98, 106, 115, 118, 13...    7.534247   \n",
       "3784                                               [10]    0.342466   \n",
       "\n",
       "      Variant Region Difference  \n",
       "0                             3  \n",
       "1                             1  \n",
       "2                             4  \n",
       "3                             1  \n",
       "4                             3  \n",
       "...                         ...  \n",
       "3780                          1  \n",
       "3781                          1  \n",
       "3782                          1  \n",
       "3783                          1  \n",
       "3784                          1  \n",
       "\n",
       "[3785 rows x 6 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_snp= pd.read_csv(\"/data/projects/GDC_Cancer_Wise/Brain/Data/Postprocessing_result/Acceptor/important_snp.tsv\", sep= \"\\t\")\n",
    "df_snp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9d48cf4f-fda6-4c7c-aa45-3103f33587fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "split_columns = df_snp.iloc[:, 0].str.split('_', expand=True)\n",
    "\n",
    "# Create a new DataFrame with the required columns\n",
    "new_df = pd.DataFrame({\n",
    "    'chr': split_columns[0],\n",
    "    'acceptor_start': split_columns[1],\n",
    "    'acceptor_end': split_columns[2]\n",
    "})\n",
    "\n",
    "# Convert 'start' and 'end' columns to numeric type\n",
    "new_df['acceptor_start'] = pd.to_numeric(new_df['acceptor_start'])\n",
    "new_df['acceptor_end'] = pd.to_numeric(new_df['acceptor_end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "106402c4-4efd-4ea7-b6ee-8049a566424d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>acceptor_start</th>\n",
       "      <th>acceptor_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr16</td>\n",
       "      <td>58267438</td>\n",
       "      <td>58267517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>73797336</td>\n",
       "      <td>73797415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr12</td>\n",
       "      <td>23190232</td>\n",
       "      <td>23190311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr4</td>\n",
       "      <td>52750790</td>\n",
       "      <td>52750869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3</td>\n",
       "      <td>175460004</td>\n",
       "      <td>175460083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3780</th>\n",
       "      <td>chr1</td>\n",
       "      <td>209623686</td>\n",
       "      <td>209623765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3781</th>\n",
       "      <td>chr6</td>\n",
       "      <td>29828502</td>\n",
       "      <td>29828581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3782</th>\n",
       "      <td>chr19</td>\n",
       "      <td>35259199</td>\n",
       "      <td>35259278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3783</th>\n",
       "      <td>chr13</td>\n",
       "      <td>110492029</td>\n",
       "      <td>110492108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3784</th>\n",
       "      <td>chr7</td>\n",
       "      <td>899451</td>\n",
       "      <td>899530</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3785 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr  acceptor_start  acceptor_end\n",
       "0     chr16        58267438      58267517\n",
       "1     chr10        73797336      73797415\n",
       "2     chr12        23190232      23190311\n",
       "3      chr4        52750790      52750869\n",
       "4      chr3       175460004     175460083\n",
       "...     ...             ...           ...\n",
       "3780   chr1       209623686     209623765\n",
       "3781   chr6        29828502      29828581\n",
       "3782  chr19        35259199      35259278\n",
       "3783  chr13       110492029     110492108\n",
       "3784   chr7          899451        899530\n",
       "\n",
       "[3785 rows x 3 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_acceptor = new_df.drop_duplicates().reset_index(drop=True)\n",
    "df_acceptor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c11f7d55-a480-480e-b65d-6b36a39b6513",
   "metadata": {},
   "outputs": [],
   "source": [
    "acceptor_bed = pybedtools.BedTool.from_dataframe(df_acceptor)\n",
    "gene_bed = pybedtools.BedTool.from_dataframe(df_gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a8c6651b-8ff0-4ffd-ac44-868812ce5056",
   "metadata": {},
   "outputs": [],
   "source": [
    "intersect_gene_acceptor = gene_bed.intersect(acceptor_bed, wa=True, wb=True)\n",
    "column_list  = df_gene.columns.to_list() + df_acceptor.columns.to_list()\n",
    "    #print(column_list)\n",
    "df_intersection = intersect_gene_acceptor.to_dataframe(names=column_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "4e0f9982-6c3b-43bb-834c-5d6e311c9610",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>chr</th>\n",
       "      <th>acceptor_start</th>\n",
       "      <th>acceptor_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>586071</td>\n",
       "      <td>827796</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000230021.10\"; gene_type \"trans...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>807176</td>\n",
       "      <td>807255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>586071</td>\n",
       "      <td>827796</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000230021.10\"; gene_type \"trans...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>809581</td>\n",
       "      <td>809660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>778747</td>\n",
       "      <td>810065</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237491.10\"; gene_type \"lncRN...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>807176</td>\n",
       "      <td>807255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>778747</td>\n",
       "      <td>810065</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237491.10\"; gene_type \"lncRN...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>809581</td>\n",
       "      <td>809660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>800879</td>\n",
       "      <td>817712</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000230092.7\"; gene_type \"transc...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>807176</td>\n",
       "      <td>807255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4842</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155137111</td>\n",
       "      <td>155156503</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000235991.4\"; gene_type \"unproc...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155156406</td>\n",
       "      <td>155156485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4843</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156025664</td>\n",
       "      <td>156027877</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227159.8\"; gene_type \"unproc...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>156027043</td>\n",
       "      <td>156027122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4844</th>\n",
       "      <td>chrY</td>\n",
       "      <td>6357219</td>\n",
       "      <td>6369921</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000229643.2\"; gene_type \"lncRNA...</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6369337</td>\n",
       "      <td>6369416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4845</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13153931</td>\n",
       "      <td>13162214</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000235462.1\"; gene_type \"transc...</td>\n",
       "      <td>chrY</td>\n",
       "      <td>13153985</td>\n",
       "      <td>13154064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4846</th>\n",
       "      <td>chrY</td>\n",
       "      <td>13234577</td>\n",
       "      <td>13480673</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000183878.16\"; gene_type \"prote...</td>\n",
       "      <td>chrY</td>\n",
       "      <td>13306211</td>\n",
       "      <td>13306290</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4847 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     seqname      start        end  source feature strand frame  \\\n",
       "0       chr1     586071     827796  HAVANA    gene      -     .   \n",
       "1       chr1     586071     827796  HAVANA    gene      -     .   \n",
       "2       chr1     778747     810065  HAVANA    gene      +     .   \n",
       "3       chr1     778747     810065  HAVANA    gene      +     .   \n",
       "4       chr1     800879     817712  HAVANA    gene      -     .   \n",
       "...      ...        ...        ...     ...     ...    ...   ...   \n",
       "4842    chrX  155137111  155156503  HAVANA    gene      +     .   \n",
       "4843    chrX  156025664  156027877  HAVANA    gene      -     .   \n",
       "4844    chrY    6357219    6369921  HAVANA    gene      -     .   \n",
       "4845    chrY   13153931   13162214  HAVANA    gene      -     .   \n",
       "4846    chrY   13234577   13480673  HAVANA    gene      -     .   \n",
       "\n",
       "                                              attribute   chr  acceptor_start  \\\n",
       "0     gene_id \"ENSG00000230021.10\"; gene_type \"trans...  chr1          807176   \n",
       "1     gene_id \"ENSG00000230021.10\"; gene_type \"trans...  chr1          809581   \n",
       "2     gene_id \"ENSG00000237491.10\"; gene_type \"lncRN...  chr1          807176   \n",
       "3     gene_id \"ENSG00000237491.10\"; gene_type \"lncRN...  chr1          809581   \n",
       "4     gene_id \"ENSG00000230092.7\"; gene_type \"transc...  chr1          807176   \n",
       "...                                                 ...   ...             ...   \n",
       "4842  gene_id \"ENSG00000235991.4\"; gene_type \"unproc...  chrX       155156406   \n",
       "4843  gene_id \"ENSG00000227159.8\"; gene_type \"unproc...  chrX       156027043   \n",
       "4844  gene_id \"ENSG00000229643.2\"; gene_type \"lncRNA...  chrY         6369337   \n",
       "4845  gene_id \"ENSG00000235462.1\"; gene_type \"transc...  chrY        13153985   \n",
       "4846  gene_id \"ENSG00000183878.16\"; gene_type \"prote...  chrY        13306211   \n",
       "\n",
       "      acceptor_end  \n",
       "0           807255  \n",
       "1           809660  \n",
       "2           807255  \n",
       "3           809660  \n",
       "4           807255  \n",
       "...            ...  \n",
       "4842     155156485  \n",
       "4843     156027122  \n",
       "4844       6369416  \n",
       "4845      13154064  \n",
       "4846      13306290  \n",
       "\n",
       "[4847 rows x 11 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_intersection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce02780e-594c-47ab-9ba8-6ec0a0218740",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
