{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cdd0a298-de34-4915-8bae-192d11cb06ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:numexpr.utils:Note: NumExpr detected 64 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n"
     ]
    }
   ],
   "source": [
    "import os, tarfile\n",
    "import pyBigWig\n",
    "import pybedtools\n",
    "import zipfile, gzip\n",
    "from gtfparse import read_gtf\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f8cf0b5f-e0cb-4ac6-b6d0-aee02c4966a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_file_path = \"/data/projects/Resources/Gencode_genome_annotation/gencode.v40.annotation.gtf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8520ea38-2cb3-4d98-bd36-8f8a77f59bd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_attributes(attribute_string):\n",
    "    # Split each attribute into key and value, then strip quotes and spaces\n",
    "    return dict(item.strip().replace('\"', '').split(' ') for item in attribute_string if item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "de06ed69-923c-4f70-8141-686c0008151f",
   "metadata": {},
   "outputs": [],
   "source": [
    "column_names = [\n",
    "    \"seqname\", \"source\", \"feature\", \"start\", \"end\",\n",
    "    \"score\", \"strand\", \"frame\", \"attribute\"\n",
    "]\n",
    "\n",
    "# Read the GTF file\n",
    "gtf_df = pd.read_csv(gtf_file_path, sep=\"\\t\", comment='#', header=None, names=column_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c42d403f-4324-4b16-b24b-c69aba6bf33c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 1: Split the 'attribute' string into a list of strings for each key-value pair\n",
    "attributes_list = gtf_df['attribute'].str.split(';')\n",
    "\n",
    "\n",
    "# Apply the function to each row's attribute list\n",
    "attributes_dicts = attributes_list.apply(parse_attributes)\n",
    "\n",
    "# Step 3: Convert the list of dictionaries into a DataFrame\n",
    "attributes_df = pd.DataFrame(list(attributes_dicts))\n",
    "\n",
    "# Step 4: Combine the new attributes DataFrame with the original gtf_df\n",
    "# This step assumes that the indexes are aligned and can be directly concatenated\n",
    "combined_df = pd.concat([gtf_df, attributes_df], axis=1)\n",
    "\n",
    "# Optionally, you can drop the original 'attribute' column if it's no longer needed\n",
    "combined_df = combined_df.drop('attribute', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8d716ea9-00ee-4cc9-a821-43e4152abb5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1;</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1;</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1;</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283857</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7499</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000387460.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>Ensembl_canonical;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283858</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7499</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000387460.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>Ensembl_canonical;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001544475.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283859</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7494;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283860</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7494</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000387461.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>Ensembl_canonical;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283861</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7494</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000387461.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>Ensembl_canonical;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001544473.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3283862 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname   source     feature  start    end score strand frame  \\\n",
       "0          chr1   HAVANA        gene  11869  14409     .      +     .   \n",
       "1          chr1   HAVANA  transcript  11869  14409     .      +     .   \n",
       "2          chr1   HAVANA        exon  11869  12227     .      +     .   \n",
       "3          chr1   HAVANA        exon  12613  12721     .      +     .   \n",
       "4          chr1   HAVANA        exon  13221  14409     .      +     .   \n",
       "...         ...      ...         ...    ...    ...   ...    ...   ...   \n",
       "3283857    chrM  ENSEMBL  transcript  15888  15953     .      +     .   \n",
       "3283858    chrM  ENSEMBL        exon  15888  15953     .      +     .   \n",
       "3283859    chrM  ENSEMBL        gene  15956  16023     .      -     .   \n",
       "3283860    chrM  ENSEMBL  transcript  15956  16023     .      -     .   \n",
       "3283861    chrM  ENSEMBL        exon  15956  16023     .      -     .   \n",
       "\n",
       "                   gene_id                           gene_type gene_name  \\\n",
       "0        ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1   \n",
       "1        ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1   \n",
       "2        ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1   \n",
       "3        ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1   \n",
       "4        ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1   \n",
       "...                    ...                                 ...       ...   \n",
       "3283857  ENSG00000210195.2                             Mt_tRNA     MT-TT   \n",
       "3283858  ENSG00000210195.2                             Mt_tRNA     MT-TT   \n",
       "3283859  ENSG00000210196.2                             Mt_tRNA     MT-TP   \n",
       "3283860  ENSG00000210196.2                             Mt_tRNA     MT-TP   \n",
       "3283861  ENSG00000210196.2                             Mt_tRNA     MT-TP   \n",
       "\n",
       "        level     hgnc_id            havana_gene      transcript_id  \\\n",
       "0           2  HGNC:37102  OTTHUMG00000000961.2;                NaN   \n",
       "1           2  HGNC:37102   OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "2           2  HGNC:37102   OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "3           2  HGNC:37102   OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "4           2  HGNC:37102   OTTHUMG00000000961.2  ENST00000456328.2   \n",
       "...       ...         ...                    ...                ...   \n",
       "3283857     3   HGNC:7499                    NaN  ENST00000387460.2   \n",
       "3283858     3   HGNC:7499                    NaN  ENST00000387460.2   \n",
       "3283859     3  HGNC:7494;                    NaN                NaN   \n",
       "3283860     3   HGNC:7494                    NaN  ENST00000387461.2   \n",
       "3283861     3   HGNC:7494                    NaN  ENST00000387461.2   \n",
       "\n",
       "              transcript_type transcript_name transcript_support_level  \\\n",
       "0                         NaN             NaN                      NaN   \n",
       "1        processed_transcript     DDX11L1-202                        1   \n",
       "2        processed_transcript     DDX11L1-202                        1   \n",
       "3        processed_transcript     DDX11L1-202                        1   \n",
       "4        processed_transcript     DDX11L1-202                        1   \n",
       "...                       ...             ...                      ...   \n",
       "3283857               Mt_tRNA       MT-TT-201                       NA   \n",
       "3283858               Mt_tRNA       MT-TT-201                       NA   \n",
       "3283859                   NaN             NaN                      NaN   \n",
       "3283860               Mt_tRNA       MT-TP-201                       NA   \n",
       "3283861               Mt_tRNA       MT-TP-201                       NA   \n",
       "\n",
       "                        tag      havana_transcript exon_number  \\\n",
       "0                       NaN                    NaN         NaN   \n",
       "1                     basic  OTTHUMT00000362751.1;         NaN   \n",
       "2                     basic  OTTHUMT00000362751.1;           1   \n",
       "3                     basic  OTTHUMT00000362751.1;           2   \n",
       "4                     basic  OTTHUMT00000362751.1;           3   \n",
       "...                     ...                    ...         ...   \n",
       "3283857  Ensembl_canonical;                    NaN         NaN   \n",
       "3283858  Ensembl_canonical;                    NaN           1   \n",
       "3283859                 NaN                    NaN         NaN   \n",
       "3283860  Ensembl_canonical;                    NaN         NaN   \n",
       "3283861  Ensembl_canonical;                    NaN           1   \n",
       "\n",
       "                   exon_id  ont protein_id ccdsid  \n",
       "0                      NaN  NaN        NaN    NaN  \n",
       "1                      NaN  NaN        NaN    NaN  \n",
       "2        ENSE00002234944.1  NaN        NaN    NaN  \n",
       "3        ENSE00003582793.1  NaN        NaN    NaN  \n",
       "4        ENSE00002312635.1  NaN        NaN    NaN  \n",
       "...                    ...  ...        ...    ...  \n",
       "3283857                NaN  NaN        NaN    NaN  \n",
       "3283858  ENSE00001544475.2  NaN        NaN    NaN  \n",
       "3283859                NaN  NaN        NaN    NaN  \n",
       "3283860                NaN  NaN        NaN    NaN  \n",
       "3283861  ENSE00001544473.2  NaN        NaN    NaN  \n",
       "\n",
       "[3283862 rows x 25 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "edd0e18a-cfdc-4b88-b467-17907cf847bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_type\n",
       "IG_C_gene                                 281\n",
       "IG_C_pseudogene                            33\n",
       "IG_D_gene                                 152\n",
       "IG_J_gene                                  76\n",
       "IG_J_pseudogene                             9\n",
       "IG_V_gene                                1134\n",
       "IG_V_pseudogene                           661\n",
       "IG_pseudogene                               3\n",
       "Mt_rRNA                                     6\n",
       "Mt_tRNA                                    66\n",
       "TEC                                      3221\n",
       "TR_C_gene                                  70\n",
       "TR_D_gene                                  16\n",
       "TR_J_gene                                 316\n",
       "TR_J_pseudogene                            12\n",
       "TR_V_gene                                 823\n",
       "TR_V_pseudogene                           123\n",
       "lncRNA                                 258393\n",
       "miRNA                                    5637\n",
       "misc_RNA                                 6636\n",
       "polymorphic_pseudogene                   2011\n",
       "processed_pseudogene                    32087\n",
       "protein_coding                        2909967\n",
       "pseudogene                                 64\n",
       "rRNA                                      141\n",
       "rRNA_pseudogene                          1491\n",
       "ribozyme                                   24\n",
       "sRNA                                       15\n",
       "scRNA                                       3\n",
       "scaRNA                                    147\n",
       "snRNA                                    5703\n",
       "snoRNA                                   2829\n",
       "transcribed_processed_pseudogene         4597\n",
       "transcribed_unitary_pseudogene           6146\n",
       "transcribed_unprocessed_pseudogene      27286\n",
       "translated_processed_pseudogene             8\n",
       "translated_unprocessed_pseudogene          44\n",
       "unitary_pseudogene                        548\n",
       "unprocessed_pseudogene                  13080\n",
       "vault_RNA                                   3\n",
       "dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df.groupby(\"gene_type\").size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9c58a4a7-3fe6-4410-95a2-1074fdcd76cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df.to_csv(\"/data/projects/Resources/Gencode_genome_annotation/gencode_annotation_V40.tsv\", sep= \"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b8d86e06-cf45-499a-9fd9-2e2a8178050b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PGO:0000019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002839.1;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:50039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>Ensembl_canonical;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29554</td>\n",
       "      <td>31097</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:52482</td>\n",
       "      <td>OTTHUMG00000000959.2</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG-202</td>\n",
       "      <td>5</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002840.1;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283698</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57209306</td>\n",
       "      <td>57210051</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000483079.6_PAR_Y</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-210</td>\n",
       "      <td>1</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058833.1;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283701</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57209887</td>\n",
       "      <td>57212186</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000496301.6_PAR_Y</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-215</td>\n",
       "      <td>2</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058827.1;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283704</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57210344</td>\n",
       "      <td>57212074</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000483286.6_PAR_Y</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>WASH6P-211</td>\n",
       "      <td>1</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058834.1;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283708</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57210591</td>\n",
       "      <td>57212074</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:31685</td>\n",
       "      <td>OTTHUMG00000022677.5</td>\n",
       "      <td>ENST00000464205.6_PAR_Y</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>WASH6P-205</td>\n",
       "      <td>2</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058835.1;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283713</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57212184</td>\n",
       "      <td>57214397</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>ENST00000507418.6_PAR_Y</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>DDX11L16-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>PAR</td>\n",
       "      <td>OTTHUMT00000058841.1;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>246587 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname   source     feature     start       end score strand frame  \\\n",
       "1          chr1   HAVANA  transcript     11869     14409     .      +     .   \n",
       "5          chr1   HAVANA  transcript     12010     13670     .      +     .   \n",
       "13         chr1   HAVANA  transcript     14404     29570     .      -     .   \n",
       "26         chr1  ENSEMBL  transcript     17369     17436     .      -     .   \n",
       "29         chr1   HAVANA  transcript     29554     31097     .      +     .   \n",
       "...         ...      ...         ...       ...       ...   ...    ...   ...   \n",
       "3283698    chrY   HAVANA  transcript  57209306  57210051     .      +     .   \n",
       "3283701    chrY   HAVANA  transcript  57209887  57212186     .      +     .   \n",
       "3283704    chrY   HAVANA  transcript  57210344  57212074     .      +     .   \n",
       "3283708    chrY   HAVANA  transcript  57210591  57212074     .      +     .   \n",
       "3283713    chrY   HAVANA  transcript  57212184  57214397     .      -     .   \n",
       "\n",
       "                          gene_id                           gene_type  \\\n",
       "1               ENSG00000223972.5  transcribed_unprocessed_pseudogene   \n",
       "5               ENSG00000223972.5  transcribed_unprocessed_pseudogene   \n",
       "13              ENSG00000227232.5              unprocessed_pseudogene   \n",
       "26              ENSG00000278267.1                               miRNA   \n",
       "29              ENSG00000243485.5                              lncRNA   \n",
       "...                           ...                                 ...   \n",
       "3283698  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "3283701  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "3283704  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "3283708  ENSG00000182484.15_PAR_Y                      protein_coding   \n",
       "3283713   ENSG00000227159.8_PAR_Y              unprocessed_pseudogene   \n",
       "\n",
       "           gene_name level     hgnc_id           havana_gene  \\\n",
       "1            DDX11L1     2  HGNC:37102  OTTHUMG00000000961.2   \n",
       "5            DDX11L1     2  HGNC:37102  OTTHUMG00000000961.2   \n",
       "13            WASH7P     2  HGNC:38034  OTTHUMG00000000958.1   \n",
       "26         MIR6859-1     3  HGNC:50039                   NaN   \n",
       "29       MIR1302-2HG     2  HGNC:52482  OTTHUMG00000000959.2   \n",
       "...              ...   ...         ...                   ...   \n",
       "3283698       WASH6P     2  HGNC:31685  OTTHUMG00000022677.5   \n",
       "3283701       WASH6P     2  HGNC:31685  OTTHUMG00000022677.5   \n",
       "3283704       WASH6P     2  HGNC:31685  OTTHUMG00000022677.5   \n",
       "3283708       WASH6P     2  HGNC:31685  OTTHUMG00000022677.5   \n",
       "3283713     DDX11L16     2  HGNC:37115  OTTHUMG00000022678.1   \n",
       "\n",
       "                   transcript_id                     transcript_type  \\\n",
       "1              ENST00000456328.2                processed_transcript   \n",
       "5              ENST00000450305.2  transcribed_unprocessed_pseudogene   \n",
       "13             ENST00000488147.1              unprocessed_pseudogene   \n",
       "26             ENST00000619216.1                               miRNA   \n",
       "29             ENST00000473358.1                              lncRNA   \n",
       "...                          ...                                 ...   \n",
       "3283698  ENST00000483079.6_PAR_Y                     retained_intron   \n",
       "3283701  ENST00000496301.6_PAR_Y                     retained_intron   \n",
       "3283704  ENST00000483286.6_PAR_Y                     retained_intron   \n",
       "3283708  ENST00000464205.6_PAR_Y                processed_transcript   \n",
       "3283713  ENST00000507418.6_PAR_Y              unprocessed_pseudogene   \n",
       "\n",
       "         transcript_name transcript_support_level                 tag  \\\n",
       "1            DDX11L1-202                        1               basic   \n",
       "5            DDX11L1-201                       NA   Ensembl_canonical   \n",
       "13            WASH7P-201                       NA   Ensembl_canonical   \n",
       "26         MIR6859-1-201                       NA  Ensembl_canonical;   \n",
       "29       MIR1302-2HG-202                        5   Ensembl_canonical   \n",
       "...                  ...                      ...                 ...   \n",
       "3283698       WASH6P-210                        1                 PAR   \n",
       "3283701       WASH6P-215                        2                 PAR   \n",
       "3283704       WASH6P-211                        1                 PAR   \n",
       "3283708       WASH6P-205                        2                 PAR   \n",
       "3283713     DDX11L16-201                       NA                 PAR   \n",
       "\n",
       "             havana_transcript exon_number exon_id          ont protein_id  \\\n",
       "1        OTTHUMT00000362751.1;         NaN     NaN          NaN        NaN   \n",
       "5        OTTHUMT00000002844.2;         NaN     NaN  PGO:0000019        NaN   \n",
       "13       OTTHUMT00000002839.1;         NaN     NaN  PGO:0000005        NaN   \n",
       "26                         NaN         NaN     NaN          NaN        NaN   \n",
       "29       OTTHUMT00000002840.1;         NaN     NaN          NaN        NaN   \n",
       "...                        ...         ...     ...          ...        ...   \n",
       "3283698  OTTHUMT00000058833.1;         NaN     NaN          NaN        NaN   \n",
       "3283701  OTTHUMT00000058827.1;         NaN     NaN          NaN        NaN   \n",
       "3283704  OTTHUMT00000058834.1;         NaN     NaN          NaN        NaN   \n",
       "3283708  OTTHUMT00000058835.1;         NaN     NaN          NaN        NaN   \n",
       "3283713  OTTHUMT00000058841.1;         NaN     NaN  PGO:0000005        NaN   \n",
       "\n",
       "        ccdsid  \n",
       "1          NaN  \n",
       "5          NaN  \n",
       "13         NaN  \n",
       "26         NaN  \n",
       "29         NaN  \n",
       "...        ...  \n",
       "3283698    NaN  \n",
       "3283701    NaN  \n",
       "3283704    NaN  \n",
       "3283708    NaN  \n",
       "3283713    NaN  \n",
       "\n",
       "[246587 rows x 25 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter rows where 'feature' is 'transcript' and 'seqname' is not 'chrM'\n",
    "gtf_transcript = combined_df[(combined_df[\"feature\"] == \"transcript\") & (combined_df[\"seqname\"] != \"chrM\")]\n",
    "\n",
    "# Display the resulting DataFrame\n",
    "gtf_transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "11332d68-96ff-4430-acd2-099cd9ca767c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_tss_regions_from_df(tss_df):\n",
    "    \"\"\"\n",
    "    Extracts TSS (+50 and -50 bp) regions from the GTF DataFrame.\n",
    "    :param df: A pandas DataFrame containing GTF data.\n",
    "    :return: A DataFrame with TSS regions.\n",
    "    \"\"\"\n",
    "\n",
    "    # Calculate TSS based on strand\n",
    "    tss_df['TSS'] = tss_df.apply(lambda x: x['start'] if x['strand'] == '+' else x['end'], axis=1)\n",
    "\n",
    "    # Adjust TSS based on strand for +50 and -50 bp\n",
    "    tss_df['start_adj'] = tss_df.apply(lambda x: x['TSS'] - 50 if x['strand'] == '+' else x['TSS'], axis=1)\n",
    "    tss_df['end_adj'] = tss_df.apply(lambda x: x['TSS'] + 50 if x['strand'] == '+' else x['TSS'] + 100, axis=1)\n",
    "\n",
    "    # Select and rename relevant columns\n",
    "    tss_regions = tss_df[['seqname', 'start_adj', 'end_adj', 'strand', 'gene_id', 'gene_type', 'gene_name', 'hgnc_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name']]\n",
    "    tss_regions.columns = ['Chromosome', 'Start', 'End', 'Strand', 'GeneID', 'gene_type', 'gene_name', 'hgnc_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name']\n",
    "\n",
    "    return tss_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ee7903b9-6683-4c34-a1e2-1c00004cba19",
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_regions_df = extract_tss_regions_from_df(combined_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "93ef4cae-5e4b-4c46-8c07-d0da487fff46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11819</td>\n",
       "      <td>11919</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11819</td>\n",
       "      <td>11919</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11819</td>\n",
       "      <td>11919</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12563</td>\n",
       "      <td>12663</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13171</td>\n",
       "      <td>13271</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283857</th>\n",
       "      <td>chrM</td>\n",
       "      <td>15838</td>\n",
       "      <td>15938</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>HGNC:7499</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000387460.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283858</th>\n",
       "      <td>chrM</td>\n",
       "      <td>15838</td>\n",
       "      <td>15938</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>HGNC:7499</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000387460.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283859</th>\n",
       "      <td>chrM</td>\n",
       "      <td>16023</td>\n",
       "      <td>16123</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>HGNC:7494;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283860</th>\n",
       "      <td>chrM</td>\n",
       "      <td>16023</td>\n",
       "      <td>16123</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>HGNC:7494</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000387461.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3283861</th>\n",
       "      <td>chrM</td>\n",
       "      <td>16023</td>\n",
       "      <td>16123</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>HGNC:7494</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000387461.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP-201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3283862 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome  Start    End Strand             GeneID  \\\n",
       "0             chr1  11819  11919      +  ENSG00000223972.5   \n",
       "1             chr1  11819  11919      +  ENSG00000223972.5   \n",
       "2             chr1  11819  11919      +  ENSG00000223972.5   \n",
       "3             chr1  12563  12663      +  ENSG00000223972.5   \n",
       "4             chr1  13171  13271      +  ENSG00000223972.5   \n",
       "...            ...    ...    ...    ...                ...   \n",
       "3283857       chrM  15838  15938      +  ENSG00000210195.2   \n",
       "3283858       chrM  15838  15938      +  ENSG00000210195.2   \n",
       "3283859       chrM  16023  16123      -  ENSG00000210196.2   \n",
       "3283860       chrM  16023  16123      -  ENSG00000210196.2   \n",
       "3283861       chrM  16023  16123      -  ENSG00000210196.2   \n",
       "\n",
       "                                  gene_type gene_name     hgnc_id  \\\n",
       "0        transcribed_unprocessed_pseudogene   DDX11L1  HGNC:37102   \n",
       "1        transcribed_unprocessed_pseudogene   DDX11L1  HGNC:37102   \n",
       "2        transcribed_unprocessed_pseudogene   DDX11L1  HGNC:37102   \n",
       "3        transcribed_unprocessed_pseudogene   DDX11L1  HGNC:37102   \n",
       "4        transcribed_unprocessed_pseudogene   DDX11L1  HGNC:37102   \n",
       "...                                     ...       ...         ...   \n",
       "3283857                             Mt_tRNA     MT-TT   HGNC:7499   \n",
       "3283858                             Mt_tRNA     MT-TT   HGNC:7499   \n",
       "3283859                             Mt_tRNA     MT-TP  HGNC:7494;   \n",
       "3283860                             Mt_tRNA     MT-TP   HGNC:7494   \n",
       "3283861                             Mt_tRNA     MT-TP   HGNC:7494   \n",
       "\n",
       "                   havana_gene      transcript_id       transcript_type  \\\n",
       "0        OTTHUMG00000000961.2;                NaN                   NaN   \n",
       "1         OTTHUMG00000000961.2  ENST00000456328.2  processed_transcript   \n",
       "2         OTTHUMG00000000961.2  ENST00000456328.2  processed_transcript   \n",
       "3         OTTHUMG00000000961.2  ENST00000456328.2  processed_transcript   \n",
       "4         OTTHUMG00000000961.2  ENST00000456328.2  processed_transcript   \n",
       "...                        ...                ...                   ...   \n",
       "3283857                    NaN  ENST00000387460.2               Mt_tRNA   \n",
       "3283858                    NaN  ENST00000387460.2               Mt_tRNA   \n",
       "3283859                    NaN                NaN                   NaN   \n",
       "3283860                    NaN  ENST00000387461.2               Mt_tRNA   \n",
       "3283861                    NaN  ENST00000387461.2               Mt_tRNA   \n",
       "\n",
       "        transcript_name  \n",
       "0                   NaN  \n",
       "1           DDX11L1-202  \n",
       "2           DDX11L1-202  \n",
       "3           DDX11L1-202  \n",
       "4           DDX11L1-202  \n",
       "...                 ...  \n",
       "3283857       MT-TT-201  \n",
       "3283858       MT-TT-201  \n",
       "3283859             NaN  \n",
       "3283860       MT-TP-201  \n",
       "3283861       MT-TP-201  \n",
       "\n",
       "[3283862 rows x 12 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss_regions_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8e7bb847-f105-4254-b83b-cb99ba732a05",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a custom aggregation function\n",
    "def custom_agg(series):\n",
    "    # If all values are the same, return any one of them; otherwise, return the list of unique values\n",
    "    if series.nunique() == 1:\n",
    "        return series.iloc[0]\n",
    "    else:\n",
    "        return list(series.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08c15241-179b-4766-ac9b-5a46442a087f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group by the columns 'Chromosome', 'Start', 'End', 'Strand'\n",
    "# and aggregate other columns\n",
    "unique_tss_df = tss_regions_df.groupby(['Chromosome', 'Start', 'End', 'Strand']).agg(custom_agg).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2158629-241a-46e0-ba4d-f582f5143111",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_tss_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a03ea461-3ee1-40bb-82b1-e44dc0166c03",
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_regions_dfATAWAWR (W = A or T, R = A or G)\n",
    "\n",
    "\n",
    "\n",
    "# TATAAAA\n",
    "# TATAAAT\n",
    "# TATATAA\n",
    "# TATATAT\n",
    "\n",
    "\n",
    "substrings = ['TATAAAA', 'TATAAAT', 'TATATAA', 'TATAAAA']\n",
    "rev_com = ['TTTTATA', 'ATTTATA', 'TTATATA', 'TTTTATA']\n",
    "#substrings =['TATA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f193d66f-e63d-486c-8540-47a0c87db9c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_TATA = df_pos[df_pos['Sequence'].apply(lambda x: any(substring in x for substring in substrings))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fe9f5edb-8c4a-4950-8e47-5b8733345a74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACCTATAAAATCCAGGTAAATATCTGATACTGGCACACAGGTTGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>ATGATATAAAACAGGTCAGAACCCTCCTGCCTGTCTGCTCAGTTCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>GATATAAAAGTGCTCCAACAAACCCGGAGCTGGCGTGCCGACCGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>TTTGTATAAAAGGTGAACTGAGATTTCATTCAGTCTACAGCTCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>GCTTTTCTTCCCTCCCAAGAGACCAGCAAGGCTCACTATAAATAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4002</th>\n",
       "      <td>CCATCTATAAATAGCTGGTCATCAGGGTTGATGAGAGGCTCGGAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4090</th>\n",
       "      <td>CTTATAAAAGGGACCCCAGGGAGGTAGCCCCTTCTACCATGTGTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4131</th>\n",
       "      <td>TTATAAAAGGGACCCCAGGGAGGTAGCCCCTTCTACCATGTGTGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4139</th>\n",
       "      <td>TTTGCTAGCATATAAATCTCTCGGAGGCAGAAAACATGTCTGAAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4299</th>\n",
       "      <td>GATGTATGCCTTAAATATAAATTTTCCTTACAATATGGATTTCTGT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3739 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Sequence\n",
       "5     ACCTATAAAATCCAGGTAAATATCTGATACTGGCACACAGGTTGGA...\n",
       "44    ATGATATAAAACAGGTCAGAACCCTCCTGCCTGTCTGCTCAGTTCA...\n",
       "65    GATATAAAAGTGCTCCAACAAACCCGGAGCTGGCGTGCCGACCGCG...\n",
       "86    TTTGTATAAAAGGTGAACTGAGATTTCATTCAGTCTACAGCTCTTG...\n",
       "103   GCTTTTCTTCCCTCCCAAGAGACCAGCAAGGCTCACTATAAATAGC...\n",
       "...                                                 ...\n",
       "4002  CCATCTATAAATAGCTGGTCATCAGGGTTGATGAGAGGCTCGGAGG...\n",
       "4090  CTTATAAAAGGGACCCCAGGGAGGTAGCCCCTTCTACCATGTGTGG...\n",
       "4131  TTATAAAAGGGACCCCAGGGAGGTAGCCCCTTCTACCATGTGTGGA...\n",
       "4139  TTTGCTAGCATATAAATCTCTCGGAGGCAGAAAACATGTCTGAAAT...\n",
       "4299  GATGTATGCCTTAAATATAAATTTTCCTTACAATATGGATTTCTGT...\n",
       "\n",
       "[3739 rows x 1 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_TATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "da105da3-29e7-4fce-b9d1-9da69096fdb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_non_TATA = df_pos[~df_pos['Sequence'].apply(lambda x: any(substring in x for substring in substrings))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a368d99e-87ad-4a4c-97ab-ca8567b9c5ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCTTTGCATGCGCACGGCCGGCCCCACCCCCGCTGTCAGCTGGAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GGACGGAGCGGGGTGGAGGGCGGGGGCGGGGCCTAGTCCTGAGAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CTCCACTTCCTGTCCCCGCGGCTCTGTGTCCCCTGCTAGCCGTAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TGGCCCCGCCCCTCCGCGCTCAGGCCCCGCCCCCAGCTCCGAGGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCCCTCCCACCCCACTGCTCCCGCTCCATTGTCTGGGAATTGCAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4319</th>\n",
       "      <td>ATGACAGTCCTGTCTCTTTAAGAAGAAAAAAAAAGGCTGCTGGGTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4320</th>\n",
       "      <td>GTGCGAGGAGTGGCGCATGCTCTGTGAGGCCGGCAGCTTCCCATTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4321</th>\n",
       "      <td>ATGACAGTCCTGTCTCTTTAAGAAGAAAAAAAAAGGCTGCTGGGTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4322</th>\n",
       "      <td>CGCGGCACAGGCGGCGGCGTCTCCAGGGGGAGCCAAGGTACGTAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4323</th>\n",
       "      <td>ATGACAGTCCTGTCTCTTTAAGAAGAAAAAAAAAGGCTGCTGGGTA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>213699 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Sequence\n",
       "0     GCTTTGCATGCGCACGGCCGGCCCCACCCCCGCTGTCAGCTGGAGG...\n",
       "1     GGACGGAGCGGGGTGGAGGGCGGGGGCGGGGCCTAGTCCTGAGAGG...\n",
       "2     CTCCACTTCCTGTCCCCGCGGCTCTGTGTCCCCTGCTAGCCGTAGG...\n",
       "3     TGGCCCCGCCCCTCCGCGCTCAGGCCCCGCCCCCAGCTCCGAGGGC...\n",
       "4     TCCCTCCCACCCCACTGCTCCCGCTCCATTGTCTGGGAATTGCAGC...\n",
       "...                                                 ...\n",
       "4319  ATGACAGTCCTGTCTCTTTAAGAAGAAAAAAAAAGGCTGCTGGGTA...\n",
       "4320  GTGCGAGGAGTGGCGCATGCTCTGTGAGGCCGGCAGCTTCCCATTG...\n",
       "4321  ATGACAGTCCTGTCTCTTTAAGAAGAAAAAAAAAGGCTGCTGGGTA...\n",
       "4322  CGCGGCACAGGCGGCGGCGTCTCCAGGGGGAGCCAAGGTACGTAGG...\n",
       "4323  ATGACAGTCCTGTCTCTTTAAGAAGAAAAAAAAAGGCTGCTGGGTA...\n",
       "\n",
       "[213699 rows x 1 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_non_TATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8269260f-ccba-4d08-aa85-ce0ec59f9dd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pos=pd.DataFrame()\n",
    "df_pos['Sequence'] = combined_csv_data['sequence'].str.upper().apply(seq2kmer, args=(6,))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6d21d321-a58c-493f-8d54-0de0fa4a6db5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCTTTG CTTTGC TTTGCA TTGCAT TGCATG GCATGC CATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GGACGG GACGGA ACGGAG CGGAGC GGAGCG GAGCGG AGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CTCCAC TCCACT CCACTT CACTTC ACTTCC CTTCCT TTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TGGCCC GGCCCC GCCCCG CCCCGC CCCGCC CCGCCC CGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCCCTC CCCTCC CCTCCC CTCCCA TCCCAC CCCACC CCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4320</th>\n",
       "      <td>GTGCGA TGCGAG GCGAGG CGAGGA GAGGAG AGGAGT GGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4321</th>\n",
       "      <td>ATGACA TGACAG GACAGT ACAGTC CAGTCC AGTCCT GTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4322</th>\n",
       "      <td>CGCGGC GCGGCA CGGCAC GGCACA GCACAG CACAGG ACAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4323</th>\n",
       "      <td>ATGACA TGACAG GACAGT ACAGTC CAGTCC AGTCCT GTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4324</th>\n",
       "      <td>CCTGAC CTGACC TGACCA GACCAG ACCAGA CCAGAT CAGA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>236975 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Sequence\n",
       "0     GCTTTG CTTTGC TTTGCA TTGCAT TGCATG GCATGC CATG...\n",
       "1     GGACGG GACGGA ACGGAG CGGAGC GGAGCG GAGCGG AGCG...\n",
       "2     CTCCAC TCCACT CCACTT CACTTC ACTTCC CTTCCT TTCC...\n",
       "3     TGGCCC GGCCCC GCCCCG CCCCGC CCCGCC CCGCCC CGCC...\n",
       "4     TCCCTC CCCTCC CCTCCC CTCCCA TCCCAC CCCACC CCAC...\n",
       "...                                                 ...\n",
       "4320  GTGCGA TGCGAG GCGAGG CGAGGA GAGGAG AGGAGT GGAG...\n",
       "4321  ATGACA TGACAG GACAGT ACAGTC CAGTCC AGTCCT GTCC...\n",
       "4322  CGCGGC GCGGCA CGGCAC GGCACA GCACAG CACAGG ACAG...\n",
       "4323  ATGACA TGACAG GACAGT ACAGTC CAGTCC AGTCCT GTCC...\n",
       "4324  CCTGAC CTGACC TGACCA GACCAG ACCAGA CCAGAT CAGA...\n",
       "\n",
       "[236975 rows x 1 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "86981274-16dc-4444-ae56-df2b19be889c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pos['Label']=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "53e66247-0acf-41d9-b6fc-02bba72480f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCTTTG CTTTGC TTTGCA TTGCAT TGCATG GCATGC CATG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GGACGG GACGGA ACGGAG CGGAGC GGAGCG GAGCGG AGCG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CTCCAC TCCACT CCACTT CACTTC ACTTCC CTTCCT TTCC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TGGCCC GGCCCC GCCCCG CCCCGC CCCGCC CCGCCC CGCC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCCCTC CCCTCC CCTCCC CTCCCA TCCCAC CCCACC CCAC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236970</th>\n",
       "      <td>GTGCGA TGCGAG GCGAGG CGAGGA GAGGAG AGGAGT GGAG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236971</th>\n",
       "      <td>ATGACA TGACAG GACAGT ACAGTC CAGTCC AGTCCT GTCC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236972</th>\n",
       "      <td>CGCGGC GCGGCA CGGCAC GGCACA GCACAG CACAGG ACAG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236973</th>\n",
       "      <td>ATGACA TGACAG GACAGT ACAGTC CAGTCC AGTCCT GTCC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236974</th>\n",
       "      <td>CCTGAC CTGACC TGACCA GACCAG ACCAGA CCAGAT CAGA...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>236975 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Sequence  Label\n",
       "0       GCTTTG CTTTGC TTTGCA TTGCAT TGCATG GCATGC CATG...      1\n",
       "1       GGACGG GACGGA ACGGAG CGGAGC GGAGCG GAGCGG AGCG...      1\n",
       "2       CTCCAC TCCACT CCACTT CACTTC ACTTCC CTTCCT TTCC...      1\n",
       "3       TGGCCC GGCCCC GCCCCG CCCCGC CCCGCC CCGCCC CGCC...      1\n",
       "4       TCCCTC CCCTCC CCTCCC CTCCCA TCCCAC CCCACC CCAC...      1\n",
       "...                                                   ...    ...\n",
       "236970  GTGCGA TGCGAG GCGAGG CGAGGA GAGGAG AGGAGT GGAG...      1\n",
       "236971  ATGACA TGACAG GACAGT ACAGTC CAGTCC AGTCCT GTCC...      1\n",
       "236972  CGCGGC GCGGCA CGGCAC GGCACA GCACAG CACAGG ACAG...      1\n",
       "236973  ATGACA TGACAG GACAGT ACAGTC CAGTCC AGTCCT GTCC...      1\n",
       "236974  CCTGAC CTGACC TGACCA GACCAG ACCAGA CCAGAT CAGA...      1\n",
       "\n",
       "[236975 rows x 2 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pos = df_pos.reset_index(drop=True)\n",
    "df_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0fd94088-8542-4675-850a-5850e95cb0ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pos.to_csv(\"/data/projects/DNABERT_data/Core_promoters/Core_promoter_regions/positive_set.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "3c073550-550d-4db9-a474-8bdf50ab0850",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d27f37d9-a1b9-45ab-a2f4-637f59be2453",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
