{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "13e86a2e-92b3-4644-a96d-547bf513e828",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, glob\n",
    "import pandas as pd\n",
    "import pysam\n",
    "import numpy as np\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c2380204-9847-4c6f-b152-114f02e0139d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_type = \"Brain\"\n",
    "intersected_base_path = \"/data/projects/GDC_Cancer_Wise/New_data/{}/Generated_files/Intersected_Data/TFBS_remaining\".format(cancer_type)\n",
    "reference_genome_path=\"/data/projects/Resources/Gencode_genome_annotation/GRCh38.primary_assembly.genome.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5a4aa801-312d-4380-adb6-0c0ebcd9ba78",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_path = \"/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/DNABERT_Data/TFBS_remaining\"\n",
    "if not os.path.exists(output_path):\n",
    "    os.makedirs(output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1e9c0e52-cf4d-48b1-9173-cd32e57d8d95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tags</th>\n",
       "      <th>config</th>\n",
       "      <th>eval_acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RBM14</td>\n",
       "      <td>{'epochs': 10, 'Dropout': 0.1, 'batch_size': 1...</td>\n",
       "      <td>0.968908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAFB</td>\n",
       "      <td>{'epochs': 10, 'Dropout': 0.1, 'batch_size': 4...</td>\n",
       "      <td>0.968090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TAF15</td>\n",
       "      <td>{'epochs': 10, 'Dropout': 0.1, 'batch_size': 4...</td>\n",
       "      <td>0.959936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPI1</td>\n",
       "      <td>{'epochs': 10, 'Dropout': 0.1, 'batch_size': 1...</td>\n",
       "      <td>0.957358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>USF1</td>\n",
       "      <td>{'epochs': 10, 'Dropout': 0.1, 'batch_size': 4...</td>\n",
       "      <td>0.951848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>ESRRA</td>\n",
       "      <td>{'epochs': 10, 'Dropout': 0.1, 'batch_size': 4...</td>\n",
       "      <td>0.851964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>ZNF652</td>\n",
       "      <td>{'epochs': 10, 'Dropout': 0.1, 'batch_size': 1...</td>\n",
       "      <td>0.851404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>ATF3</td>\n",
       "      <td>{'epochs': 10, 'Dropout': 0.1, 'batch_size': 4...</td>\n",
       "      <td>0.851240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>RBM25</td>\n",
       "      <td>{'epochs': 10, 'Dropout': 0.1, 'batch_size': 1...</td>\n",
       "      <td>0.850664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>CHD2</td>\n",
       "      <td>{'epochs': 10, 'Dropout': 0.1, 'batch_size': 4...</td>\n",
       "      <td>0.850362</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>139 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       tags                                             config  eval_acc\n",
       "0     RBM14  {'epochs': 10, 'Dropout': 0.1, 'batch_size': 1...  0.968908\n",
       "1      SAFB  {'epochs': 10, 'Dropout': 0.1, 'batch_size': 4...  0.968090\n",
       "2     TAF15  {'epochs': 10, 'Dropout': 0.1, 'batch_size': 4...  0.959936\n",
       "3      SPI1  {'epochs': 10, 'Dropout': 0.1, 'batch_size': 1...  0.957358\n",
       "4      USF1  {'epochs': 10, 'Dropout': 0.1, 'batch_size': 4...  0.951848\n",
       "..      ...                                                ...       ...\n",
       "134   ESRRA  {'epochs': 10, 'Dropout': 0.1, 'batch_size': 4...  0.851964\n",
       "135  ZNF652  {'epochs': 10, 'Dropout': 0.1, 'batch_size': 1...  0.851404\n",
       "136    ATF3  {'epochs': 10, 'Dropout': 0.1, 'batch_size': 4...  0.851240\n",
       "137   RBM25  {'epochs': 10, 'Dropout': 0.1, 'batch_size': 1...  0.850664\n",
       "138    CHD2  {'epochs': 10, 'Dropout': 0.1, 'batch_size': 4...  0.850362\n",
       "\n",
       "[139 rows x 3 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_models_df= pd.read_csv(\"/home/campus.stonybrook.edu/pdutta/Github/Postdoc/DNABERT_data_processing/TFBS/top_TFBS_models.tsv\", sep=\"\\t\") \n",
    "top_models_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c05f597d-17d5-4ea2-9dae-2903b1c0fe98",
   "metadata": {},
   "outputs": [],
   "source": [
    "def seq2kmer(seq):\n",
    "    \"\"\"\n",
    "    Convert original sequence to kmers\n",
    "    \n",
    "    Arguments:\n",
    "    seq -- str, original sequence.\n",
    "    k -- int, kmer of length k specified.\n",
    "    \n",
    "    Returns:\n",
    "    kmers -- str, kmers separated by space\n",
    "    \"\"\"\n",
    "    k=6\n",
    "    kmer = [seq[x:x+k] for x in range(len(seq)+1-k)]\n",
    "    kmers = \" \".join(kmer)\n",
    "    return kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a30b3260-fbcf-4037-9916-961c31c8c4f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sequences(df, reference_fasta):\n",
    "    data = []\n",
    "    for idx, row in df.iterrows():\n",
    "        #print(row)\n",
    "        chrom = row[0]\n",
    "        ref_start = row[1]\n",
    "        ref_end = row[2]\n",
    "        variant_start = row['START_POS']\n",
    "        variant_end = row['END_POS']\n",
    "        ref_nucleotide = row['REF']\n",
    "        alt = row[\"ALT\"]\n",
    "        \n",
    "        # Adjust for 0-based indexing in python\n",
    "        variant_pos_start = variant_start - ref_start\n",
    "        variant_pos_end = variant_end - ref_start\n",
    "        #print(ref_nucleotide , alt, variant_pos_start, variant_pos_end)\n",
    "        \n",
    "        \n",
    "        # Get reference sequence\n",
    "        #print(chrom)\n",
    "        ref_seq = reference_fasta.fetch(chrom, ref_start, ref_end)\n",
    "        #print(ref_seq)\n",
    "        \n",
    "        # Handle insertion and deletion to get the correct alt sequence\n",
    "        # Identify if the variant is an insertion or deletion\n",
    "        if len(ref_nucleotide) < len(alt):  # Insertion\n",
    "            delete_size =  len(alt) - len(ref_nucleotide)\n",
    "            #print(variant_pos_start, variant_pos_end ,delete_size)\n",
    "            alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:len(ref_seq) - delete_size]\n",
    "\n",
    "        elif len(ref_nucleotide) > len(alt):  # Deletion\n",
    "            insert_size = len(ref_nucleotide) - len(alt)\n",
    "            #print(insert_size)\n",
    "            extra_bases = reference_fasta.fetch(chrom, ref_end, ref_end + insert_size)\n",
    "            #print(extra_bases)\n",
    "            alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:] + extra_bases\n",
    "\n",
    "        else:  # SNV\n",
    "            alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:]\n",
    "        # print(ref_seq)\n",
    "        # print(alt_seq)\n",
    "        # input()\n",
    "\n",
    "\n",
    "        data.append({\n",
    "            'chr': chrom,\n",
    "            #'strand': row['strand'],\n",
    "            #'Transcript_ID': row['transcript_id'],\n",
    "            'Chip_Seq_start': row[0],\n",
    "            'Chip_Seq_end': row[1],\n",
    "            'varinat_start': variant_start,\n",
    "            'variant_end': variant_end,\n",
    "            'ref_neucleotide': ref_nucleotide,\n",
    "            'alternative_neucleotide': alt,\n",
    "            'reference_seq': ref_seq,\n",
    "            'alt_seq': alt_seq\n",
    "        })\n",
    "            \n",
    "    \n",
    "    # Convert the list of dictionaries to a DataFrame\n",
    "    new_df = pd.DataFrame(data)\n",
    "    # print(new_df.shape)\n",
    "    # print(new_df)\n",
    "    new_df = new_df.drop_duplicates().reset_index()\n",
    "    #print(new_df.shape)\n",
    "    #print(\"&*\")\n",
    "    data = []\n",
    "    merged_list = list(zip(new_df['reference_seq'], new_df['alt_seq']))\n",
    "    merged_list = [item.upper() for tup in merged_list for item in tup]\n",
    "    #print(merged_list)\n",
    "    kmer_lst = list(map(seq2kmer, merged_list))\n",
    "    df_kmer = pd.DataFrame(kmer_lst, columns=['Sequence'])\n",
    "    df_kmer['Label'] = np.random.choice([0, 1], size=len(df_kmer))\n",
    "    #print(df_kmer.shape)\n",
    "    return new_df, df_kmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ef4c248a-8b67-4dd9-8992-23b3d81450f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#####The reference file is loaded######\n"
     ]
    }
   ],
   "source": [
    "## Load the reference file\n",
    "reference_fasta = pysam.FastaFile(reference_genome_path)\n",
    "print(\"#####The reference file is loaded######\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "933aa443-d84e-48a9-b824-e5df3470e39b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/RBM14/intersected_vcf_data.pkl\n",
      "File not found for RBM14, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/SAFB/intersected_vcf_data.pkl\n",
      "File not found for SAFB, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/TAF15/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for TAF15***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/SPI1/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for SPI1***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/USF1/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for USF1***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZNF426/intersected_vcf_data.pkl\n",
      "File not found for ZNF426, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/MAFK/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for MAFK***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/FUS/intersected_vcf_data.pkl\n",
      "File not found for FUS, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/MAFF/intersected_vcf_data.pkl\n",
      "File not found for MAFF, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/THAP1/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for THAP1***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/SRSF1/intersected_vcf_data.pkl\n",
      "File not found for SRSF1, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZNF146/intersected_vcf_data.pkl\n",
      "File not found for ZNF146, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/HNRNPLL/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for HNRNPLL***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/PCBP1/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for PCBP1***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/AGO1/intersected_vcf_data.pkl\n",
      "File not found for AGO1, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/FOS/intersected_vcf_data.pkl\n",
      "File not found for FOS, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZNF654/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZNF654***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/NFE2L2/intersected_vcf_data.pkl\n",
      "File not found for NFE2L2, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/C11orf30/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for C11orf30***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/CEBPB/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for CEBPB***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/HNRNPL/intersected_vcf_data.pkl\n",
      "File not found for HNRNPL, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/EGR2/intersected_vcf_data.pkl\n",
      "File not found for EGR2, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/E2F4/intersected_vcf_data.pkl\n",
      "File not found for E2F4, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZFHX2/intersected_vcf_data.pkl\n",
      "File not found for ZFHX2, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZNF770/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZNF770***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/MITF/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for MITF***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/RAD212/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for RAD212***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/GLIS2/intersected_vcf_data.pkl\n",
      "File not found for GLIS2, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZNF143/intersected_vcf_data.pkl\n",
      "File not found for ZNF143, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/MAFG/intersected_vcf_data.pkl\n",
      "File not found for MAFG, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZNF777/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZNF777***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZNF555/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZNF555***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZNF316/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZNF316***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/USF2/intersected_vcf_data.pkl\n",
      "File not found for USF2, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/FOSL2/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for FOSL2***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/MXD3/intersected_vcf_data.pkl\n",
      "File not found for MXD3, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/BLC3/intersected_vcf_data.pkl\n",
      "File not found for BLC3, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/RBM39/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for RBM39***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZBED52/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZBED52***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/SIN3B/intersected_vcf_data.pkl\n",
      "File not found for SIN3B, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/UBTF/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for UBTF***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/RAD21M/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for RAD21M***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/NR3C1/intersected_vcf_data.pkl\n",
      "File not found for NR3C1, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/PAX5/intersected_vcf_data.pkl\n",
      "File not found for PAX5, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/FOXA2/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for FOXA2***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZBTB26/intersected_vcf_data.pkl\n",
      "File not found for ZBTB26, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/SAP30/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for SAP30***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZNF585B/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZNF585B***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZC3H11A/intersected_vcf_data.pkl\n",
      "File not found for ZC3H11A, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/AGO2/intersected_vcf_data.pkl\n",
      "File not found for AGO2, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZNF623/intersected_vcf_data.pkl\n",
      "File not found for ZNF623, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/FOXA1/intersected_vcf_data.pkl\n",
      "File not found for FOXA1, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/EGR1/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for EGR1***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/SMC3/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for SMC3***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/U2AF1/intersected_vcf_data.pkl\n",
      "File not found for U2AF1, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/MYC/intersected_vcf_data.pkl\n",
      "File not found for MYC, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/SRF/intersected_vcf_data.pkl\n",
      "File not found for SRF, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/CREB1/intersected_vcf_data.pkl\n",
      "File not found for CREB1, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/TFDP12/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for TFDP12***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/RBBP5/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for RBBP5***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZNF292/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZNF292***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/GABPB1/intersected_vcf_data.pkl\n",
      "File not found for GABPB1, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/STAT3/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for STAT3***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/MAZ/intersected_vcf_data.pkl\n",
      "File not found for MAZ, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZIC2/intersected_vcf_data.pkl\n",
      "File not found for ZIC2, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZFP36/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZFP36***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/TFAP4/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for TFAP4***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZNF121/intersected_vcf_data.pkl\n",
      "File not found for ZNF121, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/SUZ12/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for SUZ12***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/KDM5B/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for KDM5B***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/FOXP2/intersected_vcf_data.pkl\n",
      "File not found for FOXP2, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/KLF9/intersected_vcf_data.pkl\n",
      "File not found for KLF9, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/POLR2A/intersected_vcf_data.pkl\n",
      "File not found for POLR2A, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ELK1/intersected_vcf_data.pkl\n",
      "File not found for ELK1, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/RUNX3/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for RUNX3***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/SMAD42/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for SMAD42***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/CBFA2T2/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for CBFA2T2***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/XRCC5/intersected_vcf_data.pkl\n",
      "File not found for XRCC5, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/CEBPZ/intersected_vcf_data.pkl\n",
      "File not found for CEBPZ, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/NONO/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for NONO***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZNF785/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZNF785***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/MEIS2/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for MEIS2***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/TAF1/intersected_vcf_data.pkl\n",
      "File not found for TAF1, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/HNF4G/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for HNF4G***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/CEBPG/intersected_vcf_data.pkl\n",
      "File not found for CEBPG, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/HNRNPH1/intersected_vcf_data.pkl\n",
      "File not found for HNRNPH1, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZSCAN4/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZSCAN4***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZBTB6/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZBTB6***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/POLR2AphosphoS5/intersected_vcf_data.pkl\n",
      "File not found for POLR2AphosphoS5, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/GLIS1/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for GLIS1***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/NFYC/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for NFYC***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/SREBF2/intersected_vcf_data.pkl\n",
      "File not found for SREBF2, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/CBFA2T3/intersected_vcf_data.pkl\n",
      "File not found for CBFA2T3, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/POU2F2/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for POU2F2***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZFX/intersected_vcf_data.pkl\n",
      "File not found for ZFX, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/CBX1/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for CBX1***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/JUND/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for JUND***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/RARA/intersected_vcf_data.pkl\n",
      "File not found for RARA, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZSCAN5A/intersected_vcf_data.pkl\n",
      "File not found for ZSCAN5A, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/PHF20/intersected_vcf_data.pkl\n",
      "File not found for PHF20, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/PKNOX1/intersected_vcf_data.pkl\n",
      "File not found for PKNOX1, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/SETDB1/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for SETDB1***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/U2AF2/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for U2AF2***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/FOXP1/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for FOXP1***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/NR2F62/intersected_vcf_data.pkl\n",
      "File not found for NR2F62, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/RXRA/intersected_vcf_data.pkl\n",
      "File not found for RXRA, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/HBP1/intersected_vcf_data.pkl\n",
      "File not found for HBP1, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/SP5/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for SP5***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/IRF5/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for IRF5***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/SIX5/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for SIX5***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/H4K12ac/intersected_vcf_data.pkl\n",
      "File not found for H4K12ac, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZNF398/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZNF398***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/MYRF/intersected_vcf_data.pkl\n",
      "File not found for MYRF, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/HDAC6/intersected_vcf_data.pkl\n",
      "File not found for HDAC6, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/GLI4/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for GLI4***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZNF560/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZNF560***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZNF596/intersected_vcf_data.pkl\n",
      "File not found for ZNF596, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZNF521/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZNF521***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/DEAF1/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for DEAF1***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/RUNX1/intersected_vcf_data.pkl\n",
      "File not found for RUNX1, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/MNT/intersected_vcf_data.pkl\n",
      "File not found for MNT, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/SCRT2/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for SCRT2***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZNF263/intersected_vcf_data.pkl\n",
      "File not found for ZNF263, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/H2AK9ac/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for H2AK9ac***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/TEAD32/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for TEAD32***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/MBD2/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for MBD2***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZBTB8A/intersected_vcf_data.pkl\n",
      "File not found for ZBTB8A, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZFP642/intersected_vcf_data.pkl\n",
      "File not found for ZFP642, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/GTF2F1/intersected_vcf_data.pkl\n",
      "File not found for GTF2F1, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZMYM3/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZMYM3***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/HMGXB4/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for HMGXB4***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/VEZF1/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for VEZF1***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZBTB20/intersected_vcf_data.pkl\n",
      "File not found for ZBTB20, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/GATA4/intersected_vcf_data.pkl\n",
      "File not found for GATA4, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ESRRA/intersected_vcf_data.pkl\n",
      "File not found for ESRRA, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ZNF652/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ZNF652***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/ATF3/intersected_vcf_data.pkl\n",
      "****All the VCF files are loaded for ATF3***\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/RBM25/intersected_vcf_data.pkl\n",
      "File not found for RBM25, skipping...\n",
      "/data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/Intersected_Data/TFBS_remaining/CHD2/intersected_vcf_data.pkl\n",
      "File not found for CHD2, skipping...\n",
      "Missing files recorded at /data/projects/GDC_Cancer_Wise/New_data/Brain/Generated_files/DNABERT_Data/TFBS_remaining/missing_files.txt\n"
     ]
    }
   ],
   "source": [
    "missing_files = []\n",
    "\n",
    "\n",
    "\n",
    "for index, row in top_models_df.iterrows():\n",
    "    intersected_data = f\"{intersected_base_path}/{row['tags']}/intersected_vcf_data.pkl\"\n",
    "    print(intersected_data)\n",
    "    \n",
    "    try:\n",
    "        with open(intersected_data, \"rb\") as file:\n",
    "            loaded_dictionary = pickle.load(file)\n",
    "        print(f\"****All the VCF files are loaded for {row['tags']}***\") \n",
    "\n",
    "        dnabert_raw_data = {}\n",
    "        for key, value in loaded_dictionary.items():\n",
    "            new_df, df_kmer = get_sequences(value, reference_fasta)\n",
    "            dnabert_raw_data[key] = new_df\n",
    "\n",
    "            out_folder_path = f\"{output_path}/{row['tags']}/Patient_wise/{key}\"\n",
    "            if not os.path.exists(out_folder_path):\n",
    "                os.makedirs(out_folder_path)\n",
    "            df_kmer.to_csv(out_folder_path + \"/dev.tsv\", sep=\"\\t\", index=False)\n",
    "        \n",
    "        with open(f\"{output_path}/{row['tags']}/raw_vcf_data.pkl\", \"wb\") as file:\n",
    "            pickle.dump(dnabert_raw_data, file)\n",
    "\n",
    "    except FileNotFoundError:\n",
    "        print(f\"File not found for {row['tags']}, skipping...\")\n",
    "        missing_files.append(row['tags'])\n",
    "\n",
    "# Optionally, save the list of missing files/tags to a text file\n",
    "missing_files_path = f\"{output_path}/missing_files.txt\"\n",
    "with open(missing_files_path, \"w\") as f:\n",
    "    for tag in missing_files:\n",
    "        f.write(f\"{tag}\\n\")\n",
    "\n",
    "print(f\"Missing files recorded at {missing_files_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "3d3978e4-e2c0-4bc1-864d-e76c01f2e8b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(missing_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d00e115-459f-4e81-ae6a-e0b38ce19e74",
   "metadata": {},
   "outputs": [],
   "source": [
    "6"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
