{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "9807e3a4-8edb-46a2-ad4c-88777af9fadc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, glob\n",
    "import pandas as pd\n",
    "from io import StringIO\n",
    "import numpy as np\n",
    "import pysam\n",
    "import pybedtools \n",
    "from Bio import SeqIO\n",
    "pybedtools.helpers.set_tempdir(\"/data/projects/temp\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2505ee28-f7a7-474a-a4b8-755989f3f70e",
   "metadata": {},
   "outputs": [],
   "source": [
    "non_coding_region = \"Acceptor\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6c3ba47b-0edf-43de-bab5-bc0edd7f91b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '/data/projects/Enhancer/RECOMB_2024/Intersected_data/{}/all_data.tsv'.format(non_coding_region)\n",
    "output_path = \"/data/projects/Enhancer/RECOMB_2024/SNP_data/{}\".format(non_coding_region) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aafe0877-b7be-4e5e-a73f-00dab9627b51",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(output_path):\n",
    "    os.makedirs(output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "cc89bac5-1528-4192-95d4-77bda6a6e1dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def seq2kmer(seq):\n",
    "    \"\"\"\n",
    "    Convert original sequence to kmers\n",
    "    \n",
    "    Arguments:\n",
    "    seq -- str, original sequence.\n",
    "    k -- int, kmer of length k specified.\n",
    "    \n",
    "    Returns:\n",
    "    kmers -- str, kmers separated by space\n",
    "    \"\"\"\n",
    "    k=6\n",
    "    kmer = [seq[x:x+k] for x in range(len(seq)+1-k)]\n",
    "    kmers = \" \".join(kmer)\n",
    "    return kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "19aaa7a0-7309-4e97-a93e-dded3164e07f",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = SeqIO.to_dict(SeqIO.parse(\"/data/projects/Resources/Gencode_genome_annotation/GRCh38.primary_assembly.genome.fa\", \"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "0ded228f-2999-482f-9e5a-197b732e0cfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(38758, 13)\n",
      "(38184, 13)\n",
      "(38184, 13)\n",
      "(43563, 11)\n",
      "(42854, 12)\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(data_path, sep= '\\t')\n",
    "print(df.shape)\n",
    "df = df[df['No_of_variations']<3].reset_index(drop=True)\n",
    "print(df.shape)\n",
    "df = df.loc[df['Alternative_Nucleotides'] != '.'].reset_index(drop=True)\n",
    "print(df.shape)\n",
    "\n",
    "\n",
    "data= []\n",
    "#df = df.head(10)\n",
    "# Iterate over DataFrame rows\n",
    "for idx, row in df.iterrows():\n",
    "    try:\n",
    "        #print(row)\n",
    "        alts = row['Alternative_Nucleotides'].rstrip(',').split(',')\n",
    "        ref_seq = str(genome[row['chr_name']].seq[ row['Acceptor_start']:row['Acceptor_end']])\n",
    "        #ref_sequences.append(ref_sequence)\n",
    "        #print(alts)\n",
    "        #print(ref_sequence)\n",
    "\n",
    "        # Calculate the variant position relative to the fetched sequence\n",
    "        variant_pos_start = row['Variant_start'] - row['Acceptor_start']\n",
    "        variant_pos_end = row['Variant_end']  - row['Acceptor_start']\n",
    "        #print(variant_pos_start, variant_pos_end)\n",
    "\n",
    "\n",
    "        # Replace the reference nucleotide with the alternate nucleotide to get the alternate sequence\n",
    "        for alt in alts:\n",
    "            if alt == \"\":  # Deletion\n",
    "                # Fetch additional bases from genome to maintain sequence length\n",
    "                extra_bases = str(genome[row['chr_name']].seq[row['Acceptor_end']:row['Acceptor_end'] + len(row['Reference_Nucleotide'])])\n",
    "                #print(len(row['Reference_Nucleotide']))\n",
    "                alt_seq = ref_seq[:variant_pos_start] + ref_seq[variant_pos_end:] + extra_bases\n",
    "            else:  # SNPs, Insertions, and partial Deletions\n",
    "                extra_bases = str(genome[row['chr_name']].seq[row['Acceptor_end']:row['Acceptor_end'] + len(row['Reference_Nucleotide']) - len(alt)])\n",
    "                #print(extra_bases)\n",
    "                if(variant_pos_start>=0):\n",
    "                    alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:] + extra_bases\n",
    "                else:\n",
    "                    continue\n",
    "            # Append to the list as a dictionary\n",
    "            data.append({\n",
    "                'chr': row['chr_name'],\n",
    "                'strand': row['Strand'],\n",
    "                'Acceptor_coordinates': str(row['Acceptor_start'])+'-'+str(row['Acceptor_end']),\n",
    "                'Ensemble_Transcript_ID': row['Transcript_id'],\n",
    "                'dbsnp_id': row['RS_ID'],\n",
    "                'varinat_start':  row['Variant_start'],\n",
    "                'variant_end':  row['Variant_end'],\n",
    "                'ref_neucleotide': row['Reference_Nucleotide'],\n",
    "                'alternative_neucleotide': alt,\n",
    "                'reference_seq': ref_seq,\n",
    "                'alt_seq': alt_seq\n",
    "            })\n",
    "    except KeyError:\n",
    "        # If sequence is not present, append a default value\n",
    "        ref_sequences.append(\"NA\")\n",
    "        alt_sequences.append(\"NA\")\n",
    "\n",
    "        \n",
    "        \n",
    "new_df = pd.DataFrame(data)\n",
    "print(new_df.shape)\n",
    "new_df = new_df.drop_duplicates().reset_index()\n",
    "print(new_df.shape)\n",
    "merged_list = list(zip(new_df['reference_seq'], new_df['alt_seq']))\n",
    "merged_list = [item.upper() for tup in merged_list for item in tup]\n",
    "#print(merged_list)\n",
    "kmer_lst = list(map(seq2kmer, merged_list))\n",
    "df_kmer = pd.DataFrame(kmer_lst, columns=['Sequence'])\n",
    "df_kmer['Label'] = np.random.choice([0, 1], size=len(df_kmer))\n",
    "\n",
    "new_df.to_csv(output_path + \"/all_data_new.tsv\", sep=\"\\t\", index= False)\n",
    "df_kmer.to_csv(output_path + \"/dev.tsv\", sep=\"\\t\", index= False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "7f6ff90c-19f5-4d83-b2ab-7dbcf50a30f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_name</th>\n",
       "      <th>Acceptor_start</th>\n",
       "      <th>Acceptor_end</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Enhancer_start</th>\n",
       "      <th>Enhancer_end</th>\n",
       "      <th>Variant_start</th>\n",
       "      <th>Variant_end</th>\n",
       "      <th>RS_ID</th>\n",
       "      <th>Reference_Nucleotide</th>\n",
       "      <th>No_of_variations</th>\n",
       "      <th>Alternative_Nucleotides</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>108897190</td>\n",
       "      <td>108897269</td>\n",
       "      <td>ENST00000258443.7</td>\n",
       "      <td>-</td>\n",
       "      <td>108896998</td>\n",
       "      <td>108897197</td>\n",
       "      <td>108897187</td>\n",
       "      <td>108897193</td>\n",
       "      <td>rs1574362124</td>\n",
       "      <td>TCTTCT</td>\n",
       "      <td>1</td>\n",
       "      <td>TCT,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>108897190</td>\n",
       "      <td>108897269</td>\n",
       "      <td>ENST00000258443.7</td>\n",
       "      <td>-</td>\n",
       "      <td>108896998</td>\n",
       "      <td>108897197</td>\n",
       "      <td>108897191</td>\n",
       "      <td>108897192</td>\n",
       "      <td>rs566530415</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>A,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>108897190</td>\n",
       "      <td>108897269</td>\n",
       "      <td>ENST00000258443.7</td>\n",
       "      <td>-</td>\n",
       "      <td>108896998</td>\n",
       "      <td>108897197</td>\n",
       "      <td>108897193</td>\n",
       "      <td>108897194</td>\n",
       "      <td>rs121908456</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>A,T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>108897190</td>\n",
       "      <td>108897269</td>\n",
       "      <td>ENST00000258443.7</td>\n",
       "      <td>-</td>\n",
       "      <td>108896998</td>\n",
       "      <td>108897197</td>\n",
       "      <td>108897194</td>\n",
       "      <td>108897195</td>\n",
       "      <td>rs777416175</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>A,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>108897190</td>\n",
       "      <td>108897269</td>\n",
       "      <td>ENST00000258443.7</td>\n",
       "      <td>-</td>\n",
       "      <td>108896998</td>\n",
       "      <td>108897197</td>\n",
       "      <td>108897206</td>\n",
       "      <td>108897207</td>\n",
       "      <td>rs770460028</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>C,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38179</th>\n",
       "      <td>chr5</td>\n",
       "      <td>179708198</td>\n",
       "      <td>179708277</td>\n",
       "      <td>ENST00000681903.1</td>\n",
       "      <td>+</td>\n",
       "      <td>179708271</td>\n",
       "      <td>179708470</td>\n",
       "      <td>179708265</td>\n",
       "      <td>179708266</td>\n",
       "      <td>rs774523334</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>C,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38180</th>\n",
       "      <td>chr5</td>\n",
       "      <td>179708198</td>\n",
       "      <td>179708277</td>\n",
       "      <td>ENST00000681903.1</td>\n",
       "      <td>+</td>\n",
       "      <td>179708271</td>\n",
       "      <td>179708470</td>\n",
       "      <td>179708268</td>\n",
       "      <td>179708269</td>\n",
       "      <td>rs1581856354</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38181</th>\n",
       "      <td>chr5</td>\n",
       "      <td>179708198</td>\n",
       "      <td>179708277</td>\n",
       "      <td>ENST00000681903.1</td>\n",
       "      <td>+</td>\n",
       "      <td>179708271</td>\n",
       "      <td>179708470</td>\n",
       "      <td>179708269</td>\n",
       "      <td>179708270</td>\n",
       "      <td>rs759471723</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>G,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38182</th>\n",
       "      <td>chr5</td>\n",
       "      <td>179708198</td>\n",
       "      <td>179708277</td>\n",
       "      <td>ENST00000681903.1</td>\n",
       "      <td>+</td>\n",
       "      <td>179708271</td>\n",
       "      <td>179708470</td>\n",
       "      <td>179708270</td>\n",
       "      <td>179708271</td>\n",
       "      <td>rs1175755473</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>C,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38183</th>\n",
       "      <td>chr5</td>\n",
       "      <td>179708198</td>\n",
       "      <td>179708277</td>\n",
       "      <td>ENST00000681903.1</td>\n",
       "      <td>+</td>\n",
       "      <td>179708271</td>\n",
       "      <td>179708470</td>\n",
       "      <td>179708272</td>\n",
       "      <td>179708273</td>\n",
       "      <td>rs767797862</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>A,T,</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38184 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr_name  Acceptor_start  Acceptor_end      Transcript_id Strand  \\\n",
       "0         chr2       108897190     108897269  ENST00000258443.7      -   \n",
       "1         chr2       108897190     108897269  ENST00000258443.7      -   \n",
       "2         chr2       108897190     108897269  ENST00000258443.7      -   \n",
       "3         chr2       108897190     108897269  ENST00000258443.7      -   \n",
       "4         chr2       108897190     108897269  ENST00000258443.7      -   \n",
       "...        ...             ...           ...                ...    ...   \n",
       "38179     chr5       179708198     179708277  ENST00000681903.1      +   \n",
       "38180     chr5       179708198     179708277  ENST00000681903.1      +   \n",
       "38181     chr5       179708198     179708277  ENST00000681903.1      +   \n",
       "38182     chr5       179708198     179708277  ENST00000681903.1      +   \n",
       "38183     chr5       179708198     179708277  ENST00000681903.1      +   \n",
       "\n",
       "       Enhancer_start  Enhancer_end  Variant_start  Variant_end         RS_ID  \\\n",
       "0           108896998     108897197      108897187    108897193  rs1574362124   \n",
       "1           108896998     108897197      108897191    108897192   rs566530415   \n",
       "2           108896998     108897197      108897193    108897194   rs121908456   \n",
       "3           108896998     108897197      108897194    108897195   rs777416175   \n",
       "4           108896998     108897197      108897206    108897207   rs770460028   \n",
       "...               ...           ...            ...          ...           ...   \n",
       "38179       179708271     179708470      179708265    179708266   rs774523334   \n",
       "38180       179708271     179708470      179708268    179708269  rs1581856354   \n",
       "38181       179708271     179708470      179708269    179708270   rs759471723   \n",
       "38182       179708271     179708470      179708270    179708271  rs1175755473   \n",
       "38183       179708271     179708470      179708272    179708273   rs767797862   \n",
       "\n",
       "      Reference_Nucleotide  No_of_variations Alternative_Nucleotides  \n",
       "0                   TCTTCT                 1                    TCT,  \n",
       "1                        C                 1                      A,  \n",
       "2                        C                 2                    A,T,  \n",
       "3                        G                 1                      A,  \n",
       "4                        T                 1                      C,  \n",
       "...                    ...               ...                     ...  \n",
       "38179                    A                 1                      C,  \n",
       "38180                    C                 1                      T,  \n",
       "38181                    A                 1                      G,  \n",
       "38182                    A                 1                      C,  \n",
       "38183                    G                 2                    A,T,  \n",
       "\n",
       "[38184 rows x 13 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2449d935-86ff-41a2-b8bd-c9a68b195d24",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
