{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9807e3a4-8edb-46a2-ad4c-88777af9fadc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, glob\n",
    "import pandas as pd\n",
    "from io import StringIO\n",
    "import numpy as np\n",
    "import pysam\n",
    "import pybedtools \n",
    "from Bio import SeqIO\n",
    "pybedtools.helpers.set_tempdir(\"/data/projects/temp\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2505ee28-f7a7-474a-a4b8-755989f3f70e",
   "metadata": {},
   "outputs": [],
   "source": [
    "non_coding_region = \"Donor\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6c3ba47b-0edf-43de-bab5-bc0edd7f91b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '/data/projects/Enhancer/RECOMB_2024/Intersected_data/{}/all_data.tsv'.format(non_coding_region)\n",
    "output_path = \"/data/projects/Enhancer/RECOMB_2024/SNP_data/{}\".format(non_coding_region) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "aafe0877-b7be-4e5e-a73f-00dab9627b51",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(output_path):\n",
    "    os.makedirs(output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cc89bac5-1528-4192-95d4-77bda6a6e1dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def seq2kmer(seq):\n",
    "    \"\"\"\n",
    "    Convert original sequence to kmers\n",
    "    \n",
    "    Arguments:\n",
    "    seq -- str, original sequence.\n",
    "    k -- int, kmer of length k specified.\n",
    "    \n",
    "    Returns:\n",
    "    kmers -- str, kmers separated by space\n",
    "    \"\"\"\n",
    "    k=6\n",
    "    kmer = [seq[x:x+k] for x in range(len(seq)+1-k)]\n",
    "    kmers = \" \".join(kmer)\n",
    "    return kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "19aaa7a0-7309-4e97-a93e-dded3164e07f",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = SeqIO.to_dict(SeqIO.parse(\"/data/projects/Resources/Gencode_genome_annotation/GRCh38.primary_assembly.genome.fa\", \"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0ded228f-2999-482f-9e5a-197b732e0cfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(36833, 13)\n",
      "(36283, 13)\n",
      "(36283, 13)\n",
      "(41185, 11)\n",
      "(38980, 12)\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(data_path, sep= '\\t')\n",
    "print(df.shape)\n",
    "df = df[df['No_of_variations']<3].reset_index(drop=True)\n",
    "print(df.shape)\n",
    "df = df.loc[df['Alternative_Nucleotides'] != '.'].reset_index(drop=True)\n",
    "print(df.shape)\n",
    "\n",
    "\n",
    "data= []\n",
    "#df = df.head(10)\n",
    "# Iterate over DataFrame rows\n",
    "for idx, row in df.iterrows():\n",
    "    try:\n",
    "        #print(row)\n",
    "        alts = row['Alternative_Nucleotides'].rstrip(',').split(',')\n",
    "        ref_seq = str(genome[row['chr_name']].seq[ row['Donor_start']:row['Donor_end']])\n",
    "\n",
    "        # Calculate the variant position relative to the fetched sequence\n",
    "        variant_pos_start = row['Variant_start'] - row['Donor_start']\n",
    "        variant_pos_end = row['Variant_end']  - row['Donor_start']\n",
    "        #print(variant_pos_start, variant_pos_end)\n",
    "\n",
    "\n",
    "        # Replace the reference nucleotide with the alternate nucleotide to get the alternate sequence\n",
    "        for alt in alts:\n",
    "            if alt == \"\":  # Deletion\n",
    "                # Fetch additional bases from genome to maintain sequence length\n",
    "                extra_bases = str(genome[row['chr_name']].seq[row['Donor_end']:row['Donor_end'] + len(row['Reference_Nucleotide'])])\n",
    "                #print(len(row['Reference_Nucleotide']))\n",
    "                alt_seq = ref_seq[:variant_pos_start] + ref_seq[variant_pos_end:] + extra_bases\n",
    "            else:  # SNPs, Insertions, and partial Deletions\n",
    "                extra_bases = str(genome[row['chr_name']].seq[row['Donor_end']:row['Donor_end'] + len(row['Reference_Nucleotide']) - len(alt)])\n",
    "                #print(extra_bases)\n",
    "                if(variant_pos_start>=0):\n",
    "                    alt_seq = ref_seq[:variant_pos_start] + alt + ref_seq[variant_pos_end:] + extra_bases\n",
    "                else:\n",
    "                    continue\n",
    "            # Append to the list as a dictionary\n",
    "            data.append({\n",
    "                'chr': row['chr_name'],\n",
    "                'strand': row['Strand'],\n",
    "                'Donor_coordinates': str(row['Donor_start'])+'-'+str(row['Donor_end']),\n",
    "                'Ensemble_Transcript_ID': row['Transcript_id'],\n",
    "                'dbsnp_id': row['RS_ID'],\n",
    "                'varinat_start':  row['Variant_start'],\n",
    "                'variant_end':  row['Variant_end'],\n",
    "                'ref_neucleotide': row['Reference_Nucleotide'],\n",
    "                'alternative_neucleotide': alt,\n",
    "                'reference_seq': ref_seq,\n",
    "                'alt_seq': alt_seq\n",
    "            })\n",
    "    except KeyError:\n",
    "        # If sequence is not present, append a default value\n",
    "        ref_sequences.append(\"NA\")\n",
    "        alt_sequences.append(\"NA\")\n",
    "\n",
    "        \n",
    "        \n",
    "new_df = pd.DataFrame(data)\n",
    "print(new_df.shape)\n",
    "new_df = new_df.drop_duplicates().reset_index()\n",
    "print(new_df.shape)\n",
    "merged_list = list(zip(new_df['reference_seq'], new_df['alt_seq']))\n",
    "merged_list = [item.upper() for tup in merged_list for item in tup]\n",
    "#print(merged_list)\n",
    "kmer_lst = list(map(seq2kmer, merged_list))\n",
    "df_kmer = pd.DataFrame(kmer_lst, columns=['Sequence'])\n",
    "df_kmer['Label'] = np.random.choice([0, 1], size=len(df_kmer))\n",
    "\n",
    "new_df.to_csv(output_path + \"/all_data_new.tsv\", sep=\"\\t\", index= False)\n",
    "df_kmer.to_csv(output_path + \"/dev.tsv\", sep=\"\\t\", index= False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7f6ff90c-19f5-4d83-b2ab-7dbcf50a30f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTCGAT TCGATT CGATTT GATTTG ATTTGT TTTGTT TTGT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCCGAT CCGATT CGATTT GATTTG ATTTGT TTTGTT TTGT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTCGAT TCGATT CGATTT GATTTG ATTTGT TTTGTT TTGT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTGGAT TGGATT GGATTT GATTTG ATTTGT TTTGTT TTGT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTCGAT TCGATT CGATTT GATTTG ATTTGT TTTGTT TTGT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77955</th>\n",
       "      <td>AAGCCC AGCCCT GCCCTT CCCTTC CCTTCC CTTCCT TTCC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77956</th>\n",
       "      <td>AAGCCC AGCCCT GCCCTT CCCTTC CCTTCC CTTCCT TTCC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77957</th>\n",
       "      <td>AAGCCC AGCCCT GCCCTT CCCTTC CCTTCC CTTCCT TTCC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77958</th>\n",
       "      <td>AAGCCC AGCCCT GCCCTT CCCTTC CCTTCC CTTCCT TTCC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77959</th>\n",
       "      <td>AAGCCC AGCCCT GCCCTT CCCTTC CCTTCC CTTCCT TTCC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77960 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Sequence  Label\n",
       "0      TTCGAT TCGATT CGATTT GATTTG ATTTGT TTTGTT TTGT...      1\n",
       "1      TCCGAT CCGATT CGATTT GATTTG ATTTGT TTTGTT TTGT...      1\n",
       "2      TTCGAT TCGATT CGATTT GATTTG ATTTGT TTTGTT TTGT...      0\n",
       "3      TTGGAT TGGATT GGATTT GATTTG ATTTGT TTTGTT TTGT...      0\n",
       "4      TTCGAT TCGATT CGATTT GATTTG ATTTGT TTTGTT TTGT...      1\n",
       "...                                                  ...    ...\n",
       "77955  AAGCCC AGCCCT GCCCTT CCCTTC CCTTCC CTTCCT TTCC...      1\n",
       "77956  AAGCCC AGCCCT GCCCTT CCCTTC CCTTCC CTTCCT TTCC...      1\n",
       "77957  AAGCCC AGCCCT GCCCTT CCCTTC CCTTCC CTTCCT TTCC...      1\n",
       "77958  AAGCCC AGCCCT GCCCTT CCCTTC CCTTCC CTTCCT TTCC...      0\n",
       "77959  AAGCCC AGCCCT GCCCTT CCCTTC CCTTCC CTTCCT TTCC...      1\n",
       "\n",
       "[77960 rows x 2 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_kmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2449d935-86ff-41a2-b8bd-c9a68b195d24",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
