{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "48bb6eb5-6de0-4d27-848d-5c2e2eb29faf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None)\n",
    "import pybedtools\n",
    "import numpy as np\n",
    "import os\n",
    "pybedtools.helpers.set_tempdir('/data/projects/temp')\n",
    "from IPython.display import display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "f4f24891-9cb6-4b3d-b75a-13c312c4bcf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gtf= pd.read_csv(\"/data/private/pdutta/Collab_data/Jim_work/mouse_gencode_annotation/gencode_gtf_annotation.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "096b8442-f1dd-47dc-abe6-2f836f7f7d31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>mgi_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102693.2</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1918292</td>\n",
       "      <td>OTTMUSG00000049935.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102693.2</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1918292</td>\n",
       "      <td>OTTMUSG00000049935.1</td>\n",
       "      <td>ENSMUST00000193812.2</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>OTTMUST00000127109.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102693.2</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1918292</td>\n",
       "      <td>OTTMUSG00000049935.1</td>\n",
       "      <td>ENSMUST00000193812.2</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>OTTMUST00000127109.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00001343744.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064842.3</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>Gm26206</td>\n",
       "      <td>3.0</td>\n",
       "      <td>MGI:5455983</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064842.3</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>Gm26206</td>\n",
       "      <td>3.0</td>\n",
       "      <td>MGI:5455983</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUST00000082908.3</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>Gm26206-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032379</th>\n",
       "      <td>chr9</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>108691464</td>\n",
       "      <td>108691550</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1931834</td>\n",
       "      <td>OTTMUSG00000050675.2</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000128467.1</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSMUSE00000221318.4</td>\n",
       "      <td>ENSMUSP00000035218.8</td>\n",
       "      <td>CCDS23538.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032380</th>\n",
       "      <td>chr9</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>108691629</td>\n",
       "      <td>108691767</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1931834</td>\n",
       "      <td>OTTMUSG00000050675.2</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000128467.1</td>\n",
       "      <td>8.0</td>\n",
       "      <td>ENSMUSE00000221321.4</td>\n",
       "      <td>ENSMUSP00000035218.8</td>\n",
       "      <td>CCDS23538.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032381</th>\n",
       "      <td>chr9</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>108691629</td>\n",
       "      <td>108691767</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1931834</td>\n",
       "      <td>OTTMUSG00000050675.2</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000128467.1</td>\n",
       "      <td>8.0</td>\n",
       "      <td>ENSMUSE00000221321.4</td>\n",
       "      <td>ENSMUSP00000035218.8</td>\n",
       "      <td>CCDS23538.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032382</th>\n",
       "      <td>chr9</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>108691866</td>\n",
       "      <td>108691946</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1931834</td>\n",
       "      <td>OTTMUSG00000050675.2</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000128467.1</td>\n",
       "      <td>9.0</td>\n",
       "      <td>ENSMUSE00000221323.4</td>\n",
       "      <td>ENSMUSP00000035218.8</td>\n",
       "      <td>CCDS23538.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032383</th>\n",
       "      <td>chr9</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>108691866</td>\n",
       "      <td>108691946</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1032384 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname   source     feature      start        end score strand frame  \\\n",
       "0          chr1   HAVANA        gene    3143476    3144545     .      +     .   \n",
       "1          chr1   HAVANA  transcript    3143476    3144545     .      +     .   \n",
       "2          chr1   HAVANA        exon    3143476    3144545     .      +     .   \n",
       "3          chr1  ENSEMBL        gene    3172239    3172348     .      +     .   \n",
       "4          chr1  ENSEMBL  transcript    3172239    3172348     .      +     .   \n",
       "...         ...      ...         ...        ...        ...   ...    ...   ...   \n",
       "1032379    chr9   HAVANA         CDS  108691464  108691550     .      +     0   \n",
       "1032380    chr9   HAVANA        exon  108691629  108691767     .      +     .   \n",
       "1032381    chr9   HAVANA         CDS  108691629  108691767     .      +     0   \n",
       "1032382    chr9   HAVANA        exon  108691866  108691946     .      +     .   \n",
       "1032383    chr9   HAVANA         CDS  108691866  108691946     .      +     2   \n",
       "\n",
       "                      gene_id       gene_type      gene_name  level  \\\n",
       "0        ENSMUSG00000102693.2             TEC  4933401J01Rik    2.0   \n",
       "1        ENSMUSG00000102693.2             TEC  4933401J01Rik    2.0   \n",
       "2        ENSMUSG00000102693.2             TEC  4933401J01Rik    2.0   \n",
       "3        ENSMUSG00000064842.3           snRNA        Gm26206    3.0   \n",
       "4        ENSMUSG00000064842.3           snRNA        Gm26206    3.0   \n",
       "...                       ...             ...            ...    ...   \n",
       "1032379  ENSMUSG00000032598.9  protein_coding        Nckipsd    2.0   \n",
       "1032380  ENSMUSG00000032598.9  protein_coding        Nckipsd    2.0   \n",
       "1032381  ENSMUSG00000032598.9  protein_coding        Nckipsd    2.0   \n",
       "1032382  ENSMUSG00000032598.9  protein_coding        Nckipsd    2.0   \n",
       "1032383  ENSMUSG00000032598.9  protein_coding        Nckipsd    NaN   \n",
       "\n",
       "              mgi_id           havana_gene         transcript_id  \\\n",
       "0        MGI:1918292  OTTMUSG00000049935.1                   NaN   \n",
       "1        MGI:1918292  OTTMUSG00000049935.1  ENSMUST00000193812.2   \n",
       "2        MGI:1918292  OTTMUSG00000049935.1  ENSMUST00000193812.2   \n",
       "3        MGI:5455983                   NaN                   NaN   \n",
       "4        MGI:5455983                   NaN  ENSMUST00000082908.3   \n",
       "...              ...                   ...                   ...   \n",
       "1032379  MGI:1931834  OTTMUSG00000050675.2  ENSMUST00000035218.9   \n",
       "1032380  MGI:1931834  OTTMUSG00000050675.2  ENSMUST00000035218.9   \n",
       "1032381  MGI:1931834  OTTMUSG00000050675.2  ENSMUST00000035218.9   \n",
       "1032382  MGI:1931834  OTTMUSG00000050675.2  ENSMUST00000035218.9   \n",
       "1032383          NaN                   NaN  ENSMUST00000035218.9   \n",
       "\n",
       "        transcript_type    transcript_name  transcript_support_level  \\\n",
       "0                   NaN                NaN                       NaN   \n",
       "1                   TEC  4933401J01Rik-201                       NaN   \n",
       "2                   TEC  4933401J01Rik-201                       NaN   \n",
       "3                   NaN                NaN                       NaN   \n",
       "4                 snRNA        Gm26206-201                       NaN   \n",
       "...                 ...                ...                       ...   \n",
       "1032379  protein_coding        Nckipsd-201                       1.0   \n",
       "1032380  protein_coding        Nckipsd-201                       1.0   \n",
       "1032381  protein_coding        Nckipsd-201                       1.0   \n",
       "1032382  protein_coding        Nckipsd-201                       1.0   \n",
       "1032383  protein_coding        Nckipsd-201                       NaN   \n",
       "\n",
       "                       tag     havana_transcript  exon_number  \\\n",
       "0                      NaN                   NaN          NaN   \n",
       "1        Ensembl_canonical  OTTMUST00000127109.1          NaN   \n",
       "2        Ensembl_canonical  OTTMUST00000127109.1          1.0   \n",
       "3                      NaN                   NaN          NaN   \n",
       "4        Ensembl_canonical                   NaN          NaN   \n",
       "...                    ...                   ...          ...   \n",
       "1032379               CCDS  OTTMUST00000128467.1          7.0   \n",
       "1032380               CCDS  OTTMUST00000128467.1          8.0   \n",
       "1032381               CCDS  OTTMUST00000128467.1          8.0   \n",
       "1032382               CCDS  OTTMUST00000128467.1          9.0   \n",
       "1032383                NaN                   NaN          9.0   \n",
       "\n",
       "                      exon_id            protein_id       ccdsid  ont  \n",
       "0                         NaN                   NaN          NaN  NaN  \n",
       "1                         NaN                   NaN          NaN  NaN  \n",
       "2        ENSMUSE00001343744.2                   NaN          NaN  NaN  \n",
       "3                         NaN                   NaN          NaN  NaN  \n",
       "4                         NaN                   NaN          NaN  NaN  \n",
       "...                       ...                   ...          ...  ...  \n",
       "1032379  ENSMUSE00000221318.4  ENSMUSP00000035218.8  CCDS23538.1  NaN  \n",
       "1032380  ENSMUSE00000221321.4  ENSMUSP00000035218.8  CCDS23538.1  NaN  \n",
       "1032381  ENSMUSE00000221321.4  ENSMUSP00000035218.8  CCDS23538.1  NaN  \n",
       "1032382  ENSMUSE00000221323.4  ENSMUSP00000035218.8  CCDS23538.1  NaN  \n",
       "1032383                   NaN                   NaN          NaN  NaN  \n",
       "\n",
       "[1032384 rows x 25 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "d68ca5e0-898c-48b2-92f0-8eceb0ae77fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "feature\n",
       "CDS               288589\n",
       "Selenocysteine        18\n",
       "UTR               101311\n",
       "exon              469432\n",
       "gene               29795\n",
       "start_codon        32674\n",
       "stop_codon         30315\n",
       "transcript         80250\n",
       "dtype: int64"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gtf.groupby('feature').size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "4aab1fc6-a98e-46e6-9391-fdde3812df32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>mgi_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102693.2</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1918292</td>\n",
       "      <td>OTTMUSG00000049935.1</td>\n",
       "      <td>ENSMUST00000193812.2</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>OTTMUST00000127109.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00001343744.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064842.3</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>Gm26206</td>\n",
       "      <td>3.0</td>\n",
       "      <td>MGI:5455983</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUST00000082908.3</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>Gm26206-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00000522066.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3283832</td>\n",
       "      <td>3286567</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000051951.6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td>ENSMUST00000162897.2</td>\n",
       "      <td>protein_coding_CDS_not_defined</td>\n",
       "      <td>Xkr4-203</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000086625.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00000858910.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3276124</td>\n",
       "      <td>3277540</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000051951.6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td>ENSMUST00000162897.2</td>\n",
       "      <td>protein_coding_CDS_not_defined</td>\n",
       "      <td>Xkr4-203</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000086625.1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSMUSE00000866652.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3283662</td>\n",
       "      <td>3285855</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000051951.6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td>ENSMUST00000159265.2</td>\n",
       "      <td>protein_coding_CDS_not_defined</td>\n",
       "      <td>Xkr4-202</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000086624.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00000863980.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032374</th>\n",
       "      <td>chr9</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>108689418</td>\n",
       "      <td>108689908</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1931834</td>\n",
       "      <td>OTTMUSG00000050675.2</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000128467.1</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSMUSE00000472692.3</td>\n",
       "      <td>ENSMUSP00000035218.8</td>\n",
       "      <td>CCDS23538.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032376</th>\n",
       "      <td>chr9</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>108691148</td>\n",
       "      <td>108691318</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1931834</td>\n",
       "      <td>OTTMUSG00000050675.2</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000128467.1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSMUSE00000515905.3</td>\n",
       "      <td>ENSMUSP00000035218.8</td>\n",
       "      <td>CCDS23538.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032378</th>\n",
       "      <td>chr9</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>108691464</td>\n",
       "      <td>108691550</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1931834</td>\n",
       "      <td>OTTMUSG00000050675.2</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000128467.1</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSMUSE00000221318.4</td>\n",
       "      <td>ENSMUSP00000035218.8</td>\n",
       "      <td>CCDS23538.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032380</th>\n",
       "      <td>chr9</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>108691629</td>\n",
       "      <td>108691767</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1931834</td>\n",
       "      <td>OTTMUSG00000050675.2</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000128467.1</td>\n",
       "      <td>8.0</td>\n",
       "      <td>ENSMUSE00000221321.4</td>\n",
       "      <td>ENSMUSP00000035218.8</td>\n",
       "      <td>CCDS23538.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032382</th>\n",
       "      <td>chr9</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>108691866</td>\n",
       "      <td>108691946</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1931834</td>\n",
       "      <td>OTTMUSG00000050675.2</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000128467.1</td>\n",
       "      <td>9.0</td>\n",
       "      <td>ENSMUSE00000221323.4</td>\n",
       "      <td>ENSMUSP00000035218.8</td>\n",
       "      <td>CCDS23538.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>469432 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname   source feature      start        end score strand frame  \\\n",
       "2          chr1   HAVANA    exon    3143476    3144545     .      +     .   \n",
       "5          chr1  ENSEMBL    exon    3172239    3172348     .      +     .   \n",
       "8          chr1   HAVANA    exon    3283832    3286567     .      -     .   \n",
       "9          chr1   HAVANA    exon    3276124    3277540     .      -     .   \n",
       "11         chr1   HAVANA    exon    3283662    3285855     .      -     .   \n",
       "...         ...      ...     ...        ...        ...   ...    ...   ...   \n",
       "1032374    chr9   HAVANA    exon  108689418  108689908     .      +     .   \n",
       "1032376    chr9   HAVANA    exon  108691148  108691318     .      +     .   \n",
       "1032378    chr9   HAVANA    exon  108691464  108691550     .      +     .   \n",
       "1032380    chr9   HAVANA    exon  108691629  108691767     .      +     .   \n",
       "1032382    chr9   HAVANA    exon  108691866  108691946     .      +     .   \n",
       "\n",
       "                      gene_id       gene_type      gene_name  level  \\\n",
       "2        ENSMUSG00000102693.2             TEC  4933401J01Rik    2.0   \n",
       "5        ENSMUSG00000064842.3           snRNA        Gm26206    3.0   \n",
       "8        ENSMUSG00000051951.6  protein_coding           Xkr4    2.0   \n",
       "9        ENSMUSG00000051951.6  protein_coding           Xkr4    2.0   \n",
       "11       ENSMUSG00000051951.6  protein_coding           Xkr4    2.0   \n",
       "...                       ...             ...            ...    ...   \n",
       "1032374  ENSMUSG00000032598.9  protein_coding        Nckipsd    2.0   \n",
       "1032376  ENSMUSG00000032598.9  protein_coding        Nckipsd    2.0   \n",
       "1032378  ENSMUSG00000032598.9  protein_coding        Nckipsd    2.0   \n",
       "1032380  ENSMUSG00000032598.9  protein_coding        Nckipsd    2.0   \n",
       "1032382  ENSMUSG00000032598.9  protein_coding        Nckipsd    2.0   \n",
       "\n",
       "              mgi_id           havana_gene         transcript_id  \\\n",
       "2        MGI:1918292  OTTMUSG00000049935.1  ENSMUST00000193812.2   \n",
       "5        MGI:5455983                   NaN  ENSMUST00000082908.3   \n",
       "8        MGI:3528744  OTTMUSG00000026353.2  ENSMUST00000162897.2   \n",
       "9        MGI:3528744  OTTMUSG00000026353.2  ENSMUST00000162897.2   \n",
       "11       MGI:3528744  OTTMUSG00000026353.2  ENSMUST00000159265.2   \n",
       "...              ...                   ...                   ...   \n",
       "1032374  MGI:1931834  OTTMUSG00000050675.2  ENSMUST00000035218.9   \n",
       "1032376  MGI:1931834  OTTMUSG00000050675.2  ENSMUST00000035218.9   \n",
       "1032378  MGI:1931834  OTTMUSG00000050675.2  ENSMUST00000035218.9   \n",
       "1032380  MGI:1931834  OTTMUSG00000050675.2  ENSMUST00000035218.9   \n",
       "1032382  MGI:1931834  OTTMUSG00000050675.2  ENSMUST00000035218.9   \n",
       "\n",
       "                        transcript_type    transcript_name  \\\n",
       "2                                   TEC  4933401J01Rik-201   \n",
       "5                                 snRNA        Gm26206-201   \n",
       "8        protein_coding_CDS_not_defined           Xkr4-203   \n",
       "9        protein_coding_CDS_not_defined           Xkr4-203   \n",
       "11       protein_coding_CDS_not_defined           Xkr4-202   \n",
       "...                                 ...                ...   \n",
       "1032374                  protein_coding        Nckipsd-201   \n",
       "1032376                  protein_coding        Nckipsd-201   \n",
       "1032378                  protein_coding        Nckipsd-201   \n",
       "1032380                  protein_coding        Nckipsd-201   \n",
       "1032382                  protein_coding        Nckipsd-201   \n",
       "\n",
       "         transcript_support_level                tag     havana_transcript  \\\n",
       "2                             NaN  Ensembl_canonical  OTTMUST00000127109.1   \n",
       "5                             NaN  Ensembl_canonical                   NaN   \n",
       "8                             1.0                NaN  OTTMUST00000086625.1   \n",
       "9                             1.0                NaN  OTTMUST00000086625.1   \n",
       "11                            1.0                NaN  OTTMUST00000086624.1   \n",
       "...                           ...                ...                   ...   \n",
       "1032374                       1.0               CCDS  OTTMUST00000128467.1   \n",
       "1032376                       1.0               CCDS  OTTMUST00000128467.1   \n",
       "1032378                       1.0               CCDS  OTTMUST00000128467.1   \n",
       "1032380                       1.0               CCDS  OTTMUST00000128467.1   \n",
       "1032382                       1.0               CCDS  OTTMUST00000128467.1   \n",
       "\n",
       "         exon_number               exon_id            protein_id       ccdsid  \\\n",
       "2                1.0  ENSMUSE00001343744.2                   NaN          NaN   \n",
       "5                1.0  ENSMUSE00000522066.2                   NaN          NaN   \n",
       "8                1.0  ENSMUSE00000858910.2                   NaN          NaN   \n",
       "9                2.0  ENSMUSE00000866652.2                   NaN          NaN   \n",
       "11               1.0  ENSMUSE00000863980.2                   NaN          NaN   \n",
       "...              ...                   ...                   ...          ...   \n",
       "1032374          5.0  ENSMUSE00000472692.3  ENSMUSP00000035218.8  CCDS23538.1   \n",
       "1032376          6.0  ENSMUSE00000515905.3  ENSMUSP00000035218.8  CCDS23538.1   \n",
       "1032378          7.0  ENSMUSE00000221318.4  ENSMUSP00000035218.8  CCDS23538.1   \n",
       "1032380          8.0  ENSMUSE00000221321.4  ENSMUSP00000035218.8  CCDS23538.1   \n",
       "1032382          9.0  ENSMUSE00000221323.4  ENSMUSP00000035218.8  CCDS23538.1   \n",
       "\n",
       "         ont  \n",
       "2        NaN  \n",
       "5        NaN  \n",
       "8        NaN  \n",
       "9        NaN  \n",
       "11       NaN  \n",
       "...      ...  \n",
       "1032374  NaN  \n",
       "1032376  NaN  \n",
       "1032378  NaN  \n",
       "1032380  NaN  \n",
       "1032382  NaN  \n",
       "\n",
       "[469432 rows x 25 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_exon = df_gtf[df_gtf['feature']==\"exon\"]\n",
    "df_exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "25e2829a-4957-4bcf-b1c8-8b941a63496e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "source\n",
       "ENSEMBL     71786\n",
       "HAVANA     960598\n",
       "dtype: int64"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gtf.groupby('source').size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "6c347b5f-a161-4b26-8029-5af79de31eef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>mgi_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102693.2</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1918292</td>\n",
       "      <td>OTTMUSG00000049935.1</td>\n",
       "      <td>ENSMUST00000193812.2</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>OTTMUST00000127109.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00001343744.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064842.3</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>Gm26206</td>\n",
       "      <td>3.0</td>\n",
       "      <td>MGI:5455983</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUST00000082908.3</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>Gm26206-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00000522066.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3283832</td>\n",
       "      <td>3286567</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000051951.6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td>ENSMUST00000162897.2</td>\n",
       "      <td>protein_coding_CDS_not_defined</td>\n",
       "      <td>Xkr4-203</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000086625.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00000858910.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3276124</td>\n",
       "      <td>3277540</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000051951.6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td>ENSMUST00000162897.2</td>\n",
       "      <td>protein_coding_CDS_not_defined</td>\n",
       "      <td>Xkr4-203</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000086625.1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSMUSE00000866652.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3283662</td>\n",
       "      <td>3285855</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000051951.6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td>ENSMUST00000159265.2</td>\n",
       "      <td>protein_coding_CDS_not_defined</td>\n",
       "      <td>Xkr4-202</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000086624.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00000863980.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032374</th>\n",
       "      <td>chr9</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>108689418</td>\n",
       "      <td>108689908</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1931834</td>\n",
       "      <td>OTTMUSG00000050675.2</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000128467.1</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSMUSE00000472692.3</td>\n",
       "      <td>ENSMUSP00000035218.8</td>\n",
       "      <td>CCDS23538.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032376</th>\n",
       "      <td>chr9</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>108691148</td>\n",
       "      <td>108691318</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1931834</td>\n",
       "      <td>OTTMUSG00000050675.2</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000128467.1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSMUSE00000515905.3</td>\n",
       "      <td>ENSMUSP00000035218.8</td>\n",
       "      <td>CCDS23538.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032378</th>\n",
       "      <td>chr9</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>108691464</td>\n",
       "      <td>108691550</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1931834</td>\n",
       "      <td>OTTMUSG00000050675.2</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000128467.1</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSMUSE00000221318.4</td>\n",
       "      <td>ENSMUSP00000035218.8</td>\n",
       "      <td>CCDS23538.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032380</th>\n",
       "      <td>chr9</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>108691629</td>\n",
       "      <td>108691767</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1931834</td>\n",
       "      <td>OTTMUSG00000050675.2</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000128467.1</td>\n",
       "      <td>8.0</td>\n",
       "      <td>ENSMUSE00000221321.4</td>\n",
       "      <td>ENSMUSP00000035218.8</td>\n",
       "      <td>CCDS23538.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032382</th>\n",
       "      <td>chr9</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>108691866</td>\n",
       "      <td>108691946</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1931834</td>\n",
       "      <td>OTTMUSG00000050675.2</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000128467.1</td>\n",
       "      <td>9.0</td>\n",
       "      <td>ENSMUSE00000221323.4</td>\n",
       "      <td>ENSMUSP00000035218.8</td>\n",
       "      <td>CCDS23538.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>245964 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname   source feature      start        end score strand frame  \\\n",
       "2          chr1   HAVANA    exon    3143476    3144545     .      +     .   \n",
       "5          chr1  ENSEMBL    exon    3172239    3172348     .      +     .   \n",
       "8          chr1   HAVANA    exon    3283832    3286567     .      -     .   \n",
       "9          chr1   HAVANA    exon    3276124    3277540     .      -     .   \n",
       "11         chr1   HAVANA    exon    3283662    3285855     .      -     .   \n",
       "...         ...      ...     ...        ...        ...   ...    ...   ...   \n",
       "1032374    chr9   HAVANA    exon  108689418  108689908     .      +     .   \n",
       "1032376    chr9   HAVANA    exon  108691148  108691318     .      +     .   \n",
       "1032378    chr9   HAVANA    exon  108691464  108691550     .      +     .   \n",
       "1032380    chr9   HAVANA    exon  108691629  108691767     .      +     .   \n",
       "1032382    chr9   HAVANA    exon  108691866  108691946     .      +     .   \n",
       "\n",
       "                      gene_id       gene_type      gene_name  level  \\\n",
       "2        ENSMUSG00000102693.2             TEC  4933401J01Rik    2.0   \n",
       "5        ENSMUSG00000064842.3           snRNA        Gm26206    3.0   \n",
       "8        ENSMUSG00000051951.6  protein_coding           Xkr4    2.0   \n",
       "9        ENSMUSG00000051951.6  protein_coding           Xkr4    2.0   \n",
       "11       ENSMUSG00000051951.6  protein_coding           Xkr4    2.0   \n",
       "...                       ...             ...            ...    ...   \n",
       "1032374  ENSMUSG00000032598.9  protein_coding        Nckipsd    2.0   \n",
       "1032376  ENSMUSG00000032598.9  protein_coding        Nckipsd    2.0   \n",
       "1032378  ENSMUSG00000032598.9  protein_coding        Nckipsd    2.0   \n",
       "1032380  ENSMUSG00000032598.9  protein_coding        Nckipsd    2.0   \n",
       "1032382  ENSMUSG00000032598.9  protein_coding        Nckipsd    2.0   \n",
       "\n",
       "              mgi_id           havana_gene         transcript_id  \\\n",
       "2        MGI:1918292  OTTMUSG00000049935.1  ENSMUST00000193812.2   \n",
       "5        MGI:5455983                   NaN  ENSMUST00000082908.3   \n",
       "8        MGI:3528744  OTTMUSG00000026353.2  ENSMUST00000162897.2   \n",
       "9        MGI:3528744  OTTMUSG00000026353.2  ENSMUST00000162897.2   \n",
       "11       MGI:3528744  OTTMUSG00000026353.2  ENSMUST00000159265.2   \n",
       "...              ...                   ...                   ...   \n",
       "1032374  MGI:1931834  OTTMUSG00000050675.2  ENSMUST00000035218.9   \n",
       "1032376  MGI:1931834  OTTMUSG00000050675.2  ENSMUST00000035218.9   \n",
       "1032378  MGI:1931834  OTTMUSG00000050675.2  ENSMUST00000035218.9   \n",
       "1032380  MGI:1931834  OTTMUSG00000050675.2  ENSMUST00000035218.9   \n",
       "1032382  MGI:1931834  OTTMUSG00000050675.2  ENSMUST00000035218.9   \n",
       "\n",
       "                        transcript_type    transcript_name  \\\n",
       "2                                   TEC  4933401J01Rik-201   \n",
       "5                                 snRNA        Gm26206-201   \n",
       "8        protein_coding_CDS_not_defined           Xkr4-203   \n",
       "9        protein_coding_CDS_not_defined           Xkr4-203   \n",
       "11       protein_coding_CDS_not_defined           Xkr4-202   \n",
       "...                                 ...                ...   \n",
       "1032374                  protein_coding        Nckipsd-201   \n",
       "1032376                  protein_coding        Nckipsd-201   \n",
       "1032378                  protein_coding        Nckipsd-201   \n",
       "1032380                  protein_coding        Nckipsd-201   \n",
       "1032382                  protein_coding        Nckipsd-201   \n",
       "\n",
       "         transcript_support_level                tag     havana_transcript  \\\n",
       "2                             NaN  Ensembl_canonical  OTTMUST00000127109.1   \n",
       "5                             NaN  Ensembl_canonical                   NaN   \n",
       "8                             1.0                NaN  OTTMUST00000086625.1   \n",
       "9                             1.0                NaN  OTTMUST00000086625.1   \n",
       "11                            1.0                NaN  OTTMUST00000086624.1   \n",
       "...                           ...                ...                   ...   \n",
       "1032374                       1.0               CCDS  OTTMUST00000128467.1   \n",
       "1032376                       1.0               CCDS  OTTMUST00000128467.1   \n",
       "1032378                       1.0               CCDS  OTTMUST00000128467.1   \n",
       "1032380                       1.0               CCDS  OTTMUST00000128467.1   \n",
       "1032382                       1.0               CCDS  OTTMUST00000128467.1   \n",
       "\n",
       "         exon_number               exon_id            protein_id       ccdsid  \\\n",
       "2                1.0  ENSMUSE00001343744.2                   NaN          NaN   \n",
       "5                1.0  ENSMUSE00000522066.2                   NaN          NaN   \n",
       "8                1.0  ENSMUSE00000858910.2                   NaN          NaN   \n",
       "9                2.0  ENSMUSE00000866652.2                   NaN          NaN   \n",
       "11               1.0  ENSMUSE00000863980.2                   NaN          NaN   \n",
       "...              ...                   ...                   ...          ...   \n",
       "1032374          5.0  ENSMUSE00000472692.3  ENSMUSP00000035218.8  CCDS23538.1   \n",
       "1032376          6.0  ENSMUSE00000515905.3  ENSMUSP00000035218.8  CCDS23538.1   \n",
       "1032378          7.0  ENSMUSE00000221318.4  ENSMUSP00000035218.8  CCDS23538.1   \n",
       "1032380          8.0  ENSMUSE00000221321.4  ENSMUSP00000035218.8  CCDS23538.1   \n",
       "1032382          9.0  ENSMUSE00000221323.4  ENSMUSP00000035218.8  CCDS23538.1   \n",
       "\n",
       "         ont  \n",
       "2        NaN  \n",
       "5        NaN  \n",
       "8        NaN  \n",
       "9        NaN  \n",
       "11       NaN  \n",
       "...      ...  \n",
       "1032374  NaN  \n",
       "1032376  NaN  \n",
       "1032378  NaN  \n",
       "1032380  NaN  \n",
       "1032382  NaN  \n",
       "\n",
       "[245964 rows x 25 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_exon.drop_duplicates(['start','end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "680bdb2c-7607-41fb-8f50-14283d829a07",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/projects/temp/ipykernel_3503301/1634419950.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_exon['unique_id'] = df_exon['gene_id'] + '_' + df_exon['start'].astype(str) + '_' + df_exon['end'].astype(str)\n"
     ]
    }
   ],
   "source": [
    "df_exon['unique_id'] = df_exon['gene_id'] + '_' + df_exon['start'].astype(str) + '_' + df_exon['end'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "d2660dc4-bd9c-4efc-9ac6-7d565e98eed6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get a list of the existing column names\n",
    "cols = df_exon.columns.tolist()\n",
    "\n",
    "# Remove 'start' and 'end' from their current positions in the list\n",
    "cols.remove('start')\n",
    "cols.remove('end')\n",
    "\n",
    "# Insert 'start' and 'end' at the 2nd and 3rd positions\n",
    "cols.insert(1, 'start')\n",
    "cols.insert(2, 'end')\n",
    "\n",
    "# Reorder the DataFrame columns using the modified list of column names\n",
    "df_exon = df_exon[cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "32a4bf75-84c9-4276-a501-eb974d73aba8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>mgi_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>ont</th>\n",
       "      <th>unique_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102693.2</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1918292</td>\n",
       "      <td>OTTMUSG00000049935.1</td>\n",
       "      <td>ENSMUST00000193812.2</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>OTTMUST00000127109.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00001343744.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000102693.2_3143476_3144545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064842.3</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>Gm26206</td>\n",
       "      <td>3.0</td>\n",
       "      <td>MGI:5455983</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUST00000082908.3</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>Gm26206-201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00000522066.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000064842.3_3172239_3172348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3283832</td>\n",
       "      <td>3286567</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000051951.6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td>ENSMUST00000162897.2</td>\n",
       "      <td>protein_coding_CDS_not_defined</td>\n",
       "      <td>Xkr4-203</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000086625.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00000858910.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000051951.6_3283832_3286567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3276124</td>\n",
       "      <td>3277540</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000051951.6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td>ENSMUST00000162897.2</td>\n",
       "      <td>protein_coding_CDS_not_defined</td>\n",
       "      <td>Xkr4-203</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000086625.1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSMUSE00000866652.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000051951.6_3276124_3277540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3283662</td>\n",
       "      <td>3285855</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000051951.6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td>ENSMUST00000159265.2</td>\n",
       "      <td>protein_coding_CDS_not_defined</td>\n",
       "      <td>Xkr4-202</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000086624.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00000863980.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000051951.6_3283662_3285855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032374</th>\n",
       "      <td>chr9</td>\n",
       "      <td>108689418</td>\n",
       "      <td>108689908</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1931834</td>\n",
       "      <td>OTTMUSG00000050675.2</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000128467.1</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSMUSE00000472692.3</td>\n",
       "      <td>ENSMUSP00000035218.8</td>\n",
       "      <td>CCDS23538.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000032598.9_108689418_108689908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032376</th>\n",
       "      <td>chr9</td>\n",
       "      <td>108691148</td>\n",
       "      <td>108691318</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1931834</td>\n",
       "      <td>OTTMUSG00000050675.2</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000128467.1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSMUSE00000515905.3</td>\n",
       "      <td>ENSMUSP00000035218.8</td>\n",
       "      <td>CCDS23538.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000032598.9_108691148_108691318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032378</th>\n",
       "      <td>chr9</td>\n",
       "      <td>108691464</td>\n",
       "      <td>108691550</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1931834</td>\n",
       "      <td>OTTMUSG00000050675.2</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000128467.1</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSMUSE00000221318.4</td>\n",
       "      <td>ENSMUSP00000035218.8</td>\n",
       "      <td>CCDS23538.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000032598.9_108691464_108691550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032380</th>\n",
       "      <td>chr9</td>\n",
       "      <td>108691629</td>\n",
       "      <td>108691767</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1931834</td>\n",
       "      <td>OTTMUSG00000050675.2</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000128467.1</td>\n",
       "      <td>8.0</td>\n",
       "      <td>ENSMUSE00000221321.4</td>\n",
       "      <td>ENSMUSP00000035218.8</td>\n",
       "      <td>CCDS23538.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000032598.9_108691629_108691767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032382</th>\n",
       "      <td>chr9</td>\n",
       "      <td>108691866</td>\n",
       "      <td>108691946</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000032598.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1931834</td>\n",
       "      <td>OTTMUSG00000050675.2</td>\n",
       "      <td>ENSMUST00000035218.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Nckipsd-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000128467.1</td>\n",
       "      <td>9.0</td>\n",
       "      <td>ENSMUSE00000221323.4</td>\n",
       "      <td>ENSMUSP00000035218.8</td>\n",
       "      <td>CCDS23538.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000032598.9_108691866_108691946</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>469432 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname      start        end   source feature score strand frame  \\\n",
       "2          chr1    3143476    3144545   HAVANA    exon     .      +     .   \n",
       "5          chr1    3172239    3172348  ENSEMBL    exon     .      +     .   \n",
       "8          chr1    3283832    3286567   HAVANA    exon     .      -     .   \n",
       "9          chr1    3276124    3277540   HAVANA    exon     .      -     .   \n",
       "11         chr1    3283662    3285855   HAVANA    exon     .      -     .   \n",
       "...         ...        ...        ...      ...     ...   ...    ...   ...   \n",
       "1032374    chr9  108689418  108689908   HAVANA    exon     .      +     .   \n",
       "1032376    chr9  108691148  108691318   HAVANA    exon     .      +     .   \n",
       "1032378    chr9  108691464  108691550   HAVANA    exon     .      +     .   \n",
       "1032380    chr9  108691629  108691767   HAVANA    exon     .      +     .   \n",
       "1032382    chr9  108691866  108691946   HAVANA    exon     .      +     .   \n",
       "\n",
       "                      gene_id       gene_type      gene_name  level  \\\n",
       "2        ENSMUSG00000102693.2             TEC  4933401J01Rik    2.0   \n",
       "5        ENSMUSG00000064842.3           snRNA        Gm26206    3.0   \n",
       "8        ENSMUSG00000051951.6  protein_coding           Xkr4    2.0   \n",
       "9        ENSMUSG00000051951.6  protein_coding           Xkr4    2.0   \n",
       "11       ENSMUSG00000051951.6  protein_coding           Xkr4    2.0   \n",
       "...                       ...             ...            ...    ...   \n",
       "1032374  ENSMUSG00000032598.9  protein_coding        Nckipsd    2.0   \n",
       "1032376  ENSMUSG00000032598.9  protein_coding        Nckipsd    2.0   \n",
       "1032378  ENSMUSG00000032598.9  protein_coding        Nckipsd    2.0   \n",
       "1032380  ENSMUSG00000032598.9  protein_coding        Nckipsd    2.0   \n",
       "1032382  ENSMUSG00000032598.9  protein_coding        Nckipsd    2.0   \n",
       "\n",
       "              mgi_id           havana_gene         transcript_id  \\\n",
       "2        MGI:1918292  OTTMUSG00000049935.1  ENSMUST00000193812.2   \n",
       "5        MGI:5455983                   NaN  ENSMUST00000082908.3   \n",
       "8        MGI:3528744  OTTMUSG00000026353.2  ENSMUST00000162897.2   \n",
       "9        MGI:3528744  OTTMUSG00000026353.2  ENSMUST00000162897.2   \n",
       "11       MGI:3528744  OTTMUSG00000026353.2  ENSMUST00000159265.2   \n",
       "...              ...                   ...                   ...   \n",
       "1032374  MGI:1931834  OTTMUSG00000050675.2  ENSMUST00000035218.9   \n",
       "1032376  MGI:1931834  OTTMUSG00000050675.2  ENSMUST00000035218.9   \n",
       "1032378  MGI:1931834  OTTMUSG00000050675.2  ENSMUST00000035218.9   \n",
       "1032380  MGI:1931834  OTTMUSG00000050675.2  ENSMUST00000035218.9   \n",
       "1032382  MGI:1931834  OTTMUSG00000050675.2  ENSMUST00000035218.9   \n",
       "\n",
       "                        transcript_type    transcript_name  \\\n",
       "2                                   TEC  4933401J01Rik-201   \n",
       "5                                 snRNA        Gm26206-201   \n",
       "8        protein_coding_CDS_not_defined           Xkr4-203   \n",
       "9        protein_coding_CDS_not_defined           Xkr4-203   \n",
       "11       protein_coding_CDS_not_defined           Xkr4-202   \n",
       "...                                 ...                ...   \n",
       "1032374                  protein_coding        Nckipsd-201   \n",
       "1032376                  protein_coding        Nckipsd-201   \n",
       "1032378                  protein_coding        Nckipsd-201   \n",
       "1032380                  protein_coding        Nckipsd-201   \n",
       "1032382                  protein_coding        Nckipsd-201   \n",
       "\n",
       "         transcript_support_level                tag     havana_transcript  \\\n",
       "2                             NaN  Ensembl_canonical  OTTMUST00000127109.1   \n",
       "5                             NaN  Ensembl_canonical                   NaN   \n",
       "8                             1.0                NaN  OTTMUST00000086625.1   \n",
       "9                             1.0                NaN  OTTMUST00000086625.1   \n",
       "11                            1.0                NaN  OTTMUST00000086624.1   \n",
       "...                           ...                ...                   ...   \n",
       "1032374                       1.0               CCDS  OTTMUST00000128467.1   \n",
       "1032376                       1.0               CCDS  OTTMUST00000128467.1   \n",
       "1032378                       1.0               CCDS  OTTMUST00000128467.1   \n",
       "1032380                       1.0               CCDS  OTTMUST00000128467.1   \n",
       "1032382                       1.0               CCDS  OTTMUST00000128467.1   \n",
       "\n",
       "         exon_number               exon_id            protein_id       ccdsid  \\\n",
       "2                1.0  ENSMUSE00001343744.2                   NaN          NaN   \n",
       "5                1.0  ENSMUSE00000522066.2                   NaN          NaN   \n",
       "8                1.0  ENSMUSE00000858910.2                   NaN          NaN   \n",
       "9                2.0  ENSMUSE00000866652.2                   NaN          NaN   \n",
       "11               1.0  ENSMUSE00000863980.2                   NaN          NaN   \n",
       "...              ...                   ...                   ...          ...   \n",
       "1032374          5.0  ENSMUSE00000472692.3  ENSMUSP00000035218.8  CCDS23538.1   \n",
       "1032376          6.0  ENSMUSE00000515905.3  ENSMUSP00000035218.8  CCDS23538.1   \n",
       "1032378          7.0  ENSMUSE00000221318.4  ENSMUSP00000035218.8  CCDS23538.1   \n",
       "1032380          8.0  ENSMUSE00000221321.4  ENSMUSP00000035218.8  CCDS23538.1   \n",
       "1032382          9.0  ENSMUSE00000221323.4  ENSMUSP00000035218.8  CCDS23538.1   \n",
       "\n",
       "         ont                                 unique_id  \n",
       "2        NaN      ENSMUSG00000102693.2_3143476_3144545  \n",
       "5        NaN      ENSMUSG00000064842.3_3172239_3172348  \n",
       "8        NaN      ENSMUSG00000051951.6_3283832_3286567  \n",
       "9        NaN      ENSMUSG00000051951.6_3276124_3277540  \n",
       "11       NaN      ENSMUSG00000051951.6_3283662_3285855  \n",
       "...      ...                                       ...  \n",
       "1032374  NaN  ENSMUSG00000032598.9_108689418_108689908  \n",
       "1032376  NaN  ENSMUSG00000032598.9_108691148_108691318  \n",
       "1032378  NaN  ENSMUSG00000032598.9_108691464_108691550  \n",
       "1032380  NaN  ENSMUSG00000032598.9_108691629_108691767  \n",
       "1032382  NaN  ENSMUSG00000032598.9_108691866_108691946  \n",
       "\n",
       "[469432 rows x 26 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "9a6fe6ff-a3f7-46d1-8b52-d2104f18f48e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_exon.to_csv(\"/data/private/pdutta/Collab_data/Jim_work/mouse_gencode_annotation/exon_bed.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "5e850c8c-6f8e-44f0-90dd-bd35580e854f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_name</th>\n",
       "      <th>CPD_start</th>\n",
       "      <th>CPD_end</th>\n",
       "      <th>chr_name_SNV</th>\n",
       "      <th>SNV_start</th>\n",
       "      <th>SNV_end</th>\n",
       "      <th>ID</th>\n",
       "      <th>SNV_Info</th>\n",
       "      <th>CPD_original_start</th>\n",
       "      <th>CPD_original_end</th>\n",
       "      <th>distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3610212</td>\n",
       "      <td>3610271</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3610212</td>\n",
       "      <td>3610213</td>\n",
       "      <td>334</td>\n",
       "      <td>N:AG[T&gt;G]TT</td>\n",
       "      <td>3610241</td>\n",
       "      <td>3610242</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4020737</td>\n",
       "      <td>4020796</td>\n",
       "      <td>chr1</td>\n",
       "      <td>4020795</td>\n",
       "      <td>4020796</td>\n",
       "      <td>344</td>\n",
       "      <td>N:TT[C&gt;T]GT</td>\n",
       "      <td>4020766</td>\n",
       "      <td>4020767</td>\n",
       "      <td>-29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4252174</td>\n",
       "      <td>4252233</td>\n",
       "      <td>chr1</td>\n",
       "      <td>4252232</td>\n",
       "      <td>4252233</td>\n",
       "      <td>334</td>\n",
       "      <td>T:TC[C&gt;T]CC</td>\n",
       "      <td>4252203</td>\n",
       "      <td>4252204</td>\n",
       "      <td>-29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5725397</td>\n",
       "      <td>5725456</td>\n",
       "      <td>chr1</td>\n",
       "      <td>5725397</td>\n",
       "      <td>5725398</td>\n",
       "      <td>341</td>\n",
       "      <td>N:TT[C&gt;T]TG</td>\n",
       "      <td>5725426</td>\n",
       "      <td>5725427</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5779579</td>\n",
       "      <td>5779638</td>\n",
       "      <td>chr1</td>\n",
       "      <td>5779579</td>\n",
       "      <td>5779580</td>\n",
       "      <td>334</td>\n",
       "      <td>N:TT[T&gt;C]TT</td>\n",
       "      <td>5779608</td>\n",
       "      <td>5779609</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288593</th>\n",
       "      <td>chrX</td>\n",
       "      <td>166005063</td>\n",
       "      <td>166005072</td>\n",
       "      <td>chrX</td>\n",
       "      <td>166005063</td>\n",
       "      <td>166005064</td>\n",
       "      <td>334</td>\n",
       "      <td>N:TT[C&gt;T]AC</td>\n",
       "      <td>166005067</td>\n",
       "      <td>166005068</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288594</th>\n",
       "      <td>chrX</td>\n",
       "      <td>166026319</td>\n",
       "      <td>166026328</td>\n",
       "      <td>chrX</td>\n",
       "      <td>166026319</td>\n",
       "      <td>166026320</td>\n",
       "      <td>334</td>\n",
       "      <td>U:TT[T&gt;A]TA</td>\n",
       "      <td>166026323</td>\n",
       "      <td>166026324</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288595</th>\n",
       "      <td>chrX</td>\n",
       "      <td>166581272</td>\n",
       "      <td>166581281</td>\n",
       "      <td>chrX</td>\n",
       "      <td>166581272</td>\n",
       "      <td>166581273</td>\n",
       "      <td>341</td>\n",
       "      <td>U:TT[C&gt;T]CT</td>\n",
       "      <td>166581276</td>\n",
       "      <td>166581277</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288596</th>\n",
       "      <td>chrX</td>\n",
       "      <td>167351471</td>\n",
       "      <td>167351480</td>\n",
       "      <td>chrX</td>\n",
       "      <td>167351471</td>\n",
       "      <td>167351472</td>\n",
       "      <td>341</td>\n",
       "      <td>N:AG[T&gt;A]AA</td>\n",
       "      <td>167351475</td>\n",
       "      <td>167351476</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288597</th>\n",
       "      <td>chrX</td>\n",
       "      <td>168363280</td>\n",
       "      <td>168363289</td>\n",
       "      <td>chrX</td>\n",
       "      <td>168363280</td>\n",
       "      <td>168363281</td>\n",
       "      <td>341</td>\n",
       "      <td>N:GT[C&gt;T]TC</td>\n",
       "      <td>168363284</td>\n",
       "      <td>168363285</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>288598 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr_name  CPD_start    CPD_end chr_name_SNV  SNV_start    SNV_end   ID  \\\n",
       "0          chr1    3610212    3610271         chr1    3610212    3610213  334   \n",
       "1          chr1    4020737    4020796         chr1    4020795    4020796  344   \n",
       "2          chr1    4252174    4252233         chr1    4252232    4252233  334   \n",
       "3          chr1    5725397    5725456         chr1    5725397    5725398  341   \n",
       "4          chr1    5779579    5779638         chr1    5779579    5779580  334   \n",
       "...         ...        ...        ...          ...        ...        ...  ...   \n",
       "288593     chrX  166005063  166005072         chrX  166005063  166005064  334   \n",
       "288594     chrX  166026319  166026328         chrX  166026319  166026320  334   \n",
       "288595     chrX  166581272  166581281         chrX  166581272  166581273  341   \n",
       "288596     chrX  167351471  167351480         chrX  167351471  167351472  341   \n",
       "288597     chrX  168363280  168363289         chrX  168363280  168363281  341   \n",
       "\n",
       "           SNV_Info  CPD_original_start  CPD_original_end  distance  \n",
       "0       N:AG[T>G]TT             3610241           3610242        29  \n",
       "1       N:TT[C>T]GT             4020766           4020767       -29  \n",
       "2       T:TC[C>T]CC             4252203           4252204       -29  \n",
       "3       N:TT[C>T]TG             5725426           5725427        29  \n",
       "4       N:TT[T>C]TT             5779608           5779609        29  \n",
       "...             ...                 ...               ...       ...  \n",
       "288593  N:TT[C>T]AC           166005067         166005068         4  \n",
       "288594  U:TT[T>A]TA           166026323         166026324         4  \n",
       "288595  U:TT[C>T]CT           166581276         166581277         4  \n",
       "288596  N:AG[T>A]AA           167351475         167351476         4  \n",
       "288597  N:GT[C>T]TC           168363284         168363285         4  \n",
       "\n",
       "[288598 rows x 11 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_intersected_CPD_SNV = pd.read_csv(\"/data/private/pdutta/Collab_data/Jim_work/Output/UV_Plus/UV_intersected_combined_CPD_SNV.tsv\", sep= \"\\t\")\n",
    "df_intersected_CPD_SNV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "f17e2570-20d9-4549-a842-3d78e717e828",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_exon_bed = pybedtools.BedTool.from_dataframe(df_exon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "50e2907a-fc2a-435f-b5bc-5fc413452448",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1501, 37)\n",
      "Unique CPDs : (728, 37)\n",
      "Unique SNVs : (728, 37)\n",
      "Unique exons : (1337, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(916, 37)\n",
      "Unique CPDs : (437, 37)\n",
      "Unique SNVs : (428, 37)\n",
      "Unique exons : (888, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(601, 37)\n",
      "Unique CPDs : (324, 37)\n",
      "Unique SNVs : (318, 37)\n",
      "Unique exons : (587, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(539, 37)\n",
      "Unique CPDs : (293, 37)\n",
      "Unique SNVs : (291, 37)\n",
      "Unique exons : (521, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(553, 37)\n",
      "Unique CPDs : (286, 37)\n",
      "Unique SNVs : (285, 37)\n",
      "Unique exons : (543, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(519, 37)\n",
      "Unique CPDs : (255, 37)\n",
      "Unique SNVs : (253, 37)\n",
      "Unique exons : (515, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(510, 37)\n",
      "Unique CPDs : (281, 37)\n",
      "Unique SNVs : (282, 37)\n",
      "Unique exons : (508, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(431, 37)\n",
      "Unique CPDs : (224, 37)\n",
      "Unique SNVs : (222, 37)\n",
      "Unique exons : (426, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(497, 37)\n",
      "Unique CPDs : (231, 37)\n",
      "Unique SNVs : (227, 37)\n",
      "Unique exons : (490, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(343, 37)\n",
      "Unique CPDs : (196, 37)\n",
      "Unique SNVs : (194, 37)\n",
      "Unique exons : (339, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(407, 37)\n",
      "Unique CPDs : (198, 37)\n",
      "Unique SNVs : (197, 37)\n",
      "Unique exons : (406, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(392, 37)\n",
      "Unique CPDs : (207, 37)\n",
      "Unique SNVs : (202, 37)\n",
      "Unique exons : (378, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(302, 37)\n",
      "Unique CPDs : (172, 37)\n",
      "Unique SNVs : (172, 37)\n",
      "Unique exons : (302, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(319, 37)\n",
      "Unique CPDs : (177, 37)\n",
      "Unique SNVs : (178, 37)\n",
      "Unique exons : (318, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(361, 37)\n",
      "Unique CPDs : (181, 37)\n",
      "Unique SNVs : (180, 37)\n",
      "Unique exons : (351, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(326, 37)\n",
      "Unique CPDs : (160, 37)\n",
      "Unique SNVs : (159, 37)\n",
      "Unique exons : (324, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(333, 37)\n",
      "Unique CPDs : (162, 37)\n",
      "Unique SNVs : (159, 37)\n",
      "Unique exons : (324, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(323, 37)\n",
      "Unique CPDs : (140, 37)\n",
      "Unique SNVs : (138, 37)\n",
      "Unique exons : (317, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(300, 37)\n",
      "Unique CPDs : (161, 37)\n",
      "Unique SNVs : (159, 37)\n",
      "Unique exons : (297, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(246, 37)\n",
      "Unique CPDs : (127, 37)\n",
      "Unique SNVs : (124, 37)\n",
      "Unique exons : (239, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(283, 37)\n",
      "Unique CPDs : (140, 37)\n",
      "Unique SNVs : (138, 37)\n",
      "Unique exons : (278, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(338, 37)\n",
      "Unique CPDs : (163, 37)\n",
      "Unique SNVs : (162, 37)\n",
      "Unique exons : (337, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(250, 37)\n",
      "Unique CPDs : (137, 37)\n",
      "Unique SNVs : (136, 37)\n",
      "Unique exons : (248, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(309, 37)\n",
      "Unique CPDs : (131, 37)\n",
      "Unique SNVs : (130, 37)\n",
      "Unique exons : (305, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(206, 37)\n",
      "Unique CPDs : (120, 37)\n",
      "Unique SNVs : (116, 37)\n",
      "Unique exons : (201, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(175, 37)\n",
      "Unique CPDs : (106, 37)\n",
      "Unique SNVs : (105, 37)\n",
      "Unique exons : (174, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(193, 37)\n",
      "Unique CPDs : (103, 37)\n",
      "Unique SNVs : (103, 37)\n",
      "Unique exons : (193, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(219, 37)\n",
      "Unique CPDs : (118, 37)\n",
      "Unique SNVs : (117, 37)\n",
      "Unique exons : (218, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(189, 37)\n",
      "Unique CPDs : (91, 37)\n",
      "Unique SNVs : (90, 37)\n",
      "Unique exons : (188, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(159, 37)\n",
      "Unique CPDs : (93, 37)\n",
      "Unique SNVs : (92, 37)\n",
      "Unique exons : (158, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(167, 37)\n",
      "Unique CPDs : (91, 37)\n",
      "Unique SNVs : (91, 37)\n",
      "Unique exons : (167, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(142, 37)\n",
      "Unique CPDs : (80, 37)\n",
      "Unique SNVs : (79, 37)\n",
      "Unique exons : (139, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(145, 37)\n",
      "Unique CPDs : (84, 37)\n",
      "Unique SNVs : (84, 37)\n",
      "Unique exons : (144, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(168, 37)\n",
      "Unique CPDs : (97, 37)\n",
      "Unique SNVs : (95, 37)\n",
      "Unique exons : (165, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(182, 37)\n",
      "Unique CPDs : (87, 37)\n",
      "Unique SNVs : (87, 37)\n",
      "Unique exons : (181, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(212, 37)\n",
      "Unique CPDs : (84, 37)\n",
      "Unique SNVs : (84, 37)\n",
      "Unique exons : (212, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(108, 37)\n",
      "Unique CPDs : (64, 37)\n",
      "Unique SNVs : (64, 37)\n",
      "Unique exons : (108, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(186, 37)\n",
      "Unique CPDs : (84, 37)\n",
      "Unique SNVs : (82, 37)\n",
      "Unique exons : (185, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(210, 37)\n",
      "Unique CPDs : (69, 37)\n",
      "Unique SNVs : (68, 37)\n",
      "Unique exons : (205, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(143, 37)\n",
      "Unique CPDs : (76, 37)\n",
      "Unique SNVs : (76, 37)\n",
      "Unique exons : (143, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(158, 37)\n",
      "Unique CPDs : (79, 37)\n",
      "Unique SNVs : (79, 37)\n",
      "Unique exons : (158, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(118, 37)\n",
      "Unique CPDs : (67, 37)\n",
      "Unique SNVs : (67, 37)\n",
      "Unique exons : (117, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(127, 37)\n",
      "Unique CPDs : (69, 37)\n",
      "Unique SNVs : (69, 37)\n",
      "Unique exons : (127, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(116, 37)\n",
      "Unique CPDs : (64, 37)\n",
      "Unique SNVs : (63, 37)\n",
      "Unique exons : (115, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(135, 37)\n",
      "Unique CPDs : (65, 37)\n",
      "Unique SNVs : (65, 37)\n",
      "Unique exons : (135, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(143, 37)\n",
      "Unique CPDs : (72, 37)\n",
      "Unique SNVs : (72, 37)\n",
      "Unique exons : (143, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(97, 37)\n",
      "Unique CPDs : (58, 37)\n",
      "Unique SNVs : (57, 37)\n",
      "Unique exons : (96, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(114, 37)\n",
      "Unique CPDs : (59, 37)\n",
      "Unique SNVs : (59, 37)\n",
      "Unique exons : (114, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(127, 37)\n",
      "Unique CPDs : (62, 37)\n",
      "Unique SNVs : (62, 37)\n",
      "Unique exons : (127, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(92, 37)\n",
      "Unique CPDs : (55, 37)\n",
      "Unique SNVs : (54, 37)\n",
      "Unique exons : (91, 37)\n",
      "(112, 37)\n",
      "Unique CPDs : (63, 37)\n",
      "Unique SNVs : (63, 37)\n",
      "Unique exons : (112, 37)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/shared/software/anaconda/anaconda3/lib/python3.10/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 37 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    }
   ],
   "source": [
    "df_statistics_intersect = pd.DataFrame(columns=['Distance of SNVs from CPDs', 'unique_CPDs','unique_SNVs', 'unique_EXON', 'unique_Intersected_data'])\n",
    "df_list = []\n",
    "for i in range(51):\n",
    "    df_temp_CPD_SNV = df_intersected_CPD_SNV[abs(df_intersected_CPD_SNV[\"distance\"])==i]\n",
    "    temp_CPD_SNV_bed = pybedtools.BedTool.from_dataframe(df_temp_CPD_SNV)\n",
    "    # Perform the intersection\n",
    "    intersect_CPD_SNV_exon_bed = temp_CPD_SNV_bed.intersect(df_exon_bed, wa=True, wb=True)\n",
    "    df_intersect_CPD_SNV_exon = intersect_CPD_SNV_exon_bed.to_dataframe()\n",
    "    df_intersect_CPD_SNV_exon.columns = df_temp_CPD_SNV.columns.to_list() +  ['exon_' + str(x) for x in df_exon.columns.to_list()]\n",
    "    #display(df_intersect_CPD_SNV_exon)\n",
    "    print(df_intersect_CPD_SNV_exon.shape)\n",
    "    df_intersect_CPD_SNV_exon.to_csv(\"/data/private/pdutta/Collab_data/Jim_work/Output/UV_Plus/Intersected_CPD_SNV_EXON/distance_{}.tsv\".format(i), sep=\"\\t\", index=False)\n",
    "    df_list.append(df_intersect_CPD_SNV_exon)\n",
    "    print(\"Unique CPDs :\", df_intersect_CPD_SNV_exon.drop_duplicates(['CPD_start','CPD_end']).shape)\n",
    "    print(\"Unique SNVs :\", df_intersect_CPD_SNV_exon.drop_duplicates(['SNV_start','SNV_end','ID',\t'SNV_Info']).shape)\n",
    "    print(\"Unique exons :\", df_intersect_CPD_SNV_exon.drop_duplicates(['exon_start', 'exon_end','exon_gene_id','exon_transcript_id']).shape)\n",
    "    df_statistics_intersect.loc[i, ['Distance of SNVs from CPDs']] = i\n",
    "    df_statistics_intersect.loc[i, ['unique_CPDs']] = len(df_intersect_CPD_SNV_exon.drop_duplicates(['CPD_start','CPD_end']))\n",
    "    df_statistics_intersect.loc[i, ['unique_SNVs']] = len(df_intersect_CPD_SNV_exon.drop_duplicates(['SNV_start','SNV_end','ID',\t'SNV_Info']))\n",
    "    df_statistics_intersect.loc[i, ['unique_EXON']] = len(df_intersect_CPD_SNV_exon.drop_duplicates(['exon_start', 'exon_end','exon_gene_id','exon_transcript_id']))\n",
    "    df_statistics_intersect.loc[i, ['unique_Intersected_data']] = len(df_intersect_CPD_SNV_exon)\n",
    "    #input()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "d0efbaf4-531a-4432-9728-7f46743f8e30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Distance of SNVs from CPDs</th>\n",
       "      <th>unique_CPDs</th>\n",
       "      <th>unique_SNVs</th>\n",
       "      <th>unique_EXON</th>\n",
       "      <th>unique_Intersected_data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>728</td>\n",
       "      <td>728</td>\n",
       "      <td>1337</td>\n",
       "      <td>1501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>437</td>\n",
       "      <td>428</td>\n",
       "      <td>888</td>\n",
       "      <td>916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>324</td>\n",
       "      <td>318</td>\n",
       "      <td>587</td>\n",
       "      <td>601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>293</td>\n",
       "      <td>291</td>\n",
       "      <td>521</td>\n",
       "      <td>539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>286</td>\n",
       "      <td>285</td>\n",
       "      <td>543</td>\n",
       "      <td>553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>255</td>\n",
       "      <td>253</td>\n",
       "      <td>515</td>\n",
       "      <td>519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>281</td>\n",
       "      <td>282</td>\n",
       "      <td>508</td>\n",
       "      <td>510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>224</td>\n",
       "      <td>222</td>\n",
       "      <td>426</td>\n",
       "      <td>431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>231</td>\n",
       "      <td>227</td>\n",
       "      <td>490</td>\n",
       "      <td>497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>196</td>\n",
       "      <td>194</td>\n",
       "      <td>339</td>\n",
       "      <td>343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>198</td>\n",
       "      <td>197</td>\n",
       "      <td>406</td>\n",
       "      <td>407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>207</td>\n",
       "      <td>202</td>\n",
       "      <td>378</td>\n",
       "      <td>392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>172</td>\n",
       "      <td>172</td>\n",
       "      <td>302</td>\n",
       "      <td>302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>177</td>\n",
       "      <td>178</td>\n",
       "      <td>318</td>\n",
       "      <td>319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>181</td>\n",
       "      <td>180</td>\n",
       "      <td>351</td>\n",
       "      <td>361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>160</td>\n",
       "      <td>159</td>\n",
       "      <td>324</td>\n",
       "      <td>326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>162</td>\n",
       "      <td>159</td>\n",
       "      <td>324</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>140</td>\n",
       "      <td>138</td>\n",
       "      <td>317</td>\n",
       "      <td>323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>161</td>\n",
       "      <td>159</td>\n",
       "      <td>297</td>\n",
       "      <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>127</td>\n",
       "      <td>124</td>\n",
       "      <td>239</td>\n",
       "      <td>246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>140</td>\n",
       "      <td>138</td>\n",
       "      <td>278</td>\n",
       "      <td>283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>21</td>\n",
       "      <td>163</td>\n",
       "      <td>162</td>\n",
       "      <td>337</td>\n",
       "      <td>338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>22</td>\n",
       "      <td>137</td>\n",
       "      <td>136</td>\n",
       "      <td>248</td>\n",
       "      <td>250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>23</td>\n",
       "      <td>131</td>\n",
       "      <td>130</td>\n",
       "      <td>305</td>\n",
       "      <td>309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>24</td>\n",
       "      <td>120</td>\n",
       "      <td>116</td>\n",
       "      <td>201</td>\n",
       "      <td>206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>25</td>\n",
       "      <td>106</td>\n",
       "      <td>105</td>\n",
       "      <td>174</td>\n",
       "      <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>26</td>\n",
       "      <td>103</td>\n",
       "      <td>103</td>\n",
       "      <td>193</td>\n",
       "      <td>193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>27</td>\n",
       "      <td>118</td>\n",
       "      <td>117</td>\n",
       "      <td>218</td>\n",
       "      <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>28</td>\n",
       "      <td>91</td>\n",
       "      <td>90</td>\n",
       "      <td>188</td>\n",
       "      <td>189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>29</td>\n",
       "      <td>93</td>\n",
       "      <td>92</td>\n",
       "      <td>158</td>\n",
       "      <td>159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>30</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>167</td>\n",
       "      <td>167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>31</td>\n",
       "      <td>80</td>\n",
       "      <td>79</td>\n",
       "      <td>139</td>\n",
       "      <td>142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>32</td>\n",
       "      <td>84</td>\n",
       "      <td>84</td>\n",
       "      <td>144</td>\n",
       "      <td>145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>33</td>\n",
       "      <td>97</td>\n",
       "      <td>95</td>\n",
       "      <td>165</td>\n",
       "      <td>168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>34</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>181</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>35</td>\n",
       "      <td>84</td>\n",
       "      <td>84</td>\n",
       "      <td>212</td>\n",
       "      <td>212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>36</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>108</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>37</td>\n",
       "      <td>84</td>\n",
       "      <td>82</td>\n",
       "      <td>185</td>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>38</td>\n",
       "      <td>69</td>\n",
       "      <td>68</td>\n",
       "      <td>205</td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>39</td>\n",
       "      <td>76</td>\n",
       "      <td>76</td>\n",
       "      <td>143</td>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>40</td>\n",
       "      <td>79</td>\n",
       "      <td>79</td>\n",
       "      <td>158</td>\n",
       "      <td>158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>41</td>\n",
       "      <td>67</td>\n",
       "      <td>67</td>\n",
       "      <td>117</td>\n",
       "      <td>118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>42</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>127</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>43</td>\n",
       "      <td>64</td>\n",
       "      <td>63</td>\n",
       "      <td>115</td>\n",
       "      <td>116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>44</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>135</td>\n",
       "      <td>135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>45</td>\n",
       "      <td>72</td>\n",
       "      <td>72</td>\n",
       "      <td>143</td>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>46</td>\n",
       "      <td>58</td>\n",
       "      <td>57</td>\n",
       "      <td>96</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>47</td>\n",
       "      <td>59</td>\n",
       "      <td>59</td>\n",
       "      <td>114</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>48</td>\n",
       "      <td>62</td>\n",
       "      <td>62</td>\n",
       "      <td>127</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>49</td>\n",
       "      <td>55</td>\n",
       "      <td>54</td>\n",
       "      <td>91</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>50</td>\n",
       "      <td>63</td>\n",
       "      <td>63</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Distance of SNVs from CPDs unique_CPDs unique_SNVs unique_EXON  \\\n",
       "0                           0         728         728        1337   \n",
       "1                           1         437         428         888   \n",
       "2                           2         324         318         587   \n",
       "3                           3         293         291         521   \n",
       "4                           4         286         285         543   \n",
       "5                           5         255         253         515   \n",
       "6                           6         281         282         508   \n",
       "7                           7         224         222         426   \n",
       "8                           8         231         227         490   \n",
       "9                           9         196         194         339   \n",
       "10                         10         198         197         406   \n",
       "11                         11         207         202         378   \n",
       "12                         12         172         172         302   \n",
       "13                         13         177         178         318   \n",
       "14                         14         181         180         351   \n",
       "15                         15         160         159         324   \n",
       "16                         16         162         159         324   \n",
       "17                         17         140         138         317   \n",
       "18                         18         161         159         297   \n",
       "19                         19         127         124         239   \n",
       "20                         20         140         138         278   \n",
       "21                         21         163         162         337   \n",
       "22                         22         137         136         248   \n",
       "23                         23         131         130         305   \n",
       "24                         24         120         116         201   \n",
       "25                         25         106         105         174   \n",
       "26                         26         103         103         193   \n",
       "27                         27         118         117         218   \n",
       "28                         28          91          90         188   \n",
       "29                         29          93          92         158   \n",
       "30                         30          91          91         167   \n",
       "31                         31          80          79         139   \n",
       "32                         32          84          84         144   \n",
       "33                         33          97          95         165   \n",
       "34                         34          87          87         181   \n",
       "35                         35          84          84         212   \n",
       "36                         36          64          64         108   \n",
       "37                         37          84          82         185   \n",
       "38                         38          69          68         205   \n",
       "39                         39          76          76         143   \n",
       "40                         40          79          79         158   \n",
       "41                         41          67          67         117   \n",
       "42                         42          69          69         127   \n",
       "43                         43          64          63         115   \n",
       "44                         44          65          65         135   \n",
       "45                         45          72          72         143   \n",
       "46                         46          58          57          96   \n",
       "47                         47          59          59         114   \n",
       "48                         48          62          62         127   \n",
       "49                         49          55          54          91   \n",
       "50                         50          63          63         112   \n",
       "\n",
       "   unique_Intersected_data  \n",
       "0                     1501  \n",
       "1                      916  \n",
       "2                      601  \n",
       "3                      539  \n",
       "4                      553  \n",
       "5                      519  \n",
       "6                      510  \n",
       "7                      431  \n",
       "8                      497  \n",
       "9                      343  \n",
       "10                     407  \n",
       "11                     392  \n",
       "12                     302  \n",
       "13                     319  \n",
       "14                     361  \n",
       "15                     326  \n",
       "16                     333  \n",
       "17                     323  \n",
       "18                     300  \n",
       "19                     246  \n",
       "20                     283  \n",
       "21                     338  \n",
       "22                     250  \n",
       "23                     309  \n",
       "24                     206  \n",
       "25                     175  \n",
       "26                     193  \n",
       "27                     219  \n",
       "28                     189  \n",
       "29                     159  \n",
       "30                     167  \n",
       "31                     142  \n",
       "32                     145  \n",
       "33                     168  \n",
       "34                     182  \n",
       "35                     212  \n",
       "36                     108  \n",
       "37                     186  \n",
       "38                     210  \n",
       "39                     143  \n",
       "40                     158  \n",
       "41                     118  \n",
       "42                     127  \n",
       "43                     116  \n",
       "44                     135  \n",
       "45                     143  \n",
       "46                      97  \n",
       "47                     114  \n",
       "48                     127  \n",
       "49                      92  \n",
       "50                     112  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_statistics_intersect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "4bc2c44e-3818-44c9-9aa0-e08a5bbea84b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_name</th>\n",
       "      <th>CPD_start</th>\n",
       "      <th>CPD_end</th>\n",
       "      <th>chr_name_SNV</th>\n",
       "      <th>SNV_start</th>\n",
       "      <th>SNV_end</th>\n",
       "      <th>ID</th>\n",
       "      <th>SNV_Info</th>\n",
       "      <th>CPD_original_start</th>\n",
       "      <th>CPD_original_end</th>\n",
       "      <th>distance</th>\n",
       "      <th>exon_seqname</th>\n",
       "      <th>exon_start</th>\n",
       "      <th>exon_end</th>\n",
       "      <th>exon_source</th>\n",
       "      <th>exon_feature</th>\n",
       "      <th>exon_score</th>\n",
       "      <th>exon_strand</th>\n",
       "      <th>exon_frame</th>\n",
       "      <th>exon_gene_id</th>\n",
       "      <th>exon_gene_type</th>\n",
       "      <th>exon_gene_name</th>\n",
       "      <th>exon_level</th>\n",
       "      <th>exon_mgi_id</th>\n",
       "      <th>exon_havana_gene</th>\n",
       "      <th>exon_transcript_id</th>\n",
       "      <th>exon_transcript_type</th>\n",
       "      <th>exon_transcript_name</th>\n",
       "      <th>exon_transcript_support_level</th>\n",
       "      <th>exon_tag</th>\n",
       "      <th>exon_havana_transcript</th>\n",
       "      <th>exon_exon_number</th>\n",
       "      <th>exon_exon_id</th>\n",
       "      <th>exon_protein_id</th>\n",
       "      <th>exon_ccdsid</th>\n",
       "      <th>exon_ont</th>\n",
       "      <th>exon_unique_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5346484</td>\n",
       "      <td>5346485</td>\n",
       "      <td>chr1</td>\n",
       "      <td>5346484</td>\n",
       "      <td>5346485</td>\n",
       "      <td>334</td>\n",
       "      <td>N:TC[C&gt;G]TA</td>\n",
       "      <td>5346484</td>\n",
       "      <td>5346485</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>5346329</td>\n",
       "      <td>5346985</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000104352.2</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>Gm7182</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:3644877</td>\n",
       "      <td>OTTMUSG00000050143.1</td>\n",
       "      <td>ENSMUST00000193351.2</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>Gm7182-201</td>\n",
       "      <td>.</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>OTTMUST00000127487.1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSMUSE00001338734.2</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PGO:0000004</td>\n",
       "      <td>ENSMUSG00000104352.2_5346329_5346985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5378238</td>\n",
       "      <td>5378239</td>\n",
       "      <td>chr1</td>\n",
       "      <td>5378238</td>\n",
       "      <td>5378239</td>\n",
       "      <td>335</td>\n",
       "      <td>N:AT[C&gt;T]CA</td>\n",
       "      <td>5378238</td>\n",
       "      <td>5378239</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>5377962</td>\n",
       "      <td>5380240</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000104046.2</td>\n",
       "      <td>TEC</td>\n",
       "      <td>Gm37567</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:5610795</td>\n",
       "      <td>OTTMUSG00000050142.1</td>\n",
       "      <td>ENSMUST00000193241.2</td>\n",
       "      <td>TEC</td>\n",
       "      <td>Gm37567-201</td>\n",
       "      <td>.</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>OTTMUST00000127486.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00001341804.2</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000104046.2_5377962_5380240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7586433</td>\n",
       "      <td>7586434</td>\n",
       "      <td>chr1</td>\n",
       "      <td>7586433</td>\n",
       "      <td>7586434</td>\n",
       "      <td>334</td>\n",
       "      <td>N:CC[T&gt;A]CC</td>\n",
       "      <td>7586433</td>\n",
       "      <td>7586434</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>7586398</td>\n",
       "      <td>7588373</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000103067.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>Gm30414</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:5589573</td>\n",
       "      <td>OTTMUSG00000050304.1</td>\n",
       "      <td>ENSMUST00000194886.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>Gm30414-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>OTTMUST00000127738.1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSMUSE00001339812.2</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000103067.2_7586398_7588373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10522032</td>\n",
       "      <td>10522033</td>\n",
       "      <td>chr1</td>\n",
       "      <td>10522032</td>\n",
       "      <td>10522033</td>\n",
       "      <td>335</td>\n",
       "      <td>U:TT[C&gt;T]CC</td>\n",
       "      <td>10522032</td>\n",
       "      <td>10522033</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>10519470</td>\n",
       "      <td>10522214</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000103810.2</td>\n",
       "      <td>TEC</td>\n",
       "      <td>Gm38005</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:5611233</td>\n",
       "      <td>OTTMUSG00000047883.1</td>\n",
       "      <td>ENSMUST00000194845.2</td>\n",
       "      <td>TEC</td>\n",
       "      <td>Gm38005-201</td>\n",
       "      <td>.</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>OTTMUST00000123156.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00001341523.2</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000103810.2_10519470_10522214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10853212</td>\n",
       "      <td>10853213</td>\n",
       "      <td>chr1</td>\n",
       "      <td>10853212</td>\n",
       "      <td>10853213</td>\n",
       "      <td>338</td>\n",
       "      <td>N:CT[C&gt;T]TT</td>\n",
       "      <td>10853212</td>\n",
       "      <td>10853213</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>10852681</td>\n",
       "      <td>10854602</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000103448.2</td>\n",
       "      <td>TEC</td>\n",
       "      <td>Gm37133</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:5610361</td>\n",
       "      <td>OTTMUSG00000047884.1</td>\n",
       "      <td>ENSMUST00000192705.2</td>\n",
       "      <td>TEC</td>\n",
       "      <td>Gm37133-201</td>\n",
       "      <td>.</td>\n",
       "      <td>Ensembl_canonical</td>\n",
       "      <td>OTTMUST00000123157.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00001346299.2</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000103448.2_10852681_10854602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>chr9</td>\n",
       "      <td>64929376</td>\n",
       "      <td>64929477</td>\n",
       "      <td>chr9</td>\n",
       "      <td>64929376</td>\n",
       "      <td>64929377</td>\n",
       "      <td>341</td>\n",
       "      <td>N:TT[C&gt;T]CT</td>\n",
       "      <td>64929426</td>\n",
       "      <td>64929427</td>\n",
       "      <td>50</td>\n",
       "      <td>chr9</td>\n",
       "      <td>64929089</td>\n",
       "      <td>64929907</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000120656.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>Gm57019</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:7337951</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUST00000244940.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>Gm57019-201</td>\n",
       "      <td>.</td>\n",
       "      <td>TAGENE</td>\n",
       "      <td>.</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00001484772.1</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000120656.1_64929089_64929907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>chr9</td>\n",
       "      <td>75981992</td>\n",
       "      <td>75982093</td>\n",
       "      <td>chr9</td>\n",
       "      <td>75981992</td>\n",
       "      <td>75981993</td>\n",
       "      <td>334</td>\n",
       "      <td>U:TT[C&gt;T]AG</td>\n",
       "      <td>75982042</td>\n",
       "      <td>75982043</td>\n",
       "      <td>50</td>\n",
       "      <td>chr9</td>\n",
       "      <td>75981950</td>\n",
       "      <td>75982098</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000007908.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Hmgcll1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:2446108</td>\n",
       "      <td>OTTMUSG00000027403.8</td>\n",
       "      <td>ENSMUST00000008052.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Hmgcll1-201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>OTTMUST00000067779.4</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSMUSE00000219058.2</td>\n",
       "      <td>ENSMUSP00000008052.7</td>\n",
       "      <td>CCDS23349.1</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000007908.15_75981950_75982098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>chr9</td>\n",
       "      <td>75981992</td>\n",
       "      <td>75982093</td>\n",
       "      <td>chr9</td>\n",
       "      <td>75981992</td>\n",
       "      <td>75981993</td>\n",
       "      <td>334</td>\n",
       "      <td>U:TT[C&gt;T]AG</td>\n",
       "      <td>75982042</td>\n",
       "      <td>75982043</td>\n",
       "      <td>50</td>\n",
       "      <td>chr9</td>\n",
       "      <td>75981950</td>\n",
       "      <td>75983404</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000007908.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Hmgcll1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:2446108</td>\n",
       "      <td>OTTMUSG00000027403.8</td>\n",
       "      <td>ENSMUST00000121592.9</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>Hmgcll1-203</td>\n",
       "      <td>1.0</td>\n",
       "      <td>.</td>\n",
       "      <td>OTTMUST00000067781.5</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSMUSE00000720142.2</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000007908.15_75981950_75983404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>chr9</td>\n",
       "      <td>75981992</td>\n",
       "      <td>75982093</td>\n",
       "      <td>chr9</td>\n",
       "      <td>75981992</td>\n",
       "      <td>75981993</td>\n",
       "      <td>334</td>\n",
       "      <td>U:TT[C&gt;T]AG</td>\n",
       "      <td>75982042</td>\n",
       "      <td>75982043</td>\n",
       "      <td>50</td>\n",
       "      <td>chr9</td>\n",
       "      <td>75981950</td>\n",
       "      <td>75982098</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000007908.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Hmgcll1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:2446108</td>\n",
       "      <td>OTTMUSG00000027403.8</td>\n",
       "      <td>ENSMUST00000117981.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Hmgcll1-202</td>\n",
       "      <td>5.0</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTMUST00000117810.3</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ENSMUSE00000219058.2</td>\n",
       "      <td>ENSMUSP00000114045.3</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000007908.15_75981950_75982098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>chr9</td>\n",
       "      <td>87121457</td>\n",
       "      <td>87121558</td>\n",
       "      <td>chr9</td>\n",
       "      <td>87121557</td>\n",
       "      <td>87121558</td>\n",
       "      <td>342</td>\n",
       "      <td>U:TT[C&gt;T]TC</td>\n",
       "      <td>87121507</td>\n",
       "      <td>87121508</td>\n",
       "      <td>-50</td>\n",
       "      <td>chr9</td>\n",
       "      <td>87120407</td>\n",
       "      <td>87121986</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000056919.10</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Cep162</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MGI:1925343</td>\n",
       "      <td>OTTMUSG00000025816.4</td>\n",
       "      <td>ENSMUST00000144137.2</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>Cep162-203</td>\n",
       "      <td>2.0</td>\n",
       "      <td>.</td>\n",
       "      <td>OTTMUST00000078658.1</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSMUSE00000839500.2</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000056919.10_87120407_87121986</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15042 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr_name  CPD_start   CPD_end chr_name_SNV  SNV_start   SNV_end   ID  \\\n",
       "0       chr1    5346484   5346485         chr1    5346484   5346485  334   \n",
       "1       chr1    5378238   5378239         chr1    5378238   5378239  335   \n",
       "2       chr1    7586433   7586434         chr1    7586433   7586434  334   \n",
       "3       chr1   10522032  10522033         chr1   10522032  10522033  335   \n",
       "4       chr1   10853212  10853213         chr1   10853212  10853213  338   \n",
       "..       ...        ...       ...          ...        ...       ...  ...   \n",
       "107     chr9   64929376  64929477         chr9   64929376  64929377  341   \n",
       "108     chr9   75981992  75982093         chr9   75981992  75981993  334   \n",
       "109     chr9   75981992  75982093         chr9   75981992  75981993  334   \n",
       "110     chr9   75981992  75982093         chr9   75981992  75981993  334   \n",
       "111     chr9   87121457  87121558         chr9   87121557  87121558  342   \n",
       "\n",
       "        SNV_Info  CPD_original_start  CPD_original_end  distance exon_seqname  \\\n",
       "0    N:TC[C>G]TA             5346484           5346485         0         chr1   \n",
       "1    N:AT[C>T]CA             5378238           5378239         0         chr1   \n",
       "2    N:CC[T>A]CC             7586433           7586434         0         chr1   \n",
       "3    U:TT[C>T]CC            10522032          10522033         0         chr1   \n",
       "4    N:CT[C>T]TT            10853212          10853213         0         chr1   \n",
       "..           ...                 ...               ...       ...          ...   \n",
       "107  N:TT[C>T]CT            64929426          64929427        50         chr9   \n",
       "108  U:TT[C>T]AG            75982042          75982043        50         chr9   \n",
       "109  U:TT[C>T]AG            75982042          75982043        50         chr9   \n",
       "110  U:TT[C>T]AG            75982042          75982043        50         chr9   \n",
       "111  U:TT[C>T]TC            87121507          87121508       -50         chr9   \n",
       "\n",
       "     exon_start  exon_end exon_source exon_feature exon_score exon_strand  \\\n",
       "0       5346329   5346985      HAVANA         exon          .           -   \n",
       "1       5377962   5380240      HAVANA         exon          .           +   \n",
       "2       7586398   7588373      HAVANA         exon          .           +   \n",
       "3      10519470  10522214      HAVANA         exon          .           -   \n",
       "4      10852681  10854602      HAVANA         exon          .           -   \n",
       "..          ...       ...         ...          ...        ...         ...   \n",
       "107    64929089  64929907      HAVANA         exon          .           +   \n",
       "108    75981950  75982098      HAVANA         exon          .           +   \n",
       "109    75981950  75983404      HAVANA         exon          .           +   \n",
       "110    75981950  75982098      HAVANA         exon          .           +   \n",
       "111    87120407  87121986      HAVANA         exon          .           -   \n",
       "\n",
       "    exon_frame           exon_gene_id        exon_gene_type exon_gene_name  \\\n",
       "0            .   ENSMUSG00000104352.2  processed_pseudogene         Gm7182   \n",
       "1            .   ENSMUSG00000104046.2                   TEC        Gm37567   \n",
       "2            .   ENSMUSG00000103067.2                lncRNA        Gm30414   \n",
       "3            .   ENSMUSG00000103810.2                   TEC        Gm38005   \n",
       "4            .   ENSMUSG00000103448.2                   TEC        Gm37133   \n",
       "..         ...                    ...                   ...            ...   \n",
       "107          .   ENSMUSG00000120656.1                lncRNA        Gm57019   \n",
       "108          .  ENSMUSG00000007908.15        protein_coding        Hmgcll1   \n",
       "109          .  ENSMUSG00000007908.15        protein_coding        Hmgcll1   \n",
       "110          .  ENSMUSG00000007908.15        protein_coding        Hmgcll1   \n",
       "111          .  ENSMUSG00000056919.10        protein_coding         Cep162   \n",
       "\n",
       "     exon_level  exon_mgi_id      exon_havana_gene     exon_transcript_id  \\\n",
       "0           2.0  MGI:3644877  OTTMUSG00000050143.1   ENSMUST00000193351.2   \n",
       "1           2.0  MGI:5610795  OTTMUSG00000050142.1   ENSMUST00000193241.2   \n",
       "2           2.0  MGI:5589573  OTTMUSG00000050304.1   ENSMUST00000194886.2   \n",
       "3           2.0  MGI:5611233  OTTMUSG00000047883.1   ENSMUST00000194845.2   \n",
       "4           2.0  MGI:5610361  OTTMUSG00000047884.1   ENSMUST00000192705.2   \n",
       "..          ...          ...                   ...                    ...   \n",
       "107         2.0  MGI:7337951                     .   ENSMUST00000244940.1   \n",
       "108         2.0  MGI:2446108  OTTMUSG00000027403.8  ENSMUST00000008052.13   \n",
       "109         2.0  MGI:2446108  OTTMUSG00000027403.8   ENSMUST00000121592.9   \n",
       "110         2.0  MGI:2446108  OTTMUSG00000027403.8   ENSMUST00000117981.3   \n",
       "111         2.0  MGI:1925343  OTTMUSG00000025816.4   ENSMUST00000144137.2   \n",
       "\n",
       "     exon_transcript_type exon_transcript_name exon_transcript_support_level  \\\n",
       "0    processed_pseudogene           Gm7182-201                             .   \n",
       "1                     TEC          Gm37567-201                             .   \n",
       "2                  lncRNA          Gm30414-201                           1.0   \n",
       "3                     TEC          Gm38005-201                             .   \n",
       "4                     TEC          Gm37133-201                             .   \n",
       "..                    ...                  ...                           ...   \n",
       "107                lncRNA          Gm57019-201                             .   \n",
       "108        protein_coding          Hmgcll1-201                           1.0   \n",
       "109       retained_intron          Hmgcll1-203                           1.0   \n",
       "110        protein_coding          Hmgcll1-202                           5.0   \n",
       "111       retained_intron           Cep162-203                           2.0   \n",
       "\n",
       "              exon_tag exon_havana_transcript  exon_exon_number  \\\n",
       "0    Ensembl_canonical   OTTMUST00000127487.1               2.0   \n",
       "1    Ensembl_canonical   OTTMUST00000127486.1               1.0   \n",
       "2    Ensembl_canonical   OTTMUST00000127738.1               3.0   \n",
       "3    Ensembl_canonical   OTTMUST00000123156.1               1.0   \n",
       "4    Ensembl_canonical   OTTMUST00000123157.1               1.0   \n",
       "..                 ...                    ...               ...   \n",
       "107             TAGENE                      .               1.0   \n",
       "108               CCDS   OTTMUST00000067779.4               5.0   \n",
       "109                  .   OTTMUST00000067781.5               5.0   \n",
       "110              basic   OTTMUST00000117810.3               4.0   \n",
       "111                  .   OTTMUST00000078658.1               7.0   \n",
       "\n",
       "             exon_exon_id       exon_protein_id  exon_ccdsid     exon_ont  \\\n",
       "0    ENSMUSE00001338734.2                     .            .  PGO:0000004   \n",
       "1    ENSMUSE00001341804.2                     .            .            .   \n",
       "2    ENSMUSE00001339812.2                     .            .            .   \n",
       "3    ENSMUSE00001341523.2                     .            .            .   \n",
       "4    ENSMUSE00001346299.2                     .            .            .   \n",
       "..                    ...                   ...          ...          ...   \n",
       "107  ENSMUSE00001484772.1                     .            .            .   \n",
       "108  ENSMUSE00000219058.2  ENSMUSP00000008052.7  CCDS23349.1            .   \n",
       "109  ENSMUSE00000720142.2                     .            .            .   \n",
       "110  ENSMUSE00000219058.2  ENSMUSP00000114045.3            .            .   \n",
       "111  ENSMUSE00000839500.2                     .            .            .   \n",
       "\n",
       "                              exon_unique_id  \n",
       "0       ENSMUSG00000104352.2_5346329_5346985  \n",
       "1       ENSMUSG00000104046.2_5377962_5380240  \n",
       "2       ENSMUSG00000103067.2_7586398_7588373  \n",
       "3     ENSMUSG00000103810.2_10519470_10522214  \n",
       "4     ENSMUSG00000103448.2_10852681_10854602  \n",
       "..                                       ...  \n",
       "107   ENSMUSG00000120656.1_64929089_64929907  \n",
       "108  ENSMUSG00000007908.15_75981950_75982098  \n",
       "109  ENSMUSG00000007908.15_75981950_75983404  \n",
       "110  ENSMUSG00000007908.15_75981950_75982098  \n",
       "111  ENSMUSG00000056919.10_87120407_87121986  \n",
       "\n",
       "[15042 rows x 37 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_combined = pd.concat(df_list)\n",
    "df_combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "71a43cfd-760a-4aca-985b-9bea50d1bdc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_combined.to_csv(\"/data/private/pdutta/Collab_data/Jim_work/Output/UV_Plus/UV_intersected_combined_CPD_SNV_EXON.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "cdd446c3-4689-41f8-8e1e-9291d41d7a2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_index = len(df_statistics_intersect)+1\n",
    "df_statistics_intersect.loc[combined_index, ['Distance of SNVs from CPDs']] = \"Combined\"\n",
    "df_statistics_intersect.loc[combined_index, ['unique_CPDs']] = len(df_combined.drop_duplicates(['CPD_start','CPD_end']))\n",
    "df_statistics_intersect.loc[combined_index, ['unique_SNVs']] = len(df_combined.drop_duplicates(['SNV_start','SNV_end','ID',\t'SNV_Info']))\n",
    "df_statistics_intersect.loc[combined_index, ['unique_EXON']] = len(df_combined.drop_duplicates(['exon_start', 'exon_end','exon_gene_id','exon_transcript_id']))\n",
    "df_statistics_intersect.loc[combined_index, ['unique_Intersected_data']] = len(df_combined)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "34d74fb4-ed52-488d-9c86-0dd47af21a23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Distance of SNVs from CPDs</th>\n",
       "      <th>unique_CPDs</th>\n",
       "      <th>unique_SNVs</th>\n",
       "      <th>unique_EXON</th>\n",
       "      <th>unique_Intersected_data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>728</td>\n",
       "      <td>728</td>\n",
       "      <td>1337</td>\n",
       "      <td>1501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>437</td>\n",
       "      <td>428</td>\n",
       "      <td>888</td>\n",
       "      <td>916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>324</td>\n",
       "      <td>318</td>\n",
       "      <td>587</td>\n",
       "      <td>601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>293</td>\n",
       "      <td>291</td>\n",
       "      <td>521</td>\n",
       "      <td>539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>286</td>\n",
       "      <td>285</td>\n",
       "      <td>543</td>\n",
       "      <td>553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>255</td>\n",
       "      <td>253</td>\n",
       "      <td>515</td>\n",
       "      <td>519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>281</td>\n",
       "      <td>282</td>\n",
       "      <td>508</td>\n",
       "      <td>510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>224</td>\n",
       "      <td>222</td>\n",
       "      <td>426</td>\n",
       "      <td>431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>231</td>\n",
       "      <td>227</td>\n",
       "      <td>490</td>\n",
       "      <td>497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>196</td>\n",
       "      <td>194</td>\n",
       "      <td>339</td>\n",
       "      <td>343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>198</td>\n",
       "      <td>197</td>\n",
       "      <td>406</td>\n",
       "      <td>407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>207</td>\n",
       "      <td>202</td>\n",
       "      <td>378</td>\n",
       "      <td>392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>172</td>\n",
       "      <td>172</td>\n",
       "      <td>302</td>\n",
       "      <td>302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>177</td>\n",
       "      <td>178</td>\n",
       "      <td>318</td>\n",
       "      <td>319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>181</td>\n",
       "      <td>180</td>\n",
       "      <td>351</td>\n",
       "      <td>361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>160</td>\n",
       "      <td>159</td>\n",
       "      <td>324</td>\n",
       "      <td>326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>162</td>\n",
       "      <td>159</td>\n",
       "      <td>324</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>140</td>\n",
       "      <td>138</td>\n",
       "      <td>317</td>\n",
       "      <td>323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>161</td>\n",
       "      <td>159</td>\n",
       "      <td>297</td>\n",
       "      <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>127</td>\n",
       "      <td>124</td>\n",
       "      <td>239</td>\n",
       "      <td>246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>140</td>\n",
       "      <td>138</td>\n",
       "      <td>278</td>\n",
       "      <td>283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>21</td>\n",
       "      <td>163</td>\n",
       "      <td>162</td>\n",
       "      <td>337</td>\n",
       "      <td>338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>22</td>\n",
       "      <td>137</td>\n",
       "      <td>136</td>\n",
       "      <td>248</td>\n",
       "      <td>250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>23</td>\n",
       "      <td>131</td>\n",
       "      <td>130</td>\n",
       "      <td>305</td>\n",
       "      <td>309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>24</td>\n",
       "      <td>120</td>\n",
       "      <td>116</td>\n",
       "      <td>201</td>\n",
       "      <td>206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>25</td>\n",
       "      <td>106</td>\n",
       "      <td>105</td>\n",
       "      <td>174</td>\n",
       "      <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>26</td>\n",
       "      <td>103</td>\n",
       "      <td>103</td>\n",
       "      <td>193</td>\n",
       "      <td>193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>27</td>\n",
       "      <td>118</td>\n",
       "      <td>117</td>\n",
       "      <td>218</td>\n",
       "      <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>28</td>\n",
       "      <td>91</td>\n",
       "      <td>90</td>\n",
       "      <td>188</td>\n",
       "      <td>189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>29</td>\n",
       "      <td>93</td>\n",
       "      <td>92</td>\n",
       "      <td>158</td>\n",
       "      <td>159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>30</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>167</td>\n",
       "      <td>167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>31</td>\n",
       "      <td>80</td>\n",
       "      <td>79</td>\n",
       "      <td>139</td>\n",
       "      <td>142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>32</td>\n",
       "      <td>84</td>\n",
       "      <td>84</td>\n",
       "      <td>144</td>\n",
       "      <td>145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>33</td>\n",
       "      <td>97</td>\n",
       "      <td>95</td>\n",
       "      <td>165</td>\n",
       "      <td>168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>34</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>181</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>35</td>\n",
       "      <td>84</td>\n",
       "      <td>84</td>\n",
       "      <td>212</td>\n",
       "      <td>212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>36</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>108</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>37</td>\n",
       "      <td>84</td>\n",
       "      <td>82</td>\n",
       "      <td>185</td>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>38</td>\n",
       "      <td>69</td>\n",
       "      <td>68</td>\n",
       "      <td>205</td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>39</td>\n",
       "      <td>76</td>\n",
       "      <td>76</td>\n",
       "      <td>143</td>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>40</td>\n",
       "      <td>79</td>\n",
       "      <td>79</td>\n",
       "      <td>158</td>\n",
       "      <td>158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>41</td>\n",
       "      <td>67</td>\n",
       "      <td>67</td>\n",
       "      <td>117</td>\n",
       "      <td>118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>42</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>127</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>43</td>\n",
       "      <td>64</td>\n",
       "      <td>63</td>\n",
       "      <td>115</td>\n",
       "      <td>116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>44</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>135</td>\n",
       "      <td>135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>45</td>\n",
       "      <td>72</td>\n",
       "      <td>72</td>\n",
       "      <td>143</td>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>46</td>\n",
       "      <td>58</td>\n",
       "      <td>57</td>\n",
       "      <td>96</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>47</td>\n",
       "      <td>59</td>\n",
       "      <td>59</td>\n",
       "      <td>114</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>48</td>\n",
       "      <td>62</td>\n",
       "      <td>62</td>\n",
       "      <td>127</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>49</td>\n",
       "      <td>55</td>\n",
       "      <td>54</td>\n",
       "      <td>91</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>50</td>\n",
       "      <td>63</td>\n",
       "      <td>63</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Combined</td>\n",
       "      <td>7671</td>\n",
       "      <td>7605</td>\n",
       "      <td>12894</td>\n",
       "      <td>15042</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Distance of SNVs from CPDs unique_CPDs unique_SNVs unique_EXON  \\\n",
       "0                           0         728         728        1337   \n",
       "1                           1         437         428         888   \n",
       "2                           2         324         318         587   \n",
       "3                           3         293         291         521   \n",
       "4                           4         286         285         543   \n",
       "5                           5         255         253         515   \n",
       "6                           6         281         282         508   \n",
       "7                           7         224         222         426   \n",
       "8                           8         231         227         490   \n",
       "9                           9         196         194         339   \n",
       "10                         10         198         197         406   \n",
       "11                         11         207         202         378   \n",
       "12                         12         172         172         302   \n",
       "13                         13         177         178         318   \n",
       "14                         14         181         180         351   \n",
       "15                         15         160         159         324   \n",
       "16                         16         162         159         324   \n",
       "17                         17         140         138         317   \n",
       "18                         18         161         159         297   \n",
       "19                         19         127         124         239   \n",
       "20                         20         140         138         278   \n",
       "21                         21         163         162         337   \n",
       "22                         22         137         136         248   \n",
       "23                         23         131         130         305   \n",
       "24                         24         120         116         201   \n",
       "25                         25         106         105         174   \n",
       "26                         26         103         103         193   \n",
       "27                         27         118         117         218   \n",
       "28                         28          91          90         188   \n",
       "29                         29          93          92         158   \n",
       "30                         30          91          91         167   \n",
       "31                         31          80          79         139   \n",
       "32                         32          84          84         144   \n",
       "33                         33          97          95         165   \n",
       "34                         34          87          87         181   \n",
       "35                         35          84          84         212   \n",
       "36                         36          64          64         108   \n",
       "37                         37          84          82         185   \n",
       "38                         38          69          68         205   \n",
       "39                         39          76          76         143   \n",
       "40                         40          79          79         158   \n",
       "41                         41          67          67         117   \n",
       "42                         42          69          69         127   \n",
       "43                         43          64          63         115   \n",
       "44                         44          65          65         135   \n",
       "45                         45          72          72         143   \n",
       "46                         46          58          57          96   \n",
       "47                         47          59          59         114   \n",
       "48                         48          62          62         127   \n",
       "49                         49          55          54          91   \n",
       "50                         50          63          63         112   \n",
       "52                   Combined        7671        7605       12894   \n",
       "\n",
       "   unique_Intersected_data  \n",
       "0                     1501  \n",
       "1                      916  \n",
       "2                      601  \n",
       "3                      539  \n",
       "4                      553  \n",
       "5                      519  \n",
       "6                      510  \n",
       "7                      431  \n",
       "8                      497  \n",
       "9                      343  \n",
       "10                     407  \n",
       "11                     392  \n",
       "12                     302  \n",
       "13                     319  \n",
       "14                     361  \n",
       "15                     326  \n",
       "16                     333  \n",
       "17                     323  \n",
       "18                     300  \n",
       "19                     246  \n",
       "20                     283  \n",
       "21                     338  \n",
       "22                     250  \n",
       "23                     309  \n",
       "24                     206  \n",
       "25                     175  \n",
       "26                     193  \n",
       "27                     219  \n",
       "28                     189  \n",
       "29                     159  \n",
       "30                     167  \n",
       "31                     142  \n",
       "32                     145  \n",
       "33                     168  \n",
       "34                     182  \n",
       "35                     212  \n",
       "36                     108  \n",
       "37                     186  \n",
       "38                     210  \n",
       "39                     143  \n",
       "40                     158  \n",
       "41                     118  \n",
       "42                     127  \n",
       "43                     116  \n",
       "44                     135  \n",
       "45                     143  \n",
       "46                      97  \n",
       "47                     114  \n",
       "48                     127  \n",
       "49                      92  \n",
       "50                     112  \n",
       "52                   15042  "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_statistics_intersect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "56459340-9b64-4b9d-9acc-523b5157e672",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_statistics_intersect.to_csv(\"/data/private/pdutta/Collab_data/Jim_work/Output/UV_Plus/statistics_intersect_CPD_SNV_EXON.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "660ec2af-b59d-4b46-b466-8c72547e2197",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
