{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b91850f5-01f7-4588-9294-4ef6e92c8979",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pybedtools\n",
    "import numpy as np\n",
    "import os\n",
    "pybedtools.helpers.set_tempdir('/data/projects/temp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "524458f1-7005-42d4-9d64-0a96964f77d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_path_minus = \"/data/private/pdutta/Collab_data/Jim_work/CPD-seq/04-damage-1bp/Tissue_1hr_S2_CPD_1bp_sorted_minus.bed\"\n",
    "input_path_plus = \"/data/private/pdutta/Collab_data/Jim_work/CPD-seq/04-damage-1bp/Tissue_1hr_S2_CPD_1bp_sorted_plus.bed\"\n",
    "input_path_noUV_minus = \"/data/private/pdutta/Collab_data/Jim_work/CPD-seq/04-damage-1bp/Tissue_noUV_S1_CPD_1bp_sorted_minus.bed\"\n",
    "input_path_noUV_plus = \"/data/private/pdutta/Collab_data/Jim_work/CPD-seq/04-damage-1bp/Tissue_noUV_S1_CPD_1bp_sorted_plus.bed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3042b578-5f32-4207-acbb-5d8fc29efae3",
   "metadata": {},
   "outputs": [],
   "source": [
    "wgs_data_folder = \"/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d5ecfeae-629a-4c1d-b613-12517ca4a4d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_plus =pd.read_csv(input_path_plus, sep=\"\\t\", header=None)\n",
    "#df_minus =pd.read_csv(input_path_minus, sep=\"\\t\", header=None)\n",
    "#df_noUV_minus =pd.read_csv(input_path_noUV_minus, sep=\"\\t\", header=None)\n",
    "df_noUV_plus =pd.read_csv(input_path_noUV_plus, sep=\"\\t\", header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "08002d36-6375-46be-83c3-84ceb95aeb42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(99495320, 5) (15069500, 5)\n"
     ]
    }
   ],
   "source": [
    "print(df_plus.shape, df_noUV_plus.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e60b6d27-10c1-4537-8b7a-88cf47912c43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050029</td>\n",
       "      <td>3050030</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050030</td>\n",
       "      <td>3050031</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050054</td>\n",
       "      <td>3050055</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050055</td>\n",
       "      <td>3050056</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050076</td>\n",
       "      <td>3050077</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99495301</th>\n",
       "      <td>pUC19</td>\n",
       "      <td>2626</td>\n",
       "      <td>2627</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99495306</th>\n",
       "      <td>pUC19</td>\n",
       "      <td>2634</td>\n",
       "      <td>2635</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99495310</th>\n",
       "      <td>pUC19</td>\n",
       "      <td>2635</td>\n",
       "      <td>2636</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99495314</th>\n",
       "      <td>pUC19</td>\n",
       "      <td>2635</td>\n",
       "      <td>2636</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99495317</th>\n",
       "      <td>pUC19</td>\n",
       "      <td>2636</td>\n",
       "      <td>2637</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90995852 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              0        1        2  3   4\n",
       "0          chr1  3050029  3050030  +  AA\n",
       "1          chr1  3050030  3050031  +  AA\n",
       "2          chr1  3050054  3050055  +  GG\n",
       "3          chr1  3050055  3050056  +  GG\n",
       "4          chr1  3050076  3050077  +  AG\n",
       "...         ...      ...      ... ..  ..\n",
       "99495301  pUC19     2626     2627  +  GA\n",
       "99495306  pUC19     2634     2635  +  AA\n",
       "99495310  pUC19     2635     2636  +  AA\n",
       "99495314  pUC19     2635     2636  +  AG\n",
       "99495317  pUC19     2636     2637  +  AG\n",
       "\n",
       "[90995852 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_UV_plus = df_plus.drop_duplicates()\n",
    "df_UV_plus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "89e1e0cb-24b6-40ac-88e0-717b1a36a5d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050021</td>\n",
       "      <td>3050022</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050022</td>\n",
       "      <td>3050023</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050413</td>\n",
       "      <td>3050414</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050414</td>\n",
       "      <td>3050415</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050720</td>\n",
       "      <td>3050721</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15069474</th>\n",
       "      <td>pUC19</td>\n",
       "      <td>2626</td>\n",
       "      <td>2627</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15069478</th>\n",
       "      <td>pUC19</td>\n",
       "      <td>2634</td>\n",
       "      <td>2635</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15069488</th>\n",
       "      <td>pUC19</td>\n",
       "      <td>2635</td>\n",
       "      <td>2636</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15069498</th>\n",
       "      <td>pUC19</td>\n",
       "      <td>2635</td>\n",
       "      <td>2636</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15069499</th>\n",
       "      <td>pUC19</td>\n",
       "      <td>2636</td>\n",
       "      <td>2637</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14476001 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              0        1        2  3   4\n",
       "0          chr1  3050021  3050022  +  GA\n",
       "1          chr1  3050022  3050023  +  GA\n",
       "2          chr1  3050413  3050414  +  AG\n",
       "3          chr1  3050414  3050415  +  AG\n",
       "4          chr1  3050720  3050721  +  GG\n",
       "...         ...      ...      ... ..  ..\n",
       "15069474  pUC19     2626     2627  +  GA\n",
       "15069478  pUC19     2634     2635  +  AA\n",
       "15069488  pUC19     2635     2636  +  AA\n",
       "15069498  pUC19     2635     2636  +  AG\n",
       "15069499  pUC19     2636     2637  +  AG\n",
       "\n",
       "[14476001 rows x 5 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_noUV_plus = df_noUV_plus.drop_duplicates()\n",
    "df_noUV_plus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "85db1cb7-3641-4a53-8080-7fe5cc51d7f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>_merge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050029</td>\n",
       "      <td>3050030</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050030</td>\n",
       "      <td>3050031</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050054</td>\n",
       "      <td>3050055</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050055</td>\n",
       "      <td>3050056</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050076</td>\n",
       "      <td>3050077</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111903607</th>\n",
       "      <td>chrY_JH584303v1_random</td>\n",
       "      <td>155795</td>\n",
       "      <td>155796</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "      <td>right_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111903608</th>\n",
       "      <td>chrY_JH584303v1_random</td>\n",
       "      <td>156895</td>\n",
       "      <td>156896</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "      <td>right_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111903609</th>\n",
       "      <td>chrY_JH584303v1_random</td>\n",
       "      <td>156896</td>\n",
       "      <td>156897</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "      <td>right_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111903610</th>\n",
       "      <td>pUC19</td>\n",
       "      <td>1616</td>\n",
       "      <td>1617</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "      <td>right_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111903611</th>\n",
       "      <td>pUC19</td>\n",
       "      <td>1617</td>\n",
       "      <td>1618</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "      <td>right_only</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111903612 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                0        1        2  3   4      _merge\n",
       "0                            chr1  3050029  3050030  +  AA   left_only\n",
       "1                            chr1  3050030  3050031  +  AA   left_only\n",
       "2                            chr1  3050054  3050055  +  GG   left_only\n",
       "3                            chr1  3050055  3050056  +  GG   left_only\n",
       "4                            chr1  3050076  3050077  +  AG   left_only\n",
       "...                           ...      ...      ... ..  ..         ...\n",
       "111903607  chrY_JH584303v1_random   155795   155796  +  AG  right_only\n",
       "111903608  chrY_JH584303v1_random   156895   156896  +  GA  right_only\n",
       "111903609  chrY_JH584303v1_random   156896   156897  +  GA  right_only\n",
       "111903610                   pUC19     1616     1617  +  GA  right_only\n",
       "111903611                   pUC19     1617     1618  +  GA  right_only\n",
       "\n",
       "[111903612 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merged = pd.merge(df_plus, df_noUV_plus, how=\"outer\", indicator=True)\n",
    "df_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6364ef8a-bbf6-4441-ac69-5d169e810cb2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['left_only', 'both', 'right_only']\n",
       "Categories (3, object): ['left_only', 'right_only', 'both']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merged['_merge'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9e22f8ae-5148-431e-b846-44d5b65e1ff5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050029</td>\n",
       "      <td>3050030</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050030</td>\n",
       "      <td>3050031</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050054</td>\n",
       "      <td>3050055</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050055</td>\n",
       "      <td>3050056</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050076</td>\n",
       "      <td>3050077</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99327569</th>\n",
       "      <td>chrY_JH584303v1_random</td>\n",
       "      <td>157723</td>\n",
       "      <td>157724</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99327570</th>\n",
       "      <td>chrY_JH584303v1_random</td>\n",
       "      <td>157983</td>\n",
       "      <td>157984</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99327571</th>\n",
       "      <td>chrY_JH584303v1_random</td>\n",
       "      <td>157984</td>\n",
       "      <td>157985</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99327572</th>\n",
       "      <td>chrY_JH584303v1_random</td>\n",
       "      <td>158010</td>\n",
       "      <td>158011</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99327573</th>\n",
       "      <td>chrY_JH584303v1_random</td>\n",
       "      <td>158011</td>\n",
       "      <td>158012</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>96392906 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               0        1        2  3   4\n",
       "0                           chr1  3050029  3050030  +  AA\n",
       "1                           chr1  3050030  3050031  +  AA\n",
       "2                           chr1  3050054  3050055  +  GG\n",
       "3                           chr1  3050055  3050056  +  GG\n",
       "4                           chr1  3050076  3050077  +  AG\n",
       "...                          ...      ...      ... ..  ..\n",
       "99327569  chrY_JH584303v1_random   157723   157724  +  AG\n",
       "99327570  chrY_JH584303v1_random   157983   157984  +  GA\n",
       "99327571  chrY_JH584303v1_random   157984   157985  +  GA\n",
       "99327572  chrY_JH584303v1_random   158010   158011  +  GG\n",
       "99327573  chrY_JH584303v1_random   158011   158012  +  GG\n",
       "\n",
       "[96392906 rows x 5 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "only_in_df_plus = df_merged[df_merged['_merge'] == 'left_only'].drop('_merge', axis=1)\n",
    "only_in_df_plus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "71fac608-bb2c-4006-9eb3-aadeeb128e53",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050029</td>\n",
       "      <td>3050030</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050030</td>\n",
       "      <td>3050031</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050054</td>\n",
       "      <td>3050055</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050055</td>\n",
       "      <td>3050056</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050076</td>\n",
       "      <td>3050077</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88928138</th>\n",
       "      <td>chrY_JH584303v1_random</td>\n",
       "      <td>157723</td>\n",
       "      <td>157724</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88928139</th>\n",
       "      <td>chrY_JH584303v1_random</td>\n",
       "      <td>157983</td>\n",
       "      <td>157984</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88928140</th>\n",
       "      <td>chrY_JH584303v1_random</td>\n",
       "      <td>157984</td>\n",
       "      <td>157985</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88928141</th>\n",
       "      <td>chrY_JH584303v1_random</td>\n",
       "      <td>158010</td>\n",
       "      <td>158011</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88928142</th>\n",
       "      <td>chrY_JH584303v1_random</td>\n",
       "      <td>158011</td>\n",
       "      <td>158012</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88928143 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               0        1        2  3   4\n",
       "0                           chr1  3050029  3050030  +  AA\n",
       "1                           chr1  3050030  3050031  +  AA\n",
       "2                           chr1  3050054  3050055  +  GG\n",
       "3                           chr1  3050055  3050056  +  GG\n",
       "4                           chr1  3050076  3050077  +  AG\n",
       "...                          ...      ...      ... ..  ..\n",
       "88928138  chrY_JH584303v1_random   157723   157724  +  AG\n",
       "88928139  chrY_JH584303v1_random   157983   157984  +  GA\n",
       "88928140  chrY_JH584303v1_random   157984   157985  +  GA\n",
       "88928141  chrY_JH584303v1_random   158010   158011  +  GG\n",
       "88928142  chrY_JH584303v1_random   158011   158012  +  GG\n",
       "\n",
       "[88928143 rows x 5 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "only_in_df_plus.drop_duplicates().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fe01030a-38bd-49fb-8632-9d7af4608ff8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>90995852</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050021</td>\n",
       "      <td>3050022</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90995853</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050022</td>\n",
       "      <td>3050023</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90995854</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050413</td>\n",
       "      <td>3050414</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90995855</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050414</td>\n",
       "      <td>3050415</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90995856</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050721</td>\n",
       "      <td>3050722</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103404139</th>\n",
       "      <td>chrY_JH584303v1_random</td>\n",
       "      <td>155795</td>\n",
       "      <td>155796</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103404140</th>\n",
       "      <td>chrY_JH584303v1_random</td>\n",
       "      <td>156895</td>\n",
       "      <td>156896</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103404141</th>\n",
       "      <td>chrY_JH584303v1_random</td>\n",
       "      <td>156896</td>\n",
       "      <td>156897</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103404142</th>\n",
       "      <td>pUC19</td>\n",
       "      <td>1616</td>\n",
       "      <td>1617</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103404143</th>\n",
       "      <td>pUC19</td>\n",
       "      <td>1617</td>\n",
       "      <td>1618</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12408292 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                0        1        2  3   4\n",
       "90995852                     chr1  3050021  3050022  +  GA\n",
       "90995853                     chr1  3050022  3050023  +  GA\n",
       "90995854                     chr1  3050413  3050414  +  AG\n",
       "90995855                     chr1  3050414  3050415  +  AG\n",
       "90995856                     chr1  3050721  3050722  +  GG\n",
       "...                           ...      ...      ... ..  ..\n",
       "103404139  chrY_JH584303v1_random   155795   155796  +  AG\n",
       "103404140  chrY_JH584303v1_random   156895   156896  +  GA\n",
       "103404141  chrY_JH584303v1_random   156896   156897  +  GA\n",
       "103404142                   pUC19     1616     1617  +  GA\n",
       "103404143                   pUC19     1617     1618  +  GA\n",
       "\n",
       "[12408292 rows x 5 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "only_in_noUV_df_plus = df_merged[df_merged['_merge'] == 'right_only'].drop('_merge', axis=1)\n",
    "only_in_noUV_df_plus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ba068125-4fa8-4f9b-81f4-033aca87ed5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050720</td>\n",
       "      <td>3050721</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050881</td>\n",
       "      <td>3050882</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3051229</td>\n",
       "      <td>3051230</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3051533</td>\n",
       "      <td>3051534</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3051534</td>\n",
       "      <td>3051535</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90995847</th>\n",
       "      <td>pUC19</td>\n",
       "      <td>2626</td>\n",
       "      <td>2627</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90995848</th>\n",
       "      <td>pUC19</td>\n",
       "      <td>2634</td>\n",
       "      <td>2635</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90995849</th>\n",
       "      <td>pUC19</td>\n",
       "      <td>2635</td>\n",
       "      <td>2636</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90995850</th>\n",
       "      <td>pUC19</td>\n",
       "      <td>2635</td>\n",
       "      <td>2636</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90995851</th>\n",
       "      <td>pUC19</td>\n",
       "      <td>2636</td>\n",
       "      <td>2637</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2067709 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              0        1        2  3   4\n",
       "53         chr1  3050720  3050721  +  GG\n",
       "60         chr1  3050881  3050882  +  GG\n",
       "84         chr1  3051229  3051230  +  AA\n",
       "108        chr1  3051533  3051534  +  GG\n",
       "109        chr1  3051534  3051535  +  GG\n",
       "...         ...      ...      ... ..  ..\n",
       "90995847  pUC19     2626     2627  +  GA\n",
       "90995848  pUC19     2634     2635  +  AA\n",
       "90995849  pUC19     2635     2636  +  AA\n",
       "90995850  pUC19     2635     2636  +  AG\n",
       "90995851  pUC19     2636     2637  +  AG\n",
       "\n",
       "[2067709 rows x 5 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "both_in_df = df_merged[df_merged['_merge'] == 'both'].drop('_merge', axis=1)\n",
    "both_in_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e217d80d-befc-48ae-b5db-d3d261a96ad7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['chr1', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15',\n",
       "       'chr16', 'chr17', 'chr18', 'chr19', 'chr1_GL456210v1_random',\n",
       "       'chr1_GL456211v1_random', 'chr1_GL456212v1_random',\n",
       "       'chr1_GL456221v1_random', 'chr1_GL456239v1_random',\n",
       "       'chr1_MU069434v1_random', 'chr2', 'chr3', 'chr4',\n",
       "       'chr4_JH584295v1_random', 'chr5', 'chr5_GL456354v1_random',\n",
       "       'chr5_JH584296v1_random', 'chr5_JH584297v1_random',\n",
       "       'chr5_JH584298v1_random', 'chr5_JH584299v1_random', 'chr6', 'chr7',\n",
       "       'chr7_GL456219v1_random', 'chr8', 'chr9', 'chrM',\n",
       "       'chrUn_GL456359v1', 'chrUn_GL456360v1', 'chrUn_GL456366v1',\n",
       "       'chrUn_GL456367v1', 'chrUn_GL456368v1', 'chrUn_GL456370v1',\n",
       "       'chrUn_GL456372v1', 'chrUn_GL456378v1', 'chrUn_GL456379v1',\n",
       "       'chrUn_GL456381v1', 'chrUn_GL456382v1', 'chrUn_GL456383v1',\n",
       "       'chrUn_GL456385v1', 'chrUn_GL456387v1', 'chrUn_GL456389v1',\n",
       "       'chrUn_GL456390v1', 'chrUn_GL456392v1', 'chrUn_GL456394v1',\n",
       "       'chrUn_GL456396v1', 'chrUn_JH584304v1', 'chrUn_MU069435v1', 'chrX',\n",
       "       'chrX_GL456233v2_random', 'chrY', 'chrY_JH584300v1_random',\n",
       "       'chrY_JH584301v1_random', 'chrY_JH584302v1_random',\n",
       "       'chrY_JH584303v1_random', 'pUC19'], dtype=object)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "only_in_noUV_df_plus[0].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "185c17ba-31b6-4685-b611-29274d1a0cdf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0\n",
       "chr1                      912599\n",
       "chr10                     604646\n",
       "chr11                     642427\n",
       "chr12                     564011\n",
       "chr13                     570294\n",
       "                           ...  \n",
       "chrY_JH584300v1_random       480\n",
       "chrY_JH584301v1_random       530\n",
       "chrY_JH584302v1_random       362\n",
       "chrY_JH584303v1_random       369\n",
       "pUC19                          2\n",
       "Length: 62, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "only_in_noUV_df_plus.groupby([0]).size()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a12b776-c7a3-4f26-b7ff-f15839209d10",
   "metadata": {},
   "source": [
    "# WGS data (SNV)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d158f364-a4ac-4160-a47e-0ae9ef3c7c98",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/15_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/Y_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/10_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/2_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/18_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/19_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/13_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/1_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/3_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/5_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/8_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/12_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/6_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/9_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/4_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/16_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/17_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/14_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/7_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/11_seqinfo.txt\n",
      "/data/private/pdutta/Collab_data/Jim_work/Mouse_WGS/SNV/X_seqinfo.txt\n"
     ]
    }
   ],
   "source": [
    "combined_df = pd.DataFrame()\n",
    "for filename in os.listdir(wgs_data_folder):\n",
    "    # Check if the file is a txt file\n",
    "    if filename.endswith('.txt'):\n",
    "        # Construct the full filepath\n",
    "        filepath = os.path.join(wgs_data_folder, filename)\n",
    "        print(filepath)\n",
    "        \n",
    "        # Read the file into a DataFrame\n",
    "        temp_df = pd.read_csv(filepath, delimiter='\\t', header=None)  # Assuming tab-delimited, change as needed\n",
    "        \n",
    "        \n",
    "        # Append the data to the combined DataFrame\n",
    "        combined_df = pd.concat([combined_df, temp_df], ignore_index=True)\n",
    "\n",
    "# Show the combined DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2de2c7c8-e962-479f-921f-2fdefdf26709",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S319_final_vep_snpfiltered</td>\n",
       "      <td>15</td>\n",
       "      <td>3066686</td>\n",
       "      <td>N:AC[C&gt;T]GT</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S319_final_vep_snpfiltered</td>\n",
       "      <td>15</td>\n",
       "      <td>3391439</td>\n",
       "      <td>T:GG[C&gt;T]GG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S319_final_vep_snpfiltered</td>\n",
       "      <td>15</td>\n",
       "      <td>3911756</td>\n",
       "      <td>N:TC[C&gt;T]TT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S319_final_vep_snpfiltered</td>\n",
       "      <td>15</td>\n",
       "      <td>3930627</td>\n",
       "      <td>N:AT[C&gt;T]CC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S319_final_vep_snpfiltered</td>\n",
       "      <td>15</td>\n",
       "      <td>4574999</td>\n",
       "      <td>T:CC[C&gt;T]TT</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>828281</th>\n",
       "      <td>S346_final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>168111504</td>\n",
       "      <td>N:CT[C&gt;T]TT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>828282</th>\n",
       "      <td>S346_final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>168232166</td>\n",
       "      <td>N:CT[C&gt;T]TT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>828283</th>\n",
       "      <td>S346_final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>168273413</td>\n",
       "      <td>N:AA[C&gt;T]TT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>828284</th>\n",
       "      <td>S346_final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>168326651</td>\n",
       "      <td>N:TC[C&gt;T]TT</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>828285</th>\n",
       "      <td>S346_final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>168621252</td>\n",
       "      <td>U:TT[C&gt;T]TG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>828286 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 0   1          2            3  4\n",
       "0       S319_final_vep_snpfiltered  15    3066686  N:AC[C>T]GT -1\n",
       "1       S319_final_vep_snpfiltered  15    3391439  T:GG[C>T]GG  1\n",
       "2       S319_final_vep_snpfiltered  15    3911756  N:TC[C>T]TT  1\n",
       "3       S319_final_vep_snpfiltered  15    3930627  N:AT[C>T]CC  1\n",
       "4       S319_final_vep_snpfiltered  15    4574999  T:CC[C>T]TT -1\n",
       "...                            ...  ..        ...          ... ..\n",
       "828281  S346_final_vep_snpfiltered   X  168111504  N:CT[C>T]TT  1\n",
       "828282  S346_final_vep_snpfiltered   X  168232166  N:CT[C>T]TT  1\n",
       "828283  S346_final_vep_snpfiltered   X  168273413  N:AA[C>T]TT  1\n",
       "828284  S346_final_vep_snpfiltered   X  168326651  N:TC[C>T]TT -1\n",
       "828285  S346_final_vep_snpfiltered   X  168621252  U:TT[C>T]TG  1\n",
       "\n",
       "[828286 rows x 5 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ef27dd23-2581-4ba7-ac48-2a9daed99f13",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/projects/temp/ipykernel_1922124/4152682433.py:1: FutureWarning: In a future version of pandas all arguments of StringMethods.split except for the argument 'pat' will be keyword-only.\n",
      "  combined_df[['ID', '0']] = combined_df[0].str.split('_', 1, expand=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>319</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>15</td>\n",
       "      <td>3066686</td>\n",
       "      <td>N:AC[C&gt;T]GT</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>319</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>15</td>\n",
       "      <td>3391439</td>\n",
       "      <td>T:GG[C&gt;T]GG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>319</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>15</td>\n",
       "      <td>3911756</td>\n",
       "      <td>N:TC[C&gt;T]TT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>319</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>15</td>\n",
       "      <td>3930627</td>\n",
       "      <td>N:AT[C&gt;T]CC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>319</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>15</td>\n",
       "      <td>4574999</td>\n",
       "      <td>T:CC[C&gt;T]TT</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>828281</th>\n",
       "      <td>346</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>168111504</td>\n",
       "      <td>N:CT[C&gt;T]TT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>828282</th>\n",
       "      <td>346</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>168232166</td>\n",
       "      <td>N:CT[C&gt;T]TT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>828283</th>\n",
       "      <td>346</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>168273413</td>\n",
       "      <td>N:AA[C&gt;T]TT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>828284</th>\n",
       "      <td>346</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>168326651</td>\n",
       "      <td>N:TC[C&gt;T]TT</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>828285</th>\n",
       "      <td>346</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>168621252</td>\n",
       "      <td>U:TT[C&gt;T]TG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>828286 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ID                      0   1          2            3  4\n",
       "0       319  final_vep_snpfiltered  15    3066686  N:AC[C>T]GT -1\n",
       "1       319  final_vep_snpfiltered  15    3391439  T:GG[C>T]GG  1\n",
       "2       319  final_vep_snpfiltered  15    3911756  N:TC[C>T]TT  1\n",
       "3       319  final_vep_snpfiltered  15    3930627  N:AT[C>T]CC  1\n",
       "4       319  final_vep_snpfiltered  15    4574999  T:CC[C>T]TT -1\n",
       "...     ...                    ...  ..        ...          ... ..\n",
       "828281  346  final_vep_snpfiltered   X  168111504  N:CT[C>T]TT  1\n",
       "828282  346  final_vep_snpfiltered   X  168232166  N:CT[C>T]TT  1\n",
       "828283  346  final_vep_snpfiltered   X  168273413  N:AA[C>T]TT  1\n",
       "828284  346  final_vep_snpfiltered   X  168326651  N:TC[C>T]TT -1\n",
       "828285  346  final_vep_snpfiltered   X  168621252  U:TT[C>T]TG  1\n",
       "\n",
       "[828286 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df[['ID', '0']] = combined_df[0].str.split('_', 1, expand=True)\n",
    "\n",
    "# Drop the original column\n",
    "combined_df.drop(columns=[0], inplace=True)\n",
    "\n",
    "# Reorder the columns to make new_col1 and new_col2 the first and second columns\n",
    "combined_df = combined_df[['ID', '0'] + [col for col in combined_df.columns if col not in ['ID', '0']]]\n",
    "combined_df['ID'] = combined_df['ID'].str.replace('S','').astype(int)\n",
    "combined_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "27864708-912a-4dea-aee9-c12f02a302c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>15</td>\n",
       "      <td>3058069</td>\n",
       "      <td>N:CC[C&gt;T]TA</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>15</td>\n",
       "      <td>3063770</td>\n",
       "      <td>N:AA[T&gt;A]AT</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>15</td>\n",
       "      <td>3080731</td>\n",
       "      <td>N:GT[C&gt;T]TG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>15</td>\n",
       "      <td>3094450</td>\n",
       "      <td>N:AG[T&gt;G]TT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>15</td>\n",
       "      <td>3109676</td>\n",
       "      <td>N:AC[C&gt;T]CA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421491</th>\n",
       "      <td>346</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>168111504</td>\n",
       "      <td>N:CT[C&gt;T]TT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421492</th>\n",
       "      <td>346</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>168232166</td>\n",
       "      <td>N:CT[C&gt;T]TT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421493</th>\n",
       "      <td>346</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>168273413</td>\n",
       "      <td>N:AA[C&gt;T]TT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421494</th>\n",
       "      <td>346</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>168326651</td>\n",
       "      <td>N:TC[C&gt;T]TT</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421495</th>\n",
       "      <td>346</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>168621252</td>\n",
       "      <td>U:TT[C&gt;T]TG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>412073 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ID                      0   1          2            3  4\n",
       "0       334  final_vep_snpfiltered  15    3058069  N:CC[C>T]TA -1\n",
       "1       334  final_vep_snpfiltered  15    3063770  N:AA[T>A]AT -1\n",
       "2       334  final_vep_snpfiltered  15    3080731  N:GT[C>T]TG  1\n",
       "3       334  final_vep_snpfiltered  15    3094450  N:AG[T>G]TT  1\n",
       "4       334  final_vep_snpfiltered  15    3109676  N:AC[C>T]CA  1\n",
       "...     ...                    ...  ..        ...          ... ..\n",
       "421491  346  final_vep_snpfiltered   X  168111504  N:CT[C>T]TT  1\n",
       "421492  346  final_vep_snpfiltered   X  168232166  N:CT[C>T]TT  1\n",
       "421493  346  final_vep_snpfiltered   X  168273413  N:AA[C>T]TT  1\n",
       "421494  346  final_vep_snpfiltered   X  168326651  N:TC[C>T]TT -1\n",
       "421495  346  final_vep_snpfiltered   X  168621252  U:TT[C>T]TG  1\n",
       "\n",
       "[412073 rows x 6 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_interest = combined_df[(combined_df[\"ID\"] >= 332) & (combined_df[\"ID\"] <= 346)].reset_index(drop=True)\n",
    "df_interest[1] = df_interest[1].astype(str)\n",
    "df_interest = df_interest.drop_duplicates()\n",
    "df_interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9bc5bbf2-edff-4059-ab74-4e16454a77c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Custom sorting function\n",
    "def custom_chromosome_sort(chromosome):\n",
    "    if chromosome.isdigit():\n",
    "        return int(chromosome)\n",
    "    if chromosome == 'X':\n",
    "        return 23\n",
    "    if chromosome == 'Y':\n",
    "        return 24\n",
    "    return 25  # for any other value that might appear"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "01493942-52e4-4af1-b946-424a80ae60f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>1</td>\n",
       "      <td>17162732</td>\n",
       "      <td>U:TT[T&gt;A]TT</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>1</td>\n",
       "      <td>130607017</td>\n",
       "      <td>N:TC[T&gt;C]AT</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>1</td>\n",
       "      <td>130613091</td>\n",
       "      <td>N:TT[C&gt;T]AT</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>1</td>\n",
       "      <td>130653834</td>\n",
       "      <td>T:AC[C&gt;T]TT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>1</td>\n",
       "      <td>130674696</td>\n",
       "      <td>N:TT[C&gt;T]TG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412068</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>112617823</td>\n",
       "      <td>T:AT[C&gt;T]AG</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412069</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>112578837</td>\n",
       "      <td>T:AT[C&gt;T]CT</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412070</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>112683167</td>\n",
       "      <td>T:CC[C&gt;A]GG</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412071</th>\n",
       "      <td>346</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>168621252</td>\n",
       "      <td>U:TT[C&gt;T]TG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412072</th>\n",
       "      <td>341</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>Y</td>\n",
       "      <td>90741143</td>\n",
       "      <td>N:TC[T&gt;C]TG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>412073 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ID                      0  1          2            3  4\n",
       "0       334  final_vep_snpfiltered  1   17162732  U:TT[T>A]TT -1\n",
       "1       334  final_vep_snpfiltered  1  130607017  N:TC[T>C]AT -1\n",
       "2       334  final_vep_snpfiltered  1  130613091  N:TT[C>T]AT -1\n",
       "3       334  final_vep_snpfiltered  1  130653834  T:AC[C>T]TT  1\n",
       "4       334  final_vep_snpfiltered  1  130674696  N:TT[C>T]TG  1\n",
       "...     ...                    ... ..        ...          ... ..\n",
       "412068  334  final_vep_snpfiltered  X  112617823  T:AT[C>T]AG -1\n",
       "412069  334  final_vep_snpfiltered  X  112578837  T:AT[C>T]CT -1\n",
       "412070  334  final_vep_snpfiltered  X  112683167  T:CC[C>A]GG -1\n",
       "412071  346  final_vep_snpfiltered  X  168621252  U:TT[C>T]TG  1\n",
       "412072  341  final_vep_snpfiltered  Y   90741143  N:TC[T>C]TG  1\n",
       "\n",
       "[412073 rows x 6 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Apply custom sorting function\n",
    "df_interest['sort_key'] = df_interest[1].apply(custom_chromosome_sort)\n",
    "df_interest = df_interest.sort_values(by=['sort_key']).drop(columns=['sort_key']).reset_index(drop=True)\n",
    "df_interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d8ad6eb8-ff27-461e-995f-2bc096774136",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0         U:TT[T>A]TT\n",
       "1         N:TC[T>C]AT\n",
       "2         N:TT[C>T]AT\n",
       "3         T:AC[C>T]TT\n",
       "4         N:TT[C>T]TG\n",
       "             ...     \n",
       "412068    T:AT[C>T]AG\n",
       "412069    T:AT[C>T]CT\n",
       "412070    T:CC[C>A]GG\n",
       "412071    U:TT[C>T]TG\n",
       "412072    N:TC[T>C]TG\n",
       "Name: 3, Length: 412073, dtype: object"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_interest[3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f1eb1af8-019e-42aa-98a7-78a3ad12713b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>1</td>\n",
       "      <td>17162732</td>\n",
       "      <td>U:TT[T&gt;A]TT</td>\n",
       "      <td>-1</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>1</td>\n",
       "      <td>130607017</td>\n",
       "      <td>N:TC[T&gt;C]AT</td>\n",
       "      <td>-1</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>1</td>\n",
       "      <td>130613091</td>\n",
       "      <td>N:TT[C&gt;T]AT</td>\n",
       "      <td>-1</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>1</td>\n",
       "      <td>130653834</td>\n",
       "      <td>T:AC[C&gt;T]TT</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>1</td>\n",
       "      <td>130674696</td>\n",
       "      <td>N:TT[C&gt;T]TG</td>\n",
       "      <td>1</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412068</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>112617823</td>\n",
       "      <td>T:AT[C&gt;T]AG</td>\n",
       "      <td>-1</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412069</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>112578837</td>\n",
       "      <td>T:AT[C&gt;T]CT</td>\n",
       "      <td>-1</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412070</th>\n",
       "      <td>334</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>112683167</td>\n",
       "      <td>T:CC[C&gt;A]GG</td>\n",
       "      <td>-1</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412071</th>\n",
       "      <td>346</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>X</td>\n",
       "      <td>168621252</td>\n",
       "      <td>U:TT[C&gt;T]TG</td>\n",
       "      <td>1</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412072</th>\n",
       "      <td>341</td>\n",
       "      <td>final_vep_snpfiltered</td>\n",
       "      <td>Y</td>\n",
       "      <td>90741143</td>\n",
       "      <td>N:TC[T&gt;C]TG</td>\n",
       "      <td>1</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>412073 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ID                      0  1          2            3  4 type\n",
       "0       334  final_vep_snpfiltered  1   17162732  U:TT[T>A]TT -1    U\n",
       "1       334  final_vep_snpfiltered  1  130607017  N:TC[T>C]AT -1    N\n",
       "2       334  final_vep_snpfiltered  1  130613091  N:TT[C>T]AT -1    N\n",
       "3       334  final_vep_snpfiltered  1  130653834  T:AC[C>T]TT  1    T\n",
       "4       334  final_vep_snpfiltered  1  130674696  N:TT[C>T]TG  1    N\n",
       "...     ...                    ... ..        ...          ... ..  ...\n",
       "412068  334  final_vep_snpfiltered  X  112617823  T:AT[C>T]AG -1    T\n",
       "412069  334  final_vep_snpfiltered  X  112578837  T:AT[C>T]CT -1    T\n",
       "412070  334  final_vep_snpfiltered  X  112683167  T:CC[C>A]GG -1    T\n",
       "412071  346  final_vep_snpfiltered  X  168621252  U:TT[C>T]TG  1    U\n",
       "412072  341  final_vep_snpfiltered  Y   90741143  N:TC[T>C]TG  1    N\n",
       "\n",
       "[412073 rows x 7 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_interest['type'] = df_interest[3].apply(lambda x: x.split(\":\")[0])\n",
    "df_interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a2b8322b-2f4f-4094-ba4b-f75aa2deb044",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['U', 'N', 'T', 'B'], dtype=object)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_interest['type'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0b34224-bf3e-4489-a4ed-91aa205add2a",
   "metadata": {},
   "source": [
    "## Found four type of tags  \n",
    "   * __N__: This might stand for \"Normal,\" perhaps indicating that the sequence comes from normal tissue or represents a reference sequence.\n",
    "   * __T__: This could stand for \"Tumor,\" possibly indicating that the sequence comes from tumor tissue, especially if you're working on cancer genomics data.\n",
    "   * __U__: This could stand for \"Unknown,\" perhaps indicating that the origin of the sequence or the quality of the data is uncertain.\n",
    "   * __B__: This is not standard in genomics as far as my training data goes. It could be specific to the particular dataset or pipeline you're working with. It might stand for something like \"Baseline\" or \"Both\" (if it's a combined analysis of tumor and normal, for example), but that's purely speculative."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "daf0e997-7dd8-4758-83b4-c3797045d2cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "type\n",
       "B       912\n",
       "N    271852\n",
       "T     55255\n",
       "U     84054\n",
       "dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_interest.groupby('type').size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8f6eed21-6709-4165-ad87-23ca8437e00b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ID</th>\n",
       "      <th>SNV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17162731</td>\n",
       "      <td>17162732</td>\n",
       "      <td>334</td>\n",
       "      <td>U:TT[T&gt;A]TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>130607016</td>\n",
       "      <td>130607017</td>\n",
       "      <td>334</td>\n",
       "      <td>N:TC[T&gt;C]AT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>130613090</td>\n",
       "      <td>130613091</td>\n",
       "      <td>334</td>\n",
       "      <td>N:TT[C&gt;T]AT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>130653833</td>\n",
       "      <td>130653834</td>\n",
       "      <td>334</td>\n",
       "      <td>T:AC[C&gt;T]TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>130674695</td>\n",
       "      <td>130674696</td>\n",
       "      <td>334</td>\n",
       "      <td>N:TT[C&gt;T]TG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412068</th>\n",
       "      <td>chrX</td>\n",
       "      <td>112617822</td>\n",
       "      <td>112617823</td>\n",
       "      <td>334</td>\n",
       "      <td>T:AT[C&gt;T]AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412069</th>\n",
       "      <td>chrX</td>\n",
       "      <td>112578836</td>\n",
       "      <td>112578837</td>\n",
       "      <td>334</td>\n",
       "      <td>T:AT[C&gt;T]CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412070</th>\n",
       "      <td>chrX</td>\n",
       "      <td>112683166</td>\n",
       "      <td>112683167</td>\n",
       "      <td>334</td>\n",
       "      <td>T:CC[C&gt;A]GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412071</th>\n",
       "      <td>chrX</td>\n",
       "      <td>168621251</td>\n",
       "      <td>168621252</td>\n",
       "      <td>346</td>\n",
       "      <td>U:TT[C&gt;T]TG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412072</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90741142</td>\n",
       "      <td>90741143</td>\n",
       "      <td>341</td>\n",
       "      <td>N:TC[T&gt;C]TG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>412073 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr      start        end   ID          SNV\n",
       "0       chr1   17162731   17162732  334  U:TT[T>A]TT\n",
       "1       chr1  130607016  130607017  334  N:TC[T>C]AT\n",
       "2       chr1  130613090  130613091  334  N:TT[C>T]AT\n",
       "3       chr1  130653833  130653834  334  T:AC[C>T]TT\n",
       "4       chr1  130674695  130674696  334  N:TT[C>T]TG\n",
       "...      ...        ...        ...  ...          ...\n",
       "412068  chrX  112617822  112617823  334  T:AT[C>T]AG\n",
       "412069  chrX  112578836  112578837  334  T:AT[C>T]CT\n",
       "412070  chrX  112683166  112683167  334  T:CC[C>A]GG\n",
       "412071  chrX  168621251  168621252  346  U:TT[C>T]TG\n",
       "412072  chrY   90741142   90741143  341  N:TC[T>C]TG\n",
       "\n",
       "[412073 rows x 5 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_interest_bed = pd.DataFrame()\n",
    "df_interest_bed['chr'] = 'chr'+ df_interest[1].astype(str)\n",
    "df_interest_bed['start'] = df_interest[2]-1\n",
    "df_interest_bed['end'] = df_interest[2]\n",
    "df_interest_bed['ID'] = df_interest['ID']\n",
    "df_interest_bed['SNV'] = df_interest[3]\n",
    "df_interest_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "da89077d-1841-475e-99d5-480a18a2815e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ID</th>\n",
       "      <th>SNV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17162731</td>\n",
       "      <td>17162732</td>\n",
       "      <td>334</td>\n",
       "      <td>U:TT[T&gt;A]TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>130607016</td>\n",
       "      <td>130607017</td>\n",
       "      <td>334</td>\n",
       "      <td>N:TC[T&gt;C]AT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>130613090</td>\n",
       "      <td>130613091</td>\n",
       "      <td>334</td>\n",
       "      <td>N:TT[C&gt;T]AT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>130653833</td>\n",
       "      <td>130653834</td>\n",
       "      <td>334</td>\n",
       "      <td>T:AC[C&gt;T]TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>130674695</td>\n",
       "      <td>130674696</td>\n",
       "      <td>334</td>\n",
       "      <td>N:TT[C&gt;T]TG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412068</th>\n",
       "      <td>chrX</td>\n",
       "      <td>112617822</td>\n",
       "      <td>112617823</td>\n",
       "      <td>334</td>\n",
       "      <td>T:AT[C&gt;T]AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412069</th>\n",
       "      <td>chrX</td>\n",
       "      <td>112578836</td>\n",
       "      <td>112578837</td>\n",
       "      <td>334</td>\n",
       "      <td>T:AT[C&gt;T]CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412070</th>\n",
       "      <td>chrX</td>\n",
       "      <td>112683166</td>\n",
       "      <td>112683167</td>\n",
       "      <td>334</td>\n",
       "      <td>T:CC[C&gt;A]GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412071</th>\n",
       "      <td>chrX</td>\n",
       "      <td>168621251</td>\n",
       "      <td>168621252</td>\n",
       "      <td>346</td>\n",
       "      <td>U:TT[C&gt;T]TG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412072</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90741142</td>\n",
       "      <td>90741143</td>\n",
       "      <td>341</td>\n",
       "      <td>N:TC[T&gt;C]TG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>412073 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr      start        end   ID          SNV\n",
       "0       chr1   17162731   17162732  334  U:TT[T>A]TT\n",
       "1       chr1  130607016  130607017  334  N:TC[T>C]AT\n",
       "2       chr1  130613090  130613091  334  N:TT[C>T]AT\n",
       "3       chr1  130653833  130653834  334  T:AC[C>T]TT\n",
       "4       chr1  130674695  130674696  334  N:TT[C>T]TG\n",
       "...      ...        ...        ...  ...          ...\n",
       "412068  chrX  112617822  112617823  334  T:AT[C>T]AG\n",
       "412069  chrX  112578836  112578837  334  T:AT[C>T]CT\n",
       "412070  chrX  112683166  112683167  334  T:CC[C>A]GG\n",
       "412071  chrX  168621251  168621252  346  U:TT[C>T]TG\n",
       "412072  chrY   90741142   90741143  341  N:TC[T>C]TG\n",
       "\n",
       "[412073 rows x 5 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_interest_bed.drop_duplicates()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7b56e51-0015-4363-9aa8-4573f2501e9f",
   "metadata": {},
   "source": [
    "## Generate multiple UV_plus CPD_seq data(by increasing the number of nucleotides in both starting and end position) and fond the intersection with the C>T mutation data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "049976e2-72f5-41fa-80cf-fad229f38da9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(90787037, 5) (412073, 5)\n",
      "The number of SNV counts 412073 at 0 epoch\n",
      "The intersecting regions size is : (23933, 5)\n",
      "After removal of intersecting region the remaining SNVs are  388140\n",
      "#############################################################\n",
      "The number of SNV counts 388140 at 1 epoch\n",
      "The intersecting regions size is : (14631, 5)\n",
      "After removal of intersecting region the remaining SNVs are  373509\n",
      "#############################################################\n",
      "The number of SNV counts 373509 at 2 epoch\n",
      "The intersecting regions size is : (12562, 5)\n",
      "After removal of intersecting region the remaining SNVs are  360947\n",
      "#############################################################\n",
      "The number of SNV counts 360947 at 3 epoch\n",
      "The intersecting regions size is : (10613, 5)\n",
      "After removal of intersecting region the remaining SNVs are  350334\n",
      "#############################################################\n",
      "The number of SNV counts 350334 at 4 epoch\n",
      "The intersecting regions size is : (10100, 5)\n",
      "After removal of intersecting region the remaining SNVs are  340234\n",
      "#############################################################\n",
      "The number of SNV counts 340234 at 5 epoch\n",
      "The intersecting regions size is : (9958, 5)\n",
      "After removal of intersecting region the remaining SNVs are  330276\n",
      "#############################################################\n",
      "The number of SNV counts 330276 at 6 epoch\n",
      "The intersecting regions size is : (9132, 5)\n",
      "After removal of intersecting region the remaining SNVs are  321144\n",
      "#############################################################\n",
      "The number of SNV counts 321144 at 7 epoch\n",
      "The intersecting regions size is : (8545, 5)\n",
      "After removal of intersecting region the remaining SNVs are  312599\n",
      "#############################################################\n",
      "The number of SNV counts 312599 at 8 epoch\n",
      "The intersecting regions size is : (8367, 5)\n",
      "After removal of intersecting region the remaining SNVs are  304232\n",
      "#############################################################\n",
      "The number of SNV counts 304232 at 9 epoch\n",
      "The intersecting regions size is : (7892, 5)\n",
      "After removal of intersecting region the remaining SNVs are  296340\n",
      "#############################################################\n",
      "The number of SNV counts 296340 at 10 epoch\n",
      "The intersecting regions size is : (7567, 5)\n",
      "After removal of intersecting region the remaining SNVs are  288773\n",
      "#############################################################\n",
      "The number of SNV counts 288773 at 11 epoch\n",
      "The intersecting regions size is : (7333, 5)\n",
      "After removal of intersecting region the remaining SNVs are  281440\n",
      "#############################################################\n",
      "The number of SNV counts 281440 at 12 epoch\n",
      "The intersecting regions size is : (7068, 5)\n",
      "After removal of intersecting region the remaining SNVs are  274372\n",
      "#############################################################\n",
      "The number of SNV counts 274372 at 13 epoch\n",
      "The intersecting regions size is : (6655, 5)\n",
      "After removal of intersecting region the remaining SNVs are  267717\n",
      "#############################################################\n",
      "The number of SNV counts 267717 at 14 epoch\n",
      "The intersecting regions size is : (6487, 5)\n",
      "After removal of intersecting region the remaining SNVs are  261230\n",
      "#############################################################\n",
      "The number of SNV counts 261230 at 15 epoch\n",
      "The intersecting regions size is : (6083, 5)\n",
      "After removal of intersecting region the remaining SNVs are  255147\n",
      "#############################################################\n",
      "The number of SNV counts 255147 at 16 epoch\n",
      "The intersecting regions size is : (5920, 5)\n",
      "After removal of intersecting region the remaining SNVs are  249227\n",
      "#############################################################\n",
      "The number of SNV counts 249227 at 17 epoch\n",
      "The intersecting regions size is : (5718, 5)\n",
      "After removal of intersecting region the remaining SNVs are  243509\n",
      "#############################################################\n",
      "The number of SNV counts 243509 at 18 epoch\n",
      "The intersecting regions size is : (5584, 5)\n",
      "After removal of intersecting region the remaining SNVs are  237925\n",
      "#############################################################\n",
      "The number of SNV counts 237925 at 19 epoch\n",
      "The intersecting regions size is : (5198, 5)\n",
      "After removal of intersecting region the remaining SNVs are  232727\n",
      "#############################################################\n",
      "The number of SNV counts 232727 at 20 epoch\n",
      "The intersecting regions size is : (5082, 5)\n",
      "After removal of intersecting region the remaining SNVs are  227645\n",
      "#############################################################\n",
      "The number of SNV counts 227645 at 21 epoch\n",
      "The intersecting regions size is : (5004, 5)\n",
      "After removal of intersecting region the remaining SNVs are  222641\n",
      "#############################################################\n",
      "The number of SNV counts 222641 at 22 epoch\n",
      "The intersecting regions size is : (4756, 5)\n",
      "After removal of intersecting region the remaining SNVs are  217885\n",
      "#############################################################\n",
      "The number of SNV counts 217885 at 23 epoch\n",
      "The intersecting regions size is : (4755, 5)\n",
      "After removal of intersecting region the remaining SNVs are  213130\n",
      "#############################################################\n",
      "The number of SNV counts 213130 at 24 epoch\n",
      "The intersecting regions size is : (4520, 5)\n",
      "After removal of intersecting region the remaining SNVs are  208610\n",
      "#############################################################\n",
      "The number of SNV counts 208610 at 25 epoch\n",
      "The intersecting regions size is : (4370, 5)\n",
      "After removal of intersecting region the remaining SNVs are  204240\n",
      "#############################################################\n",
      "The number of SNV counts 204240 at 26 epoch\n",
      "The intersecting regions size is : (4233, 5)\n",
      "After removal of intersecting region the remaining SNVs are  200007\n",
      "#############################################################\n",
      "The number of SNV counts 200007 at 27 epoch\n",
      "The intersecting regions size is : (4036, 5)\n",
      "After removal of intersecting region the remaining SNVs are  195971\n",
      "#############################################################\n",
      "The number of SNV counts 195971 at 28 epoch\n",
      "The intersecting regions size is : (3840, 5)\n",
      "After removal of intersecting region the remaining SNVs are  192131\n",
      "#############################################################\n",
      "The number of SNV counts 192131 at 29 epoch\n",
      "The intersecting regions size is : (3913, 5)\n",
      "After removal of intersecting region the remaining SNVs are  188218\n",
      "#############################################################\n",
      "The number of SNV counts 188218 at 30 epoch\n",
      "The intersecting regions size is : (3676, 5)\n",
      "After removal of intersecting region the remaining SNVs are  184542\n",
      "#############################################################\n",
      "The number of SNV counts 184542 at 31 epoch\n",
      "The intersecting regions size is : (3604, 5)\n",
      "After removal of intersecting region the remaining SNVs are  180938\n",
      "#############################################################\n",
      "The number of SNV counts 180938 at 32 epoch\n",
      "The intersecting regions size is : (3523, 5)\n",
      "After removal of intersecting region the remaining SNVs are  177415\n",
      "#############################################################\n",
      "The number of SNV counts 177415 at 33 epoch\n",
      "The intersecting regions size is : (3432, 5)\n",
      "After removal of intersecting region the remaining SNVs are  173983\n",
      "#############################################################\n",
      "The number of SNV counts 173983 at 34 epoch\n",
      "The intersecting regions size is : (3283, 5)\n",
      "After removal of intersecting region the remaining SNVs are  170700\n",
      "#############################################################\n",
      "The number of SNV counts 170700 at 35 epoch\n",
      "The intersecting regions size is : (3267, 5)\n",
      "After removal of intersecting region the remaining SNVs are  167433\n",
      "#############################################################\n",
      "The number of SNV counts 167433 at 36 epoch\n",
      "The intersecting regions size is : (3232, 5)\n",
      "After removal of intersecting region the remaining SNVs are  164201\n",
      "#############################################################\n",
      "The number of SNV counts 164201 at 37 epoch\n",
      "The intersecting regions size is : (3046, 5)\n",
      "After removal of intersecting region the remaining SNVs are  161155\n",
      "#############################################################\n",
      "The number of SNV counts 161155 at 38 epoch\n",
      "The intersecting regions size is : (2987, 5)\n",
      "After removal of intersecting region the remaining SNVs are  158168\n",
      "#############################################################\n",
      "The number of SNV counts 158168 at 39 epoch\n",
      "The intersecting regions size is : (2935, 5)\n",
      "After removal of intersecting region the remaining SNVs are  155233\n",
      "#############################################################\n",
      "The number of SNV counts 155233 at 40 epoch\n",
      "The intersecting regions size is : (2767, 5)\n",
      "After removal of intersecting region the remaining SNVs are  152466\n",
      "#############################################################\n",
      "The number of SNV counts 152466 at 41 epoch\n",
      "The intersecting regions size is : (2810, 5)\n",
      "After removal of intersecting region the remaining SNVs are  149656\n",
      "#############################################################\n",
      "The number of SNV counts 149656 at 42 epoch\n",
      "The intersecting regions size is : (2717, 5)\n",
      "After removal of intersecting region the remaining SNVs are  146939\n",
      "#############################################################\n",
      "The number of SNV counts 146939 at 43 epoch\n",
      "The intersecting regions size is : (2666, 5)\n",
      "After removal of intersecting region the remaining SNVs are  144273\n",
      "#############################################################\n",
      "The number of SNV counts 144273 at 44 epoch\n",
      "The intersecting regions size is : (2548, 5)\n",
      "After removal of intersecting region the remaining SNVs are  141725\n",
      "#############################################################\n",
      "The number of SNV counts 141725 at 45 epoch\n",
      "The intersecting regions size is : (2468, 5)\n",
      "After removal of intersecting region the remaining SNVs are  139257\n",
      "#############################################################\n",
      "The number of SNV counts 139257 at 46 epoch\n",
      "The intersecting regions size is : (2462, 5)\n",
      "After removal of intersecting region the remaining SNVs are  136795\n",
      "#############################################################\n",
      "The number of SNV counts 136795 at 47 epoch\n",
      "The intersecting regions size is : (2404, 5)\n",
      "After removal of intersecting region the remaining SNVs are  134391\n",
      "#############################################################\n",
      "The number of SNV counts 134391 at 48 epoch\n",
      "The intersecting regions size is : (2363, 5)\n",
      "After removal of intersecting region the remaining SNVs are  132028\n",
      "#############################################################\n",
      "The number of SNV counts 132028 at 49 epoch\n",
      "The intersecting regions size is : (2280, 5)\n",
      "After removal of intersecting region the remaining SNVs are  129748\n",
      "#############################################################\n",
      "The number of SNV counts 129748 at 50 epoch\n",
      "The intersecting regions size is : (2300, 5)\n",
      "After removal of intersecting region the remaining SNVs are  127448\n",
      "#############################################################\n"
     ]
    }
   ],
   "source": [
    "# Create a list of chromosomes to keep\n",
    "chromosomes_to_keep = ['chr' + str(i) for i in range(1, 23)] + ['chrX', 'chrY']\n",
    "\n",
    "# Filter the original DataFrame to only include specified chromosomes\n",
    "df_UV_plus_interested = df_UV_plus[df_UV_plus[0].isin(chromosomes_to_keep)].reset_index(drop=True)\n",
    "print(df_UV_plus_interested.shape, df_interest_bed.shape)\n",
    "\n",
    "# Create a BedTool object from another DataFrame\n",
    "bd_interest_SNV_bed = pybedtools.BedTool.from_dataframe(df_interest_bed)\n",
    "\n",
    "# Initialize an empty DataFrame for temporary use\n",
    "df_temp = pd.DataFrame()\n",
    "df_temp['chr_name'] = df_UV_plus_interested[0]\n",
    "\n",
    "# Convert df_noUV_plus_interested to a BedTool object initially\n",
    "df_UV_plus_bed = pybedtools.BedTool.from_dataframe(df_UV_plus_interested)\n",
    "\n",
    "for i in range(51):\n",
    "    # Populate df_temp with updated start and end positions\n",
    "    df_temp['chr_start'] = df_UV_plus_interested[1] - i\n",
    "    df_temp['chr_end'] = df_UV_plus_interested[2] + i\n",
    "    print(\"The number of SNV counts {} at {} epoch\".format(bd_interest_SNV_bed.count(), i))\n",
    "    # Convert df_temp to a BedTool object\n",
    "    bd_UV_plus_bed = pybedtools.BedTool.from_dataframe(df_temp)\n",
    "    # Perform the intersection\n",
    "    intersect_CPD_SNV = bd_UV_plus_bed.intersect(bd_interest_SNV_bed, wa=True, wb=True)\n",
    "    # Convert the result back to a DataFrame and set column names\n",
    "    df_intersect_CPD_SNV = intersect_CPD_SNV.to_dataframe()\n",
    "    df_intersect_CPD_SNV.columns = ['chr_name', 'CPD_start', 'CPD_end', 'chr_name_SNV', 'SNV_start', 'SNV_end', 'ID','SNV_Info']\n",
    "    #print(df_intersect_noCPD_SNV[['chr_name_SNV', 'SNV_start', 'SNV_end', 'ID','SNV_Info']].drop_duplicates())\n",
    "    \n",
    "    # Save the result to a file\n",
    "    df_intersect_CPD_SNV.to_csv(f\"/data/private/pdutta/Collab_data/Jim_work/Output/UV_plus/intersected_CPD_SNV_UV_plus_{i}.tsv\", sep='\\t', index=False)\n",
    "    \n",
    "    # Identify intersecting regions for removal in the next iteration\n",
    "    intersecting_SNV_regions = bd_interest_SNV_bed.intersect(bd_UV_plus_bed, wa=True, u=True)\n",
    "    print(\"The intersecting regions size is :\",intersecting_SNV_regions.to_dataframe().shape)\n",
    "    #print(bd_interest_SNV_bed.to_dataframe())\n",
    "    \n",
    "    # Remove intersecting regions for the next iteration\n",
    "    bd_interest_SNV_bed = bd_interest_SNV_bed.subtract(intersecting_SNV_regions)\n",
    "    #print(bd_interest_SNV_bed.to_dataframe())\n",
    "    print(\"After removal of intersecting region the remaining SNVs are \", bd_interest_SNV_bed.count())\n",
    "    #df_noUV_plus_interested.columns = ['chr_name', 'CPD_start', 'CPD_end']\n",
    "    \n",
    "    print(\"#############################################################\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca1d04f5-34be-4f56-8310-d87310dec8c1",
   "metadata": {},
   "source": [
    "## Generate multiple NoUV_plus CPD_seq data(by increasing the number of nucleotides in both starting and end position) and fond the intersection with the C>T mutation data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "defbdce3-8635-4463-8b1d-41ac390186df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14424047, 5) (412073, 5)\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      " \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of SNV counts 412073 at 0 epoch\n",
      "The intersecting regions size is : (4055, 5)\n",
      "After removal of intersecting region the remaining SNVs are  408018\n",
      "#############################################################\n",
      "The number of SNV counts 408018 at 1 epoch\n",
      "The intersecting regions size is : (2719, 5)\n",
      "After removal of intersecting region the remaining SNVs are  405299\n",
      "#############################################################\n",
      "The number of SNV counts 405299 at 2 epoch\n",
      "The intersecting regions size is : (2540, 5)\n",
      "After removal of intersecting region the remaining SNVs are  402759\n",
      "#############################################################\n",
      "The number of SNV counts 402759 at 3 epoch\n",
      "The intersecting regions size is : (2243, 5)\n",
      "After removal of intersecting region the remaining SNVs are  400516\n",
      "#############################################################\n",
      "The number of SNV counts 400516 at 4 epoch\n",
      "The intersecting regions size is : (2029, 5)\n",
      "After removal of intersecting region the remaining SNVs are  398487\n",
      "#############################################################\n",
      "The number of SNV counts 398487 at 5 epoch\n",
      "The intersecting regions size is : (2251, 5)\n",
      "After removal of intersecting region the remaining SNVs are  396236\n",
      "#############################################################\n",
      "The number of SNV counts 396236 at 6 epoch\n",
      "The intersecting regions size is : (2142, 5)\n",
      "After removal of intersecting region the remaining SNVs are  394094\n",
      "#############################################################\n",
      "The number of SNV counts 394094 at 7 epoch\n",
      "The intersecting regions size is : (2061, 5)\n",
      "After removal of intersecting region the remaining SNVs are  392033\n",
      "#############################################################\n",
      "The number of SNV counts 392033 at 8 epoch\n",
      "The intersecting regions size is : (2109, 5)\n",
      "After removal of intersecting region the remaining SNVs are  389924\n",
      "#############################################################\n",
      "The number of SNV counts 389924 at 9 epoch\n",
      "The intersecting regions size is : (2055, 5)\n",
      "After removal of intersecting region the remaining SNVs are  387869\n",
      "#############################################################\n",
      "The number of SNV counts 387869 at 10 epoch\n",
      "The intersecting regions size is : (2055, 5)\n",
      "After removal of intersecting region the remaining SNVs are  385814\n",
      "#############################################################\n",
      "The number of SNV counts 385814 at 11 epoch\n",
      "The intersecting regions size is : (2134, 5)\n",
      "After removal of intersecting region the remaining SNVs are  383680\n",
      "#############################################################\n",
      "The number of SNV counts 383680 at 12 epoch\n",
      "The intersecting regions size is : (2013, 5)\n",
      "After removal of intersecting region the remaining SNVs are  381667\n",
      "#############################################################\n",
      "The number of SNV counts 381667 at 13 epoch\n",
      "The intersecting regions size is : (2065, 5)\n",
      "After removal of intersecting region the remaining SNVs are  379602\n",
      "#############################################################\n",
      "The number of SNV counts 379602 at 14 epoch\n",
      "The intersecting regions size is : (2040, 5)\n",
      "After removal of intersecting region the remaining SNVs are  377562\n",
      "#############################################################\n",
      "The number of SNV counts 377562 at 15 epoch\n",
      "The intersecting regions size is : (1948, 5)\n",
      "After removal of intersecting region the remaining SNVs are  375614\n",
      "#############################################################\n",
      "The number of SNV counts 375614 at 16 epoch\n",
      "The intersecting regions size is : (1948, 5)\n",
      "After removal of intersecting region the remaining SNVs are  373666\n",
      "#############################################################\n",
      "The number of SNV counts 373666 at 17 epoch\n",
      "The intersecting regions size is : (2004, 5)\n",
      "After removal of intersecting region the remaining SNVs are  371662\n",
      "#############################################################\n",
      "The number of SNV counts 371662 at 18 epoch\n",
      "The intersecting regions size is : (1942, 5)\n",
      "After removal of intersecting region the remaining SNVs are  369720\n",
      "#############################################################\n",
      "The number of SNV counts 369720 at 19 epoch\n",
      "The intersecting regions size is : (1859, 5)\n",
      "After removal of intersecting region the remaining SNVs are  367861\n",
      "#############################################################\n",
      "The number of SNV counts 367861 at 20 epoch\n",
      "The intersecting regions size is : (1874, 5)\n",
      "After removal of intersecting region the remaining SNVs are  365987\n",
      "#############################################################\n",
      "The number of SNV counts 365987 at 21 epoch\n",
      "The intersecting regions size is : (1884, 5)\n",
      "After removal of intersecting region the remaining SNVs are  364103\n",
      "#############################################################\n",
      "The number of SNV counts 364103 at 22 epoch\n",
      "The intersecting regions size is : (1875, 5)\n",
      "After removal of intersecting region the remaining SNVs are  362228\n",
      "#############################################################\n",
      "The number of SNV counts 362228 at 23 epoch\n",
      "The intersecting regions size is : (1779, 5)\n",
      "After removal of intersecting region the remaining SNVs are  360449\n",
      "#############################################################\n",
      "The number of SNV counts 360449 at 24 epoch\n",
      "The intersecting regions size is : (1818, 5)\n",
      "After removal of intersecting region the remaining SNVs are  358631\n",
      "#############################################################\n",
      "The number of SNV counts 358631 at 25 epoch\n",
      "The intersecting regions size is : (1769, 5)\n",
      "After removal of intersecting region the remaining SNVs are  356862\n",
      "#############################################################\n",
      "The number of SNV counts 356862 at 26 epoch\n",
      "The intersecting regions size is : (1764, 5)\n",
      "After removal of intersecting region the remaining SNVs are  355098\n",
      "#############################################################\n",
      "The number of SNV counts 355098 at 27 epoch\n",
      "The intersecting regions size is : (1780, 5)\n",
      "After removal of intersecting region the remaining SNVs are  353318\n",
      "#############################################################\n",
      "The number of SNV counts 353318 at 28 epoch\n",
      "The intersecting regions size is : (1788, 5)\n",
      "After removal of intersecting region the remaining SNVs are  351530\n",
      "#############################################################\n",
      "The number of SNV counts 351530 at 29 epoch\n",
      "The intersecting regions size is : (1865, 5)\n",
      "After removal of intersecting region the remaining SNVs are  349665\n",
      "#############################################################\n",
      "The number of SNV counts 349665 at 30 epoch\n",
      "The intersecting regions size is : (1773, 5)\n",
      "After removal of intersecting region the remaining SNVs are  347892\n",
      "#############################################################\n",
      "The number of SNV counts 347892 at 31 epoch\n",
      "The intersecting regions size is : (1765, 5)\n",
      "After removal of intersecting region the remaining SNVs are  346127\n",
      "#############################################################\n",
      "The number of SNV counts 346127 at 32 epoch\n",
      "The intersecting regions size is : (1787, 5)\n",
      "After removal of intersecting region the remaining SNVs are  344340\n",
      "#############################################################\n",
      "The number of SNV counts 344340 at 33 epoch\n",
      "The intersecting regions size is : (1815, 5)\n",
      "After removal of intersecting region the remaining SNVs are  342525\n",
      "#############################################################\n",
      "The number of SNV counts 342525 at 34 epoch\n",
      "The intersecting regions size is : (1768, 5)\n",
      "After removal of intersecting region the remaining SNVs are  340757\n",
      "#############################################################\n",
      "The number of SNV counts 340757 at 35 epoch\n",
      "The intersecting regions size is : (1741, 5)\n",
      "After removal of intersecting region the remaining SNVs are  339016\n",
      "#############################################################\n",
      "The number of SNV counts 339016 at 36 epoch\n",
      "The intersecting regions size is : (1756, 5)\n",
      "After removal of intersecting region the remaining SNVs are  337260\n",
      "#############################################################\n",
      "The number of SNV counts 337260 at 37 epoch\n",
      "The intersecting regions size is : (1652, 5)\n",
      "After removal of intersecting region the remaining SNVs are  335608\n",
      "#############################################################\n",
      "The number of SNV counts 335608 at 38 epoch\n",
      "The intersecting regions size is : (1678, 5)\n",
      "After removal of intersecting region the remaining SNVs are  333930\n",
      "#############################################################\n",
      "The number of SNV counts 333930 at 39 epoch\n",
      "The intersecting regions size is : (1663, 5)\n",
      "After removal of intersecting region the remaining SNVs are  332267\n",
      "#############################################################\n",
      "The number of SNV counts 332267 at 40 epoch\n",
      "The intersecting regions size is : (1634, 5)\n",
      "After removal of intersecting region the remaining SNVs are  330633\n",
      "#############################################################\n",
      "The number of SNV counts 330633 at 41 epoch\n",
      "The intersecting regions size is : (1656, 5)\n",
      "After removal of intersecting region the remaining SNVs are  328977\n",
      "#############################################################\n",
      "The number of SNV counts 328977 at 42 epoch\n",
      "The intersecting regions size is : (1571, 5)\n",
      "After removal of intersecting region the remaining SNVs are  327406\n",
      "#############################################################\n",
      "The number of SNV counts 327406 at 43 epoch\n",
      "The intersecting regions size is : (1528, 5)\n",
      "After removal of intersecting region the remaining SNVs are  325878\n",
      "#############################################################\n",
      "The number of SNV counts 325878 at 44 epoch\n",
      "The intersecting regions size is : (1591, 5)\n",
      "After removal of intersecting region the remaining SNVs are  324287\n",
      "#############################################################\n",
      "The number of SNV counts 324287 at 45 epoch\n",
      "The intersecting regions size is : (1627, 5)\n",
      "After removal of intersecting region the remaining SNVs are  322660\n",
      "#############################################################\n",
      "The number of SNV counts 322660 at 46 epoch\n",
      "The intersecting regions size is : (1652, 5)\n",
      "After removal of intersecting region the remaining SNVs are  321008\n",
      "#############################################################\n",
      "The number of SNV counts 321008 at 47 epoch\n",
      "The intersecting regions size is : (1607, 5)\n",
      "After removal of intersecting region the remaining SNVs are  319401\n",
      "#############################################################\n",
      "The number of SNV counts 319401 at 48 epoch\n",
      "The intersecting regions size is : (1529, 5)\n",
      "After removal of intersecting region the remaining SNVs are  317872\n",
      "#############################################################\n",
      "The number of SNV counts 317872 at 49 epoch\n",
      "The intersecting regions size is : (1568, 5)\n",
      "After removal of intersecting region the remaining SNVs are  316304\n",
      "#############################################################\n",
      "The number of SNV counts 316304 at 50 epoch\n",
      "The intersecting regions size is : (1525, 5)\n",
      "After removal of intersecting region the remaining SNVs are  314779\n",
      "#############################################################\n"
     ]
    }
   ],
   "source": [
    "# Create a list of chromosomes to keep\n",
    "chromosomes_to_keep = ['chr' + str(i) for i in range(1, 23)] + ['chrX', 'chrY']\n",
    "\n",
    "# Filter the original DataFrame to only include specified chromosomes\n",
    "df_noUV_plus_interested = df_noUV_plus[df_noUV_plus[0].isin(chromosomes_to_keep)].reset_index(drop=True)\n",
    "print(df_noUV_plus_interested.shape, df_interest_bed.shape)\n",
    "input()\n",
    "# Create a BedTool object from another DataFrame\n",
    "bd_interest_SNV_bed = pybedtools.BedTool.from_dataframe(df_interest_bed)\n",
    "\n",
    "# Initialize an empty DataFrame for temporary use\n",
    "df_temp = pd.DataFrame()\n",
    "df_temp['chr_name'] = df_noUV_plus_interested[0]\n",
    "\n",
    "# Convert df_noUV_plus_interested to a BedTool object initially\n",
    "df_noUV_plus_bed = pybedtools.BedTool.from_dataframe(df_noUV_plus_interested)\n",
    "\n",
    "for i in range(51):\n",
    "    # Populate df_temp with updated start and end positions\n",
    "    df_temp['chr_start'] = df_noUV_plus_interested[1] - i\n",
    "    df_temp['chr_end'] = df_noUV_plus_interested[2] + i\n",
    "    print(\"The number of SNV counts {} at {} epoch\".format(bd_interest_SNV_bed.count(), i))\n",
    "    # Convert df_temp to a BedTool object\n",
    "    bd_noUV_plus_bed = pybedtools.BedTool.from_dataframe(df_temp)\n",
    "    # Perform the intersection\n",
    "    intersect_noCPD_SNV = bd_noUV_plus_bed.intersect(bd_interest_SNV_bed, wa=True, wb=True)\n",
    "    # Convert the result back to a DataFrame and set column names\n",
    "    df_intersect_noCPD_SNV = intersect_noCPD_SNV.to_dataframe()\n",
    "    df_intersect_noCPD_SNV.columns = ['chr_name', 'CPD_start', 'CPD_end', 'chr_name_SNV', 'SNV_start', 'SNV_end', 'ID','SNV_Info']\n",
    "    #print(df_intersect_noCPD_SNV[['chr_name_SNV', 'SNV_start', 'SNV_end', 'ID','SNV_Info']].drop_duplicates())\n",
    "    \n",
    "    # Save the result to a file\n",
    "    df_intersect_noCPD_SNV.to_csv(f\"/data/private/pdutta/Collab_data/Jim_work/Output/No_UV_Plus/intersected_CPD_SNV_noUV_plus_{i}.tsv\", sep='\\t', index=False)\n",
    "    \n",
    "    # Identify intersecting regions for removal in the next iteration\n",
    "    intersecting_SNV_regions = bd_interest_SNV_bed.intersect(bd_noUV_plus_bed, wa=True, u=True)\n",
    "    print(\"The intersecting regions size is :\",intersecting_SNV_regions.to_dataframe().shape)\n",
    "    #print(bd_interest_SNV_bed.to_dataframe())\n",
    "    \n",
    "    # Remove intersecting regions for the next iteration\n",
    "    bd_interest_SNV_bed = bd_interest_SNV_bed.subtract(intersecting_SNV_regions)\n",
    "    #print(bd_interest_SNV_bed.to_dataframe())\n",
    "    print(\"After removal of intersecting region the remaining SNVs are \", bd_interest_SNV_bed.count())\n",
    "    #df_noUV_plus_interested.columns = ['chr_name', 'CPD_start', 'CPD_end']\n",
    "    \n",
    "    print(\"#############################################################\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "8727afb7-b89b-4c39-9abc-b224459e8ee5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14424047, 5) (421496, 3)\n",
      "         chr_name  chr_start   chr_end\n",
      "0            chr1    3050021   3050022\n",
      "1            chr1    3050022   3050023\n",
      "2            chr1    3050413   3050414\n",
      "3            chr1    3050414   3050415\n",
      "4            chr1    3050720   3050721\n",
      "...           ...        ...       ...\n",
      "14424042     chrY   90855678  90855679\n",
      "14424043     chrY   90855821  90855822\n",
      "14424044     chrY   90855822  90855823\n",
      "14424045     chrY   90855830  90855831\n",
      "14424046     chrY   90855831  90855832\n",
      "\n",
      "[14424047 rows x 3 columns]\n",
      "     chr_name  CPD_start    CPD_end chr_name_SNV  SNV_start    SNV_end\n",
      "0        chr1    4306970    4306971         chr1    4306970    4306971\n",
      "1        chr1    5857620    5857621         chr1    5857620    5857621\n",
      "2        chr1    7532894    7532895         chr1    7532894    7532895\n",
      "3        chr1   10057967   10057968         chr1   10057967   10057968\n",
      "4        chr1   11061577   11061578         chr1   11061577   11061578\n",
      "...       ...        ...        ...          ...        ...        ...\n",
      "4195     chrX  160068488  160068489         chrX  160068488  160068489\n",
      "4196     chrX  162290505  162290506         chrX  162290505  162290506\n",
      "4197     chrX  162903275  162903276         chrX  162903275  162903276\n",
      "4198     chrX  164756177  164756178         chrX  164756177  164756178\n",
      "4199     chrX  165686991  165686992         chrX  165686991  165686992\n",
      "\n",
      "[4200 rows x 6 columns]\n",
      "#############################################################\n",
      "         chr_name  chr_start   chr_end\n",
      "0            chr1    3050020   3050023\n",
      "1            chr1    3050021   3050024\n",
      "2            chr1    3050412   3050415\n",
      "3            chr1    3050413   3050416\n",
      "4            chr1    3050719   3050722\n",
      "...           ...        ...       ...\n",
      "14424042     chrY   90855677  90855680\n",
      "14424043     chrY   90855820  90855823\n",
      "14424044     chrY   90855821  90855824\n",
      "14424045     chrY   90855829  90855832\n",
      "14424046     chrY   90855830  90855833\n",
      "\n",
      "[14424047 rows x 3 columns]\n",
      "      chr_name  CPD_start    CPD_end chr_name_SNV  SNV_start    SNV_end\n",
      "0         chr1    4306968    4306971         chr1    4306970    4306971\n",
      "1         chr1    4306969    4306972         chr1    4306970    4306971\n",
      "2         chr1    5477064    5477067         chr1    5477064    5477065\n",
      "3         chr1    5857618    5857621         chr1    5857620    5857621\n",
      "4         chr1    5857619    5857622         chr1    5857620    5857621\n",
      "...        ...        ...        ...          ...        ...        ...\n",
      "11295     chrX  164756176  164756179         chrX  164756177  164756178\n",
      "11296     chrX  165686989  165686992         chrX  165686991  165686992\n",
      "11297     chrX  165686990  165686993         chrX  165686991  165686992\n",
      "11298     chrX  167698694  167698697         chrX  167698694  167698695\n",
      "11299     chrX  168752437  168752440         chrX  168752439  168752440\n",
      "\n",
      "[11300 rows x 6 columns]\n",
      "#############################################################\n",
      "         chr_name  chr_start   chr_end\n",
      "0            chr1    3050019   3050024\n",
      "1            chr1    3050020   3050025\n",
      "2            chr1    3050411   3050416\n",
      "3            chr1    3050412   3050417\n",
      "4            chr1    3050718   3050723\n",
      "...           ...        ...       ...\n",
      "14424042     chrY   90855676  90855681\n",
      "14424043     chrY   90855819  90855824\n",
      "14424044     chrY   90855820  90855825\n",
      "14424045     chrY   90855828  90855833\n",
      "14424046     chrY   90855829  90855834\n",
      "\n",
      "[14424047 rows x 3 columns]\n",
      "      chr_name  CPD_start    CPD_end chr_name_SNV  SNV_start    SNV_end\n",
      "0         chr1    3363328    3363333         chr1    3363332    3363333\n",
      "1         chr1    4306967    4306972         chr1    4306970    4306971\n",
      "2         chr1    4306968    4306973         chr1    4306970    4306971\n",
      "3         chr1    5477063    5477068         chr1    5477064    5477065\n",
      "4         chr1    5477064    5477069         chr1    5477064    5477065\n",
      "...        ...        ...        ...          ...        ...        ...\n",
      "16934     chrX  167698694  167698699         chrX  167698694  167698695\n",
      "16935     chrX  168445088  168445093         chrX  168445092  168445093\n",
      "16936     chrX  168447265  168447270         chrX  168447269  168447270\n",
      "16937     chrX  168752435  168752440         chrX  168752439  168752440\n",
      "16938     chrX  168752436  168752441         chrX  168752439  168752440\n",
      "\n",
      "[16939 rows x 6 columns]\n",
      "#############################################################\n",
      "         chr_name  chr_start   chr_end\n",
      "0            chr1    3050018   3050025\n",
      "1            chr1    3050019   3050026\n",
      "2            chr1    3050410   3050417\n",
      "3            chr1    3050411   3050418\n",
      "4            chr1    3050717   3050724\n",
      "...           ...        ...       ...\n",
      "14424042     chrY   90855675  90855682\n",
      "14424043     chrY   90855818  90855825\n",
      "14424044     chrY   90855819  90855826\n",
      "14424045     chrY   90855827  90855834\n",
      "14424046     chrY   90855828  90855835\n",
      "\n",
      "[14424047 rows x 3 columns]\n",
      "      chr_name  CPD_start    CPD_end chr_name_SNV  SNV_start    SNV_end\n",
      "0         chr1    3059753    3059760         chr1    3059753    3059754\n",
      "1         chr1    3363326    3363333         chr1    3363332    3363333\n",
      "2         chr1    3363327    3363334         chr1    3363332    3363333\n",
      "3         chr1    3891231    3891238         chr1    3891237    3891238\n",
      "4         chr1    3976125    3976132         chr1    3976125    3976126\n",
      "...        ...        ...        ...          ...        ...        ...\n",
      "22058     chrX  168445087  168445094         chrX  168445092  168445093\n",
      "22059     chrX  168447263  168447270         chrX  168447269  168447270\n",
      "22060     chrX  168447264  168447271         chrX  168447269  168447270\n",
      "22061     chrX  168752434  168752441         chrX  168752439  168752440\n",
      "22062     chrX  168752435  168752442         chrX  168752439  168752440\n",
      "\n",
      "[22063 rows x 6 columns]\n",
      "#############################################################\n",
      "         chr_name  chr_start   chr_end\n",
      "0            chr1    3050017   3050026\n",
      "1            chr1    3050018   3050027\n",
      "2            chr1    3050409   3050418\n",
      "3            chr1    3050410   3050419\n",
      "4            chr1    3050716   3050725\n",
      "...           ...        ...       ...\n",
      "14424042     chrY   90855674  90855683\n",
      "14424043     chrY   90855817  90855826\n",
      "14424044     chrY   90855818  90855827\n",
      "14424045     chrY   90855826  90855835\n",
      "14424046     chrY   90855827  90855836\n",
      "\n",
      "[14424047 rows x 3 columns]\n",
      "      chr_name  CPD_start    CPD_end chr_name_SNV  SNV_start    SNV_end\n",
      "0         chr1    3059752    3059761         chr1    3059753    3059754\n",
      "1         chr1    3059753    3059762         chr1    3059753    3059754\n",
      "2         chr1    3363325    3363334         chr1    3363332    3363333\n",
      "3         chr1    3363326    3363335         chr1    3363332    3363333\n",
      "4         chr1    3651240    3651249         chr1    3651240    3651241\n",
      "...        ...        ...        ...          ...        ...        ...\n",
      "26650     chrX  168447262  168447271         chrX  168447269  168447270\n",
      "26651     chrX  168447263  168447272         chrX  168447269  168447270\n",
      "26652     chrX  168545490  168545499         chrX  168545498  168545499\n",
      "26653     chrX  168752433  168752442         chrX  168752439  168752440\n",
      "26654     chrX  168752434  168752443         chrX  168752439  168752440\n",
      "\n",
      "[26655 rows x 6 columns]\n",
      "#############################################################\n",
      "         chr_name  chr_start   chr_end\n",
      "0            chr1    3050016   3050027\n",
      "1            chr1    3050017   3050028\n",
      "2            chr1    3050408   3050419\n",
      "3            chr1    3050409   3050420\n",
      "4            chr1    3050715   3050726\n",
      "...           ...        ...       ...\n",
      "14424042     chrY   90855673  90855684\n",
      "14424043     chrY   90855816  90855827\n",
      "14424044     chrY   90855817  90855828\n",
      "14424045     chrY   90855825  90855836\n",
      "14424046     chrY   90855826  90855837\n",
      "\n",
      "[14424047 rows x 3 columns]\n",
      "      chr_name  CPD_start    CPD_end chr_name_SNV  SNV_start    SNV_end\n",
      "0         chr1    3052410    3052421         chr1    3052410    3052411\n",
      "1         chr1    3059751    3059762         chr1    3059753    3059754\n",
      "2         chr1    3059752    3059763         chr1    3059753    3059754\n",
      "3         chr1    3193475    3193486         chr1    3193475    3193476\n",
      "4         chr1    3363324    3363335         chr1    3363332    3363333\n",
      "...        ...        ...        ...          ...        ...        ...\n",
      "31299     chrX  168447262  168447273         chrX  168447269  168447270\n",
      "31300     chrX  168545488  168545499         chrX  168545498  168545499\n",
      "31301     chrX  168545489  168545500         chrX  168545498  168545499\n",
      "31302     chrX  168752432  168752443         chrX  168752439  168752440\n",
      "31303     chrX  168752433  168752444         chrX  168752439  168752440\n",
      "\n",
      "[31304 rows x 6 columns]\n",
      "#############################################################\n",
      "         chr_name  chr_start   chr_end\n",
      "0            chr1    3050015   3050028\n",
      "1            chr1    3050016   3050029\n",
      "2            chr1    3050407   3050420\n",
      "3            chr1    3050408   3050421\n",
      "4            chr1    3050714   3050727\n",
      "...           ...        ...       ...\n",
      "14424042     chrY   90855672  90855685\n",
      "14424043     chrY   90855815  90855828\n",
      "14424044     chrY   90855816  90855829\n",
      "14424045     chrY   90855824  90855837\n",
      "14424046     chrY   90855825  90855838\n",
      "\n",
      "[14424047 rows x 3 columns]\n",
      "      chr_name  CPD_start    CPD_end chr_name_SNV  SNV_start    SNV_end\n",
      "0         chr1    3052409    3052422         chr1    3052410    3052411\n",
      "1         chr1    3052410    3052423         chr1    3052410    3052411\n",
      "2         chr1    3059750    3059763         chr1    3059753    3059754\n",
      "3         chr1    3059751    3059764         chr1    3059753    3059754\n",
      "4         chr1    3193474    3193487         chr1    3193475    3193476\n",
      "...        ...        ...        ...          ...        ...        ...\n",
      "36101     chrX  168447261  168447274         chrX  168447269  168447270\n",
      "36102     chrX  168545487  168545500         chrX  168545498  168545499\n",
      "36103     chrX  168545488  168545501         chrX  168545498  168545499\n",
      "36104     chrX  168752431  168752444         chrX  168752439  168752440\n",
      "36105     chrX  168752432  168752445         chrX  168752439  168752440\n",
      "\n",
      "[36106 rows x 6 columns]\n",
      "#############################################################\n",
      "         chr_name  chr_start   chr_end\n",
      "0            chr1    3050014   3050029\n",
      "1            chr1    3050015   3050030\n",
      "2            chr1    3050406   3050421\n",
      "3            chr1    3050407   3050422\n",
      "4            chr1    3050713   3050728\n",
      "...           ...        ...       ...\n",
      "14424042     chrY   90855671  90855686\n",
      "14424043     chrY   90855814  90855829\n",
      "14424044     chrY   90855815  90855830\n",
      "14424045     chrY   90855823  90855838\n",
      "14424046     chrY   90855824  90855839\n",
      "\n",
      "[14424047 rows x 3 columns]\n",
      "      chr_name  CPD_start    CPD_end chr_name_SNV  SNV_start    SNV_end\n",
      "0         chr1    3052408    3052423         chr1    3052410    3052411\n",
      "1         chr1    3052409    3052424         chr1    3052410    3052411\n",
      "2         chr1    3059749    3059764         chr1    3059753    3059754\n",
      "3         chr1    3059750    3059765         chr1    3059753    3059754\n",
      "4         chr1    3193473    3193488         chr1    3193475    3193476\n",
      "...        ...        ...        ...          ...        ...        ...\n",
      "40725     chrX  168447260  168447275         chrX  168447269  168447270\n",
      "40726     chrX  168545486  168545501         chrX  168545498  168545499\n",
      "40727     chrX  168545487  168545502         chrX  168545498  168545499\n",
      "40728     chrX  168752430  168752445         chrX  168752439  168752440\n",
      "40729     chrX  168752431  168752446         chrX  168752439  168752440\n",
      "\n",
      "[40730 rows x 6 columns]\n",
      "#############################################################\n",
      "         chr_name  chr_start   chr_end\n",
      "0            chr1    3050013   3050030\n",
      "1            chr1    3050014   3050031\n",
      "2            chr1    3050405   3050422\n",
      "3            chr1    3050406   3050423\n",
      "4            chr1    3050712   3050729\n",
      "...           ...        ...       ...\n",
      "14424042     chrY   90855670  90855687\n",
      "14424043     chrY   90855813  90855830\n",
      "14424044     chrY   90855814  90855831\n",
      "14424045     chrY   90855822  90855839\n",
      "14424046     chrY   90855823  90855840\n",
      "\n",
      "[14424047 rows x 3 columns]\n",
      "      chr_name  CPD_start    CPD_end chr_name_SNV  SNV_start    SNV_end\n",
      "0         chr1    3052407    3052424         chr1    3052410    3052411\n",
      "1         chr1    3052408    3052425         chr1    3052410    3052411\n",
      "2         chr1    3059748    3059765         chr1    3059753    3059754\n",
      "3         chr1    3059749    3059766         chr1    3059753    3059754\n",
      "4         chr1    3193472    3193489         chr1    3193475    3193476\n",
      "...        ...        ...        ...          ...        ...        ...\n",
      "45370     chrX  168447259  168447276         chrX  168447269  168447270\n",
      "45371     chrX  168545485  168545502         chrX  168545498  168545499\n",
      "45372     chrX  168545486  168545503         chrX  168545498  168545499\n",
      "45373     chrX  168752429  168752446         chrX  168752439  168752440\n",
      "45374     chrX  168752430  168752447         chrX  168752439  168752440\n",
      "\n",
      "[45375 rows x 6 columns]\n",
      "#############################################################\n",
      "         chr_name  chr_start   chr_end\n",
      "0            chr1    3050012   3050031\n",
      "1            chr1    3050013   3050032\n",
      "2            chr1    3050404   3050423\n",
      "3            chr1    3050405   3050424\n",
      "4            chr1    3050711   3050730\n",
      "...           ...        ...       ...\n",
      "14424042     chrY   90855669  90855688\n",
      "14424043     chrY   90855812  90855831\n",
      "14424044     chrY   90855813  90855832\n",
      "14424045     chrY   90855821  90855840\n",
      "14424046     chrY   90855822  90855841\n",
      "\n",
      "[14424047 rows x 3 columns]\n",
      "      chr_name  CPD_start    CPD_end chr_name_SNV  SNV_start    SNV_end\n",
      "0         chr1    3052406    3052425         chr1    3052410    3052411\n",
      "1         chr1    3052407    3052426         chr1    3052410    3052411\n",
      "2         chr1    3059747    3059766         chr1    3059753    3059754\n",
      "3         chr1    3059748    3059767         chr1    3059753    3059754\n",
      "4         chr1    3193471    3193490         chr1    3193475    3193476\n",
      "...        ...        ...        ...          ...        ...        ...\n",
      "50083     chrX  168447258  168447277         chrX  168447269  168447270\n",
      "50084     chrX  168545484  168545503         chrX  168545498  168545499\n",
      "50085     chrX  168545485  168545504         chrX  168545498  168545499\n",
      "50086     chrX  168752428  168752447         chrX  168752439  168752440\n",
      "50087     chrX  168752429  168752448         chrX  168752439  168752440\n",
      "\n",
      "[50088 rows x 6 columns]\n",
      "#############################################################\n",
      "         chr_name  chr_start   chr_end\n",
      "0            chr1    3050011   3050032\n",
      "1            chr1    3050012   3050033\n",
      "2            chr1    3050403   3050424\n",
      "3            chr1    3050404   3050425\n",
      "4            chr1    3050710   3050731\n",
      "...           ...        ...       ...\n",
      "14424042     chrY   90855668  90855689\n",
      "14424043     chrY   90855811  90855832\n",
      "14424044     chrY   90855812  90855833\n",
      "14424045     chrY   90855820  90855841\n",
      "14424046     chrY   90855821  90855842\n",
      "\n",
      "[14424047 rows x 3 columns]\n",
      "      chr_name  CPD_start    CPD_end chr_name_SNV  SNV_start    SNV_end\n",
      "0         chr1    3052405    3052426         chr1    3052410    3052411\n",
      "1         chr1    3052406    3052427         chr1    3052410    3052411\n",
      "2         chr1    3059746    3059767         chr1    3059753    3059754\n",
      "3         chr1    3059747    3059768         chr1    3059753    3059754\n",
      "4         chr1    3193470    3193491         chr1    3193475    3193476\n",
      "...        ...        ...        ...          ...        ...        ...\n",
      "54776     chrX  168447257  168447278         chrX  168447269  168447270\n",
      "54777     chrX  168545483  168545504         chrX  168545498  168545499\n",
      "54778     chrX  168545484  168545505         chrX  168545498  168545499\n",
      "54779     chrX  168752427  168752448         chrX  168752439  168752440\n",
      "54780     chrX  168752428  168752449         chrX  168752439  168752440\n",
      "\n",
      "[54781 rows x 6 columns]\n",
      "#############################################################\n"
     ]
    }
   ],
   "source": [
    "chromosomes_to_keep = ['chr' + str(i) for i in range(1, 23)] + ['chrX', 'chrY']\n",
    "df_noUV_plus_interested = df_noUV_plus[df_noUV_plus[0].isin(chromosomes_to_keep)].reset_index(drop=True)\n",
    "df_interest_SNV_bed = pybedtools.BedTool.from_dataframe(df_interest_bed)\n",
    "print(df_noUV_plus_interested.shape, df_interest_bed.shape)\n",
    "df_temp=pd.DataFrame()\n",
    "df_temp['chr_name'] = df_noUV_plus_interested[0]\n",
    "for i in range(11):\n",
    "    df_temp['chr_start'] = df_noUV_plus_interested[1]- i\n",
    "    df_temp['chr_end'] = df_noUV_plus_interested[2]+ i\n",
    "    print(df_temp)\n",
    "    df_UV_plus_bed = pybedtools.BedTool.from_dataframe(df_temp)\n",
    "    intersect_SNV_CPD = df_UV_plus_bed.intersect(df_interest_SNV_bed, wa=True, wb=True)\n",
    "    intersect_SNV_CPD_df= intersect_SNV_CPD.to_dataframe()\n",
    "    intersect_SNV_CPD_df.columns = ['chr_name', 'CPD_start', 'CPD_end', 'chr_name_SNV', 'SNV_start', 'SNV_end']\n",
    "    print(intersect_SNV_CPD_df)\n",
    "    intersect_SNV_CPD_df.to_csv(\"/data/private/pdutta/Collab_data/Jim_work/Output/No_UV_Plus/intersected_CPD_SNV_noUV_plus_\"+str(i)+\".tsv\", sep=\"\\t\", index=False)\n",
    "    print(\"#############################################################\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "dfe95c64-bd5b-4ddd-9b6f-d92269b3df06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050021</td>\n",
       "      <td>3050022</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050022</td>\n",
       "      <td>3050023</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050413</td>\n",
       "      <td>3050414</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050414</td>\n",
       "      <td>3050415</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050721</td>\n",
       "      <td>3050722</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050756</td>\n",
       "      <td>3050757</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050757</td>\n",
       "      <td>3050758</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050807</td>\n",
       "      <td>3050808</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050808</td>\n",
       "      <td>3050809</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050842</td>\n",
       "      <td>3050843</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050843</td>\n",
       "      <td>3050844</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050880</td>\n",
       "      <td>3050881</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3051228</td>\n",
       "      <td>3051229</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3051272</td>\n",
       "      <td>3051273</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3051273</td>\n",
       "      <td>3051274</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3051321</td>\n",
       "      <td>3051322</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3051322</td>\n",
       "      <td>3051323</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3051346</td>\n",
       "      <td>3051347</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3051347</td>\n",
       "      <td>3051348</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3051525</td>\n",
       "      <td>3051526</td>\n",
       "      <td>+</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       0        1        2  3   4\n",
       "0   chr1  3050021  3050022  +  GA\n",
       "1   chr1  3050022  3050023  +  GA\n",
       "2   chr1  3050413  3050414  +  AG\n",
       "3   chr1  3050414  3050415  +  AG\n",
       "4   chr1  3050721  3050722  +  GG\n",
       "5   chr1  3050756  3050757  +  GA\n",
       "6   chr1  3050757  3050758  +  GA\n",
       "7   chr1  3050807  3050808  +  GA\n",
       "8   chr1  3050808  3050809  +  GA\n",
       "9   chr1  3050842  3050843  +  AA\n",
       "10  chr1  3050843  3050844  +  AA\n",
       "11  chr1  3050880  3050881  +  GG\n",
       "12  chr1  3051228  3051229  +  AA\n",
       "13  chr1  3051272  3051273  +  AA\n",
       "14  chr1  3051273  3051274  +  AA\n",
       "15  chr1  3051321  3051322  +  AG\n",
       "16  chr1  3051322  3051323  +  AG\n",
       "17  chr1  3051346  3051347  +  GA\n",
       "18  chr1  3051347  3051348  +  GA\n",
       "19  chr1  3051525  3051526  +  AA"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_filtered_noUV_plus_only.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d418c091-023d-4556-ad84-139ada22b167",
   "metadata": {},
   "source": [
    "## Finding Intersection using `PyBedtools`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "c43bcd06-8b12-454f-bce6-0613832c6950",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>thickStart</th>\n",
       "      <th>thickEnd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr15</td>\n",
       "      <td>3895294</td>\n",
       "      <td>3895295</td>\n",
       "      <td>chr15</td>\n",
       "      <td>3895294</td>\n",
       "      <td>3895295</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr15</td>\n",
       "      <td>5579415</td>\n",
       "      <td>5579416</td>\n",
       "      <td>chr15</td>\n",
       "      <td>5579415</td>\n",
       "      <td>5579416</td>\n",
       "      <td>+</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr15</td>\n",
       "      <td>6191775</td>\n",
       "      <td>6191776</td>\n",
       "      <td>chr15</td>\n",
       "      <td>6191775</td>\n",
       "      <td>6191776</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr15</td>\n",
       "      <td>6207261</td>\n",
       "      <td>6207262</td>\n",
       "      <td>chr15</td>\n",
       "      <td>6207261</td>\n",
       "      <td>6207262</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr15</td>\n",
       "      <td>9691374</td>\n",
       "      <td>9691375</td>\n",
       "      <td>chr15</td>\n",
       "      <td>9691374</td>\n",
       "      <td>9691375</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3501</th>\n",
       "      <td>chrX</td>\n",
       "      <td>147720411</td>\n",
       "      <td>147720412</td>\n",
       "      <td>chrX</td>\n",
       "      <td>147720411</td>\n",
       "      <td>147720412</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3502</th>\n",
       "      <td>chrX</td>\n",
       "      <td>39568936</td>\n",
       "      <td>39568937</td>\n",
       "      <td>chrX</td>\n",
       "      <td>39568936</td>\n",
       "      <td>39568937</td>\n",
       "      <td>+</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3503</th>\n",
       "      <td>chrX</td>\n",
       "      <td>88746130</td>\n",
       "      <td>88746131</td>\n",
       "      <td>chrX</td>\n",
       "      <td>88746130</td>\n",
       "      <td>88746131</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3504</th>\n",
       "      <td>chrX</td>\n",
       "      <td>104886995</td>\n",
       "      <td>104886996</td>\n",
       "      <td>chrX</td>\n",
       "      <td>104886995</td>\n",
       "      <td>104886996</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3505</th>\n",
       "      <td>chrX</td>\n",
       "      <td>158712030</td>\n",
       "      <td>158712031</td>\n",
       "      <td>chrX</td>\n",
       "      <td>158712030</td>\n",
       "      <td>158712031</td>\n",
       "      <td>+</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3506 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom      start        end   name      score     strand thickStart  \\\n",
       "0     chr15    3895294    3895295  chr15    3895294    3895295          +   \n",
       "1     chr15    5579415    5579416  chr15    5579415    5579416          +   \n",
       "2     chr15    6191775    6191776  chr15    6191775    6191776          +   \n",
       "3     chr15    6207261    6207262  chr15    6207261    6207262          +   \n",
       "4     chr15    9691374    9691375  chr15    9691374    9691375          +   \n",
       "...     ...        ...        ...    ...        ...        ...        ...   \n",
       "3501   chrX  147720411  147720412   chrX  147720411  147720412          +   \n",
       "3502   chrX   39568936   39568937   chrX   39568936   39568937          +   \n",
       "3503   chrX   88746130   88746131   chrX   88746130   88746131          +   \n",
       "3504   chrX  104886995  104886996   chrX  104886995  104886996          +   \n",
       "3505   chrX  158712030  158712031   chrX  158712030  158712031          +   \n",
       "\n",
       "     thickEnd  \n",
       "0          GG  \n",
       "1          AG  \n",
       "2          GA  \n",
       "3          GA  \n",
       "4          GG  \n",
       "...       ...  \n",
       "3501       GG  \n",
       "3502       GA  \n",
       "3503       GG  \n",
       "3504       GG  \n",
       "3505       GG  \n",
       "\n",
       "[3506 rows x 8 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_interest_SNV_bed = pybedtools.BedTool.from_dataframe(df_interest_bed)\n",
    "df_noUV_plus_only_bed = pybedtools.BedTool.from_dataframe(df_filtered_noUV_plus_only)\n",
    "intersect_SNV_CPD = df_interest_SNV_bed.intersect(df_noUV_plus_only_bed, wa=True, wb=True)\n",
    "intersect_SNV_CPD.to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "124b3b46-f78d-4816-9257-a68131e70ac0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
